// @remotion/studio-server/dist/file-watcher.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.writeFileAndNotifyFileWatchers = exports.installFileWatcher = exports.getFileWatcherRegistry = exports.setFileWatcherRegistry = exports.createFileWatcherRegistry = void 0;
const node_fs_1 = __importStar(require("node:fs"));
const getRegistryKey = (file, existenceOnly) => existenceOnly ? `${file}\0existence-only` : file;
const createFileWatcherRegistry = () => {
    const sharedWatchers = new Map();
    const _installFileWatcher = ({ file, onChange, existenceOnly, }) => {
        const registryKey = getRegistryKey(file, existenceOnly);
        const existing = sharedWatchers.get(registryKey);
        if (existing) {
            existing.subscribers.add(onChange);
            return {
                exists: existing.existedBefore,
                unwatch: () => {
                    existing.subscribers.delete(onChange);
                    if (existing.subscribers.size === 0) {
                        existing.unwatch();
                        sharedWatchers.delete(registryKey);
                    }
                },
            };
        }
        const existedAtBeginning = node_fs_1.default.existsSync(file);
        const shared = {
            subscribers: new Set([onChange]),
            existedBefore: existedAtBeginning,
            lastKnownContent: null,
            unwatch: () => {
                node_fs_1.default.unwatchFile(file, listener);
            },
        };
        const listener = () => {
            const existsNow = node_fs_1.default.existsSync(file);
            let event = null;
            if (existenceOnly) {
                if (!shared.existedBefore && existsNow) {
                    shared.existedBefore = true;
                    event = {
                        type: 'created',
                        content: '',
                        originatorClientId: undefined,
                    };
                }
                else if (shared.existedBefore && !existsNow) {
                    shared.existedBefore = false;
                    event = { type: 'deleted' };
                }
            }
            else if (!shared.existedBefore && existsNow) {
                const content = (0, node_fs_1.readFileSync)(file, 'utf-8');
                shared.existedBefore = true;
                shared.lastKnownContent = content;
                event = { type: 'created', content, originatorClientId: undefined };
            }
            else if (shared.existedBefore && !existsNow) {
                shared.existedBefore = false;
                shared.lastKnownContent = null;
                event = { type: 'deleted' };
            }
            else if (existsNow) {
                if (typeof Deno !== 'undefined') {
                    return;
                }
                const content = (0, node_fs_1.readFileSync)(file, 'utf-8');
                if (shared.lastKnownContent !== null &&
                    content === shared.lastKnownContent) {
                    return;
                }
                shared.lastKnownContent = content;
                event = { type: 'changed', content, originatorClientId: undefined };
            }
            if (!event) {
                return;
            }
            for (const subscriber of shared.subscribers) {
                subscriber(event);
            }
        };
        node_fs_1.default.watchFile(file, { interval: 100 }, listener);
        sharedWatchers.set(registryKey, shared);
        return {
            exists: existedAtBeginning,
            unwatch: () => {
                shared.subscribers.delete(onChange);
                if (shared.subscribers.size === 0) {
                    shared.unwatch();
                    sharedWatchers.delete(registryKey);
                }
            },
        };
    };
    const _writeFileAndNotifyFileWatchers = (file, content, originatorClientId) => {
        (0, node_fs_1.writeFileSync)(file, content);
        const shared = sharedWatchers.get(getRegistryKey(file, false));
        if (!shared) {
            return;
        }
        shared.lastKnownContent = content;
        shared.existedBefore = true;
        for (const subscriber of shared.subscribers) {
            subscriber({ type: 'changed', content, originatorClientId });
        }
    };
    return {
        installFileWatcher: _installFileWatcher,
        writeFileAndNotifyFileWatchers: _writeFileAndNotifyFileWatchers,
    };
};
exports.createFileWatcherRegistry = createFileWatcherRegistry;
let currentRegistry = null;
const setFileWatcherRegistry = (registry) => {
    currentRegistry = registry;
    return () => {
        currentRegistry = null;
    };
};
exports.setFileWatcherRegistry = setFileWatcherRegistry;
const getFileWatcherRegistry = () => {
    if (!currentRegistry) {
        throw new Error('File watcher registry not initialized');
    }
    return currentRegistry;
};
exports.getFileWatcherRegistry = getFileWatcherRegistry;
const installFileWatcher = (options) => {
    if (!currentRegistry) {
        return {
            exists: false,
            unwatch: () => { },
        };
    }
    return (0, exports.getFileWatcherRegistry)().installFileWatcher(options);
};
exports.installFileWatcher = installFileWatcher;
const writeFileAndNotifyFileWatchers = (file, content, originatorClientId) => {
    if (!currentRegistry) {
        return;
    }
    (0, exports.getFileWatcherRegistry)().writeFileAndNotifyFileWatchers(file, content, originatorClientId);
};
exports.writeFileAndNotifyFileWatchers = writeFileAndNotifyFileWatchers;
