// @remotion/timeline-utils/dist/audio-waveform/load-waveform-peaks.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TARGET_SAMPLE_RATE = void 0;
exports.loadWaveformPeaks = loadWaveformPeaks;
const mediabunny_1 = require("mediabunny");
const constants_1 = require("./constants");
Object.defineProperty(exports, "TARGET_SAMPLE_RATE", { enumerable: true, get: function () { return constants_1.TARGET_SAMPLE_RATE; } });
const trim_audio_sample_before_zero_1 = require("./trim-audio-sample-before-zero");
const waveform_peak_processor_1 = require("./waveform-peak-processor");
const DEFAULT_PROGRESS_INTERVAL_IN_MS = 50;
const peaksCache = new Map();
async function loadWaveformPeaks(url, signal, options) {
    var _a, _b;
    const cached = peaksCache.get(url);
    if (cached) {
        (0, waveform_peak_processor_1.emitWaveformProgress)({
            peaks: cached,
            completedPeaks: cached.length,
            totalPeaks: cached.length,
            final: true,
            onProgress: options === null || options === void 0 ? void 0 : options.onProgress,
        });
        return cached;
    }
    const input = new mediabunny_1.Input({
        formats: mediabunny_1.ALL_FORMATS,
        source: new mediabunny_1.UrlSource(url),
    });
    try {
        const audioTrack = await input.getPrimaryAudioTrack();
        if (!audioTrack) {
            return new Float32Array(0);
        }
        if (await audioTrack.isLive()) {
            throw new Error('Live streams are not currently supported by Remotion. Sorry! Source: ' +
                url);
        }
        if (await audioTrack.isRelativeToUnixEpoch()) {
            throw new Error('Streams with UNIX timestamps are not currently supported by Remotion. Sorry! Source: ' +
                url);
        }
        const sampleRate = await audioTrack.getSampleRate();
        const durationInSeconds = (_a = (await audioTrack.getDurationFromMetadata({ skipLiveWait: true }))) !== null && _a !== void 0 ? _a : (await audioTrack.computeDuration({ skipLiveWait: true }));
        const totalPeaks = Math.ceil(durationInSeconds * constants_1.TARGET_SAMPLE_RATE);
        const samplesPerPeak = Math.max(1, Math.floor(sampleRate / constants_1.TARGET_SAMPLE_RATE));
        const sink = new mediabunny_1.AudioSampleSink(audioTrack);
        const processor = (0, waveform_peak_processor_1.createWaveformPeakProcessor)({
            totalPeaks,
            samplesPerPeak,
            onProgress: options === null || options === void 0 ? void 0 : options.onProgress,
            progressIntervalInMs: (_b = options === null || options === void 0 ? void 0 : options.progressIntervalInMs) !== null && _b !== void 0 ? _b : DEFAULT_PROGRESS_INTERVAL_IN_MS,
            now: () => Date.now(),
        });
        for await (const sample of sink.samples()) {
            if (signal.aborted) {
                sample.close();
                return new Float32Array(0);
            }
            const startFrame = (0, trim_audio_sample_before_zero_1.getAudioSampleStartFrameAtTimelineZero)(sample);
            if (startFrame === null) {
                sample.close();
                continue;
            }
            const frameCount = sample.numberOfFrames - startFrame;
            if (frameCount <= 0) {
                sample.close();
                continue;
            }
            const bytesNeeded = sample.allocationSize({
                format: 'f32',
                planeIndex: 0,
                frameOffset: startFrame,
                frameCount,
            });
            const floats = new Float32Array(bytesNeeded / 4);
            sample.copyTo(floats, {
                format: 'f32',
                planeIndex: 0,
                frameOffset: startFrame,
                frameCount,
            });
            const channels = Math.max(1, sample.numberOfChannels);
            sample.close();
            processor.processSampleChunk(floats, channels);
        }
        processor.finalize();
        const { peaks } = processor;
        peaksCache.set(url, peaks);
        return peaks;
    }
    finally {
        input.dispose();
    }
}
