// @remotion/cli/dist/config/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ConfigInternals = exports.Config = void 0;
const client_1 = require("@remotion/renderer/client");
const studio_server_1 = require("@remotion/studio-server");
const browser_1 = require("./browser");
const buffer_state_delay_in_milliseconds_1 = require("./buffer-state-delay-in-milliseconds");
const concurrency_1 = require("./concurrency");
const entry_point_1 = require("./entry-point");
const env_file_1 = require("./env-file");
const ffmpeg_override_1 = require("./ffmpeg-override");
const image_sequence_1 = require("./image-sequence");
const metadata_1 = require("./metadata");
const output_location_1 = require("./output-location");
const output_location_2 = require("./output-location");
const override_webpack_1 = require("./override-webpack");
const override_webpack_2 = require("./override-webpack");
const preview_server_1 = require("./preview-server");
const preview_server_2 = require("./preview-server");
const still_frame_1 = require("./still-frame");
const webpack_caching_1 = require("./webpack-caching");
const webpack_poll_1 = require("./webpack-poll");
const { allowHtmlInCanvasOption, benchmarkConcurrenciesOption, concurrencyOption, offthreadVideoCacheSizeInBytesOption, x264Option, audioBitrateOption, videoBitrateOption, scaleOption, crfOption, jpegQualityOption, enforceAudioOption, overwriteOption, chromeModeOption, mutedOption, videoCodecOption, colorSpaceOption, disallowParallelEncodingOption, deleteAfterOption, folderExpiryOption, enableMultiprocessOnLinuxOption, glOption, gopSizeOption, headlessOption, numberOfGifLoopsOption, beepOnFinishOption, encodingMaxRateOption, encodingBufferSizeOption, reproOption, enableLambdaInsights, logLevelOption, delayRenderTimeoutInMillisecondsOption, publicDirOption, binariesDirectoryOption, preferLosslessOption, framesOption, forSeamlessAacConcatenationOption, audioCodecOption, publicPathOption, hardwareAccelerationOption, audioLatencyHintOption, enableCrossSiteIsolationOption, imageSequencePatternOption, darkModeOption, askAIOption, publicLicenseKeyOption, experimentalClientSideRenderingOption, keyboardShortcutsOption, forceNewStudioOption, numberOfSharedAudioTagsOption, ipv4Option, pixelFormatOption, browserExecutableOption, everyNthFrameOption, proResProfileOption, stillImageFormatOption, videoImageFormatOption, userAgentOption, disableWebSecurityOption, ignoreCertificateErrorsOption, overrideHeightOption, overrideWidthOption, overrideFpsOption, overrideDurationOption, rspackOption, outDirOption, webpackPollOption, imageSequenceOption, bundleCacheOption, envFileOption, runsOption, noOpenOption, sampleRateOption, previewSampleRateOption, } = client_1.BrowserSafeApis.options;
exports.Config = {
    get Bundling() {
        throw new Error('The config format has changed. Change `Config.Bundling.*()` calls to `Config.*()` in your config file.');
    },
    get Rendering() {
        throw new Error('The config format has changed. Change `Config.Rendering.*()` calls to `Config.*()` in your config file.');
    },
    get Output() {
        throw new Error('The config format has changed. Change `Config.Output.*()` calls to `Config.*()` in your config file.');
    },
    get Log() {
        throw new Error('The config format has changed. Change `Config.Log.*()` calls to `Config.*()` in your config file.');
    },
    get Preview() {
        throw new Error('The config format has changed. Change `Config.Preview.*()` calls to `Config.*()` in your config file.');
    },
    get Puppeteer() {
        throw new Error('The config format has changed. Change `Config.Puppeteer.*()` calls to `Config.*()` in your config file.');
    },
    setMaxTimelineTracks: studio_server_1.StudioServerInternals.setMaxTimelineTracks,
    setKeyboardShortcutsEnabled: keyboardShortcutsOption.setConfig,
    setExperimentalClientSideRenderingEnabled: experimentalClientSideRenderingOption.setConfig,
    setAllowHtmlInCanvasEnabled: allowHtmlInCanvasOption.setConfig,
    setExperimentalRspackEnabled: rspackOption.setConfig,
    setNumberOfSharedAudioTags: numberOfSharedAudioTagsOption.setConfig,
    setWebpackPollingInMilliseconds: webpackPollOption.setConfig,
    setShouldOpenBrowser: noOpenOption.setConfig,
    setBufferStateDelayInMilliseconds: buffer_state_delay_in_milliseconds_1.setBufferStateDelayInMilliseconds,
    overrideWebpackConfig: override_webpack_2.overrideWebpackConfig,
    setCachingEnabled: bundleCacheOption.setConfig,
    setPort: preview_server_2.setPort,
    setStudioPort: preview_server_2.setStudioPort,
    setRendererPort: preview_server_2.setRendererPort,
    setPublicDir: publicDirOption.setConfig,
    setEntryPoint: entry_point_1.setEntryPoint,
    setLogLevel: logLevelOption.setConfig,
    setLevel: logLevelOption.setConfig,
    setBrowserExecutable: browserExecutableOption.setConfig,
    setTimeoutInMilliseconds: delayRenderTimeoutInMillisecondsOption.setConfig,
    setDelayRenderTimeoutInMilliseconds: delayRenderTimeoutInMillisecondsOption.setConfig,
    setChromiumDisableWebSecurity: disableWebSecurityOption.setConfig,
    setChromiumIgnoreCertificateErrors: ignoreCertificateErrorsOption.setConfig,
    setChromiumHeadlessMode: headlessOption.setConfig,
    setChromiumOpenGlRenderer: glOption.setConfig,
    setChromiumUserAgent: userAgentOption.setConfig,
    setDotEnvLocation: envFileOption.setConfig,
    setConcurrency: concurrencyOption.setConfig,
    setChromiumMultiProcessOnLinux: enableMultiprocessOnLinuxOption.setConfig,
    setChromiumDarkMode: darkModeOption.setConfig,
    setQuality: () => {
        throw new Error('setQuality() has been renamed - use setJpegQuality() instead.');
    },
    setImageFormat: () => {
        throw new Error('Config.setImageFormat() has been renamed - use Config.setVideoImageFormat() instead (default "jpeg"). For rendering stills, use Config.setStillImageFormat() (default "png")');
    },
    setJpegQuality: jpegQualityOption.setConfig,
    setStillImageFormat: stillImageFormatOption.setConfig,
    setVideoImageFormat: videoImageFormatOption.setConfig,
    setMetadata: metadata_1.setMetadata,
    setEncodingMaxRate: encodingMaxRateOption.setConfig,
    setEncodingBufferSize: encodingBufferSizeOption.setConfig,
    setFrameRange: framesOption.setConfig,
    setScale: scaleOption.setConfig,
    setEveryNthFrame: everyNthFrameOption.setConfig,
    setNumberOfGifLoops: numberOfGifLoopsOption.setConfig,
    setMuted: mutedOption.setConfig,
    setEnforceAudioTrack: enforceAudioOption.setConfig,
    setOutputLocation: output_location_2.setOutputLocation,
    setOverwriteOutput: overwriteOption.setConfig,
    setChromeMode: chromeModeOption.setConfig,
    setPixelFormat: pixelFormatOption.setConfig,
    setCodec: videoCodecOption.setConfig,
    setCrf: crfOption.setConfig,
    setGopSize: gopSizeOption.setConfig,
    setImageSequence: imageSequenceOption.setConfig,
    setProResProfile: proResProfileOption.setConfig,
    setX264Preset: x264Option.setConfig,
    setAudioBitrate: audioBitrateOption.setConfig,
    setVideoBitrate: videoBitrateOption.setConfig,
    setAudioLatencyHint: audioLatencyHintOption.setConfig,
    setForSeamlessAacConcatenation: forSeamlessAacConcatenationOption.setConfig,
    overrideHeight: overrideHeightOption.setConfig,
    overrideWidth: overrideWidthOption.setConfig,
    overrideFps: overrideFpsOption.setConfig,
    overrideDuration: overrideDurationOption.setConfig,
    overrideFfmpegCommand: ffmpeg_override_1.setFfmpegOverrideFunction,
    setAudioCodec: audioCodecOption.setConfig,
    setOffthreadVideoCacheSizeInBytes: offthreadVideoCacheSizeInBytesOption.setConfig,
    setDeleteAfter: deleteAfterOption.setConfig,
    setColorSpace: colorSpaceOption.setConfig,
    setDisallowParallelEncoding: disallowParallelEncodingOption.setConfig,
    setBeepOnFinish: beepOnFinishOption.setConfig,
    setEnableFolderExpiry: folderExpiryOption.setConfig,
    setRepro: reproOption.setConfig,
    setLambdaInsights: enableLambdaInsights.setConfig,
    setBinariesDirectory: binariesDirectoryOption.setConfig,
    setPreferLosslessAudio: preferLosslessOption.setConfig,
    setPublicPath: publicPathOption.setConfig,
    setImageSequencePattern: imageSequencePatternOption.setConfig,
    setHardwareAcceleration: hardwareAccelerationOption.setConfig,
    setEnableCrossSiteIsolation: enableCrossSiteIsolationOption.setConfig,
    setAskAIEnabled: askAIOption.setConfig,
    setPublicLicenseKey: publicLicenseKeyOption.setConfig,
    setForceNewStudioEnabled: forceNewStudioOption.setConfig,
    setIPv4: ipv4Option.setConfig,
    setBundleOutDir: outDirOption.setConfig,
    setBenchmarkRuns: runsOption.setConfig,
    setBenchmarkConcurrencies: benchmarkConcurrenciesOption.setConfig,
    setSampleRate: sampleRateOption.setConfig,
    setPreviewSampleRate: previewSampleRateOption.setConfig,
};
exports.ConfigInternals = {
    getBrowser: browser_1.getBrowser,
    getStudioPort: preview_server_1.getStudioPort,
    getRendererPortFromConfigFile: preview_server_1.getRendererPortFromConfigFile,
    getRendererPortFromConfigFileAndCliFlag: preview_server_1.getRendererPortFromConfigFileAndCliFlag,
    getConcurrency: concurrency_1.getConcurrency,
    getStillFrame: still_frame_1.getStillFrame,
    getShouldOutputImageSequence: image_sequence_1.getShouldOutputImageSequence,
    getDotEnvLocation: env_file_1.getDotEnvLocation,
    getWebpackOverrideFn: override_webpack_1.getWebpackOverrideFn,
    getWebpackCaching: webpack_caching_1.getWebpackCaching,
    getOutputLocation: output_location_1.getOutputLocation,
    setStillFrame: still_frame_1.setStillFrame,
    getMaxTimelineTracks: studio_server_1.StudioServerInternals.getMaxTimelineTracks,
    defaultOverrideFunction: override_webpack_1.defaultOverrideFunction,
    getFfmpegOverrideFunction: ffmpeg_override_1.getFfmpegOverrideFunction,
    getMetadata: metadata_1.getMetadata,
    getEntryPoint: entry_point_1.getEntryPoint,
    getWebpackPolling: webpack_poll_1.getWebpackPolling,
    getBufferStateDelayInMilliseconds: buffer_state_delay_in_milliseconds_1.getBufferStateDelayInMilliseconds,
    getOutputCodecOrUndefined: client_1.BrowserSafeApis.getOutputCodecOrUndefined,
};
