// @remotion/player/dist/cjs/use-playback.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.usePlayback = void 0;
const react_1 = require("react");
/* eslint-disable @typescript-eslint/no-use-before-define */
const react_2 = require("react");
const remotion_1 = require("remotion");
const browser_mediasession_js_1 = require("./browser-mediasession.js");
const calculate_next_frame_js_1 = require("./calculate-next-frame.js");
const is_backgrounded_js_1 = require("./is-backgrounded.js");
const set_global_time_anchor_js_1 = require("./set-global-time-anchor.js");
const use_player_js_1 = require("./use-player.js");
const shouldForceAnchorChange = (newState) => {
    if (newState === 'suspended' || newState === 'running-to-suspended') {
        return true;
    }
    if (newState === 'closed' ||
        newState === 'interrupted' ||
        newState === 'running' ||
        newState === 'suspended-to-running') {
        return false;
    }
    throw new Error(`Unexpected audio context state: ${newState}`);
};
const usePlayback = ({ loop, playbackRate, moveToBeginningWhenEnded, inFrame, outFrame, browserMediaControlsBehavior, getCurrentFrame, muted, }) => {
    const config = remotion_1.Internals.useUnsafeVideoConfig();
    const frame = remotion_1.Internals.Timeline.useTimelinePosition();
    const { playing, pause, emitter, isPlaying } = (0, use_player_js_1.usePlayer)();
    const setFrame = remotion_1.Internals.Timeline.useTimelineSetFrame();
    const sharedAudioContext = (0, react_2.useContext)(remotion_1.Internals.SharedAudioContext);
    const logLevel = remotion_1.Internals.useLogLevel();
    // requestAnimationFrame() does not work if the tab is not active.
    // This means that audio will keep playing even if it has ended.
    // In that case, we use setTimeout() instead.
    const isBackgroundedRef = (0, is_backgrounded_js_1.useIsBackgrounded)();
    const lastTimeUpdateTimestamp = (0, react_2.useRef)(0);
    const context = (0, react_2.useContext)(remotion_1.Internals.BufferingContextReact);
    if (!context) {
        throw new Error('Missing the buffering context. Most likely you have a Remotion version mismatch.');
    }
    (0, browser_mediasession_js_1.useBrowserMediaSession)({
        browserMediaControlsBehavior,
        playbackRate,
        videoConfig: config,
    });
    // Update time anchor when seeking:
    // If the user clicked on a different time in the timeline, we need to re-sync the anchor
    (0, react_1.useLayoutEffect)(() => {
        if (!sharedAudioContext) {
            return;
        }
        if (!sharedAudioContext.audioContext) {
            return;
        }
        if (!config) {
            return;
        }
        if (muted) {
            return;
        }
        const changed = (0, set_global_time_anchor_js_1.setGlobalTimeAnchor)({
            audioContext: sharedAudioContext.audioContext,
            audioSyncAnchor: sharedAudioContext.audioSyncAnchor,
            absoluteTimeInSeconds: frame / config.fps,
            globalPlaybackRate: playbackRate,
            logLevel,
            force: false,
        });
        if (changed) {
            sharedAudioContext.audioSyncAnchorEmitter.dispatch('changed');
        }
    }, [config, frame, logLevel, playbackRate, sharedAudioContext, muted]);
    // When the audio context is suspended, we use the opportunity to
    // re-anchor the time to be exact.
    (0, react_1.useLayoutEffect)(() => {
        const audioContext = sharedAudioContext === null || sharedAudioContext === void 0 ? void 0 : sharedAudioContext.audioContext;
        if (!audioContext) {
            return;
        }
        if (!config) {
            return;
        }
        if (muted) {
            return;
        }
        const callback = () => {
            const newState = sharedAudioContext === null || sharedAudioContext === void 0 ? void 0 : sharedAudioContext.getAudioContextState();
            if (newState && shouldForceAnchorChange(newState)) {
                (0, set_global_time_anchor_js_1.setGlobalTimeAnchor)({
                    audioContext,
                    audioSyncAnchor: sharedAudioContext.audioSyncAnchor,
                    absoluteTimeInSeconds: getCurrentFrame() / config.fps,
                    globalPlaybackRate: playbackRate,
                    logLevel,
                    force: true,
                });
            }
        };
        audioContext === null || audioContext === void 0 ? void 0 : audioContext.addEventListener('statechange', callback);
        return () => {
            audioContext === null || audioContext === void 0 ? void 0 : audioContext.removeEventListener('statechange', callback);
        };
    }, [
        config,
        getCurrentFrame,
        logLevel,
        muted,
        playbackRate,
        sharedAudioContext,
    ]);
    (0, react_2.useEffect)(() => {
        var _a;
        if (!config) {
            return;
        }
        if (!playing) {
            (_a = sharedAudioContext === null || sharedAudioContext === void 0 ? void 0 : sharedAudioContext.suspend) === null || _a === void 0 ? void 0 : _a.call(sharedAudioContext);
            return;
        }
        let hasBeenStopped = false;
        let reqAnimFrameCall = null;
        let startedTime = performance.now();
        let framesAdvanced = 0;
        const cancelQueuedFrame = () => {
            if (reqAnimFrameCall !== null) {
                if (reqAnimFrameCall.type === 'raf') {
                    cancelAnimationFrame(reqAnimFrameCall.id);
                }
                else {
                    clearTimeout(reqAnimFrameCall.id);
                }
            }
        };
        const stop = () => {
            hasBeenStopped = true;
            cancelQueuedFrame();
        };
        const callback = () => {
            var _a, _b;
            if (hasBeenStopped) {
                return;
            }
            if (!isPlaying()) {
                (_a = sharedAudioContext === null || sharedAudioContext === void 0 ? void 0 : sharedAudioContext.suspend) === null || _a === void 0 ? void 0 : _a.call(sharedAudioContext);
                return;
            }
            if (!muted && !context.buffering.current) {
                (_b = sharedAudioContext === null || sharedAudioContext === void 0 ? void 0 : sharedAudioContext.resume) === null || _b === void 0 ? void 0 : _b.call(sharedAudioContext);
            }
            const time = performance.now() - startedTime;
            const actualLastFrame = outFrame !== null && outFrame !== void 0 ? outFrame : config.durationInFrames - 1;
            const actualFirstFrame = inFrame !== null && inFrame !== void 0 ? inFrame : 0;
            const currentFrame = getCurrentFrame();
            const { nextFrame, framesToAdvance, hasEnded } = (0, calculate_next_frame_js_1.calculateNextFrame)({
                time,
                currentFrame,
                playbackSpeed: playbackRate,
                fps: config.fps,
                actualFirstFrame,
                actualLastFrame,
                framesAdvanced,
                shouldLoop: loop,
            });
            framesAdvanced += framesToAdvance;
            if (nextFrame !== getCurrentFrame() &&
                (!hasEnded || moveToBeginningWhenEnded) &&
                !context.buffering.current) {
                setFrame((c) => ({ ...c, [config.id]: nextFrame }));
            }
            if (hasEnded) {
                stop();
                pause();
                emitter.dispatchEnded();
                return;
            }
            queueNextFrame();
        };
        const queueNextFrame = () => {
            var _a, _b;
            var _c;
            const getIsResumingAudioContext = (_c = (_a = sharedAudioContext === null || sharedAudioContext === void 0 ? void 0 : sharedAudioContext.getIsResumingAudioContext) === null || _a === void 0 ? void 0 : _a.call(sharedAudioContext)) !== null && _c !== void 0 ? _c : null;
            if (getIsResumingAudioContext !== null && !muted) {
                getIsResumingAudioContext.then(() => {
                    startedTime = performance.now();
                    framesAdvanced = 0;
                    queueNextFrame();
                });
                return;
            }
            if (context.buffering.current) {
                if (!muted) {
                    (_b = sharedAudioContext === null || sharedAudioContext === void 0 ? void 0 : sharedAudioContext.suspend) === null || _b === void 0 ? void 0 : _b.call(sharedAudioContext);
                }
                const stopListening = context.listenForResume(() => {
                    stopListening.remove();
                    startedTime = performance.now();
                    framesAdvanced = 0;
                    queueNextFrame();
                });
                return;
            }
            if (isBackgroundedRef.current) {
                reqAnimFrameCall = {
                    type: 'timeout',
                    // Note: Most likely, this will not be 1000 / fps, but the browser will throttle it to ~1/sec.
                    id: setTimeout(callback, 1000 / config.fps),
                };
                return;
            }
            reqAnimFrameCall = { type: 'raf', id: requestAnimationFrame(callback) };
        };
        queueNextFrame();
        const onVisibilityChange = () => {
            if (document.visibilityState === 'visible') {
                return;
            }
            // If tab goes into the background, cancel requestAnimationFrame() and update immediately.
            // , so the transition to setTimeout() can be fulfilled.
            cancelQueuedFrame();
            callback();
        };
        window.addEventListener('visibilitychange', onVisibilityChange);
        return () => {
            window.removeEventListener('visibilitychange', onVisibilityChange);
            stop();
        };
    }, [
        config,
        loop,
        pause,
        playing,
        setFrame,
        emitter,
        playbackRate,
        inFrame,
        outFrame,
        moveToBeginningWhenEnded,
        isBackgroundedRef,
        getCurrentFrame,
        context,
        isPlaying,
        sharedAudioContext,
        logLevel,
        muted,
    ]);
    (0, react_2.useEffect)(() => {
        const now = performance.now();
        const timeSinceLastUpdate = now - lastTimeUpdateTimestamp.current;
        if (timeSinceLastUpdate >= 250) {
            emitter.dispatchTimeUpdate({ frame });
            lastTimeUpdateTimestamp.current = now;
            return;
        }
        const timeoutId = setTimeout(() => {
            emitter.dispatchTimeUpdate({ frame });
            lastTimeUpdateTimestamp.current = performance.now();
        }, 250 - timeSinceLastUpdate);
        return () => clearTimeout(timeoutId);
    }, [emitter, frame]);
    (0, react_2.useEffect)(() => {
        emitter.dispatchFrameUpdate({ frame });
    }, [emitter, frame]);
};
exports.usePlayback = usePlayback;
