// @remotion/studio/dist/components/EditorRuler/Ruler.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const jsx_runtime_1 = require("react/jsx-runtime");
const react_1 = require("react");
const remotion_1 = require("remotion");
const colors_1 = require("../../helpers/colors");
const editor_guide_selection_1 = require("../../helpers/editor-guide-selection");
const editor_ruler_1 = require("../../helpers/editor-ruler");
const editor_guides_1 = require("../../state/editor-guides");
const editor_rulers_1 = require("../../state/editor-rulers");
const ForceSpecificCursor_1 = require("../ForceSpecificCursor");
const should_clear_selection_on_pointer_down_1 = require("../Timeline/should-clear-selection-on-pointer-down");
const TimelineSelection_1 = require("../Timeline/TimelineSelection");
const makeGuideId = () => {
    return Math.random().toString(36).substring(7);
};
const Ruler = ({ scale, points, originOffset, startMarking, size, markingGaps, orientation, }) => {
    const rulerCanvasRef = (0, react_1.useRef)(null);
    const isVerticalRuler = orientation === 'vertical';
    const { shouldCreateGuideRef, setGuidesList, draggingGuideId, hoveredGuideId, setDraggingGuideId, guidesList, setEditorShowGuides, } = (0, react_1.useContext)(editor_guides_1.EditorShowGuidesContext);
    const { selectedItems } = (0, TimelineSelection_1.useTimelineSelection)();
    const unsafeVideoConfig = remotion_1.Internals.useUnsafeVideoConfig();
    if (!unsafeVideoConfig) {
        throw new Error('Video config not set');
    }
    const cursor = isVerticalRuler ? 'ew-resize' : 'ns-resize';
    const guideHighlight = (0, react_1.useMemo)(() => (0, editor_guide_selection_1.getRulerGuideHighlight)({
        guidesList,
        selectedItems,
        hoveredGuideId,
        draggingGuideId,
    }), [draggingGuideId, guidesList, hoveredGuideId, selectedItems]);
    const rulerWidth = isVerticalRuler ? editor_rulers_1.RULER_WIDTH : size.width - editor_rulers_1.RULER_WIDTH;
    const rulerHeight = isVerticalRuler ? size.height - editor_rulers_1.RULER_WIDTH : editor_rulers_1.RULER_WIDTH;
    (0, react_1.useEffect)(() => {
        (0, editor_ruler_1.drawMarkingOnRulerCanvas)({
            scale,
            points,
            startMarking,
            originOffset,
            markingGaps,
            orientation,
            rulerCanvasRef,
            guideHighlight,
            canvasHeight: rulerHeight * window.devicePixelRatio,
            canvasWidth: rulerWidth * window.devicePixelRatio,
        });
    }, [
        scale,
        points,
        startMarking,
        originOffset,
        markingGaps,
        orientation,
        guideHighlight,
        size,
        rulerHeight,
        rulerWidth,
    ]);
    const rulerStyle = (0, react_1.useMemo)(() => ({
        position: 'absolute',
        background: colors_1.BACKGROUND,
        width: rulerWidth,
        height: rulerHeight,
        left: isVerticalRuler ? 0 : 'unset',
        top: isVerticalRuler ? 'unset' : 0,
        borderBottom: isVerticalRuler ? undefined : '1px solid ' + colors_1.RULER_COLOR,
        borderRight: isVerticalRuler ? '1px solid ' + colors_1.RULER_COLOR : undefined,
        cursor,
    }), [rulerWidth, rulerHeight, cursor, isVerticalRuler]);
    const onPointerDown = (0, react_1.useCallback)((e) => {
        if (e.button !== 0) {
            return;
        }
        e.preventDefault();
        // Prevent deselection of currently selected items
        e.stopPropagation();
        shouldCreateGuideRef.current = true;
        (0, ForceSpecificCursor_1.forceSpecificCursor)(cursor);
        const guideId = makeGuideId();
        setEditorShowGuides(() => true);
        setDraggingGuideId(() => guideId);
        setGuidesList((prevState) => {
            return [
                ...prevState,
                {
                    orientation,
                    position: -originOffset,
                    show: false,
                    id: guideId,
                    compositionId: unsafeVideoConfig.id,
                },
            ];
        });
    }, [
        shouldCreateGuideRef,
        setEditorShowGuides,
        setDraggingGuideId,
        setGuidesList,
        orientation,
        originOffset,
        unsafeVideoConfig.id,
        cursor,
    ]);
    return (jsx_runtime_1.jsx("canvas", { ref: rulerCanvasRef, width: rulerWidth * window.devicePixelRatio, height: rulerHeight * window.devicePixelRatio, style: rulerStyle, [should_clear_selection_on_pointer_down_1.PREVENT_CLEAR_SELECTION_ON_POINTER_DOWN_ATTR]: 'true', onPointerDown: onPointerDown }));
};
exports.default = Ruler;
