// @remotion/studio/dist/components/Timeline/call-move-keyframe.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.callMoveKeyframes = void 0;
const studio_shared_1 = require("@remotion/studio-shared");
const call_api_1 = require("../call-api");
const groupByNodePath = (keyframes) => {
    var _a;
    const groups = new Map();
    for (const keyframe of keyframes) {
        const key = JSON.stringify(keyframe.nodePath);
        const group = (_a = groups.get(key)) !== null && _a !== void 0 ? _a : [];
        group.push(keyframe);
        groups.set(key, group);
    }
    return [...groups.values()];
};
const callMoveKeyframes = ({ sequenceKeyframes, effectKeyframes, setPropStatuses, clientId, }) => {
    if (sequenceKeyframes.length === 0 && effectKeyframes.length === 0) {
        return Promise.resolve();
    }
    for (const keyframes of groupByNodePath(sequenceKeyframes)) {
        const [firstKeyframe] = keyframes;
        if (!firstKeyframe) {
            continue;
        }
        setPropStatuses(firstKeyframe.nodePath, (prev) => (0, studio_shared_1.optimisticMoveSequenceKeyframes)({
            previous: prev,
            keyframes: keyframes.map((keyframe) => ({
                fieldKey: keyframe.fieldKey,
                fromFrame: keyframe.fromFrame,
                toFrame: keyframe.toFrame,
            })),
        }));
    }
    for (const keyframes of groupByNodePath(effectKeyframes)) {
        const [firstKeyframe] = keyframes;
        if (!firstKeyframe) {
            continue;
        }
        setPropStatuses(firstKeyframe.nodePath, (prev) => (0, studio_shared_1.optimisticMoveEffectKeyframes)({
            previous: prev,
            keyframes: keyframes.map((keyframe) => ({
                effectIndex: keyframe.effectIndex,
                fieldKey: keyframe.fieldKey,
                fromFrame: keyframe.fromFrame,
                toFrame: keyframe.toFrame,
            })),
        }));
    }
    return (0, call_api_1.callApi)('/api/move-keyframes', {
        sequenceKeyframes: sequenceKeyframes.map((keyframe) => ({
            fileName: keyframe.fileName,
            nodePath: keyframe.nodePath,
            key: keyframe.fieldKey,
            fromFrame: keyframe.fromFrame,
            toFrame: keyframe.toFrame,
            schema: keyframe.schema,
        })),
        effectKeyframes: effectKeyframes.map((keyframe) => ({
            fileName: keyframe.fileName,
            sequenceNodePath: keyframe.nodePath,
            effectIndex: keyframe.effectIndex,
            key: keyframe.fieldKey,
            fromFrame: keyframe.fromFrame,
            toFrame: keyframe.toFrame,
            schema: keyframe.schema,
        })),
        clientId,
    }).then(() => undefined);
};
exports.callMoveKeyframes = callMoveKeyframes;
