// @remotion/studio/dist/components/Timeline/call-delete-keyframe.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.callDeleteKeyframes = exports.callDeleteEffectKeyframe = exports.callDeleteSequenceKeyframe = void 0;
const studio_shared_1 = require("@remotion/studio-shared");
const call_api_1 = require("../call-api");
const save_prop_queue_1 = require("./save-prop-queue");
const groupByNodePath = (keyframes) => {
    var _a;
    const groups = new Map();
    for (const keyframe of keyframes) {
        const key = JSON.stringify(keyframe.nodePath);
        const group = (_a = groups.get(key)) !== null && _a !== void 0 ? _a : [];
        group.push(keyframe);
        groups.set(key, group);
    }
    return [...groups.values()];
};
const callDeleteSequenceKeyframe = ({ fileName, nodePath, fieldKey, sourceFrame, schema, setPropStatuses, clientId, }) => {
    return (0, save_prop_queue_1.enqueueSavePropChange)({
        nodePath,
        setPropStatuses,
        applyOptimistic: (prev) => (0, studio_shared_1.optimisticDeleteSequenceKeyframe)({
            previous: prev,
            fieldKey,
            frame: sourceFrame,
        }),
        apiCall: () => (0, call_api_1.callApi)('/api/delete-keyframes', {
            sequenceKeyframes: [
                {
                    fileName,
                    nodePath,
                    key: fieldKey,
                    frame: sourceFrame,
                    schema,
                },
            ],
            effectKeyframes: [],
            clientId,
        }),
        errorLabel: 'Could not delete keyframe',
    });
};
exports.callDeleteSequenceKeyframe = callDeleteSequenceKeyframe;
const callDeleteEffectKeyframe = ({ fileName, nodePath, effectIndex, fieldKey, sourceFrame, schema, setPropStatuses, clientId, }) => {
    return (0, save_prop_queue_1.enqueueSavePropChange)({
        nodePath,
        setPropStatuses,
        applyOptimistic: (prev) => (0, studio_shared_1.optimisticDeleteEffectKeyframe)({
            previous: prev,
            effectIndex,
            fieldKey,
            frame: sourceFrame,
        }),
        apiCall: () => (0, call_api_1.callApi)('/api/delete-keyframes', {
            sequenceKeyframes: [],
            effectKeyframes: [
                {
                    fileName,
                    sequenceNodePath: nodePath,
                    effectIndex,
                    key: fieldKey,
                    frame: sourceFrame,
                    schema,
                },
            ],
            clientId,
        }),
        errorLabel: 'Could not delete keyframe',
    });
};
exports.callDeleteEffectKeyframe = callDeleteEffectKeyframe;
const callDeleteKeyframes = ({ sequenceKeyframes, effectKeyframes, setPropStatuses, clientId, }) => {
    if (sequenceKeyframes.length === 0 && effectKeyframes.length === 0) {
        return Promise.resolve();
    }
    for (const keyframes of groupByNodePath(sequenceKeyframes)) {
        const [firstKeyframe] = keyframes;
        if (!firstKeyframe) {
            continue;
        }
        setPropStatuses(firstKeyframe.nodePath, (prev) => (0, studio_shared_1.optimisticDeleteSequenceKeyframes)({
            previous: prev,
            keyframes: keyframes.map((keyframe) => ({
                fieldKey: keyframe.fieldKey,
                frame: keyframe.sourceFrame,
            })),
        }));
    }
    for (const keyframes of groupByNodePath(effectKeyframes)) {
        const [firstKeyframe] = keyframes;
        if (!firstKeyframe) {
            continue;
        }
        setPropStatuses(firstKeyframe.nodePath, (prev) => (0, studio_shared_1.optimisticDeleteEffectKeyframes)({
            previous: prev,
            keyframes: keyframes.map((keyframe) => ({
                effectIndex: keyframe.effectIndex,
                fieldKey: keyframe.fieldKey,
                frame: keyframe.sourceFrame,
            })),
        }));
    }
    return (0, call_api_1.callApi)('/api/delete-keyframes', {
        sequenceKeyframes: sequenceKeyframes.map((keyframe) => ({
            fileName: keyframe.fileName,
            nodePath: keyframe.nodePath,
            key: keyframe.fieldKey,
            frame: keyframe.sourceFrame,
            schema: keyframe.schema,
        })),
        effectKeyframes: effectKeyframes.map((keyframe) => ({
            fileName: keyframe.fileName,
            sequenceNodePath: keyframe.nodePath,
            effectIndex: keyframe.effectIndex,
            key: keyframe.fieldKey,
            frame: keyframe.sourceFrame,
            schema: keyframe.schema,
        })),
        clientId,
    }).then(() => undefined);
};
exports.callDeleteKeyframes = callDeleteKeyframes;
