// @remotion/studio/dist/components/Timeline/call-update-keyframe-settings.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.callUpdateEffectKeyframeSettings = exports.callUpdateSequenceKeyframeSettings = void 0;
const studio_shared_1 = require("@remotion/studio-shared");
const call_api_1 = require("../call-api");
const apply_effect_response_to_prop_statuses_1 = require("./apply-effect-response-to-prop-statuses");
const save_prop_queue_1 = require("./save-prop-queue");
const callUpdateSequenceKeyframeSettings = ({ fileName, nodePath, fieldKey, settings, schema, setPropStatuses, clientId, }) => {
    return (0, save_prop_queue_1.enqueueSavePropChange)({
        nodePath,
        setPropStatuses,
        applyOptimistic: (prev) => (0, studio_shared_1.optimisticUpdateSequenceKeyframeSettings)({
            previous: prev,
            fieldKey,
            settings,
        }),
        apiCall: () => (0, call_api_1.callApi)('/api/update-sequence-keyframe-settings', {
            fileName,
            nodePath,
            key: fieldKey,
            settings,
            schema,
            clientId,
        }),
        errorLabel: 'Could not update keyframe settings',
    });
};
exports.callUpdateSequenceKeyframeSettings = callUpdateSequenceKeyframeSettings;
const callUpdateEffectKeyframeSettings = ({ fileName, nodePath, effectIndex, fieldKey, settings, schema, setPropStatuses, clientId, }) => {
    return (0, save_prop_queue_1.enqueueSavePropChange)({
        nodePath,
        setPropStatuses,
        applyOptimistic: (prev) => (0, studio_shared_1.optimisticUpdateEffectKeyframeSettings)({
            previous: prev,
            effectIndex,
            fieldKey,
            settings,
        }),
        applyServerResponse: (prev, response) => (0, apply_effect_response_to_prop_statuses_1.applyEffectResponseToPropStatuses)({ previous: prev, response }),
        apiCall: () => (0, call_api_1.callApi)('/api/update-effect-keyframe-settings', {
            fileName,
            sequenceNodePath: nodePath,
            effectIndex,
            key: fieldKey,
            settings,
            schema,
            clientId,
        }),
        errorLabel: 'Could not update keyframe settings',
    });
};
exports.callUpdateEffectKeyframeSettings = callUpdateEffectKeyframeSettings;
