// @remotion/studio/dist/components/Timeline/call-add-keyframe.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.callAddEffectKeyframe = exports.callAddKeyframes = exports.callAddSequenceKeyframe = void 0;
const studio_shared_1 = require("@remotion/studio-shared");
const call_api_1 = require("../call-api");
const apply_effect_response_to_prop_statuses_1 = require("./apply-effect-response-to-prop-statuses");
const save_prop_queue_1 = require("./save-prop-queue");
const groupByNodePath = (keyframes) => {
    var _a;
    const groups = new Map();
    for (const keyframe of keyframes) {
        const key = JSON.stringify(keyframe.nodePath);
        const group = (_a = groups.get(key)) !== null && _a !== void 0 ? _a : [];
        group.push(keyframe);
        groups.set(key, group);
    }
    return [...groups.values()];
};
const callAddSequenceKeyframe = ({ fileName, nodePath, fieldKey, sourceFrame, value, schema, setPropStatuses, clientId, }) => {
    return (0, save_prop_queue_1.enqueueSavePropChange)({
        nodePath,
        setPropStatuses,
        applyOptimistic: (prev) => (0, studio_shared_1.optimisticAddSequenceKeyframe)({
            previous: prev,
            fieldKey,
            frame: sourceFrame,
            value,
            schema,
        }),
        apiCall: () => (0, call_api_1.callApi)('/api/add-sequence-keyframe', {
            fileName,
            nodePath,
            key: fieldKey,
            frame: sourceFrame,
            value: JSON.stringify(value),
            schema,
            clientId,
        }),
        errorLabel: 'Could not add keyframe',
    });
};
exports.callAddSequenceKeyframe = callAddSequenceKeyframe;
const callAddKeyframes = ({ sequenceKeyframes, effectKeyframes, setPropStatuses, clientId, }) => {
    if (sequenceKeyframes.length === 0 && effectKeyframes.length === 0) {
        return Promise.resolve();
    }
    for (const keyframes of groupByNodePath(sequenceKeyframes)) {
        const [firstKeyframe] = keyframes;
        if (!firstKeyframe) {
            continue;
        }
        setPropStatuses(firstKeyframe.nodePath, (prev) => keyframes.reduce((current, keyframe) => (0, studio_shared_1.optimisticAddSequenceKeyframe)({
            previous: current,
            fieldKey: keyframe.fieldKey,
            frame: keyframe.sourceFrame,
            value: keyframe.value,
            schema: keyframe.schema,
        }), prev));
    }
    for (const keyframes of groupByNodePath(effectKeyframes)) {
        const [firstKeyframe] = keyframes;
        if (!firstKeyframe) {
            continue;
        }
        setPropStatuses(firstKeyframe.nodePath, (prev) => keyframes.reduce((current, keyframe) => (0, studio_shared_1.optimisticAddEffectKeyframe)({
            previous: current,
            effectIndex: keyframe.effectIndex,
            fieldKey: keyframe.fieldKey,
            frame: keyframe.sourceFrame,
            value: keyframe.value,
            schema: keyframe.schema,
        }), prev));
    }
    return (0, call_api_1.callApi)('/api/add-keyframes', {
        sequenceKeyframes: sequenceKeyframes.map((keyframe) => ({
            fileName: keyframe.fileName,
            nodePath: keyframe.nodePath,
            key: keyframe.fieldKey,
            frame: keyframe.sourceFrame,
            value: JSON.stringify(keyframe.value),
            schema: keyframe.schema,
        })),
        effectKeyframes: effectKeyframes.map((keyframe) => ({
            fileName: keyframe.fileName,
            sequenceNodePath: keyframe.nodePath,
            effectIndex: keyframe.effectIndex,
            key: keyframe.fieldKey,
            frame: keyframe.sourceFrame,
            value: JSON.stringify(keyframe.value),
            schema: keyframe.schema,
        })),
        clientId,
    }).then(() => undefined);
};
exports.callAddKeyframes = callAddKeyframes;
const callAddEffectKeyframe = ({ fileName, nodePath, effectIndex, fieldKey, sourceFrame, value, schema, setPropStatuses, clientId, }) => {
    return (0, save_prop_queue_1.enqueueSavePropChange)({
        nodePath,
        setPropStatuses,
        applyOptimistic: (prev) => (0, studio_shared_1.optimisticAddEffectKeyframe)({
            previous: prev,
            effectIndex,
            fieldKey,
            frame: sourceFrame,
            value,
            schema,
        }),
        applyServerResponse: (prev, response) => (0, apply_effect_response_to_prop_statuses_1.applyEffectResponseToPropStatuses)({ previous: prev, response }),
        apiCall: () => (0, call_api_1.callApi)('/api/add-effect-keyframe', {
            fileName,
            sequenceNodePath: nodePath,
            effectIndex,
            key: fieldKey,
            frame: sourceFrame,
            value: JSON.stringify(value),
            schema,
            clientId,
        }),
        errorLabel: 'Could not add keyframe',
    });
};
exports.callAddEffectKeyframe = callAddEffectKeyframe;
