// @remotion/studio/dist/components/Timeline/delete-selected-keyframe.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.deleteSelectedKeyframes = exports.deleteSelectedKeyframe = void 0;
const call_delete_keyframe_1 = require("./call-delete-keyframe");
const find_track_for_node_path_info_1 = require("./find-track-for-node-path-info");
const parse_keyframe_field_from_node_path_1 = require("./parse-keyframe-field-from-node-path");
const getSelectedKeyframeDeletion = ({ nodePathInfo, frame, sequences, overrideIdsToNodePaths, }) => {
    var _a, _b;
    const field = (0, parse_keyframe_field_from_node_path_1.parseKeyframeFieldFromNodePath)(nodePathInfo.auxiliaryKeys);
    if (field === null) {
        return null;
    }
    const track = (0, find_track_for_node_path_info_1.findTrackForNodePathInfo)({
        sequences,
        overrideIdsToNodePaths,
        nodePathInfo,
    });
    const sequence = (_a = track === null || track === void 0 ? void 0 : track.sequence) !== null && _a !== void 0 ? _a : null;
    if (!(sequence === null || sequence === void 0 ? void 0 : sequence.controls)) {
        return null;
    }
    const sourceFrame = frame - ((_b = track === null || track === void 0 ? void 0 : track.keyframeDisplayOffset) !== null && _b !== void 0 ? _b : 0);
    const nodePath = nodePathInfo.sequenceSubscriptionKey;
    const fileName = nodePath.absolutePath;
    if (field.type === 'effect') {
        const effect = sequence.effects[field.effectIndex];
        if (!effect) {
            return null;
        }
        return {
            type: 'effect',
            fileName,
            nodePath,
            effectIndex: field.effectIndex,
            fieldKey: field.fieldKey,
            sourceFrame,
            schema: effect.schema,
        };
    }
    return {
        type: 'sequence',
        fileName,
        nodePath,
        fieldKey: field.fieldKey,
        sourceFrame,
        schema: sequence.controls.schema,
    };
};
const deleteSelectedKeyframe = ({ nodePathInfo, frame, sequences, overrideIdsToNodePaths, setPropStatuses, clientId, }) => {
    const deletion = getSelectedKeyframeDeletion({
        nodePathInfo,
        frame,
        sequences,
        overrideIdsToNodePaths,
    });
    if (deletion === null) {
        return null;
    }
    if (deletion.type === 'effect') {
        return (0, call_delete_keyframe_1.callDeleteEffectKeyframe)({
            ...deletion,
            setPropStatuses,
            clientId,
        });
    }
    return (0, call_delete_keyframe_1.callDeleteSequenceKeyframe)({
        ...deletion,
        setPropStatuses,
        clientId,
    });
};
exports.deleteSelectedKeyframe = deleteSelectedKeyframe;
const deleteSelectedKeyframes = ({ keyframes, sequences, overrideIdsToNodePaths, setPropStatuses, clientId, }) => {
    const deletions = keyframes
        .map((keyframe) => getSelectedKeyframeDeletion({
        nodePathInfo: keyframe.nodePathInfo,
        frame: keyframe.frame,
        sequences,
        overrideIdsToNodePaths,
    }))
        .filter((deletion) => deletion !== null);
    if (deletions.length === 0) {
        return null;
    }
    return (0, call_delete_keyframe_1.callDeleteKeyframes)({
        sequenceKeyframes: deletions.filter((deletion) => deletion.type === 'sequence'),
        effectKeyframes: deletions.filter((deletion) => deletion.type === 'effect'),
        setPropStatuses,
        clientId,
    });
};
exports.deleteSelectedKeyframes = deleteSelectedKeyframes;
