// @remotion/studio/dist/components/Timeline/duplicate-selected-timeline-item.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.duplicateSelectedTimelineItems = exports.isDuplicatableEffectSelection = exports.isDuplicatableSequenceRowSelection = exports.duplicateSequencesFromSource = void 0;
const call_api_1 = require("../call-api");
const NotificationCenter_1 = require("../Notifications/NotificationCenter");
const confirmDuplicatingProgrammaticallyDuplicatedSequences = (nodePathInfos, confirm) => {
    if (nodePathInfos.length === 0) {
        return Promise.resolve(true);
    }
    if (nodePathInfos.length === 1) {
        const [nodePathInfo] = nodePathInfos;
        return confirm({
            title: 'Duplicate sequence?',
            message: 'This sequence is programmatically duplicated ' +
                nodePathInfo.numberOfSequencesWithThisNodePath +
                ' times in the code. Duplicating inserts another copy. Continue?',
            confirmLabel: 'Duplicate',
        });
    }
    return confirm({
        title: 'Duplicate sequences?',
        message: nodePathInfos.length +
            ' selected sequences are programmatically duplicated in the code. Duplicating inserts another copy of each. Continue?',
        confirmLabel: 'Duplicate',
    });
};
const duplicateSequence = (nodePathInfo) => {
    const nodePath = nodePathInfo.sequenceSubscriptionKey;
    return (0, call_api_1.callApi)('/api/duplicate-jsx-node', {
        fileName: nodePath.absolutePath,
        nodePath: nodePath.nodePath,
    });
};
const duplicateSequencesFromSource = (nodePathInfos, confirm) => {
    const programmaticallyDuplicated = nodePathInfos.filter((nodePathInfo) => nodePathInfo.numberOfSequencesWithThisNodePath > 1);
    const regular = nodePathInfos.filter((nodePathInfo) => nodePathInfo.numberOfSequencesWithThisNodePath <= 1);
    return confirmDuplicatingProgrammaticallyDuplicatedSequences(programmaticallyDuplicated, confirm).then((shouldDuplicateProgrammaticSequences) => {
        const toDuplicate = [...regular];
        if (shouldDuplicateProgrammaticSequences) {
            toDuplicate.push(...programmaticallyDuplicated);
        }
        if (toDuplicate.length === 0) {
            return Promise.resolve();
        }
        return Promise.all(toDuplicate.map(duplicateSequence))
            .then((results) => {
            const failedResult = results.find((result) => !result.success);
            if (failedResult && !failedResult.success) {
                (0, NotificationCenter_1.showNotification)(failedResult.reason, 4000);
                return;
            }
            (0, NotificationCenter_1.showNotification)(toDuplicate.length === 1
                ? 'Duplicated sequence in source file'
                : 'Duplicated sequences in source files', 2000);
        })
            .catch((err) => {
            (0, NotificationCenter_1.showNotification)(err.message, 4000);
        });
    });
};
exports.duplicateSequencesFromSource = duplicateSequencesFromSource;
const isDuplicatableSequenceRowSelection = (selection) => selection.type === 'sequence';
exports.isDuplicatableSequenceRowSelection = isDuplicatableSequenceRowSelection;
const isDuplicatableEffectSelection = (selection) => selection.type === 'sequence-effect';
exports.isDuplicatableEffectSelection = isDuplicatableEffectSelection;
const duplicateEffectsFromSource = (effects) => {
    return (0, call_api_1.callApi)('/api/duplicate-effect', effects.map((effect) => {
        const nodePath = effect.nodePathInfo.sequenceSubscriptionKey;
        return {
            fileName: nodePath.absolutePath,
            sequenceNodePath: nodePath,
            effectIndex: effect.i,
        };
    }))
        .then((result) => {
        if (result.success) {
            (0, NotificationCenter_1.showNotification)(effects.length === 1
                ? 'Duplicated effect in source file'
                : 'Duplicated effects in source files', 2000);
        }
        else {
            (0, NotificationCenter_1.showNotification)(result.reason, 4000);
        }
    })
        .catch((err) => {
        (0, NotificationCenter_1.showNotification)(err.message, 4000);
    });
};
const duplicateSelectedTimelineItems = ({ selections, confirm, }) => {
    const sequenceSelections = selections.filter(exports.isDuplicatableSequenceRowSelection);
    if (sequenceSelections.length > 0) {
        return (0, exports.duplicateSequencesFromSource)(sequenceSelections.map((selection) => selection.nodePathInfo), confirm);
    }
    const effectSelections = selections.filter(exports.isDuplicatableEffectSelection);
    if (effectSelections.length === 0) {
        return null;
    }
    return duplicateEffectsFromSource(effectSelections);
};
exports.duplicateSelectedTimelineItems = duplicateSelectedTimelineItems;
