// @remotion/studio/dist/components/Timeline/get-node-keyframes.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getNodeHasKeyframes = exports.getNodeKeyframes = void 0;
const remotion_1 = require("remotion");
const get_timeline_keyframes_1 = require("./get-timeline-keyframes");
const hasOverride = (overrides, key) => Object.prototype.hasOwnProperty.call(overrides, key);
const withDragOverrideKeyframe = ({ propStatus, keyframeDisplayOffset, timelinePosition, dragOverrideValue, hasDragOverride, }) => {
    if ((dragOverrideValue === null || dragOverrideValue === void 0 ? void 0 : dragOverrideValue.type) === 'keyframed') {
        return (0, get_timeline_keyframes_1.getTimelineKeyframes)(dragOverrideValue.status, keyframeDisplayOffset);
    }
    const keyframes = (0, get_timeline_keyframes_1.getTimelineKeyframes)(propStatus, keyframeDisplayOffset);
    if (!hasDragOverride || (propStatus === null || propStatus === void 0 ? void 0 : propStatus.status) !== 'keyframed') {
        return keyframes;
    }
    const dragKeyframe = {
        frame: timelinePosition,
        value: dragOverrideValue === null || dragOverrideValue === void 0 ? void 0 : dragOverrideValue.value,
    };
    const existingIndex = keyframes.findIndex((keyframe) => keyframe.frame === timelinePosition);
    if (existingIndex !== -1) {
        return keyframes.map((keyframe, index) => index === existingIndex ? dragKeyframe : keyframe);
    }
    return [...keyframes, dragKeyframe].sort((first, second) => first.frame - second.frame);
};
const getNodeKeyframes = ({ node, nodePath, propStatuses, keyframeDisplayOffset, getDragOverrides, getEffectDragOverrides, timelinePosition, }) => {
    var _a, _b;
    if (node.kind !== 'field' || node.field === null) {
        return [];
    }
    if (node.field.kind === 'sequence-field') {
        const dragOverrides = getDragOverrides(nodePath);
        return withDragOverrideKeyframe({
            propStatus: (_a = remotion_1.Internals.getPropStatusesCtx(propStatuses, nodePath)) === null || _a === void 0 ? void 0 : _a[node.field.key],
            keyframeDisplayOffset,
            timelinePosition,
            dragOverrideValue: dragOverrides[node.field.key],
            hasDragOverride: hasOverride(dragOverrides, node.field.key),
        });
    }
    const effectStatus = remotion_1.Internals.getEffectPropStatusesCtx({
        propStatuses,
        nodePath,
        effectIndex: node.field.effectIndex,
    });
    const effectDragOverrides = getEffectDragOverrides(nodePath, node.field.effectIndex);
    return withDragOverrideKeyframe({
        propStatus: effectStatus.type === 'can-update-effect'
            ? (_b = effectStatus.props) === null || _b === void 0 ? void 0 : _b[node.field.key]
            : null,
        keyframeDisplayOffset,
        timelinePosition,
        dragOverrideValue: effectDragOverrides[node.field.key],
        hasDragOverride: hasOverride(effectDragOverrides, node.field.key),
    });
};
exports.getNodeKeyframes = getNodeKeyframes;
const getNodeHasKeyframes = ({ node, nodePath, propStatuses, getDragOverrides, getEffectDragOverrides, }) => {
    var _a, _b, _c, _d, _e, _f;
    if (node.kind !== 'field' || node.field === null) {
        return false;
    }
    if (node.field.kind === 'sequence-field') {
        const dragOverrides = getDragOverrides(nodePath);
        if (((_a = dragOverrides[node.field.key]) === null || _a === void 0 ? void 0 : _a.type) === 'keyframed') {
            return true;
        }
        return (((_c = (_b = remotion_1.Internals.getPropStatusesCtx(propStatuses, nodePath)) === null || _b === void 0 ? void 0 : _b[node.field.key]) === null || _c === void 0 ? void 0 : _c.status) === 'keyframed');
    }
    const effectDragOverrides = getEffectDragOverrides(nodePath, node.field.effectIndex);
    if (((_d = effectDragOverrides[node.field.key]) === null || _d === void 0 ? void 0 : _d.type) === 'keyframed') {
        return true;
    }
    const effectStatus = remotion_1.Internals.getEffectPropStatusesCtx({
        propStatuses,
        nodePath,
        effectIndex: node.field.effectIndex,
    });
    return (effectStatus.type === 'can-update-effect' &&
        ((_f = (_e = effectStatus.props) === null || _e === void 0 ? void 0 : _e[node.field.key]) === null || _f === void 0 ? void 0 : _f.status) === 'keyframed');
};
exports.getNodeHasKeyframes = getNodeHasKeyframes;
