// @remotion/studio/dist/components/Timeline/KeyframeSettingsModal.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.KeyframeSettingsModal = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const react_1 = require("react");
const remotion_1 = require("remotion");
const client_id_1 = require("../../helpers/client-id");
const colors_1 = require("../../helpers/colors");
const modals_1 = require("../../state/modals");
const Button_1 = require("../Button");
const layout_1 = require("../layout");
const ModalButton_1 = require("../ModalButton");
const ModalFooter_1 = require("../ModalFooter");
const ModalHeader_1 = require("../ModalHeader");
const ComboBox_1 = require("../NewComposition/ComboBox");
const DismissableModal_1 = require("../NewComposition/DismissableModal");
const InputDragger_1 = require("../NewComposition/InputDragger");
const call_update_keyframe_settings_1 = require("./call-update-keyframe-settings");
const container = {
    padding: 16,
    width: 520,
};
const row = {
    alignItems: 'center',
    display: 'flex',
    justifyContent: 'space-between',
    marginBottom: 12,
};
const label = {
    fontSize: 13,
};
const helperText = {
    color: 'rgba(255, 255, 255, 0.6)',
    fontSize: 12,
    lineHeight: 1.4,
    marginTop: 4,
};
const comboStyle = {
    minWidth: 150,
};
const posterizeInputStyle = {
    backgroundColor: colors_1.INPUT_BACKGROUND,
    borderRadius: 4,
    minWidth: 64,
    textAlign: 'right',
};
const extrapolateOptions = [
    'extend',
    'clamp',
    'identity',
    'wrap',
];
const labelForExtrapolate = (value) => value[0].toUpperCase() + value.slice(1);
const getExtrapolateValues = (onSelect) => {
    return extrapolateOptions.map((value) => ({
        type: 'item',
        id: value,
        keyHint: null,
        label: labelForExtrapolate(value),
        leftItem: null,
        disabled: false,
        onClick: () => onSelect(value),
        quickSwitcherLabel: null,
        subMenu: null,
        value,
    }));
};
const KeyframeSettingsModal = ({ state }) => {
    var _a;
    const { setSelectedModal } = (0, react_1.useContext)(modals_1.ModalsContext);
    const { setPropStatuses } = (0, react_1.useContext)(remotion_1.Internals.VisualModeSettersContext);
    const { previewServerState } = (0, react_1.useContext)(client_id_1.StudioServerConnectionCtx);
    const [left, setLeft] = (0, react_1.useState)(state.status.clamping.left);
    const [right, setRight] = (0, react_1.useState)(state.status.clamping.right);
    const [posterize, setPosterize] = (0, react_1.useState)((_a = state.status.posterize) !== null && _a !== void 0 ? _a : 0);
    const [saving, setSaving] = (0, react_1.useState)(false);
    const canEditClamping = state.status.interpolationFunction === 'interpolate';
    const close = (0, react_1.useCallback)(() => {
        setSelectedModal(null);
    }, [setSelectedModal]);
    const leftOptions = (0, react_1.useMemo)(() => getExtrapolateValues(setLeft), []);
    const rightOptions = (0, react_1.useMemo)(() => getExtrapolateValues(setRight), []);
    const onPosterizeChange = (0, react_1.useCallback)((value) => {
        setPosterize(Math.max(0, Math.round(value)));
    }, []);
    const posterizeFormatter = (0, react_1.useCallback)((value) => {
        const numericValue = Number(value);
        return String(Math.round(numericValue));
    }, []);
    const onSave = (0, react_1.useCallback)(() => {
        if (previewServerState.type !== 'connected') {
            return;
        }
        const settings = {
            type: 'settings',
            clamping: canEditClamping ? { left, right } : undefined,
            posterize: posterize <= 0 ? undefined : posterize,
        };
        setSaving(true);
        const promise = state.effectIndex === null
            ? (0, call_update_keyframe_settings_1.callUpdateSequenceKeyframeSettings)({
                fileName: state.fileName,
                nodePath: state.nodePath,
                fieldKey: state.fieldKey,
                settings,
                schema: state.schema,
                setPropStatuses,
                clientId: previewServerState.clientId,
            })
            : (0, call_update_keyframe_settings_1.callUpdateEffectKeyframeSettings)({
                fileName: state.fileName,
                nodePath: state.nodePath,
                effectIndex: state.effectIndex,
                fieldKey: state.fieldKey,
                settings,
                schema: state.schema,
                setPropStatuses,
                clientId: previewServerState.clientId,
            });
        promise.then(close).finally(() => {
            setSaving(false);
        });
    }, [
        canEditClamping,
        close,
        left,
        posterize,
        previewServerState,
        right,
        setPropStatuses,
        state,
    ]);
    const saveDisabled = saving || previewServerState.type !== 'connected';
    return (jsx_runtime_1.jsxs(DismissableModal_1.DismissableModal, { children: [
            jsx_runtime_1.jsx(ModalHeader_1.ModalHeader, { title: `Keyframe settings: ${state.fieldLabel}` }), jsx_runtime_1.jsxs("div", { style: container, children: [canEditClamping ? (jsx_runtime_1.jsxs(jsx_runtime_1.Fragment, { children: [
                            jsx_runtime_1.jsxs("div", { style: row, children: [
                                    jsx_runtime_1.jsx("div", { style: label, children: "Extrapolate left" }), jsx_runtime_1.jsx(ComboBox_1.Combobox, { values: leftOptions, selectedId: left, title: "Extrapolate left", style: comboStyle })
                                ] }), jsx_runtime_1.jsxs("div", { style: row, children: [
                                    jsx_runtime_1.jsx("div", { style: label, children: "Extrapolate right" }), jsx_runtime_1.jsx(ComboBox_1.Combobox, { values: rightOptions, selectedId: right, title: "Extrapolate right", style: comboStyle })
                                ] })
                        ] })) : null, jsx_runtime_1.jsxs("div", { style: row, children: [
                            jsx_runtime_1.jsxs("div", { children: [
                                    jsx_runtime_1.jsx("div", { style: label, children: "Posterize" }), jsx_runtime_1.jsx("div", { style: helperText, children: "Use 0 to turn posterization off." })
                                ] }), jsx_runtime_1.jsx(InputDragger_1.InputDragger, { type: "number", value: posterize, status: "ok", onValueChange: onPosterizeChange, onValueChangeEnd: onPosterizeChange, onTextChange: () => undefined, min: 0, step: 1, formatter: posterizeFormatter, rightAlign: true, style: posterizeInputStyle })
                        ] })
                ] }), jsx_runtime_1.jsx(ModalFooter_1.ModalFooterContainer, { children: jsx_runtime_1.jsxs(layout_1.Row, { justify: "flex-end", align: "center", children: [
                        jsx_runtime_1.jsx(Button_1.Button, { onClick: close, children: "Cancel" }), jsx_runtime_1.jsx(layout_1.Spacing, { x: 1 }), jsx_runtime_1.jsx(ModalButton_1.ModalButton, { onClick: onSave, disabled: saveDisabled, children: "Save" })
                    ] }) })
        ] }));
};
exports.KeyframeSettingsModal = KeyframeSettingsModal;
