// @remotion/studio/dist/components/Timeline/reset-selected-timeline-props.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.resetSelectedTimelineProps = exports.getTimelinePropResetTargets = void 0;
const studio_shared_1 = require("@remotion/studio-shared");
const remotion_1 = require("remotion");
const find_track_for_node_path_info_1 = require("./find-track-for-node-path-info");
const save_effect_prop_1 = require("./save-effect-prop");
const save_sequence_prop_1 = require("./save-sequence-prop");
const isPropResetSelection = (selection) => selection.type === 'sequence-prop' ||
    selection.type === 'sequence-effect-prop';
function assertPropResetSelections(selections) {
    for (const selection of selections) {
        if (!isPropResetSelection(selection)) {
            throw new Error(`Assertion failed: Cannot reset timeline selection of type ${selection.type}`);
        }
    }
}
const isVisibleFieldSchema = (fieldSchema) => fieldSchema !== undefined && fieldSchema.type !== 'hidden';
const isNonDefaultCodeValue = ({ propStatus, defaultValue, }) => JSON.stringify(propStatus !== null && propStatus !== void 0 ? propStatus : defaultValue) !== JSON.stringify(defaultValue);
const isResettablePropStatus = ({ propStatus, defaultValue, }) => {
    if (!propStatus) {
        return false;
    }
    if (defaultValue === undefined) {
        return false;
    }
    if (propStatus.status === 'keyframed') {
        return true;
    }
    if (propStatus.status === 'computed') {
        return false;
    }
    return isNonDefaultCodeValue({
        propStatus: propStatus.codeValue,
        defaultValue,
    });
};
const getDefaultValue = (fieldSchema) => fieldSchema.default !== undefined
    ? JSON.stringify(fieldSchema.default)
    : null;
const getActiveFieldSchema = ({ schema, key, resolveValue, }) => {
    return remotion_1.Internals.flattenActiveSchema(schema, resolveValue)[key];
};
const getTimelinePropResetTargets = ({ selections, sequences, overrideIdsToNodePaths, propStatuses, }) => {
    var _a;
    const propSelections = selections.filter(isPropResetSelection);
    if (propSelections.length === 0) {
        return null;
    }
    if (propSelections.length !== selections.length) {
        return null;
    }
    assertPropResetSelections(selections);
    const resetTargets = [];
    for (const selection of selections) {
        const track = (0, find_track_for_node_path_info_1.findTrackForNodePathInfo)({
            sequences,
            overrideIdsToNodePaths,
            nodePathInfo: selection.nodePathInfo,
        });
        const sequence = (_a = track === null || track === void 0 ? void 0 : track.sequence) !== null && _a !== void 0 ? _a : null;
        if (!sequence) {
            continue;
        }
        const nodePath = selection.nodePathInfo.sequenceSubscriptionKey;
        if (selection.type === 'sequence-prop') {
            if (!sequence.controls) {
                continue;
            }
            const sequencePropStatus = remotion_1.Internals.getPropStatusesCtx(propStatuses, nodePath);
            const { merged: sequenceValuesDotNotation } = remotion_1.Internals.computeEffectiveSchemaValuesDotNotation({
                schema: sequence.controls.schema,
                currentValue: sequence.controls.currentRuntimeValueDotNotation,
                overrideValues: {},
                propStatus: sequencePropStatus,
                frame: null,
            });
            const sequenceFieldSchema = getActiveFieldSchema({
                schema: sequence.controls.schema,
                key: selection.key,
                resolveValue: (key) => sequenceValuesDotNotation[key],
            });
            const selectedPropStatus = sequencePropStatus === null || sequencePropStatus === void 0 ? void 0 : sequencePropStatus[selection.key];
            if (!isVisibleFieldSchema(sequenceFieldSchema) ||
                !isResettablePropStatus({
                    propStatus: selectedPropStatus,
                    defaultValue: sequenceFieldSchema.default,
                })) {
                continue;
            }
            resetTargets.push({
                type: 'sequence-prop',
                fileName: nodePath.absolutePath,
                nodePath,
                fieldKey: selection.key,
                value: sequenceFieldSchema.default,
                defaultValue: getDefaultValue(sequenceFieldSchema),
                schema: sequence.controls.schema,
            });
            continue;
        }
        const effect = sequence.effects[selection.i];
        const field = effect
            ? (0, studio_shared_1.getEffectFieldsToShow)({
                effect,
                effectIndex: selection.i,
                nodePath,
                propStatuses,
                getEffectDragOverrides: () => ({}),
            }).find((candidate) => candidate.key === selection.key)
            : null;
        const fieldSchema = field === null || field === void 0 ? void 0 : field.fieldSchema;
        const effectStatus = remotion_1.Internals.getEffectPropStatusesCtx({
            propStatuses,
            nodePath,
            effectIndex: selection.i,
        });
        const propStatus = effectStatus.type === 'can-update-effect'
            ? effectStatus.props[selection.key]
            : null;
        if (!effect ||
            !isVisibleFieldSchema(fieldSchema) ||
            !isResettablePropStatus({
                propStatus,
                defaultValue: fieldSchema.default,
            })) {
            continue;
        }
        resetTargets.push({
            type: 'effect-prop',
            fileName: nodePath.absolutePath,
            nodePath,
            effectIndex: selection.i,
            fieldKey: selection.key,
            value: fieldSchema.default,
            defaultValue: getDefaultValue(fieldSchema),
            schema: effect.schema,
        });
    }
    return resetTargets;
};
exports.getTimelinePropResetTargets = getTimelinePropResetTargets;
const resetSelectedTimelineProps = ({ selections, sequences, overrideIdsToNodePaths, propStatuses, setPropStatuses, clientId, }) => {
    const resetTargets = (0, exports.getTimelinePropResetTargets)({
        selections,
        sequences,
        overrideIdsToNodePaths,
        propStatuses,
    });
    if (resetTargets === null || resetTargets.length === 0) {
        return null;
    }
    const sequencePropTargets = resetTargets.filter((target) => target.type === 'sequence-prop');
    const effectPropTargets = resetTargets.filter((target) => target.type === 'effect-prop');
    const resetPromises = [];
    if (sequencePropTargets.length > 0) {
        resetPromises.push((0, save_sequence_prop_1.saveSequenceProps)({
            changes: sequencePropTargets.map((target) => ({
                fileName: target.fileName,
                nodePath: target.nodePath,
                fieldKey: target.fieldKey,
                value: target.value,
                defaultValue: target.defaultValue,
                schema: target.schema,
            })),
            setPropStatuses,
            clientId,
            undoLabel: sequencePropTargets.length > 1
                ? 'Reset selected sequence props'
                : 'Reset sequence prop',
            redoLabel: sequencePropTargets.length > 1
                ? 'Reapply selected sequence props'
                : 'Reapply sequence prop',
        }));
    }
    for (const target of effectPropTargets) {
        resetPromises.push((0, save_effect_prop_1.saveEffectProp)({
            type: 'value',
            fileName: target.fileName,
            nodePath: target.nodePath,
            effectIndex: target.effectIndex,
            fieldKey: target.fieldKey,
            value: target.value,
            defaultValue: target.defaultValue,
            schema: target.schema,
            setPropStatuses,
            clientId,
        }));
    }
    return Promise.all(resetPromises).then(() => undefined);
};
exports.resetSelectedTimelineProps = resetSelectedTimelineProps;
