// @remotion/studio/dist/components/Timeline/save-effect-prop.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.saveEffectProp = void 0;
const studio_shared_1 = require("@remotion/studio-shared");
const call_api_1 = require("../call-api");
const apply_effect_response_to_prop_statuses_1 = require("./apply-effect-response-to-prop-statuses");
const save_prop_queue_1 = require("./save-prop-queue");
const saveEffectProp = (input) => {
    const { fileName, nodePath, effectIndex, fieldKey, defaultValue, schema, setPropStatuses, clientId, } = input;
    return (0, save_prop_queue_1.enqueueSavePropChange)({
        nodePath,
        setPropStatuses,
        applyOptimistic: (prev) => input.type === 'effect-param'
            ? prev
            : (0, studio_shared_1.optimisticUpdateForEffectPropStatuses)({
                previous: prev,
                effectIndex,
                fieldKey,
                value: input.value,
                schema,
            }),
        applyServerResponse: (prev, response) => (0, apply_effect_response_to_prop_statuses_1.applyEffectResponseToPropStatuses)({ previous: prev, response }),
        apiCall: () => (0, call_api_1.callApi)('/api/save-effect-props', input.type === 'effect-param'
            ? {
                type: 'effect-param',
                fileName,
                sequenceNodePath: nodePath,
                effectIndex,
                key: fieldKey,
                effectParam: input.effectParam,
                defaultValue,
                schema,
                clientId,
            }
            : {
                type: 'value',
                fileName,
                sequenceNodePath: nodePath,
                effectIndex,
                key: fieldKey,
                value: JSON.stringify(input.value),
                defaultValue,
                schema,
                clientId,
            }),
        errorLabel: 'Could not save effect prop',
    });
};
exports.saveEffectProp = saveEffectProp;
