// @remotion/studio/dist/components/Timeline/save-sequence-prop.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.saveSequenceProps = void 0;
const studio_shared_1 = require("@remotion/studio-shared");
const call_api_1 = require("../call-api");
const save_prop_queue_1 = require("./save-prop-queue");
const saveSequenceProps = ({ changes, setPropStatuses, clientId, undoLabel, redoLabel, }) => {
    if (changes.length === 0) {
        return Promise.resolve();
    }
    if (changes.length === 1) {
        const change = changes[0];
        if (change === undefined) {
            throw new Error('Expected a sequence prop change');
        }
        return (0, save_prop_queue_1.enqueueSavePropChange)({
            nodePath: change.nodePath,
            setPropStatuses,
            applyOptimistic: (prev) => (0, studio_shared_1.optimisticUpdateForPropStatuses)({
                previous: prev,
                fieldKey: change.fieldKey,
                value: change.value,
                schema: change.schema,
            }),
            apiCall: () => (0, call_api_1.callApi)('/api/save-sequence-props', {
                edits: [
                    {
                        fileName: change.fileName,
                        nodePath: change.nodePath,
                        key: change.fieldKey,
                        value: JSON.stringify(change.value),
                        defaultValue: change.defaultValue,
                        schema: change.schema,
                    },
                ],
                clientId,
                undoLabel,
                redoLabel,
            }),
            errorLabel: 'Could not save sequence prop',
        });
    }
    for (const change of changes) {
        setPropStatuses(change.nodePath, (prev) => (0, studio_shared_1.optimisticUpdateForPropStatuses)({
            previous: prev,
            fieldKey: change.fieldKey,
            value: change.value,
            schema: change.schema,
        }));
    }
    return (0, call_api_1.callApi)('/api/save-sequence-props', {
        edits: changes.map((change) => {
            return {
                fileName: change.fileName,
                nodePath: change.nodePath,
                key: change.fieldKey,
                value: JSON.stringify(change.value),
                defaultValue: change.defaultValue,
                schema: change.schema,
            };
        }),
        clientId,
        undoLabel,
        redoLabel,
    }).then(() => undefined);
};
exports.saveSequenceProps = saveSequenceProps;
