// @remotion/studio/dist/components/Timeline/TimelineFieldLabel.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TimelineFieldLabel = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const react_1 = require("react");
const timeline_field_row_layout_1 = require("./timeline-field-row-layout");
const TimelineSelection_1 = require("./TimelineSelection");
const fieldNameBase = {
    fontSize: 12,
    color: 'rgba(255, 255, 255, 0.8)',
    userSelect: 'none',
    whiteSpace: 'nowrap',
    overflow: 'hidden',
    textOverflow: 'ellipsis',
    flex: 1,
    minWidth: 0,
};
const TimelineFieldLabel = ({ rowDepth, selected, label }) => {
    const labelRowStyle = (0, react_1.useMemo)(() => ({
        ...(0, timeline_field_row_layout_1.getTimelineFieldLabelRowStyle)(rowDepth),
        ...(0, TimelineSelection_1.getTimelineSelectedLabelStyle)(selected, true),
        alignSelf: 'stretch',
    }), [rowDepth, selected]);
    const fieldNameStyle = (0, react_1.useMemo)(() => ({
        ...fieldNameBase,
        color: (0, TimelineSelection_1.getTimelineColor)(selected, true),
    }), [selected]);
    return (jsx_runtime_1.jsx("div", { style: labelRowStyle, children: jsx_runtime_1.jsx("span", { style: fieldNameStyle, title: label, children: label }) }));
};
exports.TimelineFieldLabel = TimelineFieldLabel;
