// @remotion/studio/dist/components/Timeline/TimelineKeyframeEasingLine.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.TimelineKeyframeEasingLine = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const studio_shared_1 = require("@remotion/studio-shared");
const react_1 = __importStar(require("react"));
const remotion_1 = require("remotion");
const client_id_1 = require("../../helpers/client-id");
const colors_1 = require("../../helpers/colors");
const get_left_of_timeline_slider_1 = require("../../helpers/get-left-of-timeline-slider");
const timeline_layout_1 = require("../../helpers/timeline-layout");
const ContextMenu_1 = require("../ContextMenu");
const TimelineSelection_1 = require("./TimelineSelection");
const TimelineWidthProvider_1 = require("./TimelineWidthProvider");
const update_selected_easing_1 = require("./update-selected-easing");
const use_timeline_keyframe_drag_1 = require("./use-timeline-keyframe-drag");
const hitTargetHeight = 12;
const lineHeight = 2;
const easingLineButton = {
    background: 'none',
    border: 'none',
    height: hitTargetHeight,
    padding: 0,
    position: 'absolute',
    transform: 'translateY(-50%)',
};
const easingLine = {
    backgroundColor: 'rgba(255, 255, 255, 0.1)',
    borderRadius: lineHeight / 2,
    height: lineHeight,
    left: 0,
    position: 'absolute',
    right: 0,
    top: '50%',
    transform: 'translateY(-50%)',
};
const TimelineKeyframeEasingLineUnmemoized = ({ fromFrame, toFrame, rowHeight, nodePathInfo, segmentIndex }) => {
    const buttonRef = (0, react_1.useRef)(null);
    const videoConfig = (0, remotion_1.useVideoConfig)();
    const timelineWidth = (0, react_1.useContext)(TimelineWidthProvider_1.TimelineWidthContext);
    const { selected, onSelect, selectable, selectionItem } = (0, TimelineSelection_1.useTimelineEasingSelection)({
        nodePathInfo,
        fromFrame,
        toFrame,
        segmentIndex,
    });
    (0, TimelineSelection_1.useTimelineMarqueeSelectableItem)(selectionItem, buttonRef);
    const { previewServerState } = (0, react_1.useContext)(client_id_1.StudioServerConnectionCtx);
    const sequencesRef = (0, react_1.useContext)(remotion_1.Internals.SequenceManagerRefContext);
    const propStatusesRef = (0, react_1.useContext)(remotion_1.Internals.VisualModePropStatusesRefContext);
    const { setPropStatuses } = (0, react_1.useContext)(remotion_1.Internals.VisualModeSettersContext);
    const { overrideIdToNodePathMappings } = (0, react_1.useContext)(remotion_1.Internals.OverrideIdsToNodePathsGettersContext);
    const currentSelection = (0, TimelineSelection_1.useCurrentTimelineSelectionStateAsRef)();
    const getTargetSelections = (0, react_1.useCallback)(() => {
        const selectedEasings = (0, update_selected_easing_1.getEasingSelections)(currentSelection.current.selectedItems);
        return selected ? selectedEasings : [selectionItem];
    }, [currentSelection, selected, selectionItem]);
    const updateEasing = (0, react_1.useCallback)((easing) => {
        if (previewServerState.type !== 'connected') {
            return;
        }
        const promise = (0, update_selected_easing_1.updateSelectedTimelineEasings)({
            selections: getTargetSelections(),
            sequences: sequencesRef.current,
            overrideIdsToNodePaths: overrideIdToNodePathMappings,
            propStatuses: propStatusesRef.current,
            setPropStatuses,
            clientId: previewServerState.clientId,
            easing,
        });
        promise === null || promise === void 0 ? void 0 : promise.catch(() => undefined);
    }, [
        getTargetSelections,
        overrideIdToNodePathMappings,
        previewServerState,
        propStatusesRef,
        sequencesRef,
        setPropStatuses,
    ]);
    const contextMenuValues = (0, react_1.useMemo)(() => {
        return [
            {
                type: 'item',
                id: 'linear',
                keyHint: null,
                label: 'Linear',
                leftItem: null,
                disabled: previewServerState.type !== 'connected',
                onClick: () => updateEasing(studio_shared_1.LINEAR_KEYFRAME_EASING),
                quickSwitcherLabel: null,
                subMenu: null,
                value: 'linear',
            },
            ...studio_shared_1.KEYFRAME_EASING_PRESETS.map((preset) => ({
                type: 'item',
                id: preset.id,
                keyHint: null,
                label: preset.label,
                leftItem: null,
                disabled: previewServerState.type !== 'connected',
                onClick: () => updateEasing(preset.easing),
                quickSwitcherLabel: null,
                subMenu: null,
                value: preset.id,
            })),
        ];
    }, [previewServerState.type, updateEasing]);
    const onOpenContextMenu = (0, react_1.useCallback)((event) => {
        if (!selectable) {
            return false;
        }
        if (!selected) {
            onSelect({
                shiftKey: event.shiftKey,
                toggleKey: event.metaKey || event.ctrlKey,
            });
        }
        return contextMenuValues;
    }, [contextMenuValues, onSelect, selectable, selected]);
    const style = (0, react_1.useMemo)(() => {
        if (timelineWidth === null) {
            return null;
        }
        const fromLeft = (0, get_left_of_timeline_slider_1.getXPositionOfItemInTimelineImperatively)(fromFrame, videoConfig.durationInFrames, timelineWidth) - timeline_layout_1.TIMELINE_PADDING;
        const toLeft = (0, get_left_of_timeline_slider_1.getXPositionOfItemInTimelineImperatively)(toFrame, videoConfig.durationInFrames, timelineWidth) - timeline_layout_1.TIMELINE_PADDING;
        const left = Math.min(fromLeft, toLeft);
        const width = Math.abs(toLeft - fromLeft);
        if (width === 0) {
            return null;
        }
        return {
            ...easingLineButton,
            left,
            pointerEvents: selectable ? 'auto' : 'none',
            top: rowHeight / 2,
            width,
        };
    }, [
        fromFrame,
        rowHeight,
        selectable,
        timelineWidth,
        toFrame,
        videoConfig.durationInFrames,
    ]);
    const lineStyle = (0, react_1.useMemo)(() => ({
        ...easingLine,
        outline: selected ? `1px solid ${colors_1.BLUE}` : undefined,
    }), [selected]);
    const onPointerDown = (0, use_timeline_keyframe_drag_1.useTimelineEasingKeyframeDrag)({
        onSelect,
        selectable,
        selected,
        selectionItem,
    });
    if (style === null) {
        return null;
    }
    return (jsx_runtime_1.jsxs(jsx_runtime_1.Fragment, { children: [
            jsx_runtime_1.jsx("button", { ref: buttonRef, [TimelineSelection_1.TIMELINE_MARQUEE_ITEM_ATTR]: true, type: "button", style: style, title: `Easing from frame ${fromFrame} to ${toFrame}`, "aria-label": `Select easing from frame ${fromFrame} to ${toFrame}`, onPointerDown: selectable ? onPointerDown : undefined, children: jsx_runtime_1.jsx("div", { style: lineStyle }) }), jsx_runtime_1.jsx(ContextMenu_1.ContextMenuForTarget, { triggerRef: buttonRef, values: contextMenuValues, onOpen: onOpenContextMenu })
        ] }));
};
exports.TimelineKeyframeEasingLine = react_1.default.memo(TimelineKeyframeEasingLineUnmemoized);
