// @remotion/studio/dist/components/Timeline/TimelineEnumField.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TimelineEnumField = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const react_1 = require("react");
const ComboBox_1 = require("../NewComposition/ComboBox");
const comboboxStyle = {
    marginLeft: 8,
};
const TimelineEnumField = ({ field, propStatus, effectiveValue, onSave, onDragValueChange, onDragEnd, }) => {
    const { fieldSchema } = field;
    if (fieldSchema.type !== 'enum') {
        throw new Error('TimelineEnumField rendered for non-enum field');
    }
    const variantKeys = Object.keys(fieldSchema.variants);
    const current = String(effectiveValue !== null && effectiveValue !== void 0 ? effectiveValue : fieldSchema.default);
    const onSelect = (0, react_1.useCallback)((newValue) => {
        if (newValue === propStatus.codeValue) {
            return;
        }
        onDragValueChange(newValue);
        onSave(newValue).finally(() => {
            onDragEnd();
        });
    }, [propStatus, onSave, onDragValueChange, onDragEnd]);
    const items = (0, react_1.useMemo)(() => {
        return variantKeys.map((key) => ({
            type: 'item',
            id: key,
            value: key,
            label: key,
            onClick: () => onSelect(key),
            keyHint: null,
            leftItem: null,
            subMenu: null,
            quickSwitcherLabel: null,
            disabled: false,
        }));
    }, [variantKeys, onSelect]);
    return (jsx_runtime_1.jsx(ComboBox_1.Combobox, { small: true, title: field.key, selectedId: current, values: items, style: comboboxStyle }));
};
exports.TimelineEnumField = TimelineEnumField;
