// @remotion/studio/dist/components/Timeline/timeline-rotation-utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.serializeCssRotation = exports.parseCssRotationToDegrees = void 0;
const timeline_field_utils_1 = require("./timeline-field-utils");
const unitPattern = /^([+-]?(?:\d+\.?\d*|\.\d+))(deg|rad|turn|grad)$/;
const unitToDegrees = {
    deg: 1,
    rad: 180 / Math.PI,
    turn: 360,
    grad: 360 / 400,
};
const parseCssRotationToDegrees = (value) => {
    const match = value.trim().match(unitPattern);
    if (match) {
        return (0, timeline_field_utils_1.normalizeTimelineNumber)(Number(match[1]) * unitToDegrees[match[2]]);
    }
    if (typeof DOMMatrix === 'undefined') {
        return 0;
    }
    try {
        const m = new DOMMatrix(`rotate(${value})`);
        return (0, timeline_field_utils_1.normalizeTimelineNumber)(Math.atan2(m.b, m.a) * (180 / Math.PI));
    }
    catch (_a) {
        return 0;
    }
};
exports.parseCssRotationToDegrees = parseCssRotationToDegrees;
const serializeCssRotation = (value, decimalPlaces = 6) => {
    const factor = 10 ** decimalPlaces;
    const rounded = Math.round((0, timeline_field_utils_1.normalizeTimelineNumber)(value) * factor) / factor;
    return `${Object.is(rounded, -0) ? 0 : rounded}deg`;
};
exports.serializeCssRotation = serializeCssRotation;
