// @remotion/studio/dist/components/Timeline/TimelineScaleField.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TimelineScaleField = exports.getLinkedScale = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const react_1 = require("react");
const no_react_1 = require("remotion/no-react");
const colors_1 = require("../../helpers/colors");
const scale_lock_1 = require("../../state/scale-lock");
const InputDragger_1 = require("../NewComposition/InputDragger");
const timeline_field_display_utils_1 = require("./timeline-field-display-utils");
const timeline_field_utils_1 = require("./timeline-field-utils");
const TimelineLayerEye_1 = require("./TimelineLayerEye");
const leftDraggerStyle = {
    paddingLeft: 0,
};
const rightDraggerStyle = {
    paddingRight: 0,
};
const containerStyle = {
    alignItems: 'center',
    display: 'flex',
    gap: 4,
};
const toggleStyle = {
    ...TimelineLayerEye_1.timelineLayerIconContainer,
    border: 'none',
    color: colors_1.LIGHT_COLOR,
    cursor: 'default',
    marginRight: 0,
    padding: 0,
};
const linkIconStyle = {
    width: 14,
    height: 14,
    pointerEvents: 'none',
};
const gapStyle = {
    marginLeft: -6,
    marginRight: -6,
};
const clamp = (value, min, max) => {
    return Math.min(max, Math.max(min, value));
};
const getLinkedScale = ({ axis, newValue, baseX, baseY, min, max, }) => {
    const drivingBase = axis === 'x' ? baseX : baseY;
    const linkedBase = axis === 'x' ? baseY : baseX;
    if (drivingBase === 0 || linkedBase === 0) {
        const clamped = (0, timeline_field_utils_1.normalizeTimelineNumber)(clamp(newValue, min, max));
        return [clamped, clamped];
    }
    let factor = newValue / drivingBase;
    let driving = newValue;
    let linked = linkedBase * factor;
    const clampedLinked = clamp(linked, min, max);
    if (clampedLinked !== linked) {
        factor = clampedLinked / linkedBase;
        linked = clampedLinked;
        driving = drivingBase * factor;
    }
    const clampedDriving = clamp(driving, min, max);
    const normalizedDriving = (0, timeline_field_utils_1.normalizeTimelineNumber)(clampedDriving);
    const normalizedLinked = (0, timeline_field_utils_1.normalizeTimelineNumber)(linked);
    return axis === 'x'
        ? [normalizedDriving, normalizedLinked]
        : [normalizedLinked, normalizedDriving];
};
exports.getLinkedScale = getLinkedScale;
const valuesEqual = (left, right) => {
    return JSON.stringify(left) === JSON.stringify(right);
};
const LinkToggle = ({ linked, onToggle }) => {
    const onPointerDown = (0, react_1.useCallback)((e) => {
        if (e.button !== 0) {
            return;
        }
        e.stopPropagation();
        onToggle();
    }, [onToggle]);
    return (jsx_runtime_1.jsx("button", { type: "button", style: toggleStyle, onPointerDown: onPointerDown, title: linked ? 'Unlink scale axes' : 'Link scale axes', "aria-label": linked ? 'Unlink scale axes' : 'Link scale axes', children: linked ? (jsx_runtime_1.jsx("svg", { xmlns: "http://www.w3.org/2000/svg", viewBox: "0 0 640 640", style: linkIconStyle, children: jsx_runtime_1.jsx("path", { fill: "currentcolor", d: "M32 320C32 214 118 128 224 128L288 128L288 192L224 192C153.3 192 96 249.3 96 320C96 390.7 153.3 448 224 448L288 448L288 512L224 512C118 512 32 426 32 320zM608 320C608 426 522 512 416 512L352 512L352 448L416 448C486.7 448 544 390.7 544 320C544 249.3 486.7 192 416 192L352 192L352 128L416 128C522 128 608 214 608 320zM224 288L448 288L448 352L192 352L192 288L224 288z" }) })) : null }));
};
const TimelineScaleField = ({ field, propStatus, effectiveValue, onSave, onDragValueChange, onDragEnd, scaleLockNodePath, }) => {
    var _a, _b;
    const [dragX, setDragX] = (0, react_1.useState)(null);
    const [dragY, setDragY] = (0, react_1.useState)(null);
    const dragStartRef = (0, react_1.useRef)(null);
    const { getScaleLockState, setScaleLockState } = (0, react_1.useContext)(scale_lock_1.ScaleLockContext);
    const [codeX, codeY, codeZ] = (0, react_1.useMemo)(() => no_react_1.NoReactInternals.parseScaleValue(effectiveValue), [effectiveValue]);
    const defaultLinked = codeX === codeY;
    const linked = getScaleLockState({
        nodePath: scaleLockNodePath,
        fieldKey: field.key,
        defaultValue: defaultLinked,
    });
    const configuredStep = field.fieldSchema.type === 'scale' ? field.fieldSchema.step : undefined;
    const step = configuredStep !== null && configuredStep !== void 0 ? configuredStep : 0.01;
    const min = field.fieldSchema.type === 'scale'
        ? ((_a = field.fieldSchema.min) !== null && _a !== void 0 ? _a : -Infinity)
        : -Infinity;
    const max = field.fieldSchema.type === 'scale'
        ? ((_b = field.fieldSchema.max) !== null && _b !== void 0 ? _b : Infinity)
        : Infinity;
    const formatter = (0, react_1.useCallback)((v) => {
        return (0, timeline_field_display_utils_1.formatTimelineFieldValueForDisplay)({
            fieldSchema: field.fieldSchema,
            value: v,
        });
    }, [field.fieldSchema]);
    const getDragStart = (0, react_1.useCallback)(() => {
        if (dragStartRef.current === null) {
            dragStartRef.current = [dragX !== null && dragX !== void 0 ? dragX : codeX, dragY !== null && dragY !== void 0 ? dragY : codeY];
        }
        return dragStartRef.current;
    }, [codeX, codeY, dragX, dragY]);
    const serialize = (0, react_1.useCallback)((x, y) => {
        return no_react_1.NoReactInternals.serializeScaleValue([x, y, codeZ]);
    }, [codeZ]);
    const onXChange = (0, react_1.useCallback)((newVal) => {
        if (linked) {
            const [baseX, baseY] = getDragStart();
            const [newX, newY] = (0, exports.getLinkedScale)({
                axis: 'x',
                newValue: newVal,
                baseX,
                baseY,
                min,
                max,
            });
            setDragX(newX);
            setDragY(newY);
            onDragValueChange(serialize(newX, newY));
            return;
        }
        setDragX(newVal);
        const currentY = dragY !== null && dragY !== void 0 ? dragY : codeY;
        onDragValueChange(serialize(newVal, currentY));
    }, [
        codeY,
        dragY,
        getDragStart,
        linked,
        max,
        min,
        onDragValueChange,
        serialize,
    ]);
    const onXChangeEnd = (0, react_1.useCallback)((newVal) => {
        var _a, _b;
        var _c, _d;
        const [newX, newY] = linked
            ? (0, exports.getLinkedScale)({
                axis: 'x',
                newValue: newVal,
                baseX: (_c = (_a = dragStartRef.current) === null || _a === void 0 ? void 0 : _a[0]) !== null && _c !== void 0 ? _c : codeX,
                baseY: (_d = (_b = dragStartRef.current) === null || _b === void 0 ? void 0 : _b[1]) !== null && _d !== void 0 ? _d : codeY,
                min,
                max,
            })
            : [newVal, dragY !== null && dragY !== void 0 ? dragY : codeY];
        const newScale = serialize(newX, newY);
        const clearDragState = () => {
            dragStartRef.current = null;
            setDragX(null);
            setDragY(null);
            onDragEnd();
        };
        if (!valuesEqual(newScale, propStatus.codeValue)) {
            onSave(newScale).finally(clearDragState);
        }
        else {
            clearDragState();
        }
    }, [
        codeX,
        codeY,
        dragY,
        linked,
        max,
        min,
        onDragEnd,
        onSave,
        propStatus,
        serialize,
    ]);
    const onXTextChange = (0, react_1.useCallback)((newVal) => {
        const parsed = Number(newVal);
        if (Number.isNaN(parsed)) {
            return;
        }
        const [newX, newY] = linked
            ? (0, exports.getLinkedScale)({
                axis: 'x',
                newValue: parsed,
                baseX: codeX,
                baseY: codeY,
                min,
                max,
            })
            : [parsed, dragY !== null && dragY !== void 0 ? dragY : codeY];
        const newScale = serialize(newX, newY);
        if (!valuesEqual(newScale, propStatus.codeValue)) {
            setDragX(newX);
            setDragY(newY);
            onSave(newScale).finally(() => {
                dragStartRef.current = null;
                setDragX(null);
                setDragY(null);
            });
        }
    }, [codeX, codeY, dragY, linked, max, min, onSave, propStatus, serialize]);
    const onYChange = (0, react_1.useCallback)((newVal) => {
        if (linked) {
            const [baseX, baseY] = getDragStart();
            const [newX, newY] = (0, exports.getLinkedScale)({
                axis: 'y',
                newValue: newVal,
                baseX,
                baseY,
                min,
                max,
            });
            setDragX(newX);
            setDragY(newY);
            onDragValueChange(serialize(newX, newY));
            return;
        }
        setDragY(newVal);
        const currentX = dragX !== null && dragX !== void 0 ? dragX : codeX;
        onDragValueChange(serialize(currentX, newVal));
    }, [
        codeX,
        dragX,
        getDragStart,
        linked,
        max,
        min,
        onDragValueChange,
        serialize,
    ]);
    const onYChangeEnd = (0, react_1.useCallback)((newVal) => {
        var _a, _b;
        var _c, _d;
        const [newX, newY] = linked
            ? (0, exports.getLinkedScale)({
                axis: 'y',
                newValue: newVal,
                baseX: (_c = (_a = dragStartRef.current) === null || _a === void 0 ? void 0 : _a[0]) !== null && _c !== void 0 ? _c : codeX,
                baseY: (_d = (_b = dragStartRef.current) === null || _b === void 0 ? void 0 : _b[1]) !== null && _d !== void 0 ? _d : codeY,
                min,
                max,
            })
            : [dragX !== null && dragX !== void 0 ? dragX : codeX, newVal];
        const newScale = serialize(newX, newY);
        const clearDragState = () => {
            dragStartRef.current = null;
            setDragX(null);
            setDragY(null);
            onDragEnd();
        };
        if (!valuesEqual(newScale, propStatus.codeValue)) {
            onSave(newScale).finally(clearDragState);
        }
        else {
            clearDragState();
        }
    }, [
        codeX,
        codeY,
        dragX,
        linked,
        max,
        min,
        onDragEnd,
        onSave,
        propStatus,
        serialize,
    ]);
    const onYTextChange = (0, react_1.useCallback)((newVal) => {
        const parsed = Number(newVal);
        if (Number.isNaN(parsed)) {
            return;
        }
        const [newX, newY] = linked
            ? (0, exports.getLinkedScale)({
                axis: 'y',
                newValue: parsed,
                baseX: codeX,
                baseY: codeY,
                min,
                max,
            })
            : [dragX !== null && dragX !== void 0 ? dragX : codeX, parsed];
        const newScale = serialize(newX, newY);
        if (!valuesEqual(newScale, propStatus.codeValue)) {
            setDragX(newX);
            setDragY(newY);
            onSave(newScale).finally(() => {
                dragStartRef.current = null;
                setDragX(null);
                setDragY(null);
            });
        }
    }, [codeX, codeY, dragX, linked, max, min, onSave, propStatus, serialize]);
    const onToggleLink = (0, react_1.useCallback)(() => {
        setScaleLockState({
            nodePath: scaleLockNodePath,
            fieldKey: field.key,
            linked: !linked,
        });
    }, [field.key, linked, scaleLockNodePath, setScaleLockState]);
    return (jsx_runtime_1.jsxs("span", { style: containerStyle, children: [
            jsx_runtime_1.jsx(InputDragger_1.InputDragger, { type: "number", value: dragX !== null && dragX !== void 0 ? dragX : codeX, style: leftDraggerStyle, status: "ok", small: true, onValueChange: onXChange, onValueChangeEnd: onXChangeEnd, onTextChange: onXTextChange, min: min, max: max, step: step, formatter: formatter, rightAlign: false }), jsx_runtime_1.jsx("div", { style: gapStyle }), jsx_runtime_1.jsx(InputDragger_1.InputDragger, { type: "number", value: dragY !== null && dragY !== void 0 ? dragY : codeY, style: rightDraggerStyle, status: "ok", small: true, onValueChange: onYChange, onValueChangeEnd: onYChangeEnd, onTextChange: onYTextChange, min: min, max: max, step: step, formatter: formatter, rightAlign: false }), jsx_runtime_1.jsx(LinkToggle, { linked: linked, onToggle: onToggleLink })
        ] }));
};
exports.TimelineScaleField = TimelineScaleField;
