// @remotion/studio/dist/components/Timeline/TimelineTimeIndicators.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TimelineTimeIndicators = exports.TimelineTimePadding = exports.TimelineTimePlaceholders = exports.TIMELINE_TIME_INDICATOR_HEIGHT = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const react_1 = require("react");
const remotion_1 = require("remotion");
const colors_1 = require("../../helpers/colors");
const timeline_layout_1 = require("../../helpers/timeline-layout");
const render_frame_1 = require("../../state/render-frame");
const TimeValue_1 = require("../TimeValue");
const timeline_refs_1 = require("./timeline-refs");
const timeline_scroll_logic_1 = require("./timeline-scroll-logic");
const TimelineSelection_1 = require("./TimelineSelection");
const TimelineWidthProvider_1 = require("./TimelineWidthProvider");
exports.TIMELINE_TIME_INDICATOR_HEIGHT = 39;
const container = {
    height: exports.TIMELINE_TIME_INDICATOR_HEIGHT,
    position: 'absolute',
    backgroundColor: TimelineSelection_1.TIMELINE_TICKS_BACKGROUND,
    top: 0,
    borderBottom: `${timeline_layout_1.TIMELINE_ITEM_BORDER_BOTTOM}px solid ${colors_1.TIMELINE_TRACK_SEPARATOR}`,
};
const tick = {
    width: 1,
    backgroundColor: 'rgba(255, 255, 255, 0.15)',
    height: 20,
    position: 'absolute',
};
const secondTick = {
    ...tick,
    height: 15,
};
const TICK_LABEL_FONT_SIZE = 12;
const TICK_LABEL_MARGIN_LEFT = 8;
const TICK_LABEL_MIN_GAP = 16;
const tickLabel = {
    fontSize: TICK_LABEL_FONT_SIZE,
    marginLeft: TICK_LABEL_MARGIN_LEFT,
    marginTop: 7,
    color: colors_1.LIGHT_TEXT,
};
const timeValue = {
    height: exports.TIMELINE_TIME_INDICATOR_HEIGHT,
    position: 'absolute',
    top: 0,
    width: 'calc(100% + 1.5px)',
    paddingLeft: 10,
    display: 'flex',
    alignItems: 'center',
    backgroundColor: colors_1.BACKGROUND,
    borderBottom: `${timeline_layout_1.TIMELINE_ITEM_BORDER_BOTTOM}px solid ${colors_1.TIMELINE_TRACK_SEPARATOR}`,
};
const TimelineTimePlaceholders = () => {
    return (jsx_runtime_1.jsx("div", { style: timeValue, children: jsx_runtime_1.jsx(TimeValue_1.TimeValue, {}) }));
};
exports.TimelineTimePlaceholders = TimelineTimePlaceholders;
const TimelineTimePadding = () => {
    return (jsx_runtime_1.jsx("div", { style: {
            height: exports.TIMELINE_TIME_INDICATOR_HEIGHT,
        } }));
};
exports.TimelineTimePadding = TimelineTimePadding;
const TimelineTimeIndicators = () => {
    const sliderTrack = (0, react_1.useContext)(TimelineWidthProvider_1.TimelineWidthContext);
    const video = remotion_1.Internals.useVideo();
    if (sliderTrack === null) {
        return null;
    }
    if (video === null) {
        return null;
    }
    return (jsx_runtime_1.jsx(TimelineTimeIndicatorsInner, { durationInFrames: video.durationInFrames, fps: video.fps, windowWidth: sliderTrack }));
};
exports.TimelineTimeIndicators = TimelineTimeIndicators;
const TimelineTimeIndicatorsInner = ({ windowWidth, durationInFrames, fps }) => {
    const ref = (0, react_1.useRef)(null);
    (0, react_1.useEffect)(() => {
        const currentRef = ref.current;
        if (!currentRef) {
            return;
        }
        const { current } = timeline_refs_1.timelineVerticalScroll;
        if (!current) {
            return;
        }
        const onScroll = () => {
            currentRef.style.top = current.scrollTop + 'px';
        };
        current.addEventListener('scroll', onScroll);
        return () => {
            current.removeEventListener('scroll', onScroll);
        };
    }, []);
    const style = (0, react_1.useMemo)(() => {
        return {
            ...container,
            width: windowWidth,
            overflow: 'hidden',
            pointerEvents: 'none',
        };
    }, [windowWidth]);
    const ticks = (0, react_1.useMemo)(() => {
        const frameInterval = (0, timeline_scroll_logic_1.getFrameIncrementFromWidth)(durationInFrames, windowWidth);
        const MIN_SPACING_BETWEEN_TICKS_PX = 5;
        const maxTickLabelWidth = (0, render_frame_1.renderFrame)(durationInFrames - 1, fps).length *
            TICK_LABEL_FONT_SIZE *
            0.6;
        const minSpacingBetweenTickLabelsPx = TICK_LABEL_MARGIN_LEFT + maxTickLabelWidth + TICK_LABEL_MIN_GAP;
        const seconds = Math.floor(durationInFrames / fps);
        const secondMarkerEveryNth = Math.ceil(minSpacingBetweenTickLabelsPx / (frameInterval * fps));
        const frameMarkerEveryNth = Math.ceil(MIN_SPACING_BETWEEN_TICKS_PX / frameInterval);
        // Big ticks showing for every second
        const secondTicks = new Array(seconds)
            .fill(true)
            .map((_, index) => {
            return {
                frame: index * fps,
                style: {
                    ...secondTick,
                    left: frameInterval * index * fps + timeline_layout_1.TIMELINE_PADDING,
                },
                showTime: index > 0,
            };
        })
            .filter((_, idx) => idx % secondMarkerEveryNth === 0);
        const frameTicks = new Array(durationInFrames)
            .fill(true)
            .map((_, index) => {
            return {
                frame: index,
                style: {
                    ...tick,
                    left: frameInterval * index + timeline_layout_1.TIMELINE_PADDING,
                    height: index % fps === 0
                        ? 10
                        : (index / frameMarkerEveryNth) % 2 === 0
                            ? 5
                            : 2,
                },
                showTime: false,
            };
        })
            .filter((_, idx) => idx % frameMarkerEveryNth === 0);
        // Merge and deduplicate ticks
        const hasTicks = [];
        return [...secondTicks, ...frameTicks].filter((t) => {
            const alreadyUsed = hasTicks.find((ht) => ht === t.frame) !== undefined;
            hasTicks.push(t.frame);
            return !alreadyUsed;
        });
    }, [durationInFrames, fps, windowWidth]);
    return (jsx_runtime_1.jsx("div", { ref: ref, style: style, children: ticks.map((t) => {
            return (jsx_runtime_1.jsx("div", { style: t.style, children: t.showTime ? (jsx_runtime_1.jsx("div", { style: tickLabel, children: (0, render_frame_1.renderFrame)(t.frame, fps) })) : null }, t.frame));
        }) }));
};
