// @remotion/studio/dist/components/Timeline/TimelineKeyframeDiamond.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.TimelineKeyframeDiamond = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const react_1 = __importStar(require("react"));
const remotion_1 = require("remotion");
const colors_1 = require("../../helpers/colors");
const get_left_of_timeline_slider_1 = require("../../helpers/get-left-of-timeline-slider");
const timeline_layout_1 = require("../../helpers/timeline-layout");
const TimelineKeyframeDiamondIcon_1 = require("./TimelineKeyframeDiamondIcon");
const TimelineKeyframeDragState_1 = require("./TimelineKeyframeDragState");
const TimelineSelection_1 = require("./TimelineSelection");
const TimelineWidthProvider_1 = require("./TimelineWidthProvider");
const use_timeline_keyframe_drag_1 = require("./use-timeline-keyframe-drag");
const diamondSize = 12;
const diamondBase = {
    alignItems: 'center',
    background: 'none',
    border: 'none',
    display: 'flex',
    height: diamondSize,
    justifyContent: 'center',
    padding: 0,
    position: 'absolute',
    width: diamondSize,
};
const TimelineKeyframeDiamondUnmemoized = ({ frame, rowHeight, nodePathInfo }) => {
    const videoConfig = (0, remotion_1.useVideoConfig)();
    const timelineWidth = (0, react_1.useContext)(TimelineWidthProvider_1.TimelineWidthContext);
    const ref = (0, react_1.useRef)(null);
    const { selected, onSelect, selectable, selectionItem } = (0, TimelineSelection_1.useTimelineKeyframeSelection)(nodePathInfo, frame);
    (0, TimelineSelection_1.useTimelineMarqueeSelectableItem)(selectionItem, ref);
    const { isKeyframeDragging } = (0, TimelineKeyframeDragState_1.useTimelineKeyframeDragState)();
    const visuallySelected = selected || isKeyframeDragging({ nodePathInfo, frame });
    const style = (0, react_1.useMemo)(() => {
        if (timelineWidth === null) {
            return null;
        }
        return {
            ...diamondBase,
            left: (0, get_left_of_timeline_slider_1.getXPositionOfItemInTimelineImperatively)(frame, videoConfig.durationInFrames, timelineWidth) - timeline_layout_1.TIMELINE_PADDING,
            pointerEvents: 'auto',
            top: rowHeight / 2,
            transform: 'translate(-50%, -50%)',
        };
    }, [frame, rowHeight, timelineWidth, videoConfig.durationInFrames]);
    const onPointerDown = (0, use_timeline_keyframe_drag_1.useTimelineKeyframeDrag)({
        frame,
        nodePathInfo,
        onSelect,
        selectable,
        selected,
    });
    if (style === null) {
        return null;
    }
    return (jsx_runtime_1.jsx("button", { ref: ref, [TimelineSelection_1.TIMELINE_MARQUEE_ITEM_ATTR]: true, type: "button", style: style, title: `Keyframe at frame ${frame}`, "aria-label": `Select keyframe at frame ${frame}`, onPointerDown: selectable ? onPointerDown : undefined, children: jsx_runtime_1.jsx(TimelineKeyframeDiamondIcon_1.TimelineKeyframeDiamondIcon, { color: colors_1.LIGHT_TEXT, selected: visuallySelected, size: diamondSize }) }));
};
exports.TimelineKeyframeDiamond = react_1.default.memo(TimelineKeyframeDiamondUnmemoized);
