// @remotion/studio/dist/components/Timeline/TimelineEffectPropItem.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TimelineEffectPropItem = exports.TimelineEffectPropValue = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const studio_shared_1 = require("@remotion/studio-shared");
const react_1 = require("react");
const remotion_1 = require("remotion");
const client_id_1 = require("../../helpers/client-id");
const modals_1 = require("../../state/modals");
const call_api_1 = require("../call-api");
const ContextMenu_1 = require("../ContextMenu");
const call_add_keyframe_1 = require("./call-add-keyframe");
const get_timeline_keyframes_1 = require("./get-timeline-keyframes");
const save_effect_prop_1 = require("./save-effect-prop");
const save_prop_queue_1 = require("./save-prop-queue");
const timeline_field_row_layout_1 = require("./timeline-field-row-layout");
const TimelineExpandArrowButton_1 = require("./TimelineExpandArrowButton");
const TimelineFieldLabel_1 = require("./TimelineFieldLabel");
const TimelineKeyframeControls_1 = require("./TimelineKeyframeControls");
const TimelineKeyframedValue_1 = require("./TimelineKeyframedValue");
const TimelineLayerEye_1 = require("./TimelineLayerEye");
const TimelineRowChrome_1 = require("./TimelineRowChrome");
const TimelineSchemaField_1 = require("./TimelineSchemaField");
const TimelineSelection_1 = require("./TimelineSelection");
const fieldRowBase = {};
const isKeyframedStatus = (status) => {
    return status.status === 'keyframed';
};
const isResettableStatus = ({ status, defaultValue, }) => {
    var _a;
    if (defaultValue === undefined) {
        return false;
    }
    if (status.status === 'keyframed') {
        return true;
    }
    if (status.status === 'computed') {
        return false;
    }
    const effectiveCodeValue = (_a = status.codeValue) !== null && _a !== void 0 ? _a : defaultValue;
    return JSON.stringify(effectiveCodeValue) !== JSON.stringify(defaultValue);
};
const TimelineEffectPropValue = ({ field, nodePath, validatedLocation, sourceFrame }) => {
    var _a;
    var _b;
    const { setEffectDragOverrides, clearEffectDragOverrides, setPropStatuses } = (0, react_1.useContext)(remotion_1.Internals.VisualModeSettersContext);
    const { getEffectDragOverrides } = (0, react_1.useContext)(remotion_1.Internals.VisualModeDragOverridesContext);
    const { propStatuses: visualModePropStatuses } = (0, react_1.useContext)(remotion_1.Internals.VisualModePropStatusesContext);
    const { previewServerState } = (0, react_1.useContext)(client_id_1.StudioServerConnectionCtx);
    const clientId = previewServerState.type === 'connected'
        ? previewServerState.clientId
        : null;
    const effectStatus = remotion_1.Internals.getEffectPropStatusesCtx({
        propStatuses: visualModePropStatuses,
        nodePath,
        effectIndex: field.effectIndex,
    });
    const propStatus = effectStatus.type === 'can-update-effect'
        ? ((_b = (_a = effectStatus.props) === null || _a === void 0 ? void 0 : _a[field.key]) !== null && _b !== void 0 ? _b : null)
        : null;
    const onDragValueChange = (0, react_1.useCallback)((value) => {
        const nextDragOverrideValue = propStatus !== null && isKeyframedStatus(propStatus)
            ? remotion_1.Internals.makeKeyframedDragOverride({
                status: propStatus,
                frame: sourceFrame,
                value,
            })
            : remotion_1.Internals.makeStaticDragOverride(value);
        setEffectDragOverrides(nodePath, field.effectIndex, field.key, nextDragOverrideValue);
    }, [
        field.effectIndex,
        field.key,
        nodePath,
        propStatus,
        setEffectDragOverrides,
        sourceFrame,
    ]);
    const onDragEnd = (0, react_1.useCallback)(() => {
        clearEffectDragOverrides(nodePath, field.effectIndex);
    }, [clearEffectDragOverrides, nodePath, field.effectIndex]);
    const dragOverrideValue = (0, react_1.useMemo)(() => {
        const overrides = getEffectDragOverrides(nodePath, field.effectIndex);
        return overrides[field.key];
    }, [getEffectDragOverrides, nodePath, field.effectIndex, field.key]);
    const onSave = (0, react_1.useCallback)((value) => {
        if (!validatedLocation) {
            return Promise.reject(new Error('Cannot save'));
        }
        if (!propStatus) {
            return Promise.reject(new Error('Cannot save'));
        }
        if (propStatus.status !== 'static') {
            return Promise.reject(new Error('Cannot save'));
        }
        if (!clientId) {
            return Promise.reject(new Error('Not connected to studio server'));
        }
        const defaultValue = field.fieldSchema.default !== undefined
            ? JSON.stringify(field.fieldSchema.default)
            : null;
        const stringifiedValue = JSON.stringify(value);
        if (value === propStatus.codeValue) {
            return Promise.resolve();
        }
        if (defaultValue === stringifiedValue &&
            propStatus.codeValue === undefined) {
            return Promise.resolve();
        }
        return (0, save_prop_queue_1.enqueueSavePropChange)({
            nodePath,
            setPropStatuses,
            applyOptimistic: (prev) => (0, studio_shared_1.optimisticUpdateForEffectPropStatuses)({
                previous: prev,
                effectIndex: field.effectIndex,
                fieldKey: field.key,
                value,
                schema: field.effectSchema,
            }),
            apiCall: () => (0, call_api_1.callApi)('/api/save-effect-props', {
                type: 'value',
                fileName: validatedLocation.source,
                sequenceNodePath: nodePath,
                effectIndex: field.effectIndex,
                key: field.key,
                value: stringifiedValue,
                defaultValue,
                schema: field.effectSchema,
                clientId,
            }),
            errorLabel: 'Could not save effect prop',
        });
    }, [
        clientId,
        field.effectIndex,
        field.effectSchema,
        field.fieldSchema.default,
        field.key,
        nodePath,
        propStatus,
        setPropStatuses,
        validatedLocation,
    ]);
    const onSaveKeyframed = (0, react_1.useCallback)((value, frame) => {
        if (!validatedLocation) {
            return Promise.reject(new Error('Cannot save'));
        }
        if (!clientId) {
            return Promise.reject(new Error('Not connected to studio server'));
        }
        return (0, call_add_keyframe_1.callAddEffectKeyframe)({
            fileName: validatedLocation.source,
            nodePath,
            effectIndex: field.effectIndex,
            fieldKey: field.key,
            sourceFrame: frame,
            value,
            schema: field.effectSchema,
            setPropStatuses,
            clientId,
        });
    }, [
        clientId,
        field.effectIndex,
        field.effectSchema,
        field.key,
        nodePath,
        setPropStatuses,
        validatedLocation,
    ]);
    if (field.fieldSchema.type === 'scale') {
        throw new Error(`Effects do not support scale fields: ${field.key}`);
    }
    if (effectStatus.type === 'cannot-update-effect') {
        if (effectStatus.reason === 'computed') {
            return jsx_runtime_1.jsx(TimelineSchemaField_1.UnsupportedStatus, { label: "computed" });
        }
        if (effectStatus.reason === 'not-call-expression') {
            return jsx_runtime_1.jsx(TimelineSchemaField_1.UnsupportedStatus, { label: "not inline" });
        }
        if (effectStatus.reason === 'not-found') {
            return jsx_runtime_1.jsx(TimelineSchemaField_1.UnsupportedStatus, { label: "not found in code" });
        }
        throw new Error(`Unsupported effect status: ${effectStatus.reason}`);
    }
    if (effectStatus.type === 'cannot-update-sequence') {
        if (effectStatus.reason === 'not-found') {
            return jsx_runtime_1.jsx(TimelineSchemaField_1.UnsupportedStatus, { label: "not found in code" });
        }
        if (effectStatus.reason === 'error') {
            return jsx_runtime_1.jsx(TimelineSchemaField_1.UnsupportedStatus, { label: "error" });
        }
        throw new Error(`Unsupported effect status: ${effectStatus.reason}`);
    }
    if (propStatus === null) {
        return null;
    }
    if (isKeyframedStatus(propStatus)) {
        return (jsx_runtime_1.jsx(TimelineKeyframedValue_1.TimelineKeyframedValue, { field: field, propStatus: propStatus, sourceFrame: sourceFrame, dragOverrideValue: dragOverrideValue, onSave: onSaveKeyframed, onDragValueChange: onDragValueChange, onDragEnd: onDragEnd, scaleLockNodePath: nodePath }));
    }
    if (propStatus.status === 'computed') {
        return jsx_runtime_1.jsx(TimelineSchemaField_1.UnsupportedStatus, { label: (0, get_timeline_keyframes_1.getComputedStatusLabel)(propStatus) });
    }
    const effectiveValue = remotion_1.Internals.getEffectiveVisualModeValue({
        propStatus,
        dragOverrideValue,
        defaultValue: field.fieldSchema.default,
        frame: sourceFrame,
        shouldResortToDefaultValueIfUndefined: true,
    });
    return (jsx_runtime_1.jsx(TimelineSchemaField_1.TimelineFieldValue, { field: field, propStatus: propStatus, onSave: onSave, onDragValueChange: onDragValueChange, onDragEnd: onDragEnd, effectiveValue: effectiveValue, scaleLockNodePath: null }));
};
exports.TimelineEffectPropValue = TimelineEffectPropValue;
const TimelineEffectPropItem = ({ field, validatedLocation, rowDepth, nodePath, nodePathInfo, keyframeDisplayOffset, keyframeControlsMode = 'timeline', }) => {
    var _a;
    var _b, _c;
    const { previewServerState } = (0, react_1.useContext)(client_id_1.StudioServerConnectionCtx);
    const { setSelectedModal } = (0, react_1.useContext)(modals_1.ModalsContext);
    const { setPropStatuses } = (0, react_1.useContext)(remotion_1.Internals.VisualModeSettersContext);
    const { propStatuses } = (0, react_1.useContext)(remotion_1.Internals.VisualModePropStatusesContext);
    const { getEffectDragOverrides } = (0, react_1.useContext)(remotion_1.Internals.VisualModeDragOverridesContext);
    const selection = (0, TimelineSelection_1.useTimelineRowSelection)(nodePathInfo);
    const timelinePosition = remotion_1.Internals.Timeline.useTimelinePosition();
    const sourceFrame = timelinePosition - keyframeDisplayOffset;
    const style = (0, react_1.useMemo)(() => {
        return {
            ...fieldRowBase,
            height: field.rowHeight,
        };
    }, [field.rowHeight]);
    const effectStatus = (0, react_1.useMemo)(() => remotion_1.Internals.getEffectPropStatusesCtx({
        propStatuses,
        nodePath,
        effectIndex: field.effectIndex,
    }), [propStatuses, nodePath, field.effectIndex]);
    const propStatus = effectStatus.type === 'can-update-effect'
        ? ((_b = (_a = effectStatus.props) === null || _a === void 0 ? void 0 : _a[field.key]) !== null && _b !== void 0 ? _b : null)
        : null;
    const dragOverrideValue = (0, react_1.useMemo)(() => {
        const overrides = getEffectDragOverrides(nodePath, field.effectIndex);
        return overrides[field.key];
    }, [getEffectDragOverrides, nodePath, field.effectIndex, field.key]);
    const keyframable = (0, studio_shared_1.isSchemaFieldKeyframable)({
        schema: field.effectSchema,
        key: field.key,
    });
    const keyframeControls = propStatus !== null &&
        (keyframeControlsMode === 'inspector'
            ? keyframable
            : (0, TimelineKeyframeControls_1.shouldShowTimelineKeyframeControls)({
                propStatus,
                selected: selection.selected,
                keyframable,
            })) ? (jsx_runtime_1.jsx(TimelineKeyframeControls_1.TimelineKeyframeControls, { fieldKey: field.key, propStatus: propStatus, nodePath: nodePath, fileName: validatedLocation.source, keyframeDisplayOffset: keyframeDisplayOffset, defaultValue: field.fieldSchema.default, dragOverrideValue: dragOverrideValue, schema: field.effectSchema, effectIndex: field.effectIndex, nodePathInfo: nodePathInfo, mode: keyframeControlsMode })) : null;
    const canResetToDefault = (0, react_1.useMemo)(() => {
        if (!propStatus || propStatus.status === 'computed') {
            return false;
        }
        return isResettableStatus({
            status: propStatus,
            defaultValue: field.fieldSchema.default,
        });
    }, [field.fieldSchema.default, propStatus]);
    const canPerformReset = previewServerState.type === 'connected' &&
        propStatus !== null &&
        propStatus.status !== 'computed';
    const canShowReset = canPerformReset && field.fieldSchema.default !== undefined;
    const onReset = (0, react_1.useCallback)(() => {
        if (!canShowReset ||
            !canResetToDefault ||
            previewServerState.type !== 'connected') {
            return;
        }
        const defaultValue = field.fieldSchema.default !== undefined
            ? JSON.stringify(field.fieldSchema.default)
            : null;
        (0, save_effect_prop_1.saveEffectProp)({
            type: 'value',
            fileName: validatedLocation.source,
            nodePath,
            effectIndex: field.effectIndex,
            fieldKey: field.key,
            value: field.fieldSchema.default,
            defaultValue,
            schema: field.effectSchema,
            setPropStatuses,
            clientId: previewServerState.clientId,
        });
    }, [
        canResetToDefault,
        canShowReset,
        field.effectIndex,
        field.effectSchema,
        field.fieldSchema.default,
        field.key,
        nodePath,
        previewServerState,
        setPropStatuses,
        validatedLocation.source,
    ]);
    const onOpenKeyframeSettings = (0, react_1.useCallback)(() => {
        var _a;
        if (propStatus === null || !isKeyframedStatus(propStatus)) {
            return;
        }
        setSelectedModal({
            type: 'keyframe-settings',
            fileName: validatedLocation.source,
            nodePath,
            fieldKey: field.key,
            fieldLabel: (_a = field.description) !== null && _a !== void 0 ? _a : field.key,
            status: propStatus,
            schema: field.effectSchema,
            effectIndex: field.effectIndex,
        });
    }, [
        field.description,
        field.effectIndex,
        field.effectSchema,
        field.key,
        nodePath,
        propStatus,
        setSelectedModal,
        validatedLocation.source,
    ]);
    const contextMenuValues = (0, react_1.useMemo)(() => {
        const values = [
            {
                type: 'item',
                id: 'reset-effect-field',
                keyHint: null,
                label: 'Reset',
                leftItem: null,
                disabled: !canShowReset,
                onClick: onReset,
                quickSwitcherLabel: null,
                subMenu: null,
                value: 'reset-effect-field',
            },
        ];
        if (propStatus !== null && isKeyframedStatus(propStatus)) {
            values.push({
                type: 'item',
                id: 'keyframe-settings-effect-field',
                keyHint: null,
                label: 'Keyframe settings...',
                leftItem: null,
                disabled: previewServerState.type !== 'connected',
                onClick: onOpenKeyframeSettings,
                quickSwitcherLabel: null,
                subMenu: null,
                value: 'keyframe-settings-effect-field',
            });
        }
        return values;
    }, [
        canShowReset,
        onOpenKeyframeSettings,
        onReset,
        previewServerState,
        propStatus,
    ]);
    const row = (jsx_runtime_1.jsxs(TimelineRowChrome_1.TimelineRowChrome, { depth: rowDepth, eye: jsx_runtime_1.jsx(TimelineLayerEye_1.TimelineLayerEyeSpacer, {}), keyframeControls: keyframeControls, arrow: jsx_runtime_1.jsx(TimelineExpandArrowButton_1.TimelineExpandArrowSpacer, {}), style: style, selected: selection.selected, selectable: selection.selectable, selectionItem: selection.selectionItem, onSelect: selection.onSelect, showSelectedBackground: true, containsSelection: false, outerHeight: null, children: [
            jsx_runtime_1.jsx(TimelineFieldLabel_1.TimelineFieldLabel, { rowDepth: rowDepth, selected: selection.selected, label: (_c = field.description) !== null && _c !== void 0 ? _c : field.key }), jsx_runtime_1.jsx("div", { style: timeline_field_row_layout_1.timelineFieldValueColumnStyle, children: jsx_runtime_1.jsx(exports.TimelineEffectPropValue, { field: field, nodePath: nodePath, validatedLocation: validatedLocation, sourceFrame: sourceFrame }) })
        ] }));
    return (jsx_runtime_1.jsx(ContextMenu_1.ContextMenu, { values: contextMenuValues, onOpen: selection.selectable ? selection.onSelect : null, children: row }));
};
exports.TimelineEffectPropItem = TimelineEffectPropItem;
