// @remotion/studio/dist/components/Timeline/TimelineEffectItem.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TimelineEffectItem = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const react_1 = require("react");
const remotion_1 = require("remotion");
const client_id_1 = require("../../helpers/client-id");
const timeline_layout_1 = require("../../helpers/timeline-layout");
const call_api_1 = require("../call-api");
const ContextMenu_1 = require("../ContextMenu");
const NotificationCenter_1 = require("../Notifications/NotificationCenter");
const save_effect_prop_1 = require("./save-effect-prop");
const TimelineExpandArrowButton_1 = require("./TimelineExpandArrowButton");
const TimelineLayerEye_1 = require("./TimelineLayerEye");
const TimelineRowChrome_1 = require("./TimelineRowChrome");
const TimelineSelection_1 = require("./TimelineSelection");
const EFFECT_REORDER_MIME_TYPE = 'application/remotion-effect-reorder';
let currentEffectDrag = null;
const rowLabel = {
    fontSize: 12,
    color: 'rgba(255, 255, 255, 0.8)',
    userSelect: 'none',
};
const rowStyle = {
    height: timeline_layout_1.TREE_GROUP_ROW_HEIGHT,
    cursor: 'default',
};
const reorderWrapper = {
    position: 'relative',
};
const reorderLineBase = {
    backgroundColor: '#0b84ff',
    height: 2,
    left: 0,
    pointerEvents: 'none',
    position: 'absolute',
    right: 0,
    zIndex: 1,
};
const hasEffectReorderDragType = (dataTransfer) => {
    return Array.from(dataTransfer.types).includes(EFFECT_REORDER_MIME_TYPE);
};
const getEffectReorderDragData = (dataTransfer) => {
    if (currentEffectDrag) {
        return currentEffectDrag;
    }
    const value = dataTransfer.getData(EFFECT_REORDER_MIME_TYPE);
    if (!value) {
        return null;
    }
    try {
        const parsed = JSON.parse(value);
        if (typeof parsed.nodePathKey === 'string' &&
            typeof parsed.effectIndex === 'number') {
            return parsed;
        }
    }
    catch (_a) {
        return null;
    }
    return null;
};
const getDestinationIndex = ({ fromIndex, insertionIndex, }) => {
    return insertionIndex > fromIndex ? insertionIndex - 1 : insertionIndex;
};
const TimelineEffectItem = ({ label, nodePathInfo, effectIndex, effectSchema, documentationLink, nodePath, validatedLocation, rowDepth, getIsExpanded, toggleTrack, }) => {
    var _a;
    var _b;
    const { previewServerState } = (0, react_1.useContext)(client_id_1.StudioServerConnectionCtx);
    const previewConnected = previewServerState.type === 'connected';
    const { propStatuses } = (0, react_1.useContext)(remotion_1.Internals.VisualModePropStatusesContext);
    const { setPropStatuses } = (0, react_1.useContext)(remotion_1.Internals.VisualModeSettersContext);
    const selection = (0, TimelineSelection_1.useTimelineRowSelection)(nodePathInfo);
    const containsSelection = (0, TimelineSelection_1.useTimelineRowContainsSelection)(nodePathInfo);
    const [dropIndicator, setDropIndicator] = (0, react_1.useState)(null);
    const effectStatus = (0, react_1.useMemo)(() => remotion_1.Internals.getEffectPropStatusesCtx({
        propStatuses,
        nodePath,
        effectIndex,
    }), [propStatuses, nodePath, effectIndex]);
    const disabledStatus = effectStatus.type === 'can-update-effect'
        ? ((_b = (_a = effectStatus.props) === null || _a === void 0 ? void 0 : _a.disabled) !== null && _b !== void 0 ? _b : null)
        : null;
    const isDisabled = (0, react_1.useMemo)(() => {
        if ((disabledStatus === null || disabledStatus === void 0 ? void 0 : disabledStatus.status) === 'static') {
            return Boolean(disabledStatus.codeValue);
        }
        return false;
    }, [disabledStatus]);
    const canToggle = previewConnected && (disabledStatus === null || disabledStatus === void 0 ? void 0 : disabledStatus.status) === 'static';
    const deleteDisabled = !previewConnected ||
        effectStatus.type !== 'can-update-effect' ||
        !validatedLocation.source;
    const canReorder = previewConnected &&
        effectStatus.type === 'can-update-effect' &&
        Boolean(validatedLocation.source);
    const nodePathKey = (0, react_1.useMemo)(() => remotion_1.Internals.makeSequencePropsSubscriptionKey(nodePath), [nodePath]);
    const onDeleteEffectFromSource = (0, react_1.useCallback)(async () => {
        if (deleteDisabled) {
            return;
        }
        try {
            const result = await (0, call_api_1.callApi)('/api/delete-effect', [
                {
                    type: 'single-effect',
                    fileName: validatedLocation.source,
                    sequenceNodePath: nodePath,
                    effectIndex,
                },
            ]);
            if (result.success) {
                (0, NotificationCenter_1.showNotification)('Removed effect from source file', 2000);
            }
            else {
                (0, NotificationCenter_1.showNotification)(result.reason, 4000);
            }
        }
        catch (err) {
            (0, NotificationCenter_1.showNotification)(err.message, 4000);
        }
    }, [deleteDisabled, effectIndex, nodePath, validatedLocation.source]);
    const contextMenuValues = (0, react_1.useMemo)(() => {
        if (!previewConnected) {
            return [];
        }
        const items = [];
        if (documentationLink) {
            items.push({
                type: 'item',
                id: 'open-effect-docs',
                keyHint: null,
                label: 'Open effect docs',
                leftItem: null,
                disabled: false,
                onClick: () => {
                    window.open(documentationLink, '_blank', 'noopener,noreferrer');
                },
                quickSwitcherLabel: null,
                subMenu: null,
                value: 'open-effect-docs',
            });
            items.push({
                type: 'divider',
                id: 'open-effect-docs-divider',
            });
        }
        items.push({
            type: 'item',
            id: 'delete-effect',
            keyHint: null,
            label: 'Delete',
            leftItem: null,
            disabled: deleteDisabled,
            onClick: () => {
                if (deleteDisabled) {
                    return;
                }
                onDeleteEffectFromSource();
            },
            quickSwitcherLabel: null,
            subMenu: null,
            value: 'delete-effect',
        });
        return items;
    }, [
        deleteDisabled,
        documentationLink,
        onDeleteEffectFromSource,
        previewConnected,
    ]);
    const onToggle = (0, react_1.useCallback)((type) => {
        if (!canToggle || previewServerState.type !== 'connected') {
            return;
        }
        const newValue = type !== 'enable';
        const fieldSchema = effectSchema.disabled;
        const defaultValue = fieldSchema && fieldSchema.type === 'boolean'
            ? JSON.stringify(fieldSchema.default)
            : null;
        (0, save_effect_prop_1.saveEffectProp)({
            type: 'value',
            fileName: validatedLocation.source,
            nodePath,
            effectIndex,
            fieldKey: 'disabled',
            value: newValue,
            defaultValue,
            schema: effectSchema,
            setPropStatuses,
            clientId: previewServerState.clientId,
        });
    }, [
        canToggle,
        effectIndex,
        effectSchema,
        nodePath,
        previewServerState,
        setPropStatuses,
        validatedLocation.source,
    ]);
    const isExpanded = getIsExpanded(nodePathInfo);
    const labelStyle = (0, react_1.useMemo)(() => {
        return {
            ...rowLabel,
            ...(0, TimelineSelection_1.getTimelineSelectedLabelStyle)(selection.selected, true),
            alignSelf: 'stretch',
            alignItems: 'center',
            color: (0, TimelineSelection_1.getTimelineColor)(selection.selected, true),
            display: 'inline-flex',
            marginRight: timeline_layout_1.EXPANDED_SECTION_PADDING_RIGHT,
            minWidth: 0,
            boxShadow: containsSelection && !selection.selected
                ? `inset 0 0 0 2px ${TimelineSelection_1.TIMELINE_SELECTED_LABEL_BACKGROUND}`
                : undefined,
        };
    }, [containsSelection, selection.selected]);
    const getDropTarget = (0, react_1.useCallback)((e) => {
        const dragData = getEffectReorderDragData(e.dataTransfer);
        if (!dragData || dragData.nodePathKey !== nodePathKey) {
            return null;
        }
        const rect = e.currentTarget.getBoundingClientRect();
        const before = e.clientY < rect.top + rect.height / 2;
        const insertionIndex = before ? effectIndex : effectIndex + 1;
        const toIndex = getDestinationIndex({
            fromIndex: dragData.effectIndex,
            insertionIndex,
        });
        if (toIndex === dragData.effectIndex) {
            return null;
        }
        return {
            dragData,
            toIndex,
            indicator: before ? 'before' : 'after',
        };
    }, [effectIndex, nodePathKey]);
    const onDragStart = (0, react_1.useCallback)((e) => {
        if (!canReorder) {
            e.preventDefault();
            return;
        }
        const dragData = { nodePathKey, effectIndex };
        currentEffectDrag = dragData;
        e.dataTransfer.effectAllowed = 'move';
        e.dataTransfer.setData(EFFECT_REORDER_MIME_TYPE, JSON.stringify(dragData));
        e.stopPropagation();
    }, [canReorder, effectIndex, nodePathKey]);
    const onDragEnd = (0, react_1.useCallback)(() => {
        currentEffectDrag = null;
        setDropIndicator(null);
    }, []);
    const onDragOver = (0, react_1.useCallback)((e) => {
        if (!canReorder || !hasEffectReorderDragType(e.dataTransfer)) {
            return;
        }
        const dropTarget = getDropTarget(e);
        if (!dropTarget) {
            setDropIndicator(null);
            return;
        }
        e.preventDefault();
        e.stopPropagation();
        e.dataTransfer.dropEffect = 'move';
        setDropIndicator(dropTarget.indicator);
    }, [canReorder, getDropTarget]);
    const onDragLeave = (0, react_1.useCallback)((e) => {
        if (e.currentTarget.contains(e.relatedTarget)) {
            return;
        }
        setDropIndicator(null);
    }, []);
    const onDrop = (0, react_1.useCallback)(async (e) => {
        if (!canReorder ||
            previewServerState.type !== 'connected' ||
            !validatedLocation.source) {
            return;
        }
        const dropTarget = getDropTarget(e);
        if (!dropTarget) {
            setDropIndicator(null);
            return;
        }
        e.preventDefault();
        e.stopPropagation();
        setDropIndicator(null);
        currentEffectDrag = null;
        try {
            const result = await (0, call_api_1.callApi)('/api/reorder-effect', {
                fileName: validatedLocation.source,
                sequenceNodePath: nodePath,
                fromIndex: dropTarget.dragData.effectIndex,
                toIndex: dropTarget.toIndex,
                clientId: previewServerState.clientId,
            });
            if (result.success) {
                (0, NotificationCenter_1.showNotification)('Reordered effect', 2000);
            }
            else {
                (0, NotificationCenter_1.showNotification)(result.reason, 4000);
            }
        }
        catch (err) {
            (0, NotificationCenter_1.showNotification)(err.message, 4000);
        }
    }, [
        canReorder,
        getDropTarget,
        nodePath,
        previewServerState,
        validatedLocation.source,
    ]);
    const reorderLineStyle = (0, react_1.useMemo)(() => {
        if (!dropIndicator) {
            return null;
        }
        return {
            ...reorderLineBase,
            ...(dropIndicator === 'before' ? { top: -1 } : { bottom: -1 }),
        };
    }, [dropIndicator]);
    const row = (jsx_runtime_1.jsx(TimelineRowChrome_1.TimelineRowChrome, { depth: rowDepth, eye: canToggle ? (jsx_runtime_1.jsx(TimelineLayerEye_1.TimelineLayerEye, { type: "effect", hidden: isDisabled, onInvoked: onToggle })) : (jsx_runtime_1.jsx(TimelineLayerEye_1.TimelineLayerEyeSpacer, {})), arrow: jsx_runtime_1.jsx(TimelineExpandArrowButton_1.TimelineExpandArrowButton, { isExpanded: isExpanded, onClick: () => toggleTrack(nodePathInfo), label: `${label} section`, disabled: false }), style: rowStyle, selected: selection.selected, selectable: selection.selectable, selectionItem: selection.selectionItem, onSelect: selection.onSelect, showSelectedBackground: true, containsSelection: containsSelection, outerHeight: null, children: jsx_runtime_1.jsx("span", { title: label, style: labelStyle, children: label }) }));
    const draggableRow = canReorder ? (jsx_runtime_1.jsxs("div", { draggable: true, onDragStart: onDragStart, onDragEnd: onDragEnd, onDragOver: onDragOver, onDragLeave: onDragLeave, onDrop: onDrop, style: reorderWrapper, children: [reorderLineStyle ? jsx_runtime_1.jsx("div", { style: reorderLineStyle }) : null, row] })) : (row);
    return previewConnected ? (jsx_runtime_1.jsx(ContextMenu_1.ContextMenu, { values: contextMenuValues, onOpen: selection.selectable ? selection.onSelect : null, children: draggableRow })) : (draggableRow);
};
exports.TimelineEffectItem = TimelineEffectItem;
