// @remotion/studio/dist/components/Timeline/TimelineTranslateField.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TimelineTranslateField = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const react_1 = require("react");
const InputDragger_1 = require("../NewComposition/InputDragger");
const timeline_field_display_utils_1 = require("./timeline-field-display-utils");
const timeline_field_utils_1 = require("./timeline-field-utils");
const timeline_translate_utils_1 = require("./timeline-translate-utils");
const leftDraggerStyle = {
    paddingLeft: 0,
};
const rightDraggerStyle = {
    paddingRight: 0,
};
const containerStyle = {
    display: 'flex',
    gap: 4,
};
const TimelineTranslateField = ({ field, propStatus, effectiveValue, onSave, onDragValueChange, onDragEnd, }) => {
    const [dragX, setDragX] = (0, react_1.useState)(null);
    const [dragY, setDragY] = (0, react_1.useState)(null);
    const [codeX, codeY] = (0, react_1.useMemo)(() => (0, timeline_translate_utils_1.parseTranslate)(String(effectiveValue !== null && effectiveValue !== void 0 ? effectiveValue : '0px 0px')), [effectiveValue]);
    const configuredStep = field.fieldSchema.type === 'translate' ? field.fieldSchema.step : undefined;
    const step = configuredStep !== null && configuredStep !== void 0 ? configuredStep : 1;
    const decimalPlaces = (0, react_1.useMemo)(() => (0, timeline_field_utils_1.getTimelineDisplayDecimalPlaces)({
        defaultDecimalPlaces: 1,
        step: configuredStep,
    }), [configuredStep]);
    const formatter = (0, react_1.useCallback)((v) => {
        return (0, timeline_field_display_utils_1.formatTimelineFieldValueForDisplay)({
            fieldSchema: field.fieldSchema,
            value: v,
        });
    }, [field.fieldSchema]);
    // --- X callbacks ---
    const onXChange = (0, react_1.useCallback)((newVal) => {
        setDragX(newVal);
        const currentY = dragY !== null && dragY !== void 0 ? dragY : codeY;
        onDragValueChange((0, timeline_translate_utils_1.serializeTranslate)(newVal, currentY, decimalPlaces));
    }, [onDragValueChange, dragY, codeY, decimalPlaces]);
    const onXChangeEnd = (0, react_1.useCallback)((newVal) => {
        const currentY = dragY !== null && dragY !== void 0 ? dragY : codeY;
        const newStr = (0, timeline_translate_utils_1.serializeTranslate)(newVal, currentY, decimalPlaces);
        if (newStr !== propStatus.codeValue) {
            onSave(newStr).finally(() => {
                setDragX(null);
                onDragEnd();
            });
        }
        else {
            setDragX(null);
            onDragEnd();
        }
    }, [dragY, codeY, decimalPlaces, propStatus, onSave, onDragEnd]);
    const onXTextChange = (0, react_1.useCallback)((newVal) => {
        const parsed = Number(newVal);
        if (!Number.isNaN(parsed)) {
            const currentY = dragY !== null && dragY !== void 0 ? dragY : codeY;
            const newStr = (0, timeline_translate_utils_1.serializeTranslate)(parsed, currentY, decimalPlaces);
            if (newStr !== propStatus.codeValue) {
                setDragX(parsed);
                onSave(newStr).finally(() => {
                    setDragX(null);
                });
            }
        }
    }, [propStatus, dragY, codeY, decimalPlaces, onSave]);
    // --- Y callbacks ---
    const onYChange = (0, react_1.useCallback)((newVal) => {
        setDragY(newVal);
        const currentX = dragX !== null && dragX !== void 0 ? dragX : codeX;
        onDragValueChange((0, timeline_translate_utils_1.serializeTranslate)(currentX, newVal, decimalPlaces));
    }, [onDragValueChange, dragX, codeX, decimalPlaces]);
    const onYChangeEnd = (0, react_1.useCallback)((newVal) => {
        const currentX = dragX !== null && dragX !== void 0 ? dragX : codeX;
        const newStr = (0, timeline_translate_utils_1.serializeTranslate)(currentX, newVal, decimalPlaces);
        if (newStr !== propStatus.codeValue) {
            onSave(newStr).finally(() => {
                setDragY(null);
                onDragEnd();
            });
        }
        else {
            setDragY(null);
            onDragEnd();
        }
    }, [dragX, codeX, decimalPlaces, propStatus, onSave, onDragEnd]);
    const onYTextChange = (0, react_1.useCallback)((newVal) => {
        const parsed = Number(newVal);
        if (!Number.isNaN(parsed)) {
            const currentX = dragX !== null && dragX !== void 0 ? dragX : codeX;
            const newStr = (0, timeline_translate_utils_1.serializeTranslate)(currentX, parsed, decimalPlaces);
            if (newStr !== propStatus.codeValue) {
                setDragY(parsed);
                onSave(newStr).finally(() => {
                    setDragY(null);
                });
            }
        }
    }, [propStatus, onSave, dragX, codeX, decimalPlaces]);
    return (jsx_runtime_1.jsxs("span", { style: containerStyle, children: [
            jsx_runtime_1.jsx(InputDragger_1.InputDragger, { type: "number", value: dragX !== null && dragX !== void 0 ? dragX : codeX, style: leftDraggerStyle, status: "ok", small: true, onValueChange: onXChange, onValueChangeEnd: onXChangeEnd, onTextChange: onXTextChange, min: -Infinity, max: Infinity, step: step, formatter: formatter, rightAlign: false, snapToStep: false, dragDecimalPlaces: decimalPlaces }), jsx_runtime_1.jsx("div", { style: { marginLeft: -6, marginRight: -6 } }), jsx_runtime_1.jsx(InputDragger_1.InputDragger, { type: "number", value: dragY !== null && dragY !== void 0 ? dragY : codeY, style: rightDraggerStyle, status: "ok", small: true, onValueChange: onYChange, onValueChangeEnd: onYChangeEnd, onTextChange: onYTextChange, min: -Infinity, max: Infinity, step: step, formatter: formatter, rightAlign: false, snapToStep: false, dragDecimalPlaces: decimalPlaces })
        ] }));
};
exports.TimelineTranslateField = TimelineTranslateField;
