// @remotion/studio/dist/components/Timeline/TimelineExpandedSection.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.TimelineExpandedSection = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const react_1 = __importStar(require("react"));
const colors_1 = require("../../helpers/colors");
const timeline_layout_1 = require("../../helpers/timeline-layout");
const TimelineExpandedRow_1 = require("./TimelineExpandedRow");
const use_timeline_expanded_tree_1 = require("./use-timeline-expanded-tree");
const expandedSectionBase = {
    color: 'white',
    fontFamily: 'Arial, Helvetica, sans-serif',
    fontSize: 12,
    display: 'flex',
    flexDirection: 'column',
    borderBottom: `1px solid ${colors_1.TIMELINE_TRACK_SEPARATOR}`,
};
const separator = {
    height: 0,
    borderBottom: `1px solid ${colors_1.TIMELINE_TRACK_SEPARATOR}`,
};
const TimelineExpandedSection = ({ sequence, validatedLocation, nodePathInfo, nestedDepth, keyframeDisplayOffset, }) => {
    const { filteredTree, getIsExpanded, toggleTrack } = (0, use_timeline_expanded_tree_1.useTimelineExpandedTree)({
        sequence,
        nodePathInfo,
    });
    const flat = (0, react_1.useMemo)(() => (0, timeline_layout_1.flattenVisibleTreeNodes)({ nodes: filteredTree, getIsExpanded }), [filteredTree, getIsExpanded]);
    const expandedHeight = (0, react_1.useMemo)(() => {
        const totalRowsHeight = flat.reduce((sum, { node }) => sum + (0, timeline_layout_1.getTreeRowHeight)(node), 0);
        const separators = Math.max(0, flat.length - 1);
        return totalRowsHeight + separators;
    }, [flat]);
    const style = (0, react_1.useMemo)(() => {
        return {
            ...expandedSectionBase,
            height: expandedHeight,
        };
    }, [expandedHeight]);
    const { schema } = sequence.controls;
    if (flat.length === 0) {
        return null;
    }
    return (jsx_runtime_1.jsx("div", { style: style, children: flat.map(({ node, depth }, i) => {
            return (jsx_runtime_1.jsxs(react_1.default.Fragment, { children: [i > 0 ? jsx_runtime_1.jsx("div", { style: separator }) : null, jsx_runtime_1.jsx(TimelineExpandedRow_1.TimelineExpandedRow, { node: node, depth: depth, nestedDepth: nestedDepth, getIsExpanded: getIsExpanded, toggleTrack: toggleTrack, validatedLocation: validatedLocation, nodePath: nodePathInfo.sequenceSubscriptionKey, schema: schema, keyframeDisplayOffset: keyframeDisplayOffset })
                ] }, JSON.stringify(node.nodePathInfo)));
        }) }));
};
exports.TimelineExpandedSection = TimelineExpandedSection;
