// @remotion/studio/dist/components/Timeline/TimelineMediaInfo.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TimelineMediaInfo = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const react_1 = require("react");
const remotion_1 = require("remotion");
const colors_1 = require("../../helpers/colors");
const timeline_asset_link_1 = require("./timeline-asset-link");
const lineStyle = {
    whiteSpace: 'nowrap',
    overflow: 'hidden',
    textOverflow: 'ellipsis',
    minWidth: 0,
    fontSize: 12,
    color: colors_1.VERY_LIGHT_TEXT,
    display: 'inline-block',
};
const useAssetLink = (src) => {
    const linkInfo = (0, react_1.useMemo)(() => (0, timeline_asset_link_1.getTimelineAssetLinkInfo)(src), [src]);
    const fileNameStyle = (0, react_1.useMemo)(() => ({
        ...lineStyle,
        color: colors_1.VERY_LIGHT_TEXT,
        textDecoration: 'none',
        display: 'inline-block',
        overflow: 'hidden',
        whiteSpace: 'pre',
        textOverflow: 'ellipsis',
        userSelect: 'none',
        WebkitUserSelect: 'none',
    }), []);
    return {
        linkInfo,
        fileNameStyle,
    };
};
const TimelineMediaInfo = ({ src }) => {
    const fileName = (0, react_1.useMemo)(() => remotion_1.Internals.getAssetDisplayName(src), [src]);
    const { linkInfo, fileNameStyle } = useAssetLink(src);
    return (jsx_runtime_1.jsx("div", { style: fileNameStyle, title: linkInfo ? linkInfo.title : fileName, children: fileName }));
};
exports.TimelineMediaInfo = TimelineMediaInfo;
