// @remotion/studio/dist/components/Timeline/TimelineKeyframeDragState.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.useTimelineKeyframeDragState = exports.TimelineKeyframeDragStateProvider = exports.getTimelineKeyframeDragKey = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const react_1 = require("react");
const timeline_node_path_key_1 = require("../../helpers/timeline-node-path-key");
const emptyDraggedKeyframes = new Set();
const getTimelineKeyframeDragKey = ({ nodePathInfo, frame, }) => {
    return `${(0, timeline_node_path_key_1.timelineNodePathInfoToKey)(nodePathInfo)}.keyframe.${frame}`;
};
exports.getTimelineKeyframeDragKey = getTimelineKeyframeDragKey;
const TimelineKeyframeDragStateContext = (0, react_1.createContext)({
    clearDraggedKeyframes: () => undefined,
    isKeyframeDragging: () => false,
    setDraggedKeyframes: () => undefined,
});
const TimelineKeyframeDragStateProvider = ({ children }) => {
    const [draggedKeys, setDraggedKeys] = (0, react_1.useState)(emptyDraggedKeyframes);
    const setDraggedKeyframes = (0, react_1.useCallback)((keyframes) => {
        setDraggedKeys(new Set(keyframes.map(exports.getTimelineKeyframeDragKey)));
    }, []);
    const clearDraggedKeyframes = (0, react_1.useCallback)(() => {
        setDraggedKeys(emptyDraggedKeyframes);
    }, []);
    const isKeyframeDragging = (0, react_1.useCallback)((keyframe) => {
        return draggedKeys.has((0, exports.getTimelineKeyframeDragKey)(keyframe));
    }, [draggedKeys]);
    const value = (0, react_1.useMemo)(() => ({
        clearDraggedKeyframes,
        isKeyframeDragging,
        setDraggedKeyframes,
    }), [clearDraggedKeyframes, isKeyframeDragging, setDraggedKeyframes]);
    return (jsx_runtime_1.jsx(TimelineKeyframeDragStateContext.Provider, { value: value, children: children }));
};
exports.TimelineKeyframeDragStateProvider = TimelineKeyframeDragStateProvider;
const useTimelineKeyframeDragState = () => {
    return (0, react_1.useContext)(TimelineKeyframeDragStateContext);
};
exports.useTimelineKeyframeDragState = useTimelineKeyframeDragState;
