// @remotion/studio/dist/components/Timeline/TimelineInOutDragHandler.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.TimelineInOutDragHandler = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const player_1 = require("@remotion/player");
const react_1 = __importStar(require("react"));
const remotion_1 = require("remotion");
const get_left_of_timeline_slider_1 = require("../../helpers/get-left-of-timeline-slider");
const in_out_1 = require("../../state/in-out");
const z_index_1 = require("../../state/z-index");
const ContextMenu_1 = require("../ContextMenu");
const ForceSpecificCursor_1 = require("../ForceSpecificCursor");
const TimelineInOutToggle_1 = require("../TimelineInOutToggle");
const timeline_refs_1 = require("./timeline-refs");
const timeline_scroll_logic_1 = require("./timeline-scroll-logic");
const TimelineInOutPointer_1 = require("./TimelineInOutPointer");
const TimelineInOutPointerHandle_1 = require("./TimelineInOutPointerHandle");
const TimelineWidthProvider_1 = require("./TimelineWidthProvider");
const getClientXWithScroll = (x) => {
    var _a;
    return x + ((_a = timeline_refs_1.scrollableRef.current) === null || _a === void 0 ? void 0 : _a.scrollLeft);
};
const TimelineInOutDragHandler = () => {
    const video = remotion_1.Internals.useUnsafeVideoConfig();
    const { canvasContent } = (0, react_1.useContext)(remotion_1.Internals.CompositionManager);
    if (!canvasContent || canvasContent.type !== 'composition') {
        return null;
    }
    return video ? jsx_runtime_1.jsx(TimelineInOutDragHandlerInnerMemo, {}) : null;
};
exports.TimelineInOutDragHandler = TimelineInOutDragHandler;
const TimelineInOutDragHandlerInner = () => {
    var _a;
    var _b, _c;
    const videoConfig = (0, remotion_1.useVideoConfig)();
    const size = player_1.PlayerInternals.useElementSize(timeline_refs_1.scrollableRef, {
        triggerOnWindowResize: true,
        shouldApplyCssTransforms: true,
    });
    const { isHighestContext } = (0, z_index_1.useZIndex)();
    const [inOutDragging, setInOutDragging] = (0, react_1.useState)({
        dragging: false,
    });
    const timelineWidth = (0, react_1.useContext)(TimelineWidthProvider_1.TimelineWidthContext);
    const get = (0, react_1.useCallback)((frame) => {
        if (timelineWidth === null) {
            throw new Error('timeline width is not yet determined');
        }
        return (0, get_left_of_timeline_slider_1.getXPositionOfItemInTimelineImperatively)(frame, videoConfig.durationInFrames, timelineWidth);
    }, [timelineWidth, videoConfig.durationInFrames]);
    const width = (_b = (_a = timeline_refs_1.scrollableRef.current) === null || _a === void 0 ? void 0 : _a.scrollWidth) !== null && _b !== void 0 ? _b : 0;
    const left = (_c = size === null || size === void 0 ? void 0 : size.left) !== null && _c !== void 0 ? _c : 0;
    const { inFrame, outFrame } = (0, in_out_1.useTimelineInOutFramePosition)();
    const { setInAndOutFrames } = (0, in_out_1.useTimelineSetInOutFramePosition)();
    const onInPointerDown = (0, react_1.useCallback)((e) => {
        if (e.button !== 0) {
            return;
        }
        if (!isHighestContext) {
            return;
        }
        if (!videoConfig) {
            return;
        }
        if (inFrame === null) {
            throw new Error('expected in frame');
        }
        e.stopPropagation();
        document.body.style.userSelect = 'none';
        document.body.style.webkitUserSelect = 'none';
        const inMarker = get(inFrame);
        const outMarker = outFrame === null ? Infinity : get(outFrame - 1);
        (0, ForceSpecificCursor_1.forceSpecificCursor)('ew-resize');
        setInOutDragging({
            dragging: 'in',
            initialOffset: getClientXWithScroll(e.clientX),
            boundaries: [-Infinity, outMarker - inMarker],
        });
    }, [isHighestContext, videoConfig, inFrame, get, outFrame]);
    const onOutPointerDown = (0, react_1.useCallback)((e) => {
        if (e.button !== 0) {
            return;
        }
        if (!isHighestContext) {
            return;
        }
        if (!videoConfig) {
            return;
        }
        if (outFrame === null) {
            throw new Error('expected out frame');
        }
        e.stopPropagation();
        document.body.style.userSelect = 'none';
        document.body.style.webkitUserSelect = 'none';
        const outMarker = get(outFrame);
        const inMarker = inFrame === null ? -Infinity : get(inFrame + 1);
        (0, ForceSpecificCursor_1.forceSpecificCursor)('ew-resize');
        setInOutDragging({
            dragging: 'out',
            initialOffset: getClientXWithScroll(e.clientX),
            boundaries: [inMarker - outMarker, Infinity],
        });
    }, [isHighestContext, videoConfig, inFrame, get, outFrame]);
    const onPointerMoveInOut = (0, react_1.useCallback)((e) => {
        if (!videoConfig) {
            return;
        }
        if (!inOutDragging.dragging) {
            return;
        }
        const offset = Math.max(inOutDragging.boundaries[0], Math.min(inOutDragging.boundaries[1], getClientXWithScroll(e.clientX) - inOutDragging.initialOffset));
        if (inOutDragging.dragging === 'in') {
            if (!TimelineInOutPointerHandle_1.inPointerHandle.current) {
                throw new Error('in pointer handle');
            }
            if (!TimelineInOutPointer_1.inMarkerAreaRef.current) {
                throw new Error('expected inMarkerAreaRef');
            }
            if (!inFrame) {
                throw new Error('expected inframes');
            }
            TimelineInOutPointerHandle_1.inPointerHandle.current.style.transform = `translateX(${get(inFrame) + offset}px)`;
            TimelineInOutPointer_1.inMarkerAreaRef.current.style.width =
                String(get(inFrame) + offset) + 'px';
        }
        if (inOutDragging.dragging === 'out') {
            if (!TimelineInOutPointerHandle_1.outPointerHandle.current) {
                throw new Error('in pointer handle');
            }
            if (!TimelineInOutPointer_1.outMarkerAreaRef.current) {
                throw new Error('in outMarkerAreaRef');
            }
            if (!outFrame) {
                throw new Error('expected outframes');
            }
            TimelineInOutPointerHandle_1.outPointerHandle.current.style.transform = `translateX(${get(outFrame) + offset}px)`;
            TimelineInOutPointer_1.outMarkerAreaRef.current.style.left =
                String(get(outFrame) + offset) + 'px';
            TimelineInOutPointer_1.outMarkerAreaRef.current.style.width =
                String(width - get(outFrame) - offset) + 'px';
        }
    }, [get, inFrame, inOutDragging, outFrame, videoConfig, width]);
    const onPointerUpInOut = (0, react_1.useCallback)((e) => {
        document.body.style.userSelect = '';
        document.body.style.webkitUserSelect = '';
        (0, ForceSpecificCursor_1.stopForcingSpecificCursor)();
        if (!videoConfig) {
            return;
        }
        if (!inOutDragging.dragging) {
            return;
        }
        setInOutDragging({
            dragging: false,
        });
        const frame = (0, timeline_scroll_logic_1.getFrameFromX)({
            clientX: getClientXWithScroll(e.clientX) - left,
            durationInFrames: videoConfig.durationInFrames,
            width,
            extrapolate: 'extend',
        });
        if (inOutDragging.dragging === 'in') {
            if (frame < 1) {
                return setInAndOutFrames((prev) => {
                    var _a;
                    return ({
                        ...prev,
                        [videoConfig.id]: {
                            ...((_a = prev[videoConfig.id]) !== null && _a !== void 0 ? _a : TimelineInOutToggle_1.defaultInOutValue),
                            inFrame: null,
                        },
                    });
                });
            }
            const maxFrame = outFrame === null ? Infinity : outFrame - 1;
            setInAndOutFrames((prev) => {
                var _a;
                return ({
                    ...prev,
                    [videoConfig.id]: {
                        ...((_a = prev[videoConfig.id]) !== null && _a !== void 0 ? _a : TimelineInOutToggle_1.defaultInOutValue),
                        inFrame: Math.min(maxFrame, frame),
                    },
                });
            });
        }
        else {
            if (frame > videoConfig.durationInFrames - 2) {
                return setInAndOutFrames((prev) => {
                    var _a;
                    return ({
                        ...prev,
                        [videoConfig.id]: {
                            ...((_a = prev[videoConfig.id]) !== null && _a !== void 0 ? _a : TimelineInOutToggle_1.defaultInOutValue),
                            outFrame: null,
                        },
                    });
                });
            }
            const minFrame = inFrame === null ? -Infinity : inFrame + 1;
            setInAndOutFrames((prev) => {
                var _a;
                return ({
                    ...prev,
                    [videoConfig.id]: {
                        ...((_a = prev[videoConfig.id]) !== null && _a !== void 0 ? _a : TimelineInOutToggle_1.defaultInOutValue),
                        outFrame: Math.max(minFrame, frame),
                    },
                });
            });
        }
    }, [
        inFrame,
        inOutDragging.dragging,
        left,
        outFrame,
        setInAndOutFrames,
        videoConfig,
        width,
    ]);
    (0, react_1.useEffect)(() => {
        if (inOutDragging.dragging === false) {
            return;
        }
        window.addEventListener('pointermove', onPointerMoveInOut);
        window.addEventListener('pointerup', onPointerUpInOut);
        return () => {
            window.removeEventListener('pointermove', onPointerMoveInOut);
            window.removeEventListener('pointerup', onPointerUpInOut);
        };
    }, [inOutDragging.dragging, onPointerMoveInOut, onPointerUpInOut]);
    const inContextMenu = (0, react_1.useMemo)(() => {
        return [
            {
                id: 'hide-in',
                keyHint: null,
                label: 'Clear In marker',
                leftItem: null,
                onClick: (_, e) => {
                    e === null || e === void 0 ? void 0 : e.stopPropagation();
                    e === null || e === void 0 ? void 0 : e.preventDefault();
                    setInAndOutFrames((prev) => {
                        var _a;
                        return ({
                            ...prev,
                            [videoConfig.id]: {
                                ...((_a = prev[videoConfig.id]) !== null && _a !== void 0 ? _a : TimelineInOutToggle_1.defaultInOutValue),
                                inFrame: null,
                            },
                        });
                    });
                },
                quickSwitcherLabel: null,
                subMenu: null,
                type: 'item',
                value: 'hide-in',
            },
        ];
    }, [setInAndOutFrames, videoConfig.id]);
    const outContextMenu = (0, react_1.useMemo)(() => {
        return [
            {
                id: 'hide-out',
                keyHint: null,
                label: 'Clear Out marker',
                leftItem: null,
                onClick: (_, e) => {
                    e === null || e === void 0 ? void 0 : e.stopPropagation();
                    e === null || e === void 0 ? void 0 : e.preventDefault();
                    setInAndOutFrames((prev) => {
                        var _a;
                        return ({
                            ...prev,
                            [videoConfig.id]: {
                                ...((_a = prev[videoConfig.id]) !== null && _a !== void 0 ? _a : TimelineInOutToggle_1.defaultInOutValue),
                                outFrame: null,
                            },
                        });
                    });
                },
                quickSwitcherLabel: null,
                subMenu: null,
                type: 'item',
                value: 'hide-out',
            },
        ];
    }, [setInAndOutFrames, videoConfig.id]);
    return (jsx_runtime_1.jsxs(jsx_runtime_1.Fragment, { children: [inFrame !== null && (jsx_runtime_1.jsx(ContextMenu_1.ContextMenu, { values: inContextMenu, onOpen: null, children: jsx_runtime_1.jsx(TimelineInOutPointerHandle_1.TimelineInOutPointerHandle, { type: "in", atFrame: inFrame, dragging: inOutDragging.dragging === 'in', onPointerDown: onInPointerDown }) })), outFrame !== null && (jsx_runtime_1.jsx(ContextMenu_1.ContextMenu, { values: outContextMenu, onOpen: null, children: jsx_runtime_1.jsx(TimelineInOutPointerHandle_1.TimelineInOutPointerHandle, { type: "out", dragging: inOutDragging.dragging === 'out', atFrame: outFrame, onPointerDown: onOutPointerDown }) }))] }));
};
const TimelineInOutDragHandlerInnerMemo = react_1.default.memo(TimelineInOutDragHandlerInner);
