// @remotion/studio/dist/components/Timeline/TimelineDragHandler.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.TimelineDragHandler = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const player_1 = require("@remotion/player");
const react_1 = __importStar(require("react"));
const remotion_1 = require("remotion");
const timeline_layout_1 = require("../../helpers/timeline-layout");
const timeline_zoom_1 = require("../../state/timeline-zoom");
const z_index_1 = require("../../state/z-index");
const is_menu_item_1 = require("../Menu/is-menu-item");
const timeline_refs_1 = require("./timeline-refs");
const timeline_scroll_logic_1 = require("./timeline-scroll-logic");
const TimelineSelection_1 = require("./TimelineSelection");
const TimelineSlider_1 = require("./TimelineSlider");
const TimelineTimeIndicators_1 = require("./TimelineTimeIndicators");
const inner = {
    overflowY: 'auto',
    overflowX: 'hidden',
};
const container = {
    userSelect: 'none',
    WebkitUserSelect: 'none',
    position: 'absolute',
    height: '100%',
    top: 0,
};
const style = {
    width: '100%',
    height: '100%',
    userSelect: 'none',
    WebkitUserSelect: 'none',
    position: 'absolute',
};
const getClientXWithScroll = (x) => {
    var _a;
    return x + ((_a = timeline_refs_1.scrollableRef.current) === null || _a === void 0 ? void 0 : _a.scrollLeft);
};
const TimelineDragHandler = () => {
    const video = remotion_1.Internals.useUnsafeVideoConfig();
    const { zoom: zoomMap } = (0, react_1.useContext)(timeline_zoom_1.TimelineZoomCtx);
    const { canvasContent } = (0, react_1.useContext)(remotion_1.Internals.CompositionManager);
    const containerStyle = (0, react_1.useMemo)(() => {
        var _a;
        if (!canvasContent || canvasContent.type !== 'composition') {
            return {};
        }
        const zoom = (_a = zoomMap[canvasContent.compositionId]) !== null && _a !== void 0 ? _a : timeline_zoom_1.TIMELINE_MIN_ZOOM;
        return {
            ...container,
            width: 100 * zoom + '%',
            height: TimelineTimeIndicators_1.TIMELINE_TIME_INDICATOR_HEIGHT,
        };
    }, [canvasContent, zoomMap]);
    if (!canvasContent || canvasContent.type !== 'composition') {
        return null;
    }
    return (jsx_runtime_1.jsx("div", { ref: timeline_refs_1.sliderAreaRef, style: containerStyle, [TimelineSelection_1.TIMELINE_SCRUBBER_ATTR]: true, children: video ? jsx_runtime_1.jsx(TimelineDragHandlerInnerMemo, {}) : null }));
};
exports.TimelineDragHandler = TimelineDragHandler;
const TimelineDragHandlerInner = () => {
    var _a;
    var _b, _c;
    const videoConfig = (0, remotion_1.useVideoConfig)();
    const size = player_1.PlayerInternals.useElementSize(timeline_refs_1.scrollableRef, {
        triggerOnWindowResize: true,
        shouldApplyCssTransforms: true,
    });
    const { isHighestContext } = (0, z_index_1.useZIndex)();
    const setFrame = remotion_1.Internals.useTimelineSetFrame();
    const width = (_b = (_a = timeline_refs_1.scrollableRef.current) === null || _a === void 0 ? void 0 : _a.scrollWidth) !== null && _b !== void 0 ? _b : 0;
    const left = (_c = size === null || size === void 0 ? void 0 : size.left) !== null && _c !== void 0 ? _c : 0;
    const [dragging, setDragging] = (0, react_1.useState)({
        dragging: false,
    });
    const { playing, play, pause, seek } = player_1.PlayerInternals.usePlayer();
    const scroller = (0, react_1.useRef)(null);
    const stopInterval = () => {
        if (scroller.current) {
            clearInterval(scroller.current);
            scroller.current = null;
        }
    };
    const onPointerDown = (0, react_1.useCallback)((e) => {
        if (e.button !== 0) {
            return;
        }
        if (!isHighestContext) {
            return;
        }
        stopInterval();
        if (!videoConfig) {
            return;
        }
        e.stopPropagation();
        document.body.style.userSelect = 'none';
        document.body.style.webkitUserSelect = 'none';
        const frame = (0, timeline_scroll_logic_1.getFrameFromX)({
            clientX: getClientXWithScroll(e.clientX) - left,
            durationInFrames: videoConfig.durationInFrames,
            width,
            extrapolate: 'clamp',
        });
        seek(frame);
        setDragging({
            dragging: true,
            wasPlaying: playing,
        });
        pause();
    }, [isHighestContext, videoConfig, left, width, seek, playing, pause]);
    const onPointerMoveScrubbing = (0, react_1.useCallback)((e) => {
        var _a;
        if (!videoConfig) {
            return;
        }
        if (!dragging.dragging) {
            return;
        }
        const isRightOfArea = e.clientX >=
            ((_a = timeline_refs_1.scrollableRef.current) === null || _a === void 0 ? void 0 : _a.clientWidth) +
                left -
                timeline_layout_1.TIMELINE_PADDING;
        const isLeftOfArea = e.clientX <= left;
        const frame = (0, timeline_scroll_logic_1.getFrameFromX)({
            clientX: getClientXWithScroll(e.clientX) - left,
            durationInFrames: videoConfig.durationInFrames,
            width,
            extrapolate: 'clamp',
        });
        if (isLeftOfArea && (0, timeline_scroll_logic_1.canScrollTimelineIntoDirection)().canScrollLeft) {
            if (scroller.current) {
                return;
            }
            const scrollEvery = () => {
                var _a;
                if (!(0, timeline_scroll_logic_1.canScrollTimelineIntoDirection)().canScrollLeft) {
                    stopInterval();
                    return;
                }
                const nextFrame = (0, timeline_scroll_logic_1.getFrameWhileScrollingLeft)({
                    durationInFrames: videoConfig.durationInFrames,
                    width,
                });
                const scrollPos = (0, timeline_scroll_logic_1.getScrollPositionForCursorOnLeftEdge)({
                    nextFrame,
                    durationInFrames: videoConfig.durationInFrames,
                });
                (_a = TimelineSlider_1.redrawTimelineSliderFast.current) === null || _a === void 0 ? void 0 : _a.draw(nextFrame);
                seek(nextFrame);
                (0, timeline_scroll_logic_1.scrollToTimelineXOffset)(scrollPos);
            };
            scrollEvery();
            scroller.current = setInterval(() => {
                scrollEvery();
            }, 100);
        }
        else if (isRightOfArea &&
            (0, timeline_scroll_logic_1.canScrollTimelineIntoDirection)().canScrollRight) {
            if (scroller.current) {
                return;
            }
            const scrollEvery = () => {
                var _a;
                if (!(0, timeline_scroll_logic_1.canScrollTimelineIntoDirection)().canScrollRight) {
                    stopInterval();
                    return;
                }
                const nextFrame = (0, timeline_scroll_logic_1.getFrameWhileScrollingRight)({
                    durationInFrames: videoConfig.durationInFrames,
                    width,
                });
                const scrollPos = (0, timeline_scroll_logic_1.getScrollPositionForCursorOnRightEdge)({
                    nextFrame,
                    durationInFrames: videoConfig.durationInFrames,
                });
                (_a = TimelineSlider_1.redrawTimelineSliderFast.current) === null || _a === void 0 ? void 0 : _a.draw(nextFrame);
                seek(nextFrame);
                (0, timeline_scroll_logic_1.scrollToTimelineXOffset)(scrollPos);
            };
            scrollEvery();
            scroller.current = setInterval(() => {
                scrollEvery();
            }, 100);
        }
        else {
            stopInterval();
            seek(frame);
        }
    }, [videoConfig, dragging.dragging, left, width, seek]);
    const onPointerUpScrubbing = (0, react_1.useCallback)((e) => {
        stopInterval();
        document.body.style.userSelect = '';
        document.body.style.webkitUserSelect = '';
        if (!videoConfig) {
            return;
        }
        if (!dragging.dragging) {
            return;
        }
        setDragging({
            dragging: false,
        });
        const frame = (0, timeline_scroll_logic_1.getFrameFromX)({
            clientX: getClientXWithScroll(e.clientX) - left,
            durationInFrames: videoConfig.durationInFrames,
            width,
            extrapolate: 'clamp',
        });
        setFrame((c) => {
            const newObj = { ...c, [videoConfig.id]: frame };
            remotion_1.Internals.persistCurrentFrame(newObj);
            return newObj;
        });
        if (dragging.wasPlaying) {
            play();
        }
    }, [dragging, left, play, videoConfig, setFrame, width]);
    (0, react_1.useEffect)(() => {
        if (!dragging.dragging) {
            return;
        }
        window.addEventListener('pointermove', onPointerMoveScrubbing);
        window.addEventListener('pointerup', onPointerUpScrubbing);
        return () => {
            window.removeEventListener('pointermove', onPointerMoveScrubbing);
            window.removeEventListener('pointerup', onPointerUpScrubbing);
        };
    }, [dragging.dragging, onPointerMoveScrubbing, onPointerUpScrubbing]);
    const ref = (0, react_1.useRef)(null);
    (0, react_1.useEffect)(() => {
        const currentRef = ref.current;
        if (!currentRef) {
            return;
        }
        const { current } = timeline_refs_1.timelineVerticalScroll;
        if (!current) {
            return;
        }
        const onScroll = () => {
            currentRef.style.top = current.scrollTop + 'px';
        };
        current.addEventListener('scroll', onScroll);
        return () => {
            current.removeEventListener('scroll', onScroll);
        };
    }, []);
    return (jsx_runtime_1.jsx("div", { ref: ref, style: style, onPointerDown: onPointerDown, children: jsx_runtime_1.jsx("div", { style: inner, className: is_menu_item_1.VERTICAL_SCROLLBAR_CLASSNAME }) }));
};
const TimelineDragHandlerInnerMemo = react_1.default.memo(TimelineDragHandlerInner);
