// @remotion/studio/dist/components/Timeline/TimelineExpandedRow.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TimelineExpandedRow = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const react_1 = __importDefault(require("react"));
const timeline_layout_1 = require("../../helpers/timeline-layout");
const timeline_row_layout_1 = require("./timeline-row-layout");
const TimelineEffectItem_1 = require("./TimelineEffectItem");
const TimelineEffectPropItem_1 = require("./TimelineEffectPropItem");
const TimelineExpandArrowButton_1 = require("./TimelineExpandArrowButton");
const TimelineLayerEye_1 = require("./TimelineLayerEye");
const TimelineRowChrome_1 = require("./TimelineRowChrome");
const TimelineSelection_1 = require("./TimelineSelection");
const TimelineSequencePropItem_1 = require("./TimelineSequencePropItem");
const rowLabel = {
    fontSize: 12,
    color: 'rgba(255, 255, 255, 0.8)',
    userSelect: 'none',
};
const TimelineExpandedRow = ({ node, depth, nestedDepth, rowDepthBase, getIsExpanded, toggleTrack, validatedLocation, nodePath, schema, keyframeDisplayOffset, keyframeControlsMode, }) => {
    const rowDepth = (rowDepthBase !== null && rowDepthBase !== void 0 ? rowDepthBase : (0, timeline_row_layout_1.getExpandedRowDepth)({ nestedDepth, treeDepth: 0 })) + depth;
    const selection = (0, TimelineSelection_1.useTimelineRowSelection)(node.nodePathInfo);
    const labelStyle = react_1.default.useMemo(() => ({
        ...rowLabel,
        ...(0, TimelineSelection_1.getTimelineSelectedLabelStyle)(selection.selected, true),
        alignSelf: 'stretch',
        alignItems: 'center',
        color: (0, TimelineSelection_1.getTimelineColor)(selection.selected, true),
        display: 'flex',
        flex: 1,
        minWidth: 0,
        paddingRight: timeline_layout_1.EXPANDED_SECTION_PADDING_RIGHT,
    }), [selection.selected]);
    if (node.kind === 'group') {
        if (node.effectInfo) {
            return (
            // A single effect
            jsx_runtime_1.jsx(TimelineEffectItem_1.TimelineEffectItem, { label: node.label, nodePathInfo: node.nodePathInfo, effectIndex: node.effectInfo.effectIndex, effectSchema: node.effectInfo.effectSchema, documentationLink: node.effectInfo.documentationLink, nodePath: nodePath, validatedLocation: validatedLocation, rowDepth: rowDepth, getIsExpanded: getIsExpanded, toggleTrack: toggleTrack }));
        }
        // Group like "Effects"
        const isExpanded = getIsExpanded(node.nodePathInfo);
        return (jsx_runtime_1.jsx(TimelineRowChrome_1.TimelineRowChrome, { depth: rowDepth, eye: jsx_runtime_1.jsx(TimelineLayerEye_1.TimelineLayerEyeSpacer, {}), arrow: jsx_runtime_1.jsx(TimelineExpandArrowButton_1.TimelineExpandArrowButton, { isExpanded: isExpanded, onClick: () => toggleTrack(node.nodePathInfo), label: `${node.label} section`, disabled: false }), style: {
                height: timeline_layout_1.TREE_GROUP_ROW_HEIGHT,
            }, selected: selection.selected, selectable: selection.selectable, selectionItem: selection.selectionItem, onSelect: selection.onSelect, showSelectedBackground: true, containsSelection: false, outerHeight: null, children: jsx_runtime_1.jsx("span", { style: labelStyle, children: node.label }) }));
    }
    if (node.field) {
        if (node.field.kind === 'effect-field') {
            return (jsx_runtime_1.jsx(TimelineEffectPropItem_1.TimelineEffectPropItem, { field: node.field, validatedLocation: validatedLocation, rowDepth: rowDepth, nodePath: nodePath, nodePathInfo: node.nodePathInfo, keyframeDisplayOffset: keyframeDisplayOffset, keyframeControlsMode: keyframeControlsMode }));
        }
        if (node.field.kind === 'sequence-field') {
            return (jsx_runtime_1.jsx(TimelineSequencePropItem_1.TimelineSequencePropItem, { field: node.field, validatedLocation: validatedLocation, rowDepth: rowDepth, nodePath: nodePath, nodePathInfo: node.nodePathInfo, schema: schema, keyframeDisplayOffset: keyframeDisplayOffset, keyframeControlsMode: keyframeControlsMode }));
        }
        throw new Error('Unexpected field kind: ' + JSON.stringify(node.field));
    }
    return (jsx_runtime_1.jsx(TimelineRowChrome_1.TimelineRowChrome, { depth: rowDepth, eye: jsx_runtime_1.jsx(TimelineLayerEye_1.TimelineLayerEyeSpacer, {}), arrow: jsx_runtime_1.jsx(TimelineExpandArrowButton_1.TimelineExpandArrowSpacer, {}), style: {
            height: (0, timeline_layout_1.getTreeRowHeight)(node),
        }, selected: selection.selected, selectable: selection.selectable, selectionItem: selection.selectionItem, onSelect: selection.onSelect, showSelectedBackground: true, containsSelection: false, outerHeight: null, children: jsx_runtime_1.jsx("span", { style: labelStyle, children: node.label }) }));
};
exports.TimelineExpandedRow = TimelineExpandedRow;
