// @remotion/studio/dist/components/Timeline/TimelineTransformOriginField.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TimelineTransformOriginField = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const react_1 = require("react");
const InputDragger_1 = require("../NewComposition/InputDragger");
const timeline_field_display_utils_1 = require("./timeline-field-display-utils");
const timeline_field_utils_1 = require("./timeline-field-utils");
const TimelineSchemaField_1 = require("./TimelineSchemaField");
const transform_origin_utils_1 = require("./transform-origin-utils");
const leftDraggerStyle = {
    paddingLeft: 0,
};
const rightDraggerStyle = {
    paddingRight: 0,
};
const containerStyle = {
    display: 'flex',
    gap: 4,
};
const TimelineTransformOriginField = ({ field, propStatus, effectiveValue, onSave, onDragValueChange, onDragEnd, }) => {
    const [dragX, setDragX] = (0, react_1.useState)(null);
    const [dragY, setDragY] = (0, react_1.useState)(null);
    const parsed = (0, react_1.useMemo)(() => (0, transform_origin_utils_1.parseTransformOrigin)(effectiveValue), [effectiveValue]);
    const percent = (0, react_1.useMemo)(() => (parsed === null ? null : (0, transform_origin_utils_1.parsedTransformOriginToPercent)(parsed)), [parsed]);
    const configuredStep = field.fieldSchema.type === 'transform-origin'
        ? field.fieldSchema.step
        : undefined;
    const step = configuredStep !== null && configuredStep !== void 0 ? configuredStep : 1;
    const decimalPlaces = (0, react_1.useMemo)(() => (0, timeline_field_utils_1.getTimelineDisplayDecimalPlaces)({
        defaultDecimalPlaces: 2,
        step: configuredStep,
    }), [configuredStep]);
    const formatter = (0, react_1.useCallback)((v) => {
        return (0, timeline_field_display_utils_1.formatTimelineFieldValueForDisplay)({
            fieldSchema: field.fieldSchema,
            value: v,
        });
    }, [field.fieldSchema]);
    const serialize = (0, react_1.useCallback)((x, y) => {
        var _a;
        return (0, transform_origin_utils_1.serializeTransformOrigin)({
            uv: (0, transform_origin_utils_1.transformOriginPercentToUv)([x, y]),
            z: (_a = parsed === null || parsed === void 0 ? void 0 : parsed.z) !== null && _a !== void 0 ? _a : null,
            decimalPlaces,
        });
    }, [decimalPlaces, parsed === null || parsed === void 0 ? void 0 : parsed.z]);
    const onXChange = (0, react_1.useCallback)((newVal) => {
        if (percent === null) {
            return;
        }
        setDragX(newVal);
        const currentY = dragY !== null && dragY !== void 0 ? dragY : percent[1];
        onDragValueChange(serialize(newVal, currentY));
    }, [dragY, onDragValueChange, percent, serialize]);
    const onXChangeEnd = (0, react_1.useCallback)((newVal) => {
        if (percent === null) {
            return;
        }
        const currentY = dragY !== null && dragY !== void 0 ? dragY : percent[1];
        const value = serialize(newVal, currentY);
        if (value !== propStatus.codeValue) {
            onSave(value).finally(() => {
                setDragX(null);
                onDragEnd();
            });
        }
        else {
            setDragX(null);
            onDragEnd();
        }
    }, [dragY, onDragEnd, onSave, percent, propStatus.codeValue, serialize]);
    const onXTextChange = (0, react_1.useCallback)((newVal) => {
        if (percent === null) {
            return;
        }
        const parsedValue = Number(newVal);
        if (!Number.isNaN(parsedValue)) {
            const currentY = dragY !== null && dragY !== void 0 ? dragY : percent[1];
            const value = serialize(parsedValue, currentY);
            if (value !== propStatus.codeValue) {
                setDragX(parsedValue);
                onSave(value).finally(() => {
                    setDragX(null);
                });
            }
        }
    }, [dragY, onSave, percent, propStatus.codeValue, serialize]);
    const onYChange = (0, react_1.useCallback)((newVal) => {
        if (percent === null) {
            return;
        }
        setDragY(newVal);
        const currentX = dragX !== null && dragX !== void 0 ? dragX : percent[0];
        onDragValueChange(serialize(currentX, newVal));
    }, [dragX, onDragValueChange, percent, serialize]);
    const onYChangeEnd = (0, react_1.useCallback)((newVal) => {
        if (percent === null) {
            return;
        }
        const currentX = dragX !== null && dragX !== void 0 ? dragX : percent[0];
        const value = serialize(currentX, newVal);
        if (value !== propStatus.codeValue) {
            onSave(value).finally(() => {
                setDragY(null);
                onDragEnd();
            });
        }
        else {
            setDragY(null);
            onDragEnd();
        }
    }, [dragX, onDragEnd, onSave, percent, propStatus.codeValue, serialize]);
    const onYTextChange = (0, react_1.useCallback)((newVal) => {
        if (percent === null) {
            return;
        }
        const parsedValue = Number(newVal);
        if (!Number.isNaN(parsedValue)) {
            const currentX = dragX !== null && dragX !== void 0 ? dragX : percent[0];
            const value = serialize(currentX, parsedValue);
            if (value !== propStatus.codeValue) {
                setDragY(parsedValue);
                onSave(value).finally(() => {
                    setDragY(null);
                });
            }
        }
    }, [dragX, onSave, percent, propStatus.codeValue, serialize]);
    if (percent === null) {
        return jsx_runtime_1.jsx(TimelineSchemaField_1.UnsupportedStatus, { label: "unsupported origin" });
    }
    return (jsx_runtime_1.jsxs("span", { style: containerStyle, children: [
            jsx_runtime_1.jsx(InputDragger_1.InputDragger, { type: "number", value: dragX !== null && dragX !== void 0 ? dragX : percent[0], style: leftDraggerStyle, status: "ok", small: true, onValueChange: onXChange, onValueChangeEnd: onXChangeEnd, onTextChange: onXTextChange, min: -Infinity, max: Infinity, step: step, formatter: formatter, rightAlign: false, snapToStep: false, dragDecimalPlaces: decimalPlaces }), jsx_runtime_1.jsx("div", { style: { marginLeft: -6, marginRight: -6 } }), jsx_runtime_1.jsx(InputDragger_1.InputDragger, { type: "number", value: dragY !== null && dragY !== void 0 ? dragY : percent[1], style: rightDraggerStyle, status: "ok", small: true, onValueChange: onYChange, onValueChangeEnd: onYChangeEnd, onTextChange: onYTextChange, min: -Infinity, max: Infinity, step: step, formatter: formatter, rightAlign: false, snapToStep: false, dragDecimalPlaces: decimalPlaces })
        ] }));
};
exports.TimelineTransformOriginField = TimelineTransformOriginField;
