// @remotion/studio/dist/components/Timeline/TimelineExpandArrowButton.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.TimelineExpandArrowSpacer = exports.TimelineExpandArrowButton = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const react_1 = __importStar(require("react"));
const arrowButton = {
    background: 'none',
    border: 'none',
    color: 'white',
    cursor: 'pointer',
    padding: 0,
    display: 'flex',
    alignItems: 'center',
    justifyContent: 'center',
    width: 12,
    height: 12,
    flexShrink: 0,
    fontSize: 8,
    marginRight: 4,
    userSelect: 'none',
    outline: 'none',
    lineHeight: 1,
};
const arrowSpacer = {
    ...arrowButton,
    cursor: 'default',
};
const svgStyle = { display: 'block' };
const TimelineExpandArrowButton = ({ isExpanded, onClick, label, disabled = false }) => {
    const handleClick = react_1.default.useCallback((e) => {
        e.stopPropagation();
        onClick();
    }, [onClick]);
    const stopPropagation = react_1.default.useCallback((e) => {
        e.stopPropagation();
    }, []);
    const style = (0, react_1.useMemo)(() => {
        return {
            ...arrowButton,
            transform: isExpanded ? 'rotate(90deg)' : 'rotate(0deg)',
            cursor: disabled ? 'default' : 'pointer',
            opacity: disabled ? 0.5 : 1,
        };
    }, [isExpanded, disabled]);
    return (jsx_runtime_1.jsx("button", { type: "button", style: style, onClick: handleClick, onDoubleClick: stopPropagation, onPointerDown: stopPropagation, disabled: disabled, "aria-expanded": isExpanded, "aria-label": `${isExpanded ? 'Collapse' : 'Expand'} ${label}`, children: jsx_runtime_1.jsx("svg", { width: "12", height: "12", viewBox: "0 0 8 8", style: svgStyle, children: jsx_runtime_1.jsx("path", { d: "M2 1L6 4L2 7Z", fill: "#ccc" }) }) }));
};
exports.TimelineExpandArrowButton = TimelineExpandArrowButton;
const TimelineExpandArrowSpacer = () => {
    return jsx_runtime_1.jsx("div", { style: arrowSpacer });
};
exports.TimelineExpandArrowSpacer = TimelineExpandArrowSpacer;
