// @remotion/studio/dist/components/Timeline/TimelineDeleteKeybindings.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TimelineDeleteKeybindings = void 0;
const studio_shared_1 = require("@remotion/studio-shared");
const react_1 = require("react");
const remotion_1 = require("remotion");
const client_id_1 = require("../../helpers/client-id");
const use_keybinding_1 = require("../../helpers/use-keybinding");
const editor_guides_1 = require("../../state/editor-guides");
const ConfirmationDialog_1 = require("../ConfirmationDialog");
const delete_selected_timeline_item_1 = require("./delete-selected-timeline-item");
const duplicate_selected_timeline_item_1 = require("./duplicate-selected-timeline-item");
const reset_selected_timeline_props_1 = require("./reset-selected-timeline-props");
const TimelineSelection_1 = require("./TimelineSelection");
const update_selected_easing_1 = require("./update-selected-easing");
const TimelineDeleteKeybindings = () => {
    const keybindings = (0, use_keybinding_1.useKeybinding)();
    const { previewServerState } = (0, react_1.useContext)(client_id_1.StudioServerConnectionCtx);
    const sequencesRef = (0, react_1.useContext)(remotion_1.Internals.SequenceManagerRefContext);
    const { overrideIdToNodePathMappings } = (0, react_1.useContext)(remotion_1.Internals.OverrideIdsToNodePathsGettersContext);
    const propStatusesRef = (0, react_1.useContext)(remotion_1.Internals.VisualModePropStatusesRefContext);
    const { setPropStatuses } = (0, react_1.useContext)(remotion_1.Internals.VisualModeSettersContext);
    const { setGuidesList } = (0, react_1.useContext)(editor_guides_1.EditorShowGuidesContext);
    const { canSelect } = (0, TimelineSelection_1.useTimelineSelection)();
    const currentSelection = (0, TimelineSelection_1.useCurrentTimelineSelectionStateAsRef)();
    const confirm = (0, ConfirmationDialog_1.useConfirmationDialog)();
    (0, react_1.useEffect)(() => {
        if (!canSelect || previewServerState.type !== 'connected') {
            return;
        }
        const { clientId } = previewServerState;
        const handleDelete = () => {
            const { selectedItems, clearSelection, selectItems } = currentSelection.current;
            const sequences = sequencesRef.current;
            const propStatuses = propStatusesRef.current;
            if (selectedItems.length === 0) {
                return;
            }
            const selectedGuide = selectedItems.find((item) => item.type === 'guide');
            if (selectedGuide) {
                setGuidesList((prevGuides) => {
                    const newGuides = prevGuides.filter((guide) => guide.id !== selectedGuide.guideId);
                    (0, editor_guides_1.persistGuidesList)(newGuides);
                    return newGuides;
                });
                clearSelection();
                return;
            }
            const deletePromise = (0, delete_selected_timeline_item_1.deleteSelectedTimelineItems)({
                selections: selectedItems,
                sequences,
                overrideIdsToNodePaths: overrideIdToNodePathMappings,
                setPropStatuses,
                clientId,
                confirm,
            });
            if (deletePromise !== null) {
                deletePromise
                    .then((deleted) => {
                    if (deleted) {
                        const nextSelection = (0, delete_selected_timeline_item_1.getTimelineSelectionAfterDeletingItems)(selectedItems);
                        if (nextSelection.length === 0) {
                            clearSelection();
                        }
                        else {
                            selectItems(nextSelection);
                        }
                    }
                })
                    .catch(() => undefined);
                return;
            }
            const easingSelections = (0, update_selected_easing_1.getEasingSelections)(selectedItems);
            if (easingSelections.length === selectedItems.length) {
                const resetEasingPromise = (0, update_selected_easing_1.updateSelectedTimelineEasings)({
                    selections: easingSelections,
                    sequences,
                    overrideIdsToNodePaths: overrideIdToNodePathMappings,
                    propStatuses,
                    setPropStatuses,
                    clientId,
                    easing: studio_shared_1.LINEAR_KEYFRAME_EASING,
                });
                if (resetEasingPromise !== null) {
                    resetEasingPromise.catch(() => undefined);
                    return;
                }
            }
            const resetPromise = (0, reset_selected_timeline_props_1.resetSelectedTimelineProps)({
                selections: selectedItems,
                sequences,
                overrideIdsToNodePaths: overrideIdToNodePathMappings,
                propStatuses,
                setPropStatuses,
                clientId,
            });
            if (resetPromise !== null) {
                resetPromise.catch(() => undefined);
            }
        };
        const backspace = keybindings.registerKeybinding({
            event: 'keydown',
            key: 'Backspace',
            callback: handleDelete,
            commandCtrlKey: false,
            preventDefault: true,
            triggerIfInputFieldFocused: false,
            keepRegisteredWhenNotHighestContext: false,
        });
        const deleteKey = keybindings.registerKeybinding({
            event: 'keydown',
            key: 'Delete',
            callback: handleDelete,
            commandCtrlKey: false,
            preventDefault: true,
            triggerIfInputFieldFocused: false,
            keepRegisteredWhenNotHighestContext: false,
        });
        const duplicate = keybindings.registerKeybinding({
            event: 'keydown',
            key: 'd',
            callback: () => {
                const { selectedItems } = currentSelection.current;
                if (selectedItems.length === 0) {
                    return;
                }
                const duplicatePromise = (0, duplicate_selected_timeline_item_1.duplicateSelectedTimelineItems)({
                    selections: selectedItems,
                    confirm,
                });
                if (duplicatePromise === null) {
                    return;
                }
                duplicatePromise.catch(() => undefined);
            },
            commandCtrlKey: true,
            preventDefault: true,
            triggerIfInputFieldFocused: false,
            keepRegisteredWhenNotHighestContext: false,
        });
        return () => {
            backspace.unregister();
            deleteKey.unregister();
            duplicate.unregister();
        };
    }, [
        canSelect,
        confirm,
        currentSelection,
        keybindings,
        overrideIdToNodePathMappings,
        propStatusesRef,
        previewServerState,
        sequencesRef,
        setGuidesList,
        setPropStatuses,
    ]);
    return null;
};
exports.TimelineDeleteKeybindings = TimelineDeleteKeybindings;
