// @remotion/studio/dist/components/Timeline/timeline-field-utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.draggerStyle = exports.normalizeTimelineNumber = exports.formatTimelineNumber = exports.roundToDecimalPlaces = exports.getTimelineDisplayDecimalPlaces = exports.getDecimalPlaces = void 0;
const getDecimalPlaces = (num) => {
    const [coefficient, exponent] = String(num).toLowerCase().split('e');
    const decimalIndex = coefficient.indexOf('.');
    const coefficientDecimals = decimalIndex === -1 ? 0 : coefficient.length - decimalIndex - 1;
    if (exponent === undefined) {
        return coefficientDecimals;
    }
    return Math.max(0, coefficientDecimals - Number(exponent));
};
exports.getDecimalPlaces = getDecimalPlaces;
const getTimelineDisplayDecimalPlaces = ({ defaultDecimalPlaces, step, }) => {
    return Math.max(defaultDecimalPlaces, step === undefined ? 0 : (0, exports.getDecimalPlaces)(step));
};
exports.getTimelineDisplayDecimalPlaces = getTimelineDisplayDecimalPlaces;
const roundToDecimalPlaces = (value, decimalPlaces) => {
    const factor = 10 ** decimalPlaces;
    const rounded = Math.round(value * factor) / factor;
    return Object.is(rounded, -0) ? 0 : rounded;
};
exports.roundToDecimalPlaces = roundToDecimalPlaces;
const formatTimelineNumber = ({ decimalPlaces, fixed, value, }) => {
    const rounded = (0, exports.roundToDecimalPlaces)(Number(value), decimalPlaces);
    return fixed && decimalPlaces > 0
        ? rounded.toFixed(decimalPlaces)
        : String(rounded);
};
exports.formatTimelineNumber = formatTimelineNumber;
const normalizeTimelineNumber = (value) => {
    return (0, exports.roundToDecimalPlaces)(value, 6);
};
exports.normalizeTimelineNumber = normalizeTimelineNumber;
exports.draggerStyle = {
    width: 80,
};
