// @remotion/studio/dist/components/Timeline/TimelineScrollable.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TimelineScrollable = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const react_1 = require("react");
const is_menu_item_1 = require("../Menu/is-menu-item");
const timeline_refs_1 = require("./timeline-refs");
const TimelineSelection_1 = require("./TimelineSelection");
const outer = {
    width: '100%',
    height: '100%',
    overflowX: 'auto',
    overflowY: 'hidden',
    position: 'relative',
    backgroundColor: TimelineSelection_1.TIMELINE_BACKGROUND,
};
const marqueeStyle = {
    backgroundColor: 'rgba(70, 130, 255, 0.16)',
    border: '1px solid rgba(70, 130, 255, 0.75)',
    boxSizing: 'border-box',
    pointerEvents: 'none',
    position: 'fixed',
    zIndex: 10,
};
const TimelineScrollable = ({ children }) => {
    const { marqueeRect, onPointerDownCapture } = (0, TimelineSelection_1.useTimelineMarqueeSelection)();
    const containerStyle = (0, react_1.useMemo)(() => {
        return {
            width: '100%',
            minHeight: '100%',
        };
    }, []);
    return (jsx_runtime_1.jsxs("div", { ref: timeline_refs_1.scrollableRef, style: outer, className: is_menu_item_1.HORIZONTAL_SCROLLBAR_CLASSNAME, onPointerDownCapture: onPointerDownCapture, children: [
            jsx_runtime_1.jsx("div", { style: containerStyle, children: children }), marqueeRect === null ? null : (jsx_runtime_1.jsx("div", { style: {
                    ...marqueeStyle,
                    height: marqueeRect.bottom - marqueeRect.top,
                    left: marqueeRect.left,
                    top: marqueeRect.top,
                    width: marqueeRect.right - marqueeRect.left,
                } }))] }));
};
exports.TimelineScrollable = TimelineScrollable;
