// @remotion/studio/dist/components/Timeline/transform-origin-utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.transformOriginPercentToUv = exports.parsedTransformOriginToPercent = exports.serializeTransformOrigin = exports.parsedTransformOriginToUv = exports.axisValueToUv = exports.parseTransformOrigin = void 0;
const timeline_field_utils_1 = require("./timeline-field-utils");
const cssNumberWithUnit = /^([+-]?(?:\d+\.?\d*|\.\d+))(px|%)$/i;
const cssLength = /^([+-]?(?:\d+\.?\d*|\.\d+))[a-zA-Z]+$/;
const keywords = new Set(['left', 'center', 'right', 'top', 'bottom']);
const center = { value: 50, unit: '%' };
const keywordOptions = (keyword) => {
    if (keyword === 'left') {
        return [{ axis: 'x', value: { value: 0, unit: '%' } }];
    }
    if (keyword === 'right') {
        return [{ axis: 'x', value: { value: 100, unit: '%' } }];
    }
    if (keyword === 'top') {
        return [{ axis: 'y', value: { value: 0, unit: '%' } }];
    }
    if (keyword === 'bottom') {
        return [{ axis: 'y', value: { value: 100, unit: '%' } }];
    }
    return [
        { axis: 'x', value: center },
        { axis: 'y', value: center },
    ];
};
const parseLengthPercentage = (token) => {
    const match = cssNumberWithUnit.exec(token);
    if (!match) {
        return null;
    }
    const value = Number(match[1]);
    if (!Number.isFinite(value)) {
        return null;
    }
    return { value, unit: match[2].toLowerCase() };
};
const parseToken = (token) => {
    const lower = token.toLowerCase();
    if (keywords.has(lower)) {
        return { type: 'keyword', keyword: lower };
    }
    const value = parseLengthPercentage(token);
    return value === null ? null : { type: 'length-percentage', value };
};
const parseTwoKeywords = (first, second) => {
    var _a;
    const candidates = [];
    for (const firstOption of keywordOptions(first)) {
        for (const secondOption of keywordOptions(second)) {
            if (firstOption.axis === secondOption.axis) {
                continue;
            }
            candidates.push(firstOption.axis === 'x'
                ? [firstOption.value, secondOption.value]
                : [secondOption.value, firstOption.value]);
        }
    }
    return (_a = candidates[0]) !== null && _a !== void 0 ? _a : null;
};
const parseXY = (parts) => {
    if (parts.length === 1) {
        const token = parseToken(parts[0]);
        if (token === null) {
            return null;
        }
        if (token.type === 'length-percentage') {
            return [token.value, center];
        }
        if (token.keyword === 'top' || token.keyword === 'bottom') {
            return [center, keywordOptions(token.keyword)[0].value];
        }
        return [keywordOptions(token.keyword)[0].value, center];
    }
    const first = parseToken(parts[0]);
    const second = parseToken(parts[1]);
    if (first === null || second === null) {
        return null;
    }
    if (first.type === 'length-percentage' &&
        second.type === 'length-percentage') {
        return [first.value, second.value];
    }
    if (first.type === 'keyword' && second.type === 'keyword') {
        return parseTwoKeywords(first.keyword, second.keyword);
    }
    const keyword = first.type === 'keyword'
        ? first
        : second.type === 'keyword'
            ? second
            : null;
    const length = first.type === 'length-percentage'
        ? first.value
        : second.type === 'length-percentage'
            ? second.value
            : null;
    if (keyword === null || length === null) {
        return null;
    }
    const keywordIsFirst = first.type === 'keyword';
    if (keyword.keyword === 'left' || keyword.keyword === 'right') {
        return keywordIsFirst
            ? [keywordOptions(keyword.keyword)[0].value, length]
            : null;
    }
    if (keyword.keyword === 'top' || keyword.keyword === 'bottom') {
        return [length, keywordOptions(keyword.keyword)[0].value];
    }
    return keywordIsFirst ? [center, length] : [length, center];
};
const isSupportedZ = (token) => {
    if (token.includes('%')) {
        return false;
    }
    return cssLength.test(token);
};
const parseTransformOrigin = (value) => {
    var _a;
    if (typeof value !== 'string') {
        return null;
    }
    const parts = value.trim().split(/\s+/);
    if (parts.length < 1 || parts.length > 3 || parts[0] === '') {
        return null;
    }
    const xy = parseXY(parts.length === 1 ? [parts[0]] : [parts[0], parts[1]]);
    if (xy === null) {
        return null;
    }
    const z = (_a = parts[2]) !== null && _a !== void 0 ? _a : null;
    if (z !== null && !isSupportedZ(z)) {
        return null;
    }
    return { x: xy[0], y: xy[1], z };
};
exports.parseTransformOrigin = parseTransformOrigin;
const axisValueToUv = (axis, size) => {
    if (!Number.isFinite(size) || size <= 0) {
        return null;
    }
    if (axis.unit === '%') {
        return axis.value / 100;
    }
    return axis.value / size;
};
exports.axisValueToUv = axisValueToUv;
const parsedTransformOriginToUv = ({ parsed, width, height, }) => {
    const x = (0, exports.axisValueToUv)(parsed.x, width);
    const y = (0, exports.axisValueToUv)(parsed.y, height);
    return x === null || y === null ? null : [x, y];
};
exports.parsedTransformOriginToUv = parsedTransformOriginToUv;
const formatPercent = (value, decimalPlaces) => {
    const rounded = (0, timeline_field_utils_1.roundToDecimalPlaces)((0, timeline_field_utils_1.normalizeTimelineNumber)(value * 100), decimalPlaces);
    return String(Object.is(rounded, -0) ? 0 : rounded);
};
const serializeTransformOrigin = ({ uv, z, decimalPlaces = 2, }) => {
    const xy = `${formatPercent(uv[0], decimalPlaces)}% ${formatPercent(uv[1], decimalPlaces)}%`;
    return z === null ? xy : `${xy} ${z}`;
};
exports.serializeTransformOrigin = serializeTransformOrigin;
const parsedTransformOriginToPercent = (parsed) => {
    if (parsed.x.unit !== '%' || parsed.y.unit !== '%') {
        return null;
    }
    return [parsed.x.value, parsed.y.value];
};
exports.parsedTransformOriginToPercent = parsedTransformOriginToPercent;
const transformOriginPercentToUv = (percent) => [percent[0] / 100, percent[1] / 100];
exports.transformOriginPercentToUv = transformOriginPercentToUv;
