// @remotion/studio/dist/components/Timeline/use-sequence-freeze-frame-menu-item.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.useSequenceFreezeFrameMenuItem = exports.shouldShowFreezeFrameMenuItem = void 0;
const react_1 = require("react");
const save_sequence_prop_1 = require("./save-sequence-prop");
const shouldShowFreezeFrameMenuItem = (sequence) => {
    return sequence.type !== 'audio';
};
exports.shouldShowFreezeFrameMenuItem = shouldShowFreezeFrameMenuItem;
const useSequenceFreezeFrameMenuItem = ({ clientId, nodePath, propStatusesForOverride, sequence, sequenceFrameOffset, setPropStatuses, timelinePosition, validatedSource, }) => {
    const freezeStatus = propStatusesForOverride === null || propStatusesForOverride === void 0 ? void 0 : propStatusesForOverride.freeze;
    const isFrozen = (freezeStatus === null || freezeStatus === void 0 ? void 0 : freezeStatus.status) === 'static' &&
        typeof freezeStatus.codeValue === 'number';
    const canToggleFreeze = clientId !== null &&
        Boolean(sequence.controls) &&
        nodePath !== null &&
        validatedSource !== null &&
        freezeStatus !== undefined &&
        freezeStatus !== null &&
        freezeStatus.status === 'static';
    const onToggleFreezeFrame = (0, react_1.useCallback)(() => {
        if (!canToggleFreeze ||
            !sequence.controls ||
            nodePath === null ||
            validatedSource === null ||
            clientId === null) {
            return;
        }
        const rawFreezeFrame = Math.round(timelinePosition - sequence.from + sequenceFrameOffset);
        const maxFrame = Number.isFinite(sequence.duration)
            ? Math.max(0, sequence.duration - 1)
            : rawFreezeFrame;
        const freezeFrame = Math.min(Math.max(0, rawFreezeFrame), maxFrame);
        const remove = isFrozen;
        (0, save_sequence_prop_1.saveSequenceProps)({
            changes: [
                {
                    fileName: validatedSource,
                    nodePath,
                    fieldKey: 'freeze',
                    value: remove ? null : freezeFrame,
                    defaultValue: null,
                    schema: sequence.controls.schema,
                },
            ],
            setPropStatuses,
            clientId,
            undoLabel: remove ? 'Unfreeze sequence' : 'Freeze sequence',
            redoLabel: remove ? 'Freeze sequence again' : 'Unfreeze sequence again',
        });
    }, [
        canToggleFreeze,
        clientId,
        isFrozen,
        nodePath,
        sequence.controls,
        sequence.duration,
        sequence.from,
        sequenceFrameOffset,
        setPropStatuses,
        timelinePosition,
        validatedSource,
    ]);
    return (0, react_1.useMemo)(() => (0, exports.shouldShowFreezeFrameMenuItem)(sequence)
        ? {
            type: 'item',
            id: 'toggle-freeze-frame',
            keyHint: null,
            label: isFrozen ? 'Unfreeze frame' : 'Freeze frame',
            leftItem: null,
            disabled: !canToggleFreeze,
            onClick: onToggleFreezeFrame,
            quickSwitcherLabel: null,
            subMenu: null,
            value: 'toggle-freeze-frame',
        }
        : null, [canToggleFreeze, isFrozen, onToggleFreezeFrame, sequence]);
};
exports.useSequenceFreezeFrameMenuItem = useSequenceFreezeFrameMenuItem;
