// @remotion/studio/dist/components/Timeline/use-expanded-track-keyframe-rows.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.useExpandedTrackKeyframeRows = void 0;
const react_1 = require("react");
const remotion_1 = require("remotion");
const timeline_layout_1 = require("../../helpers/timeline-layout");
const timeline_node_path_key_1 = require("../../helpers/timeline-node-path-key");
const ExpandedTracksProvider_1 = require("../ExpandedTracksProvider");
const get_node_keyframes_1 = require("./get-node-keyframes");
const timeline_expanded_filter_1 = require("./timeline-expanded-filter");
const TimelineSelection_1 = require("./TimelineSelection");
const canEditEasingForInterpolationFunction = (interpolationFunction) => interpolationFunction === 'interpolate' ||
    interpolationFunction === 'interpolateColors';
const getNodeCanEditEasing = ({ node, nodePath, propStatuses, }) => {
    var _a;
    if (node.kind !== 'field' || node.field === null) {
        return false;
    }
    if (node.field.kind === 'sequence-field') {
        const sequencePropStatus = (_a = remotion_1.Internals.getPropStatusesCtx(propStatuses, nodePath)) === null || _a === void 0 ? void 0 : _a[node.field.key];
        return ((sequencePropStatus === null || sequencePropStatus === void 0 ? void 0 : sequencePropStatus.status) === 'keyframed' &&
            canEditEasingForInterpolationFunction(sequencePropStatus.interpolationFunction));
    }
    const effectStatus = remotion_1.Internals.getEffectPropStatusesCtx({
        propStatuses,
        nodePath,
        effectIndex: node.field.effectIndex,
    });
    const effectPropStatus = effectStatus.type === 'can-update-effect'
        ? effectStatus.props[node.field.key]
        : null;
    return ((effectPropStatus === null || effectPropStatus === void 0 ? void 0 : effectPropStatus.status) === 'keyframed' &&
        canEditEasingForInterpolationFunction(effectPropStatus.interpolationFunction));
};
const useExpandedTrackKeyframeRows = ({ sequence, nodePathInfo, keyframeDisplayOffset, }) => {
    const { getIsExpanded } = (0, react_1.useContext)(ExpandedTracksProvider_1.ExpandedTracksGetterContext);
    const { propStatuses } = (0, react_1.useContext)(remotion_1.Internals.VisualModePropStatusesContext);
    const { getDragOverrides, getEffectDragOverrides } = (0, react_1.useContext)(remotion_1.Internals.VisualModeDragOverridesContext);
    const { selectedItems } = (0, TimelineSelection_1.useTimelineSelection)();
    const timelinePosition = remotion_1.Internals.Timeline.useTimelinePosition();
    const tree = (0, react_1.useMemo)(() => (0, timeline_layout_1.buildTimelineTree)({
        sequence,
        nodePathInfo,
        getDragOverrides,
        getEffectDragOverrides,
        propStatuses,
    }), [
        propStatuses,
        getDragOverrides,
        getEffectDragOverrides,
        nodePathInfo,
        sequence,
    ]);
    const selectedRowKeys = (0, react_1.useMemo)(() => (0, timeline_expanded_filter_1.getSelectedTimelineExpandedRowKeys)(selectedItems), [selectedItems]);
    const filteredTree = (0, react_1.useMemo)(() => (0, timeline_expanded_filter_1.filterTimelineExpandedTree)({
        nodes: tree,
        shouldShowNode: (node) => (0, timeline_expanded_filter_1.isTimelineExpandedNodeSelected)({
            nodePathInfo: node.nodePathInfo,
            selectedRowKeys,
        }) ||
            (0, get_node_keyframes_1.getNodeHasKeyframes)({
                node,
                nodePath: nodePathInfo.sequenceSubscriptionKey,
                propStatuses,
                getDragOverrides,
                getEffectDragOverrides,
            }),
    }), [
        getDragOverrides,
        getEffectDragOverrides,
        nodePathInfo.sequenceSubscriptionKey,
        propStatuses,
        selectedRowKeys,
        tree,
    ]);
    const flat = (0, react_1.useMemo)(() => (0, timeline_layout_1.flattenVisibleTreeNodes)({ nodes: filteredTree, getIsExpanded }), [filteredTree, getIsExpanded]);
    const expandedHeight = (0, react_1.useMemo)(() => {
        const totalRowsHeight = flat.reduce((sum, { node }) => sum + (0, timeline_layout_1.getTreeRowHeight)(node), 0);
        const separators = Math.max(0, flat.length - 1);
        return totalRowsHeight + separators;
    }, [flat]);
    const nodeKeyframes = (0, react_1.useMemo)(() => new Map(flat.map(({ node }) => [
        (0, timeline_node_path_key_1.timelineNodePathInfoToKey)(node.nodePathInfo),
        (0, get_node_keyframes_1.getNodeKeyframes)({
            node,
            nodePath: nodePathInfo.sequenceSubscriptionKey,
            propStatuses,
            keyframeDisplayOffset,
            getDragOverrides,
            getEffectDragOverrides,
            timelinePosition,
        }),
    ])), [
        flat,
        getDragOverrides,
        getEffectDragOverrides,
        keyframeDisplayOffset,
        nodePathInfo.sequenceSubscriptionKey,
        propStatuses,
        timelinePosition,
    ]);
    const rows = (0, react_1.useMemo)(() => flat.map(({ node }) => {
        var _a;
        const rowKey = (0, timeline_node_path_key_1.timelineNodePathInfoToKey)(node.nodePathInfo);
        return {
            height: (0, timeline_layout_1.getTreeRowHeight)(node),
            keyframes: (_a = nodeKeyframes.get(rowKey)) !== null && _a !== void 0 ? _a : [],
            canEditEasing: getNodeCanEditEasing({
                node,
                nodePath: nodePathInfo.sequenceSubscriptionKey,
                propStatuses,
            }),
            rowKey,
            nodePathInfo: node.nodePathInfo,
        };
    }), [propStatuses, flat, nodePathInfo.sequenceSubscriptionKey, nodeKeyframes]);
    return { rows, expandedHeight };
};
exports.useExpandedTrackKeyframeRows = useExpandedTrackKeyframeRows;
