// @remotion/studio/dist/components/Timeline/use-timeline-height.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.useTimelineHeight = void 0;
const react_1 = require("react");
const remotion_1 = require("remotion");
const client_id_1 = require("../../helpers/client-id");
const timeline_layout_1 = require("../../helpers/timeline-layout");
const ExpandedTracksProvider_1 = require("../ExpandedTracksProvider");
const get_node_keyframes_1 = require("./get-node-keyframes");
const MaxTimelineTracks_1 = require("./MaxTimelineTracks");
const timeline_expanded_filter_1 = require("./timeline-expanded-filter");
const TimelineSelection_1 = require("./TimelineSelection");
const TimelineTimeIndicators_1 = require("./TimelineTimeIndicators");
const useTimelineHeight = ({ shown, hasBeenCut, }) => {
    const { getIsExpanded } = (0, react_1.useContext)(ExpandedTracksProvider_1.ExpandedTracksGetterContext);
    const { previewServerState } = (0, react_1.useContext)(client_id_1.StudioServerConnectionCtx);
    const { propStatuses } = (0, react_1.useContext)(remotion_1.Internals.VisualModePropStatusesContext);
    const { getDragOverrides, getEffectDragOverrides } = (0, react_1.useContext)(remotion_1.Internals.VisualModeDragOverridesContext);
    const { selectedItems } = (0, TimelineSelection_1.useTimelineSelection)();
    const previewServerConnected = previewServerState.type === 'connected';
    const selectedRowKeys = (0, react_1.useMemo)(() => (0, timeline_expanded_filter_1.getSelectedTimelineExpandedRowKeys)(selectedItems), [selectedItems]);
    return (0, react_1.useMemo)(() => {
        const tracksHeight = shown.reduce((acc, track) => {
            const isExpanded = previewServerConnected &&
                track.nodePathInfo !== null &&
                getIsExpanded(track.nodePathInfo);
            const layerHeight = (0, timeline_layout_1.getTimelineLayerHeight)(track.sequence.type) +
                timeline_layout_1.TIMELINE_ITEM_BORDER_BOTTOM;
            const expandedHeight = (() => {
                if (!isExpanded || track.nodePathInfo === null) {
                    return 0;
                }
                const { nodePathInfo } = track;
                const tree = (0, timeline_layout_1.buildTimelineTree)({
                    sequence: track.sequence,
                    nodePathInfo,
                    getDragOverrides,
                    getEffectDragOverrides,
                    propStatuses,
                });
                const filteredTree = (0, timeline_expanded_filter_1.filterTimelineExpandedTree)({
                    nodes: tree,
                    shouldShowNode: (node) => (0, timeline_expanded_filter_1.isTimelineExpandedNodeSelected)({
                        nodePathInfo: node.nodePathInfo,
                        selectedRowKeys,
                    }) ||
                        (0, get_node_keyframes_1.getNodeHasKeyframes)({
                            node,
                            nodePath: nodePathInfo.sequenceSubscriptionKey,
                            propStatuses,
                            getDragOverrides,
                            getEffectDragOverrides,
                        }),
                });
                const flat = (0, timeline_layout_1.flattenVisibleTreeNodes)({
                    nodes: filteredTree,
                    getIsExpanded,
                });
                if (flat.length === 0) {
                    return 0;
                }
                const totalRowsHeight = flat.reduce((sum, { node }) => sum + (0, timeline_layout_1.getTreeRowHeight)(node), 0);
                const separators = Math.max(0, flat.length - 1);
                return totalRowsHeight + separators + timeline_layout_1.TIMELINE_ITEM_BORDER_BOTTOM;
            })();
            return acc + layerHeight + expandedHeight;
        }, 0);
        return (tracksHeight +
            timeline_layout_1.TIMELINE_ITEM_BORDER_BOTTOM +
            (hasBeenCut ? MaxTimelineTracks_1.MAX_TIMELINE_TRACKS_NOTICE_HEIGHT : 0) +
            TimelineTimeIndicators_1.TIMELINE_TIME_INDICATOR_HEIGHT);
    }, [
        shown,
        hasBeenCut,
        previewServerConnected,
        getIsExpanded,
        propStatuses,
        getDragOverrides,
        getEffectDragOverrides,
        selectedRowKeys,
    ]);
};
exports.useTimelineHeight = useTimelineHeight;
