// @remotion/studio/dist/components/Timeline/use-timeline-expanded-tree.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.useTimelineExpandedTree = void 0;
const react_1 = require("react");
const remotion_1 = require("remotion");
const timeline_layout_1 = require("../../helpers/timeline-layout");
const ExpandedTracksProvider_1 = require("../ExpandedTracksProvider");
const get_node_keyframes_1 = require("./get-node-keyframes");
const timeline_expanded_filter_1 = require("./timeline-expanded-filter");
const TimelineSelection_1 = require("./TimelineSelection");
const useTimelineExpandedTree = ({ sequence, nodePathInfo, }) => {
    const { getIsExpanded } = (0, react_1.useContext)(ExpandedTracksProvider_1.ExpandedTracksGetterContext);
    const { toggleTrack } = (0, react_1.useContext)(ExpandedTracksProvider_1.ExpandedTracksSetterContext);
    const { propStatuses: visualModePropStatuses } = (0, react_1.useContext)(remotion_1.Internals.VisualModePropStatusesContext);
    const { getDragOverrides, getEffectDragOverrides } = (0, react_1.useContext)(remotion_1.Internals.VisualModeDragOverridesContext);
    const { selectedItems } = (0, TimelineSelection_1.useTimelineSelection)();
    const tree = (0, react_1.useMemo)(() => (0, timeline_layout_1.buildTimelineTree)({
        sequence,
        nodePathInfo,
        getDragOverrides,
        getEffectDragOverrides,
        propStatuses: visualModePropStatuses,
    }), [
        sequence,
        nodePathInfo,
        getDragOverrides,
        getEffectDragOverrides,
        visualModePropStatuses,
    ]);
    const selectedRowKeys = (0, react_1.useMemo)(() => (0, timeline_expanded_filter_1.getSelectedTimelineExpandedRowKeys)(selectedItems), [selectedItems]);
    const filteredTree = (0, react_1.useMemo)(() => (0, timeline_expanded_filter_1.filterTimelineExpandedTree)({
        nodes: tree,
        shouldShowNode: (node) => (0, timeline_expanded_filter_1.isTimelineExpandedNodeSelected)({
            nodePathInfo: node.nodePathInfo,
            selectedRowKeys,
        }) ||
            (0, get_node_keyframes_1.getNodeHasKeyframes)({
                node,
                nodePath: nodePathInfo.sequenceSubscriptionKey,
                propStatuses: visualModePropStatuses,
                getDragOverrides,
                getEffectDragOverrides,
            }),
    }), [
        getDragOverrides,
        getEffectDragOverrides,
        nodePathInfo.sequenceSubscriptionKey,
        selectedRowKeys,
        tree,
        visualModePropStatuses,
    ]);
    return {
        filteredTree,
        getIsExpanded,
        propStatuses: visualModePropStatuses,
        toggleTrack,
        tree,
    };
};
exports.useTimelineExpandedTree = useTimelineExpandedTree;
