// @remotion/studio/dist/components/Timeline/use-sequence-props-subscription.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.useSequencePropsSubscription = void 0;
const studio_shared_1 = require("@remotion/studio-shared");
const react_1 = require("react");
const remotion_1 = require("remotion");
const client_id_1 = require("../../helpers/client-id");
const ExpandedTracksProvider_1 = require("../ExpandedTracksProvider");
const sequence_props_subscription_store_1 = require("./sequence-props-subscription-store");
const useSequencePropsSubscription = ({ originalLocation, overrideId, componentIdentity, schema, effects, }) => {
    var _a, _b, _c;
    const { setPropStatuses } = (0, react_1.useContext)(remotion_1.Internals.VisualModeSettersContext);
    const { setOverrideIdToNodePath } = (0, react_1.useContext)(remotion_1.Internals.OverrideIdsToNodePathsSettersContext);
    const { overrideIdToNodePathMappings } = (0, react_1.useContext)(remotion_1.Internals.OverrideIdsToNodePathsGettersContext);
    const { migrateExpandedTracksForSubscriptionKey } = (0, react_1.useContext)(ExpandedTracksProvider_1.ExpandedTracksSetterContext);
    const { previewServerState: state } = (0, react_1.useContext)(client_id_1.StudioServerConnectionCtx);
    const previousNodePathRef = (0, react_1.useRef)(null);
    const overrideIdToNodePathMappingsRef = (0, react_1.useRef)(overrideIdToNodePathMappings);
    overrideIdToNodePathMappingsRef.current = overrideIdToNodePathMappings;
    const clientId = state.type === 'connected' ? state.clientId : undefined;
    const effectsSignature = (0, react_1.useMemo)(() => effects.map((effect) => (0, studio_shared_1.getAllSchemaKeys)(effect).join('\0')).join('\0\0'), [effects]);
    const validatedLocation = (0, react_1.useMemo)(() => {
        var _a;
        if (!originalLocation ||
            !originalLocation.source ||
            !originalLocation.line) {
            return null;
        }
        return {
            source: originalLocation.source,
            line: originalLocation.line,
            column: (_a = originalLocation.column) !== null && _a !== void 0 ? _a : 0,
        };
    }, [originalLocation]);
    const locationSource = (_a = validatedLocation === null || validatedLocation === void 0 ? void 0 : validatedLocation.source) !== null && _a !== void 0 ? _a : null;
    const locationLine = (_b = validatedLocation === null || validatedLocation === void 0 ? void 0 : validatedLocation.line) !== null && _b !== void 0 ? _b : null;
    const locationColumn = (_c = validatedLocation === null || validatedLocation === void 0 ? void 0 : validatedLocation.column) !== null && _c !== void 0 ? _c : null;
    (0, react_1.useEffect)(() => {
        var _a, _b;
        if (!clientId ||
            !locationSource ||
            !locationLine ||
            locationColumn === null ||
            !schema) {
            return;
        }
        const nodePathAtResubscribe = (_a = overrideIdToNodePathMappingsRef.current[overrideId]) !== null && _a !== void 0 ? _a : null;
        const { release } = (0, sequence_props_subscription_store_1.acquireSequencePropsSubscription)({
            fileName: locationSource,
            line: locationLine,
            column: locationColumn,
            schema,
            componentIdentity,
            effects,
            nodePath: (_b = nodePathAtResubscribe === null || nodePathAtResubscribe === void 0 ? void 0 : nodePathAtResubscribe.nodePath) !== null && _b !== void 0 ? _b : null,
            clientId,
            applyOnce: (result) => {
                if (!result.success) {
                    return;
                }
                setPropStatuses(result.nodePath, () => result.status);
            },
            applyEach: (result) => {
                var _a;
                if (!result.success) {
                    return;
                }
                const newNodePath = result.nodePath;
                const newNodePathKey = (0, studio_shared_1.stringifySequenceSubscriptionKey)(newNodePath);
                const previousNodePath = (_a = previousNodePathRef.current) !== null && _a !== void 0 ? _a : nodePathAtResubscribe;
                const previousNodePathKey = previousNodePath
                    ? (0, studio_shared_1.stringifySequenceSubscriptionKey)(previousNodePath)
                    : null;
                if (previousNodePathKey === newNodePathKey) {
                    return;
                }
                if (previousNodePath) {
                    migrateExpandedTracksForSubscriptionKey(previousNodePath, newNodePath);
                }
                previousNodePathRef.current = newNodePath;
                setOverrideIdToNodePath(overrideId, newNodePath);
            },
        });
        return () => {
            release();
        };
    }, [
        clientId,
        componentIdentity,
        effects,
        effectsSignature,
        locationColumn,
        locationLine,
        locationSource,
        migrateExpandedTracksForSubscriptionKey,
        overrideId,
        schema,
        setPropStatuses,
        setOverrideIdToNodePath,
    ]);
};
exports.useSequencePropsSubscription = useSequencePropsSubscription;
