// @remotion/studio-server/dist/preview-server/routes/log-updates/formatting.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.formatAddition = exports.formatDeletion = exports.formatPropChangeDelta = exports.formatPropDelta = exports.addedPrefixIfNoColor = exports.strikeThroughOrRemovedPrefix = exports.strikeThrough = exports.inlineAddition = exports.numberValue = exports.stringValue = exports.punctuation = exports.equals = exports.attrName = exports.bg = exports.fg = exports.colorValue = exports.colorEnabled = void 0;
const renderer_1 = require("@remotion/renderer");
const colorEnabled = () => renderer_1.RenderInternals.chalk.enabled();
exports.colorEnabled = colorEnabled;
const colorValue = (str) => {
    if ((str.startsWith("'") && str.endsWith("'")) ||
        (str.startsWith('"') && str.endsWith('"'))) {
        return (0, exports.stringValue)(str);
    }
    if (/^-?\d+(\.\d+)?$/.test(str)) {
        return (0, exports.numberValue)(str);
    }
    return (0, exports.punctuation)(str);
};
exports.colorValue = colorValue;
// eslint-disable-next-line no-control-regex
const stripAnsi = (str) => str.replace(/\u001b\[[0-9;]*m/g, '');
// 24-bit ANSI helpers
const fg = (r, g, b, str) => (0, exports.colorEnabled)() ? `\u001b[38;2;${r};${g};${b}m${str}\u001b[39m` : str;
exports.fg = fg;
const bg = (r, g, b, str) => (0, exports.colorEnabled)() ? `\u001b[48;2;${r};${g};${b}m${str}\u001b[49m` : str;
exports.bg = bg;
// Monokai-inspired syntax colors
const attrName = (str) => (0, exports.fg)(166, 226, 46, str);
exports.attrName = attrName;
const equals = (str) => (0, exports.fg)(249, 38, 114, str);
exports.equals = equals;
// Use the terminal's default foreground color so punctuation is visible in both
// light and dark terminal themes.
const punctuation = (str) => str;
exports.punctuation = punctuation;
const stringValue = (str) => (0, exports.fg)(230, 219, 116, str);
exports.stringValue = stringValue;
const numberValue = (str) => (0, exports.fg)(174, 129, 255, str);
exports.numberValue = numberValue;
const inlineAddition = (str) => (0, exports.fg)(166, 226, 46, str);
exports.inlineAddition = inlineAddition;
const strikeThrough = (str) => `\u001b[9m\u001b[38;2;255;85;85m${stripAnsi(str)}\u001b[39m\u001b[29m`;
exports.strikeThrough = strikeThrough;
const strikeThroughOrRemovedPrefix = (str) => (0, exports.colorEnabled)() ? (0, exports.strikeThrough)(str) : 'removed: ' + str;
exports.strikeThroughOrRemovedPrefix = strikeThroughOrRemovedPrefix;
const addedPrefixIfNoColor = (str) => (0, exports.colorEnabled)() ? str : 'added: ' + stripAnsi(str);
exports.addedPrefixIfNoColor = addedPrefixIfNoColor;
// Format key={value} with Monokai syntax highlighting
const formatSimpleProp = (key, value) => {
    return `${(0, exports.attrName)(key)}${(0, exports.equals)('=')}${(0, exports.punctuation)('{')}${(0, exports.colorValue)(value)}${(0, exports.punctuation)('}')}`;
};
// Format parentKey={{childKey: value}} with Monokai syntax highlighting
const formatNestedProp = (parentKey, childKey, value) => {
    return `${(0, exports.attrName)(parentKey)}${(0, exports.equals)('=')}${(0, exports.punctuation)('{{')}${(0, exports.punctuation)(childKey)}${(0, exports.punctuation)(':')} ${(0, exports.colorValue)(value)}${(0, exports.punctuation)('}}')}`;
};
const formatValueDelta = ({ oldValueString, newValueString, }) => {
    const withoutUnchangedOptions = shortenUnchangedInterpolateOptions({
        oldValueString,
        newValueString,
    });
    const shortened = removeUnchangedInterpolateOptionProperties(withoutUnchangedOptions);
    const inlineDelta = formatInlineAdditionOrRemoval(shortened);
    if (inlineDelta !== null) {
        return inlineDelta;
    }
    return `${(0, exports.colorValue)(shortened.oldValueString)} ${(0, exports.punctuation)('→')} ${(0, exports.colorValue)(shortened.newValueString)}`;
};
const formatSimplePropChange = ({ key, oldValueString, newValueString, }) => {
    return `${(0, exports.attrName)(key)}${(0, exports.equals)('=')}${(0, exports.punctuation)('{')}${formatValueDelta({ oldValueString, newValueString })}${(0, exports.punctuation)('}')}`;
};
const formatNestedPropChange = ({ key, oldValueString, newValueString, }) => {
    const dotIdx = key.indexOf('.');
    return `${(0, exports.attrName)(key.slice(0, dotIdx))}${(0, exports.equals)('=')}${(0, exports.punctuation)('{{')}${(0, exports.punctuation)(key.slice(dotIdx + 1))}${(0, exports.punctuation)(':')} ${formatValueDelta({ oldValueString, newValueString })}${(0, exports.punctuation)('}}')}`;
};
const formatPropDelta = ({ key, valueString }) => {
    const dotIdx = key.indexOf('.');
    if (dotIdx === -1) {
        return formatSimpleProp(key, valueString);
    }
    return formatNestedProp(key.slice(0, dotIdx), key.slice(dotIdx + 1), valueString);
};
exports.formatPropDelta = formatPropDelta;
const formatPropChangeDelta = ({ key, oldValueString, newValueString, }) => {
    const dotIdx = key.indexOf('.');
    if (dotIdx === -1) {
        return formatSimplePropChange({ key, oldValueString, newValueString });
    }
    return formatNestedPropChange({ key, oldValueString, newValueString });
};
exports.formatPropChangeDelta = formatPropChangeDelta;
const formatDeletion = (prop) => {
    const formatted = (0, exports.formatPropDelta)(prop);
    return (0, exports.strikeThroughOrRemovedPrefix)(formatted);
};
exports.formatDeletion = formatDeletion;
const formatAddition = (prop) => {
    const formatted = (0, exports.formatPropDelta)(prop);
    return (0, exports.addedPrefixIfNoColor)(formatted);
};
exports.formatAddition = formatAddition;
const callStart = 'interpolate(';
const normalizeArg = (arg) => {
    return arg
        .replace(/\s+/g, ' ')
        .replace(/,(\s*[}\]])/g, '$1')
        .trim();
};
const shortenUnchangedOptionProperties = new Set([
    'extrapolateLeft',
    'extrapolateRight',
]);
const splitTopLevelArgs = (argsSource) => {
    const args = [];
    let depth = 0;
    let quote = null;
    let start = 0;
    for (let i = 0; i < argsSource.length; i++) {
        const char = argsSource[i];
        const previous = argsSource[i - 1];
        if (quote) {
            if (char === quote && previous !== '\\') {
                quote = null;
            }
            continue;
        }
        if (char === "'" || char === '"' || char === '`') {
            quote = char;
            continue;
        }
        if (char === '(' || char === '[' || char === '{') {
            depth++;
            continue;
        }
        if (char === ')' || char === ']' || char === '}') {
            depth--;
            continue;
        }
        if (char === ',' && depth === 0) {
            args.push(argsSource.slice(start, i).trim());
            start = i + 1;
        }
    }
    args.push(argsSource.slice(start).trim());
    return args;
};
const getCommonPrefixLength = (oldValueString, newValueString) => {
    const maxLength = Math.min(oldValueString.length, newValueString.length);
    let index = 0;
    while (index < maxLength && oldValueString[index] === newValueString[index]) {
        index++;
    }
    return index;
};
const getCommonSuffixLength = ({ oldValueString, newValueString, prefixLength, }) => {
    const maxLength = Math.min(oldValueString.length, newValueString.length) - prefixLength;
    let index = 0;
    while (index < maxLength &&
        oldValueString[oldValueString.length - 1 - index] ===
            newValueString[newValueString.length - 1 - index]) {
        index++;
    }
    return index;
};
const minSharedInlineDeltaChars = 12;
const formatInlineAdditionOrRemoval = ({ oldValueString, newValueString, }) => {
    if (!(0, exports.colorEnabled)()) {
        return null;
    }
    const prefixLength = getCommonPrefixLength(oldValueString, newValueString);
    const suffixLength = getCommonSuffixLength({
        oldValueString,
        newValueString,
        prefixLength,
    });
    if (prefixLength + suffixLength < minSharedInlineDeltaChars) {
        return null;
    }
    const oldMiddleEnd = oldValueString.length - suffixLength;
    const newMiddleEnd = newValueString.length - suffixLength;
    const oldMiddle = oldValueString.slice(prefixLength, oldMiddleEnd);
    const newMiddle = newValueString.slice(prefixLength, newMiddleEnd);
    if (oldMiddle.length === 0 && newMiddle.length > 0) {
        return `${newValueString.slice(0, prefixLength)}${(0, exports.inlineAddition)(newMiddle)}${newValueString.slice(newMiddleEnd)}`;
    }
    if (newMiddle.length === 0 && oldMiddle.length > 0) {
        return `${oldValueString.slice(0, prefixLength)}${(0, exports.strikeThrough)(oldMiddle)}${oldValueString.slice(oldMiddleEnd)}`;
    }
    return null;
};
const getObjectPropertyKey = (propertySource) => {
    const colonIndex = propertySource.indexOf(':');
    if (colonIndex === -1) {
        return null;
    }
    const key = propertySource.slice(0, colonIndex).trim();
    if (/^[A-Za-z_$][\w$]*$/.test(key)) {
        return key;
    }
    if ((key.startsWith("'") && key.endsWith("'")) ||
        (key.startsWith('"') && key.endsWith('"'))) {
        return key.slice(1, -1);
    }
    return null;
};
const parseTopLevelObjectProperties = (objectSource) => {
    const trimmed = objectSource.trim();
    if (!trimmed.startsWith('{') || !trimmed.endsWith('}')) {
        return null;
    }
    const properties = splitTopLevelArgs(trimmed.slice(1, -1))
        .filter(Boolean)
        .map((propertySource) => {
        const key = getObjectPropertyKey(propertySource);
        if (key === null) {
            return null;
        }
        const colonIndex = propertySource.indexOf(':');
        return {
            key,
            value: normalizeArg(propertySource.slice(colonIndex + 1)),
            source: propertySource.trim(),
        };
    });
    if (properties.some((property) => property === null)) {
        return null;
    }
    return properties;
};
const parseInterpolateCall = (valueString) => {
    const trimmed = valueString.trim();
    if (!trimmed.startsWith(callStart) || !trimmed.endsWith(')')) {
        return null;
    }
    let depth = 0;
    let quote = null;
    for (let i = 'interpolate'.length; i < trimmed.length; i++) {
        const char = trimmed[i];
        const previous = trimmed[i - 1];
        if (quote) {
            if (char === quote && previous !== '\\') {
                quote = null;
            }
            continue;
        }
        if (char === "'" || char === '"' || char === '`') {
            quote = char;
            continue;
        }
        if (char === '(' || char === '[' || char === '{') {
            depth++;
            continue;
        }
        if (char === ')' || char === ']' || char === '}') {
            depth--;
            if (depth === 0 && i !== trimmed.length - 1) {
                return null;
            }
        }
    }
    if (depth !== 0) {
        return null;
    }
    return {
        args: splitTopLevelArgs(trimmed.slice(callStart.length, -1)),
    };
};
const shortenUnchangedInterpolateOptions = ({ oldValueString, newValueString, }) => {
    const oldCall = parseInterpolateCall(oldValueString);
    const newCall = parseInterpolateCall(newValueString);
    if (!oldCall ||
        !newCall ||
        oldCall.args.length <= 3 ||
        newCall.args.length <= 3) {
        return { oldValueString, newValueString };
    }
    const oldOptions = oldCall.args.slice(3).map(normalizeArg);
    const newOptions = newCall.args.slice(3).map(normalizeArg);
    if (oldOptions.length !== newOptions.length ||
        oldOptions.some((option, index) => option !== newOptions[index])) {
        return { oldValueString, newValueString };
    }
    return {
        oldValueString: `interpolate(${oldCall.args.slice(0, 3).join(', ')})`,
        newValueString: `interpolate(${newCall.args.slice(0, 3).join(', ')})`,
    };
};
const removeUnchangedInterpolateOptionProperties = ({ oldValueString, newValueString, }) => {
    const oldCall = parseInterpolateCall(oldValueString);
    const newCall = parseInterpolateCall(newValueString);
    if (!oldCall ||
        !newCall ||
        oldCall.args.length <= 3 ||
        newCall.args.length <= 3) {
        return { oldValueString, newValueString };
    }
    const oldProperties = parseTopLevelObjectProperties(oldCall.args[3]);
    const newProperties = parseTopLevelObjectProperties(newCall.args[3]);
    if (!oldProperties || !newProperties) {
        return { oldValueString, newValueString };
    }
    const propertiesToRemove = [...shortenUnchangedOptionProperties].filter((propertyName) => {
        const oldProperty = oldProperties.find((property) => property.key === propertyName);
        const newProperty = newProperties.find((property) => property.key === propertyName);
        return (oldProperty !== undefined &&
            newProperty !== undefined &&
            oldProperty.value === newProperty.value);
    });
    if (propertiesToRemove.length === 0) {
        return { oldValueString, newValueString };
    }
    const removeProperties = (call, properties) => {
        const nextOptions = properties.filter((property) => !propertiesToRemove.includes(property.key));
        const nextArgs = nextOptions.length === 0
            ? [...call.args.slice(0, 3), ...call.args.slice(4)]
            : [
                ...call.args.slice(0, 3),
                `{${nextOptions.map((property) => property.source).join(', ')}}`,
                ...call.args.slice(4),
            ];
        return `interpolate(${nextArgs.join(', ')})`;
    };
    return {
        oldValueString: removeProperties(oldCall, oldProperties),
        newValueString: removeProperties(newCall, newProperties),
    };
};
