// @remotion/studio-server/dist/preview-server/routes/log-updates/log-update.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.logUpdate = exports.normalizeQuotes = exports.warnAboutPrettierOnce = void 0;
const renderer_1 = require("@remotion/renderer");
const format_prop_change_1 = require("./format-prop-change");
let warnedAboutPrettier = false;
const warnAboutPrettierOnce = (logLevel) => {
    if (warnedAboutPrettier) {
        return;
    }
    warnedAboutPrettier = true;
    renderer_1.RenderInternals.Log.warn({ indent: false, logLevel }, renderer_1.RenderInternals.chalk.yellow('Could not format with Prettier. File will need to be formatted manually.'));
};
exports.warnAboutPrettierOnce = warnAboutPrettierOnce;
const normalizeQuotes = (str) => {
    if (str.length >= 2 &&
        ((str.startsWith("'") && str.endsWith("'")) ||
            (str.startsWith('"') && str.endsWith('"')))) {
        return `'${str.slice(1, -1)}'`;
    }
    return str;
};
exports.normalizeQuotes = normalizeQuotes;
const logUpdate = ({ fileRelativeToRoot, line, key, oldValueString, newValueString, defaultValueString, formatted, logLevel, removedProps, addedProps, }) => {
    const locationLabel = `${fileRelativeToRoot}:${line}`;
    const propChange = (0, format_prop_change_1.formatPropChange)({
        key,
        oldValueString: (0, exports.normalizeQuotes)(oldValueString),
        newValueString: (0, exports.normalizeQuotes)(newValueString),
        defaultValueString: defaultValueString !== null ? (0, exports.normalizeQuotes)(defaultValueString) : null,
        removedProps,
        addedProps,
    });
    renderer_1.RenderInternals.Log.info({ indent: false, logLevel }, `${renderer_1.RenderInternals.chalk.blueBright(`${locationLabel}`)} ${propChange}`);
    if (!formatted) {
        (0, exports.warnAboutPrettierOnce)(logLevel);
    }
};
exports.logUpdate = logUpdate;
