// @remotion/studio-server/dist/preview-server/default-props-watchers.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.unsubscribeClientDefaultPropsWatchers = exports.unsubscribeFromDefaultPropsWatchers = exports.subscribeToDefaultPropsWatchers = void 0;
const file_watcher_1 = require("../file-watcher");
const live_events_1 = require("./live-events");
const can_update_default_props_1 = require("./routes/can-update-default-props");
// Global file watcher — at most one for the root file
let globalWatcher = null;
// Per-compositionId set of subscribed clientIds
const subscriptions = {};
// Cached config for recomputation on file change
let watcherConfig = null;
const ensureGlobalWatcher = (rootFile) => {
    if (globalWatcher) {
        globalWatcher.refCount++;
        return;
    }
    const { unwatch } = (0, file_watcher_1.installFileWatcher)({
        file: rootFile,
        existenceOnly: false,
        onChange: (event) => {
            if (event.type === 'deleted') {
                return;
            }
            if (!watcherConfig) {
                return;
            }
            const compositionIds = Object.keys(subscriptions);
            for (const compositionId of compositionIds) {
                if (subscriptions[compositionId].size === 0) {
                    continue;
                }
                const clientIds = [...subscriptions[compositionId]];
                const newResult = (0, can_update_default_props_1.computeCanUpdateDefaultPropsFromContent)(event.content, compositionId);
                (0, live_events_1.waitForLiveEventsListener)().then((listener) => {
                    const updateEvent = {
                        type: 'default-props-updatable-changed',
                        compositionId,
                        result: newResult,
                    };
                    for (const cId of clientIds) {
                        listener.sendEventToClientId(cId, updateEvent);
                    }
                });
            }
        },
    });
    globalWatcher = { unwatch, rootFile, refCount: 1 };
};
const releaseGlobalWatcher = () => {
    if (!globalWatcher) {
        return;
    }
    globalWatcher.refCount--;
    if (globalWatcher.refCount <= 0) {
        globalWatcher.unwatch();
        globalWatcher = null;
        watcherConfig = null;
    }
};
const subscribeToDefaultPropsWatchers = async ({ compositionId, clientId, remotionRoot, entryPoint, }) => {
    var _a;
    const { result, rootFile } = await (0, can_update_default_props_1.computeCanUpdateDefaultProps)({
        compositionId,
        remotionRoot,
        entryPoint,
    });
    // Remove from any previous subscription for this client+composition
    if ((_a = subscriptions[compositionId]) === null || _a === void 0 ? void 0 : _a.has(clientId)) {
        subscriptions[compositionId].delete(clientId);
        releaseGlobalWatcher();
    }
    if (rootFile) {
        watcherConfig = { remotionRoot, entryPoint };
        if (!subscriptions[compositionId]) {
            subscriptions[compositionId] = new Set();
        }
        subscriptions[compositionId].add(clientId);
        ensureGlobalWatcher(rootFile);
    }
    return result;
};
exports.subscribeToDefaultPropsWatchers = subscribeToDefaultPropsWatchers;
const unsubscribeFromDefaultPropsWatchers = ({ compositionId, clientId, }) => {
    var _a;
    if (!((_a = subscriptions[compositionId]) === null || _a === void 0 ? void 0 : _a.has(clientId))) {
        return;
    }
    subscriptions[compositionId].delete(clientId);
    if (subscriptions[compositionId].size === 0) {
        delete subscriptions[compositionId];
    }
    releaseGlobalWatcher();
};
exports.unsubscribeFromDefaultPropsWatchers = unsubscribeFromDefaultPropsWatchers;
const unsubscribeClientDefaultPropsWatchers = (clientId) => {
    for (const compositionId of Object.keys(subscriptions)) {
        if (subscriptions[compositionId].has(clientId)) {
            subscriptions[compositionId].delete(clientId);
            if (subscriptions[compositionId].size === 0) {
                delete subscriptions[compositionId];
            }
            releaseGlobalWatcher();
        }
    }
};
exports.unsubscribeClientDefaultPropsWatchers = unsubscribeClientDefaultPropsWatchers;
