// @remotion/studio-server/dist/preview-server/undo-stack.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.pushToUndoStack = pushToUndoStack;
exports.pushTransactionToUndoStack = pushTransactionToUndoStack;
exports.printUndoHint = printUndoHint;
exports.pushToRedoStack = pushToRedoStack;
exports.suppressUndoStackInvalidation = suppressUndoStackInvalidation;
exports.popUndo = popUndo;
exports.popRedo = popRedo;
exports.getUndoStack = getUndoStack;
exports.getRedoStack = getRedoStack;
exports.clearUndoStackForTests = clearUndoStackForTests;
const node_fs_1 = require("node:fs");
const renderer_1 = require("@remotion/renderer");
const parse_ast_1 = require("../codemods/parse-ast");
const read_visual_control_values_1 = require("../codemods/read-visual-control-values");
const file_watcher_1 = require("../file-watcher");
const format_log_file_location_1 = require("./format-log-file-location");
const live_events_1 = require("./live-events");
const watch_ignore_next_change_1 = require("./watch-ignore-next-change");
const MAX_ENTRIES = 100;
const undoStack = [];
const redoStack = [];
const suppressedWrites = new Map();
const watchers = new Map();
let storedLogLevel = 'info';
let storedRemotionRoot = null;
let printedUndoHint = false;
function broadcastState() {
    const undoFile = undoStack.length > 0 ? undoStack[undoStack.length - 1].filePath : null;
    const redoFile = redoStack.length > 0 ? redoStack[redoStack.length - 1].filePath : null;
    (0, live_events_1.waitForLiveEventsListener)().then((listener) => {
        listener.sendEventToClient({
            type: 'undo-redo-stack-changed',
            undoFile,
            redoFile,
        });
    });
}
const entryTouchesFile = (entry, filePath) => {
    return entry.snapshots.some((snapshot) => snapshot.filePath === filePath);
};
const getEntryFilePaths = (entry) => {
    return entry.snapshots.map((snapshot) => snapshot.filePath);
};
const makeUndoEntry = ({ snapshots, description, entryType, suppressHmrOnFileRestore, }) => {
    if (snapshots.length === 0) {
        throw new Error('Cannot create an undo entry without snapshots');
    }
    return {
        filePath: snapshots[0].filePath,
        logLine: snapshots[0].logLine,
        snapshots,
        description,
        entryType,
        suppressHmrOnFileRestore,
    };
};
function pushToUndoStack({ filePath, oldContents, newContents, logLevel, remotionRoot, logLine, description, entryType, suppressHmrOnFileRestore, }) {
    pushTransactionToUndoStack({
        snapshots: [
            {
                filePath,
                oldContents,
                newContents,
                logLine,
            },
        ],
        logLevel,
        remotionRoot,
        description,
        entryType,
        suppressHmrOnFileRestore,
    });
}
function pushTransactionToUndoStack({ snapshots, logLevel, remotionRoot, description, entryType, suppressHmrOnFileRestore, }) {
    storedLogLevel = logLevel;
    storedRemotionRoot = remotionRoot;
    const entry = makeUndoEntry({
        snapshots,
        description,
        entryType,
        suppressHmrOnFileRestore,
    });
    undoStack.push(entry);
    if (undoStack.length > MAX_ENTRIES) {
        undoStack.shift();
    }
    redoStack.length = 0;
    renderer_1.RenderInternals.Log.verbose({ indent: false, logLevel }, renderer_1.RenderInternals.chalk.gray(`Undo stack: added entry for ${entry.filePath} (${undoStack.length} items)`));
    for (const filePath of getEntryFilePaths(entry)) {
        ensureWatching(filePath);
    }
    broadcastState();
}
function printUndoHint(logLevel) {
    if (!printedUndoHint) {
        printedUndoHint = true;
        const shortcut = process.platform === 'darwin' ? 'Cmd+Z' : 'Ctrl+Z';
        renderer_1.RenderInternals.Log.info({ indent: false, logLevel }, renderer_1.RenderInternals.chalk.gray(`Tip: ${shortcut} in Studio to undo`));
    }
}
function pushToRedoStack({ filePath, oldContents, newContents, logLine, description, entryType, suppressHmrOnFileRestore, }) {
    const entry = makeUndoEntry({
        snapshots: [
            {
                filePath,
                oldContents,
                newContents,
                logLine,
            },
        ],
        description,
        entryType,
        suppressHmrOnFileRestore,
    });
    redoStack.push(entry);
    if (redoStack.length > MAX_ENTRIES) {
        redoStack.shift();
    }
    renderer_1.RenderInternals.Log.verbose({ indent: false, logLevel: storedLogLevel }, renderer_1.RenderInternals.chalk.gray(`Redo stack: added entry for ${entry.filePath} (${redoStack.length} items)`));
    for (const watchedFilePath of getEntryFilePaths(entry)) {
        ensureWatching(watchedFilePath);
    }
    broadcastState();
}
function suppressUndoStackInvalidation(filePath) {
    var _a;
    suppressedWrites.set(filePath, ((_a = suppressedWrites.get(filePath)) !== null && _a !== void 0 ? _a : 0) + 1);
}
function ensureWatching(filePath) {
    if (watchers.has(filePath)) {
        return;
    }
    const watcher = (0, file_watcher_1.installFileWatcher)({
        file: filePath,
        existenceOnly: false,
        onChange: () => {
            var _a;
            const count = (_a = suppressedWrites.get(filePath)) !== null && _a !== void 0 ? _a : 0;
            if (count > 0) {
                if (count === 1) {
                    suppressedWrites.delete(filePath);
                }
                else {
                    suppressedWrites.set(filePath, count - 1);
                }
                return;
            }
            invalidateForFile(filePath);
        },
    });
    watchers.set(filePath, watcher);
}
function invalidateForFile(filePath) {
    let changed = false;
    let lastUndoIndex = -1;
    for (let i = undoStack.length - 1; i >= 0; i--) {
        if (entryTouchesFile(undoStack[i], filePath)) {
            lastUndoIndex = i;
            break;
        }
    }
    if (lastUndoIndex !== -1) {
        const removed = lastUndoIndex + 1;
        undoStack.splice(0, removed);
        changed = true;
        renderer_1.RenderInternals.Log.verbose({ indent: false, logLevel: storedLogLevel }, renderer_1.RenderInternals.chalk.gray(`Undo stack: ${filePath} was externally modified, removed ${removed} entries (${undoStack.length} items)`));
    }
    let lastRedoIndex = -1;
    for (let i = redoStack.length - 1; i >= 0; i--) {
        if (entryTouchesFile(redoStack[i], filePath)) {
            lastRedoIndex = i;
            break;
        }
    }
    if (lastRedoIndex !== -1) {
        const removed = lastRedoIndex + 1;
        redoStack.splice(0, removed);
        changed = true;
        renderer_1.RenderInternals.Log.verbose({ indent: false, logLevel: storedLogLevel }, renderer_1.RenderInternals.chalk.gray(`Redo stack: ${filePath} was externally modified, removed ${removed} entries (${redoStack.length} items)`));
    }
    cleanupWatchers();
    if (changed) {
        broadcastState();
    }
}
function cleanupWatchers() {
    const filesInStacks = new Set([
        ...undoStack.flatMap(getEntryFilePaths),
        ...redoStack.flatMap(getEntryFilePaths),
    ]);
    for (const [filePath, watcher] of watchers) {
        if (!filesInStacks.has(filePath)) {
            watcher.unwatch();
            watchers.delete(filePath);
        }
    }
}
function emitVisualControlChanges(fileContents) {
    try {
        const ast = (0, parse_ast_1.parseAst)(fileContents);
        const values = (0, read_visual_control_values_1.readVisualControlValues)(ast);
        if (values.length > 0) {
            (0, live_events_1.waitForLiveEventsListener)().then((listener) => {
                listener.sendEventToClient({
                    type: 'visual-control-values-changed',
                    values,
                });
            });
        }
    }
    catch (_a) {
        // File might not contain visual controls or might not be parseable
    }
}
function logFileAction(action, filePath, logLine) {
    const locationLabel = storedRemotionRoot !== null
        ? (0, format_log_file_location_1.formatLogFileLocation)({
            remotionRoot: storedRemotionRoot,
            absolutePath: filePath,
            line: logLine,
        })
        : `${filePath}:${logLine}`;
    renderer_1.RenderInternals.Log.info({ indent: false, logLevel: storedLogLevel }, `${renderer_1.RenderInternals.chalk.blueBright(`${locationLabel}`)} ${action}`);
}
function popUndo() {
    const entry = undoStack.pop();
    if (!entry) {
        return { success: false, reason: 'Nothing to undo' };
    }
    const redoSnapshots = entry.snapshots.map((snapshot) => {
        var _a;
        return {
            ...snapshot,
            newContents: (_a = snapshot.newContents) !== null && _a !== void 0 ? _a : ((0, node_fs_1.existsSync)(snapshot.filePath)
                ? (0, node_fs_1.readFileSync)(snapshot.filePath, 'utf-8')
                : null),
        };
    });
    redoStack.push(makeUndoEntry({
        snapshots: redoSnapshots,
        description: entry.description,
        entryType: entry.entryType,
        suppressHmrOnFileRestore: entry.suppressHmrOnFileRestore,
    }));
    if (redoStack.length > MAX_ENTRIES) {
        redoStack.shift();
    }
    for (const snapshot of entry.snapshots) {
        suppressUndoStackInvalidation(snapshot.filePath);
        if (entry.suppressHmrOnFileRestore) {
            (0, watch_ignore_next_change_1.suppressBundlerUpdateForFile)(snapshot.filePath);
        }
        if (snapshot.oldContents === null) {
            (0, node_fs_1.rmSync)(snapshot.filePath, { force: true });
        }
        else {
            (0, file_watcher_1.writeFileAndNotifyFileWatchers)(snapshot.filePath, snapshot.oldContents, undefined);
        }
    }
    renderer_1.RenderInternals.Log.verbose({ indent: false, logLevel: storedLogLevel }, renderer_1.RenderInternals.chalk.gray(`Undo: restored ${entry.filePath} (undo: ${undoStack.length}, redo: ${redoStack.length})`));
    logFileAction(entry.description.undoMessage, entry.filePath, entry.logLine);
    if (entry.entryType === 'visual-control') {
        for (const snapshot of entry.snapshots) {
            if (snapshot.oldContents !== null) {
                emitVisualControlChanges(snapshot.oldContents);
            }
        }
    }
    for (const filePath of getEntryFilePaths(entry)) {
        ensureWatching(filePath);
    }
    broadcastState();
    return { success: true };
}
function popRedo() {
    const entry = redoStack.pop();
    if (!entry) {
        return { success: false, reason: 'Nothing to redo' };
    }
    const snapshotsWithNewContents = [];
    for (const snapshot of entry.snapshots) {
        if (snapshot.newContents === null) {
            return { success: false, reason: 'Redo entry is incomplete' };
        }
        snapshotsWithNewContents.push({
            ...snapshot,
            newContents: snapshot.newContents,
        });
    }
    undoStack.push(makeUndoEntry({
        snapshots: snapshotsWithNewContents,
        description: entry.description,
        entryType: entry.entryType,
        suppressHmrOnFileRestore: entry.suppressHmrOnFileRestore,
    }));
    if (undoStack.length > MAX_ENTRIES) {
        undoStack.shift();
    }
    for (const snapshot of snapshotsWithNewContents) {
        suppressUndoStackInvalidation(snapshot.filePath);
        if (entry.suppressHmrOnFileRestore) {
            (0, watch_ignore_next_change_1.suppressBundlerUpdateForFile)(snapshot.filePath);
        }
        (0, file_watcher_1.writeFileAndNotifyFileWatchers)(snapshot.filePath, snapshot.newContents, undefined);
    }
    renderer_1.RenderInternals.Log.verbose({ indent: false, logLevel: storedLogLevel }, renderer_1.RenderInternals.chalk.gray(`Redo: restored ${entry.filePath} (undo: ${undoStack.length}, redo: ${redoStack.length})`));
    logFileAction(entry.description.redoMessage, entry.filePath, entry.logLine);
    if (entry.entryType === 'visual-control') {
        for (const snapshot of entry.snapshots) {
            if (snapshot.newContents !== null) {
                emitVisualControlChanges(snapshot.newContents);
            }
        }
    }
    for (const filePath of getEntryFilePaths(entry)) {
        ensureWatching(filePath);
    }
    broadcastState();
    return { success: true };
}
/*
 * Keep stack accessors typed as readonly arrays so callers can only inspect
 * whether undo/redo is available and which file represents the top entry.
 */
function getUndoStack() {
    return undoStack;
}
function getRedoStack() {
    return redoStack;
}
function clearUndoStackForTests() {
    undoStack.length = 0;
    redoStack.length = 0;
    suppressedWrites.clear();
    cleanupWatchers();
}
