// @remotion/studio-server/dist/preview-server/hot-middleware/index.js
"use strict";
/**
 * Source code is adapted from
 * https://github.com/webpack-contrib/webpack-hot-middleware#readme
 * and rewritten in TypeScript. This file is MIT licensed
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.setupWebpackHmr = void 0;
const renderer_1 = require("@remotion/renderer");
const setupWebpackHmr = (compiler, logLevel, liveEventsServer) => {
    let latestStats = null;
    const publishHmr = (hmrEvent) => {
        liveEventsServer.sendEventToClient({ type: 'hmr', hmrEvent });
    };
    compiler.hooks.invalid.tap('remotion', onInvalid);
    compiler.hooks.done.tap('remotion', onDone);
    function onInvalid() {
        latestStats = null;
        renderer_1.RenderInternals.Log.info({ indent: false, logLevel }, 'Building...');
        publishHmr({
            action: 'building',
        });
    }
    function onDone(statsResult) {
        // Keep hold of latest stats so they can be propagated to new clients
        latestStats = statsResult;
        publishStats('built', latestStats, publishHmr);
    }
    liveEventsServer.addNewClientListener(() => {
        if (latestStats) {
            publishStats('sync', latestStats, publishHmr);
        }
    });
};
exports.setupWebpackHmr = setupWebpackHmr;
function publishStats(action, statsResult, publishHmr) {
    const stats = statsResult.toJson({
        all: false,
        cached: true,
        children: true,
        modules: true,
        timings: true,
        hash: true,
    });
    // For multi-compiler, stats will be an object with a 'children' array of stats
    const bundles = extractBundles(stats);
    bundles.forEach((_stats) => {
        let name = _stats.name || '';
        // Fallback to compilation name in case of 1 bundle (if it exists)
        if (bundles.length === 1 && !name && statsResult.compilation) {
            name = statsResult.compilation.name || '';
        }
        publishHmr({
            name,
            action,
            time: _stats.time,
            hash: _stats.hash,
            warnings: _stats.warnings || [],
            errors: _stats.errors || [],
            modules: buildModuleMap(_stats.modules),
        });
    });
}
function extractBundles(stats) {
    var _a;
    // Stats has modules, single bundle
    if (stats.modules)
        return [stats];
    // Stats has children, multiple bundles
    if ((_a = stats.children) === null || _a === void 0 ? void 0 : _a.length)
        return stats.children;
    // Not sure, assume single
    return [stats];
}
function buildModuleMap(modules) {
    const map = {};
    if (!modules) {
        return map;
    }
    modules.forEach((module) => {
        const id = module.id;
        map[id] = module.name;
    });
    return map;
}
