// @remotion/studio/dist/components/ColorPicker/AlphaSlider.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AlphaSlider = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const react_1 = require("react");
const color_conversion_1 = require("../../helpers/color-conversion");
const checker_1 = require("./checker");
const SLIDER_HEIGHT = 12;
const HANDLE_WIDTH = 8;
const wrapperStyle = {
    position: 'relative',
    height: SLIDER_HEIGHT,
    width: '100%',
    borderRadius: 3,
    overflow: 'hidden',
    cursor: 'ew-resize',
    backgroundColor: checker_1.CHECKER_BACKGROUND_COLOR,
    backgroundImage: checker_1.CHECKER_BACKGROUND_IMAGE,
    backgroundSize: checker_1.CHECKER_BACKGROUND_SIZE,
    backgroundPosition: checker_1.CHECKER_BACKGROUND_POSITION,
    touchAction: 'none',
};
const innerStyle = {
    position: 'absolute',
    inset: 0,
};
const AlphaSlider = ({ hue, saturation, value, alpha, onChange, onChangeComplete }) => {
    const ref = (0, react_1.useRef)(null);
    const opaqueColor = (0, react_1.useMemo)(() => {
        const { r, g, b } = (0, color_conversion_1.hsvToRgb)({ h: hue, s: saturation, v: value });
        return `rgb(${r}, ${g}, ${b})`;
    }, [hue, saturation, value]);
    const gradientStyle = (0, react_1.useMemo)(() => {
        return {
            ...innerStyle,
            background: `linear-gradient(to right, rgba(0, 0, 0, 0), ${opaqueColor})`,
        };
    }, [opaqueColor]);
    const updateFromEvent = (0, react_1.useCallback)((clientX, isFinal) => {
        const { current } = ref;
        if (!current) {
            return;
        }
        const rect = current.getBoundingClientRect();
        const next = (0, color_conversion_1.clamp)((clientX - rect.left) / rect.width, 0, 1);
        if (isFinal) {
            onChangeComplete(next);
        }
        else {
            onChange(next);
        }
    }, [onChange, onChangeComplete]);
    const onPointerDown = (0, react_1.useCallback)((e) => {
        if (e.button !== 0) {
            return;
        }
        e.preventDefault();
        updateFromEvent(e.clientX, false);
        const onMove = (ev) => {
            updateFromEvent(ev.clientX, false);
        };
        const onUp = (ev) => {
            window.removeEventListener('pointermove', onMove);
            window.removeEventListener('pointerup', onUp);
            updateFromEvent(ev.clientX, true);
        };
        window.addEventListener('pointermove', onMove);
        window.addEventListener('pointerup', onUp);
    }, [updateFromEvent]);
    const handleStyle = (0, react_1.useMemo)(() => {
        return {
            position: 'absolute',
            top: -2,
            left: `${alpha * 100}%`,
            marginLeft: -HANDLE_WIDTH / 2,
            width: HANDLE_WIDTH,
            height: SLIDER_HEIGHT + 4,
            borderRadius: 2,
            border: '2px solid #fff',
            boxShadow: '0 0 0 1px rgba(0, 0, 0, 0.6)',
            pointerEvents: 'none',
        };
    }, [alpha]);
    return (jsx_runtime_1.jsxs("div", { ref: ref, style: wrapperStyle, onPointerDown: onPointerDown, children: [
            jsx_runtime_1.jsx("div", { style: gradientStyle }), jsx_runtime_1.jsx("div", { style: handleStyle })
        ] }));
};
exports.AlphaSlider = AlphaSlider;
