// @remotion/studio/dist/components/ColorPicker/ColorPickerPopup.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ColorPickerPopup = exports.parseEyeDropperColor = exports.POPUP_WIDTH = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const react_1 = require("react");
const color_conversion_1 = require("../../helpers/color-conversion");
const colors_1 = require("../../helpers/colors");
const eyedropper_1 = require("../../icons/eyedropper");
const z_index_1 = require("../../state/z-index");
const AlphaSlider_1 = require("./AlphaSlider");
const checker_1 = require("./checker");
const HueSlider_1 = require("./HueSlider");
const SaturationValueArea_1 = require("./SaturationValueArea");
exports.POPUP_WIDTH = 240;
const POPUP_PADDING = 12;
const popupShellStyle = {
    width: exports.POPUP_WIDTH,
    padding: POPUP_PADDING,
    background: '#1f2428',
    border: '1px solid rgba(0, 0, 0, 0.6)',
    borderRadius: 4,
    boxShadow: '0 4px 16px rgba(0, 0, 0, 0.5)',
    display: 'flex',
    flexDirection: 'column',
    gap: 10,
    color: 'white',
    fontSize: 12,
    userSelect: 'none',
    WebkitUserSelect: 'none',
};
const sliderRowStyle = {
    display: 'flex',
    alignItems: 'center',
    gap: 8,
};
const previewSwatchStyle = {
    position: 'relative',
    width: 28,
    height: 28,
    flex: '0 0 auto',
    borderRadius: 14,
    overflow: 'hidden',
    border: '1px solid rgba(0, 0, 0, 0.6)',
    backgroundColor: checker_1.CHECKER_BACKGROUND_COLOR,
    backgroundImage: checker_1.CHECKER_BACKGROUND_IMAGE,
    backgroundSize: checker_1.CHECKER_BACKGROUND_SIZE,
    backgroundPosition: checker_1.CHECKER_BACKGROUND_POSITION,
};
const previewSwatchFillBase = {
    position: 'absolute',
    inset: 0,
};
const slidersColumnStyle = {
    display: 'flex',
    flexDirection: 'column',
    gap: 6,
    flex: 1,
};
const inputsRowStyle = {
    display: 'flex',
    gap: 6,
    alignItems: 'flex-end',
};
const inputColumnStyle = {
    display: 'flex',
    flexDirection: 'column',
    flex: 1,
    minWidth: 0,
};
const labelStyle = {
    color: colors_1.LIGHT_TEXT,
    fontSize: 10,
    textTransform: 'uppercase',
    letterSpacing: 0.5,
    marginBottom: 2,
};
const baseInputStyle = {
    width: '100%',
    background: colors_1.INPUT_BACKGROUND,
    border: '1px solid rgba(0, 0, 0, 0.6)',
    borderRadius: 3,
    color: 'white',
    padding: '4px 6px',
    fontSize: 12,
    textAlign: 'center',
    boxSizing: 'border-box',
};
const eyedropperButtonStyle = {
    background: 'transparent',
    border: '1px solid rgba(0, 0, 0, 0.6)',
    borderRadius: 3,
    padding: 0,
    cursor: 'pointer',
    color: colors_1.LIGHT_TEXT,
    width: 28,
    height: 28,
    display: 'inline-flex',
    alignItems: 'center',
    justifyContent: 'center',
    flex: '0 0 auto',
};
const hasEyeDropper = () => typeof window !== 'undefined' && 'EyeDropper' in window;
const parseEyeDropperColor = (pickedColor) => {
    const parsed = (0, color_conversion_1.parseAnyColor)(pickedColor);
    return {
        ...parsed,
        a: 255,
    };
};
exports.parseEyeDropperColor = parseEyeDropperColor;
const ChannelInput = ({ label, channel, value, min, max, onCommit, }) => {
    const [draft, setDraft] = (0, react_1.useState)(String(Math.round(value)));
    const { tabIndex } = (0, z_index_1.useZIndex)();
    (0, react_1.useEffect)(() => {
        setDraft(String(Math.round(value)));
    }, [value]);
    const onChange = (0, react_1.useCallback)((e) => {
        setDraft(e.target.value);
    }, []);
    const onBlur = (0, react_1.useCallback)(() => {
        const parsed = Number(draft);
        if (Number.isNaN(parsed)) {
            setDraft(String(Math.round(value)));
            return;
        }
        const clamped = Math.max(min, Math.min(max, parsed));
        setDraft(String(Math.round(clamped)));
        onCommit(channel, clamped);
    }, [channel, draft, max, min, onCommit, value]);
    const onKeyDown = (0, react_1.useCallback)((e) => {
        if (e.key === 'Enter') {
            e.target.blur();
        }
    }, []);
    return (jsx_runtime_1.jsxs("label", { style: inputColumnStyle, children: [
            jsx_runtime_1.jsx("span", { style: labelStyle, children: label }), jsx_runtime_1.jsx("input", { type: "text", inputMode: "numeric", value: draft, onChange: onChange, onBlur: onBlur, onKeyDown: onKeyDown, style: baseInputStyle, tabIndex: tabIndex })
        ] }));
};
const HexInput = ({ value, onCommit }) => {
    const [draft, setDraft] = (0, react_1.useState)(value);
    const { tabIndex } = (0, z_index_1.useZIndex)();
    (0, react_1.useEffect)(() => {
        setDraft(value);
    }, [value]);
    const onChange = (0, react_1.useCallback)((e) => {
        setDraft(e.target.value);
    }, []);
    const onBlur = (0, react_1.useCallback)(() => {
        try {
            const parsed = (0, color_conversion_1.parseAnyColor)(draft);
            onCommit((0, color_conversion_1.formatRgba)(parsed));
        }
        catch (_a) {
            setDraft(value);
        }
    }, [draft, onCommit, value]);
    const onKeyDown = (0, react_1.useCallback)((e) => {
        if (e.key === 'Enter') {
            e.target.blur();
        }
    }, []);
    const style = (0, react_1.useMemo)(() => {
        return {
            ...baseInputStyle,
            textAlign: 'left',
        };
    }, []);
    return (jsx_runtime_1.jsxs("label", { style: inputColumnStyle, children: [
            jsx_runtime_1.jsx("span", { style: labelStyle, children: "Hex" }), jsx_runtime_1.jsx("input", { type: "text", value: draft, onChange: onChange, onBlur: onBlur, onKeyDown: onKeyDown, style: style, autoComplete: "off", spellCheck: false, tabIndex: tabIndex })
        ] }));
};
const ColorPickerPopup = ({ value, onChange, onChangeComplete }) => {
    // useZIndex is intentionally read inside the popup (which is wrapped by
    // HigherZIndex) so child inputs receive the popup's tabIndex rather than
    // the trigger's, which would resolve to -1 once the popup is open.
    const { tabIndex } = (0, z_index_1.useZIndex)();
    // HSV is the source of truth while the picker is open. We seed it from
    // the incoming value but keep our own state so that hue and saturation
    // don't snap back to 0 when the user drags into pure black/white.
    const [hsva, setHsva] = (0, react_1.useState)(() => (0, color_conversion_1.rgbaToHsva)((0, color_conversion_1.parseAnyColor)(value)));
    const lastEmittedRef = (0, react_1.useRef)((0, color_conversion_1.formatRgba)((0, color_conversion_1.parseAnyColor)(value)));
    (0, react_1.useEffect)(() => {
        // Sync from external value changes that didn't originate from us.
        const incoming = (0, color_conversion_1.formatRgba)((0, color_conversion_1.parseAnyColor)(value));
        if (incoming === lastEmittedRef.current) {
            return;
        }
        lastEmittedRef.current = incoming;
        setHsva((0, color_conversion_1.rgbaToHsva)((0, color_conversion_1.parseAnyColor)(value)));
    }, [value]);
    const rgba = (0, react_1.useMemo)(() => (0, color_conversion_1.hsvaToRgba)(hsva), [hsva]);
    const formatted = (0, react_1.useMemo)(() => (0, color_conversion_1.formatRgba)(rgba), [rgba]);
    const emit = (0, react_1.useCallback)((next, mode) => {
        setHsva(next);
        const nextRgba = (0, color_conversion_1.hsvaToRgba)(next);
        const nextFormatted = (0, color_conversion_1.formatRgba)(nextRgba);
        lastEmittedRef.current = nextFormatted;
        if (mode === 'complete') {
            onChangeComplete(nextFormatted);
        }
        else {
            onChange(nextFormatted);
        }
    }, [onChange, onChangeComplete]);
    const onSvChange = (0, react_1.useCallback)((next) => {
        emit({ ...hsva, s: next.s, v: next.v }, 'change');
    }, [emit, hsva]);
    const onSvComplete = (0, react_1.useCallback)((next) => {
        emit({ ...hsva, s: next.s, v: next.v }, 'complete');
    }, [emit, hsva]);
    const onHueChange = (0, react_1.useCallback)((next) => {
        emit({ ...hsva, h: next }, 'change');
    }, [emit, hsva]);
    const onHueComplete = (0, react_1.useCallback)((next) => {
        emit({ ...hsva, h: next }, 'complete');
    }, [emit, hsva]);
    const onAlphaChange = (0, react_1.useCallback)((next) => {
        emit({ ...hsva, a: next }, 'change');
    }, [emit, hsva]);
    const onAlphaComplete = (0, react_1.useCallback)((next) => {
        emit({ ...hsva, a: next }, 'complete');
    }, [emit, hsva]);
    const onChannelCommit = (0, react_1.useCallback)((channel, next) => {
        if (channel === 'a-percent') {
            const clamped = Math.max(0, Math.min(100, next));
            emit({ ...hsva, a: clamped / 100 }, 'complete');
            return;
        }
        const updatedRgba = { ...rgba, [channel]: next };
        const newHsva = (0, color_conversion_1.rgbaToHsva)(updatedRgba);
        // Preserve hue when transitioning through achromatic colors.
        if (newHsva.s === 0) {
            newHsva.h = hsva.h;
        }
        emit(newHsva, 'complete');
    }, [emit, hsva, rgba]);
    const onHexCommit = (0, react_1.useCallback)((next) => {
        const parsed = (0, color_conversion_1.parseAnyColor)(next);
        const newHsva = (0, color_conversion_1.rgbaToHsva)(parsed);
        if (newHsva.s === 0) {
            newHsva.h = hsva.h;
        }
        emit(newHsva, 'complete');
    }, [emit, hsva.h]);
    const onPickWithEyeDropper = (0, react_1.useCallback)(() => {
        const Ctor = window.EyeDropper;
        if (!Ctor) {
            return;
        }
        const dropper = new Ctor();
        dropper
            .open()
            .then((result) => {
            const parsed = (0, exports.parseEyeDropperColor)(result.sRGBHex);
            const newHsva = (0, color_conversion_1.rgbaToHsva)(parsed);
            if (newHsva.s === 0) {
                newHsva.h = hsva.h;
            }
            emit(newHsva, 'complete');
        })
            .catch(() => {
            // Aborted; ignore.
        });
    }, [emit, hsva.h]);
    const previewFill = (0, react_1.useMemo)(() => {
        return {
            ...previewSwatchFillBase,
            backgroundColor: formatted,
        };
    }, [formatted]);
    const showEyeDropper = hasEyeDropper();
    return (jsx_runtime_1.jsxs("div", { style: popupShellStyle, children: [
            jsx_runtime_1.jsx(SaturationValueArea_1.SaturationValueArea, { hue: hsva.h, saturation: hsva.s, value: hsva.v, onChange: onSvChange, onChangeComplete: onSvComplete }), jsx_runtime_1.jsxs("div", { style: sliderRowStyle, children: [showEyeDropper ? (jsx_runtime_1.jsx("button", { type: "button", style: eyedropperButtonStyle, onClick: onPickWithEyeDropper, tabIndex: tabIndex, title: "Pick color from screen", "aria-label": "Pick color from screen", children: jsx_runtime_1.jsx(eyedropper_1.EyedropperIcon, { style: { width: 14, height: 14 }, color: "currentColor" }) })) : null, jsx_runtime_1.jsx("div", { style: previewSwatchStyle, title: formatted, children: jsx_runtime_1.jsx("div", { style: previewFill }) }), jsx_runtime_1.jsxs("div", { style: slidersColumnStyle, children: [
                            jsx_runtime_1.jsx(HueSlider_1.HueSlider, { hue: hsva.h, onChange: onHueChange, onChangeComplete: onHueComplete }), jsx_runtime_1.jsx(AlphaSlider_1.AlphaSlider, { hue: hsva.h, saturation: hsva.s, value: hsva.v, alpha: hsva.a, onChange: onAlphaChange, onChangeComplete: onAlphaComplete })
                        ] })
                ] }), jsx_runtime_1.jsxs("div", { style: inputsRowStyle, children: [
                    jsx_runtime_1.jsx(HexInput, { value: formatted, onCommit: onHexCommit }), jsx_runtime_1.jsx(ChannelInput, { label: "R", channel: "r", value: rgba.r, min: 0, max: 255, onCommit: onChannelCommit }), jsx_runtime_1.jsx(ChannelInput, { label: "G", channel: "g", value: rgba.g, min: 0, max: 255, onCommit: onChannelCommit }), jsx_runtime_1.jsx(ChannelInput, { label: "B", channel: "b", value: rgba.b, min: 0, max: 255, onCommit: onChannelCommit }), jsx_runtime_1.jsx(ChannelInput, { label: "A%", channel: "a-percent", value: Math.round(hsva.a * 100), min: 0, max: 100, onCommit: onChannelCommit })
                ] })
        ] }));
};
exports.ColorPickerPopup = ColorPickerPopup;
