// @remotion/studio/dist/components/ColorPicker/ColorPicker.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ColorPicker = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const player_1 = require("@remotion/player");
const react_1 = require("react");
const react_dom_1 = __importDefault(require("react-dom"));
const z_index_1 = require("../../state/z-index");
const is_menu_item_1 = require("../Menu/is-menu-item");
const portals_1 = require("../Menu/portals");
const styles_1 = require("../Menu/styles");
const checker_1 = require("./checker");
const ColorPickerPopup_1 = require("./ColorPickerPopup");
// Class name used to opt the swatch button out of the global
// `button:focus` inset box-shadow defined in inject-css.ts.
const SWATCH_CLASSNAME = '__remotion_color_swatch';
const swatchBaseStyle = {
    position: 'relative',
    display: 'inline-block',
    overflow: 'hidden',
    padding: 0,
    margin: 0,
    border: 'none',
    cursor: 'pointer',
    backgroundColor: checker_1.CHECKER_BACKGROUND_COLOR,
    backgroundImage: checker_1.CHECKER_BACKGROUND_IMAGE,
    backgroundSize: checker_1.CHECKER_BACKGROUND_SIZE,
    backgroundPosition: checker_1.CHECKER_BACKGROUND_POSITION,
    boxSizing: 'border-box',
};
const fillStyle = {
    position: 'absolute',
    inset: 0,
    display: 'block',
};
const ColorPicker = ({ value, onChange, onChangeComplete, disabled, width = 45, height = 25, borderRadius = 3, title, name, className, style: customStyle, }) => {
    const [opened, setOpened] = (0, react_1.useState)(false);
    const triggerRef = (0, react_1.useRef)(null);
    const { tabIndex, currentZIndex } = (0, z_index_1.useZIndex)();
    const size = player_1.PlayerInternals.useElementSize(triggerRef, {
        triggerOnWindowResize: true,
        shouldApplyCssTransforms: true,
    });
    const refresh = size === null || size === void 0 ? void 0 : size.refresh;
    const onHide = (0, react_1.useCallback)(() => {
        setOpened(false);
    }, []);
    const swatchFill = (0, react_1.useMemo)(() => {
        return {
            ...fillStyle,
            backgroundColor: value,
        };
    }, [value]);
    const swatchStyle = (0, react_1.useMemo)(() => {
        return {
            ...swatchBaseStyle,
            width,
            height,
            borderRadius,
            cursor: disabled ? 'not-allowed' : 'pointer',
            opacity: disabled ? 0.5 : 1,
            ...(customStyle !== null && customStyle !== void 0 ? customStyle : {}),
        };
    }, [borderRadius, customStyle, disabled, height, width]);
    // Toggle on pointerdown (matches Combobox) so the state flips before the
    // HigherZIndex outside-click detection runs on pointerup. If we toggled in
    // onClick, the popup would close in pointerup and immediately re-open in
    // the click that follows.
    const onTriggerPointerDown = (0, react_1.useCallback)(() => {
        if (disabled) {
            return;
        }
        setOpened((prev) => {
            if (!prev) {
                refresh === null || refresh === void 0 ? void 0 : refresh();
            }
            return !prev;
        });
    }, [disabled, refresh]);
    const onTriggerClick = (0, react_1.useCallback)((e) => {
        if (disabled) {
            return;
        }
        // Keyboard-initiated activations (Enter/Space) don't fire pointerdown,
        // so handle them here. Mouse clicks are already handled by pointerdown
        // above; ignore them.
        const isKeyboardInitiated = e.detail === 0;
        if (!isKeyboardInitiated) {
            return;
        }
        setOpened((prev) => {
            if (!prev) {
                refresh === null || refresh === void 0 ? void 0 : refresh();
            }
            return !prev;
        });
    }, [disabled, refresh]);
    const portalStyle = (0, react_1.useMemo)(() => {
        if (!opened || !size) {
            return null;
        }
        const margin = 6;
        const popupHeight = 250;
        const popupWidth = ColorPickerPopup_1.POPUP_WIDTH;
        const spaceBelow = size.windowSize.height - (size.top + size.height);
        const spaceAbove = size.top;
        const openBelow = spaceBelow >= popupHeight + margin || spaceBelow >= spaceAbove;
        const spaceRight = size.windowSize.width - size.left;
        // Try aligning the popup to the trigger's left edge first; fall back
        // to right-aligning if it would clip the viewport.
        const left = spaceRight >= popupWidth + margin
            ? size.left
            : Math.max(margin, size.windowSize.width - popupWidth - margin);
        if (openBelow) {
            return {
                ...styles_1.menuContainerTowardsBottom,
                top: size.top + size.height + margin,
                left,
                background: 'transparent',
                boxShadow: 'none',
            };
        }
        return {
            ...styles_1.menuContainerTowardsTop,
            bottom: size.windowSize.height - size.top + margin,
            left,
            background: 'transparent',
            boxShadow: 'none',
        };
    }, [opened, size]);
    (0, react_1.useEffect)(() => {
        if (!opened) {
            return;
        }
        const onWindowBlur = () => setOpened(false);
        window.addEventListener('blur', onWindowBlur);
        return () => window.removeEventListener('blur', onWindowBlur);
    }, [opened]);
    return (jsx_runtime_1.jsxs(jsx_runtime_1.Fragment, { children: [
            jsx_runtime_1.jsx("button", { ref: triggerRef, type: "button", className: [is_menu_item_1.MENU_INITIATOR_CLASSNAME, SWATCH_CLASSNAME, className]
                    .filter(Boolean)
                    .join(' '), disabled: disabled, name: name, title: title !== null && title !== void 0 ? title : value, tabIndex: tabIndex, style: swatchStyle, onPointerDown: onTriggerPointerDown, onClick: onTriggerClick, children: jsx_runtime_1.jsx("span", { style: swatchFill }) }), portalStyle
                ? react_dom_1.default.createPortal(jsx_runtime_1.jsx("div", { style: styles_1.fullScreenOverlay, children: jsx_runtime_1.jsx("div", { style: styles_1.outerPortal, className: "css-reset", children: jsx_runtime_1.jsx(z_index_1.HigherZIndex, { onOutsideClick: onHide, onEscape: onHide, children: jsx_runtime_1.jsx("div", { style: portalStyle, children: jsx_runtime_1.jsx(ColorPickerPopup_1.ColorPickerPopup, { value: value, onChange: onChange, onChangeComplete: onChangeComplete }) }) }) }) }), (0, portals_1.getPortal)(currentZIndex))
                : null] }));
};
exports.ColorPicker = ColorPicker;
