// @remotion/studio/dist/components/ColorPicker/HueSlider.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.HueSlider = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const react_1 = require("react");
const color_conversion_1 = require("../../helpers/color-conversion");
const SLIDER_HEIGHT = 12;
const HANDLE_WIDTH = 8;
const containerStyle = {
    position: 'relative',
    height: SLIDER_HEIGHT,
    width: '100%',
    borderRadius: 3,
    cursor: 'ew-resize',
    background: 'linear-gradient(to right, #f00 0%, #ff0 17%, #0f0 33%, #0ff 50%, #00f 67%, #f0f 83%, #f00 100%)',
    touchAction: 'none',
};
const HueSlider = ({ hue, onChange, onChangeComplete }) => {
    const ref = (0, react_1.useRef)(null);
    const updateFromEvent = (0, react_1.useCallback)((clientX, isFinal) => {
        const { current } = ref;
        if (!current) {
            return;
        }
        const rect = current.getBoundingClientRect();
        const x = (0, color_conversion_1.clamp)((clientX - rect.left) / rect.width, 0, 1);
        const next = x * 360;
        if (isFinal) {
            onChangeComplete(next);
        }
        else {
            onChange(next);
        }
    }, [onChange, onChangeComplete]);
    const onPointerDown = (0, react_1.useCallback)((e) => {
        if (e.button !== 0) {
            return;
        }
        e.preventDefault();
        updateFromEvent(e.clientX, false);
        const onMove = (ev) => {
            updateFromEvent(ev.clientX, false);
        };
        const onUp = (ev) => {
            window.removeEventListener('pointermove', onMove);
            window.removeEventListener('pointerup', onUp);
            updateFromEvent(ev.clientX, true);
        };
        window.addEventListener('pointermove', onMove);
        window.addEventListener('pointerup', onUp);
    }, [updateFromEvent]);
    const handleStyle = (0, react_1.useMemo)(() => {
        return {
            position: 'absolute',
            top: -2,
            left: `${(hue / 360) * 100}%`,
            marginLeft: -HANDLE_WIDTH / 2,
            width: HANDLE_WIDTH,
            height: SLIDER_HEIGHT + 4,
            borderRadius: 2,
            border: '2px solid #fff',
            boxShadow: '0 0 0 1px rgba(0, 0, 0, 0.6)',
            pointerEvents: 'none',
        };
    }, [hue]);
    return (jsx_runtime_1.jsx("div", { ref: ref, style: containerStyle, onPointerDown: onPointerDown, children: jsx_runtime_1.jsx("div", { style: handleStyle }) }));
};
exports.HueSlider = HueSlider;
