// @remotion/studio/dist/components/ColorPicker/SaturationValueArea.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SaturationValueArea = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const react_1 = require("react");
const color_conversion_1 = require("../../helpers/color-conversion");
const AREA_HEIGHT = 140;
const containerStyle = {
    position: 'relative',
    width: '100%',
    height: AREA_HEIGHT,
    cursor: 'crosshair',
    borderRadius: 3,
    overflow: 'hidden',
    touchAction: 'none',
};
const valueOverlay = {
    position: 'absolute',
    inset: 0,
    background: 'linear-gradient(to top, #000, transparent)',
};
const saturationOverlay = {
    position: 'absolute',
    inset: 0,
    background: 'linear-gradient(to right, #fff, transparent)',
};
const SaturationValueArea = ({ hue, saturation, value, onChange, onChangeComplete }) => {
    const containerRef = (0, react_1.useRef)(null);
    const baseColor = (0, react_1.useMemo)(() => {
        const { r, g, b } = (0, color_conversion_1.hsvToRgb)({ h: hue, s: 1, v: 1 });
        return `rgb(${r}, ${g}, ${b})`;
    }, [hue]);
    const updateFromEvent = (0, react_1.useCallback)((clientX, clientY, isFinal) => {
        const { current } = containerRef;
        if (!current) {
            return;
        }
        const rect = current.getBoundingClientRect();
        const x = (0, color_conversion_1.clamp)((clientX - rect.left) / rect.width, 0, 1);
        const y = (0, color_conversion_1.clamp)((clientY - rect.top) / rect.height, 0, 1);
        const next = { s: x, v: 1 - y };
        if (isFinal) {
            onChangeComplete(next);
        }
        else {
            onChange(next);
        }
    }, [onChange, onChangeComplete]);
    const onPointerDown = (0, react_1.useCallback)((e) => {
        var _a, _b;
        if (e.button !== 0) {
            return;
        }
        e.preventDefault();
        (_b = (_a = e.target).setPointerCapture) === null || _b === void 0 ? void 0 : _b.call(_a, e.pointerId);
        updateFromEvent(e.clientX, e.clientY, false);
        const onMove = (ev) => {
            updateFromEvent(ev.clientX, ev.clientY, false);
        };
        const onUp = (ev) => {
            window.removeEventListener('pointermove', onMove);
            window.removeEventListener('pointerup', onUp);
            updateFromEvent(ev.clientX, ev.clientY, true);
        };
        window.addEventListener('pointermove', onMove);
        window.addEventListener('pointerup', onUp);
    }, [updateFromEvent]);
    const containerWithBg = (0, react_1.useMemo)(() => {
        return {
            ...containerStyle,
            backgroundColor: baseColor,
        };
    }, [baseColor]);
    const handleStyle = (0, react_1.useMemo)(() => {
        return {
            position: 'absolute',
            left: `${saturation * 100}%`,
            top: `${(1 - value) * 100}%`,
            width: 12,
            height: 12,
            marginLeft: -6,
            marginTop: -6,
            borderRadius: '50%',
            border: '2px solid #fff',
            boxShadow: '0 0 0 1px rgba(0, 0, 0, 0.6)',
            pointerEvents: 'none',
        };
    }, [saturation, value]);
    return (jsx_runtime_1.jsxs("div", { ref: containerRef, style: containerWithBg, onPointerDown: onPointerDown, children: [
            jsx_runtime_1.jsx("div", { style: saturationOverlay }), jsx_runtime_1.jsx("div", { style: valueOverlay }), jsx_runtime_1.jsx("div", { style: handleStyle })
        ] }));
};
exports.SaturationValueArea = SaturationValueArea;
