// @remotion/studio/dist/helpers/color-conversion.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.hsvaToRgba = exports.rgbaToHsva = exports.hsvToRgb = exports.rgbToHsv = exports.formatRgba = exports.formatRgbHex = exports.parseAnyColor = exports.clamp = void 0;
const no_react_1 = require("remotion/no-react");
const clamp = (value, min, max) => {
    return Math.min(Math.max(value, min), max);
};
exports.clamp = clamp;
const componentToHex = (c) => {
    return Math.round((0, exports.clamp)(c, 0, 255))
        .toString(16)
        .padStart(2, '0');
};
// Always returns a normalized {r, g, b, a (0-255)} object,
// regardless of whether the input was a hex, rgb(), rgba(), or named color.
// Falls back to opaque black when parsing fails.
const parseAnyColor = (input) => {
    try {
        const argb = no_react_1.NoReactInternals.processColor(input)
            .toString(16)
            .padStart(8, '0');
        const a = parseInt(argb.slice(0, 2), 16);
        const r = parseInt(argb.slice(2, 4), 16);
        const g = parseInt(argb.slice(4, 6), 16);
        const b = parseInt(argb.slice(6, 8), 16);
        return { r, g, b, a };
    }
    catch (_a) {
        return { r: 0, g: 0, b: 0, a: 255 };
    }
};
exports.parseAnyColor = parseAnyColor;
const formatRgbHex = ({ r, g, b }) => {
    return `#${componentToHex(r)}${componentToHex(g)}${componentToHex(b)}`;
};
exports.formatRgbHex = formatRgbHex;
const formatRgba = ({ r, g, b, a }) => {
    if (a >= 255) {
        return (0, exports.formatRgbHex)({ r, g, b });
    }
    const opacity = Number((a / 255).toFixed(2));
    return `rgba(${Math.round(r)}, ${Math.round(g)}, ${Math.round(b)}, ${opacity})`;
};
exports.formatRgba = formatRgba;
const rgbToHsv = ({ r, g, b, }) => {
    const rn = r / 255;
    const gn = g / 255;
    const bn = b / 255;
    const max = Math.max(rn, gn, bn);
    const min = Math.min(rn, gn, bn);
    const delta = max - min;
    let h = 0;
    if (delta !== 0) {
        if (max === rn) {
            h = ((gn - bn) / delta) % 6;
        }
        else if (max === gn) {
            h = (bn - rn) / delta + 2;
        }
        else {
            h = (rn - gn) / delta + 4;
        }
        h *= 60;
        if (h < 0) {
            h += 360;
        }
    }
    const s = max === 0 ? 0 : delta / max;
    const v = max;
    return { h, s, v, a: 1 };
};
exports.rgbToHsv = rgbToHsv;
const hsvToRgb = ({ h, s, v, }) => {
    const c = v * s;
    const hh = (h % 360) / 60;
    const x = c * (1 - Math.abs((hh % 2) - 1));
    let r1 = 0;
    let g1 = 0;
    let b1 = 0;
    if (hh >= 0 && hh < 1) {
        r1 = c;
        g1 = x;
    }
    else if (hh >= 1 && hh < 2) {
        r1 = x;
        g1 = c;
    }
    else if (hh >= 2 && hh < 3) {
        g1 = c;
        b1 = x;
    }
    else if (hh >= 3 && hh < 4) {
        g1 = x;
        b1 = c;
    }
    else if (hh >= 4 && hh < 5) {
        r1 = x;
        b1 = c;
    }
    else {
        r1 = c;
        b1 = x;
    }
    const m = v - c;
    return {
        r: Math.round((r1 + m) * 255),
        g: Math.round((g1 + m) * 255),
        b: Math.round((b1 + m) * 255),
    };
};
exports.hsvToRgb = hsvToRgb;
const rgbaToHsva = (rgba) => {
    const { h, s, v } = (0, exports.rgbToHsv)(rgba);
    return { h, s, v, a: rgba.a / 255 };
};
exports.rgbaToHsva = rgbaToHsva;
const hsvaToRgba = (hsva) => {
    const { r, g, b } = (0, exports.hsvToRgb)(hsva);
    return { r, g, b, a: Math.round(hsva.a * 255) };
};
exports.hsvaToRgba = hsvaToRgba;
