// @remotion/studio/dist/helpers/open-in-editor.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.openCompositionComponentInEditor = exports.preloadCompositionComponentInfo = exports.loadCompositionComponentInfo = exports.useCachedCompositionComponentInfo = exports.getCachedCompositionComponentInfo = exports.subscribeToCompositionComponentInfo = exports.openOriginalPositionInEditor = exports.openInEditor = void 0;
const react_1 = require("react");
const call_api_1 = require("../components/call-api");
const openInEditor = (stack) => {
    const { originalFileName, originalLineNumber, originalColumnNumber, originalFunctionName, originalScriptCode, } = stack;
    return (0, call_api_1.callApi)('/api/open-in-editor', {
        stack: {
            originalFileName,
            originalLineNumber,
            originalColumnNumber,
            originalFunctionName,
            originalScriptCode,
        },
    });
};
exports.openInEditor = openInEditor;
const openOriginalPositionInEditor = async (originalPosition) => {
    await (0, exports.openInEditor)({
        originalColumnNumber: originalPosition.column,
        originalFileName: originalPosition.source,
        originalFunctionName: null,
        originalLineNumber: originalPosition.line,
        originalScriptCode: null,
    });
};
exports.openOriginalPositionInEditor = openOriginalPositionInEditor;
const componentResolutionCache = new Map();
const componentResolutionResults = new Map();
const componentResolutionListeners = new Set();
const getComponentResolutionCacheKey = ({ compositionFile, compositionId, }) => {
    return `${compositionFile}::${compositionId}`;
};
const notifyComponentResolutionListeners = () => {
    for (const listener of componentResolutionListeners) {
        listener();
    }
};
const subscribeToCompositionComponentInfo = (listener) => {
    componentResolutionListeners.add(listener);
    return () => {
        componentResolutionListeners.delete(listener);
    };
};
exports.subscribeToCompositionComponentInfo = subscribeToCompositionComponentInfo;
const getCachedCompositionComponentInfo = ({ compositionFile, compositionId, }) => {
    var _a;
    return ((_a = componentResolutionResults.get(getComponentResolutionCacheKey({ compositionFile, compositionId }))) !== null && _a !== void 0 ? _a : null);
};
exports.getCachedCompositionComponentInfo = getCachedCompositionComponentInfo;
const useCachedCompositionComponentInfo = ({ compositionFile, compositionId, }) => {
    return (0, react_1.useSyncExternalStore)(exports.subscribeToCompositionComponentInfo, () => {
        if (compositionFile === null || compositionId === null) {
            return null;
        }
        return (0, exports.getCachedCompositionComponentInfo)({
            compositionFile,
            compositionId,
        });
    }, () => null);
};
exports.useCachedCompositionComponentInfo = useCachedCompositionComponentInfo;
const loadCompositionComponentInfo = async ({ compositionFile, compositionId, }) => {
    const cacheKey = getComponentResolutionCacheKey({
        compositionFile,
        compositionId,
    });
    const existing = componentResolutionCache.get(cacheKey);
    if (existing) {
        return existing;
    }
    const promise = (async () => {
        const body = await (0, call_api_1.callApi)('/api/composition-component-info', {
            compositionFile,
            compositionId,
        });
        const result = {
            location: body.location,
            canAddSequence: body.canAddSequence,
        };
        componentResolutionResults.set(cacheKey, result);
        notifyComponentResolutionListeners();
        return result;
    })();
    componentResolutionCache.set(cacheKey, promise);
    try {
        return await promise;
    }
    catch (err) {
        componentResolutionCache.delete(cacheKey);
        componentResolutionResults.delete(cacheKey);
        notifyComponentResolutionListeners();
        throw err;
    }
};
exports.loadCompositionComponentInfo = loadCompositionComponentInfo;
const preloadCompositionComponentInfo = ({ compositionFile, compositionId, }) => {
    (0, exports.loadCompositionComponentInfo)({
        compositionFile,
        compositionId,
    }).catch(() => undefined);
};
exports.preloadCompositionComponentInfo = preloadCompositionComponentInfo;
const openCompositionComponentInEditor = async ({ compositionFile, compositionId, }) => {
    const info = await (0, exports.loadCompositionComponentInfo)({
        compositionFile,
        compositionId,
    });
    await (0, exports.openOriginalPositionInEditor)(info.location);
};
exports.openCompositionComponentInEditor = openCompositionComponentInEditor;
