// @remotion/studio/dist/helpers/timeline-layout.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getTimelineLayerHeight = exports.TIMELINE_LIST_ITEM_ROW_HEIGHT = exports.TIMELINE_LAYER_HEIGHT_DEFAULT = exports.TIMELINE_LAYER_HEIGHT_AUDIO = exports.TIMELINE_LAYER_HEIGHT_IMAGE = exports.TIMELINE_LAYER_HEIGHT_VIDEO = exports.TIMELINE_VIDEO_INFO_WAVEFORM_HEIGHT = exports.TIMELINE_LAYER_FILMSTRIP_HEIGHT = exports.getExpandedTrackHeight = exports.getTreeRowHeight = exports.flattenVisibleTreeNodes = exports.buildTimelineTree = exports.EXPANDED_SECTION_PADDING_RIGHT = exports.TREE_GROUP_ROW_HEIGHT = exports.TIMELINE_TRACK_EXPANDED_HEIGHT = exports.TIMELINE_ITEM_BORDER_BOTTOM = exports.TIMELINE_BORDER = exports.TIMELINE_PADDING = exports.SCHEMA_FIELD_ROW_HEIGHT = exports.getFieldsToShow = exports.getEffectFieldsToShow = void 0;
const studio_shared_1 = require("@remotion/studio-shared");
const studio_shared_2 = require("@remotion/studio-shared");
Object.defineProperty(exports, "getEffectFieldsToShow", { enumerable: true, get: function () { return studio_shared_2.getEffectFieldsToShow; } });
Object.defineProperty(exports, "getFieldsToShow", { enumerable: true, get: function () { return studio_shared_2.getFieldsToShow; } });
Object.defineProperty(exports, "SCHEMA_FIELD_ROW_HEIGHT", { enumerable: true, get: function () { return studio_shared_2.SCHEMA_FIELD_ROW_HEIGHT; } });
exports.TIMELINE_PADDING = 16;
exports.TIMELINE_BORDER = 1;
exports.TIMELINE_ITEM_BORDER_BOTTOM = 1;
exports.TIMELINE_TRACK_EXPANDED_HEIGHT = 100;
exports.TREE_GROUP_ROW_HEIGHT = 22;
exports.EXPANDED_SECTION_PADDING_RIGHT = 10;
const buildTimelineTree = ({ sequence, nodePathInfo, getDragOverrides, getEffectDragOverrides, propStatuses, }) => {
    var _a;
    const roots = [];
    const { sequenceSubscriptionKey, index, auxiliaryKeys, supportsEffects } = nodePathInfo;
    const controlFields = (0, studio_shared_1.getFieldsToShow)({
        schema: sequence.controls.schema,
        currentRuntimeValueDotNotation: sequence.controls.currentRuntimeValueDotNotation,
        getDragOverrides,
        propStatuses,
        nodePath: sequenceSubscriptionKey,
    });
    if (controlFields && controlFields.length > 0) {
        for (const f of controlFields) {
            roots.push({
                kind: 'field',
                nodePathInfo: {
                    sequenceSubscriptionKey,
                    auxiliaryKeys: [...auxiliaryKeys, 'controls', f.key],
                    index,
                    numberOfSequencesWithThisNodePath: 0,
                    supportsEffects,
                },
                label: (_a = f.description) !== null && _a !== void 0 ? _a : f.key,
                field: f,
            });
        }
    }
    if (sequence.effects.length > 0) {
        roots.push({
            kind: 'group',
            nodePathInfo: {
                sequenceSubscriptionKey,
                auxiliaryKeys: [...auxiliaryKeys, 'effects'],
                index,
                numberOfSequencesWithThisNodePath: 0,
                supportsEffects,
            },
            label: 'Effects',
            effectInfo: null,
            children: sequence.effects.map((effect, i) => {
                var _a;
                const effectFields = (0, studio_shared_1.getEffectFieldsToShow)({
                    effect,
                    effectIndex: i,
                    nodePath: sequenceSubscriptionKey,
                    propStatuses,
                    getEffectDragOverrides,
                });
                return {
                    kind: 'group',
                    nodePathInfo: {
                        sequenceSubscriptionKey,
                        auxiliaryKeys: [...auxiliaryKeys, 'effects', i.toString()],
                        index,
                        numberOfSequencesWithThisNodePath: 0,
                        supportsEffects,
                    },
                    label: effect.label,
                    effectInfo: {
                        effectIndex: i,
                        effectSchema: effect.schema,
                        documentationLink: (_a = effect.documentationLink) !== null && _a !== void 0 ? _a : null,
                    },
                    children: effectFields.map((f) => {
                        var _a;
                        return ({
                            kind: 'field',
                            nodePathInfo: {
                                sequenceSubscriptionKey,
                                auxiliaryKeys: [
                                    ...auxiliaryKeys,
                                    'effects',
                                    i.toString(),
                                    f.key,
                                ],
                                index,
                                numberOfSequencesWithThisNodePath: 0,
                                supportsEffects,
                            },
                            label: (_a = f.description) !== null && _a !== void 0 ? _a : f.key,
                            field: f,
                        });
                    }),
                };
            }),
        });
    }
    return roots;
};
exports.buildTimelineTree = buildTimelineTree;
const flattenVisibleTreeNodes = ({ nodes, getIsExpanded, depth = 0, }) => {
    const out = [];
    for (const node of nodes) {
        out.push({ node, depth });
        if (node.kind === 'group' && getIsExpanded(node.nodePathInfo)) {
            out.push(...(0, exports.flattenVisibleTreeNodes)({
                nodes: node.children,
                getIsExpanded,
                depth: depth + 1,
            }));
        }
    }
    return out;
};
exports.flattenVisibleTreeNodes = flattenVisibleTreeNodes;
const getTreeRowHeight = (node) => {
    if (node.kind === 'field' && node.field) {
        return node.field.rowHeight;
    }
    return exports.TREE_GROUP_ROW_HEIGHT;
};
exports.getTreeRowHeight = getTreeRowHeight;
const getExpandedTrackHeight = ({ sequence, nodePathInfo, getIsExpanded, propStatuses, }) => {
    const tree = (0, exports.buildTimelineTree)({
        sequence,
        nodePathInfo,
        // We assume that no drag overrides can change the timeline layout
        getDragOverrides: () => ({}),
        getEffectDragOverrides: () => ({}),
        propStatuses,
    });
    const flat = (0, exports.flattenVisibleTreeNodes)({ nodes: tree, getIsExpanded });
    if (flat.length === 0) {
        return exports.TIMELINE_TRACK_EXPANDED_HEIGHT;
    }
    const totalRowsHeight = flat.reduce((sum, { node }) => sum + (0, exports.getTreeRowHeight)(node), 0);
    const separators = Math.max(0, flat.length - 1);
    return totalRowsHeight + separators;
};
exports.getExpandedTrackHeight = getExpandedTrackHeight;
exports.TIMELINE_LAYER_FILMSTRIP_HEIGHT = 26;
// The waveform stripe rendered underneath the filmstrip in TimelineVideoInfo.
exports.TIMELINE_VIDEO_INFO_WAVEFORM_HEIGHT = 17;
exports.TIMELINE_LAYER_HEIGHT_VIDEO = 2 + exports.TIMELINE_LAYER_FILMSTRIP_HEIGHT + exports.TIMELINE_VIDEO_INFO_WAVEFORM_HEIGHT;
exports.TIMELINE_LAYER_HEIGHT_IMAGE = 26;
exports.TIMELINE_LAYER_HEIGHT_AUDIO = 34;
exports.TIMELINE_LAYER_HEIGHT_DEFAULT = 21;
// The horizontal row inside a timeline list item (eye + arrow + label).
exports.TIMELINE_LIST_ITEM_ROW_HEIGHT = 21;
const getTimelineLayerHeight = (type) => {
    if (type === 'video') {
        return exports.TIMELINE_LAYER_HEIGHT_VIDEO;
    }
    if (type === 'image') {
        return exports.TIMELINE_LAYER_HEIGHT_IMAGE;
    }
    if (type === 'audio') {
        return exports.TIMELINE_LAYER_HEIGHT_AUDIO;
    }
    return exports.TIMELINE_LAYER_HEIGHT_DEFAULT;
};
exports.getTimelineLayerHeight = getTimelineLayerHeight;
