// @remotion/studio/dist/helpers/use-media-metadata.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.useMediaMetadata = exports.getMediaMetadata = void 0;
const media_utils_1 = require("@remotion/media-utils");
const mediabunny_1 = require("mediabunny");
const react_1 = require("react");
const get_duration_or_compute_1 = require("./get-duration-or-compute");
const cache = new Map();
const pendingRequests = new Map();
const safeCall = async (fn) => {
    try {
        return await fn();
    }
    catch (_a) {
        return null;
    }
};
const getMediabunnyMetadata = async (src) => {
    var _a, _b;
    let input;
    try {
        input = new mediabunny_1.Input({
            formats: mediabunny_1.ALL_FORMATS,
            source: new mediabunny_1.UrlSource(src),
        });
    }
    catch (_c) {
        return null;
    }
    try {
        const [duration, format, videoTrack, audioTrack] = await Promise.all([
            safeCall(() => (0, get_duration_or_compute_1.getDurationOrCompute)(input)),
            safeCall(() => input.getFormat()),
            safeCall(() => input.getPrimaryVideoTrack()),
            safeCall(() => input.getPrimaryAudioTrack()),
        ]);
        if (duration === null) {
            return null;
        }
        const [width, height, videoCodec, packetStats, isHdr, audioCodec, sampleRate,] = await Promise.all([
            videoTrack ? safeCall(() => videoTrack.getDisplayWidth()) : null,
            videoTrack ? safeCall(() => videoTrack.getDisplayHeight()) : null,
            videoTrack ? safeCall(() => videoTrack.getCodec()) : null,
            videoTrack ? safeCall(() => videoTrack.computePacketStats(50)) : null,
            videoTrack ? safeCall(() => videoTrack.hasHighDynamicRange()) : null,
            audioTrack ? safeCall(() => audioTrack.getCodec()) : null,
            audioTrack ? safeCall(() => audioTrack.getSampleRate()) : null,
        ]);
        return {
            duration,
            format: (_a = format === null || format === void 0 ? void 0 : format.name) !== null && _a !== void 0 ? _a : null,
            width,
            height,
            videoCodec,
            audioCodec,
            fps: (_b = packetStats === null || packetStats === void 0 ? void 0 : packetStats.averagePacketRate) !== null && _b !== void 0 ? _b : null,
            isHdr,
            sampleRate,
            hasVideoTrack: Boolean(videoTrack),
            hasAudioTrack: Boolean(audioTrack),
        };
    }
    finally {
        try {
            input.dispose();
        }
        catch (_d) {
            // ignore
        }
    }
};
const getFallbackVideoMetadata = async (src) => {
    try {
        const metadata = await (0, media_utils_1.getVideoMetadata)(src);
        return {
            duration: metadata.durationInSeconds,
            format: null,
            width: metadata.width,
            height: metadata.height,
            videoCodec: null,
            audioCodec: null,
            fps: null,
            isHdr: null,
            sampleRate: null,
            hasVideoTrack: null,
            hasAudioTrack: null,
        };
    }
    catch (_a) {
        return null;
    }
};
const getMediaMetadata = (src) => {
    const cached = cache.get(src);
    if (cached) {
        return Promise.resolve(cached);
    }
    const pendingRequest = pendingRequests.get(src);
    if (pendingRequest) {
        return pendingRequest;
    }
    const request = getMediabunnyMetadata(src)
        .catch(() => null)
        .then((metadata) => metadata !== null && metadata !== void 0 ? metadata : getFallbackVideoMetadata(src))
        .then((metadata) => {
        if (metadata) {
            cache.set(src, metadata);
        }
        return metadata;
    })
        .finally(() => {
        pendingRequests.delete(src);
    });
    pendingRequests.set(src, request);
    return request;
};
exports.getMediaMetadata = getMediaMetadata;
const useMediaMetadata = (src) => {
    var _a;
    const [mediaMetadata, setMediaMetadata] = (0, react_1.useState)(src ? ((_a = cache.get(src)) !== null && _a !== void 0 ? _a : null) : null);
    (0, react_1.useEffect)(() => {
        var _a;
        const cached = src ? ((_a = cache.get(src)) !== null && _a !== void 0 ? _a : null) : null;
        setMediaMetadata(cached);
        if (!src || cached) {
            return;
        }
        let cancelled = false;
        (0, exports.getMediaMetadata)(src)
            .then((metadata) => {
            if (cancelled) {
                return;
            }
            setMediaMetadata(metadata);
        })
            .catch(() => {
            if (cancelled) {
                return;
            }
            setMediaMetadata(null);
        });
        return () => {
            cancelled = true;
        };
    }, [src]);
    return mediaMetadata;
};
exports.useMediaMetadata = useMediaMetadata;
