// @remotion/studio-shared/dist/asset-drag-data.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.parseAssetDragData = exports.makeAssetDragData = exports.ASSET_DRAG_MIME_TYPE = void 0;
exports.ASSET_DRAG_MIME_TYPE = 'application/vnd.remotion.asset+json';
const isRecord = (value) => {
    return typeof value === 'object' && value !== null && !Array.isArray(value);
};
const makeAssetDragData = (assetPath) => {
    return {
        type: 'remotion-asset',
        version: 1,
        assetPath,
    };
};
exports.makeAssetDragData = makeAssetDragData;
const parseAssetDragData = (value) => {
    try {
        const parsed = JSON.parse(value);
        if (!isRecord(parsed)) {
            return null;
        }
        if (parsed.type !== 'remotion-asset' || parsed.version !== 1) {
            return null;
        }
        if (typeof parsed.assetPath !== 'string' || parsed.assetPath.length === 0) {
            return null;
        }
        return {
            type: 'remotion-asset',
            version: 1,
            assetPath: parsed.assetPath,
        };
    }
    catch (_a) {
        return null;
    }
};
exports.parseAssetDragData = parseAssetDragData;
