// @remotion/studio-shared/dist/component-drag-data.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.parseComponentDragData = exports.makeComponentDragData = exports.areComponentProps = exports.isComponentProp = exports.isComponentPropName = exports.isComponentImportPath = exports.isComponentIdentifier = exports.COMPONENT_DRAG_MIME_TYPE = void 0;
exports.COMPONENT_DRAG_MIME_TYPE = 'application/vnd.remotion.component+json';
const isRecord = (value) => {
    return typeof value === 'object' && value !== null && !Array.isArray(value);
};
const isComponentIdentifier = (value) => {
    return typeof value === 'string' && /^[A-Z_$][A-Za-z0-9_$]*$/.test(value);
};
exports.isComponentIdentifier = isComponentIdentifier;
const isComponentImportPath = (value) => {
    return (typeof value === 'string' &&
        value.length > 0 &&
        value.length < 200 &&
        !value.includes('\\') &&
        !value.includes('\0') &&
        !value.startsWith('/') &&
        /^[A-Za-z0-9@._/-]+$/.test(value));
};
exports.isComponentImportPath = isComponentImportPath;
const isComponentPropName = (value) => {
    return (typeof value === 'string' &&
        value !== 'style' &&
        /^[A-Za-z_$][A-Za-z0-9_$]*$/.test(value));
};
exports.isComponentPropName = isComponentPropName;
const isComponentProp = (value) => {
    if (!isRecord(value)) {
        return false;
    }
    if (!(0, exports.isComponentPropName)(value.name)) {
        return false;
    }
    return (typeof value.value === 'string' ||
        typeof value.value === 'boolean' ||
        (typeof value.value === 'number' && Number.isFinite(value.value)));
};
exports.isComponentProp = isComponentProp;
const areComponentProps = (value) => {
    if (!Array.isArray(value)) {
        return false;
    }
    const seen = new Set();
    for (const prop of value) {
        if (!(0, exports.isComponentProp)(prop) || seen.has(prop.name)) {
            return false;
        }
        seen.add(prop.name);
    }
    return true;
};
exports.areComponentProps = areComponentProps;
const isComponentDimensions = (value) => {
    if (!isRecord(value)) {
        return false;
    }
    return (typeof value.width === 'number' &&
        Number.isFinite(value.width) &&
        value.width >= 0 &&
        typeof value.height === 'number' &&
        Number.isFinite(value.height) &&
        value.height >= 0);
};
const makeComponentDragData = ({ componentName, dimensions, importName, importPath, props, }) => {
    return {
        type: 'remotion-component',
        version: 1,
        component: {
            componentName,
            ...(dimensions ? { dimensions } : {}),
            importName,
            importPath,
            props,
        },
    };
};
exports.makeComponentDragData = makeComponentDragData;
const parseComponentDragData = (value) => {
    try {
        const parsed = JSON.parse(value);
        if (!isRecord(parsed)) {
            return null;
        }
        if (parsed.type !== 'remotion-component' || parsed.version !== 1) {
            return null;
        }
        if (!isRecord(parsed.component)) {
            return null;
        }
        const { componentName, dimensions, importName, importPath, props } = parsed.component;
        if (!(0, exports.isComponentIdentifier)(componentName) ||
            !(0, exports.isComponentIdentifier)(importName) ||
            !(0, exports.isComponentImportPath)(importPath) ||
            !(0, exports.areComponentProps)(props) ||
            (typeof dimensions !== 'undefined' && !isComponentDimensions(dimensions))) {
            return null;
        }
        return {
            type: 'remotion-component',
            version: 1,
            component: {
                componentName,
                ...(dimensions ? { dimensions } : {}),
                importName,
                importPath,
                props,
            },
        };
    }
    catch (_a) {
        return null;
    }
};
exports.parseComponentDragData = parseComponentDragData;
