// @remotion/studio-shared/dist/detect-file-type.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.detectFileType = exports.isImageFileType = exports.isM3u = exports.isFlac = exports.isAac = exports.isMp3 = exports.isTransportStream = exports.isIsoBaseMedia = exports.isWebm = exports.isRiffWave = exports.isRiffAvi = exports.matchesPattern = void 0;
const webmPattern = new Uint8Array([0x1a, 0x45, 0xdf, 0xa3]);
const matchesPattern = (pattern) => {
    return (data) => {
        return pattern.every((value, index) => data[index] === value);
    };
};
exports.matchesPattern = matchesPattern;
const isRiffAvi = (data) => {
    const riffPattern = new Uint8Array([0x52, 0x49, 0x46, 0x46]);
    if (!(0, exports.matchesPattern)(riffPattern)(data.subarray(0, 4))) {
        return false;
    }
    const fileType = data.subarray(8, 12);
    const aviPattern = new Uint8Array([0x41, 0x56, 0x49, 0x20]);
    return (0, exports.matchesPattern)(aviPattern)(fileType);
};
exports.isRiffAvi = isRiffAvi;
const isRiffWave = (data) => {
    const riffPattern = new Uint8Array([0x52, 0x49, 0x46, 0x46]);
    if (!(0, exports.matchesPattern)(riffPattern)(data.subarray(0, 4))) {
        return false;
    }
    const fileType = data.subarray(8, 12);
    const wavePattern = new Uint8Array([0x57, 0x41, 0x56, 0x45]);
    return (0, exports.matchesPattern)(wavePattern)(fileType);
};
exports.isRiffWave = isRiffWave;
const isWebm = (data) => {
    return (0, exports.matchesPattern)(webmPattern)(data.subarray(0, 4));
};
exports.isWebm = isWebm;
const isIsoBaseMedia = (data) => {
    const isoBaseMediaMp4Pattern = new TextEncoder().encode('ftyp');
    return (0, exports.matchesPattern)(isoBaseMediaMp4Pattern)(data.subarray(4, 8));
};
exports.isIsoBaseMedia = isIsoBaseMedia;
const isTransportStream = (data) => {
    return data[0] === 0x47 && data[188] === 0x47;
};
exports.isTransportStream = isTransportStream;
const isMp3 = (data) => {
    const mpegPattern = new Uint8Array([0xff, 0xf3]);
    const mpegPattern2 = new Uint8Array([0xff, 0xfb]);
    const id3v4Pattern = new Uint8Array([0x49, 0x44, 0x33, 4]);
    const id3v3Pattern = new Uint8Array([0x49, 0x44, 0x33, 3]);
    const id3v2Pattern = new Uint8Array([0x49, 0x44, 0x33, 2]);
    const subarray = data.subarray(0, 4);
    return ((0, exports.matchesPattern)(mpegPattern)(subarray) ||
        (0, exports.matchesPattern)(mpegPattern2)(subarray) ||
        (0, exports.matchesPattern)(id3v4Pattern)(subarray) ||
        (0, exports.matchesPattern)(id3v3Pattern)(subarray) ||
        (0, exports.matchesPattern)(id3v2Pattern)(subarray));
};
exports.isMp3 = isMp3;
const isAac = (data) => {
    const aacPattern = new Uint8Array([0xff, 0xf1]);
    return (0, exports.matchesPattern)(aacPattern)(data.subarray(0, 2));
};
exports.isAac = isAac;
const isFlac = (data) => {
    const flacPattern = new Uint8Array([0x66, 0x4c, 0x61, 0x43]);
    return (0, exports.matchesPattern)(flacPattern)(data.subarray(0, 4));
};
exports.isFlac = isFlac;
const isM3u = (data) => {
    return new TextDecoder('utf-8').decode(data.slice(0, 7)) === '#EXTM3U';
};
exports.isM3u = isM3u;
const pngSignature = [137, 80, 78, 71, 13, 10, 26, 10];
const acTLChunkType = new Uint8Array([0x61, 0x63, 0x54, 0x4c]);
const idatChunkType = new Uint8Array([0x49, 0x44, 0x41, 0x54]);
const iendChunkType = new Uint8Array([0x49, 0x45, 0x4e, 0x44]);
const getPngDimensions = (pngData) => {
    if (pngData.length < 24) {
        return null;
    }
    const view = new DataView(pngData.buffer, pngData.byteOffset);
    for (let i = 0; i < 8; i++) {
        if (pngData[i] !== pngSignature[i]) {
            return null;
        }
    }
    return {
        width: view.getUint32(16, false),
        height: view.getUint32(20, false),
    };
};
const hasApngAnimationControlChunk = (pngData) => {
    if (pngData.length < 16) {
        return false;
    }
    const view = new DataView(pngData.buffer, pngData.byteOffset, pngData.byteLength);
    let offset = 8;
    while (offset + 8 <= pngData.length) {
        const chunkLength = view.getUint32(offset, false);
        const chunkType = pngData.subarray(offset + 4, offset + 8);
        if ((0, exports.matchesPattern)(acTLChunkType)(chunkType)) {
            return true;
        }
        if ((0, exports.matchesPattern)(idatChunkType)(chunkType) ||
            (0, exports.matchesPattern)(iendChunkType)(chunkType)) {
            return false;
        }
        const nextOffset = offset + 12 + chunkLength;
        if (nextOffset <= offset) {
            return false;
        }
        offset = nextOffset;
    }
    return false;
};
const isPng = (data) => {
    const pngPattern = new Uint8Array([0x89, 0x50, 0x4e, 0x47]);
    if ((0, exports.matchesPattern)(pngPattern)(data.subarray(0, 4))) {
        const png = getPngDimensions(data);
        return {
            dimensions: png,
            type: hasApngAnimationControlChunk(data) ? 'apng' : 'png',
        };
    }
    return null;
};
const getJpegDimensions = (data) => {
    let offset = 0;
    const readUint16BE = (o) => {
        return (data[o] << 8) | data[o + 1];
    };
    if (data.length < 4 || readUint16BE(offset) !== 0xffd8) {
        return null;
    }
    offset += 2;
    while (offset + 3 < data.length) {
        if (data[offset] === 0xff) {
            const marker = data[offset + 1];
            if (marker === 0xc0 || marker === 0xc2) {
                if (offset + 8 >= data.length) {
                    return null;
                }
                const height = readUint16BE(offset + 5);
                const width = readUint16BE(offset + 7);
                return { width, height };
            }
            const length = readUint16BE(offset + 2);
            if (length <= 0) {
                return null;
            }
            offset += length + 2;
        }
        else {
            offset++;
        }
    }
    return null;
};
const isJpeg = (data) => {
    const jpegPattern = new Uint8Array([0xff, 0xd8]);
    const jpeg = (0, exports.matchesPattern)(jpegPattern)(data.subarray(0, 2));
    if (!jpeg) {
        return null;
    }
    const dim = getJpegDimensions(data);
    return { dimensions: dim, type: 'jpeg' };
};
const getGifDimensions = (data) => {
    if (data.length < 10) {
        return null;
    }
    const view = new DataView(data.buffer, data.byteOffset);
    const width = view.getUint16(6, true);
    const height = view.getUint16(8, true);
    return { width, height };
};
const isGif = (data) => {
    const gifPattern = new Uint8Array([0x47, 0x49, 0x46, 0x38]);
    if ((0, exports.matchesPattern)(gifPattern)(data.subarray(0, 4))) {
        return { type: 'gif', dimensions: getGifDimensions(data) };
    }
    return null;
};
const getWebPDimensions = (bytes) => {
    if (bytes.length < 30) {
        return null;
    }
    if (bytes[0] !== 0x52 ||
        bytes[1] !== 0x49 ||
        bytes[2] !== 0x46 ||
        bytes[3] !== 0x46 ||
        bytes[8] !== 0x57 ||
        bytes[9] !== 0x45 ||
        bytes[10] !== 0x42 ||
        bytes[11] !== 0x50) {
        return null;
    }
    if (bytes[12] === 0x56 && bytes[13] === 0x50 && bytes[14] === 0x38) {
        if (bytes[15] === 0x20) {
            return {
                width: bytes[26] | ((bytes[27] << 8) & 0x3fff),
                height: bytes[28] | ((bytes[29] << 8) & 0x3fff),
            };
        }
    }
    if (bytes[12] === 0x56 &&
        bytes[13] === 0x50 &&
        bytes[14] === 0x38 &&
        bytes[15] === 0x4c) {
        return {
            width: 1 + (bytes[21] | ((bytes[22] & 0x3f) << 8)),
            height: 1 +
                (((bytes[22] & 0xc0) >> 6) |
                    (bytes[23] << 2) |
                    ((bytes[24] & 0x0f) << 10)),
        };
    }
    if (bytes[12] === 0x56 &&
        bytes[13] === 0x50 &&
        bytes[14] === 0x38 &&
        bytes[15] === 0x58) {
        return {
            width: 1 + (bytes[24] | (bytes[25] << 8) | (bytes[26] << 16)),
            height: 1 + (bytes[27] | (bytes[28] << 8) | (bytes[29] << 16)),
        };
    }
    return null;
};
const isWebp = (data) => {
    const webpPattern = new Uint8Array([0x52, 0x49, 0x46, 0x46]);
    if ((0, exports.matchesPattern)(webpPattern)(data.subarray(0, 4))) {
        return {
            type: 'webp',
            dimensions: getWebPDimensions(data),
        };
    }
    return null;
};
const getBmpDimensions = (bmpData) => {
    if (bmpData.length < 26) {
        return null;
    }
    const view = new DataView(bmpData.buffer, bmpData.byteOffset);
    return {
        width: view.getUint32(18, true),
        height: Math.abs(view.getInt32(22, true)),
    };
};
const isBmp = (data) => {
    const bmpPattern = new Uint8Array([0x42, 0x4d]);
    if ((0, exports.matchesPattern)(bmpPattern)(data.subarray(0, 2))) {
        const bmp = getBmpDimensions(data);
        return { dimensions: bmp, type: 'bmp' };
    }
    return null;
};
const isPdf = (data) => {
    if (data.length < 4) {
        return null;
    }
    const pdfPattern = new Uint8Array([0x25, 0x50, 0x44, 0x46]);
    return (0, exports.matchesPattern)(pdfPattern)(data.subarray(0, 4)) ? { type: 'pdf' } : null;
};
const isImageFileType = (fileType) => {
    return (fileType.type === 'jpeg' ||
        fileType.type === 'webp' ||
        fileType.type === 'gif' ||
        fileType.type === 'png' ||
        fileType.type === 'apng' ||
        fileType.type === 'bmp');
};
exports.isImageFileType = isImageFileType;
const detectFileType = (data) => {
    if ((0, exports.isRiffWave)(data)) {
        return { type: 'wav' };
    }
    if ((0, exports.isRiffAvi)(data)) {
        return { type: 'riff' };
    }
    if ((0, exports.isAac)(data)) {
        return { type: 'aac' };
    }
    if ((0, exports.isFlac)(data)) {
        return { type: 'flac' };
    }
    if ((0, exports.isM3u)(data)) {
        return { type: 'm3u' };
    }
    const webp = isWebp(data);
    if (webp) {
        return webp;
    }
    if ((0, exports.isWebm)(data)) {
        return { type: 'webm' };
    }
    if ((0, exports.isIsoBaseMedia)(data)) {
        return { type: 'iso-base-media' };
    }
    if ((0, exports.isTransportStream)(data)) {
        return { type: 'transport-stream' };
    }
    if ((0, exports.isMp3)(data)) {
        return { type: 'mp3' };
    }
    const gif = isGif(data);
    if (gif) {
        return gif;
    }
    const png = isPng(data);
    if (png) {
        return png;
    }
    const pdf = isPdf(data);
    if (pdf) {
        return pdf;
    }
    const bmp = isBmp(data);
    if (bmp) {
        return bmp;
    }
    const jpeg = isJpeg(data);
    if (jpeg) {
        return jpeg;
    }
    return { type: 'unknown' };
};
exports.detectFileType = detectFileType;
