// @remotion/studio-shared/dist/easing-clipboard-data.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.parseEasingClipboardData = exports.parseEasingClipboardDataResult = void 0;
const isRecord = (value) => {
    return typeof value === 'object' && value !== null && !Array.isArray(value);
};
const isFiniteNumber = (value) => {
    return typeof value === 'number' && Number.isFinite(value);
};
const isEasing = (value) => {
    return (isRecord(value) &&
        (value.type === 'linear' ||
            (value.type === 'bezier' &&
                isFiniteNumber(value.x1) &&
                isFiniteNumber(value.y1) &&
                isFiniteNumber(value.x2) &&
                isFiniteNumber(value.y2)) ||
            (value.type === 'spring' &&
                isFiniteNumber(value.damping) &&
                isFiniteNumber(value.mass) &&
                isFiniteNumber(value.stiffness) &&
                typeof value.overshootClamping === 'boolean')));
};
const parseEasingClipboardDataResult = (value) => {
    try {
        const parsed = JSON.parse(value);
        if (!isRecord(parsed)) {
            return { status: 'invalid' };
        }
        if (parsed.remotionClipboard !== 'easing') {
            return { status: 'invalid' };
        }
        if (parsed.version !== 1) {
            return {
                status: 'unsupported-version',
                version: parsed.version,
            };
        }
        if (parsed.type !== 'easing' || !isEasing(parsed.easing)) {
            return { status: 'invalid' };
        }
        return {
            status: 'valid',
            data: {
                type: 'easing',
                version: 1,
                remotionClipboard: 'easing',
                easing: parsed.easing,
            },
        };
    }
    catch (_a) {
        return { status: 'invalid' };
    }
};
exports.parseEasingClipboardDataResult = parseEasingClipboardDataResult;
const parseEasingClipboardData = (value) => {
    const result = (0, exports.parseEasingClipboardDataResult)(value);
    if (result.status !== 'valid') {
        return null;
    }
    return result.data;
};
exports.parseEasingClipboardData = parseEasingClipboardData;
