// @remotion/studio-shared/dist/effect-clipboard-data.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.parseEffectPropClipboardData = exports.parseEffectPropClipboardDataResult = exports.parseEffectClipboardData = exports.parseEffectClipboardDataResult = void 0;
const keyframe_interpolation_function_1 = require("./keyframe-interpolation-function");
const isRecord = (value) => {
    return typeof value === 'object' && value !== null && !Array.isArray(value);
};
const extrapolateTypes = new Set(['extend', 'identity', 'clamp', 'wrap']);
const isFiniteNumber = (value) => {
    return typeof value === 'number' && Number.isFinite(value);
};
const isEasing = (value) => {
    return (isRecord(value) &&
        (value.type === 'linear' ||
            (value.type === 'bezier' &&
                isFiniteNumber(value.x1) &&
                isFiniteNumber(value.y1) &&
                isFiniteNumber(value.x2) &&
                isFiniteNumber(value.y2)) ||
            (value.type === 'spring' &&
                isFiniteNumber(value.damping) &&
                isFiniteNumber(value.mass) &&
                isFiniteNumber(value.stiffness) &&
                (value.allowTail === undefined ||
                    value.allowTail === null ||
                    typeof value.allowTail === 'boolean') &&
                (value.durationRestThreshold === undefined ||
                    value.durationRestThreshold === null ||
                    isFiniteNumber(value.durationRestThreshold)) &&
                typeof value.overshootClamping === 'boolean')));
};
const normalizeEasing = (easing) => {
    var _a, _b;
    if (easing.type !== 'spring') {
        return easing;
    }
    return {
        ...easing,
        allowTail: (_a = easing.allowTail) !== null && _a !== void 0 ? _a : null,
        durationRestThreshold: (_b = easing.durationRestThreshold) !== null && _b !== void 0 ? _b : null,
    };
};
const normalizeParam = (param) => {
    if (param.type === 'static') {
        return param;
    }
    return {
        ...param,
        easing: param.easing.map(normalizeEasing),
    };
};
const normalizeSnapshot = (snapshot) => {
    return {
        ...snapshot,
        params: Object.fromEntries(Object.entries(snapshot.params).map(([key, param]) => [
            key,
            normalizeParam(param),
        ])),
    };
};
const isKeyframe = (value) => {
    return isRecord(value) && isFiniteNumber(value.frame) && 'value' in value;
};
const isClamping = (value) => {
    return (isRecord(value) &&
        typeof value.left === 'string' &&
        extrapolateTypes.has(value.left) &&
        typeof value.right === 'string' &&
        extrapolateTypes.has(value.right));
};
const isEffectClipboardParam = (value) => {
    if (!isRecord(value)) {
        return false;
    }
    if (value.type === 'static') {
        return 'value' in value;
    }
    if (value.type !== 'keyframed') {
        return false;
    }
    const { posterize } = value;
    const easingLength = Array.isArray(value.keyframes) && value.keyframes.length > 0
        ? value.keyframes.length - 1
        : null;
    return (typeof value.interpolationFunction === 'string' &&
        (0, keyframe_interpolation_function_1.isKeyframeInterpolationFunction)(value.interpolationFunction) &&
        Array.isArray(value.keyframes) &&
        value.keyframes.length > 0 &&
        value.keyframes.every(isKeyframe) &&
        Array.isArray(value.easing) &&
        value.easing.length === easingLength &&
        value.easing.every(isEasing) &&
        isClamping(value.clamping) &&
        (posterize === undefined || (isFiniteNumber(posterize) && posterize > 0)));
};
const isEffectClipboardSnapshotV3 = (value) => {
    if (!isRecord(value)) {
        return false;
    }
    return (typeof value.callee === 'string' &&
        typeof value.importPath === 'string' &&
        isRecord(value.params) &&
        Object.values(value.params).every(isEffectClipboardParam));
};
const parseEffectClipboardDataResult = (value) => {
    try {
        const parsed = JSON.parse(value);
        if (!isRecord(parsed)) {
            return { status: 'invalid' };
        }
        if (parsed.remotionClipboard !== 'effects') {
            return { status: 'invalid' };
        }
        if (parsed.version !== 3) {
            return {
                status: 'unsupported-version',
                version: parsed.version,
            };
        }
        if (parsed.type !== 'effects-additive' &&
            parsed.type !== 'effects-replacing') {
            return { status: 'invalid' };
        }
        if (!Array.isArray(parsed.effects)) {
            return { status: 'invalid' };
        }
        const effects = [];
        for (const effect of parsed.effects) {
            if (!isEffectClipboardSnapshotV3(effect)) {
                return { status: 'invalid' };
            }
            effects.push(normalizeSnapshot(effect));
        }
        return {
            status: 'valid',
            data: {
                type: parsed.type,
                version: 3,
                remotionClipboard: 'effects',
                effects,
            },
        };
    }
    catch (_a) {
        return { status: 'invalid' };
    }
};
exports.parseEffectClipboardDataResult = parseEffectClipboardDataResult;
const parseEffectClipboardData = (value) => {
    const result = (0, exports.parseEffectClipboardDataResult)(value);
    if (result.status !== 'valid') {
        return null;
    }
    return result.data;
};
exports.parseEffectClipboardData = parseEffectClipboardData;
const parseEffectPropClipboardDataResult = (value) => {
    try {
        const parsed = JSON.parse(value);
        if (!isRecord(parsed)) {
            return { status: 'invalid' };
        }
        if (parsed.remotionClipboard !== 'effect-prop') {
            return { status: 'invalid' };
        }
        if (parsed.version !== 1) {
            return {
                status: 'unsupported-version',
                version: parsed.version,
            };
        }
        if (parsed.type !== 'effect-prop') {
            return { status: 'invalid' };
        }
        if (!isRecord(parsed.effect)) {
            return { status: 'invalid' };
        }
        if (typeof parsed.effect.callee !== 'string' ||
            typeof parsed.effect.importPath !== 'string') {
            return { status: 'invalid' };
        }
        if (typeof parsed.key !== 'string') {
            return { status: 'invalid' };
        }
        if (!isEffectClipboardParam(parsed.param)) {
            return { status: 'invalid' };
        }
        return {
            status: 'valid',
            data: {
                type: 'effect-prop',
                version: 1,
                remotionClipboard: 'effect-prop',
                effect: {
                    callee: parsed.effect.callee,
                    importPath: parsed.effect.importPath,
                },
                key: parsed.key,
                param: normalizeParam(parsed.param),
            },
        };
    }
    catch (_a) {
        return { status: 'invalid' };
    }
};
exports.parseEffectPropClipboardDataResult = parseEffectPropClipboardDataResult;
const parseEffectPropClipboardData = (value) => {
    const result = (0, exports.parseEffectPropClipboardDataResult)(value);
    if (result.status !== 'valid') {
        return null;
    }
    return result.data;
};
exports.parseEffectPropClipboardData = parseEffectPropClipboardData;
