// @remotion/studio-shared/dist/keyframe-easing-presets.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.KEYFRAME_EASING_PRESETS = exports.LINEAR_KEYFRAME_EASING = void 0;
exports.LINEAR_KEYFRAME_EASING = { type: 'linear' };
exports.KEYFRAME_EASING_PRESETS = [
    {
        id: 'ease-in',
        label: 'Ease in',
        easing: { type: 'bezier', x1: 0.42, y1: 0, x2: 1, y2: 1 },
    },
    {
        id: 'ease-out',
        label: 'Ease out',
        easing: { type: 'bezier', x1: 0, y1: 0, x2: 0.58, y2: 1 },
    },
    {
        id: 'ease-in-out',
        label: 'Ease in-out',
        easing: { type: 'bezier', x1: 0.42, y1: 0, x2: 0.58, y2: 1 },
    },
    {
        id: 'tail-spring',
        label: 'Tail spring',
        easing: {
            type: 'spring',
            allowTail: true,
            damping: 200,
            durationRestThreshold: 0.02,
            mass: 1,
            overshootClamping: false,
            stiffness: 100,
        },
    },
    {
        id: 'spring',
        label: 'Spring',
        easing: {
            type: 'spring',
            allowTail: true,
            damping: 10,
            durationRestThreshold: 0.02,
            mass: 1,
            overshootClamping: false,
            stiffness: 100,
        },
    },
    {
        id: 'bouncy-spring',
        label: 'Bouncy spring',
        easing: {
            type: 'spring',
            allowTail: true,
            damping: 5,
            durationRestThreshold: 0.02,
            mass: 1,
            overshootClamping: false,
            stiffness: 120,
        },
    },
];
