// @remotion/studio-shared/dist/keyframe-interpolation-function.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getKeyframeInterpolationFunction = exports.getKeyframeInterpolationFunctionForSchemaField = exports.isSchemaFieldKeyframable = exports.isInteractivitySchemaFieldKeyframable = exports.isKeyframeInterpolationFunction = exports.keyframeInterpolationFunctions = void 0;
exports.keyframeInterpolationFunctions = [
    'interpolate',
    'interpolateColors',
];
const isKeyframeInterpolationFunction = (name) => {
    return exports.keyframeInterpolationFunctions.includes(name);
};
exports.isKeyframeInterpolationFunction = isKeyframeInterpolationFunction;
const isInteractivitySchemaFieldKeyframable = (field) => {
    if (!field) {
        return true;
    }
    if (field.type === 'array' ||
        field.type === 'boolean' ||
        field.type === 'enum') {
        return false;
    }
    return field.keyframable !== false;
};
exports.isInteractivitySchemaFieldKeyframable = isInteractivitySchemaFieldKeyframable;
const findFieldInSchema = (schema, key) => {
    if (key in schema) {
        return schema[key];
    }
    for (const field of Object.values(schema)) {
        if (field.type !== 'enum') {
            continue;
        }
        for (const variant of Object.values(field.variants)) {
            const found = findFieldInSchema(variant, key);
            if (found) {
                return found;
            }
        }
    }
    return undefined;
};
const isSchemaFieldKeyframable = ({ schema, key, }) => {
    const field = schema ? findFieldInSchema(schema, key) : undefined;
    return (0, exports.isInteractivitySchemaFieldKeyframable)(field);
};
exports.isSchemaFieldKeyframable = isSchemaFieldKeyframable;
const getKeyframeInterpolationFunctionForSchemaField = ({ schema, key, }) => {
    const field = schema ? findFieldInSchema(schema, key) : undefined;
    if ((field === null || field === void 0 ? void 0 : field.type) === 'color') {
        return 'interpolateColors';
    }
    if ((field === null || field === void 0 ? void 0 : field.type) === 'scale' ||
        (field === null || field === void 0 ? void 0 : field.type) === 'translate' ||
        (field === null || field === void 0 ? void 0 : field.type) === 'transform-origin' ||
        (field === null || field === void 0 ? void 0 : field.type) === 'rotation-css') {
        return 'interpolate';
    }
    return null;
};
exports.getKeyframeInterpolationFunctionForSchemaField = getKeyframeInterpolationFunctionForSchemaField;
const getKeyframeInterpolationFunction = ({ schema, key, staticValue, newValue, }) => {
    const schemaFunction = (0, exports.getKeyframeInterpolationFunctionForSchemaField)({
        schema,
        key,
    });
    if (schemaFunction) {
        return schemaFunction;
    }
    return typeof staticValue === 'string' && typeof newValue === 'string'
        ? 'interpolateColors'
        : 'interpolate';
};
exports.getKeyframeInterpolationFunction = getKeyframeInterpolationFunction;
