// @remotion/studio-shared/dist/optimistic-move-keyframe.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.optimisticMoveEffectKeyframes = exports.optimisticMoveSequenceKeyframes = exports.moveKeyframesInPropStatus = exports.canMoveKeyframesWithoutCollisions = void 0;
const getMoveMap = (moves) => {
    const moveMap = new Map();
    for (const move of moves) {
        if (move.fromFrame === move.toFrame) {
            continue;
        }
        if (moveMap.has(move.fromFrame)) {
            return null;
        }
        moveMap.set(move.fromFrame, move.toFrame);
    }
    return moveMap;
};
const getMovedKeyframes = ({ status, moves, }) => {
    if (status.status !== 'keyframed') {
        return null;
    }
    const moveMap = getMoveMap(moves);
    if (moveMap === null) {
        return null;
    }
    if (moveMap.size === 0) {
        return { keyframes: status.keyframes, removedKeyframeIndexes: [] };
    }
    const frames = new Set(status.keyframes.map((keyframe) => keyframe.frame));
    for (const fromFrame of moveMap.keys()) {
        if (!frames.has(fromFrame)) {
            return null;
        }
    }
    const movedFromFrames = new Set(moveMap.keys());
    const movedToFrames = new Set(moveMap.values());
    const removedKeyframeIndexes = [];
    const nextKeyframes = status.keyframes.flatMap((keyframe, index) => {
        const movedFrame = moveMap.get(keyframe.frame);
        if (movedFrame !== undefined) {
            return [{ ...keyframe, frame: movedFrame }];
        }
        if (movedToFrames.has(keyframe.frame) &&
            !movedFromFrames.has(keyframe.frame)) {
            removedKeyframeIndexes.push(index);
            return [];
        }
        return [keyframe];
    });
    const nextFrames = new Set();
    for (const keyframe of nextKeyframes) {
        if (nextFrames.has(keyframe.frame)) {
            return null;
        }
        nextFrames.add(keyframe.frame);
    }
    return {
        keyframes: nextKeyframes.sort((a, b) => a.frame - b.frame),
        removedKeyframeIndexes,
    };
};
const removeEasingForRemovedKeyframes = ({ easing, removedKeyframeIndexes, }) => {
    const nextEasing = [...easing];
    for (const removedKeyframeIndex of [...removedKeyframeIndexes].sort((a, b) => b - a)) {
        if (nextEasing.length === 0) {
            break;
        }
        const easingIndexToRemove = removedKeyframeIndex === 0 ? 0 : removedKeyframeIndex - 1;
        nextEasing.splice(easingIndexToRemove, 1);
    }
    return nextEasing;
};
const canMoveKeyframesWithoutCollisions = ({ status, moves, }) => {
    return getMovedKeyframes({ status, moves }) !== null;
};
exports.canMoveKeyframesWithoutCollisions = canMoveKeyframesWithoutCollisions;
const moveKeyframesInPropStatus = ({ status, moves, }) => {
    if (status.status !== 'keyframed') {
        return status;
    }
    const moved = getMovedKeyframes({ status, moves });
    if (moved === null ||
        (moved.removedKeyframeIndexes.length === 0 &&
            moved.keyframes === status.keyframes)) {
        return status;
    }
    const easing = removeEasingForRemovedKeyframes({
        easing: status.easing,
        removedKeyframeIndexes: moved.removedKeyframeIndexes,
    });
    return {
        ...status,
        keyframes: moved.keyframes,
        easing,
    };
};
exports.moveKeyframesInPropStatus = moveKeyframesInPropStatus;
const optimisticMoveSequenceKeyframes = ({ previous, keyframes, }) => {
    var _a;
    if (!previous.canUpdate) {
        return previous;
    }
    const movesByField = new Map();
    for (const keyframe of keyframes) {
        const moves = (_a = movesByField.get(keyframe.fieldKey)) !== null && _a !== void 0 ? _a : [];
        moves.push(keyframe);
        movesByField.set(keyframe.fieldKey, moves);
    }
    const props = { ...previous.props };
    for (const [fieldKey, moves] of movesByField) {
        const status = props[fieldKey];
        if (!status) {
            continue;
        }
        props[fieldKey] = (0, exports.moveKeyframesInPropStatus)({ status, moves });
    }
    return {
        ...previous,
        props,
    };
};
exports.optimisticMoveSequenceKeyframes = optimisticMoveSequenceKeyframes;
const optimisticMoveEffectKeyframes = ({ previous, keyframes, }) => {
    var _a;
    if (!previous.canUpdate) {
        return previous;
    }
    const movesByEffect = new Map();
    for (const keyframe of keyframes) {
        const moves = (_a = movesByEffect.get(keyframe.effectIndex)) !== null && _a !== void 0 ? _a : [];
        moves.push(keyframe);
        movesByEffect.set(keyframe.effectIndex, moves);
    }
    const effects = previous.effects.map((effect) => {
        var _a;
        if (!effect.canUpdate) {
            return effect;
        }
        const movesForEffect = movesByEffect.get(effect.effectIndex);
        if (!movesForEffect) {
            return effect;
        }
        const props = { ...effect.props };
        const movesByField = new Map();
        for (const move of movesForEffect) {
            const moves = (_a = movesByField.get(move.fieldKey)) !== null && _a !== void 0 ? _a : [];
            moves.push(move);
            movesByField.set(move.fieldKey, moves);
        }
        for (const [fieldKey, moves] of movesByField) {
            const status = props[fieldKey];
            if (!status) {
                continue;
            }
            props[fieldKey] = (0, exports.moveKeyframesInPropStatus)({ status, moves });
        }
        return {
            ...effect,
            props,
        };
    });
    return {
        ...previous,
        effects,
    };
};
exports.optimisticMoveEffectKeyframes = optimisticMoveEffectKeyframes;
