// @remotion/studio-shared/dist/optimistic-update-keyframe-settings.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.optimisticUpdateEffectKeyframeSettings = exports.optimisticUpdateSequenceKeyframeSettings = void 0;
const keyframe_easing_presets_1 = require("./keyframe-easing-presets");
const updateEasing = ({ easing, segmentCount, segmentIndex, value, }) => {
    if (!Number.isInteger(segmentIndex) ||
        segmentIndex < 0 ||
        segmentIndex >= segmentCount) {
        throw new Error('Cannot update easing: segment index out of range');
    }
    const nextEasing = Array.from({ length: segmentCount }, (_, index) => {
        var _a;
        return (_a = easing[index]) !== null && _a !== void 0 ? _a : keyframe_easing_presets_1.LINEAR_KEYFRAME_EASING;
    });
    nextEasing[segmentIndex] = value;
    return nextEasing;
};
const applySettingsToStatus = (status, settings) => {
    if (!status || status.status !== 'keyframed') {
        throw new Error('Expected keyframed status');
    }
    return {
        ...status,
        ...(settings.type === 'settings' && settings.clamping
            ? { clamping: settings.clamping }
            : {}),
        ...(settings.type === 'settings' ? { posterize: settings.posterize } : {}),
        ...(settings.type === 'easing'
            ? {
                easing: updateEasing({
                    easing: status.easing,
                    segmentCount: Math.max(0, status.keyframes.length - 1),
                    segmentIndex: settings.segmentIndex,
                    value: settings.easing,
                }),
            }
            : {}),
    };
};
const optimisticUpdateSequenceKeyframeSettings = ({ previous, fieldKey, settings, }) => {
    if (!previous.canUpdate) {
        return previous;
    }
    const status = previous.props[fieldKey];
    if (!status || status.status !== 'keyframed') {
        return previous;
    }
    return {
        ...previous,
        props: {
            ...previous.props,
            [fieldKey]: applySettingsToStatus(status, settings),
        },
    };
};
exports.optimisticUpdateSequenceKeyframeSettings = optimisticUpdateSequenceKeyframeSettings;
const optimisticUpdateEffectKeyframeSettings = ({ previous, effectIndex, fieldKey, settings, }) => {
    if (!previous.canUpdate) {
        return previous;
    }
    const targetIndex = previous.effects.findIndex((effect) => effect.effectIndex === effectIndex);
    if (targetIndex === -1) {
        return previous;
    }
    const target = previous.effects[targetIndex];
    if (!target.canUpdate) {
        return previous;
    }
    const status = target.props[fieldKey];
    if (!status || status.status !== 'keyframed') {
        return previous;
    }
    const effects = [...previous.effects];
    effects[targetIndex] = {
        ...target,
        props: {
            ...target.props,
            [fieldKey]: applySettingsToStatus(status, settings),
        },
    };
    return {
        ...previous,
        effects,
    };
};
exports.optimisticUpdateEffectKeyframeSettings = optimisticUpdateEffectKeyframeSettings;
