// @remotion/studio-shared/dist/optimistic-delete-keyframe.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.optimisticDeleteEffectKeyframes = exports.optimisticDeleteEffectKeyframe = exports.optimisticDeleteSequenceKeyframes = exports.optimisticDeleteSequenceKeyframe = void 0;
const removeKeyframeFromPropStatus = ({ status, frame, }) => {
    if (status.status !== 'keyframed') {
        return status;
    }
    const index = status.keyframes.findIndex((kf) => kf.frame === frame);
    if (index === -1) {
        return status;
    }
    const keyframes = status.keyframes.filter((_, i) => i !== index);
    if (keyframes.length === 0) {
        return {
            status: 'static',
            codeValue: status.keyframes[index].value,
        };
    }
    // Easing holds one segment per gap between consecutive keyframes
    // (keyframes.length - 1 entries). Drop the segment adjacent to the removed
    // keyframe so the invariant keeps holding until the server responds.
    const easing = [...status.easing];
    if (easing.length > 0) {
        const easingIndexToRemove = index === 0 ? 0 : index - 1;
        easing.splice(easingIndexToRemove, 1);
    }
    return {
        ...status,
        keyframes,
        easing,
    };
};
const optimisticDeleteSequenceKeyframe = ({ previous, fieldKey, frame, }) => {
    if (!previous.canUpdate) {
        return previous;
    }
    const status = previous.props[fieldKey];
    if (!status) {
        return previous;
    }
    return {
        ...previous,
        props: {
            ...previous.props,
            [fieldKey]: removeKeyframeFromPropStatus({ status, frame }),
        },
    };
};
exports.optimisticDeleteSequenceKeyframe = optimisticDeleteSequenceKeyframe;
const optimisticDeleteSequenceKeyframes = ({ previous, keyframes, }) => {
    return keyframes.reduce((current, keyframe) => (0, exports.optimisticDeleteSequenceKeyframe)({
        previous: current,
        fieldKey: keyframe.fieldKey,
        frame: keyframe.frame,
    }), previous);
};
exports.optimisticDeleteSequenceKeyframes = optimisticDeleteSequenceKeyframes;
const optimisticDeleteEffectKeyframe = ({ previous, effectIndex, fieldKey, frame, }) => {
    if (!previous.canUpdate) {
        return previous;
    }
    const targetIndex = previous.effects.findIndex((e) => e.effectIndex === effectIndex);
    if (targetIndex === -1) {
        return previous;
    }
    const target = previous.effects[targetIndex];
    if (!target.canUpdate) {
        return previous;
    }
    const status = target.props[fieldKey];
    if (!status) {
        return previous;
    }
    const updatedEffect = {
        ...target,
        props: {
            ...target.props,
            [fieldKey]: removeKeyframeFromPropStatus({ status, frame }),
        },
    };
    const effects = [...previous.effects];
    effects[targetIndex] = updatedEffect;
    return {
        ...previous,
        effects,
    };
};
exports.optimisticDeleteEffectKeyframe = optimisticDeleteEffectKeyframe;
const optimisticDeleteEffectKeyframes = ({ previous, keyframes, }) => {
    return keyframes.reduce((current, keyframe) => (0, exports.optimisticDeleteEffectKeyframe)({
        previous: current,
        effectIndex: keyframe.effectIndex,
        fieldKey: keyframe.fieldKey,
        frame: keyframe.frame,
    }), previous);
};
exports.optimisticDeleteEffectKeyframes = optimisticDeleteEffectKeyframes;
