// @remotion/studio-shared/dist/optimistic-add-keyframe.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.optimisticAddEffectKeyframe = exports.optimisticAddSequenceKeyframe = void 0;
const keyframe_easing_presets_1 = require("./keyframe-easing-presets");
const keyframe_interpolation_function_1 = require("./keyframe-interpolation-function");
const getEasingIndexToDuplicate = ({ insertedKeyframeIndex, easingLength, keyframeCount, }) => {
    const isSplittingExistingSegment = insertedKeyframeIndex > 0 && insertedKeyframeIndex < keyframeCount - 1;
    if (!isSplittingExistingSegment || easingLength === 0) {
        return null;
    }
    return Math.min(insertedKeyframeIndex - 1, easingLength - 1);
};
const addKeyframeToPropStatus = ({ status, fieldKey, frame, value, schema, }) => {
    var _a;
    if (status.status === 'keyframed') {
        const existingIndex = status.keyframes.findIndex((kf) => kf.frame === frame);
        if (existingIndex !== -1) {
            const updatedKeyframes = status.keyframes.map((keyframe, index) => index === existingIndex ? { frame, value } : keyframe);
            return {
                ...status,
                keyframes: updatedKeyframes,
            };
        }
        const keyframes = [...status.keyframes, { frame, value }].sort((first, second) => first.frame - second.frame);
        const easing = [...status.easing];
        const insertedKeyframeIndex = keyframes.findIndex((keyframe) => keyframe.frame === frame);
        const easingIndexToDuplicate = getEasingIndexToDuplicate({
            insertedKeyframeIndex,
            easingLength: easing.length,
            keyframeCount: keyframes.length,
        });
        const easingToDuplicate = easingIndexToDuplicate === null
            ? keyframe_easing_presets_1.LINEAR_KEYFRAME_EASING
            : easing[easingIndexToDuplicate];
        easing.splice(insertedKeyframeIndex, 0, easingToDuplicate);
        while (easing.length < keyframes.length - 1) {
            easing.push(keyframe_easing_presets_1.LINEAR_KEYFRAME_EASING);
        }
        return {
            ...status,
            keyframes,
            easing,
        };
    }
    if (status.status === 'static') {
        const staticValue = (_a = status.codeValue) !== null && _a !== void 0 ? _a : value;
        return {
            status: 'keyframed',
            interpolationFunction: (0, keyframe_interpolation_function_1.getKeyframeInterpolationFunction)({
                schema,
                key: fieldKey,
                staticValue,
                newValue: value,
            }),
            keyframes: [{ frame, value }],
            easing: [],
            clamping: { left: 'clamp', right: 'clamp' },
            posterize: undefined,
        };
    }
    return status;
};
const findFieldInSchema = (schema, key) => {
    if (key in schema) {
        return schema[key];
    }
    for (const field of Object.values(schema)) {
        if (field.type !== 'enum') {
            continue;
        }
        for (const variant of Object.values(field.variants)) {
            const found = findFieldInSchema(variant, key);
            if (found) {
                return found;
            }
        }
    }
    return undefined;
};
const getMissingPropStatus = ({ schema, fieldKey, }) => {
    const field = schema ? findFieldInSchema(schema, fieldKey) : undefined;
    if (field && field.type !== 'hidden' && field.default !== undefined) {
        return {
            status: 'static',
            codeValue: field.default,
        };
    }
    return {
        status: 'static',
        codeValue: undefined,
    };
};
const optimisticAddSequenceKeyframe = ({ previous, fieldKey, frame, value, schema, }) => {
    var _a;
    if (!previous.canUpdate) {
        return previous;
    }
    if (!(0, keyframe_interpolation_function_1.isSchemaFieldKeyframable)({ schema: schema !== null && schema !== void 0 ? schema : null, key: fieldKey })) {
        return previous;
    }
    const status = (_a = previous.props[fieldKey]) !== null && _a !== void 0 ? _a : getMissingPropStatus({ schema: schema !== null && schema !== void 0 ? schema : null, fieldKey });
    return {
        ...previous,
        props: {
            ...previous.props,
            [fieldKey]: addKeyframeToPropStatus({
                status,
                fieldKey,
                frame,
                value,
                schema: schema !== null && schema !== void 0 ? schema : null,
            }),
        },
    };
};
exports.optimisticAddSequenceKeyframe = optimisticAddSequenceKeyframe;
const optimisticAddEffectKeyframe = ({ previous, effectIndex, fieldKey, frame, value, schema, }) => {
    var _a;
    if (!previous.canUpdate) {
        return previous;
    }
    if (!(0, keyframe_interpolation_function_1.isSchemaFieldKeyframable)({ schema: schema !== null && schema !== void 0 ? schema : null, key: fieldKey })) {
        return previous;
    }
    const targetIndex = previous.effects.findIndex((e) => e.effectIndex === effectIndex);
    if (targetIndex === -1) {
        return previous;
    }
    const target = previous.effects[targetIndex];
    if (!target.canUpdate) {
        return previous;
    }
    const status = (_a = target.props[fieldKey]) !== null && _a !== void 0 ? _a : getMissingPropStatus({ schema: schema !== null && schema !== void 0 ? schema : null, fieldKey });
    const updatedEffect = {
        ...target,
        props: {
            ...target.props,
            [fieldKey]: addKeyframeToPropStatus({
                status,
                fieldKey,
                frame,
                value,
                schema: schema !== null && schema !== void 0 ? schema : null,
            }),
        },
    };
    const effects = [...previous.effects];
    effects[targetIndex] = updatedEffect;
    return {
        ...previous,
        effects,
    };
};
exports.optimisticAddEffectKeyframe = optimisticAddEffectKeyframe;
