// @remotion/studio-shared/dist/schema-field-info.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getEffectFieldsToShow = exports.getFieldsToShow = exports.getSchemaFieldGroup = exports.SCHEMA_FIELD_GROUPS = exports.SCHEMA_FIELD_ROW_HEIGHT = void 0;
const remotion_1 = require("remotion");
const no_react_1 = require("remotion/no-react");
exports.SCHEMA_FIELD_ROW_HEIGHT = 22;
exports.SCHEMA_FIELD_GROUPS = [
    { id: 'controls', label: 'Controls' },
    { id: 'transforms', label: 'Transform' },
    { id: 'text', label: 'Text' },
];
const schemaFieldGroupOrder = exports.SCHEMA_FIELD_GROUPS.reduce((acc, group, index) => {
    acc[group.id] = index;
    return acc;
}, {});
const TRANSFORM_FIELD_KEYS = new Set([
    'style.transformOrigin',
    'style.translate',
    'style.scale',
    'style.rotate',
    'style.opacity',
]);
const TEXT_FIELD_KEYS = new Set([
    'style.color',
    'style.fontSize',
    'style.lineHeight',
    'style.fontWeight',
    'style.fontStyle',
    'style.letterSpacing',
    'style.textAlign',
]);
const getSchemaFieldGroup = (key) => {
    if (TRANSFORM_FIELD_KEYS.has(key)) {
        return 'transforms';
    }
    if (TEXT_FIELD_KEYS.has(key)) {
        return 'text';
    }
    return 'controls';
};
exports.getSchemaFieldGroup = getSchemaFieldGroup;
const sortSchemaFields = (fields) => {
    return fields
        .map((field, index) => ({ field, index }))
        .sort((a, b) => {
        const groupDiff = schemaFieldGroupOrder[a.field.group] -
            schemaFieldGroupOrder[b.field.group];
        return groupDiff === 0 ? a.index - b.index : groupDiff;
    })
        .map(({ field }) => field);
};
const SUPPORTED_SCHEMA_TYPES = [
    'number',
    'boolean',
    'rotation-css',
    'rotation-degrees',
    'translate',
    'transform-origin',
    'scale',
    'uv-coordinate',
    'color',
    'array',
    'enum',
    'hidden',
];
const getArrayRowCount = ({ fieldSchema, value, }) => {
    var _a, _b;
    const items = Array.isArray(value)
        ? value
        : Array.isArray(fieldSchema.default)
            ? fieldSchema.default
            : Array.from({ length: (_a = fieldSchema.minLength) !== null && _a !== void 0 ? _a : 0 });
    const canAdd = items.length < ((_b = fieldSchema.maxLength) !== null && _b !== void 0 ? _b : Infinity);
    return Math.max(1, items.length + (canAdd ? 1 : 0));
};
const getSchemaFieldRowHeight = ({ fieldSchema, value, }) => {
    if (fieldSchema.type === 'array') {
        return (getArrayRowCount({
            fieldSchema,
            value,
        }) * exports.SCHEMA_FIELD_ROW_HEIGHT);
    }
    return exports.SCHEMA_FIELD_ROW_HEIGHT;
};
const getEffectFieldValue = ({ key, dragOverrides, effectStatus, }) => {
    const dragOverride = remotion_1.Internals.getStaticDragOverrideValue(dragOverrides[key]);
    if (dragOverride !== undefined) {
        return dragOverride;
    }
    if ((effectStatus === null || effectStatus === void 0 ? void 0 : effectStatus.type) !== 'can-update-effect') {
        return undefined;
    }
    const propStatus = effectStatus.props[key];
    if ((propStatus === null || propStatus === void 0 ? void 0 : propStatus.status) !== 'static') {
        return undefined;
    }
    return propStatus.codeValue;
};
const getFieldsToShow = ({ getDragOverrides, propStatuses, nodePath, schema, currentRuntimeValueDotNotation, }) => {
    const { merged: valuesDotNotation } = remotion_1.Internals.computeEffectiveSchemaValuesDotNotation({
        schema,
        currentValue: currentRuntimeValueDotNotation,
        overrideValues: getDragOverrides(nodePath),
        propStatus: remotion_1.Internals.getPropStatusesCtx(propStatuses, nodePath),
        frame: null,
    });
    const activeSchema = remotion_1.Internals.flattenActiveSchema(schema, (key) => valuesDotNotation[key]);
    const fields = Object.entries(activeSchema)
        .map(([key, fieldSchema]) => {
        const typeName = fieldSchema.type;
        if (SUPPORTED_SCHEMA_TYPES.indexOf(typeName) === -1) {
            throw new Error(`Unsupported field type: ${typeName}`);
        }
        if (typeName === 'hidden') {
            return null;
        }
        if (fieldSchema.type === 'number' && fieldSchema.hiddenFromList) {
            return null;
        }
        // `hidden` is represented as the eye/speaker icon on the timeline track,
        // so we don't render it as a regular field in the expanded section.
        if (key === 'hidden') {
            return null;
        }
        return {
            kind: 'sequence-field',
            key,
            description: fieldSchema.description,
            typeName,
            rowHeight: getSchemaFieldRowHeight({
                fieldSchema,
                value: valuesDotNotation[key],
            }),
            fieldSchema,
            group: (0, exports.getSchemaFieldGroup)(key),
        };
    })
        .filter(no_react_1.NoReactInternals.truthy);
    return sortSchemaFields(fields);
};
exports.getFieldsToShow = getFieldsToShow;
const getEffectFieldsToShow = ({ effect, effectIndex, nodePath, propStatuses, getEffectDragOverrides, }) => {
    const effectStatus = nodePath === null
        ? null
        : remotion_1.Internals.getEffectPropStatusesCtx({
            propStatuses,
            nodePath,
            effectIndex,
        });
    const dragOverrides = nodePath === null ? {} : getEffectDragOverrides(nodePath, effectIndex);
    const activeSchema = remotion_1.Internals.flattenActiveSchema(effect.schema, (key) => {
        return getEffectFieldValue({ key, dragOverrides, effectStatus });
    });
    const fields = Object.entries(activeSchema)
        .map(([key, fieldSchema]) => {
        const typeName = fieldSchema.type;
        if (typeName === 'hidden') {
            return null;
        }
        if (fieldSchema.type === 'number' && fieldSchema.hiddenFromList) {
            return null;
        }
        // `disabled` is represented as the eye icon on the effect timeline row,
        // so we don't render it as a regular field in the expanded section.
        if (key === 'disabled') {
            return null;
        }
        if (SUPPORTED_SCHEMA_TYPES.indexOf(typeName) === -1) {
            throw new Error(`Unsupported field type: ${typeName}`);
        }
        return {
            kind: 'effect-field',
            key,
            description: fieldSchema.description,
            typeName,
            rowHeight: getSchemaFieldRowHeight({
                fieldSchema,
                value: getEffectFieldValue({ key, dragOverrides, effectStatus }),
            }),
            fieldSchema,
            effectSchema: effect.schema,
            effectIndex,
            group: (0, exports.getSchemaFieldGroup)(key),
        };
    })
        .filter(no_react_1.NoReactInternals.truthy);
    return sortSchemaFields(fields);
};
exports.getEffectFieldsToShow = getEffectFieldsToShow;
