// @remotion/renderer/dist/options/allow-html-in-canvas.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.allowHtmlInCanvasOption = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
let allowHtmlInCanvasEnabled = false;
const cliFlag = 'allow-html-in-canvas';
exports.allowHtmlInCanvasOption = {
    name: 'Allow HTML-in-canvas for client-side rendering',
    cliFlag,
    description: () => (jsx_runtime_1.jsxs(jsx_runtime_1.Fragment, { children: ["When client-side rendering is enabled in the Studio, allow the experimental Chromium HTML-in-canvas API to be used for capturing frames. See", ' ', jsx_runtime_1.jsx("a", { href: "/docs/client-side-rendering/html-in-canvas", children: "HTML-in-canvas docs" }),
            "."] })),
    ssrName: null,
    docLink: 'https://www.remotion.dev/docs/client-side-rendering/html-in-canvas',
    type: false,
    getValue: ({ commandLine }) => {
        if (commandLine[cliFlag] !== null) {
            return {
                value: commandLine[cliFlag],
                source: 'cli',
            };
        }
        return {
            value: allowHtmlInCanvasEnabled,
            source: 'config',
        };
    },
    setConfig(value) {
        allowHtmlInCanvasEnabled = value;
    },
    id: cliFlag,
};
