// @remotion/renderer/dist/options/preview-sample-rate.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.previewSampleRateOption = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const cliFlag = 'preview-sample-rate';
let currentPreviewSampleRate = null;
exports.previewSampleRateOption = {
    name: 'Preview Sample Rate',
    cliFlag,
    description: () => (jsx_runtime_1.jsxs(jsx_runtime_1.Fragment, { children: ["Controls the sample rate used for audio playback during preview. When unset, Remotion uses ",
            jsx_runtime_1.jsx("code", { children: "48000" }),
            " Hz."] })),
    ssrName: null,
    docLink: 'https://www.remotion.dev/docs/config#setpreviewsamplerate',
    type: null,
    getValue: ({ commandLine }) => {
        if (commandLine[cliFlag] !== undefined) {
            return { value: commandLine[cliFlag], source: 'cli' };
        }
        if (currentPreviewSampleRate !== null) {
            return { value: currentPreviewSampleRate, source: 'config file' };
        }
        return { value: null, source: 'default' };
    },
    setConfig: (value) => {
        currentPreviewSampleRate = value;
    },
    id: cliFlag,
};
