// @remotion/studio/dist/components/NewComposition/CodemodFooter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CodemodFooter = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const react_1 = require("react");
const ShortcutHint_1 = require("../../error-overlay/remotion-overlay/ShortcutHint");
const resolved_stack_to_symbolicated_1 = require("../../helpers/resolved-stack-to-symbolicated");
const use_keybinding_1 = require("../../helpers/use-keybinding");
const modals_1 = require("../../state/modals");
const layout_1 = require("../layout");
const ModalButton_1 = require("../ModalButton");
const NotificationCenter_1 = require("../Notifications/NotificationCenter");
const actions_1 = require("../RenderQueue/actions");
const use_resolved_stack_1 = require("../Timeline/use-resolved-stack");
const DiffPreview_1 = require("./DiffPreview");
const CodemodFooter = ({ codemod, stack, valid, loadingNotification, successNotification, errorNotification, genericSubmitLabel, submitLabel, onSuccess, fallbackToRootFile = false, }) => {
    var _a;
    const [submitting, setSubmitting] = (0, react_1.useState)(false);
    const { setSelectedModal } = (0, react_1.useContext)(modals_1.ModalsContext);
    const [codemodStatus, setCanApplyCodemod] = (0, react_1.useState)({
        type: 'loading',
    });
    const resolvedLocation = (0, use_resolved_stack_1.useResolvedStack)(stack);
    const symbolicatedStack = (0, react_1.useMemo)(() => (0, resolved_stack_to_symbolicated_1.resolvedStackToSymbolicated)(resolvedLocation), [resolvedLocation]);
    const relativeFilePath = (_a = symbolicatedStack === null || symbolicatedStack === void 0 ? void 0 : symbolicatedStack.originalFileName) !== null && _a !== void 0 ? _a : null;
    const trigger = (0, react_1.useCallback)(() => {
        setSubmitting(true);
        setSelectedModal(null);
        const notification = (0, NotificationCenter_1.showNotification)(loadingNotification, null);
        (0, actions_1.applyCodemod)({
            codemod,
            dryRun: false,
            symbolicatedStack,
            signal: new AbortController().signal,
        })
            .then(() => {
            notification.replaceContent(successNotification, 2000);
            onSuccess === null || onSuccess === void 0 ? void 0 : onSuccess();
        })
            .catch((err) => {
            notification.replaceContent(`${errorNotification}: ${err.message}`, 2000);
        });
    }, [
        codemod,
        errorNotification,
        loadingNotification,
        onSuccess,
        setSelectedModal,
        successNotification,
        symbolicatedStack,
    ]);
    const getCanApplyCodemod = (0, react_1.useCallback)(async (signal) => {
        const res = await (0, actions_1.applyCodemod)({
            codemod,
            dryRun: true,
            symbolicatedStack,
            signal,
        });
        if (res.success) {
            setCanApplyCodemod({ type: 'success', diff: res.diff });
        }
        else {
            setCanApplyCodemod({
                type: 'fail',
                error: res.reason,
            });
        }
    }, [codemod, symbolicatedStack]);
    (0, react_1.useEffect)(() => {
        if (!stack) {
            if (fallbackToRootFile) {
                const rootFileAbortController = new AbortController();
                let rootFileAborted = false;
                getCanApplyCodemod(rootFileAbortController.signal)
                    .then(() => undefined)
                    .catch((err) => {
                    if (rootFileAborted) {
                        return;
                    }
                    (0, NotificationCenter_1.showNotification)(`${errorNotification}: ${err.message}`, 3000);
                });
                return () => {
                    rootFileAborted = true;
                    rootFileAbortController.abort();
                };
            }
            setCanApplyCodemod({
                type: 'fail',
                error: 'Could not determine where this item is defined',
            });
            return;
        }
        if (!(0, use_resolved_stack_1.hasResolvedStack)(stack)) {
            return;
        }
        if (!symbolicatedStack) {
            setCanApplyCodemod({
                type: 'fail',
                error: 'Could not resolve the source location of this item',
            });
            return;
        }
        const abortController = new AbortController();
        let aborted = false;
        getCanApplyCodemod(abortController.signal)
            .then(() => undefined)
            .catch((err) => {
            if (aborted) {
                return;
            }
            (0, NotificationCenter_1.showNotification)(`${errorNotification}: ${err.message}`, 3000);
        });
        return () => {
            aborted = true;
            abortController.abort();
        };
    }, [
        errorNotification,
        fallbackToRootFile,
        getCanApplyCodemod,
        stack,
        symbolicatedStack,
    ]);
    const disabled = !valid ||
        submitting ||
        (symbolicatedStack === null && !fallbackToRootFile) ||
        codemodStatus.type !== 'success';
    const { registerKeybinding } = (0, use_keybinding_1.useKeybinding)();
    (0, react_1.useEffect)(() => {
        if (disabled) {
            return;
        }
        const enter = registerKeybinding({
            callback() {
                trigger();
            },
            commandCtrlKey: false,
            key: 'Enter',
            event: 'keydown',
            preventDefault: true,
            triggerIfInputFieldFocused: true,
            keepRegisteredWhenNotHighestContext: false,
        });
        return () => {
            enter.unregister();
        };
    }, [disabled, registerKeybinding, trigger, valid]);
    return (jsx_runtime_1.jsxs(layout_1.Row, { align: "center", children: [
            jsx_runtime_1.jsx(DiffPreview_1.CodemodDiffPreview, { status: codemodStatus }), jsx_runtime_1.jsx(layout_1.Flex, {}), jsx_runtime_1.jsx(layout_1.Spacing, { block: true, x: 2 }), jsx_runtime_1.jsxs(ModalButton_1.ModalButton, { onClick: trigger, disabled: disabled, children: [relativeFilePath
                        ? submitLabel({ relativeRootPath: relativeFilePath })
                        : genericSubmitLabel, jsx_runtime_1.jsx(ShortcutHint_1.ShortcutHint, { keyToPress: "\u21B5", cmdOrCtrl: false })
                ] })
        ] }));
};
exports.CodemodFooter = CodemodFooter;
