// @remotion/studio/dist/components/NewComposition/DeleteFolder.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DeleteFolder = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const react_1 = require("react");
const get_folder_id_1 = require("../../helpers/get-folder-id");
const styles_1 = require("../Menu/styles");
const ModalFooter_1 = require("../ModalFooter");
const ModalHeader_1 = require("../ModalHeader");
const CodemodFooter_1 = require("./CodemodFooter");
const DismissableModal_1 = require("./DismissableModal");
const content = {
    padding: 16,
    fontSize: 14,
    flex: 1,
    minWidth: 500,
};
const DeleteFolder = ({ folderName, parentName, stack }) => {
    const codemod = (0, react_1.useMemo)(() => {
        return {
            type: 'delete-folder',
            folderName,
            parentName,
        };
    }, [folderName, parentName]);
    const onSubmit = (0, react_1.useCallback)((e) => {
        e.preventDefault();
    }, []);
    const folderId = (0, get_folder_id_1.getFolderId)({ folderName, parentName });
    return (jsx_runtime_1.jsxs(DismissableModal_1.DismissableModal, { children: [
            jsx_runtime_1.jsx(ModalHeader_1.ModalHeader, { title: 'Delete folder' }), jsx_runtime_1.jsxs("form", { onSubmit: onSubmit, children: [
                    jsx_runtime_1.jsxs("div", { style: content, children: ["Do you want to delete the", ' ', jsx_runtime_1.jsx("code", { style: styles_1.inlineCodeSnippet, children: "Folder" }),
                            " with ID ", '"', folderId, '"', "?",
                            jsx_runtime_1.jsx("br", {}),
                            "The compositions and nested folders inside it will stay in your code."] }), jsx_runtime_1.jsx(ModalFooter_1.ModalFooterContainer, { children: jsx_runtime_1.jsx(CodemodFooter_1.CodemodFooter, { errorNotification: `Could not delete folder`, loadingNotification: 'Deleting folder', successNotification: `Deleted folder ${folderId}`, genericSubmitLabel: `Delete`, submitLabel: ({ relativeRootPath }) => `Delete from ${relativeRootPath}`, codemod: codemod, stack: stack, valid: true, onSuccess: null }) })
                ] })
        ] }));
};
exports.DeleteFolder = DeleteFolder;
