// @remotion/studio/dist/components/NewComposition/NewComposition.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.NewComposition = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const react_1 = require("react");
const remotion_1 = require("remotion");
const url_state_1 = require("../../helpers/url-state");
const validate_new_comp_data_1 = require("../../helpers/validate-new-comp-data");
const layout_1 = require("../layout");
const ModalFooter_1 = require("../ModalFooter");
const ModalHeader_1 = require("../ModalHeader");
const layout_2 = require("../RenderModal/layout");
const CodemodFooter_1 = require("./CodemodFooter");
const DismissableModal_1 = require("./DismissableModal");
const InputDragger_1 = require("./InputDragger");
const NewCompDuration_1 = require("./NewCompDuration");
const RemInput_1 = require("./RemInput");
const ValidationMessage_1 = require("./ValidationMessage");
const content = {
    padding: 12,
    paddingRight: 12,
    flex: 1,
    fontSize: 13,
    minWidth: 500,
};
const toPascalCase = (value) => {
    var _a;
    const words = (_a = value.match(/[a-zA-Z0-9]+/g)) !== null && _a !== void 0 ? _a : [];
    const candidate = words
        .map((word) => `${word.charAt(0).toUpperCase()}${word.slice(1)}`)
        .join('');
    if (!candidate) {
        return 'NewComposition';
    }
    if (/^[0-9]/.test(candidate)) {
        return `Composition${candidate}`;
    }
    return candidate;
};
const waitForComposition = (compositionId) => {
    return new Promise((resolve) => {
        const started = Date.now();
        const interval = window.setInterval(() => {
            var _a;
            var _b;
            const compositionNames = (_b = (_a = window.remotion_getCompositionNames) === null || _a === void 0 ? void 0 : _a.call(window)) !== null && _b !== void 0 ? _b : [];
            if (compositionNames.includes(compositionId) ||
                Date.now() - started > 10000) {
                window.clearInterval(interval);
                resolve();
            }
        }, 100);
    });
};
const getUniqueCompositionName = (compositions) => {
    let counter = 1;
    while (true) {
        const name = counter === 1 ? 'NewComposition' : `NewComposition${counter}`;
        const err = (0, validate_new_comp_data_1.validateCompositionName)(name, compositions);
        if (!err) {
            return name;
        }
        counter++;
    }
};
const NewCompositionLoaded = () => {
    const { compositions } = (0, react_1.useContext)(remotion_1.Internals.CompositionManager);
    const [newId, setName] = (0, react_1.useState)(() => getUniqueCompositionName(compositions));
    const [selectedFrameRate, setFrameRate] = (0, react_1.useState)(30);
    const [size, setSize] = (0, react_1.useState)(() => ({
        width: 1920,
        height: 1080,
    }));
    const [durationInFrames, setDurationInFrames] = (0, react_1.useState)(150);
    const onWidthChanged = (0, react_1.useCallback)((newValue) => {
        setSize((s) => {
            return {
                height: s.height,
                width: Number(newValue),
            };
        });
    }, []);
    const onWidthDirectlyChanged = (0, react_1.useCallback)((newWidth) => {
        setSize((s) => {
            return {
                height: s.height,
                width: newWidth,
            };
        });
    }, []);
    const onHeightDirectlyChanged = (0, react_1.useCallback)((newHeight) => {
        setSize((s) => {
            return {
                width: s.width,
                height: newHeight,
            };
        });
    }, []);
    const onHeightChanged = (0, react_1.useCallback)((newValue) => {
        setSize((s) => {
            return {
                width: s.width,
                height: Number(newValue),
            };
        });
    }, []);
    const onNameChange = (0, react_1.useCallback)((e) => {
        setName(e.target.value);
    }, []);
    const onTextFpsChange = (0, react_1.useCallback)((newFps) => {
        setFrameRate(Number(newFps));
    }, []);
    const onFpsChange = (0, react_1.useCallback)((newFps) => {
        setFrameRate(newFps);
    }, []);
    const compNameErrMessage = (0, validate_new_comp_data_1.validateCompositionName)(newId, compositions);
    const compWidthErrMessage = (0, validate_new_comp_data_1.validateCompositionDimension)('Width', size.width);
    const compHeightErrMessage = (0, validate_new_comp_data_1.validateCompositionDimension)('Height', size.height);
    const componentName = toPascalCase(newId);
    const valid = compNameErrMessage === null &&
        compWidthErrMessage === null &&
        compHeightErrMessage === null;
    const codemod = (0, react_1.useMemo)(() => {
        return {
            type: 'new-composition',
            newDurationInFrames: Number(durationInFrames),
            newFps: Number(selectedFrameRate),
            newHeight: Number(size.height),
            newWidth: Number(size.width),
            newId,
            componentName,
            componentImportPath: `./${componentName}`,
        };
    }, [
        componentName,
        durationInFrames,
        newId,
        selectedFrameRate,
        size.height,
        size.width,
    ]);
    const onSuccess = (0, react_1.useCallback)(() => {
        waitForComposition(newId).then(() => {
            (0, url_state_1.pushUrl)(`/${newId}`);
        });
    }, [newId]);
    const onSubmit = (0, react_1.useCallback)((e) => {
        e.preventDefault();
    }, []);
    return (jsx_runtime_1.jsxs(jsx_runtime_1.Fragment, { children: [
            jsx_runtime_1.jsx(ModalHeader_1.ModalHeader, { title: "New composition" }), jsx_runtime_1.jsxs("form", { onSubmit: onSubmit, children: [
                    jsx_runtime_1.jsxs("div", { style: content, children: [
                            jsx_runtime_1.jsxs("div", { style: layout_2.optionRow, children: [
                                    jsx_runtime_1.jsx("div", { style: layout_2.label, children: "ID" }), jsx_runtime_1.jsx("div", { style: layout_2.rightRow, children: jsx_runtime_1.jsxs("div", { children: [
                                                jsx_runtime_1.jsx(RemInput_1.RemotionInput, { value: newId, onChange: onNameChange, type: "text", autoFocus: true, placeholder: "Composition ID", status: "ok", rightAlign: true }), compNameErrMessage ? (jsx_runtime_1.jsxs(jsx_runtime_1.Fragment, { children: [
                                                        jsx_runtime_1.jsx(layout_1.Spacing, { y: 1, block: true }), jsx_runtime_1.jsx(ValidationMessage_1.ValidationMessage, { align: "flex-start", message: compNameErrMessage, type: "error" })
                                                    ] })) : null] }) })
                                ] }), jsx_runtime_1.jsxs("div", { style: layout_2.optionRow, children: [
                                    jsx_runtime_1.jsx("div", { style: layout_2.label, children: "Width" }), jsx_runtime_1.jsxs("div", { style: layout_2.rightRow, children: [
                                            jsx_runtime_1.jsx(InputDragger_1.InputDragger, { type: "number", value: size.width, placeholder: "Width", onTextChange: onWidthChanged, name: "width", step: 2, min: 2, required: true, status: "ok", formatter: (w) => `${w}px`, max: 100000000, onValueChange: onWidthDirectlyChanged, rightAlign: false }), compWidthErrMessage ? (jsx_runtime_1.jsxs(jsx_runtime_1.Fragment, { children: [
                                                    jsx_runtime_1.jsx(layout_1.Spacing, { y: 1, block: true }), jsx_runtime_1.jsx(ValidationMessage_1.ValidationMessage, { align: "flex-start", message: compWidthErrMessage, type: "error" })
                                                ] })) : null] })
                                ] }), jsx_runtime_1.jsxs("div", { style: layout_2.optionRow, children: [
                                    jsx_runtime_1.jsx("div", { style: layout_2.label, children: "Height" }), jsx_runtime_1.jsxs("div", { style: layout_2.rightRow, children: [
                                            jsx_runtime_1.jsx(InputDragger_1.InputDragger, { type: "number", value: size.height, onTextChange: onHeightChanged, placeholder: "Height", name: "height", step: 2, required: true, formatter: (h) => `${h}px`, min: 2, status: "ok", max: 100000000, onValueChange: onHeightDirectlyChanged, rightAlign: false }), compHeightErrMessage ? (jsx_runtime_1.jsxs(jsx_runtime_1.Fragment, { children: [
                                                    jsx_runtime_1.jsx(layout_1.Spacing, { y: 1, block: true }), jsx_runtime_1.jsx(ValidationMessage_1.ValidationMessage, { align: "flex-start", message: compHeightErrMessage, type: "error" })
                                                ] })) : null] })
                                ] }), jsx_runtime_1.jsx(NewCompDuration_1.NewCompDuration, { durationInFrames: durationInFrames, setDurationInFrames: setDurationInFrames }), jsx_runtime_1.jsxs("div", { style: layout_2.optionRow, children: [
                                    jsx_runtime_1.jsx("div", { style: layout_2.label, children: "FPS" }), jsx_runtime_1.jsx("div", { style: layout_2.rightRow, children: jsx_runtime_1.jsx(InputDragger_1.InputDragger, { type: "number", value: selectedFrameRate, onTextChange: onTextFpsChange, placeholder: "Frame rate (fps)", name: "fps", min: 1, required: true, status: "ok", max: 240, step: 0.01, onValueChange: onFpsChange, rightAlign: false }) })
                                ] })
                        ] }), jsx_runtime_1.jsx(ModalFooter_1.ModalFooterContainer, { children: jsx_runtime_1.jsx(CodemodFooter_1.CodemodFooter, { loadingNotification: "Creating composition...", errorNotification: "Could not create composition", successNotification: `Created ${newId}`, genericSubmitLabel: "Add to root file", submitLabel: ({ relativeRootPath }) => `Add to ${relativeRootPath}`, codemod: codemod, stack: null, valid: valid, onSuccess: onSuccess, fallbackToRootFile: true }) })
                ] })
        ] }));
};
const NewComposition = () => {
    return (jsx_runtime_1.jsx(DismissableModal_1.DismissableModal, { children: jsx_runtime_1.jsx(NewCompositionLoaded, {}) }));
};
exports.NewComposition = NewComposition;
