// @remotion/studio/dist/components/NewComposition/RenameFolder.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RenameFolder = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const react_1 = require("react");
const remotion_1 = require("remotion");
const get_folder_id_1 = require("../../helpers/get-folder-id");
const validate_folder_rename_1 = require("../../helpers/validate-folder-rename");
const layout_1 = require("../layout");
const ModalFooter_1 = require("../ModalFooter");
const ModalHeader_1 = require("../ModalHeader");
const layout_2 = require("../RenderModal/layout");
const CodemodFooter_1 = require("./CodemodFooter");
const DismissableModal_1 = require("./DismissableModal");
const RemInput_1 = require("./RemInput");
const ValidationMessage_1 = require("./ValidationMessage");
const content = {
    padding: 12,
    paddingRight: 12,
    flex: 1,
    fontSize: 13,
    minWidth: 500,
};
const RenameFolder = ({ folderName, parentName, stack }) => {
    const { folders } = (0, react_1.useContext)(remotion_1.Internals.CompositionManager);
    const [newName, setName] = (0, react_1.useState)(folderName);
    const inputRef = (0, react_1.useRef)(null);
    (0, react_1.useEffect)(() => {
        const input = inputRef.current;
        if (!input)
            return;
        input.select();
    }, []);
    const onNameChange = (0, react_1.useCallback)((e) => {
        setName(e.target.value);
    }, []);
    const folderNameErrMessage = (0, validate_folder_rename_1.validateFolderRename)({
        folders,
        newName,
        originalName: folderName,
        parentName,
    });
    const valid = folderNameErrMessage === null && folderName !== newName;
    const codemod = (0, react_1.useMemo)(() => {
        return {
            type: 'rename-folder',
            folderName,
            parentName,
            newName,
        };
    }, [folderName, newName, parentName]);
    const onSubmit = (0, react_1.useCallback)((e) => {
        e.preventDefault();
    }, []);
    const folderId = (0, get_folder_id_1.getFolderId)({ folderName, parentName });
    return (jsx_runtime_1.jsxs(DismissableModal_1.DismissableModal, { children: [
            jsx_runtime_1.jsx(ModalHeader_1.ModalHeader, { title: `Rename ${folderId}` }), jsx_runtime_1.jsxs("form", { onSubmit: onSubmit, children: [
                    jsx_runtime_1.jsx("div", { style: content, children: jsx_runtime_1.jsxs("div", { style: layout_2.optionRow, children: [
                                jsx_runtime_1.jsx("div", { style: layout_2.label, children: "Name" }), jsx_runtime_1.jsx("div", { style: layout_2.rightRow, children: jsx_runtime_1.jsxs("div", { children: [
                                            jsx_runtime_1.jsx(RemInput_1.RemotionInput, { ref: inputRef, value: newName, onChange: onNameChange, type: "text", autoFocus: true, placeholder: "Folder name", status: "ok", rightAlign: true }), folderNameErrMessage ? (jsx_runtime_1.jsxs(jsx_runtime_1.Fragment, { children: [
                                                    jsx_runtime_1.jsx(layout_1.Spacing, { y: 1, block: true }), jsx_runtime_1.jsx(ValidationMessage_1.ValidationMessage, { align: "flex-start", message: folderNameErrMessage, type: "error" })
                                                ] })) : null] }) })
                            ] }) }), jsx_runtime_1.jsx(ModalFooter_1.ModalFooterContainer, { children: jsx_runtime_1.jsx(CodemodFooter_1.CodemodFooter, { loadingNotification: 'Renaming folder...', errorNotification: 'Could not rename folder', successNotification: `Renamed folder to ${newName}`, genericSubmitLabel: 'Rename', submitLabel: ({ relativeRootPath }) => `Modify ${relativeRootPath}`, codemod: codemod, stack: stack, valid: valid, onSuccess: null }) })
                ] })
        ] }));
};
exports.RenameFolder = RenameFolder;
