// @remotion/studio/dist/components/NewComposition/RenameStaticFile.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RenameStaticFileModal = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const react_1 = require("react");
const remotion_1 = require("remotion");
const rename_static_file_1 = require("../../api/rename-static-file");
const url_state_1 = require("../../helpers/url-state");
const modals_1 = require("../../state/modals");
const Button_1 = require("../Button");
const layout_1 = require("../layout");
const ModalFooter_1 = require("../ModalFooter");
const ModalHeader_1 = require("../ModalHeader");
const NotificationCenter_1 = require("../Notifications/NotificationCenter");
const layout_2 = require("../RenderModal/layout");
const use_static_files_1 = require("../use-static-files");
const DismissableModal_1 = require("./DismissableModal");
const RemInput_1 = require("./RemInput");
const ValidationMessage_1 = require("./ValidationMessage");
const content = {
    padding: 12,
    paddingRight: 12,
    flex: 1,
    fontSize: 13,
    minWidth: 500,
};
const getParent = (relativePath) => {
    const slashIndex = relativePath.lastIndexOf('/');
    return slashIndex === -1 ? '' : relativePath.slice(0, slashIndex);
};
const getBaseName = (relativePath) => {
    const slashIndex = relativePath.lastIndexOf('/');
    return slashIndex === -1 ? relativePath : relativePath.slice(slashIndex + 1);
};
const RenameStaticFileModal = ({ relativePath }) => {
    const { setSelectedModal } = (0, react_1.useContext)(modals_1.ModalsContext);
    const { canvasContent } = (0, react_1.useContext)(remotion_1.Internals.CompositionManager);
    const { setCanvasContent } = (0, react_1.useContext)(remotion_1.Internals.CompositionSetters);
    const staticFiles = (0, use_static_files_1.useStaticFiles)();
    const [newName, setNewName] = (0, react_1.useState)(() => getBaseName(relativePath));
    const [submitting, setSubmitting] = (0, react_1.useState)(false);
    const inputRef = (0, react_1.useRef)(null);
    (0, react_1.useEffect)(() => {
        const input = inputRef.current;
        if (!input)
            return;
        const dotIndex = newName.lastIndexOf('.');
        const stemEnd = dotIndex === -1 ? newName.length : dotIndex;
        input.setSelectionRange(0, stemEnd);
        // eslint-disable-next-line react-hooks/exhaustive-deps
    }, []);
    const parent = (0, react_1.useMemo)(() => getParent(relativePath), [relativePath]);
    const newRelativePath = (0, react_1.useMemo)(() => {
        return [parent, newName].filter(Boolean).join('/');
    }, [newName, parent]);
    const changed = newRelativePath !== relativePath;
    const validationMessage = (0, react_1.useMemo)(() => {
        const trimmedName = newName.trim();
        if (trimmedName.length === 0) {
            return 'Name cannot be empty';
        }
        if (trimmedName !== newName) {
            return 'Name cannot start or end with whitespace';
        }
        if (newName.includes('/') || newName.includes('\\')) {
            return 'Name cannot include slashes';
        }
        const existingFile = staticFiles.find((file) => {
            return file.name === newRelativePath && file.name !== relativePath;
        });
        if (existingFile) {
            return 'An asset with this name already exists';
        }
        return null;
    }, [newName, newRelativePath, relativePath, staticFiles]);
    const valid = changed && validationMessage === null;
    const onNameChange = (0, react_1.useCallback)((e) => {
        setNewName(e.target.value);
    }, []);
    const onRename = (0, react_1.useCallback)(() => {
        if (!valid) {
            return;
        }
        setSubmitting(true);
        const notification = (0, NotificationCenter_1.showNotification)(`Renaming ${relativePath}...`, null);
        (0, rename_static_file_1.renameStaticFile)({
            oldRelativePath: relativePath,
            newRelativePath,
        })
            .then(() => {
            setSelectedModal(null);
            if ((canvasContent === null || canvasContent === void 0 ? void 0 : canvasContent.type) === 'asset' &&
                canvasContent.asset === relativePath) {
                setCanvasContent({ type: 'asset', asset: newRelativePath });
                (0, url_state_1.pushUrl)(`/assets/${newRelativePath}`);
            }
            notification.replaceContent(`Renamed to ${newRelativePath}`, 2000);
        })
            .catch((err) => {
            setSubmitting(false);
            notification.replaceContent(`Could not rename ${relativePath}: ${err.message}`, 3000);
        });
    }, [
        canvasContent,
        newRelativePath,
        relativePath,
        setCanvasContent,
        setSelectedModal,
        valid,
    ]);
    const onSubmit = (0, react_1.useCallback)((e) => {
        e.preventDefault();
        onRename();
    }, [onRename]);
    return (jsx_runtime_1.jsxs(DismissableModal_1.DismissableModal, { children: [
            jsx_runtime_1.jsx(ModalHeader_1.ModalHeader, { title: `Rename ${relativePath}` }), jsx_runtime_1.jsxs("form", { onSubmit: onSubmit, children: [
                    jsx_runtime_1.jsx("div", { style: content, children: jsx_runtime_1.jsxs("div", { style: layout_2.optionRow, children: [
                                jsx_runtime_1.jsx("div", { style: layout_2.label, children: "Name" }), jsx_runtime_1.jsx("div", { style: layout_2.rightRow, children: jsx_runtime_1.jsxs("div", { children: [
                                            jsx_runtime_1.jsx(RemInput_1.RemotionInput, { ref: inputRef, value: newName, onChange: onNameChange, type: "text", autoFocus: true, autoComplete: "off", "data-1p-ignore": true, placeholder: "Asset name", status: validationMessage ? 'error' : 'ok', rightAlign: true }), validationMessage ? (jsx_runtime_1.jsxs(jsx_runtime_1.Fragment, { children: [
                                                    jsx_runtime_1.jsx(layout_1.Spacing, { y: 1, block: true }), jsx_runtime_1.jsx(ValidationMessage_1.ValidationMessage, { align: "flex-start", message: validationMessage, type: "error" })
                                                ] })) : null] }) })
                            ] }) }), jsx_runtime_1.jsx(ModalFooter_1.ModalFooterContainer, { children: jsx_runtime_1.jsxs(layout_1.Row, { align: "center", justify: "flex-end", children: [
                                jsx_runtime_1.jsx(Button_1.Button, { onClick: () => setSelectedModal(null), children: "Cancel" }), jsx_runtime_1.jsx(layout_1.Spacing, { x: 1 }), jsx_runtime_1.jsx(Button_1.Button, { onClick: onRename, disabled: !valid || submitting, children: "Rename" })
                            ] }) })
                ] })
        ] }));
};
exports.RenameStaticFileModal = RenameStaticFileModal;
