// @remotion/studio/dist/error-overlay/remotion-overlay/ErrorLoader.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ErrorLoader = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const studio_shared_1 = require("@remotion/studio-shared");
const react_1 = require("react");
const error_origin_1 = require("../error-origin");
const listen_to_runtime_errors_1 = require("../react-overlay/listen-to-runtime-errors");
const ErrorDisplay_1 = require("./ErrorDisplay");
const ErrorTitle_1 = require("./ErrorTitle");
const log_studio_error_1 = require("./log-studio-error");
const container = {
    width: '100%',
    maxWidth: 1000,
    paddingLeft: 14,
    paddingRight: 14,
    marginLeft: 'auto',
    marginRight: 'auto',
    fontFamily: 'SF Pro Text, sans-serif',
    paddingTop: '5vh',
};
const errorWhileErrorStyle = {
    color: 'white',
    lineHeight: 1.5,
    whiteSpace: 'pre',
};
const errorWhileSymbolicatingStyle = {
    color: 'white',
    lineHeight: 1.5,
    marginTop: 24,
    opacity: 0.7,
};
const shouldLogError = (error) => {
    return (!(0, error_origin_1.wasErrorLoggedByServer)(error) && (0, studio_shared_1.getLocationFromBuildError)(error) === null);
};
const shouldIncludeFrameInServerLog = (frame) => {
    var _a, _b, _c, _d;
    return !(((_a = frame.originalFileName) === null || _a === void 0 ? void 0 : _a.includes('node_modules')) ||
        ((_b = frame.originalFileName) === null || _b === void 0 ? void 0 : _b.startsWith('webpack/')) ||
        ((_c = frame.originalFileName) === null || _c === void 0 ? void 0 : _c.includes('/bundler/dist/fast-refresh/')) ||
        ((_d = frame.originalFileName) === null || _d === void 0 ? void 0 : _d.includes('bundler/dist/fast-refresh/')));
};
const logSymbolicatedStudioError = (record) => {
    const name = typeof record.error.name === 'string' ? record.error.name : null;
    const message = typeof record.error.message === 'string' ? record.error.message : '';
    const filteredStackFrames = record.stackFrames.filter(shouldIncludeFrameInServerLog);
    const stackFrames = filteredStackFrames.length > 0
        ? filteredStackFrames
        : record.stackFrames[0]
            ? [record.stackFrames[0]]
            : [];
    (0, log_studio_error_1.logStudioErrorData)({
        name,
        message,
        stack: typeof record.error.stack === 'string' ? record.error.stack : null,
        symbolicatedStackFrames: stackFrames.length > 0 ? stackFrames : null,
    });
};
const ErrorLoader = ({ error, keyboardShortcuts, onRetry, canHaveDismissButton, calculateMetadata, }) => {
    var _a;
    const [state, setState] = (0, react_1.useState)({
        type: 'loading',
    });
    (0, react_1.useEffect)(() => {
        (0, listen_to_runtime_errors_1.getErrorRecord)(error)
            .then((record) => {
            if (record) {
                if (shouldLogError(error)) {
                    logSymbolicatedStudioError(record);
                }
                setState({
                    type: 'symbolicated',
                    record,
                });
            }
            else {
                if (shouldLogError(error)) {
                    (0, log_studio_error_1.logStudioError)(error);
                }
                setState({
                    type: 'no-record',
                });
            }
        })
            .catch((err) => {
            if (shouldLogError(error)) {
                (0, log_studio_error_1.logStudioError)(error);
            }
            setState({
                err,
                type: 'error',
            });
        });
    }, [error]);
    if (state.type === 'loading') {
        return (jsx_runtime_1.jsx("div", { style: container, children: jsx_runtime_1.jsx(ErrorTitle_1.ErrorTitle, { symbolicating: true, name: error.name, message: error.message, canHaveDismissButton: canHaveDismissButton }) }));
    }
    if (state.type === 'error') {
        return (jsx_runtime_1.jsxs("div", { style: container, children: [
                jsx_runtime_1.jsx(ErrorTitle_1.ErrorTitle, { symbolicating: false, name: error.name, message: error.message, canHaveDismissButton: canHaveDismissButton }), jsx_runtime_1.jsx("div", { style: errorWhileErrorStyle, children: (_a = error.stack) !== null && _a !== void 0 ? _a : 'Check the Terminal and browser console for error messages.' }), jsx_runtime_1.jsxs("div", { style: errorWhileSymbolicatingStyle, children: ["Could not symbolicate the stack trace: ", state.err.message] })
            ] }));
    }
    if (state.type === 'no-record') {
        return (jsx_runtime_1.jsxs("div", { style: container, children: [
                jsx_runtime_1.jsx(ErrorTitle_1.ErrorTitle, { symbolicating: false, name: error.name, message: error.message, canHaveDismissButton: canHaveDismissButton }), jsx_runtime_1.jsx("div", { style: errorWhileErrorStyle, children: "Check the Terminal and browser console for error messages." })
            ] }));
    }
    return (jsx_runtime_1.jsx("div", { style: container, children: jsx_runtime_1.jsx(ErrorDisplay_1.ErrorDisplay, { keyboardShortcuts: keyboardShortcuts, display: state.record, onRetry: onRetry, canHaveDismissButton: canHaveDismissButton, calculateMetadata: calculateMetadata }) }));
};
exports.ErrorLoader = ErrorLoader;
