// @remotion/studio/dist/error-overlay/remotion-overlay/log-studio-error.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.logStudioError = exports.logStudioErrorData = void 0;
const call_api_1 = require("../../components/call-api");
const loggedErrors = new Set();
const maxLoggedErrors = 100;
const logStudioErrorData = (data) => {
    const key = JSON.stringify([data.name, data.message, data.stack]);
    if (loggedErrors.has(key)) {
        return;
    }
    if (loggedErrors.size >= maxLoggedErrors) {
        loggedErrors.clear();
    }
    loggedErrors.add(key);
    (0, call_api_1.callApi)('/api/log-studio-error', data).catch(() => undefined);
};
exports.logStudioErrorData = logStudioErrorData;
const logStudioError = (error) => {
    (0, exports.logStudioErrorData)({
        name: typeof error.name === 'string' ? error.name : null,
        message: typeof error.message === 'string' ? error.message : '',
        stack: typeof error.stack === 'string' ? error.stack : null,
        symbolicatedStackFrames: null,
    });
};
exports.logStudioError = logStudioError;
