// @remotion/studio/dist/components/EditorGuides/Guide.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const jsx_runtime_1 = require("react/jsx-runtime");
const react_1 = require("react");
const no_react_1 = require("remotion/no-react");
const editor_guide_selection_1 = require("../../helpers/editor-guide-selection");
const editor_guides_1 = require("../../state/editor-guides");
const editor_rulers_1 = require("../../state/editor-rulers");
const ContextMenu_1 = require("../ContextMenu");
const ForceSpecificCursor_1 = require("../ForceSpecificCursor");
const should_clear_selection_on_pointer_down_1 = require("../Timeline/should-clear-selection-on-pointer-down");
const TimelineSelection_1 = require("../Timeline/TimelineSelection");
const PADDING_FOR_EASY_DRAG = 4;
const GuideComp = ({ guide, canvasDimensions, scale }) => {
    const { shouldCreateGuideRef, shouldDeleteGuideRef, setGuidesList, setDraggingGuideId, setHoveredGuideId, hoveredGuideId, draggingGuideId, } = (0, react_1.useContext)(editor_guides_1.EditorShowGuidesContext);
    const { clearSelection, onSelect, selected } = (0, TimelineSelection_1.useTimelineGuideSelection)(guide.id);
    const pointerDownPositionRef = (0, react_1.useRef)(null);
    const hasMovedGuideRef = (0, react_1.useRef)(false);
    const onPointerEnter = (0, react_1.useCallback)(() => {
        setHoveredGuideId(() => guide.id);
    }, [guide.id, setHoveredGuideId]);
    const onPointerLeave = (0, react_1.useCallback)(() => {
        setHoveredGuideId(() => null);
    }, [setHoveredGuideId]);
    const isVerticalGuide = guide.orientation === 'vertical';
    const guideStyle = (0, react_1.useMemo)(() => {
        const canvasPosition = isVerticalGuide
            ? canvasDimensions.left
            : canvasDimensions.top;
        const guidePosition = guide.position * scale + canvasPosition;
        return {
            position: 'absolute',
            width: `${isVerticalGuide ? '1px' : '100%'}`,
            height: `${isVerticalGuide ? '100%' : '1px'}`,
            left: `${isVerticalGuide ? guidePosition - PADDING_FOR_EASY_DRAG : 0}px`,
            top: `${isVerticalGuide ? 0 : guidePosition - PADDING_FOR_EASY_DRAG}px`,
            cursor: `${isVerticalGuide ? 'ew-resize' : 'ns-resize'}`,
            padding: isVerticalGuide
                ? `0 ${PADDING_FOR_EASY_DRAG}px`
                : `${PADDING_FOR_EASY_DRAG}px 0`,
        };
    }, [guide, scale, canvasDimensions, isVerticalGuide]);
    const guideContentStyle = (0, react_1.useMemo)(() => {
        return {
            position: 'relative',
            minWidth: `${isVerticalGuide ? '1px' : `calc(100% + ${editor_rulers_1.RULER_WIDTH}px`}`,
            minHeight: `${isVerticalGuide ? `calc(100% + ${editor_rulers_1.RULER_WIDTH}px` : '1px'}`,
            top: `${isVerticalGuide ? `-${editor_rulers_1.RULER_WIDTH}px` : '0px'}`,
            left: `${isVerticalGuide ? '0px' : `-${editor_rulers_1.RULER_WIDTH}px`}`,
            display: guide.show ? 'block' : 'none',
            backgroundColor: (0, editor_guide_selection_1.getEditorGuideColor)({
                selected,
                active: hoveredGuideId === guide.id || draggingGuideId === guide.id,
            }),
        };
    }, [
        isVerticalGuide,
        guide.show,
        hoveredGuideId,
        guide.id,
        selected,
        draggingGuideId,
    ]);
    const onPointerDown = (0, react_1.useCallback)((e) => {
        if (e.button !== 0) {
            return;
        }
        e.stopPropagation();
        e.currentTarget.setPointerCapture(e.pointerId);
        hasMovedGuideRef.current = false;
        pointerDownPositionRef.current = {
            guideId: guide.id,
            clientX: e.clientX,
            clientY: e.clientY,
        };
        shouldCreateGuideRef.current = false;
        setDraggingGuideId(() => guide.id);
    }, [guide.id, setDraggingGuideId, shouldCreateGuideRef]);
    const onMouseDown = (0, react_1.useCallback)((e) => {
        if (e.button !== 0) {
            return;
        }
        e.preventDefault();
        e.stopPropagation();
        hasMovedGuideRef.current = false;
        pointerDownPositionRef.current = {
            guideId: guide.id,
            clientX: e.clientX,
            clientY: e.clientY,
        };
        shouldCreateGuideRef.current = false;
        setDraggingGuideId(() => guide.id);
    }, [guide.id, setDraggingGuideId, shouldCreateGuideRef]);
    const updateHasMovedGuide = (0, react_1.useCallback)((clientX, clientY) => {
        const pointerDownPosition = pointerDownPositionRef.current;
        if (pointerDownPosition === null) {
            return;
        }
        if (!(0, editor_guide_selection_1.isGuidePointerUpAClick)({
            pointerDownPosition,
            guideId: guide.id,
            clientX,
            clientY,
        })) {
            hasMovedGuideRef.current = true;
        }
    }, [guide.id]);
    const onPointerMove = (0, react_1.useCallback)((e) => {
        updateHasMovedGuide(e.clientX, e.clientY);
    }, [updateHasMovedGuide]);
    const onMouseMove = (0, react_1.useCallback)((e) => {
        updateHasMovedGuide(e.clientX, e.clientY);
    }, [updateHasMovedGuide]);
    const finishGuideInteraction = (0, react_1.useCallback)(() => {
        const shouldDeleteGuide = shouldDeleteGuideRef.current;
        setGuidesList((prevState) => {
            const newGuides = shouldDeleteGuide
                ? prevState.filter((candidate) => candidate.id !== guide.id)
                : prevState;
            (0, editor_guides_1.persistGuidesList)(newGuides);
            return newGuides;
        });
        if (shouldDeleteGuide && selected) {
            clearSelection();
        }
        shouldDeleteGuideRef.current = false;
        shouldCreateGuideRef.current = false;
        (0, ForceSpecificCursor_1.stopForcingSpecificCursor)();
        setDraggingGuideId(() => null);
    }, [
        clearSelection,
        guide.id,
        selected,
        setDraggingGuideId,
        setGuidesList,
        shouldCreateGuideRef,
        shouldDeleteGuideRef,
    ]);
    const onPointerUp = (0, react_1.useCallback)((e) => {
        if (e.currentTarget.hasPointerCapture(e.pointerId)) {
            e.currentTarget.releasePointerCapture(e.pointerId);
        }
        const pointerDownPosition = pointerDownPositionRef.current;
        pointerDownPositionRef.current = null;
        const shouldDeleteGuide = shouldDeleteGuideRef.current;
        finishGuideInteraction();
        if (shouldDeleteGuide) {
            return;
        }
        if ((0, editor_guide_selection_1.isGuidePointerUpAClick)({
            pointerDownPosition,
            guideId: guide.id,
            clientX: e.clientX,
            clientY: e.clientY,
        })) {
            onSelect();
        }
    }, [finishGuideInteraction, guide.id, onSelect, shouldDeleteGuideRef]);
    const onMouseUp = (0, react_1.useCallback)((e) => {
        const pointerDownPosition = pointerDownPositionRef.current;
        pointerDownPositionRef.current = null;
        const shouldDeleteGuide = shouldDeleteGuideRef.current;
        finishGuideInteraction();
        if (shouldDeleteGuide) {
            return;
        }
        if ((0, editor_guide_selection_1.isGuidePointerUpAClick)({
            pointerDownPosition,
            guideId: guide.id,
            clientX: e.clientX,
            clientY: e.clientY,
        })) {
            onSelect();
        }
    }, [finishGuideInteraction, guide.id, onSelect, shouldDeleteGuideRef]);
    const onClick = (0, react_1.useCallback)((e) => {
        if (e.button !== 0) {
            return;
        }
        e.preventDefault();
        e.stopPropagation();
    }, []);
    const onPointerCancel = (0, react_1.useCallback)(() => {
        pointerDownPositionRef.current = null;
        finishGuideInteraction();
    }, [finishGuideInteraction]);
    const isActive = selected || hoveredGuideId === guide.id;
    const activeClassName = isActive ? '__remotion_editor_guide_selected' : null;
    const guideClassName = (0, react_1.useMemo)(() => {
        return ['__remotion_editor_guide_content', activeClassName]
            .filter(no_react_1.NoReactInternals.truthy)
            .join(' ');
    }, [activeClassName]);
    const values = (0, react_1.useMemo)(() => {
        return [
            {
                id: '1',
                keyHint: null,
                label: 'Remove guide',
                leftItem: null,
                onClick: () => {
                    setGuidesList((prevState) => {
                        const newGuides = prevState.filter((candidate) => {
                            return candidate.id !== guide.id;
                        });
                        (0, editor_guides_1.persistGuidesList)(newGuides);
                        return newGuides;
                    });
                    if (selected) {
                        clearSelection();
                    }
                },
                quickSwitcherLabel: null,
                subMenu: null,
                type: 'item',
                value: 'remove',
            },
        ];
    }, [clearSelection, guide.id, selected, setGuidesList]);
    return (jsx_runtime_1.jsx(ContextMenu_1.ContextMenu, { values: values, onOpen: null, children: jsx_runtime_1.jsx("div", { style: guideStyle, onPointerDown: onPointerDown, onPointerMove: onPointerMove, onPointerUp: onPointerUp, onPointerCancel: onPointerCancel, onMouseDown: onMouseDown, onMouseMove: onMouseMove, onMouseUp: onMouseUp, onClick: onClick, className: "__remotion_editor_guide", [should_clear_selection_on_pointer_down_1.PREVENT_CLEAR_SELECTION_ON_POINTER_DOWN_ATTR]: 'true', onPointerEnter: onPointerEnter, onPointerLeave: onPointerLeave, children: jsx_runtime_1.jsx("div", { style: guideContentStyle, className: guideClassName }) }) }));
};
exports.default = (0, react_1.memo)(GuideComp);
