// @remotion/studio/dist/components/RenderModal/SchemaEditor/ZodErrorMessages.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ZodErrorMessages = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const react_1 = require("react");
const colors_1 = require("../../../helpers/colors");
const layout_1 = require("../../layout");
const ValidationMessage_1 = require("../../NewComposition/ValidationMessage");
const schemaLabel = {
    fontSize: 14,
    color: colors_1.LIGHT_TEXT,
};
const compactSchemaLabel = {
    ...schemaLabel,
    fontSize: 12,
};
const jsonLabel = {
    color: 'white',
    fontSize: 13,
    fontFamily: 'sans-serif',
    display: 'flex',
    alignItems: 'center',
};
const compactJsonLabel = {
    ...jsonLabel,
    fontSize: 12,
};
const triangleStyle = {
    width: 12,
    height: 12,
    flexShrink: 0,
    fill: colors_1.FAIL_COLOR,
};
const compactTriangleStyle = {
    ...triangleStyle,
    width: 10,
    height: 10,
};
const ZodErrorMessages = ({ zodValidationResult, viewTab, size = 'default' }) => {
    if (zodValidationResult.success) {
        throw new Error('Expected error');
    }
    const style = (0, react_1.useMemo)(() => {
        if (viewTab === 'json') {
            return size === 'compact' ? compactJsonLabel : jsonLabel;
        }
        return size === 'compact' ? compactSchemaLabel : schemaLabel;
    }, [size, viewTab]);
    const code = (0, react_1.useMemo)(() => {
        return {
            ...(size === 'compact' ? compactSchemaLabel : schemaLabel),
            fontFamily: 'monospace',
        };
    }, [size]);
    const finalTriangleStyle = size === 'compact' ? compactTriangleStyle : triangleStyle;
    if (viewTab === 'json') {
        return (jsx_runtime_1.jsx("div", { children: zodValidationResult.error.issues.map((error) => {
                return (jsx_runtime_1.jsxs("div", { style: style, children: [
                        jsx_runtime_1.jsx(ValidationMessage_1.WarningTriangle, { style: finalTriangleStyle }), jsx_runtime_1.jsx(layout_1.Spacing, { x: size === 'compact' ? 0.75 : 1 }), error.path.length === 0 ? 'Root' : error.path.join('.'), ":", ' ', error.message] }, error.path.join('.')));
            }) }));
    }
    return (jsx_runtime_1.jsx("div", { children: zodValidationResult.error.issues.map((error) => {
            return (jsx_runtime_1.jsxs("div", { style: style, children: ["-", ' ', jsx_runtime_1.jsx("code", { style: code, children: error.path.length === 0 ? 'Root' : error.path.join('.') }),
                    ": ", error.message] }, error.path.join('.')));
        }) }));
};
exports.ZodErrorMessages = ZodErrorMessages;
