// @remotion/studio/dist/components/RenderModal/SchemaEditor/ZodNonEditableValue.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ZonNonEditableValue = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const colors_1 = require("../../../helpers/colors");
const Fieldset_1 = require("./Fieldset");
const SchemaLabel_1 = require("./SchemaLabel");
const fullWidth = {
    width: '100%',
};
const emptyLabel = {
    width: '100%',
    color: colors_1.VERY_LIGHT_TEXT,
    fontFamily: 'sans-serif',
    fontSize: 14,
};
const wideEmptyLabel = {
    ...emptyLabel,
    fontSize: 12,
    lineHeight: '28px',
};
const ZonNonEditableValue = ({ jsonPath, label, mayPad }) => {
    return (jsx_runtime_1.jsxs(Fieldset_1.Fieldset, { shouldPad: mayPad, children: [
            jsx_runtime_1.jsx(SchemaLabel_1.SchemaLabel, { handleClick: null, jsonPath: jsonPath, onRemove: null, valid: true, suffix: null }), jsx_runtime_1.jsx("div", { style: fullWidth, children: jsx_runtime_1.jsx("em", { style: wideEmptyLabel, children: label }) })
        ] }));
};
exports.ZonNonEditableValue = ZonNonEditableValue;
