// remotion/dist/cjs/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Config = exports.Experimental = exports.watchStaticFile = exports.MediaPlaybackError = exports.Video = exports.OffthreadVideo = exports.Html5Video = exports.useRemotionEnvironment = exports.usePixelDensity = exports.useDelayRender = exports.useCurrentScale = exports.useCurrentFrame = exports.useBufferState = exports.staticFile = exports.Series = exports.Sequence = exports.registerRoot = exports.prefetch = exports.random = exports.interpolate = exports.assertValidInterpolatePosterizeOption = exports.assertValidInterpolateEasingOption = exports.Loop = exports.interpolateColors = exports.Interactive = exports.Img = exports.getStaticFiles = exports.getRemotionEnvironment = exports.delayRender = exports.continueRender = exports.getInputProps = exports.Composition = exports.CanvasImage = exports.isHtmlInCanvasSupported = exports.HtmlInCanvas = exports.HTML_IN_CANVAS_UNSUPPORTED_MESSAGE = exports.Solid = exports.cancelRender = exports.Html5Audio = exports.Audio = exports.Artifact = exports.createEffect = void 0;
require("./_check-rsc.js");
require("./asset-types.js");
const Clipper_js_1 = require("./Clipper.js");
const Composition_js_1 = require("./Composition.js");
const enable_sequence_stack_traces_js_1 = require("./enable-sequence-stack-traces.js");
const Folder_js_1 = require("./Folder.js");
const is_player_js_1 = require("./is-player.js");
const multiple_versions_warning_js_1 = require("./multiple-versions-warning.js");
const Null_js_1 = require("./Null.js");
const Sequence_js_1 = require("./Sequence.js");
(0, multiple_versions_warning_js_1.checkMultipleRemotionVersions)();
__exportStar(require("./AbsoluteFill.js"), exports);
__exportStar(require("./animated-image/index.js"), exports);
const index_js_1 = require("./effects/index.js");
Object.defineProperty(exports, "createEffect", { enumerable: true, get: function () { return index_js_1.createEffect; } });
const Artifact_js_1 = require("./Artifact.js");
Object.defineProperty(exports, "Artifact", { enumerable: true, get: function () { return Artifact_js_1.Artifact; } });
const index_js_2 = require("./audio/index.js");
Object.defineProperty(exports, "Audio", { enumerable: true, get: function () { return index_js_2.Audio; } });
Object.defineProperty(exports, "Html5Audio", { enumerable: true, get: function () { return index_js_2.Html5Audio; } });
const cancel_render_js_1 = require("./cancel-render.js");
Object.defineProperty(exports, "cancelRender", { enumerable: true, get: function () { return cancel_render_js_1.cancelRender; } });
const Solid_js_1 = require("./effects/Solid.js");
Object.defineProperty(exports, "Solid", { enumerable: true, get: function () { return Solid_js_1.Solid; } });
const HtmlInCanvas_js_1 = require("./HtmlInCanvas.js");
Object.defineProperty(exports, "HTML_IN_CANVAS_UNSUPPORTED_MESSAGE", { enumerable: true, get: function () { return HtmlInCanvas_js_1.HTML_IN_CANVAS_UNSUPPORTED_MESSAGE; } });
Object.defineProperty(exports, "HtmlInCanvas", { enumerable: true, get: function () { return HtmlInCanvas_js_1.HtmlInCanvas; } });
Object.defineProperty(exports, "isHtmlInCanvasSupported", { enumerable: true, get: function () { return HtmlInCanvas_js_1.isHtmlInCanvasSupported; } });
/**
 * @description Renders a static image to a `<canvas>` and applies Remotion effects.
 * @see [Documentation](https://www.remotion.dev/docs/canvasimage)
 */
const index_js_3 = require("./canvas-image/index.js");
Object.defineProperty(exports, "CanvasImage", { enumerable: true, get: function () { return index_js_3.CanvasImage; } });
const Composition_js_2 = require("./Composition.js");
Object.defineProperty(exports, "Composition", { enumerable: true, get: function () { return Composition_js_2.Composition; } });
const input_props_js_1 = require("./config/input-props.js");
Object.defineProperty(exports, "getInputProps", { enumerable: true, get: function () { return input_props_js_1.getInputProps; } });
const delay_render_js_1 = require("./delay-render.js");
Object.defineProperty(exports, "continueRender", { enumerable: true, get: function () { return delay_render_js_1.continueRender; } });
Object.defineProperty(exports, "delayRender", { enumerable: true, get: function () { return delay_render_js_1.delayRender; } });
__exportStar(require("./easing.js"), exports);
__exportStar(require("./Folder.js"), exports);
__exportStar(require("./freeze.js"), exports);
const get_remotion_environment_js_1 = require("./get-remotion-environment.js");
Object.defineProperty(exports, "getRemotionEnvironment", { enumerable: true, get: function () { return get_remotion_environment_js_1.getRemotionEnvironment; } });
const get_static_files_js_1 = require("./get-static-files.js");
Object.defineProperty(exports, "getStaticFiles", { enumerable: true, get: function () { return get_static_files_js_1.getStaticFiles; } });
__exportStar(require("./IFrame.js"), exports);
const Img_js_1 = require("./Img.js");
Object.defineProperty(exports, "Img", { enumerable: true, get: function () { return Img_js_1.Img; } });
const Interactive_js_1 = require("./Interactive.js");
Object.defineProperty(exports, "Interactive", { enumerable: true, get: function () { return Interactive_js_1.Interactive; } });
__exportStar(require("./internals.js"), exports);
const interpolate_colors_js_1 = require("./interpolate-colors.js");
Object.defineProperty(exports, "interpolateColors", { enumerable: true, get: function () { return interpolate_colors_js_1.interpolateColors; } });
const index_js_4 = require("./loop/index.js");
Object.defineProperty(exports, "Loop", { enumerable: true, get: function () { return index_js_4.Loop; } });
const no_react_1 = require("./no-react");
Object.defineProperty(exports, "assertValidInterpolateEasingOption", { enumerable: true, get: function () { return no_react_1.assertValidInterpolateEasingOption; } });
Object.defineProperty(exports, "assertValidInterpolatePosterizeOption", { enumerable: true, get: function () { return no_react_1.assertValidInterpolatePosterizeOption; } });
Object.defineProperty(exports, "interpolate", { enumerable: true, get: function () { return no_react_1.interpolate; } });
Object.defineProperty(exports, "random", { enumerable: true, get: function () { return no_react_1.random; } });
const prefetch_js_1 = require("./prefetch.js");
Object.defineProperty(exports, "prefetch", { enumerable: true, get: function () { return prefetch_js_1.prefetch; } });
const register_root_js_1 = require("./register-root.js");
Object.defineProperty(exports, "registerRoot", { enumerable: true, get: function () { return register_root_js_1.registerRoot; } });
const Sequence_js_2 = require("./Sequence.js");
Object.defineProperty(exports, "Sequence", { enumerable: true, get: function () { return Sequence_js_2.Sequence; } });
const index_js_5 = require("./series/index.js");
Object.defineProperty(exports, "Series", { enumerable: true, get: function () { return index_js_5.Series; } });
__exportStar(require("./spring/index.js"), exports);
const static_file_js_1 = require("./static-file.js");
Object.defineProperty(exports, "staticFile", { enumerable: true, get: function () { return static_file_js_1.staticFile; } });
__exportStar(require("./Still.js"), exports);
const use_buffer_state_1 = require("./use-buffer-state");
Object.defineProperty(exports, "useBufferState", { enumerable: true, get: function () { return use_buffer_state_1.useBufferState; } });
const use_current_frame_js_1 = require("./use-current-frame.js");
Object.defineProperty(exports, "useCurrentFrame", { enumerable: true, get: function () { return use_current_frame_js_1.useCurrentFrame; } });
const use_current_scale_1 = require("./use-current-scale");
Object.defineProperty(exports, "useCurrentScale", { enumerable: true, get: function () { return use_current_scale_1.useCurrentScale; } });
const use_delay_render_1 = require("./use-delay-render");
Object.defineProperty(exports, "useDelayRender", { enumerable: true, get: function () { return use_delay_render_1.useDelayRender; } });
const use_pixel_density_1 = require("./use-pixel-density");
Object.defineProperty(exports, "usePixelDensity", { enumerable: true, get: function () { return use_pixel_density_1.usePixelDensity; } });
const use_remotion_environment_js_1 = require("./use-remotion-environment.js");
Object.defineProperty(exports, "useRemotionEnvironment", { enumerable: true, get: function () { return use_remotion_environment_js_1.useRemotionEnvironment; } });
__exportStar(require("./use-video-config.js"), exports);
__exportStar(require("./version.js"), exports);
__exportStar(require("./video-config.js"), exports);
const index_js_6 = require("./video/index.js");
Object.defineProperty(exports, "Html5Video", { enumerable: true, get: function () { return index_js_6.Html5Video; } });
Object.defineProperty(exports, "OffthreadVideo", { enumerable: true, get: function () { return index_js_6.OffthreadVideo; } });
Object.defineProperty(exports, "Video", { enumerable: true, get: function () { return index_js_6.Video; } });
const MediaPlaybackError_js_1 = require("./video/MediaPlaybackError.js");
Object.defineProperty(exports, "MediaPlaybackError", { enumerable: true, get: function () { return MediaPlaybackError_js_1.MediaPlaybackError; } });
const watch_static_file_js_1 = require("./watch-static-file.js");
Object.defineProperty(exports, "watchStaticFile", { enumerable: true, get: function () { return watch_static_file_js_1.watchStaticFile; } });
exports.Experimental = {
    /**
     * @description This is a special component that will cause Remotion to only partially capture the frame of the video.
     * @see [Documentation](https://www.remotion.dev/docs/clipper)
     */
    Clipper: Clipper_js_1.Clipper,
    /**
     * @description This is a special component, that, when rendered, will skip rendering the frame altogether.
     * @see [Documentation](https://www.remotion.dev/docs/null)
     */
    Null: Null_js_1.Null,
    useIsPlayer: is_player_js_1.useIsPlayer,
};
const proxyObj = {};
exports.Config = new Proxy(proxyObj, {
    get(_, prop) {
        if (prop === 'Bundling' ||
            prop === 'Rendering' ||
            prop === 'Log' ||
            prop === 'Puppeteer' ||
            prop === 'Output') {
            return exports.Config;
        }
        return () => {
            /* eslint-disable no-console */
            console.warn('⚠️  The CLI configuration has been extracted from Remotion Core.');
            console.warn('Update the import from the config file:');
            console.warn();
            console.warn('- Delete:');
            console.warn('import {Config} from "remotion";');
            console.warn('+ Replace:');
            console.warn('import {Config} from "@remotion/cli/config";');
            console.warn();
            console.warn('For more information, see https://www.remotion.dev/docs/4-0-migration.');
            /* eslint-enable no-console */
            process.exit(1);
        };
    },
});
Sequence_js_1.Sequence.displayName = 'Sequence';
(0, enable_sequence_stack_traces_js_1.addSequenceStackTraces)(Sequence_js_1.Sequence);
(0, enable_sequence_stack_traces_js_1.addSequenceStackTraces)(Composition_js_1.Composition);
(0, enable_sequence_stack_traces_js_1.addSequenceStackTraces)(Folder_js_1.Folder);
