// remotion/dist/cjs/internals.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.Internals = void 0;
const react_1 = require("react");
const absolute_src_js_1 = require("./absolute-src.js");
const AudioForPreview_js_1 = require("./audio/AudioForPreview.js");
const shared_audio_tags_js_1 = require("./audio/shared-audio-tags.js");
const use_audio_frame_js_1 = require("./audio/use-audio-frame.js");
const buffering_js_1 = require("./buffering.js");
const calculate_media_duration_js_1 = require("./calculate-media-duration.js");
const CanUseRemotionHooks_js_1 = require("./CanUseRemotionHooks.js");
const composition_render_error_context_js_1 = require("./composition-render-error-context.js");
const CompositionManager_js_1 = require("./CompositionManager.js");
const CompositionManagerContext_js_1 = require("./CompositionManagerContext.js");
const CompositionManagerProvider_js_1 = require("./CompositionManagerProvider.js");
const CSSUtils = __importStar(require("./default-css.js"));
const default_css_js_1 = require("./default-css.js");
const EditorProps_js_1 = require("./EditorProps.js");
const create_effect_js_1 = require("./effects/create-effect.js");
const run_effect_chain_js_1 = require("./effects/run-effect-chain.js");
const use_effect_chain_state_js_1 = require("./effects/use-effect-chain-state.js");
const use_memoized_effects_js_1 = require("./effects/use-memoized-effects.js");
const webgl2_context_error_js_1 = require("./effects/webgl2-context-error.js");
const enable_sequence_stack_traces_js_1 = require("./enable-sequence-stack-traces.js");
const find_props_to_delete_js_1 = require("./find-props-to-delete.js");
const flatten_schema_js_1 = require("./flatten-schema.js");
const get_asset_file_name_js_1 = require("./get-asset-file-name.js");
const get_effective_visual_mode_value_js_1 = require("./get-effective-visual-mode-value.js");
const get_preview_dom_element_js_1 = require("./get-preview-dom-element.js");
const get_remotion_environment_js_1 = require("./get-remotion-environment.js");
const input_props_override_js_1 = require("./input-props-override.js");
const interactivity_schema_js_1 = require("./interactivity-schema.js");
const interpolate_keyframed_status_js_1 = require("./interpolate-keyframed-status.js");
const is_player_js_1 = require("./is-player.js");
const log_level_context_js_1 = require("./log-level-context.js");
const log_js_1 = require("./log.js");
const max_video_cache_size_js_1 = require("./max-video-cache-size.js");
const nonce_js_1 = require("./nonce.js");
const playback_logging_js_1 = require("./playback-logging.js");
const portal_node_js_1 = require("./portal-node.js");
const prefetch_state_js_1 = require("./prefetch-state.js");
const prefetch_js_1 = require("./prefetch.js");
const PremountContext_js_1 = require("./PremountContext.js");
const register_root_js_1 = require("./register-root.js");
const remotion_environment_context_js_1 = require("./remotion-environment-context.js");
const RemotionRoot_js_1 = require("./RemotionRoot.js");
const RenderAssetManager_js_1 = require("./RenderAssetManager.js");
const resolve_video_config_js_1 = require("./resolve-video-config.js");
const ResolveCompositionConfig_js_1 = require("./ResolveCompositionConfig.js");
const sequence_node_path_js_1 = require("./sequence-node-path.js");
const sequence_stack_traces_js_1 = require("./sequence-stack-traces.js");
const Sequence_js_1 = require("./Sequence.js");
const SequenceContext_js_1 = require("./SequenceContext.js");
const SequenceManager_js_1 = require("./SequenceManager.js");
const setup_env_variables_js_1 = require("./setup-env-variables.js");
const TimelinePosition = __importStar(require("./timeline-position-state.js"));
const timeline_position_state_js_1 = require("./timeline-position-state.js");
const TimelineContext_js_1 = require("./TimelineContext.js");
const truthy_js_1 = require("./truthy.js");
const use_current_scale_js_1 = require("./use-current-scale.js");
const use_delay_render_js_1 = require("./use-delay-render.js");
const use_lazy_component_js_1 = require("./use-lazy-component.js");
const use_media_enabled_js_1 = require("./use-media-enabled.js");
const use_media_in_timeline_js_1 = require("./use-media-in-timeline.js");
const use_pixel_density_js_1 = require("./use-pixel-density.js");
const use_schema_js_1 = require("./use-schema.js");
const use_unsafe_video_config_js_1 = require("./use-unsafe-video-config.js");
const use_video_js_1 = require("./use-video.js");
const validate_media_props_js_1 = require("./validate-media-props.js");
const validate_start_from_props_js_1 = require("./validate-start-from-props.js");
const validate_artifact_js_1 = require("./validation/validate-artifact.js");
const validate_composition_id_js_1 = require("./validation/validate-composition-id.js");
const validate_folder_name_js_1 = require("./validation/validate-folder-name.js");
const duration_state_js_1 = require("./video/duration-state.js");
const OffthreadVideo_js_1 = require("./video/OffthreadVideo.js");
const video_fragment_js_1 = require("./video/video-fragment.js");
const VideoForPreview_js_1 = require("./video/VideoForPreview.js");
const volume_position_state_js_1 = require("./volume-position-state.js");
const volume_prop_js_1 = require("./volume-prop.js");
const volume_safeguard_js_1 = require("./volume-safeguard.js");
const watch_static_file_js_1 = require("./watch-static-file.js");
const wrap_remotion_context_js_1 = require("./wrap-remotion-context.js");
// needs to be in core package so gets deduplicated in studio
const compositionSelectorRef = (0, react_1.createRef)();
// Mark them as Internals so use don't assume this is public
// API and are less likely to use it
exports.Internals = {
    MaxMediaCacheSizeContext: max_video_cache_size_js_1.MaxMediaCacheSizeContext,
    useUnsafeVideoConfig: use_unsafe_video_config_js_1.useUnsafeVideoConfig,
    useFrameForVolumeProp: use_audio_frame_js_1.useFrameForVolumeProp,
    useTimelinePosition: TimelinePosition.useTimelinePosition,
    useAbsoluteTimelinePosition: TimelinePosition.useAbsoluteTimelinePosition,
    evaluateVolume: volume_prop_js_1.evaluateVolume,
    getAbsoluteSrc: absolute_src_js_1.getAbsoluteSrc,
    getAssetDisplayName: get_asset_file_name_js_1.getAssetDisplayName,
    Timeline: TimelinePosition,
    validateMediaTrimProps: validate_start_from_props_js_1.validateMediaTrimProps,
    validateMediaProps: validate_media_props_js_1.validateMediaProps,
    resolveTrimProps: validate_start_from_props_js_1.resolveTrimProps,
    VideoForPreview: VideoForPreview_js_1.VideoForPreview,
    CompositionManager: CompositionManagerContext_js_1.CompositionManager,
    CompositionSetters: CompositionManagerContext_js_1.CompositionSetters,
    VisualModePropStatusesContext: SequenceManager_js_1.VisualModePropStatusesContext,
    VisualModePropStatusesRefContext: SequenceManager_js_1.VisualModePropStatusesRefContext,
    VisualModeDragOverridesContext: SequenceManager_js_1.VisualModeDragOverridesContext,
    VisualModeSettersContext: SequenceManager_js_1.VisualModeSettersContext,
    SequenceManager: SequenceManager_js_1.SequenceManager,
    SequenceManagerRefContext: SequenceManager_js_1.SequenceManagerRefContext,
    SequenceStackTracesUpdateContext: sequence_stack_traces_js_1.SequenceStackTracesUpdateContext,
    baseSchema: interactivity_schema_js_1.baseSchema,
    sequenceSchema: interactivity_schema_js_1.sequenceSchema,
    SequenceWithoutSchema: Sequence_js_1.SequenceWithoutSchema,
    sequenceStyleSchema: interactivity_schema_js_1.sequenceStyleSchema,
    sequenceVisualStyleSchema: interactivity_schema_js_1.sequenceVisualStyleSchema,
    sequencePremountSchema: interactivity_schema_js_1.sequencePremountSchema,
    textSchema: interactivity_schema_js_1.textSchema,
    transformSchema: interactivity_schema_js_1.transformSchema,
    premountSchema: interactivity_schema_js_1.premountSchema,
    flattenActiveSchema: flatten_schema_js_1.flattenActiveSchema,
    getFlatSchemaWithAllKeys: flatten_schema_js_1.getFlatSchemaWithAllKeys,
    RemotionRootContexts: RemotionRoot_js_1.RemotionRootContexts,
    CompositionManagerProvider: CompositionManagerProvider_js_1.CompositionManagerProvider,
    useVideo: use_video_js_1.useVideo,
    getRoot: register_root_js_1.getRoot,
    useMediaVolumeState: volume_position_state_js_1.useMediaVolumeState,
    usePlayerMutedState: volume_position_state_js_1.usePlayerMutedState,
    useMediaInTimeline: use_media_in_timeline_js_1.useMediaInTimeline,
    useLazyComponent: use_lazy_component_js_1.useLazyComponent,
    truthy: truthy_js_1.truthy,
    SequenceContext: SequenceContext_js_1.SequenceContext,
    PremountContext: PremountContext_js_1.PremountContext,
    useRemotionContexts: wrap_remotion_context_js_1.useRemotionContexts,
    RemotionContextProvider: wrap_remotion_context_js_1.RemotionContextProvider,
    CSSUtils,
    setupEnvVariables: setup_env_variables_js_1.setupEnvVariables,
    MediaVolumeContext: volume_position_state_js_1.MediaVolumeContext,
    SetMediaVolumeContext: volume_position_state_js_1.SetMediaVolumeContext,
    getRemotionEnvironment: get_remotion_environment_js_1.getRemotionEnvironment,
    SharedAudioContext: shared_audio_tags_js_1.SharedAudioContext,
    SharedAudioContextProvider: shared_audio_tags_js_1.SharedAudioContextProvider,
    SharedAudioTagsContext: shared_audio_tags_js_1.SharedAudioTagsContext,
    SharedAudioTagsContextProvider: shared_audio_tags_js_1.SharedAudioTagsContextProvider,
    invalidCompositionErrorMessage: validate_composition_id_js_1.invalidCompositionErrorMessage,
    invalidFolderNameErrorMessage: validate_folder_name_js_1.invalidFolderNameErrorMessage,
    calculateMediaDuration: calculate_media_duration_js_1.calculateMediaDuration,
    isCompositionIdValid: validate_composition_id_js_1.isCompositionIdValid,
    isFolderNameValid: validate_folder_name_js_1.isFolderNameValid,
    getPreviewDomElement: get_preview_dom_element_js_1.getPreviewDomElement,
    compositionsRef: CompositionManager_js_1.compositionsRef,
    portalNode: portal_node_js_1.portalNode,
    waitForRoot: register_root_js_1.waitForRoot,
    SetTimelineContext: TimelineContext_js_1.SetTimelineContext,
    CanUseRemotionHooksProvider: CanUseRemotionHooks_js_1.CanUseRemotionHooksProvider,
    CanUseRemotionHooks: CanUseRemotionHooks_js_1.CanUseRemotionHooks,
    PrefetchProvider: prefetch_state_js_1.PrefetchProvider,
    DurationsContextProvider: duration_state_js_1.DurationsContextProvider,
    IsPlayerContextProvider: is_player_js_1.IsPlayerContextProvider,
    useIsPlayer: is_player_js_1.useIsPlayer,
    EditorPropsProvider: EditorProps_js_1.EditorPropsProvider,
    EditorPropsContext: EditorProps_js_1.EditorPropsContext,
    usePreload: prefetch_js_1.usePreload,
    NonceContext: nonce_js_1.NonceContext,
    resolveVideoConfig: resolve_video_config_js_1.resolveVideoConfig,
    resolveVideoConfigOrCatch: resolve_video_config_js_1.resolveVideoConfigOrCatch,
    ResolveCompositionContext: ResolveCompositionConfig_js_1.ResolveCompositionContext,
    useResolvedVideoConfig: ResolveCompositionConfig_js_1.useResolvedVideoConfig,
    resolveCompositionsRef: ResolveCompositionConfig_js_1.resolveCompositionsRef,
    REMOTION_STUDIO_CONTAINER_ELEMENT: get_preview_dom_element_js_1.REMOTION_STUDIO_CONTAINER_ELEMENT,
    RenderAssetManager: RenderAssetManager_js_1.RenderAssetManager,
    persistCurrentFrame: timeline_position_state_js_1.persistCurrentFrame,
    usePlaybackRate: timeline_position_state_js_1.usePlaybackRate,
    useTimelineContext: timeline_position_state_js_1.useTimelineContext,
    useTimelineSetFrame: timeline_position_state_js_1.useTimelineSetFrame,
    isIosSafari: video_fragment_js_1.isIosSafari,
    WATCH_REMOTION_STATIC_FILES: watch_static_file_js_1.WATCH_REMOTION_STATIC_FILES,
    addSequenceStackTraces: enable_sequence_stack_traces_js_1.addSequenceStackTraces,
    useMediaStartsAt: use_audio_frame_js_1.useMediaStartsAt,
    BufferingProvider: buffering_js_1.BufferingProvider,
    BufferingContextReact: buffering_js_1.BufferingContextReact,
    getComponentsToAddStacksTo: enable_sequence_stack_traces_js_1.getComponentsToAddStacksTo,
    CurrentScaleContext: use_current_scale_js_1.CurrentScaleContext,
    PixelDensityContext: use_pixel_density_js_1.PixelDensityContext,
    PreviewSizeContext: use_current_scale_js_1.PreviewSizeContext,
    calculateScale: use_current_scale_js_1.calculateScale,
    validateRenderAsset: validate_artifact_js_1.validateRenderAsset,
    Log: log_js_1.Log,
    LogLevelContext: log_level_context_js_1.LogLevelContext,
    useLogLevel: log_level_context_js_1.useLogLevel,
    playbackLogging: playback_logging_js_1.playbackLogging,
    timeValueRef: EditorProps_js_1.timeValueRef,
    compositionSelectorRef,
    RemotionEnvironmentContext: remotion_environment_context_js_1.RemotionEnvironmentContext,
    warnAboutTooHighVolume: volume_safeguard_js_1.warnAboutTooHighVolume,
    AudioForPreview: AudioForPreview_js_1.AudioForPreview,
    OBJECTFIT_CONTAIN_CLASS_NAME: default_css_js_1.OBJECTFIT_CONTAIN_CLASS_NAME,
    InnerOffthreadVideo: OffthreadVideo_js_1.InnerOffthreadVideo,
    useBasicMediaInTimeline: use_media_in_timeline_js_1.useBasicMediaInTimeline,
    getInputPropsOverride: input_props_override_js_1.getInputPropsOverride,
    setInputPropsOverride: input_props_override_js_1.setInputPropsOverride,
    useVideoEnabled: use_media_enabled_js_1.useVideoEnabled,
    useAudioEnabled: use_media_enabled_js_1.useAudioEnabled,
    useIsPlayerBuffering: buffering_js_1.useIsPlayerBuffering,
    TimelinePosition,
    DelayRenderContextType: use_delay_render_js_1.DelayRenderContextType,
    TimelineContext: TimelineContext_js_1.TimelineContext,
    PlaybackRateContext: TimelineContext_js_1.PlaybackRateContext,
    AbsoluteTimeContext: TimelineContext_js_1.AbsoluteTimeContext,
    RenderAssetManagerProvider: RenderAssetManager_js_1.RenderAssetManagerProvider,
    getEffectiveVisualModeValue: get_effective_visual_mode_value_js_1.getEffectiveVisualModeValue,
    CompositionRenderErrorContext: composition_render_error_context_js_1.CompositionRenderErrorContext,
    useEffectChainState: use_effect_chain_state_js_1.useEffectChainState,
    createEffectChainState: run_effect_chain_js_1.createEffectChainState,
    cleanupEffectChainState: run_effect_chain_js_1.cleanupEffectChainState,
    runEffectChain: run_effect_chain_js_1.runEffectChain,
    useMemoizedEffects: use_memoized_effects_js_1.useMemoizedEffects,
    useMemoizedEffectDefinitions: use_memoized_effects_js_1.useMemoizedEffectDefinitions,
    createEffect: create_effect_js_1.createEffect,
    createWebGLContextError: webgl2_context_error_js_1.createWebGLContextError,
    createWebGL2ContextError: webgl2_context_error_js_1.createWebGL2ContextError,
    computeEffectiveSchemaValuesDotNotation: use_schema_js_1.computeEffectiveSchemaValuesDotNotation,
    interpolateKeyframedStatus: interpolate_keyframed_status_js_1.interpolateKeyframedStatus,
    makeStaticDragOverride: use_schema_js_1.makeStaticDragOverride,
    makeKeyframedDragOverride: use_schema_js_1.makeKeyframedDragOverride,
    resolveDragOverrideValue: get_effective_visual_mode_value_js_1.resolveDragOverrideValue,
    getStaticDragOverrideValue: use_schema_js_1.getStaticDragOverrideValue,
    OverrideIdsToNodePathsGettersContext: sequence_node_path_js_1.OverrideIdsToNodePathsGettersContext,
    OverrideIdsToNodePathsSettersContext: sequence_node_path_js_1.OverrideIdsToNodePathsSettersContext,
    findPropsToDelete: find_props_to_delete_js_1.findPropsToDelete,
    makeSequencePropsSubscriptionKey: SequenceManager_js_1.makeSequencePropsSubscriptionKey,
    getPropStatusesCtx: use_memoized_effects_js_1.getPropStatusesCtx,
    getEffectPropStatusesCtx: use_memoized_effects_js_1.getEffectPropStatusesCtx,
    hiddenField: interactivity_schema_js_1.hiddenField,
    durationInFramesField: interactivity_schema_js_1.durationInFramesField,
    freezeField: interactivity_schema_js_1.freezeField,
    fromField: interactivity_schema_js_1.fromField,
};
