// remotion/dist/cjs/interpolate-keyframed-status.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.interpolateKeyframedStatus = void 0;
const bezier_js_1 = require("./bezier.js");
const easing_js_1 = require("./easing.js");
const interpolate_colors_js_1 = require("./interpolate-colors.js");
const interpolate_js_1 = require("./interpolate.js");
const easingToFn = (e) => {
    switch (e.type) {
        case 'linear':
            return easing_js_1.Easing.linear;
        case 'spring':
            return easing_js_1.Easing.spring({
                damping: e.damping,
                mass: e.mass,
                overshootClamping: e.overshootClamping,
                stiffness: e.stiffness,
            });
        case 'bezier':
            return (0, bezier_js_1.bezier)(e.x1, e.y1, e.x2, e.y2);
        default:
            throw new TypeError(`Unsupported easing: ${JSON.stringify(e)}`);
    }
};
const interpolateKeyframedStatus = ({ frame, status, }) => {
    const { keyframes, easing, clamping, interpolationFunction } = status;
    if (keyframes.length === 0) {
        return null;
    }
    const sortedKeyframes = [...keyframes].sort((a, b) => a.frame - b.frame);
    const inputRange = sortedKeyframes.map((k) => k.frame);
    const outputs = sortedKeyframes.map((k) => k.value);
    if (interpolationFunction === 'interpolateColors') {
        if (!outputs.every((v) => typeof v === 'string')) {
            return null;
        }
        if (keyframes.length === 1) {
            return outputs[0];
        }
        try {
            return (0, interpolate_colors_js_1.interpolateColors)(frame, inputRange, outputs, {
                easing: easing.map(easingToFn),
                posterize: status.posterize,
            });
        }
        catch (_a) {
            return null;
        }
    }
    if (interpolationFunction !== 'interpolate') {
        return null;
    }
    try {
        return (0, interpolate_js_1.interpolate)(frame, inputRange, outputs, {
            easing: easing.map(easingToFn),
            extrapolateLeft: clamping.left,
            extrapolateRight: clamping.right,
            posterize: status.posterize,
        });
    }
    catch (_b) {
        return null;
    }
};
exports.interpolateKeyframedStatus = interpolateKeyframedStatus;
