// remotion/dist/cjs/no-react.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.NoReactInternals = exports.random = exports.interpolate = exports.assertValidInterpolatePosterizeOption = exports.assertValidInterpolateEasingOption = void 0;
const interpolate_1 = require("./interpolate");
Object.defineProperty(exports, "assertValidInterpolateEasingOption", { enumerable: true, get: function () { return interpolate_1.assertValidInterpolateEasingOption; } });
Object.defineProperty(exports, "assertValidInterpolatePosterizeOption", { enumerable: true, get: function () { return interpolate_1.assertValidInterpolatePosterizeOption; } });
Object.defineProperty(exports, "interpolate", { enumerable: true, get: function () { return interpolate_1.interpolate; } });
const random_js_1 = require("./random.js");
Object.defineProperty(exports, "random", { enumerable: true, get: function () { return random_js_1.random; } });
const delay_render_1 = require("./delay-render");
const find_props_to_delete_1 = require("./find-props-to-delete");
const input_props_serialization_1 = require("./input-props-serialization");
const input_props_serialization_js_1 = require("./input-props-serialization.js");
const interactivity_schema_1 = require("./interactivity-schema");
const interpolate_colors_1 = require("./interpolate-colors");
const prores_profile_1 = require("./prores-profile");
const scale_value_1 = require("./scale-value");
const truthy_1 = require("./truthy");
const v5_flag_1 = require("./v5-flag");
const validate_frame_1 = require("./validate-frame");
const validate_default_codec_1 = require("./validation/validate-default-codec");
const validate_default_props_1 = require("./validation/validate-default-props");
const validate_dimensions_1 = require("./validation/validate-dimensions");
const validate_duration_in_frames_1 = require("./validation/validate-duration-in-frames");
const validate_fps_1 = require("./validation/validate-fps");
const get_current_time_1 = require("./video/get-current-time");
const offthread_video_source_1 = require("./video/offthread-video-source");
exports.NoReactInternals = {
    processColor: interpolate_colors_1.processColor,
    truthy: truthy_1.truthy,
    validateFps: validate_fps_1.validateFps,
    validateDimension: validate_dimensions_1.validateDimension,
    validateDurationInFrames: validate_duration_in_frames_1.validateDurationInFrames,
    validateDefaultAndInputProps: validate_default_props_1.validateDefaultAndInputProps,
    validateFrame: validate_frame_1.validateFrame,
    serializeJSONWithSpecialTypes: input_props_serialization_1.serializeJSONWithSpecialTypes,
    bundleName: 'bundle.js',
    bundleMapName: 'bundle.js.map',
    deserializeJSONWithSpecialTypes: input_props_serialization_1.deserializeJSONWithSpecialTypes,
    DELAY_RENDER_CALLSTACK_TOKEN: delay_render_1.DELAY_RENDER_CALLSTACK_TOKEN,
    DELAY_RENDER_RETRY_TOKEN: delay_render_1.DELAY_RENDER_RETRY_TOKEN,
    DELAY_RENDER_CLEAR_TOKEN: delay_render_1.DELAY_RENDER_CLEAR_TOKEN,
    DELAY_RENDER_ATTEMPT_TOKEN: delay_render_1.DELAY_RENDER_RETRIES_LEFT,
    getOffthreadVideoSource: offthread_video_source_1.getOffthreadVideoSource,
    getExpectedMediaFrameUncorrected: get_current_time_1.getExpectedMediaFrameUncorrected,
    ENABLE_V5_BREAKING_CHANGES: v5_flag_1.ENABLE_V5_BREAKING_CHANGES,
    MIN_NODE_VERSION: v5_flag_1.ENABLE_V5_BREAKING_CHANGES ? 18 : 16,
    MIN_BUN_VERSION: v5_flag_1.ENABLE_V5_BREAKING_CHANGES ? '1.1.3' : '1.0.3',
    colorNames: interpolate_colors_1.colorNames,
    DATE_TOKEN: input_props_serialization_js_1.DATE_TOKEN,
    FILE_TOKEN: input_props_serialization_js_1.FILE_TOKEN,
    validateCodec: validate_default_codec_1.validateCodec,
    proResProfileOptions: prores_profile_1.proResProfileOptions,
    findPropsToDelete: find_props_to_delete_1.findPropsToDelete,
    sequenceSchema: interactivity_schema_1.sequenceSchema,
    parseScaleValue: scale_value_1.parseScaleValue,
    serializeScaleValue: scale_value_1.serializeScaleValue,
};
