// remotion/dist/cjs/Sequence.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SequenceWithoutFrom = exports.Sequence = exports.SequenceWithoutSchema = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
/* eslint-disable @typescript-eslint/no-use-before-define */
const react_1 = require("react");
const AbsoluteFill_js_1 = require("./AbsoluteFill.js");
const freeze_js_1 = require("./freeze.js");
const interactivity_schema_js_1 = require("./interactivity-schema.js");
const nonce_js_1 = require("./nonce.js");
const PremountContext_js_1 = require("./PremountContext.js");
const SequenceContext_js_1 = require("./SequenceContext.js");
const SequenceManager_js_1 = require("./SequenceManager.js");
const is_inside_series_js_1 = require("./series/is-inside-series.js");
const timeline_position_state_js_1 = require("./timeline-position-state.js");
const use_current_frame_1 = require("./use-current-frame");
const use_remotion_environment_js_1 = require("./use-remotion-environment.js");
const use_video_config_js_1 = require("./use-video-config.js");
const v5_flag_js_1 = require("./v5-flag.js");
const with_interactivity_schema_js_1 = require("./with-interactivity-schema.js");
const EMPTY_EFFECTS = [];
const RegularSequenceRefForwardingFunction = ({ from = 0, trimBefore = 0, freeze, durationInFrames = Infinity, children, name, height, width, showInTimeline = true, hidden = false, controls, _remotionInternalEffects, _remotionInternalLoopDisplay: loopDisplay, _remotionInternalStack: stack, _remotionInternalDocumentationLink: documentationLink, _remotionInternalPremountDisplay: premountDisplay, _remotionInternalPostmountDisplay: postmountDisplay, _remotionInternalIsMedia: isMedia, outlineRef: passedRefForOutline, ...other }, ref) => {
    var _a, _b, _c;
    const { layout = 'absolute-fill' } = other;
    const [id] = (0, react_1.useState)(() => String(Math.random()));
    const parentSequence = (0, react_1.useContext)(SequenceContext_js_1.SequenceContext);
    const { rootId } = (0, timeline_position_state_js_1.useTimelineContext)();
    const cumulatedFrom = parentSequence
        ? parentSequence.cumulatedFrom + parentSequence.relativeFrom
        : 0;
    const nonce = (0, nonce_js_1.useNonce)();
    if (layout !== 'absolute-fill' && layout !== 'none') {
        throw new TypeError(`The layout prop of <Sequence /> expects either "absolute-fill" or "none", but you passed: ${layout}`);
    }
    // @ts-expect-error
    if (layout === 'none' && typeof other.style !== 'undefined') {
        throw new TypeError('If layout="none", you may not pass a style. Passed: ' +
            // @ts-expect-error
            JSON.stringify(other.style));
    }
    if (typeof durationInFrames !== 'number') {
        throw new TypeError(`You passed to durationInFrames an argument of type ${typeof durationInFrames}, but it must be a number.`);
    }
    if (durationInFrames <= 0) {
        throw new TypeError(`durationInFrames must be positive, but got ${durationInFrames}`);
    }
    if (typeof from !== 'number') {
        throw new TypeError(`You passed to the "from" props of your <Sequence> an argument of type ${typeof from}, but it must be a number.`);
    }
    if (!Number.isFinite(from)) {
        throw new TypeError(`The "from" prop of a sequence must be finite, but got ${from}.`);
    }
    if (typeof trimBefore !== 'number') {
        throw new TypeError(`You passed to the "trimBefore" prop of your <Sequence> an argument of type ${typeof trimBefore}, but it must be a number.`);
    }
    if (trimBefore < 0) {
        throw new TypeError(`The "trimBefore" prop of <Sequence /> must be greater than or equal to 0, but got ${trimBefore}.`);
    }
    if (Number.isNaN(trimBefore)) {
        throw new TypeError('The "trimBefore" prop of <Sequence /> must be a real number, but it is NaN.');
    }
    if (!Number.isFinite(trimBefore)) {
        throw new TypeError(`The "trimBefore" prop of <Sequence /> must be finite, but it is ${trimBefore}.`);
    }
    if (typeof freeze !== 'undefined' && freeze !== null) {
        if (typeof freeze !== 'number') {
            throw new TypeError(`The "freeze" prop of <Sequence /> must be a number, but is of type ${typeof freeze}.`);
        }
        if (Number.isNaN(freeze)) {
            throw new TypeError(`The "freeze" prop of <Sequence /> must be a real number, but it is NaN.`);
        }
        if (!Number.isFinite(freeze)) {
            throw new TypeError(`The "freeze" prop of <Sequence /> must be finite, but it is ${freeze}.`);
        }
    }
    const absoluteFrame = (0, timeline_position_state_js_1.useTimelinePosition)();
    const videoConfig = (0, use_video_config_js_1.useVideoConfig)();
    const effectiveRelativeFrom = from - trimBefore;
    const absoluteFrom = ((_a = parentSequence === null || parentSequence === void 0 ? void 0 : parentSequence.absoluteFrom) !== null && _a !== void 0 ? _a : 0) + effectiveRelativeFrom;
    const parentSequenceDuration = parentSequence
        ? Math.min(parentSequence.durationInFrames - effectiveRelativeFrom, durationInFrames)
        : durationInFrames;
    const actualDurationInFrames = Math.max(0, Math.min(videoConfig.durationInFrames - from, parentSequenceDuration));
    const { registerSequence, unregisterSequence } = (0, react_1.useContext)(SequenceManager_js_1.SequenceManager);
    const wrapperRefForOutline = (0, react_1.useRef)(null);
    const refForOutline = other.layout === 'none'
        ? (passedRefForOutline !== null && passedRefForOutline !== void 0 ? passedRefForOutline : null)
        : (passedRefForOutline !== null && passedRefForOutline !== void 0 ? passedRefForOutline : wrapperRefForOutline);
    const premounting = (0, react_1.useMemo)(() => {
        // || is intentional, ?? would not trigger on `false`
        return ((parentSequence === null || parentSequence === void 0 ? void 0 : parentSequence.premounting) ||
            Boolean(other._remotionInternalIsPremounting));
    }, [other._remotionInternalIsPremounting, parentSequence === null || parentSequence === void 0 ? void 0 : parentSequence.premounting]);
    const postmounting = (0, react_1.useMemo)(() => {
        // || is intentional, ?? would not trigger on `false`
        return ((parentSequence === null || parentSequence === void 0 ? void 0 : parentSequence.postmounting) ||
            Boolean(other._remotionInternalIsPostmounting));
    }, [other._remotionInternalIsPostmounting, parentSequence === null || parentSequence === void 0 ? void 0 : parentSequence.postmounting]);
    // `cumulatedNegativeFrom` answers: "How many frames of this media have
    // already elapsed before the first visible frame of this sequence?"
    //
    // This is intentionally based on the effective sequence start, not on adding
    // all negative `from` values. See the asset-calculation tests for:
    // - "Should calculate startFrom correctly with negative offset (Html5Audio)"
    // - "same as above, but with <Sequence from={0}> inbetween"
    // - "same as above, but a positive child offset cancels part of the negative parent offset"
    //
    // In particular, <Sequence from={-20}><Sequence from={10}> should have a
    // 10-frame pre-roll, because the positive child offset cancels part of the
    // negative parent offset. But <Sequence from={10}><Sequence from={-5}>
    // should still trim 5 frames from the media once the parent starts.
    const currentSequenceStart = cumulatedFrom + effectiveRelativeFrom;
    const parentSequenceStart = parentSequence
        ? parentSequence.cumulatedFrom + parentSequence.relativeFrom
        : 0;
    const parentFirstFrame = parentSequence
        ? parentSequenceStart - parentSequence.cumulatedNegativeFrom
        : 0;
    const firstFrame = Math.max(0, parentFirstFrame, currentSequenceStart);
    const cumulatedNegativeFrom = currentSequenceStart - firstFrame;
    const contextValue = (0, react_1.useMemo)(() => {
        var _a, _b, _c;
        return {
            absoluteFrom,
            cumulatedFrom,
            relativeFrom: effectiveRelativeFrom,
            cumulatedNegativeFrom,
            durationInFrames: actualDurationInFrames,
            parentFrom: (_a = parentSequence === null || parentSequence === void 0 ? void 0 : parentSequence.relativeFrom) !== null && _a !== void 0 ? _a : 0,
            id,
            height: (_b = height !== null && height !== void 0 ? height : parentSequence === null || parentSequence === void 0 ? void 0 : parentSequence.height) !== null && _b !== void 0 ? _b : null,
            width: (_c = width !== null && width !== void 0 ? width : parentSequence === null || parentSequence === void 0 ? void 0 : parentSequence.width) !== null && _c !== void 0 ? _c : null,
            premounting,
            postmounting,
            premountDisplay: premountDisplay !== null && premountDisplay !== void 0 ? premountDisplay : null,
            postmountDisplay: postmountDisplay !== null && postmountDisplay !== void 0 ? postmountDisplay : null,
        };
    }, [
        cumulatedFrom,
        absoluteFrom,
        effectiveRelativeFrom,
        actualDurationInFrames,
        parentSequence,
        id,
        height,
        width,
        premounting,
        postmounting,
        premountDisplay,
        postmountDisplay,
        cumulatedNegativeFrom,
    ]);
    const timelineClipName = (0, react_1.useMemo)(() => {
        return name !== null && name !== void 0 ? name : '';
    }, [name]);
    const resolvedDocumentationLink = documentationLink !== null && documentationLink !== void 0 ? documentationLink : 'https://www.remotion.dev/docs/sequence';
    const env = (0, use_remotion_environment_js_1.useRemotionEnvironment)();
    const isInsideSeries = (0, react_1.useContext)(is_inside_series_js_1.IsInsideSeriesContext);
    const inheritedStack = (_b = other === null || other === void 0 ? void 0 : other.stack) !== null && _b !== void 0 ? _b : null;
    // Our assumption: Stack doesnt' change. After we symbolicate we assign it a nodePath
    // and if it changes, it would lead to-remounting of the sequence.
    const stackRef = (0, react_1.useRef)(null);
    stackRef.current = stack !== null && stack !== void 0 ? stack : inheritedStack;
    const registeredFrozenFrame = typeof freeze === 'number' ? freeze : null;
    const registeredTrimBefore = trimBefore === 0 ? null : trimBefore;
    const parentCumulatedNegativeFrom = (_c = parentSequence === null || parentSequence === void 0 ? void 0 : parentSequence.cumulatedNegativeFrom) !== null && _c !== void 0 ? _c : 0;
    const startMediaFrom = isMedia && isMedia.type !== 'image'
        ? isMedia.data.startMediaFrom +
            parentCumulatedNegativeFrom -
            cumulatedNegativeFrom
        : null;
    const mediaFrameAtSequenceZero = isMedia && isMedia.type !== 'image'
        ? isMedia.data.startMediaFrom + parentCumulatedNegativeFrom
        : null;
    const frozenMediaFrame = isMedia && isMedia.type !== 'image' && mediaFrameAtSequenceZero !== null
        ? registeredFrozenFrame === null
            ? null
            : mediaFrameAtSequenceZero +
                (loopDisplay
                    ? registeredFrozenFrame % loopDisplay.durationInFrames
                    : registeredFrozenFrame) *
                    isMedia.data.playbackRate
        : null;
    (0, react_1.useEffect)(() => {
        var _a, _b, _c;
        if (!env.isStudio) {
            return;
        }
        if (isMedia) {
            if (isMedia.type === 'image') {
                registerSequence({
                    type: 'image',
                    controls: controls !== null && controls !== void 0 ? controls : null,
                    effects: _remotionInternalEffects !== null && _remotionInternalEffects !== void 0 ? _remotionInternalEffects : EMPTY_EFFECTS,
                    displayName: timelineClipName,
                    documentationLink: resolvedDocumentationLink,
                    duration: actualDurationInFrames,
                    from,
                    trimBefore: registeredTrimBefore,
                    id,
                    loopDisplay,
                    nonce: nonce.get(),
                    parent: (_a = parentSequence === null || parentSequence === void 0 ? void 0 : parentSequence.id) !== null && _a !== void 0 ? _a : null,
                    postmountDisplay: postmountDisplay !== null && postmountDisplay !== void 0 ? postmountDisplay : null,
                    premountDisplay: premountDisplay !== null && premountDisplay !== void 0 ? premountDisplay : null,
                    rootId,
                    showInTimeline,
                    src: isMedia.src,
                    getStack: () => stackRef.current,
                    refForOutline: refForOutline !== null && refForOutline !== void 0 ? refForOutline : null,
                    isInsideSeries,
                    frozenFrame: registeredFrozenFrame,
                });
            }
            else {
                registerSequence({
                    type: isMedia.type,
                    controls: controls !== null && controls !== void 0 ? controls : null,
                    effects: _remotionInternalEffects !== null && _remotionInternalEffects !== void 0 ? _remotionInternalEffects : EMPTY_EFFECTS,
                    displayName: timelineClipName,
                    documentationLink: resolvedDocumentationLink,
                    doesVolumeChange: isMedia.data.doesVolumeChange,
                    duration: actualDurationInFrames,
                    from,
                    trimBefore: registeredTrimBefore,
                    id,
                    loopDisplay,
                    nonce: nonce.get(),
                    parent: (_b = parentSequence === null || parentSequence === void 0 ? void 0 : parentSequence.id) !== null && _b !== void 0 ? _b : null,
                    playbackRate: isMedia.data.playbackRate,
                    postmountDisplay: postmountDisplay !== null && postmountDisplay !== void 0 ? postmountDisplay : null,
                    premountDisplay: premountDisplay !== null && premountDisplay !== void 0 ? premountDisplay : null,
                    rootId,
                    showInTimeline,
                    src: isMedia.data.src,
                    getStack: () => stackRef.current,
                    startMediaFrom: startMediaFrom !== null && startMediaFrom !== void 0 ? startMediaFrom : isMedia.data.startMediaFrom,
                    volume: isMedia.data.volumes,
                    refForOutline: refForOutline !== null && refForOutline !== void 0 ? refForOutline : null,
                    isInsideSeries,
                    frozenFrame: registeredFrozenFrame,
                    frozenMediaFrame,
                });
            }
            return () => {
                unregisterSequence(id);
            };
        }
        registerSequence({
            from,
            trimBefore: registeredTrimBefore,
            duration: actualDurationInFrames,
            id,
            displayName: timelineClipName,
            documentationLink: resolvedDocumentationLink,
            parent: (_c = parentSequence === null || parentSequence === void 0 ? void 0 : parentSequence.id) !== null && _c !== void 0 ? _c : null,
            type: 'sequence',
            rootId,
            showInTimeline,
            nonce: nonce.get(),
            loopDisplay,
            getStack: () => stackRef.current,
            premountDisplay: premountDisplay !== null && premountDisplay !== void 0 ? premountDisplay : null,
            postmountDisplay: postmountDisplay !== null && postmountDisplay !== void 0 ? postmountDisplay : null,
            controls: controls !== null && controls !== void 0 ? controls : null,
            effects: _remotionInternalEffects !== null && _remotionInternalEffects !== void 0 ? _remotionInternalEffects : EMPTY_EFFECTS,
            refForOutline: refForOutline !== null && refForOutline !== void 0 ? refForOutline : null,
            isInsideSeries,
            frozenFrame: registeredFrozenFrame,
        });
        return () => {
            unregisterSequence(id);
        };
    }, [
        durationInFrames,
        id,
        name,
        registerSequence,
        timelineClipName,
        unregisterSequence,
        parentSequence === null || parentSequence === void 0 ? void 0 : parentSequence.id,
        actualDurationInFrames,
        rootId,
        from,
        trimBefore,
        registeredTrimBefore,
        showInTimeline,
        nonce,
        loopDisplay,
        premountDisplay,
        postmountDisplay,
        env.isStudio,
        controls,
        _remotionInternalEffects,
        isMedia,
        resolvedDocumentationLink,
        refForOutline,
        isInsideSeries,
        registeredFrozenFrame,
        startMediaFrom,
        frozenMediaFrame,
    ]);
    // Ceil to support floats
    // https://github.com/remotion-dev/remotion/issues/2958
    const endThreshold = Math.ceil(cumulatedFrom + from + durationInFrames - 1);
    const content = absoluteFrame < cumulatedFrom + from
        ? null
        : absoluteFrame > endThreshold
            ? null
            : children;
    const frozenContent = content === null || typeof freeze === 'undefined' || freeze === null ? (content) : (jsx_runtime_1.jsx(freeze_js_1.Freeze, { frame: freeze, children: content }));
    const styleIfThere = other.layout === 'none' ? undefined : other.style;
    const sequenceRef = (0, react_1.useCallback)((node) => {
        wrapperRefForOutline.current = node;
        if (typeof ref === 'function') {
            ref(node);
        }
        else if (ref) {
            ref.current = node;
        }
    }, [ref]);
    const defaultStyle = (0, react_1.useMemo)(() => {
        return {
            flexDirection: undefined,
            ...(width ? { width } : {}),
            ...(height ? { height } : {}),
            ...(styleIfThere !== null && styleIfThere !== void 0 ? styleIfThere : {}),
        };
    }, [height, styleIfThere, width]);
    if (ref !== null && layout === 'none') {
        throw new TypeError('It is not supported to pass both a `ref` and `layout="none"` to <Sequence />.');
    }
    if (hidden) {
        return null;
    }
    return (jsx_runtime_1.jsx(SequenceContext_js_1.SequenceContext.Provider, { value: contextValue, children: frozenContent === null ? null : other.layout === 'none' ? (frozenContent) : (jsx_runtime_1.jsx(AbsoluteFill_js_1.AbsoluteFill, { ref: sequenceRef, style: defaultStyle, className: other.className, children: frozenContent })) }));
};
const RegularSequence = (0, react_1.forwardRef)(RegularSequenceRefForwardingFunction);
const PremountedPostmountedSequenceRefForwardingFunction = (props, ref) => {
    const parentPremountContext = (0, react_1.useContext)(PremountContext_js_1.PremountContext);
    const frame = (0, use_current_frame_1.useCurrentFrame)() - parentPremountContext.premountFramesRemaining;
    if (props.layout === 'none') {
        throw new Error('`<Sequence>` with `premountFor` and `postmountFor` props does not support layout="none"');
    }
    const { style: passedStyle, from = 0, durationInFrames = Infinity, premountFor = 0, postmountFor = 0, styleWhilePremounted, styleWhilePostmounted, ...otherProps } = props;
    const endThreshold = Math.ceil(from + durationInFrames - 1);
    const premountingActive = frame < from && frame >= from - premountFor;
    const postmountingActive = frame > endThreshold && frame <= endThreshold + postmountFor;
    // Determine which freeze frame to use
    const freezeFrame = premountingActive
        ? from
        : postmountingActive
            ? from + durationInFrames - 1
            : 0;
    const isFreezingActive = premountingActive || postmountingActive;
    const style = (0, react_1.useMemo)(() => {
        var _a;
        return {
            ...passedStyle,
            opacity: premountingActive || postmountingActive ? 0 : 1,
            pointerEvents: premountingActive || postmountingActive
                ? 'none'
                : ((_a = passedStyle === null || passedStyle === void 0 ? void 0 : passedStyle.pointerEvents) !== null && _a !== void 0 ? _a : undefined),
            ...(premountingActive ? styleWhilePremounted : {}),
            ...(postmountingActive ? styleWhilePostmounted : {}),
        };
    }, [
        passedStyle,
        premountingActive,
        postmountingActive,
        styleWhilePremounted,
        styleWhilePostmounted,
    ]);
    return (jsx_runtime_1.jsx(freeze_js_1.Freeze, { frame: freezeFrame, active: isFreezingActive, children: jsx_runtime_1.jsx(SequenceInner, { ref: ref, from: from, durationInFrames: durationInFrames, style: style, _remotionInternalPremountDisplay: premountFor, _remotionInternalPostmountDisplay: postmountFor, _remotionInternalIsPremounting: premountingActive, _remotionInternalIsPostmounting: postmountingActive, ...otherProps }) }));
};
const PremountedPostmountedSequence = (0, react_1.forwardRef)(PremountedPostmountedSequenceRefForwardingFunction);
const SequenceRefForwardingFunction = (props, ref) => {
    var _a;
    const env = (0, use_remotion_environment_js_1.useRemotionEnvironment)();
    const { fps } = (0, use_video_config_js_1.useVideoConfig)();
    if (props.layout !== 'none' && !env.isRendering) {
        const effectivePremountFor = v5_flag_js_1.ENABLE_V5_BREAKING_CHANGES
            ? ((_a = props.premountFor) !== null && _a !== void 0 ? _a : fps)
            : props.premountFor;
        if (effectivePremountFor || props.postmountFor) {
            return (jsx_runtime_1.jsx(PremountedPostmountedSequence, { ref: ref, ...props, premountFor: effectivePremountFor }));
        }
    }
    return jsx_runtime_1.jsx(RegularSequence, { ...props, ref: ref });
};
const SequenceInner = (0, react_1.forwardRef)(SequenceRefForwardingFunction);
exports.SequenceWithoutSchema = SequenceInner;
/*
 * @description A component that time-shifts its children and wraps them in an absolutely positioned <div>.
 * @see [Documentation](https://www.remotion.dev/docs/sequence)
 */
exports.Sequence = (0, with_interactivity_schema_js_1.withInteractivitySchema)({
    Component: SequenceInner,
    componentName: '<Sequence>',
    componentIdentity: 'dev.remotion.remotion.Sequence',
    schema: interactivity_schema_js_1.sequenceSchema,
    supportsEffects: false,
});
exports.SequenceWithoutFrom = (0, with_interactivity_schema_js_1.withInteractivitySchema)({
    Component: SequenceInner,
    componentName: '<Sequence>',
    componentIdentity: null,
    schema: interactivity_schema_js_1.sequenceSchemaWithoutFrom,
    supportsEffects: false,
});
