// @remotion/timeline-utils/dist/audio-waveform-worker.js
"use strict";
/// <reference lib="webworker" />
Object.defineProperty(exports, "__esModule", { value: true });
const draw_peaks_1 = require("./audio-waveform/draw-peaks");
const load_waveform_peaks_1 = require("./audio-waveform/load-waveform-peaks");
const slice_waveform_peaks_1 = require("./audio-waveform/slice-waveform-peaks");
const loop_display_1 = require("./loop-display");
let canvas = null;
let currentController = null;
let latestRequestId = 0;
const postError = (requestId, error) => {
    const message = error instanceof Error ? error.message : 'Failed to render waveform';
    const payload = {
        type: 'error',
        requestId,
        message,
    };
    self.postMessage(payload);
};
const drawPartialWaveform = (message, peaks) => {
    if (!canvas) {
        return;
    }
    const portionPeaks = (0, slice_waveform_peaks_1.sliceWaveformPeaks)({
        durationInFrames: (0, loop_display_1.shouldTileLoopDisplay)(message.loopDisplay)
            ? message.loopDisplay.durationInFrames
            : message.durationInFrames,
        fps: message.fps,
        peaks,
        playbackRate: message.playbackRate,
        startFrom: message.startFrom,
    });
    if (!(0, loop_display_1.shouldTileLoopDisplay)(message.loopDisplay)) {
        (0, draw_peaks_1.drawBars)({
            canvas,
            peaks: portionPeaks,
            color: 'rgba(255, 255, 255, 0.6)',
            volume: message.volume,
            width: message.width,
        });
        return;
    }
    const loopWidth = (0, loop_display_1.getLoopDisplayWidth)({
        visualizationWidth: message.width,
        loopDisplay: message.loopDisplay,
    });
    const targetCanvas = new OffscreenCanvas(Math.max(1, Math.ceil(loopWidth)), message.height);
    (0, draw_peaks_1.drawBars)({
        canvas: targetCanvas,
        peaks: portionPeaks,
        color: 'rgba(255, 255, 255, 0.6)',
        volume: message.volume,
        width: targetCanvas.width,
    });
    const ctx = canvas.getContext('2d');
    if (!ctx) {
        throw new Error('Failed to get canvas context');
    }
    const pattern = ctx.createPattern(targetCanvas, 'repeat-x');
    if (!pattern) {
        return;
    }
    pattern.setTransform(new DOMMatrix().scaleSelf(loopWidth / targetCanvas.width, 1));
    ctx.clearRect(0, 0, message.width, message.height);
    ctx.fillStyle = pattern;
    ctx.fillRect(0, 0, message.width, message.height);
};
const renderWaveform = async (message) => {
    if (!canvas) {
        postError(message.requestId, new Error('Waveform canvas not initialized'));
        return;
    }
    const controller = new AbortController();
    currentController === null || currentController === void 0 ? void 0 : currentController.abort();
    currentController = controller;
    latestRequestId = message.requestId;
    try {
        canvas.width = message.width;
        canvas.height = message.height;
        const peaks = await (0, load_waveform_peaks_1.loadWaveformPeaks)(message.src, controller.signal, {
            onProgress: ({ peaks: nextPeaks }) => {
                if (controller.signal.aborted ||
                    latestRequestId !== message.requestId) {
                    return;
                }
                drawPartialWaveform(message, nextPeaks);
            },
        });
        if (controller.signal.aborted || latestRequestId !== message.requestId) {
            return;
        }
        drawPartialWaveform(message, peaks);
    }
    catch (error) {
        if (controller.signal.aborted || latestRequestId !== message.requestId) {
            return;
        }
        postError(message.requestId, error);
    }
};
self.addEventListener('message', (event) => {
    const message = event.data;
    if (message.type === 'init') {
        canvas = message.canvas;
        return;
    }
    if (message.type === 'dispose') {
        currentController === null || currentController === void 0 ? void 0 : currentController.abort();
        currentController = null;
        canvas = null;
        return;
    }
    renderWaveform(message).catch((error) => {
        postError(message.requestId, error);
    });
});
