// @remotion/timeline-utils/dist/render-frame-strip.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.fillFrameWhereItFits = exports.fillWithCachedFrames = exports.drawSlot = exports.ensureSlots = exports.calculateTimestampSlots = exports.getDurationOfOneFrame = exports.WEBCODECS_TIMESCALE = void 0;
const frame_database_1 = require("./frame-database");
exports.WEBCODECS_TIMESCALE = 1000000;
const MAX_TIME_DEVIATION = exports.WEBCODECS_TIMESCALE * 0.05;
const getDurationOfOneFrame = ({ visualizationWidth, aspectRatio, segmentDuration, frameHeight, }) => {
    const framesFitInWidthUnrounded = visualizationWidth / (frameHeight * aspectRatio);
    return (segmentDuration / framesFitInWidthUnrounded) * exports.WEBCODECS_TIMESCALE;
};
exports.getDurationOfOneFrame = getDurationOfOneFrame;
const fixRounding = (value) => {
    if (value % 1 >= 0.49999999) {
        return Math.ceil(value);
    }
    return Math.floor(value);
};
const calculateTimestampSlots = ({ visualizationWidth, fromSeconds, segmentDuration, aspectRatio, frameHeight, }) => {
    const framesFitInWidthUnrounded = visualizationWidth / (frameHeight * aspectRatio);
    const framesFitInWidth = Math.ceil(framesFitInWidthUnrounded);
    const durationOfOneFrame = (0, exports.getDurationOfOneFrame)({
        visualizationWidth,
        aspectRatio,
        segmentDuration,
        frameHeight,
    });
    const timestampTargets = [];
    for (let i = 0; i < framesFitInWidth + 1; i++) {
        const target = fromSeconds * exports.WEBCODECS_TIMESCALE + durationOfOneFrame * (i + 0.5);
        const snappedToDuration = (Math.round(fixRounding(target / durationOfOneFrame)) - 1) *
            durationOfOneFrame;
        timestampTargets.push(snappedToDuration);
    }
    return timestampTargets;
};
exports.calculateTimestampSlots = calculateTimestampSlots;
const ensureSlots = ({ filledSlots, naturalWidth, fromSeconds, toSeconds, aspectRatio, frameHeight, }) => {
    const segmentDuration = toSeconds - fromSeconds;
    const timestampTargets = (0, exports.calculateTimestampSlots)({
        visualizationWidth: naturalWidth,
        fromSeconds,
        segmentDuration,
        aspectRatio,
        frameHeight,
    });
    for (const timestamp of timestampTargets) {
        if (!filledSlots.has(timestamp)) {
            filledSlots.set(timestamp, undefined);
        }
    }
};
exports.ensureSlots = ensureSlots;
const drawSlot = ({ frame, ctx, filledSlots, visualizationWidth, timestamp, segmentDuration, fromSeconds, devicePixelRatio, frameHeight, }) => {
    const durationOfOneFrame = (0, exports.getDurationOfOneFrame)({
        visualizationWidth,
        aspectRatio: frame.displayWidth / frame.displayHeight,
        segmentDuration,
        frameHeight,
    });
    const relativeTimestamp = timestamp - fromSeconds * exports.WEBCODECS_TIMESCALE;
    const frameIndex = relativeTimestamp / durationOfOneFrame;
    const thumbnailWidth = frame.displayWidth / devicePixelRatio;
    const left = Math.floor(frameIndex * thumbnailWidth);
    const right = Math.ceil((frameIndex + 1) * thumbnailWidth);
    ctx.drawImage(frame, left, 0, right - left, frame.displayHeight / devicePixelRatio);
    filledSlots.set(timestamp, frame.timestamp);
};
exports.drawSlot = drawSlot;
const fillWithCachedFrames = ({ ctx, naturalWidth, filledSlots, src, segmentDuration, fromSeconds, devicePixelRatio, frameHeight, }) => {
    const prefix = (0, frame_database_1.getFrameDatabaseKeyPrefix)(src);
    const keys = Array.from(frame_database_1.frameDatabase.keys()).filter((k) => k.startsWith(prefix));
    const targets = Array.from(filledSlots.keys());
    for (const timestamp of targets) {
        let bestKey;
        let bestDistance = Infinity;
        for (const key of keys) {
            const distance = Math.abs((0, frame_database_1.getTimestampFromFrameDatabaseKey)(key) - timestamp);
            if (distance < bestDistance) {
                bestDistance = distance;
                bestKey = key;
            }
        }
        if (!bestKey) {
            continue;
        }
        const frame = frame_database_1.frameDatabase.get(bestKey);
        if (!frame) {
            continue;
        }
        const alreadyFilled = filledSlots.get(timestamp);
        if (alreadyFilled &&
            Math.abs(alreadyFilled - timestamp) <=
                Math.abs(frame.frame.timestamp - timestamp)) {
            continue;
        }
        frame.lastUsed = Date.now();
        (0, exports.drawSlot)({
            ctx,
            frame: frame.frame,
            filledSlots,
            visualizationWidth: naturalWidth,
            timestamp,
            segmentDuration,
            fromSeconds,
            devicePixelRatio,
            frameHeight,
        });
    }
};
exports.fillWithCachedFrames = fillWithCachedFrames;
const fillFrameWhereItFits = ({ frame, filledSlots, ctx, visualizationWidth, segmentDuration, fromSeconds, devicePixelRatio, frameHeight, }) => {
    const slots = Array.from(filledSlots.keys());
    for (let i = 0; i < slots.length; i++) {
        const slot = slots[i];
        if (Math.abs(slot - frame.timestamp) > MAX_TIME_DEVIATION) {
            continue;
        }
        const filled = filledSlots.get(slot);
        if (filled &&
            Math.abs(filled - slot) <= Math.abs(filled - frame.timestamp)) {
            continue;
        }
        (0, exports.drawSlot)({
            ctx,
            frame,
            filledSlots,
            visualizationWidth,
            timestamp: slot,
            segmentDuration,
            fromSeconds,
            devicePixelRatio,
            frameHeight,
        });
    }
};
exports.fillFrameWhereItFits = fillFrameWhereItFits;
