// @remotion/studio-server/dist/helpers/get-installed-dependencies.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getInstalledDependencies = exports.getInstalledDependenciesWithVersions = void 0;
const node_fs_1 = __importDefault(require("node:fs"));
const node_path_1 = __importDefault(require("node:path"));
const getInstalledDependenciesWithVersions = (remotionRoot) => {
    var _a, _b, _c, _d;
    const packageJsonFilePath = node_path_1.default.join(remotionRoot, 'package.json');
    const packageJson = JSON.parse(node_fs_1.default.readFileSync(packageJsonFilePath, 'utf-8'));
    return {
        dependencies: (_a = packageJson.dependencies) !== null && _a !== void 0 ? _a : {},
        devDependencies: (_b = packageJson.devDependencies) !== null && _b !== void 0 ? _b : {},
        optionalDependencies: (_c = packageJson.optionalDependencies) !== null && _c !== void 0 ? _c : {},
        peerDependencies: (_d = packageJson.peerDependencies) !== null && _d !== void 0 ? _d : {},
    };
};
exports.getInstalledDependenciesWithVersions = getInstalledDependenciesWithVersions;
const getInstalledDependencies = (remotionRoot) => {
    const deps = (0, exports.getInstalledDependenciesWithVersions)(remotionRoot);
    return {
        dependencies: Object.keys(deps.dependencies),
        devDependencies: Object.keys(deps.devDependencies),
        optionalDependencies: Object.keys(deps.optionalDependencies),
        peerDependencies: Object.keys(deps.peerDependencies),
    };
};
exports.getInstalledDependencies = getInstalledDependencies;
