// @remotion/studio-server/dist/helpers/imports.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.ensureNamedImports = exports.ensureNamedImport = exports.normalizeImportSpacing = exports.insertImportDeclaration = exports.findImportDeclaration = exports.getImportedName = void 0;
const recast = __importStar(require("recast"));
const b = recast.types.builders;
const getImportedName = (specifier) => {
    if (specifier.imported.type === 'Identifier') {
        return specifier.imported.name;
    }
    return specifier.imported.value;
};
exports.getImportedName = getImportedName;
const findImportDeclaration = (ast, sourcePath) => {
    for (const stmt of ast.program.body) {
        if (stmt.type === 'ImportDeclaration' &&
            stmt.source.type === 'StringLiteral' &&
            stmt.source.value === sourcePath) {
            return stmt;
        }
    }
    return null;
};
exports.findImportDeclaration = findImportDeclaration;
const findImportDeclarations = (ast, sourcePath) => {
    return ast.program.body.filter((stmt) => stmt.type === 'ImportDeclaration' &&
        stmt.source.type === 'StringLiteral' &&
        stmt.source.value === sourcePath);
};
const insertImportDeclaration = (ast, importDeclaration) => {
    const { body } = ast.program;
    let lastImportIndex = -1;
    for (let i = 0; i < body.length; i++) {
        if (body[i].type === 'ImportDeclaration') {
            lastImportIndex = i;
        }
    }
    body.splice(lastImportIndex + 1, 0, importDeclaration);
};
exports.insertImportDeclaration = insertImportDeclaration;
const normalizeImportSpacing = (input) => {
    return input.replace(/(import[^\n]*\n)\n+(?=import\b)/g, '$1');
};
exports.normalizeImportSpacing = normalizeImportSpacing;
const hasNamespaceSpecifier = (importDeclaration) => {
    var _a;
    return (_a = importDeclaration.specifiers) === null || _a === void 0 ? void 0 : _a.some((specifier) => specifier.type === 'ImportNamespaceSpecifier');
};
const ensureNamedImport = ({ ast, importedName, sourcePath, localName, }) => {
    var _a, _b;
    var _c, _d;
    const existingImports = findImportDeclarations(ast, sourcePath);
    for (const existingImportDeclaration of existingImports) {
        const matchingSpecifier = (_a = existingImportDeclaration.specifiers) === null || _a === void 0 ? void 0 : _a.find((importSpecifierCandidate) => importSpecifierCandidate.type === 'ImportSpecifier' &&
            (0, exports.getImportedName)(importSpecifierCandidate) === importedName);
        if (matchingSpecifier) {
            return (_c = (_b = matchingSpecifier.local) === null || _b === void 0 ? void 0 : _b.name) !== null && _c !== void 0 ? _c : importedName;
        }
    }
    const existingImport = existingImports.find((candidateImportDeclaration) => !hasNamespaceSpecifier(candidateImportDeclaration));
    if (existingImport) {
        const importSpecifier = b.importSpecifier(b.identifier(importedName), localName === importedName ? null : b.identifier(localName));
        existingImport.specifiers = [
            ...((_d = existingImport.specifiers) !== null && _d !== void 0 ? _d : []),
            importSpecifier,
        ];
        return localName;
    }
    const specifier = b.importSpecifier(b.identifier(importedName), localName === importedName ? null : b.identifier(localName));
    const importDeclaration = b.importDeclaration([specifier], b.stringLiteral(sourcePath));
    (0, exports.insertImportDeclaration)(ast, importDeclaration);
    return localName;
};
exports.ensureNamedImport = ensureNamedImport;
const ensureNamedImports = ({ ast, importedNames, sourcePath, }) => {
    var _a;
    var _b, _c, _d;
    if (importedNames.size === 0) {
        return;
    }
    const existingImports = findImportDeclarations(ast, sourcePath);
    const existingNames = new Set();
    for (const existingImportDeclaration of existingImports) {
        for (const importSpecifierCandidate of (_b = existingImportDeclaration.specifiers) !== null && _b !== void 0 ? _b : []) {
            if (importSpecifierCandidate.type !== 'ImportSpecifier') {
                continue;
            }
            const importedName = (0, exports.getImportedName)(importSpecifierCandidate);
            const localName = (_c = (_a = importSpecifierCandidate.local) === null || _a === void 0 ? void 0 : _a.name) !== null && _c !== void 0 ? _c : importedName;
            if (localName === importedName) {
                existingNames.add(importedName);
            }
        }
    }
    const existingImport = existingImports.find((candidateImportDeclaration) => !hasNamespaceSpecifier(candidateImportDeclaration));
    if (existingImport) {
        for (const importedName of importedNames) {
            if (existingNames.has(importedName)) {
                continue;
            }
            existingImport.specifiers = [
                ...((_d = existingImport.specifiers) !== null && _d !== void 0 ? _d : []),
                b.importSpecifier(b.identifier(importedName)),
            ];
            existingNames.add(importedName);
        }
        return;
    }
    const specifiers = [...importedNames]
        .filter((importedName) => !existingNames.has(importedName))
        .map((importedName) => b.importSpecifier(b.identifier(importedName)));
    if (specifiers.length === 0) {
        return;
    }
    const importDeclaration = b.importDeclaration(specifiers, b.stringLiteral(sourcePath));
    (0, exports.insertImportDeclaration)(ast, importDeclaration);
};
exports.ensureNamedImports = ensureNamedImports;
