// @remotion/studio-server/dist/helpers/import-agnostic-node-path.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.fromImportAgnosticNodePath = exports.toImportAgnosticNodePath = void 0;
const getActualIndex = ({ items, agnosticIndex, shouldIgnore, context, }) => {
    let seenIncludedNodes = 0;
    for (let i = 0; i < items.length; i++) {
        if (shouldIgnore(items[i], context)) {
            continue;
        }
        if (seenIncludedNodes === agnosticIndex) {
            return i;
        }
        seenIncludedNodes++;
    }
    return null;
};
const getAgnosticIndex = ({ items, actualIndex, shouldIgnore, context, }) => {
    if (actualIndex < 0 || actualIndex >= items.length) {
        return null;
    }
    let seenIncludedNodes = 0;
    for (let i = 0; i <= actualIndex; i++) {
        if (shouldIgnore(items[i], context)) {
            continue;
        }
        if (i === actualIndex) {
            return seenIncludedNodes;
        }
        seenIncludedNodes++;
    }
    return null;
};
const getUseCurrentFrameLocalNames = (ast) => {
    var _a;
    var _b;
    const names = new Set(['useCurrentFrame']);
    for (const statement of ast.program.body) {
        if (statement.type !== 'ImportDeclaration' ||
            statement.source.value !== 'remotion') {
            continue;
        }
        for (const specifier of statement.specifiers) {
            if (specifier.type === 'ImportSpecifier' &&
                specifier.imported.type === 'Identifier' &&
                specifier.imported.name === 'useCurrentFrame') {
                names.add((_b = (_a = specifier.local) === null || _a === void 0 ? void 0 : _a.name) !== null && _b !== void 0 ? _b : 'useCurrentFrame');
            }
        }
    }
    return names;
};
const isUseCurrentFrameCall = (node, useCurrentFrameLocalNames) => {
    return ((node === null || node === void 0 ? void 0 : node.type) === 'CallExpression' &&
        node.callee.type === 'Identifier' &&
        useCurrentFrameLocalNames.has(node.callee.name));
};
const isUseCurrentFrameStatement = (node, useCurrentFrameLocalNames) => {
    if (node.type === 'ExpressionStatement') {
        return isUseCurrentFrameCall(node.expression, useCurrentFrameLocalNames);
    }
    if (node.type !== 'VariableDeclaration' || node.declarations.length !== 1) {
        return false;
    }
    return isUseCurrentFrameCall(node.declarations[0].init, useCurrentFrameLocalNames);
};
const getIgnoredNodePredicate = (ast) => {
    const useCurrentFrameLocalNames = getUseCurrentFrameLocalNames(ast);
    return (node, { parent, property }) => {
        if (parent.type === 'Program' &&
            property === 'body' &&
            node.type === 'ImportDeclaration') {
            return true;
        }
        if (parent.type === 'BlockStatement' && property === 'body') {
            return isUseCurrentFrameStatement(node, useCurrentFrameLocalNames);
        }
        return false;
    };
};
const getChild = (value, segment) => {
    var _a, _b;
    if (value === null || value === undefined) {
        return null;
    }
    if (Array.isArray(value)) {
        return (_a = value[segment]) !== null && _a !== void 0 ? _a : null;
    }
    if (typeof value !== 'object') {
        return null;
    }
    return (_b = value[segment]) !== null && _b !== void 0 ? _b : null;
};
const transformNodePathIndices = ({ ast, nodePath, direction, }) => {
    const shouldIgnore = getIgnoredNodePredicate(ast);
    const transformed = [];
    let current = ast;
    let arrayContext = null;
    for (const segment of nodePath) {
        if (typeof segment === 'string') {
            const parent = current;
            current = getChild(current, segment);
            if (Array.isArray(current) &&
                parent &&
                typeof parent === 'object' &&
                'type' in parent) {
                arrayContext = { parent: parent, property: segment };
            }
            else {
                arrayContext = null;
            }
            transformed.push(segment);
            continue;
        }
        if (!Array.isArray(current) || !arrayContext) {
            current = getChild(current, segment);
            transformed.push(segment);
            continue;
        }
        const mappedIndex = direction === 'to-agnostic'
            ? getAgnosticIndex({
                items: current,
                actualIndex: segment,
                shouldIgnore,
                context: arrayContext,
            })
            : getActualIndex({
                items: current,
                agnosticIndex: segment,
                shouldIgnore,
                context: arrayContext,
            });
        if (mappedIndex === null) {
            if (direction === 'from-agnostic') {
                return null;
            }
            current = getChild(current, segment);
            transformed.push(segment);
            continue;
        }
        current = getChild(current, direction === 'to-agnostic' ? segment : mappedIndex);
        arrayContext = null;
        transformed.push(mappedIndex);
    }
    return transformed;
};
const toImportAgnosticNodePath = ({ ast, nodePath, }) => {
    var _a;
    return ((_a = transformNodePathIndices({ ast, nodePath, direction: 'to-agnostic' })) !== null && _a !== void 0 ? _a : nodePath);
};
exports.toImportAgnosticNodePath = toImportAgnosticNodePath;
const fromImportAgnosticNodePath = ({ ast, nodePath, }) => {
    return transformNodePathIndices({ ast, nodePath, direction: 'from-agnostic' });
};
exports.fromImportAgnosticNodePath = fromImportAgnosticNodePath;
