// @remotion/studio-server/dist/codemods/update-sequence-props/update-sequence-props.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.updateSequenceProps = exports.updateMultipleSequenceProps = exports.updateSequencePropsAst = void 0;
const recast = __importStar(require("recast"));
const no_react_1 = require("remotion/no-react");
const can_update_sequence_props_1 = require("../../preview-server/routes/can-update-sequence-props");
const format_file_content_1 = require("../format-file-content");
const parse_ast_1 = require("../parse-ast");
const update_nested_prop_1 = require("../update-nested-prop");
const b = recast.types.builders;
const removeVariantKey = ({ node, variantKey, }) => {
    var _a;
    const dotIndex = variantKey.indexOf('.');
    if (dotIndex === -1) {
        const idx = (_a = node.attributes) === null || _a === void 0 ? void 0 : _a.findIndex((a) => a.type === 'JSXAttribute' &&
            a.name.type === 'JSXIdentifier' &&
            a.name.name === variantKey);
        if (idx !== undefined && idx !== -1 && node.attributes) {
            node.attributes.splice(idx, 1);
        }
        return;
    }
    (0, update_nested_prop_1.updateNestedProp)({
        node,
        parentKey: variantKey.slice(0, dotIndex),
        childKey: variantKey.slice(dotIndex + 1),
        value: undefined,
        defaultValue: null,
        isDefault: true,
    });
};
const snapshotTopLevelAttrs = (node) => {
    var _a;
    const result = new Map();
    for (const a of (_a = node.attributes) !== null && _a !== void 0 ? _a : []) {
        if (a.type === 'JSXAttribute' && a.name.type === 'JSXIdentifier') {
            const { name } = a.name;
            const printed = recast
                .print(a)
                .code.replace(/\s+/g, ' ')
                .replace(/,(\s*[}\]])/g, '$1')
                .trim();
            const prefix = `${name}=`;
            let valueOnly = printed.startsWith(prefix)
                ? printed.slice(prefix.length)
                : printed;
            if (valueOnly.startsWith('{') && valueOnly.endsWith('}')) {
                valueOnly = valueOnly.slice(1, -1).trim();
            }
            result.set(name, valueOnly);
        }
    }
    return result;
};
const updateSequencePropsNode = ({ node, updates, schema, }) => {
    var _a, _b, _c;
    var _d, _e;
    const logLine = (_d = (_a = node.loc) === null || _a === void 0 ? void 0 : _a.start.line) !== null && _d !== void 0 ? _d : 1;
    const oldValueStrings = [];
    const initialAttrs = snapshotTopLevelAttrs(node);
    const updatedTopLevelKeys = new Set(updates.map(({ key }) => {
        const dot = key.indexOf('.');
        return dot === -1 ? key : key.slice(0, dot);
    }));
    for (const { key, value, defaultValue } of updates) {
        let oldValueString = '';
        const isDefault = defaultValue !== null &&
            JSON.stringify(value) === JSON.stringify(defaultValue);
        const dotIndex = key.indexOf('.');
        const isNested = dotIndex !== -1;
        const parentKey = isNested ? key.slice(0, dotIndex) : key;
        const childKey = isNested ? key.slice(dotIndex + 1) : '';
        if (isNested) {
            oldValueString = (0, update_nested_prop_1.updateNestedProp)({
                node,
                parentKey,
                childKey,
                value,
                defaultValue,
                isDefault,
            });
        }
        else {
            const attrIndex = (_b = node.attributes) === null || _b === void 0 ? void 0 : _b.findIndex((a) => {
                if (a.type === 'JSXSpreadAttribute') {
                    return false;
                }
                if (a.name.type === 'JSXNamespacedName') {
                    return false;
                }
                return a.name.name === key;
            });
            const attr = attrIndex !== undefined && attrIndex !== -1
                ? (_c = node.attributes) === null || _c === void 0 ? void 0 : _c[attrIndex]
                : undefined;
            if (attr && attr.type !== 'JSXSpreadAttribute' && attr.value) {
                const printed = recast.print(attr.value).code;
                // Strip JSX expression container braces, e.g. "{30}" -> "30"
                oldValueString =
                    printed.startsWith('{') && printed.endsWith('}')
                        ? printed.slice(1, -1)
                        : printed;
            }
            else if (attr && attr.type !== 'JSXSpreadAttribute' && !attr.value) {
                // JSX shorthand like `loop` (no value) is implicitly `true`
                oldValueString = 'true';
            }
            else if (!attr && defaultValue !== null) {
                oldValueString = JSON.stringify(defaultValue);
            }
            if (isDefault) {
                if (attr && attr.type !== 'JSXSpreadAttribute' && node.attributes) {
                    node.attributes.splice(attrIndex, 1);
                }
            }
            else {
                const parsed = (0, update_nested_prop_1.parseValueExpression)(value);
                const newValue = value === true ? null : b.jsxExpressionContainer(parsed);
                if (!attr || attr.type === 'JSXSpreadAttribute') {
                    const newAttr = b.jsxAttribute(b.jsxIdentifier(key), newValue);
                    if (!node.attributes) {
                        node.attributes = [];
                    }
                    node.attributes.push(newAttr);
                }
                else {
                    attr.value = newValue;
                }
            }
        }
        oldValueStrings.push(oldValueString);
        if (!isNested) {
            const fieldSchema = schema[key];
            if (fieldSchema && fieldSchema.type === 'enum') {
                const propsToDelete = no_react_1.NoReactInternals.findPropsToDelete({
                    schema,
                    key,
                    value,
                });
                for (const propToDelete of propsToDelete) {
                    removeVariantKey({ node, variantKey: propToDelete });
                }
            }
        }
    }
    const finalAttrNames = new Set();
    for (const a of (_e = node.attributes) !== null && _e !== void 0 ? _e : []) {
        if (a.type === 'JSXAttribute' && a.name.type === 'JSXIdentifier') {
            finalAttrNames.add(a.name.name);
        }
    }
    const removedProps = [];
    for (const [name, valueString] of initialAttrs) {
        if (finalAttrNames.has(name) || updatedTopLevelKeys.has(name)) {
            continue;
        }
        removedProps.push({ key: name, valueString });
    }
    return {
        oldValueStrings,
        logLine,
        removedProps,
    };
};
const updateSequencePropsAst = ({ input, nodePath, updates, schema, }) => {
    const ast = (0, parse_ast_1.parseAst)(input);
    const node = (0, can_update_sequence_props_1.findJsxElementAtNodePath)(ast, nodePath);
    if (!node) {
        throw new Error('Could not find a JSX element at the specified line to update');
    }
    const { oldValueStrings, logLine, removedProps } = updateSequencePropsNode({
        node,
        updates,
        schema,
    });
    return {
        serialized: (0, parse_ast_1.serializeAst)(ast),
        oldValueStrings,
        logLine,
        removedProps,
    };
};
exports.updateSequencePropsAst = updateSequencePropsAst;
const updateMultipleSequenceProps = async ({ input, changes, prettierConfigOverride, }) => {
    const ast = (0, parse_ast_1.parseAst)(input);
    const results = changes.map(({ nodePath, updates, schema }) => {
        const node = (0, can_update_sequence_props_1.findJsxElementAtNodePath)(ast, nodePath);
        if (!node) {
            throw new Error('Could not find a JSX element at the specified line to update');
        }
        return updateSequencePropsNode({ node, updates, schema });
    });
    const { output, formatted } = await (0, format_file_content_1.formatFileContent)({
        input: (0, parse_ast_1.serializeAst)(ast),
        prettierConfigOverride,
    });
    return {
        output,
        formatted,
        results,
    };
};
exports.updateMultipleSequenceProps = updateMultipleSequenceProps;
const updateSequenceProps = async ({ input, nodePath, updates, schema, prettierConfigOverride, }) => {
    const { serialized, oldValueStrings, logLine, removedProps } = (0, exports.updateSequencePropsAst)({
        input,
        nodePath,
        updates,
        schema,
    });
    const { output, formatted } = await (0, format_file_content_1.formatFileContent)({
        input: serialized,
        prettierConfigOverride,
    });
    return {
        output,
        oldValueStrings,
        formatted,
        logLine,
        removedProps,
    };
};
exports.updateSequenceProps = updateSequenceProps;
