// @remotion/studio/dist/components/InspectorPanel/KeyframeInspector.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.KeyframeInspector = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const react_1 = require("react");
const remotion_1 = require("remotion");
const timeline_layout_1 = require("../../helpers/timeline-layout");
const is_menu_item_1 = require("../Menu/is-menu-item");
const parse_keyframe_field_from_node_path_1 = require("../Timeline/parse-keyframe-field-from-node-path");
const TimelineEffectPropItem_1 = require("../Timeline/TimelineEffectPropItem");
const TimelineSequencePropItem_1 = require("../Timeline/TimelineSequencePropItem");
const common_1 = require("./common");
const styles_1 = require("./styles");
const use_track_for_selection_1 = require("./use-track-for-selection");
const KeyframeInspector = ({ selection }) => {
    const track = (0, use_track_for_selection_1.useTrackForSelection)(selection);
    const { propStatuses } = (0, react_1.useContext)(remotion_1.Internals.VisualModePropStatusesContext);
    const { getDragOverrides, getEffectDragOverrides } = (0, react_1.useContext)(remotion_1.Internals.VisualModeDragOverridesContext);
    const details = (0, react_1.useMemo)(() => {
        var _a;
        var _b, _c, _d, _e, _f, _g;
        if (!track || !track.sequence.controls) {
            return null;
        }
        const keyframeField = (0, parse_keyframe_field_from_node_path_1.parseKeyframeFieldFromNodePath)(selection.nodePathInfo.auxiliaryKeys);
        if (keyframeField === null) {
            return null;
        }
        const nodePath = selection.nodePathInfo.sequenceSubscriptionKey;
        const { keyframeDisplayOffset } = track;
        const sourceFrame = selection.frame - keyframeDisplayOffset;
        if (keyframeField.type === 'sequence') {
            const sequenceFields = (0, timeline_layout_1.getFieldsToShow)({
                schema: track.sequence.controls.schema,
                currentRuntimeValueDotNotation: track.sequence.controls.currentRuntimeValueDotNotation,
                getDragOverrides,
                propStatuses,
                nodePath,
            });
            const sequenceField = (_b = sequenceFields === null || sequenceFields === void 0 ? void 0 : sequenceFields.find((candidate) => candidate.key === keyframeField.fieldKey)) !== null && _b !== void 0 ? _b : null;
            const sequencePropStatus = (_c = (_a = remotion_1.Internals.getPropStatusesCtx(propStatuses, nodePath)) === null || _a === void 0 ? void 0 : _a[keyframeField.fieldKey]) !== null && _c !== void 0 ? _c : null;
            if (!sequenceField || (sequencePropStatus === null || sequencePropStatus === void 0 ? void 0 : sequencePropStatus.status) !== 'keyframed') {
                return null;
            }
            return {
                type: 'sequence',
                field: sequenceField,
                fieldLabel: (_d = sequenceField.description) !== null && _d !== void 0 ? _d : sequenceField.key,
                fileName: nodePath.absolutePath,
                nodePath,
                propStatus: sequencePropStatus,
                schema: track.sequence.controls.schema,
                sourceFrame,
            };
        }
        const effect = track.sequence.effects[keyframeField.effectIndex];
        if (!effect) {
            return null;
        }
        const effectFields = (0, timeline_layout_1.getEffectFieldsToShow)({
            effect,
            effectIndex: keyframeField.effectIndex,
            nodePath,
            propStatuses,
            getEffectDragOverrides,
        });
        const effectField = (_e = effectFields.find((candidate) => candidate.key === keyframeField.fieldKey)) !== null && _e !== void 0 ? _e : null;
        const effectStatus = remotion_1.Internals.getEffectPropStatusesCtx({
            propStatuses,
            nodePath,
            effectIndex: keyframeField.effectIndex,
        });
        const effectPropStatus = effectStatus.type === 'can-update-effect'
            ? ((_f = effectStatus.props[keyframeField.fieldKey]) !== null && _f !== void 0 ? _f : null)
            : null;
        if (!effectField || (effectPropStatus === null || effectPropStatus === void 0 ? void 0 : effectPropStatus.status) !== 'keyframed') {
            return null;
        }
        return {
            type: 'effect',
            field: effectField,
            fieldLabel: (_g = effectField.description) !== null && _g !== void 0 ? _g : effectField.key,
            nodePath,
            sourceFrame,
            validatedLocation: {
                source: nodePath.absolutePath,
                line: 1,
                column: 0,
            },
        };
    }, [
        getDragOverrides,
        getEffectDragOverrides,
        propStatuses,
        selection,
        track,
    ]);
    if (details === null) {
        return jsx_runtime_1.jsx(common_1.InspectorMessage, { children: "Keyframe unavailable" });
    }
    return (jsx_runtime_1.jsxs("div", { style: styles_1.selectedContainer, className: is_menu_item_1.VERTICAL_SCROLLBAR_CLASSNAME, children: [
            jsx_runtime_1.jsx(common_1.InspectorSectionHeader, { children: "Keyframe" }), jsx_runtime_1.jsxs("div", { style: styles_1.detailsContainer, children: [
                    jsx_runtime_1.jsx(common_1.InspectorDetailRow, { label: "Frame", children: selection.frame }), jsx_runtime_1.jsxs("div", { style: styles_1.keyframeEditorRow, children: [
                            jsx_runtime_1.jsx("div", { style: styles_1.keyframeEditorLabel, children: details.fieldLabel }), jsx_runtime_1.jsx("div", { style: styles_1.keyframeEditorValue, children: details.type === 'sequence' ? (jsx_runtime_1.jsx(TimelineSequencePropItem_1.TimelineSequenceKeyframedValue, { field: details.field, fileName: details.fileName, nodePath: details.nodePath, schema: details.schema, propStatus: details.propStatus, sourceFrame: details.sourceFrame })) : (jsx_runtime_1.jsx(TimelineEffectPropItem_1.TimelineEffectPropValue, { field: details.field, nodePath: details.nodePath, validatedLocation: details.validatedLocation, sourceFrame: details.sourceFrame })) })
                        ] })
                ] })
        ] }));
};
exports.KeyframeInspector = KeyframeInspector;
