// @remotion/studio/dist/components/InspectorPanel/SequenceSelectionInspector.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SequenceSelectionInspector = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const react_1 = require("react");
const remotion_1 = require("remotion");
const client_id_1 = require("../../helpers/client-id");
const InspectorSequenceSection_1 = require("../InspectorSequenceSection");
const InspectorSourceLocation_1 = require("../InspectorSourceLocation");
const is_menu_item_1 = require("../Menu/is-menu-item");
const TimelineSelection_1 = require("../Timeline/TimelineSelection");
const use_open_sequence_in_editor_1 = require("../Timeline/use-open-sequence-in-editor");
const common_1 = require("./common");
const styles_1 = require("./styles");
const use_track_for_selection_1 = require("./use-track-for-selection");
const useSequenceDisplayName = (track) => {
    var _a;
    var _b;
    const { propStatuses } = (0, react_1.useContext)(remotion_1.Internals.VisualModePropStatusesContext);
    const nodePath = (_b = (_a = track.nodePathInfo) === null || _a === void 0 ? void 0 : _a.sequenceSubscriptionKey) !== null && _b !== void 0 ? _b : null;
    return (0, react_1.useMemo)(() => {
        const propStatusesForOverride = nodePath
            ? remotion_1.Internals.getPropStatusesCtx(propStatuses, nodePath)
            : undefined;
        const codeNameStatus = propStatusesForOverride === null || propStatusesForOverride === void 0 ? void 0 : propStatusesForOverride.name;
        const displayName = codeNameStatus &&
            codeNameStatus.status === 'static' &&
            typeof codeNameStatus.codeValue === 'string'
            ? codeNameStatus.codeValue
            : track.sequence.displayName;
        return displayName || '<Sequence>';
    }, [nodePath, propStatuses, track.sequence.displayName]);
};
const SequenceExpandedInspector = ({ track }) => {
    const { previewServerState } = (0, react_1.useContext)(client_id_1.StudioServerConnectionCtx);
    const { selectedItems, selectItems } = (0, TimelineSelection_1.useTimelineSelection)();
    const { canOpenInEditor, openInEditor, originalLocation } = (0, use_open_sequence_in_editor_1.useOpenSequenceInEditor)(track.sequence);
    const sequenceDisplayName = useSequenceDisplayName(track);
    const { documentationLink } = track.sequence;
    const validatedLocation = (0, react_1.useMemo)(() => {
        var _a;
        if (!originalLocation ||
            !originalLocation.source ||
            !originalLocation.line) {
            return null;
        }
        return {
            source: originalLocation.source,
            line: originalLocation.line,
            column: (_a = originalLocation.column) !== null && _a !== void 0 ? _a : 0,
        };
    }, [originalLocation]);
    const openFileLocation = (0, react_1.useCallback)(() => {
        if (!canOpenInEditor) {
            return;
        }
        openInEditor();
    }, [canOpenInEditor, openInEditor]);
    const openDocumentationLink = (0, react_1.useCallback)(() => {
        if (!documentationLink) {
            return;
        }
        window.open(documentationLink, '_blank', 'noopener,noreferrer');
    }, [documentationLink]);
    const subtitleStyle = (0, react_1.useMemo)(() => {
        return {
            ...styles_1.sequenceHeaderSubtitle,
            cursor: documentationLink ? 'pointer' : 'default',
        };
    }, [documentationLink]);
    const sequenceSelection = (0, react_1.useMemo)(() => {
        if (!track.nodePathInfo) {
            return null;
        }
        return {
            type: 'sequence',
            nodePathInfo: track.nodePathInfo,
        };
    }, [track.nodePathInfo]);
    const sequenceSelected = (0, react_1.useMemo)(() => {
        if (sequenceSelection === null || selectedItems.length !== 1) {
            return false;
        }
        return ((0, TimelineSelection_1.getTimelineSelectionKey)(selectedItems[0]) ===
            (0, TimelineSelection_1.getTimelineSelectionKey)(sequenceSelection));
    }, [selectedItems, sequenceSelection]);
    const selectSequenceOnInspectorPointerDown = (0, react_1.useCallback)((event) => {
        if (event.button !== 0 || sequenceSelection === null) {
            return;
        }
        event.stopPropagation();
        if (sequenceSelected) {
            return;
        }
        selectItems([sequenceSelection], { reveal: true });
    }, [selectItems, sequenceSelected, sequenceSelection]);
    if (previewServerState.type !== 'connected') {
        return jsx_runtime_1.jsx(common_1.InspectorMessage, { children: "Studio server disconnected" });
    }
    if (!track.nodePathInfo ||
        !(0, InspectorSequenceSection_1.hasSequenceControls)(track.sequence) ||
        !validatedLocation) {
        return jsx_runtime_1.jsx(common_1.InspectorMessage, { children: "Sequence inspector unavailable" });
    }
    return (jsx_runtime_1.jsxs("div", { style: styles_1.selectedContainer, className: is_menu_item_1.VERTICAL_SCROLLBAR_CLASSNAME, onPointerDown: selectSequenceOnInspectorPointerDown, children: [
            jsx_runtime_1.jsxs("div", { style: styles_1.sequenceHeader, children: [
                    jsx_runtime_1.jsx("div", { style: styles_1.sequenceHeaderTitle, children: sequenceDisplayName }), documentationLink ? (jsx_runtime_1.jsx("button", { type: "button", style: subtitleStyle, title: "Open component docs", onClick: openDocumentationLink, children: track.sequence.controls.componentName })) : (jsx_runtime_1.jsx("div", { style: subtitleStyle, children: track.sequence.controls.componentName })), jsx_runtime_1.jsx(InspectorSourceLocation_1.InspectorSourceLocation, { location: validatedLocation, canOpen: canOpenInEditor, onOpen: openFileLocation })
                ] }), jsx_runtime_1.jsx(InspectorSequenceSection_1.InspectorSequenceSection, { sequence: track.sequence, validatedLocation: validatedLocation, nodePathInfo: track.nodePathInfo, keyframeDisplayOffset: track.keyframeDisplayOffset, renderSectionHeader: (children) => (jsx_runtime_1.jsx(common_1.InspectorSectionHeader, { children: children })) })] }));
};
const SequenceSelectionInspector = ({ selection }) => {
    const track = (0, use_track_for_selection_1.useTrackForSelection)(selection);
    if (!track) {
        return jsx_runtime_1.jsx(common_1.InspectorMessage, { children: "Sequence inspector unavailable" });
    }
    const stackKey = track.sequence.getStack();
    return (jsx_runtime_1.jsx(SequenceExpandedInspector, { track: track }, stackKey !== null && stackKey !== void 0 ? stackKey : track.sequence.id));
};
exports.SequenceSelectionInspector = SequenceSelectionInspector;
