// @remotion/timeline-utils/dist/esm/index.mjs
// src/audio-waveform/constants.ts
var TARGET_SAMPLE_RATE = 100;
// src/audio-waveform/parse-color.ts
var colorCache = new Map;
var parseColor = (color) => {
  const cached = colorCache.get(color);
  if (cached)
    return cached;
  const ctx = new OffscreenCanvas(1, 1).getContext("2d");
  ctx.fillStyle = color;
  ctx.fillRect(0, 0, 1, 1);
  const [r, g, b, a] = ctx.getImageData(0, 0, 1, 1).data;
  const result = [r, g, b, a];
  colorCache.set(color, result);
  return result;
};

// src/audio-waveform/draw-peaks.ts
var CLIPPING_COLOR = "#FF7F50";
var getVolumeAtBar = ({
  barIndex,
  numBars,
  volume
}) => {
  if (typeof volume === "number") {
    return volume;
  }
  if (volume.length === 0) {
    return 1;
  }
  if (volume.length === 1 || numBars <= 1) {
    return volume[0];
  }
  const volumeIndex = Math.round(barIndex / (numBars - 1) * (volume.length - 1));
  return volume[volumeIndex] ?? 1;
};
var drawBars = ({
  canvas,
  color,
  peaks,
  volume,
  width
}) => {
  const ctx = canvas.getContext("2d");
  if (!ctx) {
    throw new Error("Failed to get canvas context");
  }
  const { height } = canvas;
  const w = canvas.width;
  if (w === 0 || height === 0) {
    return;
  }
  ctx.clearRect(0, 0, w, height);
  const [r, g, b, a] = parseColor(color);
  const [cr, cg, cb, ca] = parseColor(CLIPPING_COLOR);
  const imageData = ctx.createImageData(w, height);
  const { data } = imageData;
  const numBars = width;
  for (let barIndex = 0;barIndex < numBars; barIndex++) {
    const x = barIndex;
    if (x >= w)
      break;
    const peakIndex = Math.floor(barIndex / numBars * peaks.length);
    const peak = peaks[peakIndex] || 0;
    const barVolume = getVolumeAtBar({ barIndex, numBars, volume });
    const scaledPeak = peak * barVolume;
    const halfBar = Math.max(0, Math.min(height / 2, scaledPeak * height / 2));
    if (halfBar === 0)
      continue;
    const mid = height / 2;
    const barY = Math.round(mid - halfBar);
    const barEnd = Math.round(mid + halfBar);
    const isClipping = scaledPeak > 1;
    const clipTopEnd = isClipping ? Math.min(barY + 2, barEnd) : barY;
    const clipBotStart = isClipping ? Math.max(barEnd - 2, barY) : barEnd;
    for (let y = barY;y < clipTopEnd; y++) {
      const idx = (y * w + x) * 4;
      data[idx] = cr;
      data[idx + 1] = cg;
      data[idx + 2] = cb;
      data[idx + 3] = ca;
    }
    for (let y = clipTopEnd;y < clipBotStart; y++) {
      const idx = (y * w + x) * 4;
      data[idx] = r;
      data[idx + 1] = g;
      data[idx + 2] = b;
      data[idx + 3] = a;
    }
    for (let y = clipBotStart;y < barEnd; y++) {
      const idx = (y * w + x) * 4;
      data[idx] = cr;
      data[idx + 1] = cg;
      data[idx + 2] = cb;
      data[idx + 3] = ca;
    }
  }
  ctx.putImageData(imageData, 0, 0);
};
// src/audio-waveform/load-waveform-peaks.ts
import { ALL_FORMATS, AudioSampleSink, Input, UrlSource } from "mediabunny";

// src/audio-waveform/trim-audio-sample-before-zero.ts
var getAudioSampleStartFrameAtTimelineZero = (sample) => {
  if (sample.timestamp + sample.duration <= 0) {
    return null;
  }
  if (sample.timestamp >= 0) {
    return 0;
  }
  return Math.min(sample.numberOfFrames, Math.ceil(-sample.timestamp * sample.sampleRate));
};

// src/audio-waveform/waveform-peak-processor.ts
var emitWaveformProgress = ({
  completedPeaks,
  final,
  onProgress,
  peaks,
  totalPeaks
}) => {
  onProgress?.({
    peaks,
    completedPeaks,
    totalPeaks,
    final
  });
};
var createWaveformPeakProcessor = ({
  totalPeaks,
  samplesPerPeak,
  onProgress,
  progressIntervalInMs,
  now
}) => {
  const peaks = new Float32Array(totalPeaks);
  let peakIndex = 0;
  let peakMax = 0;
  let sampleInPeak = 0;
  let lastProgressAt = 0;
  let lastProgressPeak = 0;
  const emitProgress = (force) => {
    const timestamp = now();
    if (!force && peakIndex === lastProgressPeak && sampleInPeak === 0) {
      return;
    }
    if (!force && timestamp - lastProgressAt < progressIntervalInMs) {
      return;
    }
    lastProgressAt = timestamp;
    lastProgressPeak = peakIndex;
    emitWaveformProgress({
      peaks,
      completedPeaks: peakIndex,
      totalPeaks,
      final: force,
      onProgress
    });
  };
  return {
    peaks,
    processSampleChunk: (floats, channels) => {
      const frameCount = Math.floor(floats.length / Math.max(1, channels));
      for (let frame = 0;frame < frameCount; frame++) {
        let framePeak = 0;
        for (let channel = 0;channel < channels; channel++) {
          const sampleIndex = frame * channels + channel;
          const abs = Math.abs(floats[sampleIndex] ?? 0);
          if (abs > framePeak) {
            framePeak = abs;
          }
        }
        if (framePeak > peakMax) {
          peakMax = framePeak;
        }
        sampleInPeak++;
        if (sampleInPeak >= samplesPerPeak) {
          if (peakIndex < totalPeaks) {
            peaks[peakIndex] = peakMax;
          }
          peakIndex++;
          peakMax = 0;
          sampleInPeak = 0;
        }
      }
      emitProgress(false);
    },
    finalize: () => {
      if (sampleInPeak > 0 && peakIndex < totalPeaks) {
        peaks[peakIndex] = peakMax;
        peakIndex++;
      }
      emitProgress(true);
    }
  };
};

// src/audio-waveform/load-waveform-peaks.ts
var DEFAULT_PROGRESS_INTERVAL_IN_MS = 50;
var peaksCache = new Map;
async function loadWaveformPeaks(url, signal, options) {
  const cached = peaksCache.get(url);
  if (cached) {
    emitWaveformProgress({
      peaks: cached,
      completedPeaks: cached.length,
      totalPeaks: cached.length,
      final: true,
      onProgress: options?.onProgress
    });
    return cached;
  }
  const input = new Input({
    formats: ALL_FORMATS,
    source: new UrlSource(url)
  });
  try {
    const audioTrack = await input.getPrimaryAudioTrack();
    if (!audioTrack) {
      return new Float32Array(0);
    }
    if (await audioTrack.isLive()) {
      throw new Error("Live streams are not currently supported by Remotion. Sorry! Source: " + url);
    }
    if (await audioTrack.isRelativeToUnixEpoch()) {
      throw new Error("Streams with UNIX timestamps are not currently supported by Remotion. Sorry! Source: " + url);
    }
    const sampleRate = await audioTrack.getSampleRate();
    const durationInSeconds = await audioTrack.getDurationFromMetadata({ skipLiveWait: true }) ?? await audioTrack.computeDuration({ skipLiveWait: true });
    const totalPeaks = Math.ceil(durationInSeconds * TARGET_SAMPLE_RATE);
    const samplesPerPeak = Math.max(1, Math.floor(sampleRate / TARGET_SAMPLE_RATE));
    const sink = new AudioSampleSink(audioTrack);
    const processor = createWaveformPeakProcessor({
      totalPeaks,
      samplesPerPeak,
      onProgress: options?.onProgress,
      progressIntervalInMs: options?.progressIntervalInMs ?? DEFAULT_PROGRESS_INTERVAL_IN_MS,
      now: () => Date.now()
    });
    for await (const sample of sink.samples()) {
      if (signal.aborted) {
        sample.close();
        return new Float32Array(0);
      }
      const startFrame = getAudioSampleStartFrameAtTimelineZero(sample);
      if (startFrame === null) {
        sample.close();
        continue;
      }
      const frameCount = sample.numberOfFrames - startFrame;
      if (frameCount <= 0) {
        sample.close();
        continue;
      }
      const bytesNeeded = sample.allocationSize({
        format: "f32",
        planeIndex: 0,
        frameOffset: startFrame,
        frameCount
      });
      const floats = new Float32Array(bytesNeeded / 4);
      sample.copyTo(floats, {
        format: "f32",
        planeIndex: 0,
        frameOffset: startFrame,
        frameCount
      });
      const channels = Math.max(1, sample.numberOfChannels);
      sample.close();
      processor.processSampleChunk(floats, channels);
    }
    processor.finalize();
    const { peaks } = processor;
    peaksCache.set(url, peaks);
    return peaks;
  } finally {
    input.dispose();
  }
}
// src/audio-waveform/make-audio-waveform-worker.ts
var makeAudioWaveformWorker = () => {
  return new Worker(new URL("./audio-waveform-worker.mjs", import.meta.url), {
    type: "module"
  });
};
// src/audio-waveform/slice-waveform-peaks.ts
var sliceWaveformPeaks = ({
  durationInFrames,
  fps,
  peaks,
  playbackRate,
  startFrom
}) => {
  if (peaks.length === 0) {
    return peaks;
  }
  const startTimeInSeconds = startFrom / fps;
  const durationInSeconds = durationInFrames / fps * playbackRate;
  const startPeakIndex = Math.floor(startTimeInSeconds * TARGET_SAMPLE_RATE);
  const endPeakIndex = Math.ceil((startTimeInSeconds + durationInSeconds) * TARGET_SAMPLE_RATE);
  return peaks.subarray(Math.max(0, startPeakIndex), Math.min(peaks.length, endPeakIndex));
};
// src/image-thumbnail/get-scaled-image-thumbnail-dimensions.ts
var getScaledImageThumbnailDimensions = ({
  naturalWidth,
  naturalHeight,
  canvasHeight
}) => {
  if (naturalWidth === 0 || naturalHeight === 0 || canvasHeight === 0) {
    return {
      scaledWidth: 0,
      scaledHeight: 0
    };
  }
  const scale = canvasHeight / naturalHeight;
  return {
    scaledWidth: naturalWidth * scale,
    scaledHeight: canvasHeight
  };
};

// src/image-thumbnail/draw-repeating-image-thumbnail.ts
var createOffscreenCanvas = (width, height) => {
  if (typeof document !== "undefined") {
    const canvas = document.createElement("canvas");
    canvas.width = width;
    canvas.height = height;
    return canvas;
  }
  if (typeof OffscreenCanvas !== "undefined") {
    return new OffscreenCanvas(width, height);
  }
  throw new Error("No canvas implementation available");
};
var drawRepeatingImageThumbnail = ({
  canvas,
  image
}) => {
  const ctx = canvas.getContext("2d");
  if (!ctx) {
    throw new Error("Failed to get canvas context");
  }
  const { width, height } = canvas;
  if (width === 0 || height === 0) {
    return;
  }
  const { scaledWidth, scaledHeight } = getScaledImageThumbnailDimensions({
    naturalWidth: image.naturalWidth,
    naturalHeight: image.naturalHeight,
    canvasHeight: height
  });
  if (scaledWidth === 0 || scaledHeight === 0) {
    return;
  }
  const offscreen = createOffscreenCanvas(scaledWidth, scaledHeight);
  const offCtx = offscreen.getContext("2d");
  if (!offCtx) {
    throw new Error("Failed to get offscreen canvas context");
  }
  offCtx.drawImage(image, 0, 0, scaledWidth, scaledHeight);
  const pattern = ctx.createPattern(offscreen, "repeat-x");
  if (!pattern) {
    return;
  }
  ctx.fillStyle = pattern;
  ctx.fillRect(0, 0, width, height);
};
// src/extract-frames.ts
import {
  ALL_FORMATS as ALL_FORMATS2,
  Input as Input2,
  InputDisposedError,
  UrlSource as UrlSource2,
  VideoSampleSink
} from "mediabunny";

// src/get-duration-or-compute.ts
var getDurationOrCompute = async (input) => {
  return await input.getDurationFromMetadata(undefined, {
    skipLiveWait: true
  }) ?? input.computeDuration(undefined, { skipLiveWait: true });
};

// src/extract-frames.ts
async function extractFrames({
  src,
  timestampsInSeconds,
  onVideoSample,
  signal
}) {
  const input = new Input2({
    formats: ALL_FORMATS2,
    source: new UrlSource2(src)
  });
  const dispose = () => {
    input.dispose();
  };
  if (signal) {
    signal.addEventListener("abort", dispose, { once: true });
  }
  try {
    const [durationInSeconds, format, videoTrack] = await Promise.all([
      getDurationOrCompute(input),
      input.getFormat(),
      input.getPrimaryVideoTrack()
    ]);
    if (!videoTrack) {
      throw new Error("No video track found in the input");
    }
    if (await videoTrack.isLive()) {
      throw new Error("Live streams are not currently supported by Remotion. Sorry! Source: " + src);
    }
    if (await videoTrack.isRelativeToUnixEpoch()) {
      throw new Error("Streams with UNIX timestamps are not currently supported by Remotion. Sorry! Source: " + src);
    }
    const timestamps = typeof timestampsInSeconds === "function" ? await timestampsInSeconds({
      track: {
        width: await videoTrack.getDisplayWidth(),
        height: await videoTrack.getDisplayHeight()
      },
      container: format.name,
      durationInSeconds
    }) : timestampsInSeconds;
    if (timestamps.length === 0) {
      return;
    }
    const sink = new VideoSampleSink(videoTrack);
    const sampleIterator = sink.samplesAtTimestamps(timestamps);
    try {
      for await (const videoSample of sampleIterator) {
        if (signal?.aborted) {
          videoSample?.close();
          break;
        }
        if (!videoSample) {
          continue;
        }
        onVideoSample(videoSample);
      }
    } finally {
      try {
        await sampleIterator.return?.();
      } catch {}
    }
  } catch (error) {
    if (error instanceof InputDisposedError) {
      return;
    }
    throw error;
  } finally {
    dispose();
    if (signal) {
      signal.removeEventListener("abort", dispose);
    }
  }
}
// src/frame-database.ts
var KEY_SEPARATOR = "|";
var makeFrameDatabaseKey = (src, timestamp) => `${src}${KEY_SEPARATOR}${timestamp}`;
var getFrameDatabaseKeyPrefix = (src) => {
  return `${src}${KEY_SEPARATOR}`;
};
var MAX_CACHE_SIZE_BYTES = 10 * 1024 * 1024;
var totalCacheSize = 0;
var frameDatabase = new Map;
var aspectRatioCache = new Map;
var getTimestampFromFrameDatabaseKey = (key) => {
  const split = key.split(KEY_SEPARATOR);
  return Number(split[split.length - 1]);
};
var getAspectRatioFromCache = (src) => {
  const cached = aspectRatioCache.get(src);
  if (cached) {
    return cached;
  }
  return null;
};
var evictLRU = () => {
  while (totalCacheSize > MAX_CACHE_SIZE_BYTES && frameDatabase.size > 0) {
    let oldestKey;
    let oldestTime = Infinity;
    for (const [key, candidate] of frameDatabase) {
      if (candidate.lastUsed < oldestTime) {
        oldestTime = candidate.lastUsed;
        oldestKey = key;
      }
    }
    if (!oldestKey) {
      break;
    }
    const entry = frameDatabase.get(oldestKey);
    if (entry) {
      totalCacheSize -= entry.size;
      entry.frame.close();
      frameDatabase.delete(oldestKey);
    }
  }
};
var addFrameToCache = (key, frame) => {
  const existing = frameDatabase.get(key);
  if (existing) {
    totalCacheSize -= existing.size;
    existing.frame.close();
  }
  const size = frame.allocationSize();
  totalCacheSize += size;
  frameDatabase.set(key, {
    frame,
    lastUsed: Date.now(),
    size
  });
  evictLRU();
};
// src/loop-display.ts
var shouldTileLoopDisplay = (loopDisplay) => {
  return loopDisplay !== undefined && loopDisplay.numberOfTimes > 1;
};
var getLoopDisplayWidth = ({
  visualizationWidth,
  loopDisplay
}) => {
  if (!shouldTileLoopDisplay(loopDisplay)) {
    return visualizationWidth;
  }
  return visualizationWidth / loopDisplay.numberOfTimes;
};
// src/render-frame-strip.ts
var WEBCODECS_TIMESCALE = 1e6;
var MAX_TIME_DEVIATION = WEBCODECS_TIMESCALE * 0.05;
var getDurationOfOneFrame = ({
  visualizationWidth,
  aspectRatio,
  segmentDuration,
  frameHeight
}) => {
  const framesFitInWidthUnrounded = visualizationWidth / (frameHeight * aspectRatio);
  return segmentDuration / framesFitInWidthUnrounded * WEBCODECS_TIMESCALE;
};
var fixRounding = (value) => {
  if (value % 1 >= 0.49999999) {
    return Math.ceil(value);
  }
  return Math.floor(value);
};
var calculateTimestampSlots = ({
  visualizationWidth,
  fromSeconds,
  segmentDuration,
  aspectRatio,
  frameHeight
}) => {
  const framesFitInWidthUnrounded = visualizationWidth / (frameHeight * aspectRatio);
  const framesFitInWidth = Math.ceil(framesFitInWidthUnrounded);
  const durationOfOneFrame = getDurationOfOneFrame({
    visualizationWidth,
    aspectRatio,
    segmentDuration,
    frameHeight
  });
  const timestampTargets = [];
  for (let i = 0;i < framesFitInWidth + 1; i++) {
    const target = fromSeconds * WEBCODECS_TIMESCALE + durationOfOneFrame * (i + 0.5);
    const snappedToDuration = (Math.round(fixRounding(target / durationOfOneFrame)) - 1) * durationOfOneFrame;
    timestampTargets.push(snappedToDuration);
  }
  return timestampTargets;
};
var ensureSlots = ({
  filledSlots,
  naturalWidth,
  fromSeconds,
  toSeconds,
  aspectRatio,
  frameHeight
}) => {
  const segmentDuration = toSeconds - fromSeconds;
  const timestampTargets = calculateTimestampSlots({
    visualizationWidth: naturalWidth,
    fromSeconds,
    segmentDuration,
    aspectRatio,
    frameHeight
  });
  for (const timestamp of timestampTargets) {
    if (!filledSlots.has(timestamp)) {
      filledSlots.set(timestamp, undefined);
    }
  }
};
var drawSlot = ({
  frame,
  ctx,
  filledSlots,
  visualizationWidth,
  timestamp,
  segmentDuration,
  fromSeconds,
  devicePixelRatio,
  frameHeight
}) => {
  const durationOfOneFrame = getDurationOfOneFrame({
    visualizationWidth,
    aspectRatio: frame.displayWidth / frame.displayHeight,
    segmentDuration,
    frameHeight
  });
  const relativeTimestamp = timestamp - fromSeconds * WEBCODECS_TIMESCALE;
  const frameIndex = relativeTimestamp / durationOfOneFrame;
  const thumbnailWidth = frame.displayWidth / devicePixelRatio;
  const left = Math.floor(frameIndex * thumbnailWidth);
  const right = Math.ceil((frameIndex + 1) * thumbnailWidth);
  ctx.drawImage(frame, left, 0, right - left, frame.displayHeight / devicePixelRatio);
  filledSlots.set(timestamp, frame.timestamp);
};
var fillWithCachedFrames = ({
  ctx,
  naturalWidth,
  filledSlots,
  src,
  segmentDuration,
  fromSeconds,
  devicePixelRatio,
  frameHeight
}) => {
  const prefix = getFrameDatabaseKeyPrefix(src);
  const keys = Array.from(frameDatabase.keys()).filter((k) => k.startsWith(prefix));
  const targets = Array.from(filledSlots.keys());
  for (const timestamp of targets) {
    let bestKey;
    let bestDistance = Infinity;
    for (const key of keys) {
      const distance = Math.abs(getTimestampFromFrameDatabaseKey(key) - timestamp);
      if (distance < bestDistance) {
        bestDistance = distance;
        bestKey = key;
      }
    }
    if (!bestKey) {
      continue;
    }
    const frame = frameDatabase.get(bestKey);
    if (!frame) {
      continue;
    }
    const alreadyFilled = filledSlots.get(timestamp);
    if (alreadyFilled && Math.abs(alreadyFilled - timestamp) <= Math.abs(frame.frame.timestamp - timestamp)) {
      continue;
    }
    frame.lastUsed = Date.now();
    drawSlot({
      ctx,
      frame: frame.frame,
      filledSlots,
      visualizationWidth: naturalWidth,
      timestamp,
      segmentDuration,
      fromSeconds,
      devicePixelRatio,
      frameHeight
    });
  }
};
var fillFrameWhereItFits = ({
  frame,
  filledSlots,
  ctx,
  visualizationWidth,
  segmentDuration,
  fromSeconds,
  devicePixelRatio,
  frameHeight
}) => {
  const slots = Array.from(filledSlots.keys());
  for (let i = 0;i < slots.length; i++) {
    const slot = slots[i];
    if (Math.abs(slot - frame.timestamp) > MAX_TIME_DEVIATION) {
      continue;
    }
    const filled = filledSlots.get(slot);
    if (filled && Math.abs(filled - slot) <= Math.abs(filled - frame.timestamp)) {
      continue;
    }
    drawSlot({
      ctx,
      frame,
      filledSlots,
      visualizationWidth,
      timestamp: slot,
      segmentDuration,
      fromSeconds,
      devicePixelRatio,
      frameHeight
    });
  }
};
// src/resize-video-frame.ts
var calculateNewDimensionsFromScale = ({
  width,
  height,
  scale
}) => {
  const scaledWidth = Math.round(width * scale);
  const scaledHeight = Math.round(height * scale);
  return {
    width: scaledWidth,
    height: scaledHeight
  };
};
var resizeVideoFrame = ({
  frame,
  scale
}) => {
  if (scale === 1) {
    return frame;
  }
  const { width, height } = calculateNewDimensionsFromScale({
    height: frame.displayHeight,
    width: frame.displayWidth,
    scale
  });
  const canvas = new OffscreenCanvas(width, height);
  const ctx = canvas.getContext("2d");
  if (!ctx) {
    throw new Error("Could not get 2d context");
  }
  canvas.width = width;
  canvas.height = height;
  ctx.scale(scale, scale);
  ctx.drawImage(frame, 0, 0);
  return new VideoFrame(canvas, {
    displayHeight: height,
    displayWidth: width,
    duration: frame.duration ?? undefined,
    timestamp: frame.timestamp
  });
};
export {
  sliceWaveformPeaks,
  shouldTileLoopDisplay,
  resizeVideoFrame,
  makeFrameDatabaseKey,
  makeAudioWaveformWorker,
  loadWaveformPeaks,
  getTimestampFromFrameDatabaseKey,
  getScaledImageThumbnailDimensions,
  getLoopDisplayWidth,
  getFrameDatabaseKeyPrefix,
  getDurationOfOneFrame,
  getAspectRatioFromCache,
  frameDatabase,
  fillWithCachedFrames,
  fillFrameWhereItFits,
  extractFrames,
  ensureSlots,
  emitWaveformProgress,
  drawSlot,
  drawRepeatingImageThumbnail,
  drawBars,
  createWaveformPeakProcessor,
  calculateTimestampSlots,
  aspectRatioCache,
  addFrameToCache,
  WEBCODECS_TIMESCALE,
  TARGET_SAMPLE_RATE
};
