// @remotion/studio/dist/components/RenderModal/DataEditor.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataEditor = exports.useDataEditorWarningVisibility = exports.useDataEditorWarnings = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const react_1 = __importStar(require("react"));
const remotion_1 = require("remotion");
const no_react_1 = require("remotion/no-react");
const client_id_1 = require("../../helpers/client-id");
const colors_1 = require("../../helpers/colors");
const CompactExplanation_1 = require("../CompactExplanation");
const get_zod_if_possible_1 = require("../get-zod-if-possible");
const InspectorPanelLayout_1 = require("../InspectorPanelLayout");
const layout_1 = require("../layout");
const ValidationMessage_1 = require("../NewComposition/ValidationMessage");
const SegmentedControl_1 = require("../SegmentedControl");
const get_render_modal_warnings_1 = require("./get-render-modal-warnings");
const RenderModalJSONPropsEditor_1 = require("./RenderModalJSONPropsEditor");
const SchemaEditor_1 = require("./SchemaEditor/SchemaEditor");
const SchemaErrorMessages_1 = require("./SchemaEditor/SchemaErrorMessages");
const zod_schema_type_1 = require("./SchemaEditor/zod-schema-type");
const WarningIndicatorButton_1 = require("./WarningIndicatorButton");
const errorExplanation = {
    fontSize: 14,
    color: colors_1.LIGHT_TEXT,
    fontFamily: 'sans-serif',
    lineHeight: 1.5,
};
const explainer = {
    display: 'flex',
    flex: 1,
    flexDirection: 'column',
    padding: 12,
    alignItems: 'flex-start',
    textAlign: 'left',
};
const outer = {
    display: 'flex',
    flexDirection: 'column',
    flex: 1,
    overflow: 'hidden',
    backgroundColor: colors_1.BACKGROUND,
};
const inspectorOuter = {
    ...outer,
    flex: 'none',
    overflow: 'visible',
};
const controlContainer = {
    flexDirection: 'column',
    display: 'flex',
    padding: 12,
    borderBottom: `1px solid ${colors_1.BORDER_COLOR}`,
};
const inspectorControlContainer = {
    ...controlContainer,
    borderBottom: 'none',
};
const tabWrapper = {
    display: 'flex',
    marginBottom: '4px',
    flexDirection: 'row',
    alignItems: 'center',
};
const resolveLinkStyle = {
    color: colors_1.BLUE,
    fontSize: 13,
    fontFamily: 'sans-serif',
    textDecoration: 'underline',
    whiteSpace: 'nowrap',
};
const compactResolveLinkStyle = {
    ...resolveLinkStyle,
    fontSize: 12,
};
const persistanceKey = 'remotion.show-render-modalwarning';
const getPersistedShowWarningState = () => {
    const val = localStorage.getItem(persistanceKey);
    if (!val) {
        return true;
    }
    return val === 'true';
};
const setPersistedShowWarningState = (val) => {
    localStorage.setItem(persistanceKey, String(Boolean(val)));
};
const useDataEditorWarnings = ({ canSaveDefaultProps, defaultProps, mode, propsEditType, showCannotSaveDefaultPropsWarning, }) => {
    const inJSONEditor = mode === 'json';
    const serializedJSON = (0, react_1.useMemo)(() => {
        if (!inJSONEditor) {
            return null;
        }
        return no_react_1.NoReactInternals.serializeJSONWithSpecialTypes({
            data: defaultProps,
            indent: 2,
            staticBase: window.remotion_staticBase,
        });
    }, [inJSONEditor, defaultProps]);
    const cliProps = (0, remotion_1.getInputProps)();
    const warnings = (0, react_1.useMemo)(() => {
        return (0, get_render_modal_warnings_1.getRenderModalWarnings)({
            canSaveDefaultProps,
            cliProps,
            isCustomDateUsed: serializedJSON ? serializedJSON.customDateUsed : false,
            customFileUsed: serializedJSON ? serializedJSON.customFileUsed : false,
            inJSONEditor,
            propsEditType,
            jsMapUsed: serializedJSON ? serializedJSON.mapUsed : false,
            jsSetUsed: serializedJSON ? serializedJSON.setUsed : false,
            showCannotSaveDefaultPropsWarning,
        });
    }, [
        cliProps,
        canSaveDefaultProps,
        inJSONEditor,
        propsEditType,
        serializedJSON,
        showCannotSaveDefaultPropsWarning,
    ]);
    return { serializedJSON, warnings };
};
exports.useDataEditorWarnings = useDataEditorWarnings;
const useDataEditorWarningVisibility = () => {
    const [showWarning, setShowWarningWithoutPersistance] = (0, react_1.useState)(() => getPersistedShowWarningState());
    const setShowWarning = (0, react_1.useCallback)((val) => {
        setShowWarningWithoutPersistance((prevVal) => {
            if (typeof val === 'boolean') {
                setPersistedShowWarningState(val);
                return val;
            }
            const nextVal = val(prevVal);
            setPersistedShowWarningState(nextVal);
            return nextVal;
        });
    }, []);
    return { setShowWarning, showWarning };
};
exports.useDataEditorWarningVisibility = useDataEditorWarningVisibility;
const DataEditor = ({ unresolvedComposition, defaultProps, setDefaultProps, propsEditType, canSaveDefaultProps, schemaErrorMode = 'full', layout = 'default', mode: controlledMode, onModeChange, hideModeControls = false, warnings: controlledWarnings, showWarning: controlledShowWarning, setShowWarning: controlledSetShowWarning, hideWarningButton = false, }) => {
    const [internalMode, setInternalMode] = (0, react_1.useState)('schema');
    const mode = controlledMode !== null && controlledMode !== void 0 ? controlledMode : internalMode;
    const setMode = (0, react_1.useCallback)((nextMode) => {
        if (onModeChange) {
            onModeChange(nextMode);
            return;
        }
        setInternalMode(nextMode);
    }, [onModeChange]);
    const { setShowWarning: internalSetShowWarning, showWarning: internalShowWarning, } = (0, exports.useDataEditorWarningVisibility)();
    const showWarning = controlledShowWarning !== null && controlledShowWarning !== void 0 ? controlledShowWarning : internalShowWarning;
    const setShowWarning = controlledSetShowWarning !== null && controlledSetShowWarning !== void 0 ? controlledSetShowWarning : internalSetShowWarning;
    const { serializedJSON, warnings: computedWarnings } = (0, exports.useDataEditorWarnings)({
        canSaveDefaultProps,
        defaultProps,
        mode,
        propsEditType,
        showCannotSaveDefaultPropsWarning: true,
    });
    const warnings = controlledWarnings !== null && controlledWarnings !== void 0 ? controlledWarnings : computedWarnings;
    const jsonEditorSetValue = (0, react_1.useCallback)((newProps) => {
        setDefaultProps(typeof newProps === 'function' ? newProps : () => newProps, { shouldSave: false });
    }, [setDefaultProps]);
    const onSave = (0, react_1.useCallback)(() => {
        setDefaultProps((p) => p, { shouldSave: true });
    }, [setDefaultProps]);
    const z = (0, get_zod_if_possible_1.useZodIfPossible)();
    const schema = (0, react_1.useMemo)(() => {
        if (!z) {
            return 'no-zod';
        }
        if (!unresolvedComposition.schema) {
            return 'no-schema';
        }
        if (!(typeof unresolvedComposition.schema
            .safeParse === 'function')) {
            throw new Error('A value which is not a Zod schema was passed to `schema`');
        }
        return unresolvedComposition.schema;
    }, [unresolvedComposition.schema, z]);
    const zodValidationResult = (0, react_1.useMemo)(() => {
        if (schema === 'no-zod') {
            return 'no-zod';
        }
        if (schema === 'no-schema') {
            return 'no-schema';
        }
        return (0, zod_schema_type_1.zodSafeParse)(schema, defaultProps);
    }, [defaultProps, schema]);
    const { previewServerState } = (0, react_1.useContext)(client_id_1.StudioServerConnectionCtx);
    const modeItems = (0, react_1.useMemo)(() => {
        return [
            {
                key: 'schema',
                label: 'Schema',
                onClick: () => {
                    setMode('schema');
                },
                selected: mode === 'schema',
            },
            {
                key: 'json',
                label: 'JSON',
                onClick: () => {
                    setMode('json');
                },
                selected: mode === 'json',
            },
        ];
    }, [mode, setMode]);
    const connectionStatus = previewServerState.type;
    if (connectionStatus === 'disconnected') {
        const message = 'The studio server has disconnected. Reconnect to edit the schema.';
        if (schemaErrorMode === 'compact') {
            return jsx_runtime_1.jsx(CompactExplanation_1.CompactExplanation, { children: message });
        }
        return (jsx_runtime_1.jsxs("div", { style: explainer, children: [
                jsx_runtime_1.jsx(layout_1.Spacing, { y: 5 }), jsx_runtime_1.jsx("div", { style: errorExplanation, children: message }), jsx_runtime_1.jsx(layout_1.Spacing, { y: 2, block: true })
            ] }));
    }
    if (schema === 'no-zod') {
        return jsx_runtime_1.jsx(SchemaErrorMessages_1.ZodNotInstalled, { mode: schemaErrorMode });
    }
    if (schema === 'no-schema') {
        return jsx_runtime_1.jsx(SchemaErrorMessages_1.NoSchemaDefined, { mode: schemaErrorMode });
    }
    if (!z) {
        throw new Error('expected zod');
    }
    if (zodValidationResult === 'no-zod') {
        throw new Error('expected zod');
    }
    if (zodValidationResult === 'no-schema') {
        throw new Error('expected schema');
    }
    const typeName = (0, zod_schema_type_1.getZodSchemaType)(schema);
    if (typeName === 'any') {
        return jsx_runtime_1.jsx(SchemaErrorMessages_1.NoSchemaDefined, { mode: schemaErrorMode });
    }
    if (!unresolvedComposition.defaultProps) {
        return jsx_runtime_1.jsx(SchemaErrorMessages_1.NoDefaultProps, { mode: schemaErrorMode });
    }
    const shouldRenderControlRow = !hideModeControls || (!hideWarningButton && warnings.length > 0);
    const shouldRenderWarningMessages = showWarning && warnings.length > 0;
    const compactLayout = layout === 'inspector';
    return (jsx_runtime_1.jsxs("div", { style: layout === 'inspector' ? inspectorOuter : outer, children: [shouldRenderControlRow || shouldRenderWarningMessages ? (jsx_runtime_1.jsxs("div", { style: layout === 'inspector'
                    ? inspectorControlContainer
                    : controlContainer, children: [shouldRenderControlRow ? (jsx_runtime_1.jsxs("div", { style: tabWrapper, children: [hideModeControls ? null : (jsx_runtime_1.jsx(SegmentedControl_1.SegmentedControl, { items: modeItems, needsWrapping: false })), jsx_runtime_1.jsx(layout_1.Flex, {}), !hideWarningButton && warnings.length > 0 ? (jsx_runtime_1.jsx(WarningIndicatorButton_1.WarningIndicatorButton, { setShowWarning: setShowWarning, showWarning: showWarning, warningCount: warnings.length })) : null] })) : null, shouldRenderWarningMessages
                        ? warnings.map((warning) => (jsx_runtime_1.jsxs(react_1.default.Fragment, { children: [
                                jsx_runtime_1.jsx(layout_1.Spacing, { y: 1 }), jsx_runtime_1.jsx(ValidationMessage_1.ValidationMessage, { message: warning.message, align: "flex-start", type: "warning", size: compactLayout ? 'compact' : 'default', action: warning.resolveLink ? (jsx_runtime_1.jsx("a", { href: warning.resolveLink, target: "_blank", rel: "noopener noreferrer", style: compactLayout
                                            ? compactResolveLinkStyle
                                            : resolveLinkStyle, children: "Resolve" })) : null })] }, warning.id)))
                        : null] })) : null, mode === 'schema' ? (jsx_runtime_1.jsx(SchemaEditor_1.SchemaEditor, { value: defaultProps, setValue: setDefaultProps, schema: schema, scrollableContainer: !compactLayout, contentInset: compactLayout ? InspectorPanelLayout_1.INSPECTOR_PANEL_HORIZONTAL_PADDING : undefined, errorMode: schemaErrorMode })) : (jsx_runtime_1.jsx(RenderModalJSONPropsEditor_1.RenderModalJSONPropsEditor, { value: defaultProps !== null && defaultProps !== void 0 ? defaultProps : {}, setValue: jsonEditorSetValue, onSave: onSave, serializedJSON: serializedJSON, defaultProps: unresolvedComposition.defaultProps, schema: schema, compositionId: unresolvedComposition.id, layout: layout }))] }));
};
exports.DataEditor = DataEditor;
