// @remotion/studio/dist/components/RenderModal/RenderModalJSONPropsEditor.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.RenderModalJSONPropsEditor = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const react_1 = __importStar(require("react"));
const react_2 = require("react");
const no_react_1 = require("remotion/no-react");
const client_id_1 = require("../../helpers/client-id");
const colors_1 = require("../../helpers/colors");
const Button_1 = require("../Button");
const layout_1 = require("../layout");
const RemTextarea_1 = require("../NewComposition/RemTextarea");
const ValidationMessage_1 = require("../NewComposition/ValidationMessage");
const deep_equal_1 = require("./SchemaEditor/deep-equal");
const zod_schema_type_1 = require("./SchemaEditor/zod-schema-type");
const ZodErrorMessages_1 = require("./SchemaEditor/ZodErrorMessages");
const style = {
    fontFamily: 'monospace',
    flex: 1,
};
const inspectorStyle = {
    fontFamily: 'monospace',
    boxSizing: 'border-box',
    flex: 'none',
    fontSize: 11,
    lineHeight: '15px',
    minHeight: 220,
    overflowY: 'hidden',
    ...{
        fieldSizing: 'content',
    },
};
const scrollable = {
    padding: '8px 12px',
    display: 'flex',
    flexDirection: 'column',
    flex: 1,
};
const inspectorScrollable = {
    ...scrollable,
    flex: 'none',
};
const parseJS = (value, schema) => {
    try {
        const zodValidation = (0, zod_schema_type_1.zodSafeParse)(schema, value);
        return {
            str: JSON.stringify(value, null, 2),
            value,
            validJSON: true,
            zodValidation,
        };
    }
    catch (e) {
        return {
            str: JSON.stringify(value, null, 2),
            validJSON: false,
            error: e.message,
        };
    }
};
const parseJSON = (str, schema) => {
    try {
        const value = no_react_1.NoReactInternals.deserializeJSONWithSpecialTypes(str);
        const zodValidation = (0, zod_schema_type_1.zodSafeParse)(schema, value);
        return { str, value, validJSON: true, zodValidation };
    }
    catch (e) {
        return { str, validJSON: false, error: e.message };
    }
};
const RenderModalJSONPropsEditor = ({ setValue, value, defaultProps, onSave, serializedJSON, schema, compositionId, layout = 'default', }) => {
    if (serializedJSON === null) {
        throw new Error('expecting serializedJSON to be defined');
    }
    const { subscribeToEvent } = (0, react_2.useContext)(client_id_1.StudioServerConnectionCtx);
    const [localValue, setLocalValue] = react_1.default.useState(() => {
        return parseJSON(serializedJSON.serializedString, schema);
    });
    (0, react_1.useEffect)(() => {
        const unsub = subscribeToEvent('default-props-updatable-changed', (e) => {
            if (e.type !== 'default-props-updatable-changed') {
                return;
            }
            if (e.compositionId !== compositionId) {
                return;
            }
            const { result } = e;
            if (result.canUpdate) {
                const deserialized = no_react_1.NoReactInternals.deserializeJSONWithSpecialTypes(JSON.stringify(result.currentDefaultProps));
                const nextState = parseJS(deserialized, schema);
                setLocalValue(nextState);
            }
        });
        return () => {
            unsub();
        };
    }, [subscribeToEvent, compositionId, schema]);
    (0, react_1.useEffect)(() => {
        setLocalValue((prev) => {
            if (prev.validJSON && (0, deep_equal_1.deepEqual)(value, prev.value)) {
                return prev;
            }
            return parseJS(value, schema);
        });
    }, [value, schema]);
    const onPretty = (0, react_1.useCallback)(() => {
        if (!localValue.validJSON) {
            return;
        }
        const parsed = JSON.parse(localValue.str);
        setLocalValue({ ...localValue, str: JSON.stringify(parsed, null, 2) });
    }, [localValue, setLocalValue]);
    const onChange = (0, react_1.useCallback)((e) => {
        const parsed = parseJSON(e.target.value, schema);
        setLocalValue(parsed);
        if (parsed.validJSON && parsed.zodValidation.success) {
            setValue(parsed.value);
        }
    }, [schema, setValue]);
    const hasError = (0, react_1.useMemo)(() => {
        return !localValue.validJSON || !localValue.zodValidation.success;
    }, [localValue]);
    const hasChanged = (0, react_1.useMemo)(() => {
        return !(0, deep_equal_1.deepEqual)(value, defaultProps);
    }, [defaultProps, value]);
    const onQuickSave = (0, react_1.useCallback)(() => {
        if (hasChanged && !hasError) {
            onSave();
        }
    }, [hasChanged, hasError, onSave]);
    // If schema is changed in code
    (0, react_1.useEffect)(() => {
        setLocalValue((v) => parseJSON(v.str, schema));
    }, [schema]);
    const reset = (0, react_1.useCallback)(() => {
        setValue(defaultProps);
        setLocalValue(parseJSON(JSON.stringify(defaultProps, null, 2), schema));
    }, [defaultProps, schema, setValue]);
    const textAreaStyle = (0, react_1.useMemo)(() => {
        const baseStyle = layout === 'inspector' ? inspectorStyle : style;
        if (!hasError) {
            return baseStyle;
        }
        return {
            ...baseStyle,
            borderColor: colors_1.FAIL_COLOR,
        };
    }, [hasError, layout]);
    const buttonSize = layout === 'inspector' ? 'condensed' : 'default';
    return (jsx_runtime_1.jsxs("div", { style: layout === 'inspector' ? inspectorScrollable : scrollable, children: [
            jsx_runtime_1.jsx(RemTextarea_1.RemTextarea, { onChange: onChange, onBlur: onQuickSave, value: localValue.str, status: localValue.validJSON ? 'ok' : 'error', style: textAreaStyle, small: layout === 'inspector' }), jsx_runtime_1.jsx(layout_1.Spacing, { y: 1 }), jsx_runtime_1.jsx("div", { "data-testid": "json-props-error", children: localValue.validJSON === false ? (jsx_runtime_1.jsx(ValidationMessage_1.ValidationMessage, { align: "flex-start", message: localValue.error, type: "error", size: layout === 'inspector' ? 'compact' : 'default' })) : localValue.zodValidation.success === false ? (jsx_runtime_1.jsx(ZodErrorMessages_1.ZodErrorMessages, { zodValidationResult: localValue.zodValidation, viewTab: "json", size: layout === 'inspector' ? 'compact' : 'default' })) : null }), jsx_runtime_1.jsx(layout_1.Spacing, { y: 1 }), jsx_runtime_1.jsxs(layout_1.Row, { children: [
                    jsx_runtime_1.jsx(Button_1.Button, { disabled: !(hasChanged || !localValue.validJSON), onClick: reset, size: buttonSize, children: "Reset" }), jsx_runtime_1.jsx(layout_1.Flex, {}), jsx_runtime_1.jsx(Button_1.Button, { disabled: !localValue.validJSON, onClick: onPretty, size: buttonSize, children: "Format" }), jsx_runtime_1.jsx(layout_1.Spacing, { x: 1 })
                ] })
        ] }));
};
exports.RenderModalJSONPropsEditor = RenderModalJSONPropsEditor;
