// @remotion/studio/dist/components/AudioWaveform.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AudioWaveform = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const timeline_utils_1 = require("@remotion/timeline-utils");
const react_1 = require("react");
const remotion_1 = require("remotion");
const colors_1 = require("../helpers/colors");
const timeline_layout_1 = require("../helpers/timeline-layout");
const EMPTY_PEAKS = new Float32Array(0);
// Recreate the canvas after Fast Refresh because an already transferred canvas
// cannot be handed to OffscreenCanvas again.
const canRetryCanvasTransfer = (err) => {
    return err instanceof DOMException && err.name === 'InvalidStateError';
};
const canUseAudioWaveformWorker = () => {
    if (typeof Worker === 'undefined' ||
        typeof OffscreenCanvas === 'undefined' ||
        typeof HTMLCanvasElement === 'undefined') {
        return false;
    }
    return 'transferControlToOffscreen' in HTMLCanvasElement.prototype;
};
const getContainerStyle = (height) => {
    return {
        display: 'flex',
        flexDirection: 'row',
        alignItems: 'center',
        position: 'relative',
        width: '100%',
        height,
    };
};
const errorMessage = {
    fontSize: 13,
    paddingTop: 6,
    paddingBottom: 6,
    paddingLeft: 12,
    paddingRight: 12,
    alignSelf: 'flex-start',
    maxWidth: 450,
    opacity: 0.75,
};
const getWaveformErrorMessage = () => {
    return new Error('No waveform available. The audio could not be decoded or may not support CORS.');
};
const waveformCanvasStyle = {
    pointerEvents: 'none',
    width: '100%',
    height: '100%',
};
const volumeCanvasStyle = {
    position: 'absolute',
};
const parseVolume = (volume) => {
    if (typeof volume === 'number') {
        return volume;
    }
    return volume.split(',').map((v) => Number(v));
};
const drawLoopedWaveform = ({ canvas, peaks, volume, visualizationWidth, loopWidth, }) => {
    const h = canvas.height;
    const w = Math.ceil(visualizationWidth);
    const targetCanvas = document.createElement('canvas');
    targetCanvas.width = Math.max(1, Math.ceil(loopWidth));
    targetCanvas.height = h;
    (0, timeline_utils_1.drawBars)({
        canvas: targetCanvas,
        peaks,
        color: 'rgba(255, 255, 255, 0.6)',
        volume,
        width: targetCanvas.width,
    });
    canvas.width = w;
    canvas.height = h;
    const ctx = canvas.getContext('2d');
    if (!ctx) {
        throw new Error('Failed to get canvas context');
    }
    const pattern = ctx.createPattern(targetCanvas, 'repeat-x');
    if (!pattern) {
        return;
    }
    pattern.setTransform(new DOMMatrix().scaleSelf(loopWidth / targetCanvas.width, 1));
    ctx.clearRect(0, 0, w, h);
    ctx.fillStyle = pattern;
    ctx.fillRect(0, 0, w, h);
};
const AudioWaveform = ({ src, height, startFrom, durationInFrames, visualizationWidth, volume, doesVolumeChange, playbackRate, loopDisplay, }) => {
    const [peaks, setPeaks] = (0, react_1.useState)(null);
    const [error, setError] = (0, react_1.useState)(null);
    const [waveformCanvasKey, setWaveformCanvasKey] = (0, react_1.useState)(0);
    const canUseWorkerPath = (0, react_1.useMemo)(() => canUseAudioWaveformWorker(), []);
    const vidConf = remotion_1.Internals.useUnsafeVideoConfig();
    if (vidConf === null) {
        throw new Error('Expected video config');
    }
    const waveformCanvas = (0, react_1.useRef)(null);
    const volumeCanvas = (0, react_1.useRef)(null);
    const waveformWorker = (0, react_1.useRef)(null);
    const hasTransferredCanvas = (0, react_1.useRef)(false);
    const latestRequestId = (0, react_1.useRef)(0);
    const shouldRenderVolumeOverlay = doesVolumeChange && typeof volume === 'string';
    const parsedVolume = (0, react_1.useMemo)(() => parseVolume(volume), [volume]);
    (0, react_1.useEffect)(() => {
        if (canUseWorkerPath) {
            return;
        }
        const controller = new AbortController();
        setPeaks(null);
        setError(null);
        (0, timeline_utils_1.loadWaveformPeaks)(src, controller.signal)
            .then((p) => {
            if (!controller.signal.aborted) {
                setPeaks(p);
            }
        })
            .catch((err) => {
            if (!controller.signal.aborted) {
                setError(err);
            }
        });
        return () => controller.abort();
    }, [canUseWorkerPath, src]);
    (0, react_1.useEffect)(() => {
        if (!canUseWorkerPath) {
            return;
        }
        const canvasElement = waveformCanvas.current;
        if (!canvasElement || hasTransferredCanvas.current) {
            return;
        }
        const worker = (0, timeline_utils_1.makeAudioWaveformWorker)();
        let workerFailed = false;
        waveformWorker.current = worker;
        worker.addEventListener('message', (event) => {
            if (event.data.type === 'error') {
                if (event.data.requestId !== latestRequestId.current) {
                    return;
                }
                setError(new Error(event.data.message));
            }
        });
        worker.addEventListener('error', (event) => {
            event.preventDefault();
            workerFailed = true;
            if (worker !== waveformWorker.current) {
                return;
            }
            worker.terminate();
            waveformWorker.current = null;
            hasTransferredCanvas.current = false;
            setError(getWaveformErrorMessage());
        });
        let offscreen;
        try {
            offscreen = canvasElement.transferControlToOffscreen();
        }
        catch (err) {
            worker.terminate();
            waveformWorker.current = null;
            if (canRetryCanvasTransfer(err)) {
                setWaveformCanvasKey((key) => key + 1);
                return;
            }
            throw err;
        }
        hasTransferredCanvas.current = true;
        worker.postMessage({ type: 'init', canvas: offscreen }, [offscreen]);
        return () => {
            if (!workerFailed) {
                worker.postMessage({ type: 'dispose' });
            }
            worker.terminate();
            waveformWorker.current = null;
            hasTransferredCanvas.current = false;
        };
    }, [canUseWorkerPath, waveformCanvasKey]);
    const portionPeaks = (0, react_1.useMemo)(() => {
        if (canUseWorkerPath || !peaks) {
            return null;
        }
        return (0, timeline_utils_1.sliceWaveformPeaks)({
            durationInFrames: (0, timeline_utils_1.shouldTileLoopDisplay)(loopDisplay)
                ? loopDisplay.durationInFrames
                : durationInFrames,
            fps: vidConf.fps,
            peaks,
            playbackRate,
            startFrom,
        });
    }, [
        canUseWorkerPath,
        durationInFrames,
        loopDisplay,
        peaks,
        playbackRate,
        startFrom,
        vidConf.fps,
    ]);
    (0, react_1.useEffect)(() => {
        const { current: canvasElement } = waveformCanvas;
        if (!canvasElement) {
            return;
        }
        const h = height;
        const w = Math.ceil(visualizationWidth);
        if (canUseWorkerPath) {
            const worker = waveformWorker.current;
            if (!worker || !hasTransferredCanvas.current) {
                return;
            }
            latestRequestId.current += 1;
            setError(null);
            const message = {
                type: 'render',
                requestId: latestRequestId.current,
                src,
                width: w,
                height: h,
                volume: parsedVolume,
                startFrom,
                durationInFrames,
                fps: vidConf.fps,
                playbackRate,
                loopDisplay,
            };
            worker.postMessage(message);
            return;
        }
        canvasElement.width = w;
        canvasElement.height = h;
        if ((0, timeline_utils_1.shouldTileLoopDisplay)(loopDisplay)) {
            drawLoopedWaveform({
                canvas: canvasElement,
                peaks: portionPeaks !== null && portionPeaks !== void 0 ? portionPeaks : EMPTY_PEAKS,
                volume: parsedVolume,
                visualizationWidth,
                loopWidth: (0, timeline_utils_1.getLoopDisplayWidth)({
                    visualizationWidth,
                    loopDisplay,
                }),
            });
        }
        else {
            (0, timeline_utils_1.drawBars)({
                canvas: canvasElement,
                peaks: portionPeaks !== null && portionPeaks !== void 0 ? portionPeaks : EMPTY_PEAKS,
                color: 'rgba(255, 255, 255, 0.6)',
                volume: parsedVolume,
                width: w,
            });
        }
    }, [
        canUseWorkerPath,
        durationInFrames,
        height,
        loopDisplay,
        playbackRate,
        parsedVolume,
        portionPeaks,
        src,
        startFrom,
        vidConf.fps,
        visualizationWidth,
        waveformCanvasKey,
    ]);
    (0, react_1.useEffect)(() => {
        if (!shouldRenderVolumeOverlay) {
            return;
        }
        const { current: volumeCanvasElement } = volumeCanvas;
        if (!volumeCanvasElement) {
            return;
        }
        const h = height;
        const context = volumeCanvasElement.getContext('2d');
        if (!context) {
            return;
        }
        volumeCanvasElement.width = Math.ceil(visualizationWidth);
        volumeCanvasElement.height = h;
        context.clearRect(0, 0, visualizationWidth, h);
        if (!Array.isArray(parsedVolume)) {
            return;
        }
        context.beginPath();
        context.moveTo(0, h);
        parsedVolume.forEach((v, index) => {
            const x = parsedVolume.length <= 1
                ? 0
                : (index / (parsedVolume.length - 1)) * visualizationWidth;
            const y = (1 - v) * (h - timeline_layout_1.TIMELINE_BORDER * 2) + 1;
            if (index === 0) {
                context.moveTo(x, y);
            }
            else {
                context.lineTo(x, y);
            }
        });
        context.strokeStyle = colors_1.LIGHT_TRANSPARENT;
        context.stroke();
    }, [height, parsedVolume, shouldRenderVolumeOverlay, visualizationWidth]);
    if (error) {
        return (jsx_runtime_1.jsx("div", { style: getContainerStyle(height), children: jsx_runtime_1.jsx("div", { style: errorMessage, children: "No waveform available. The audio could not be decoded or may not support CORS." }) }));
    }
    if (!canUseWorkerPath && !peaks) {
        return null;
    }
    return (jsx_runtime_1.jsxs("div", { style: getContainerStyle(height), children: [
            jsx_runtime_1.jsx("canvas", { ref: waveformCanvas, style: waveformCanvasStyle }, waveformCanvasKey), shouldRenderVolumeOverlay ? (jsx_runtime_1.jsx("canvas", { ref: volumeCanvas, style: volumeCanvasStyle })) : null] }));
};
exports.AudioWaveform = AudioWaveform;
