// @remotion/studio/dist/components/Button.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Button = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const react_1 = require("react");
const colors_1 = require("../helpers/colors");
const button = {
    border: `1px solid ${colors_1.INPUT_BORDER_COLOR_UNHOVERED}`,
    borderRadius: 4,
    backgroundColor: colors_1.INPUT_BACKGROUND,
    appearance: 'none',
    fontFamily: 'inherit',
    fontSize: 14,
    color: 'white',
    flexDirection: 'row',
};
const ButtonRefForwardFunction = ({ children, onClick, title, disabled, size = 'default', style, id, autoFocus, buttonContainerStyle, onPointerDown, }, ref) => {
    const combined = (0, react_1.useMemo)(() => {
        return {
            ...button,
            ...(size === 'compact' ? { fontSize: 12 } : null),
            ...(size === 'condensed' ? { fontSize: 11 } : null),
            ...(style !== null && style !== void 0 ? style : {}),
        };
    }, [size, style]);
    const buttonContainer = (0, react_1.useMemo)(() => {
        return {
            padding: size === 'condensed' ? '2px 7px' : size === 'compact' ? '5px 8px' : 10,
            cursor: disabled ? 'inherit' : 'pointer',
            fontSize: size === 'condensed' ? 11 : size === 'compact' ? 12 : 14,
            lineHeight: size === 'condensed' ? '14px' : size === 'compact' ? '14px' : undefined,
            opacity: disabled ? 0.7 : 1,
            ...(buttonContainerStyle !== null && buttonContainerStyle !== void 0 ? buttonContainerStyle : {}),
        };
    }, [buttonContainerStyle, disabled, size]);
    return (jsx_runtime_1.jsx("button", { ref: ref, id: id, style: combined, type: "button", disabled: disabled, onClick: onClick, onPointerDown: onPointerDown, autoFocus: autoFocus, title: title, children: jsx_runtime_1.jsx("div", { className: "css-reset", style: buttonContainer, children: children }) }));
};
exports.Button = (0, react_1.forwardRef)(ButtonRefForwardFunction);
