// @remotion/studio/dist/components/CompSelectorRef.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CompSelectorRef = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const react_1 = require("react");
const remotion_1 = require("remotion");
const create_folder_tree_1 = require("../helpers/create-folder-tree");
const persist_boolean_map_1 = require("../helpers/persist-boolean-map");
const persist_open_folders_1 = require("../helpers/persist-open-folders");
const folders_1 = require("../state/folders");
const InitialCompositionLoader_1 = require("./InitialCompositionLoader");
const CompSelectorRef = ({ children }) => {
    const { compositions } = (0, react_1.useContext)(remotion_1.Internals.CompositionManager);
    const { compositionFoldersExpanded, setCompositionFoldersExpanded } = (0, react_1.useContext)(folders_1.FolderContext);
    const selectComposition = (0, InitialCompositionLoader_1.useSelectComposition)();
    const toggleFolder = (0, react_1.useCallback)((folderName, parentName) => {
        setCompositionFoldersExpanded((p) => {
            const key = (0, persist_open_folders_1.openFolderKey)({ folderName, parentName });
            const foldersExpandedState = (0, persist_boolean_map_1.toggleBooleanMapKey)(p, key);
            (0, persist_open_folders_1.persistExpandedFolders)('compositions', foldersExpandedState);
            return foldersExpandedState;
        });
    }, [setCompositionFoldersExpanded]);
    (0, react_1.useImperativeHandle)(remotion_1.Internals.compositionSelectorRef, () => {
        return {
            expandComposition: (compName) => {
                const compositionToExpand = compositions.find((c) => c.id === compName);
                if (!compositionToExpand) {
                    return;
                }
                const { folderName, parentFolderName } = compositionToExpand;
                if (folderName === null) {
                    return;
                }
                setCompositionFoldersExpanded((previousState) => {
                    const keysToExpand = (0, create_folder_tree_1.getKeysToExpand)(folderName, parentFolderName);
                    const foldersExpandedState = {
                        ...previousState,
                    };
                    for (const key of keysToExpand) {
                        foldersExpandedState[key] = true;
                    }
                    (0, persist_open_folders_1.persistExpandedFolders)('compositions', foldersExpandedState);
                    return foldersExpandedState;
                });
            },
            selectComposition: (compName) => {
                const comp = compositions.find((c) => c.id === compName);
                if (!comp) {
                    throw new Error(`Composition ${compName} not found`);
                }
                selectComposition(comp, true);
            },
            toggleFolder: (folderName, parentName) => {
                toggleFolder(folderName, parentName);
            },
        };
    }, [
        compositions,
        selectComposition,
        setCompositionFoldersExpanded,
        toggleFolder,
    ]);
    const contextValue = (0, react_1.useMemo)(() => {
        return {
            foldersExpanded: compositionFoldersExpanded,
            setFoldersExpanded: (foldersExpanded) => {
                setCompositionFoldersExpanded(foldersExpanded);
                (0, persist_open_folders_1.persistExpandedFolders)('compositions', foldersExpanded);
            },
            toggleFolder,
        };
    }, [compositionFoldersExpanded, setCompositionFoldersExpanded, toggleFolder]);
    return (jsx_runtime_1.jsx(persist_open_folders_1.ExpandedFoldersContext.Provider, { value: contextValue, children: children }));
};
exports.CompSelectorRef = CompSelectorRef;
