// @remotion/studio/dist/components/composition-menu-items.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getCompositionMenuItems = void 0;
const no_react_1 = require("remotion/no-react");
const format_file_location_1 = require("../helpers/format-file-location");
const open_in_editor_1 = require("../helpers/open-in-editor");
const NotificationCenter_1 = require("./Notifications/NotificationCenter");
const getCompositionMenuItems = ({ composition, connectionStatus, resolvedLocation, setSelectedModal, closeMenu, readOnlyStudio, }) => {
    const editorName = window.remotion_editorName;
    const fileLocation = (0, format_file_location_1.formatFileLocation)({
        location: resolvedLocation,
        root: window.remotion_cwd,
    });
    const showInEditorDisabled = !composition || connectionStatus !== 'connected' || !resolvedLocation;
    const openComponentInEditorDisabled = showInEditorDisabled || !(resolvedLocation === null || resolvedLocation === void 0 ? void 0 : resolvedLocation.source);
    const copyFileLocationDisabled = !composition || !fileLocation;
    return [
        editorName
            ? {
                id: 'show-in-editor',
                keyHint: null,
                label: `Show composition in ${editorName}`,
                leftItem: null,
                onClick: async () => {
                    closeMenu();
                    if (!composition || !resolvedLocation) {
                        return;
                    }
                    try {
                        await (0, open_in_editor_1.openOriginalPositionInEditor)(resolvedLocation);
                    }
                    catch (err) {
                        (0, NotificationCenter_1.showNotification)(err.message, 2000);
                    }
                },
                quickSwitcherLabel: `Show composition in ${editorName}`,
                subMenu: null,
                type: 'item',
                value: 'show-in-editor',
                disabled: showInEditorDisabled,
            }
            : null,
        {
            id: 'copy-file-location',
            keyHint: null,
            label: `Copy file location`,
            leftItem: null,
            onClick: () => {
                closeMenu();
                if (!fileLocation) {
                    return;
                }
                navigator.clipboard
                    .writeText(fileLocation)
                    .then(() => {
                    (0, NotificationCenter_1.showNotification)('Copied file location to clipboard', 1000);
                })
                    .catch((err) => {
                    (0, NotificationCenter_1.showNotification)(`Could not copy to clipboard: ${err.message}`, 1000);
                });
            },
            quickSwitcherLabel: 'Copy composition file location',
            subMenu: null,
            type: 'item',
            value: 'copy-file-location',
            disabled: copyFileLocationDisabled,
        },
        editorName
            ? {
                id: 'open-component-in-editor',
                keyHint: null,
                label: `Show component in ${editorName}`,
                leftItem: null,
                onClick: async () => {
                    closeMenu();
                    if (!composition || !(resolvedLocation === null || resolvedLocation === void 0 ? void 0 : resolvedLocation.source)) {
                        return;
                    }
                    try {
                        await (0, open_in_editor_1.openCompositionComponentInEditor)({
                            compositionFile: resolvedLocation.source,
                            compositionId: composition.id,
                        });
                    }
                    catch (err) {
                        (0, NotificationCenter_1.showNotification)(err.message, 2000);
                    }
                },
                quickSwitcherLabel: `Show composition component in ${editorName}`,
                subMenu: null,
                type: 'item',
                value: 'open-component-in-editor',
                disabled: openComponentInEditorDisabled,
            }
            : null,
        editorName || fileLocation
            ? {
                type: 'divider',
                id: 'show-in-editor-divider',
            }
            : null,
        {
            id: 'new',
            keyHint: null,
            label: `New...`,
            leftItem: null,
            onClick: () => {
                closeMenu();
                setSelectedModal({
                    type: 'new-comp',
                });
            },
            quickSwitcherLabel: 'New composition...',
            subMenu: null,
            type: 'item',
            value: 'new',
            disabled: readOnlyStudio,
        },
        {
            id: 'rename',
            keyHint: null,
            label: `Rename...`,
            leftItem: null,
            onClick: () => {
                if (!composition) {
                    return;
                }
                closeMenu();
                setSelectedModal({
                    type: 'rename-comp',
                    compositionId: composition.id,
                });
            },
            quickSwitcherLabel: 'Rename composition...',
            subMenu: null,
            type: 'item',
            value: 'rename',
            disabled: !composition || readOnlyStudio,
        },
        {
            id: 'duplicate',
            keyHint: null,
            label: `Duplicate...`,
            leftItem: null,
            onClick: () => {
                if (!composition) {
                    return;
                }
                closeMenu();
                setSelectedModal({
                    type: 'duplicate-comp',
                    compositionId: composition.id,
                    compositionType: composition.durationInFrames === 1 ? 'still' : 'composition',
                });
            },
            quickSwitcherLabel: 'Duplicate composition...',
            subMenu: null,
            type: 'item',
            value: 'duplicate',
            disabled: !composition || readOnlyStudio,
        },
        {
            id: 'delete',
            keyHint: null,
            label: `Delete...`,
            leftItem: null,
            onClick: () => {
                if (!composition) {
                    return;
                }
                closeMenu();
                setSelectedModal({
                    type: 'delete-comp',
                    compositionId: composition.id,
                });
            },
            quickSwitcherLabel: 'Delete composition...',
            subMenu: null,
            type: 'item',
            value: 'delete',
            disabled: !composition || readOnlyStudio,
        },
        {
            type: 'divider',
            id: 'copy-id-divider',
        },
        {
            id: 'copy-id',
            keyHint: null,
            label: `Copy ID`,
            leftItem: null,
            onClick: () => {
                if (!composition) {
                    return;
                }
                closeMenu();
                navigator.clipboard
                    .writeText(composition.id)
                    .then(() => {
                    (0, NotificationCenter_1.showNotification)('Copied to clipboard', 1000);
                })
                    .catch((err) => {
                    (0, NotificationCenter_1.showNotification)(`Could not copy to clipboard: ${err.message}`, 1000);
                });
            },
            quickSwitcherLabel: 'Copy composition ID',
            subMenu: null,
            type: 'item',
            value: 'copy-id',
            disabled: !composition,
        },
    ].filter(no_react_1.NoReactInternals.truthy);
};
exports.getCompositionMenuItems = getCompositionMenuItems;
