// @remotion/studio/dist/components/ConfirmationDialog.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ConfirmationDialog = exports.useConfirmationDialog = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const react_1 = require("react");
const ShortcutHint_1 = require("../error-overlay/remotion-overlay/ShortcutHint");
const modals_1 = require("../state/modals");
const Button_1 = require("./Button");
const layout_1 = require("./layout");
const ModalButton_1 = require("./ModalButton");
const ModalContainer_1 = require("./ModalContainer");
const ModalFooter_1 = require("./ModalFooter");
const ModalHeader_1 = require("./ModalHeader");
const content = {
    padding: 16,
    fontSize: 14,
    flex: 1,
    minWidth: 420,
    maxWidth: 560,
    lineHeight: 1.4,
};
const useConfirmationDialog = () => {
    const { setSelectedModal } = (0, react_1.useContext)(modals_1.ModalsContext);
    return (0, react_1.useCallback)((options) => {
        return new Promise((resolve) => {
            var _a, _b;
            let settled = false;
            const settle = (result) => {
                if (settled) {
                    return;
                }
                settled = true;
                resolve(result);
            };
            setSelectedModal({
                type: 'confirmation-dialog',
                id: String(Math.random()),
                title: options.title,
                message: options.message,
                confirmLabel: (_a = options.confirmLabel) !== null && _a !== void 0 ? _a : 'Continue',
                cancelLabel: (_b = options.cancelLabel) !== null && _b !== void 0 ? _b : 'Cancel',
                onConfirm: () => settle(true),
                onCancel: () => settle(false),
            });
        });
    }, [setSelectedModal]);
};
exports.useConfirmationDialog = useConfirmationDialog;
const ConfirmationDialog = ({ state }) => {
    const { setSelectedModal } = (0, react_1.useContext)(modals_1.ModalsContext);
    const settledRef = (0, react_1.useRef)(false);
    const closeCurrentModal = (0, react_1.useCallback)(() => {
        setSelectedModal((modal) => (modal === null || modal === void 0 ? void 0 : modal.type) === 'confirmation-dialog' && modal.id === state.id
            ? null
            : modal);
    }, [setSelectedModal, state.id]);
    const settle = (0, react_1.useCallback)((confirmed) => {
        if (settledRef.current) {
            return;
        }
        settledRef.current = true;
        closeCurrentModal();
        if (confirmed) {
            state.onConfirm();
        }
        else {
            state.onCancel();
        }
    }, [closeCurrentModal, state]);
    (0, react_1.useEffect)(() => {
        return () => {
            if (settledRef.current) {
                return;
            }
            settledRef.current = true;
            state.onCancel();
        };
    }, [state]);
    const onCancel = (0, react_1.useCallback)(() => {
        settle(false);
    }, [settle]);
    const onConfirm = (0, react_1.useCallback)(() => {
        settle(true);
    }, [settle]);
    const onSubmit = (0, react_1.useCallback)((e) => {
        e.preventDefault();
        onConfirm();
    }, [onConfirm]);
    const cancelStyle = (0, react_1.useMemo)(() => {
        return {
            minWidth: 90,
        };
    }, []);
    return (jsx_runtime_1.jsxs(ModalContainer_1.ModalContainer, { onOutsideClick: onCancel, onEscape: onCancel, children: [
            jsx_runtime_1.jsx(ModalHeader_1.ModalHeader, { title: state.title, onClose: onCancel }), jsx_runtime_1.jsxs("form", { onSubmit: onSubmit, children: [
                    jsx_runtime_1.jsx("div", { style: content, children: state.message }), jsx_runtime_1.jsx(ModalFooter_1.ModalFooterContainer, { children: jsx_runtime_1.jsxs(layout_1.Row, { align: "center", children: [
                                jsx_runtime_1.jsx(layout_1.Flex, {}), jsx_runtime_1.jsx(Button_1.Button, { onClick: onCancel, style: cancelStyle, children: state.cancelLabel }), jsx_runtime_1.jsx(layout_1.Spacing, { x: 1 }), jsx_runtime_1.jsxs(ModalButton_1.ModalButton, { onClick: onConfirm, autoFocus: true, children: [state.confirmLabel, jsx_runtime_1.jsx(ShortcutHint_1.ShortcutHint, { keyToPress: "\u21B5", cmdOrCtrl: false })
                                    ] })
                            ] }) })
                ] })
        ] }));
};
exports.ConfirmationDialog = ConfirmationDialog;
