// @remotion/studio/dist/components/ContextMenu.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ContextMenuForTarget = exports.ContextMenu = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const player_1 = require("@remotion/player");
const react_1 = __importStar(require("react"));
const react_dom_1 = __importDefault(require("react-dom"));
const mobile_layout_1 = require("../helpers/mobile-layout");
const noop_1 = require("../helpers/noop");
const z_index_1 = require("../state/z-index");
const portals_1 = require("./Menu/portals");
const styles_1 = require("./Menu/styles");
const MenuContent_1 = require("./NewComposition/MenuContent");
const ContextMenuPortal = ({ sizeSource, currentZIndex, onHide, opened, values }) => {
    const size = player_1.PlayerInternals.useElementSize(sizeSource, {
        triggerOnWindowResize: true,
        shouldApplyCssTransforms: true,
    });
    const isMobileLayout = (0, mobile_layout_1.useMobileLayout)();
    const spaceToBottom = (0, react_1.useMemo)(() => {
        if (size) {
            return size.windowSize.height - opened.top;
        }
        return 0;
    }, [opened.top, size]);
    const spaceToTop = (0, react_1.useMemo)(() => {
        if (size) {
            return opened.top;
        }
        return 0;
    }, [opened.top, size]);
    const portalStyle = (0, react_1.useMemo)(() => {
        if (!size) {
            return;
        }
        const spaceToRight = size.windowSize.width - size.left;
        const spaceToLeft = size.left + size.width;
        const minSpaceRequired = isMobileLayout
            ? styles_1.MAX_MOBILE_MENU_WIDTH
            : styles_1.MAX_MENU_WIDTH;
        const verticalLayout = spaceToTop > spaceToBottom ? 'bottom' : 'top';
        const canOpenOnLeft = spaceToLeft >= minSpaceRequired;
        const canOpenOnRight = spaceToRight >= minSpaceRequired;
        const horizontalLayout = canOpenOnRight ? 'left' : 'right';
        return {
            ...styles_1.menuContainerTowardsTop,
            ...(verticalLayout === 'top'
                ? {
                    top: opened.top,
                }
                : {
                    bottom: size.windowSize.height - opened.top,
                }),
            ...(horizontalLayout === 'left'
                ? {
                    left: opened.left,
                }
                : {
                    right: canOpenOnLeft ? size.windowSize.width - opened.left : 0,
                }),
        };
    }, [
        opened.left,
        opened.top,
        size,
        isMobileLayout,
        spaceToTop,
        spaceToBottom,
    ]);
    // Prevent deselection of a selected item
    const onMenuPointerDown = (0, react_1.useCallback)((e) => {
        e.stopPropagation();
    }, []);
    if (!portalStyle) {
        return null;
    }
    return react_dom_1.default.createPortal(jsx_runtime_1.jsx("div", { style: styles_1.fullScreenOverlay, children: jsx_runtime_1.jsx("div", { style: styles_1.outerPortal, className: "css-reset", children: jsx_runtime_1.jsx(z_index_1.HigherZIndex, { onOutsideClick: onHide, onEscape: onHide, children: jsx_runtime_1.jsx("div", { style: portalStyle, onPointerDown: onMenuPointerDown, children: jsx_runtime_1.jsx(MenuContent_1.MenuContent, { onNextMenu: noop_1.noop, onPreviousMenu: noop_1.noop, values: values, onHide: onHide, leaveLeftSpace: true, preselectIndex: false, topItemCanBeUnselected: false, fixedHeight: null }) }) }) }) }), (0, portals_1.getPortal)(currentZIndex));
};
exports.ContextMenu = react_1.default.forwardRef(({ children, values, onOpen, style = undefined, className = undefined, onPointerDown = undefined, }, forwardedRef) => {
    const ref = (0, react_1.useRef)(null);
    const [opened, setOpened] = (0, react_1.useState)({ type: 'not-open' });
    const { currentZIndex } = (0, z_index_1.useZIndex)();
    const setRef = (0, react_1.useCallback)((node) => {
        ref.current = node;
        if (typeof forwardedRef === 'function') {
            forwardedRef(node);
            return;
        }
        if (forwardedRef) {
            forwardedRef.current = node;
        }
    }, [forwardedRef]);
    (0, react_1.useEffect)(() => {
        const { current } = ref;
        if (!current) {
            return;
        }
        const onClick = (e) => {
            e.preventDefault();
            e.stopPropagation();
            if ((onOpen === null || onOpen === void 0 ? void 0 : onOpen()) === false) {
                return false;
            }
            setOpened({ type: 'open', left: e.clientX, top: e.clientY });
            return false;
        };
        current.addEventListener('contextmenu', onClick);
        return () => {
            current.removeEventListener('contextmenu', onClick);
        };
    }, [onOpen]);
    const onHide = (0, react_1.useCallback)(() => {
        setOpened({ type: 'not-open' });
    }, []);
    return (jsx_runtime_1.jsxs(jsx_runtime_1.Fragment, { children: [
            jsx_runtime_1.jsx("div", { ref: setRef, onContextMenu: () => false, style: style, className: className, onPointerDown: onPointerDown, children: children }), opened.type === 'open' ? (jsx_runtime_1.jsx(ContextMenuPortal, { sizeSource: ref, currentZIndex: currentZIndex, onHide: onHide, opened: opened, values: values })) : null] }));
});
exports.ContextMenu.displayName = 'ContextMenu';
const ContextMenuForTarget = ({ triggerRef, values, onOpen }) => {
    const [opened, setOpened] = (0, react_1.useState)({ type: 'not-open' });
    const [openedValues, setOpenedValues] = (0, react_1.useState)(values);
    const [body, setBody] = (0, react_1.useState)(null);
    const { currentZIndex } = (0, z_index_1.useZIndex)();
    (0, react_1.useEffect)(() => {
        // Access document.body after mount so importing this component stays safe
        // in DOM-less test environments.
        setBody(document.body);
    }, []);
    (0, react_1.useEffect)(() => {
        const { current } = triggerRef;
        if (!current) {
            return;
        }
        const onClick = async (event) => {
            const e = event;
            e.preventDefault();
            e.stopPropagation();
            const result = await (onOpen === null || onOpen === void 0 ? void 0 : onOpen(e));
            if (result === false) {
                return false;
            }
            const nextValues = Array.isArray(result) ? result : values;
            if (nextValues.length === 0) {
                return false;
            }
            setOpenedValues(nextValues);
            setOpened({ type: 'open', left: e.clientX, top: e.clientY });
            return false;
        };
        current.addEventListener('contextmenu', onClick);
        return () => {
            current.removeEventListener('contextmenu', onClick);
        };
    }, [onOpen, triggerRef, values]);
    const onHide = (0, react_1.useCallback)(() => {
        setOpened({ type: 'not-open' });
    }, []);
    return opened.type === 'open' ? (jsx_runtime_1.jsx(ContextMenuPortal, { sizeSource: body, currentZIndex: currentZIndex, onHide: onHide, opened: opened, values: [...openedValues] })) : null;
};
exports.ContextMenuForTarget = ContextMenuForTarget;
