// @remotion/studio/dist/components/Editor.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.Editor = exports.BUFFER_STATE_DELAY_IN_MILLISECONDS = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const player_1 = require("@remotion/player");
const react_1 = __importStar(require("react"));
const remotion_1 = require("remotion");
const colors_1 = require("../helpers/colors");
const noop_1 = require("../helpers/noop");
const canvas_ref_1 = require("../state/canvas-ref");
const scale_lock_1 = require("../state/scale-lock");
const timeline_zoom_1 = require("../state/timeline-zoom");
const z_index_1 = require("../state/z-index");
const canvas_capture_enabled_1 = require("./canvas-capture-enabled");
const EditorContent_1 = require("./EditorContent");
const ForceSpecificCursor_1 = require("./ForceSpecificCursor");
const Modals_1 = require("./Modals");
const NotificationCenter_1 = require("./Notifications/NotificationCenter");
const RenderErrorContext_1 = require("./RenderErrorContext");
const SequencePropsSubscriptionProvider_1 = require("./SequencePropsSubscriptionProvider");
const StudioCanvasCapture_1 = require("./StudioCanvasCapture");
const TopPanel_1 = require("./TopPanel");
const background = {
    backgroundColor: colors_1.BACKGROUND,
    display: 'flex',
    width: '100%',
    height: '100%',
    flexDirection: 'column',
    position: 'absolute',
};
const DEFAULT_BUFFER_STATE_DELAY_IN_MILLISECONDS = 300;
exports.BUFFER_STATE_DELAY_IN_MILLISECONDS = typeof process.env.BUFFER_STATE_DELAY_IN_MILLISECONDS === 'undefined' ||
    process.env.BUFFER_STATE_DELAY_IN_MILLISECONDS === null
    ? DEFAULT_BUFFER_STATE_DELAY_IN_MILLISECONDS
    : Number(process.env.BUFFER_STATE_DELAY_IN_MILLISECONDS);
const Editor = ({ Root, readOnlyStudio }) => {
    const [drawElement, setDrawElement] = (0, react_1.useState)(null);
    const size = player_1.PlayerInternals.useElementSize(drawElement, {
        triggerOnWindowResize: false,
        shouldApplyCssTransforms: true,
    });
    const [canvasMounted, setCanvasMounted] = react_1.default.useState(false);
    const onMounted = (0, react_1.useCallback)(() => {
        setCanvasMounted(true);
    }, []);
    // Use a callback ref so the late-mounted canvas container triggers a render
    // and useElementSize() can observe it. See GitHub issue #8098.
    const setDrawRef = (0, react_1.useCallback)((node) => {
        canvas_ref_1.drawRef.current = node;
        setDrawElement(node);
    }, []);
    const value = (0, react_1.useMemo)(() => {
        if (!size) {
            return null;
        }
        return {
            type: 'canvas-size',
            canvasSize: size,
        };
    }, [size]);
    const MemoRoot = (0, react_1.useMemo)(() => {
        return react_1.default.memo(Root);
    }, [Root]);
    const [renderError, setRenderError] = (0, react_1.useState)(null);
    const clearError = (0, react_1.useCallback)(() => {
        setRenderError(null);
    }, []);
    const compositionRenderErrorContextValue = (0, react_1.useMemo)(() => ({ setError: setRenderError, clearError }), [clearError]);
    const renderErrorContextValue = (0, react_1.useMemo)(() => ({ error: renderError }), [renderError]);
    const editor = (jsx_runtime_1.jsx(z_index_1.HigherZIndex, { onEscape: noop_1.noop, onOutsideClick: noop_1.noop, children: jsx_runtime_1.jsx(timeline_zoom_1.TimelineZoomContext, { children: jsx_runtime_1.jsxs(SequencePropsSubscriptionProvider_1.SequencePropsSubscriptionProvider, { children: [
                    jsx_runtime_1.jsxs(remotion_1.Internals.CurrentScaleContext.Provider, { value: value, children: [
                            jsx_runtime_1.jsx(ForceSpecificCursor_1.ForceSpecificCursor, {}), jsx_runtime_1.jsx(scale_lock_1.ScaleLockProvider, { children: jsx_runtime_1.jsxs("div", { style: background, children: [
                                        jsx_runtime_1.jsx(remotion_1.Internals.CompositionRenderErrorContext.Provider, { value: compositionRenderErrorContextValue, children: canvasMounted ? jsx_runtime_1.jsx(MemoRoot, {}) : null }), jsx_runtime_1.jsx(remotion_1.Internals.CanUseRemotionHooksProvider, { children: jsx_runtime_1.jsx(RenderErrorContext_1.RenderErrorContext.Provider, { value: renderErrorContextValue, children: jsx_runtime_1.jsx(EditorContent_1.EditorContent, { readOnlyStudio: readOnlyStudio, children: jsx_runtime_1.jsx(TopPanel_1.TopPanel, { drawRef: setDrawRef, bufferStateDelayInMilliseconds: exports.BUFFER_STATE_DELAY_IN_MILLISECONDS, onMounted: onMounted, readOnlyStudio: readOnlyStudio }) }) }) })
                                    ] }) })
                        ] }), jsx_runtime_1.jsx(Modals_1.Modals, { readOnlyStudio: readOnlyStudio }), jsx_runtime_1.jsx(NotificationCenter_1.NotificationCenter, {})
                ] }) }) }));
    return canvas_capture_enabled_1.CANVAS_CAPTURE_ENABLED ? (jsx_runtime_1.jsx(StudioCanvasCapture_1.StudioCanvasCapture, { density: 1.5, children: editor })) : (editor);
};
exports.Editor = Editor;
