// @remotion/studio/dist/components/EffectPickerModal.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EffectPickerModal = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const studio_shared_1 = require("@remotion/studio-shared");
const react_1 = require("react");
const colors_1 = require("../helpers/colors");
const use_keybinding_1 = require("../helpers/use-keybinding");
const modals_1 = require("../state/modals");
const ContextMenu_1 = require("./ContextMenu");
const effect_drag_and_drop_1 = require("./effect-drag-and-drop");
const effect_picker_search_1 = require("./effect-picker-search");
const layout_1 = require("./layout");
const is_menu_item_1 = require("./Menu/is-menu-item");
const DismissableModal_1 = require("./NewComposition/DismissableModal");
const RemInput_1 = require("./NewComposition/RemInput");
const shared_1 = require("./QuickSwitcher/shared");
const container = {
    width: 500,
};
const content = {
    padding: '12px 16px 10px',
};
const inputStyle = {
    width: '100%',
};
const resultList = {
    height: 320,
    overflowY: 'auto',
    paddingBottom: 10,
};
const noResults = {
    color: colors_1.LIGHT_TEXT,
    fontSize: 13,
    padding: '12px 16px',
};
const resultContainer = {
    cursor: 'pointer',
    display: 'flex',
    flexDirection: 'row',
    alignItems: 'center',
    padding: '7px 16px',
};
const labelContainer = {
    flex: 1,
    minWidth: 0,
};
const label = {
    fontSize: shared_1.QUICK_SWITCHER_RESULT_LABEL_FONT_SIZE,
    overflow: 'hidden',
    textOverflow: 'ellipsis',
    whiteSpace: 'nowrap',
};
const category = {
    color: colors_1.LIGHT_TEXT,
    fontSize: 11,
    flexShrink: 0,
};
const EffectPickerResult = ({ item, selected, onSelected }) => {
    const [hovered, setHovered] = (0, react_1.useState)(false);
    const ref = (0, react_1.useRef)(null);
    (0, shared_1.useScrollIntoViewOnSelected)(ref, selected);
    const style = (0, react_1.useMemo)(() => {
        return {
            ...resultContainer,
            backgroundColor: (0, colors_1.getBackgroundFromHoverState)({ hovered, selected }),
        };
    }, [hovered, selected]);
    const labelStyle = (0, react_1.useMemo)(() => {
        return {
            ...label,
            color: selected || hovered ? 'white' : colors_1.LIGHT_TEXT,
        };
    }, [hovered, selected]);
    const onClick = (0, react_1.useCallback)(() => {
        onSelected(item);
    }, [item, onSelected]);
    const contextMenuValues = (0, react_1.useMemo)(() => {
        const documentationLink = (0, studio_shared_1.getEffectDocumentationLink)(item);
        return [
            {
                type: 'item',
                id: `open-${item.id}-docs`,
                keyHint: null,
                label: 'Open effect docs',
                leftItem: null,
                disabled: false,
                onClick: () => {
                    window.open(documentationLink, '_blank', 'noopener,noreferrer');
                },
                quickSwitcherLabel: null,
                subMenu: null,
                value: `open-${item.id}-docs`,
            },
        ];
    }, [item]);
    return (jsx_runtime_1.jsx(ContextMenu_1.ContextMenu, { values: contextMenuValues, onOpen: null, children: jsx_runtime_1.jsxs("div", { ref: ref, style: style, onClick: onClick, onMouseEnter: () => setHovered(true), onMouseLeave: () => setHovered(false), children: [
                jsx_runtime_1.jsx("div", { style: labelContainer, children: jsx_runtime_1.jsx("div", { style: labelStyle, children: item.label }) }), jsx_runtime_1.jsx(layout_1.Spacing, { x: 1 }), jsx_runtime_1.jsx("div", { style: category, children: item.category })
            ] }) }));
};
const EffectPickerContent = ({ state }) => {
    const { setSelectedModal } = (0, react_1.useContext)(modals_1.ModalsContext);
    const [query, setQuery] = (0, react_1.useState)('');
    const [selectedIndex, setSelectedIndex] = (0, react_1.useState)(0);
    const inputRef = (0, react_1.useRef)(null);
    const keybindings = (0, use_keybinding_1.useKeybinding)();
    const results = (0, react_1.useMemo)(() => {
        return (0, effect_picker_search_1.filterEffectCatalog)({ items: studio_shared_1.EFFECT_CATALOG, query });
    }, [query]);
    const selectedIndexRounded = results.length === 0 ? -1 : (0, shared_1.loopIndex)(selectedIndex, results.length);
    const selectItem = (0, react_1.useCallback)((item) => {
        setSelectedModal(null);
        (0, effect_drag_and_drop_1.addEffectToSequence)({
            clientId: state.clientId,
            effect: item.effect,
            fileName: state.fileName,
            nodePath: state.nodePath,
        });
    }, [setSelectedModal, state.clientId, state.fileName, state.nodePath]);
    const onArrowDown = (0, react_1.useCallback)(() => {
        setSelectedIndex((i) => i + 1);
    }, []);
    const onArrowUp = (0, react_1.useCallback)(() => {
        setSelectedIndex((i) => i - 1);
    }, []);
    const onEnter = (0, react_1.useCallback)(() => {
        if (selectedIndexRounded === -1) {
            return;
        }
        selectItem(results[selectedIndexRounded]);
    }, [results, selectItem, selectedIndexRounded]);
    (0, react_1.useEffect)(() => {
        const downBinding = keybindings.registerKeybinding({
            key: 'ArrowDown',
            callback: onArrowDown,
            commandCtrlKey: false,
            event: 'keydown',
            preventDefault: true,
            triggerIfInputFieldFocused: true,
            keepRegisteredWhenNotHighestContext: false,
        });
        const upBinding = keybindings.registerKeybinding({
            key: 'ArrowUp',
            callback: onArrowUp,
            commandCtrlKey: false,
            event: 'keydown',
            preventDefault: true,
            triggerIfInputFieldFocused: true,
            keepRegisteredWhenNotHighestContext: false,
        });
        const enterBinding = keybindings.registerKeybinding({
            key: 'Enter',
            callback: onEnter,
            commandCtrlKey: false,
            event: 'keydown',
            preventDefault: true,
            triggerIfInputFieldFocused: true,
            keepRegisteredWhenNotHighestContext: false,
        });
        return () => {
            downBinding.unregister();
            upBinding.unregister();
            enterBinding.unregister();
        };
    }, [keybindings, onArrowDown, onArrowUp, onEnter]);
    const onTextChange = (0, react_1.useCallback)((e) => {
        setQuery(e.target.value);
        setSelectedIndex(0);
    }, []);
    return (jsx_runtime_1.jsxs("div", { style: container, children: [
            jsx_runtime_1.jsx("div", { style: content, children: jsx_runtime_1.jsx(RemInput_1.RemotionInput, { ref: inputRef, type: "text", style: inputStyle, autoFocus: true, status: "ok", value: query, onChange: onTextChange, placeholder: "Search effects...", rightAlign: false }) }), jsx_runtime_1.jsx("div", { style: resultList, className: is_menu_item_1.VERTICAL_SCROLLBAR_CLASSNAME, children: results.length === 0 ? (jsx_runtime_1.jsx("div", { style: noResults, children: "No effects found" })) : (results.map((item, i) => {
                    return (jsx_runtime_1.jsx(EffectPickerResult, { item: item, selected: selectedIndexRounded === i, onSelected: selectItem }, item.id));
                })) })
        ] }));
};
const EffectPickerModal = ({ state }) => {
    return (jsx_runtime_1.jsx(DismissableModal_1.DismissableModal, { children: jsx_runtime_1.jsx(EffectPickerContent, { state: state }) }));
};
exports.EffectPickerModal = EffectPickerModal;
