// @remotion/studio/dist/components/effect-drag-and-drop.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.addEffectToSequence = exports.addEffectFromDragData = exports.getEffectDragData = exports.hasExplicitEffectDragType = exports.hasEffectDragType = void 0;
const studio_shared_1 = require("@remotion/studio-shared");
const install_required_package_1 = require("../helpers/install-required-package");
const call_api_1 = require("./call-api");
const NotificationCenter_1 = require("./Notifications/NotificationCenter");
const hasEffectDragType = (dataTransfer) => {
    const types = Array.from(dataTransfer.types);
    const hasExplicitEffectType = types.includes(studio_shared_1.EFFECT_DRAG_MIME_TYPE);
    if (!hasExplicitEffectType && hasImportableAssetDragType(dataTransfer)) {
        return false;
    }
    return types.some((type) => type === studio_shared_1.EFFECT_DRAG_MIME_TYPE || isGenericDragType(type));
};
exports.hasEffectDragType = hasEffectDragType;
const hasExplicitEffectDragType = (dataTransfer) => {
    return Array.from(dataTransfer.types).includes(studio_shared_1.EFFECT_DRAG_MIME_TYPE);
};
exports.hasExplicitEffectDragType = hasExplicitEffectDragType;
const isGenericDragType = (type) => {
    return type === 'application/json' || type === 'text/plain';
};
const hasImportableAssetDragType = (dataTransfer) => {
    return Array.from(dataTransfer.types).some((type) => type === 'Files' ||
        type === studio_shared_1.ASSET_DRAG_MIME_TYPE ||
        type === studio_shared_1.COMPONENT_DRAG_MIME_TYPE ||
        type === studio_shared_1.SFX_DRAG_MIME_TYPE ||
        type === 'text/uri-list' ||
        type === 'text/html');
};
const getEffectDragData = (dataTransfer) => {
    for (const type of [
        studio_shared_1.EFFECT_DRAG_MIME_TYPE,
        'application/json',
        'text/plain',
    ]) {
        const value = dataTransfer.getData(type);
        if (!value) {
            continue;
        }
        const parsed = (0, studio_shared_1.parseEffectDragData)(value);
        if (parsed) {
            return parsed;
        }
    }
    return null;
};
exports.getEffectDragData = getEffectDragData;
const addEffectFromDragData = ({ clientId, dragData, fileName, nodePath, }) => {
    return (0, exports.addEffectToSequence)({
        clientId,
        effect: dragData.effect,
        fileName,
        nodePath,
    });
};
exports.addEffectFromDragData = addEffectFromDragData;
const addEffectToSequence = async ({ clientId, effect, fileName, nodePath, }) => {
    try {
        const requiredPackage = (0, studio_shared_1.getRequiredPackageForEffectImportPath)(effect.importPath);
        await (0, install_required_package_1.installRequiredPackages)(requiredPackage ? [requiredPackage] : []);
        const result = await (0, call_api_1.callApi)('/api/add-effect', {
            fileName,
            sequenceNodePath: nodePath,
            effectName: effect.name,
            effectImportPath: effect.importPath,
            effectConfig: effect.config,
            clientId,
        });
        if (result.success) {
            (0, NotificationCenter_1.showNotification)(`Added ${effect.name}()`, 2000);
        }
        else {
            (0, NotificationCenter_1.showNotification)(result.reason, 4000);
        }
    }
    catch (err) {
        (0, NotificationCenter_1.showNotification)(err.message, 4000);
    }
};
exports.addEffectToSequence = addEffectToSequence;
