// @remotion/studio/dist/components/ExpandedTracksProvider.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ExpandedTracksProvider = exports.ExpandedTracksSetterContext = exports.ExpandedTracksGetterContext = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const react_1 = require("react");
const migrate_expanded_tracks_for_subscription_key_1 = require("../helpers/migrate-expanded-tracks-for-subscription-key");
const timeline_node_path_key_1 = require("../helpers/timeline-node-path-key");
const SESSION_STORAGE_KEY = 'remotion.editor.expandedTracks';
const onlyCollapsedTrackValues = (state) => {
    const result = {};
    for (const [key, value] of Object.entries(state)) {
        if (value === false) {
            result[key] = false;
        }
    }
    return result;
};
const loadExpandedTracks = () => {
    if (typeof window === 'undefined') {
        return {};
    }
    try {
        const raw = window.sessionStorage.getItem(SESSION_STORAGE_KEY);
        if (raw === null) {
            return {};
        }
        const parsed = JSON.parse(raw);
        if (!parsed || typeof parsed !== 'object') {
            return {};
        }
        return onlyCollapsedTrackValues(parsed);
    }
    catch (_a) {
        return {};
    }
};
const persistExpandedTracks = (state) => {
    if (typeof window === 'undefined') {
        return;
    }
    try {
        window.sessionStorage.setItem(SESSION_STORAGE_KEY, JSON.stringify(onlyCollapsedTrackValues(state)));
    }
    catch (_a) {
        // Ignore quota errors or disabled storage.
    }
};
exports.ExpandedTracksGetterContext = (0, react_1.createContext)({
    getIsExpanded: () => {
        throw new Error('ExpandedTracksGetterContext not initialized');
    },
});
exports.ExpandedTracksSetterContext = (0, react_1.createContext)({
    expandParentTracks: () => {
        throw new Error('ExpandedTracksSetterContext not initialized');
    },
    toggleTrack: () => {
        throw new Error('ExpandedTracksSetterContext not initialized');
    },
    migrateExpandedTracksForSubscriptionKey: () => {
        throw new Error('ExpandedTracksSetterContext not initialized');
    },
});
const ExpandedTracksProvider = ({ children }) => {
    const [expandedTracks, setExpandedTracks] = (0, react_1.useState)(loadExpandedTracks);
    const expandParentTracks = (0, react_1.useCallback)((nodePathInfo) => {
        const keysToExpand = [];
        for (let i = 0; i < nodePathInfo.auxiliaryKeys.length; i++) {
            keysToExpand.push((0, timeline_node_path_key_1.timelineNodePathInfoToKey)({
                ...nodePathInfo,
                auxiliaryKeys: nodePathInfo.auxiliaryKeys.slice(0, i),
            }));
        }
        if (keysToExpand.length === 0) {
            return;
        }
        setExpandedTracks((prev) => {
            let changed = false;
            const next = { ...prev };
            for (const key of keysToExpand) {
                if (next[key] === false) {
                    delete next[key];
                    changed = true;
                }
            }
            if (!changed) {
                return prev;
            }
            persistExpandedTracks(next);
            return next;
        });
    }, []);
    const toggleTrack = (0, react_1.useCallback)((nodePathInfo) => {
        setExpandedTracks((prev) => {
            var _a;
            const key = (0, timeline_node_path_key_1.timelineNodePathInfoToKey)(nodePathInfo);
            const next = { ...prev };
            const isExpanded = (_a = next[key]) !== null && _a !== void 0 ? _a : true;
            if (isExpanded) {
                next[key] = false;
            }
            else {
                delete next[key];
            }
            persistExpandedTracks(next);
            return next;
        });
    }, []);
    const migrateExpandedTracks = (0, react_1.useCallback)((oldKey, newKey) => {
        setExpandedTracks((prev) => {
            const next = (0, migrate_expanded_tracks_for_subscription_key_1.migrateExpandedTracksForSubscriptionKey)(prev, oldKey, newKey);
            if (!next) {
                return prev;
            }
            persistExpandedTracks(next);
            return next;
        });
    }, []);
    const getterValue = (0, react_1.useMemo)(() => ({
        getIsExpanded: (nodePathInfo) => {
            var _a;
            return (_a = expandedTracks[(0, timeline_node_path_key_1.timelineNodePathInfoToKey)(nodePathInfo)]) !== null && _a !== void 0 ? _a : true;
        },
    }), [expandedTracks]);
    const setterValue = (0, react_1.useMemo)(() => ({
        expandParentTracks,
        toggleTrack,
        migrateExpandedTracksForSubscriptionKey: migrateExpandedTracks,
    }), [expandParentTracks, toggleTrack, migrateExpandedTracks]);
    return (jsx_runtime_1.jsx(exports.ExpandedTracksSetterContext.Provider, { value: setterValue, children: jsx_runtime_1.jsx(exports.ExpandedTracksGetterContext.Provider, { value: getterValue, children: children }) }));
};
exports.ExpandedTracksProvider = ExpandedTracksProvider;
