// @remotion/studio/dist/components/folder-menu-items.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getFolderMenuItems = void 0;
const no_react_1 = require("remotion/no-react");
const format_file_location_1 = require("../helpers/format-file-location");
const get_folder_id_1 = require("../helpers/get-folder-id");
const open_in_editor_1 = require("../helpers/open-in-editor");
const NotificationCenter_1 = require("./Notifications/NotificationCenter");
const getFolderMenuItems = ({ closeMenu, connectionStatus, folder, readOnlyStudio, resolvedLocation, setSelectedModal, }) => {
    const editorName = window.remotion_editorName;
    const folderId = (0, get_folder_id_1.getFolderId)({
        folderName: folder.name,
        parentName: folder.parent,
    });
    const fileLocation = (0, format_file_location_1.formatFileLocation)({
        location: resolvedLocation,
        root: window.remotion_cwd,
    });
    const showInEditorDisabled = connectionStatus !== 'connected' || !resolvedLocation;
    const copyFileLocationDisabled = !fileLocation;
    const codemodDisabled = readOnlyStudio || !folder.stack;
    return [
        editorName
            ? {
                id: 'show-folder-in-editor',
                keyHint: null,
                label: `Show folder in ${editorName}`,
                leftItem: null,
                onClick: async () => {
                    closeMenu();
                    if (!resolvedLocation) {
                        return;
                    }
                    try {
                        await (0, open_in_editor_1.openOriginalPositionInEditor)(resolvedLocation);
                    }
                    catch (err) {
                        (0, NotificationCenter_1.showNotification)(err.message, 2000);
                    }
                },
                quickSwitcherLabel: `Show folder in ${editorName}`,
                subMenu: null,
                type: 'item',
                value: 'show-folder-in-editor',
                disabled: showInEditorDisabled,
            }
            : null,
        {
            id: 'copy-folder-file-location',
            keyHint: null,
            label: `Copy file location`,
            leftItem: null,
            onClick: () => {
                closeMenu();
                if (!fileLocation) {
                    return;
                }
                navigator.clipboard
                    .writeText(fileLocation)
                    .then(() => {
                    (0, NotificationCenter_1.showNotification)('Copied file location to clipboard', 1000);
                })
                    .catch((err) => {
                    (0, NotificationCenter_1.showNotification)(`Could not copy to clipboard: ${err.message}`, 1000);
                });
            },
            quickSwitcherLabel: 'Copy folder file location',
            subMenu: null,
            type: 'item',
            value: 'copy-folder-file-location',
            disabled: copyFileLocationDisabled,
        },
        editorName || fileLocation
            ? {
                type: 'divider',
                id: 'show-folder-in-editor-divider',
            }
            : null,
        {
            id: 'rename-folder',
            keyHint: null,
            label: `Rename...`,
            leftItem: null,
            onClick: () => {
                closeMenu();
                setSelectedModal({
                    type: 'rename-folder',
                    folderName: folder.name,
                    parentName: folder.parent,
                    stack: folder.stack,
                });
            },
            quickSwitcherLabel: 'Rename folder...',
            subMenu: null,
            type: 'item',
            value: 'rename-folder',
            disabled: codemodDisabled,
        },
        {
            id: 'delete-folder',
            keyHint: null,
            label: `Delete...`,
            leftItem: null,
            onClick: () => {
                closeMenu();
                setSelectedModal({
                    type: 'delete-folder',
                    folderName: folder.name,
                    parentName: folder.parent,
                    stack: folder.stack,
                });
            },
            quickSwitcherLabel: 'Delete folder...',
            subMenu: null,
            type: 'item',
            value: 'delete-folder',
            disabled: codemodDisabled,
        },
        {
            type: 'divider',
            id: 'copy-folder-id-divider',
        },
        {
            id: 'copy-folder-id',
            keyHint: null,
            label: `Copy ID`,
            leftItem: null,
            onClick: () => {
                closeMenu();
                navigator.clipboard
                    .writeText(folderId)
                    .then(() => {
                    (0, NotificationCenter_1.showNotification)('Copied to clipboard', 1000);
                })
                    .catch((err) => {
                    (0, NotificationCenter_1.showNotification)(`Could not copy to clipboard: ${err.message}`, 1000);
                });
            },
            quickSwitcherLabel: 'Copy folder ID',
            subMenu: null,
            type: 'item',
            value: 'copy-folder-id',
            disabled: false,
        },
    ].filter(no_react_1.NoReactInternals.truthy);
};
exports.getFolderMenuItems = getFolderMenuItems;
