// @remotion/studio/dist/components/ForceSpecificCursor.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.ForceSpecificCursor = exports.stopForcingSpecificCursor = exports.forceSpecificCursor = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const react_1 = __importStar(require("react"));
const ref = react_1.default.createRef();
const forceSpecificCursor = (cursor) => {
    if (!ref.current) {
        throw new Error('ForceSpecificCursor is not mounted');
    }
    ref.current.style.cursor = cursor;
    ref.current.style.pointerEvents = 'auto';
};
exports.forceSpecificCursor = forceSpecificCursor;
const stopForcingSpecificCursor = () => {
    if (!ref.current) {
        throw new Error('ForceSpecificCursor is not mounted');
    }
    ref.current.style.cursor = '';
    ref.current.style.pointerEvents = 'none';
};
exports.stopForcingSpecificCursor = stopForcingSpecificCursor;
const Z_INDEX_FORCE_SPECIFIC_CURSOR = 100;
const ForceSpecificCursor = () => {
    const style = (0, react_1.useMemo)(() => {
        return {
            position: 'fixed',
            inset: 0,
            zIndex: Z_INDEX_FORCE_SPECIFIC_CURSOR,
            pointerEvents: 'none',
        };
    }, []);
    return jsx_runtime_1.jsx("div", { ref: ref, style: style });
};
exports.ForceSpecificCursor = ForceSpecificCursor;
