// @remotion/studio/dist/components/GlobalKeybindings.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.GlobalKeybindings = void 0;
const react_1 = require("react");
const remotion_1 = require("remotion");
const calculate_timeline_1 = require("../helpers/calculate-timeline");
const client_id_1 = require("../helpers/client-id");
const show_browser_rendering_1 = require("../helpers/show-browser-rendering");
const timeline_node_path_key_1 = require("../helpers/timeline-node-path-key");
const use_keybinding_1 = require("../helpers/use-keybinding");
const checkerboard_1 = require("../state/checkerboard");
const modals_1 = require("../state/modals");
const AskAiModal_1 = require("./AskAiModal");
const CompositionSelector_1 = require("./CompositionSelector");
const NotificationCenter_1 = require("./Notifications/NotificationCenter");
const TimelineSelection_1 = require("./Timeline/TimelineSelection");
const sequencePropShortcuts = {
    p: 'style.translate',
    r: 'style.rotate',
    s: 'style.scale',
    t: 'style.opacity',
};
const hasOwnProperty = (obj, key) => Object.prototype.hasOwnProperty.call(obj, key);
const GlobalKeybindings = ({ readOnlyStudio }) => {
    const keybindings = (0, use_keybinding_1.useKeybinding)();
    const { setSelectedModal } = (0, react_1.useContext)(modals_1.ModalsContext);
    const { setCheckerboard } = (0, react_1.useContext)(checkerboard_1.CheckerboardContext);
    const currentSelection = (0, TimelineSelection_1.useCurrentTimelineSelectionStateAsRef)();
    const { sequences } = (0, react_1.useContext)(remotion_1.Internals.SequenceManager);
    const videoConfig = remotion_1.Internals.useUnsafeVideoConfig();
    const { overrideIdToNodePathMappings } = (0, react_1.useContext)(remotion_1.Internals.OverrideIdsToNodePathsGettersContext);
    const { type } = (0, react_1.useContext)(client_id_1.StudioServerConnectionCtx).previewServerState;
    const { navigateToNextComposition, navigateToPreviousComposition } = (0, CompositionSelector_1.useCompositionNavigation)();
    const video = remotion_1.Internals.useVideo();
    const timeline = (0, react_1.useMemo)(() => {
        if (videoConfig === null) {
            return [];
        }
        return (0, calculate_timeline_1.calculateTimeline)({
            sequences,
            overrideIdsToNodePaths: overrideIdToNodePathMappings,
        });
    }, [overrideIdToNodePathMappings, sequences, videoConfig]);
    const selectSequenceProp = (0, react_1.useCallback)((fieldKey) => {
        const { selectedItems, selectItems } = currentSelection.current;
        if (selectedItems.length !== 1) {
            return false;
        }
        const [selection] = selectedItems;
        if (selection.type !== 'sequence' && selection.type !== 'sequence-prop') {
            return false;
        }
        const selectedTrackKey = (0, TimelineSelection_1.getTimelineSequenceSelectionKey)(selection.nodePathInfo);
        const track = timeline.find((candidate) => candidate.nodePathInfo !== null &&
            (0, timeline_node_path_key_1.timelineNodePathInfoToKey)(candidate.nodePathInfo) ===
                selectedTrackKey);
        if (!(track === null || track === void 0 ? void 0 : track.sequence.controls)) {
            return false;
        }
        const { schema, currentRuntimeValueDotNotation } = track.sequence.controls;
        if (!hasOwnProperty(schema, fieldKey) &&
            !hasOwnProperty(currentRuntimeValueDotNotation, fieldKey)) {
            return false;
        }
        selectItems([
            {
                type: 'sequence-prop',
                nodePathInfo: {
                    ...selection.nodePathInfo,
                    auxiliaryKeys: ['controls', fieldKey],
                },
                key: fieldKey,
            },
        ], { reveal: true });
        return true;
    }, [currentSelection, timeline]);
    const openRenderModal = (0, react_1.useCallback)(() => {
        if (!video) {
            return;
        }
        if (type !== 'connected' && !show_browser_rendering_1.SHOW_BROWSER_RENDERING && !readOnlyStudio) {
            (0, NotificationCenter_1.showNotification)('Studio server is offline', 2000);
            return;
        }
        const renderButton = document.getElementById('render-modal-button');
        renderButton.click();
    }, [readOnlyStudio, type, video]);
    (0, react_1.useEffect)(() => {
        const onSequencePropKey = (event) => {
            const key = event.key.toLowerCase();
            const fieldKey = sequencePropShortcuts[key];
            if (!fieldKey) {
                return;
            }
            if (selectSequenceProp(fieldKey)) {
                event.preventDefault();
                return;
            }
            if (key === 't') {
                setCheckerboard((c) => !c);
                event.preventDefault();
                return;
            }
            if (key === 'r') {
                openRenderModal();
                event.preventDefault();
            }
        };
        const nKey = keybindings.registerKeybinding({
            event: 'keypress',
            key: 'n',
            callback: () => {
                (0, NotificationCenter_1.showNotification)(`To make a new composition, right-click an existing one and select "Duplicate"`, 5000);
            },
            commandCtrlKey: false,
            preventDefault: true,
            triggerIfInputFieldFocused: false,
            keepRegisteredWhenNotHighestContext: false,
        });
        const cmdKKey = keybindings.registerKeybinding({
            event: 'keydown',
            key: 'k',
            callback: () => {
                setSelectedModal({
                    type: 'quick-switcher',
                    mode: 'compositions',
                    invocationTimestamp: Date.now(),
                });
            },
            triggerIfInputFieldFocused: true,
            keepRegisteredWhenNotHighestContext: false,
            commandCtrlKey: true,
            preventDefault: true,
        });
        const cmdSKey = keybindings.registerKeybinding({
            event: 'keydown',
            key: 's',
            callback: () => {
                (0, NotificationCenter_1.showNotification)('Remotion saves automatically', 2000);
            },
            triggerIfInputFieldFocused: true,
            keepRegisteredWhenNotHighestContext: false,
            commandCtrlKey: true,
            preventDefault: true,
        });
        const cmdIKey = process.env.ASK_AI_ENABLED
            ? keybindings.registerKeybinding({
                event: 'keydown',
                key: 'i',
                callback: () => {
                    var _a;
                    (_a = AskAiModal_1.askAiModalRef.current) === null || _a === void 0 ? void 0 : _a.toggle();
                },
                triggerIfInputFieldFocused: true,
                keepRegisteredWhenNotHighestContext: true,
                commandCtrlKey: true,
                preventDefault: true,
            })
            : null;
        const sequencePropKeys = Object.keys(sequencePropShortcuts).map((key) => keybindings.registerKeybinding({
            event: 'keydown',
            key,
            callback: onSequencePropKey,
            commandCtrlKey: false,
            preventDefault: false,
            triggerIfInputFieldFocused: false,
            keepRegisteredWhenNotHighestContext: false,
        }));
        const questionMark = keybindings.registerKeybinding({
            event: 'keypress',
            key: '?',
            callback: () => {
                setSelectedModal({
                    type: 'quick-switcher',
                    mode: 'docs',
                    invocationTimestamp: Date.now(),
                });
            },
            commandCtrlKey: false,
            preventDefault: true,
            triggerIfInputFieldFocused: false,
            keepRegisteredWhenNotHighestContext: false,
        });
        const pageDown = keybindings.registerKeybinding({
            event: 'keydown',
            key: 'PageDown',
            callback: navigateToNextComposition,
            commandCtrlKey: false,
            preventDefault: true,
            triggerIfInputFieldFocused: false,
            keepRegisteredWhenNotHighestContext: false,
        });
        const pageUp = keybindings.registerKeybinding({
            event: 'keydown',
            key: 'PageUp',
            callback: navigateToPreviousComposition,
            commandCtrlKey: false,
            preventDefault: true,
            triggerIfInputFieldFocused: false,
            keepRegisteredWhenNotHighestContext: false,
        });
        return () => {
            nKey.unregister();
            for (const sequencePropKey of sequencePropKeys) {
                sequencePropKey.unregister();
            }
            questionMark.unregister();
            cmdKKey.unregister();
            cmdSKey.unregister();
            cmdIKey === null || cmdIKey === void 0 ? void 0 : cmdIKey.unregister();
            pageDown.unregister();
            pageUp.unregister();
        };
    }, [
        keybindings,
        openRenderModal,
        selectSequenceProp,
        setCheckerboard,
        setSelectedModal,
        navigateToNextComposition,
        navigateToPreviousComposition,
    ]);
    return null;
};
exports.GlobalKeybindings = GlobalKeybindings;
