// @remotion/studio/dist/components/InspectorSourceLocation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.InspectorSourceLocation = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const react_1 = require("react");
const colors_1 = require("../helpers/colors");
const format_file_location_1 = require("../helpers/format-file-location");
const source_attribution_1 = require("./Timeline/TimelineStack/source-attribution");
const sourceLocationStyle = {
    alignSelf: 'flex-start',
    appearance: 'none',
    backgroundColor: colors_1.BACKGROUND,
    border: 'none',
    boxSizing: 'border-box',
    display: 'block',
    fontFamily: 'sans-serif',
    fontSize: 12,
    height: 18,
    lineHeight: '18px',
    margin: 0,
    maxWidth: '100%',
    overflow: 'hidden',
    padding: 0,
    textAlign: 'left',
    textDecoration: 'none',
    textOverflow: 'ellipsis',
    whiteSpace: 'nowrap',
    width: 'fit-content',
};
const InspectorSourceLocation = ({ location, canOpen, onOpen }) => {
    const [hovered, setHovered] = (0, react_1.useState)(false);
    const validLocation = (0, react_1.useMemo)(() => {
        if (!(location === null || location === void 0 ? void 0 : location.source) || location.line === null) {
            return null;
        }
        return location;
    }, [location]);
    const fileLocation = (0, react_1.useMemo)(() => {
        if (!validLocation) {
            return null;
        }
        return (0, format_file_location_1.formatFileLocation)({
            location: validLocation,
            root: window.remotion_cwd,
        });
    }, [validLocation]);
    const label = (0, react_1.useMemo)(() => {
        if (!validLocation) {
            return null;
        }
        return (0, source_attribution_1.getOriginalSourceAttribution)(validLocation);
    }, [validLocation]);
    const style = (0, react_1.useMemo)(() => {
        return {
            ...sourceLocationStyle,
            color: hovered ? colors_1.LIGHT_COLOR : colors_1.LIGHT_TEXT,
            cursor: canOpen ? 'pointer' : 'default',
        };
    }, [canOpen, hovered]);
    const onClick = (0, react_1.useCallback)(() => {
        if (!canOpen) {
            return;
        }
        onOpen();
    }, [canOpen, onOpen]);
    if (!label) {
        return null;
    }
    return (jsx_runtime_1.jsx("button", { type: "button", style: style, title: fileLocation !== null && fileLocation !== void 0 ? fileLocation : undefined, onClick: onClick, onPointerEnter: () => setHovered(true), onPointerLeave: () => setHovered(false), children: label }));
};
exports.InspectorSourceLocation = InspectorSourceLocation;
