// @remotion/studio/dist/components/load-canvas-content-from-url.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.deriveCanvasContentFromUrl = exports.deriveCanvasContentFromRoute = void 0;
const url_state_1 = require("../helpers/url-state");
const deriveCanvasContentFromRoute = (route) => {
    const substrings = route.split('/').filter(Boolean);
    // CJK-named composition IDs are not automatically reselected after page refresh
    const lastPart = substrings[substrings.length - 1];
    if (substrings[0] === 'assets') {
        if (!lastPart || substrings.length === 1) {
            return null;
        }
        return {
            type: 'asset',
            asset: decodeURIComponent(route.substring('/assets/'.length)),
        };
    }
    if (substrings[0] === 'outputs') {
        return {
            type: 'output',
            path: decodeURIComponent(route.substring('/outputs/'.length)),
        };
    }
    if (lastPart) {
        return {
            type: 'composition',
            compositionId: decodeURIComponent(lastPart),
        };
    }
    return null;
};
exports.deriveCanvasContentFromRoute = deriveCanvasContentFromRoute;
const deriveCanvasContentFromUrl = () => {
    return (0, exports.deriveCanvasContentFromRoute)((0, url_state_1.getRoute)());
};
exports.deriveCanvasContentFromUrl = deriveCanvasContentFromUrl;
