// @remotion/studio/dist/components/selected-outline-uv.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getSelectedUvHandles = exports.roundNumericUvEllipseValue = exports.getUvCoordinateForPoint = exports.getUvEllipseInteractiveControls = exports.getUvHandleConnectionEllipses = exports.getUvHandleConnectionLines = exports.getUvHandlePosition = exports.tuplesEqual = void 0;
exports.constrainUv = constrainUv;
exports.roundUvCoordinate = roundUvCoordinate;
const remotion_1 = require("remotion");
const selected_outline_geometry_1 = require("./selected-outline-geometry");
const timeline_field_utils_1 = require("./Timeline/timeline-field-utils");
const parseUvCoordinate = (value) => {
    if (Array.isArray(value) &&
        value.length === 2 &&
        value.every((item) => typeof item === 'number' && Number.isFinite(item))) {
        return [value[0], value[1]];
    }
    return null;
};
const parseFiniteNumber = (value) => {
    if (typeof value === 'number' && Number.isFinite(value)) {
        return value;
    }
    return null;
};
const getResizeCursor = (degrees) => {
    const normalizedDegrees = ((degrees % 180) + 180) % 180;
    const snappedDegrees = Math.round(normalizedDegrees / 45) * 45;
    if (snappedDegrees === 0 || snappedDegrees === 180) {
        return 'ew-resize';
    }
    if (snappedDegrees === 45) {
        return 'nwse-resize';
    }
    if (snappedDegrees === 90) {
        return 'ns-resize';
    }
    return 'nesw-resize';
};
const getRotatedEllipseUv = ({ center, dimensions, width, height, rotation, angle, radiusOffset = 0, }) => {
    const rotationRadians = (rotation / 180) * Math.PI;
    const cosRotation = Math.cos(rotationRadians);
    const sinRotation = Math.sin(rotationRadians);
    if (dimensions !== null &&
        dimensions !== undefined &&
        dimensions.width > 0 &&
        dimensions.height > 0) {
        const radiusX = width * dimensions.width * 0.5 + radiusOffset * dimensions.width;
        const radiusY = height * dimensions.height * 0.5 + radiusOffset * dimensions.height;
        const pixelLocalX = Math.cos(angle) * radiusX;
        const pixelLocalY = Math.sin(angle) * radiusY;
        const rotatedX = pixelLocalX * cosRotation - pixelLocalY * sinRotation;
        const rotatedY = pixelLocalX * sinRotation + pixelLocalY * cosRotation;
        return [
            center[0] + rotatedX / dimensions.width,
            center[1] + rotatedY / dimensions.height,
        ];
    }
    const radiusU = width / 2 + radiusOffset;
    const radiusV = height / 2 + radiusOffset;
    const localX = Math.cos(angle) * radiusU;
    const localY = Math.sin(angle) * radiusV;
    return [
        center[0] + localX * cosRotation - localY * sinRotation,
        center[1] + localX * sinRotation + localY * cosRotation,
    ];
};
const getUvEllipseValues = (handle) => {
    var _a, _b, _c, _d, _e, _f;
    var _g, _h, _j, _k;
    const { visual } = handle.fieldSchema;
    if ((visual === null || visual === void 0 ? void 0 : visual.type) !== 'ellipse') {
        return null;
    }
    const ellipseMetadata = visual;
    const controls = handle.ellipseControls;
    const width = (_g = controls === null || controls === void 0 ? void 0 : controls.width.value) !== null && _g !== void 0 ? _g : parseFiniteNumber((_a = handle.effectValues) === null || _a === void 0 ? void 0 : _a[ellipseMetadata.width]);
    const height = (_h = controls === null || controls === void 0 ? void 0 : controls.height.value) !== null && _h !== void 0 ? _h : parseFiniteNumber((_b = handle.effectValues) === null || _b === void 0 ? void 0 : _b[ellipseMetadata.height]);
    const rotation = ellipseMetadata.rotation
        ? ((_j = (_c = controls === null || controls === void 0 ? void 0 : controls.rotation) === null || _c === void 0 ? void 0 : _c.value) !== null && _j !== void 0 ? _j : parseFiniteNumber((_d = handle.effectValues) === null || _d === void 0 ? void 0 : _d[ellipseMetadata.rotation]))
        : 0;
    const innerScale = ellipseMetadata.innerScale
        ? ((_k = (_e = controls === null || controls === void 0 ? void 0 : controls.innerScale) === null || _e === void 0 ? void 0 : _e.value) !== null && _k !== void 0 ? _k : parseFiniteNumber((_f = handle.effectValues) === null || _f === void 0 ? void 0 : _f[ellipseMetadata.innerScale]))
        : null;
    if (width === null || height === null || rotation === null) {
        return null;
    }
    return { width, height, rotation, innerScale };
};
const tuplesEqual = (left, right) => {
    if (!Array.isArray(left) || left.length !== 2) {
        return false;
    }
    return left[0] === right[0] && left[1] === right[1];
};
exports.tuplesEqual = tuplesEqual;
const getBilinearUvHandlePosition = (points, uv) => {
    const [tl, tr, br, bl] = points;
    const top = (0, selected_outline_geometry_1.mixPoint)(tl, tr, uv[0]);
    const bottom = (0, selected_outline_geometry_1.mixPoint)(bl, br, uv[0]);
    return (0, selected_outline_geometry_1.mixPoint)(top, bottom, uv[1]);
};
const projectiveEpsilon = 0.000001;
const getProjectiveTransform = (points) => {
    const [tl, tr, br, bl] = points;
    const dx1 = tr.x - br.x;
    const dx2 = bl.x - br.x;
    const dx3 = tl.x - tr.x + br.x - bl.x;
    const dy1 = tr.y - br.y;
    const dy2 = bl.y - br.y;
    const dy3 = tl.y - tr.y + br.y - bl.y;
    let g = 0;
    let h = 0;
    if (Math.abs(dx3) > projectiveEpsilon || Math.abs(dy3) > projectiveEpsilon) {
        const determinant = dx1 * dy2 - dx2 * dy1;
        if (Math.abs(determinant) < projectiveEpsilon) {
            return null;
        }
        g = (dx3 * dy2 - dx2 * dy3) / determinant;
        h = (dx1 * dy3 - dx3 * dy1) / determinant;
    }
    return {
        a: tr.x - tl.x + g * tr.x,
        b: bl.x - tl.x + h * bl.x,
        c: tl.x,
        d: tr.y - tl.y + g * tr.y,
        e: bl.y - tl.y + h * bl.y,
        f: tl.y,
        g,
        h,
    };
};
const applyProjectiveTransform = (transform, uv) => {
    const denominator = transform.g * uv[0] + transform.h * uv[1] + 1;
    return {
        x: (transform.a * uv[0] + transform.b * uv[1] + transform.c) / denominator,
        y: (transform.d * uv[0] + transform.e * uv[1] + transform.f) / denominator,
    };
};
const getUvHandlePosition = (points, uv) => {
    const transform = getProjectiveTransform(points);
    return transform === null
        ? getBilinearUvHandlePosition(points, uv)
        : applyProjectiveTransform(transform, uv);
};
exports.getUvHandlePosition = getUvHandlePosition;
const getUvHandleConnectionLines = ({ handles, points, }) => {
    const handlesByField = new Map(handles.map((handle) => [
        `${handle.effectIndex}\u0000${handle.fieldKey}`,
        handle,
    ]));
    const seenPairs = new Set();
    const lines = [];
    for (const handle of handles) {
        const { visual } = handle.fieldSchema;
        if ((visual === null || visual === void 0 ? void 0 : visual.type) !== 'line' || visual.to === handle.fieldKey) {
            continue;
        }
        const target = handlesByField.get(`${handle.effectIndex}\u0000${visual.to}`);
        if (target === undefined) {
            continue;
        }
        const pairKey = [
            handle.effectIndex,
            ...[handle.fieldKey, visual.to].sort(),
        ].join('\u0000');
        if (seenPairs.has(pairKey)) {
            continue;
        }
        seenPairs.add(pairKey);
        lines.push({
            key: `${handle.effectIndex}-${handle.fieldKey}-${visual.to}`,
            from: (0, exports.getUvHandlePosition)(points, handle.value),
            to: (0, exports.getUvHandlePosition)(points, target.value),
        });
    }
    return lines;
};
exports.getUvHandleConnectionLines = getUvHandleConnectionLines;
const getUvHandleConnectionEllipses = ({ handles, dimensions, points, }) => {
    var _a, _b;
    const ellipses = [];
    for (const handle of handles) {
        const { visual } = handle.fieldSchema;
        if ((visual === null || visual === void 0 ? void 0 : visual.type) === 'ellipse') {
            const ellipseMetadata = visual;
            const ellipseValues = getUvEllipseValues(handle);
            if (ellipseValues === null) {
                continue;
            }
            const rotatedEllipsePoints = [];
            for (let i = 0; i <= 64; i++) {
                const angle = (i / 64) * Math.PI * 2;
                const uv = getRotatedEllipseUv({
                    center: handle.value,
                    dimensions,
                    width: ellipseValues.width,
                    height: ellipseValues.height,
                    rotation: ellipseValues.rotation,
                    angle,
                });
                rotatedEllipsePoints.push((0, exports.getUvHandlePosition)(points, uv));
            }
            ellipses.push({
                key: [
                    handle.effectIndex,
                    handle.fieldKey,
                    ellipseMetadata.width,
                    ellipseMetadata.height,
                    (_a = ellipseMetadata.rotation) !== null && _a !== void 0 ? _a : 'rotation',
                ].join('-'),
                points: rotatedEllipsePoints,
            });
            if (ellipseMetadata.innerScale !== undefined &&
                ellipseValues.innerScale !== null) {
                const startEllipsePoints = [];
                for (let i = 0; i <= 64; i++) {
                    const angle = (i / 64) * Math.PI * 2;
                    const uv = getRotatedEllipseUv({
                        center: handle.value,
                        dimensions,
                        width: ellipseValues.width * ellipseValues.innerScale,
                        height: ellipseValues.height * ellipseValues.innerScale,
                        rotation: ellipseValues.rotation,
                        angle,
                    });
                    startEllipsePoints.push((0, exports.getUvHandlePosition)(points, uv));
                }
                ellipses.push({
                    key: [
                        handle.effectIndex,
                        handle.fieldKey,
                        ellipseMetadata.width,
                        ellipseMetadata.height,
                        (_b = ellipseMetadata.rotation) !== null && _b !== void 0 ? _b : 'rotation',
                        ellipseMetadata.innerScale,
                    ].join('-'),
                    points: startEllipsePoints,
                });
            }
            continue;
        }
    }
    return ellipses;
};
exports.getUvHandleConnectionEllipses = getUvHandleConnectionEllipses;
const getUvEllipseInteractiveControls = ({ handles, dimensions, points, }) => {
    var _a;
    const controls = [];
    for (const handle of handles) {
        if (((_a = handle.fieldSchema.visual) === null || _a === void 0 ? void 0 : _a.type) !== 'ellipse' ||
            handle.ellipseControls === null ||
            handle.ellipseControls === undefined) {
            continue;
        }
        const ellipseValues = getUvEllipseValues(handle);
        if (ellipseValues === null) {
            continue;
        }
        const center = (0, exports.getUvHandlePosition)(points, handle.value);
        const { rotation } = ellipseValues;
        const widthPoint = (0, exports.getUvHandlePosition)(points, getRotatedEllipseUv({
            center: handle.value,
            dimensions,
            width: ellipseValues.width,
            height: 0,
            rotation,
            angle: 0,
        }));
        const heightPoint = (0, exports.getUvHandlePosition)(points, getRotatedEllipseUv({
            center: handle.value,
            dimensions,
            width: 0,
            height: ellipseValues.height,
            rotation,
            angle: Math.PI / 2,
        }));
        const rotationPoint = handle.ellipseControls.rotation === null
            ? null
            : (0, exports.getUvHandlePosition)(points, getRotatedEllipseUv({
                center: handle.value,
                dimensions,
                width: 0,
                height: Math.max(ellipseValues.height, 0.001),
                rotation,
                angle: Math.PI / 2,
                radiusOffset: 0.08,
            }));
        const startPoint = handle.ellipseControls.innerScale === null ||
            ellipseValues.innerScale === null
            ? null
            : (0, exports.getUvHandlePosition)(points, getRotatedEllipseUv({
                center: handle.value,
                dimensions,
                width: ellipseValues.width * ellipseValues.innerScale,
                height: 0,
                rotation,
                angle: 0,
            }));
        controls.push({
            handle,
            center,
            width: ellipseValues.width,
            height: ellipseValues.height,
            rotation,
            resize: [
                {
                    key: `${handle.effectIndex}-${handle.fieldKey}-width`,
                    axis: 'width',
                    field: handle.ellipseControls.width,
                    position: widthPoint,
                    cursor: getResizeCursor(rotation),
                },
                {
                    key: `${handle.effectIndex}-${handle.fieldKey}-height`,
                    axis: 'height',
                    field: handle.ellipseControls.height,
                    position: heightPoint,
                    cursor: getResizeCursor(rotation + 90),
                },
            ],
            rotationControl: handle.ellipseControls.rotation === null || rotationPoint === null
                ? null
                : {
                    key: `${handle.effectIndex}-${handle.fieldKey}-rotation`,
                    field: handle.ellipseControls.rotation,
                    position: rotationPoint,
                    cursor: getResizeCursor(rotation + 90),
                },
            startControl: handle.ellipseControls.innerScale === null || startPoint === null
                ? null
                : {
                    key: `${handle.effectIndex}-${handle.fieldKey}-start`,
                    field: handle.ellipseControls.innerScale,
                    position: startPoint,
                    cursor: getResizeCursor(rotation),
                },
        });
    }
    return controls;
};
exports.getUvEllipseInteractiveControls = getUvEllipseInteractiveControls;
const vectorBetween = (from, to) => {
    return { x: to.x - from.x, y: to.y - from.y };
};
const getBilinearUvCoordinateForPoint = (points, point) => {
    const [tl, tr, br, bl] = points;
    let u = 0.5;
    let v = 0.5;
    for (let i = 0; i < 8; i++) {
        const current = getBilinearUvHandlePosition(points, [u, v]);
        const errorX = current.x - point.x;
        const errorY = current.y - point.y;
        if (Math.abs(errorX) + Math.abs(errorY) < 0.001) {
            break;
        }
        const du = {
            x: (0, selected_outline_geometry_1.mix)(tr.x - tl.x, br.x - bl.x, v),
            y: (0, selected_outline_geometry_1.mix)(tr.y - tl.y, br.y - bl.y, v),
        };
        const dv = vectorBetween((0, selected_outline_geometry_1.mixPoint)(tl, tr, u), (0, selected_outline_geometry_1.mixPoint)(bl, br, u));
        const determinant = du.x * dv.y - du.y * dv.x;
        if (Math.abs(determinant) < 0.000001) {
            break;
        }
        u -= (errorX * dv.y - errorY * dv.x) / determinant;
        v -= (du.x * errorY - du.y * errorX) / determinant;
    }
    return [u, v];
};
const getUvCoordinateForPoint = (points, point) => {
    const transform = getProjectiveTransform(points);
    if (transform === null) {
        return getBilinearUvCoordinateForPoint(points, point);
    }
    const determinant = transform.a * (transform.e - transform.f * transform.h) -
        transform.b * (transform.d - transform.f * transform.g) +
        transform.c * (transform.d * transform.h - transform.e * transform.g);
    if (Math.abs(determinant) < projectiveEpsilon) {
        return getBilinearUvCoordinateForPoint(points, point);
    }
    const inverseA = transform.e - transform.f * transform.h;
    const inverseB = transform.c * transform.h - transform.b;
    const inverseC = transform.b * transform.f - transform.c * transform.e;
    const inverseD = transform.f * transform.g - transform.d;
    const inverseE = transform.a - transform.c * transform.g;
    const inverseF = transform.c * transform.d - transform.a * transform.f;
    const inverseG = transform.d * transform.h - transform.e * transform.g;
    const inverseH = transform.b * transform.g - transform.a * transform.h;
    const inverseI = transform.a * transform.e - transform.b * transform.d;
    const denominator = inverseG * point.x + inverseH * point.y + inverseI;
    if (Math.abs(denominator) < projectiveEpsilon) {
        return getBilinearUvCoordinateForPoint(points, point);
    }
    return [
        (inverseA * point.x + inverseB * point.y + inverseC) / denominator,
        (inverseD * point.x + inverseE * point.y + inverseF) / denominator,
    ];
};
exports.getUvCoordinateForPoint = getUvCoordinateForPoint;
function constrainUv(value, schema) {
    var _a, _b;
    const min = (_a = schema.min) !== null && _a !== void 0 ? _a : -Infinity;
    const max = (_b = schema.max) !== null && _b !== void 0 ? _b : Infinity;
    return [(0, selected_outline_geometry_1.clamp)(value[0], min, max), (0, selected_outline_geometry_1.clamp)(value[1], min, max)];
}
function roundUvCoordinate(value, schema) {
    const decimalPlaces = (0, timeline_field_utils_1.getTimelineDisplayDecimalPlaces)({
        defaultDecimalPlaces: 3,
        step: schema.step,
    });
    return [
        (0, timeline_field_utils_1.roundToDecimalPlaces)(value[0], decimalPlaces),
        (0, timeline_field_utils_1.roundToDecimalPlaces)(value[1], decimalPlaces),
    ];
}
const roundNumericUvEllipseValue = (value, schema) => {
    var _a, _b;
    const min = (_a = schema.min) !== null && _a !== void 0 ? _a : -Infinity;
    const max = (_b = schema.max) !== null && _b !== void 0 ? _b : Infinity;
    const decimalPlaces = (0, timeline_field_utils_1.getTimelineDisplayDecimalPlaces)({
        defaultDecimalPlaces: schema.type === 'rotation-degrees' ? 1 : 3,
        step: schema.step,
    });
    return (0, timeline_field_utils_1.roundToDecimalPlaces)((0, selected_outline_geometry_1.clamp)(value, min, max), decimalPlaces);
};
exports.roundNumericUvEllipseValue = roundNumericUvEllipseValue;
const getNumericUvEllipseControlField = ({ activeSchema, effectStatus, effectValues, fieldKey, }) => {
    const fieldSchema = activeSchema[fieldKey];
    if ((fieldSchema === null || fieldSchema === void 0 ? void 0 : fieldSchema.type) !== 'number' &&
        (fieldSchema === null || fieldSchema === void 0 ? void 0 : fieldSchema.type) !== 'rotation-degrees') {
        return null;
    }
    const propStatus = effectStatus.props[fieldKey];
    if ((propStatus === null || propStatus === void 0 ? void 0 : propStatus.status) !== 'static' && (propStatus === null || propStatus === void 0 ? void 0 : propStatus.status) !== 'keyframed') {
        return null;
    }
    const value = parseFiniteNumber(effectValues[fieldKey]);
    if (value === null) {
        return null;
    }
    return {
        fieldKey,
        fieldSchema,
        fieldDefault: fieldSchema.default,
        propStatus,
        value,
    };
};
const getUvEllipseControls = ({ activeSchema, effectStatus, effectValues, fieldSchema, }) => {
    const { visual } = fieldSchema;
    if ((visual === null || visual === void 0 ? void 0 : visual.type) !== 'ellipse') {
        return null;
    }
    const width = getNumericUvEllipseControlField({
        activeSchema,
        effectStatus,
        effectValues,
        fieldKey: visual.width,
    });
    const height = getNumericUvEllipseControlField({
        activeSchema,
        effectStatus,
        effectValues,
        fieldKey: visual.height,
    });
    const rotation = visual.rotation === undefined
        ? null
        : getNumericUvEllipseControlField({
            activeSchema,
            effectStatus,
            effectValues,
            fieldKey: visual.rotation,
        });
    const innerScale = visual.innerScale === undefined
        ? null
        : getNumericUvEllipseControlField({
            activeSchema,
            effectStatus,
            effectValues,
            fieldKey: visual.innerScale,
        });
    if (width === null || height === null) {
        return null;
    }
    return { width, height, rotation, innerScale };
};
const getSelectedUvHandles = ({ propStatuses, clientId, getEffectDragOverrides, nodePath, selectedEffects, sequence, sourceFrame, }) => {
    if (clientId === null || selectedEffects === undefined) {
        return [];
    }
    const handles = [];
    for (const [effectIndex, selectedFields] of selectedEffects) {
        const effect = sequence.effects[effectIndex];
        if (!effect) {
            continue;
        }
        const effectStatus = remotion_1.Internals.getEffectPropStatusesCtx({
            propStatuses,
            nodePath,
            effectIndex,
        });
        if (effectStatus.type !== 'can-update-effect') {
            continue;
        }
        const shouldShowUvHandles = selectedFields.allFields || selectedFields.fieldKeys.size > 0;
        if (!shouldShowUvHandles) {
            continue;
        }
        const dragOverrides = getEffectDragOverrides(nodePath, effectIndex);
        const activeSchema = remotion_1.Internals.flattenActiveSchema(effect.schema, (key) => {
            const propStatus = effectStatus.props[key];
            if ((propStatus === null || propStatus === void 0 ? void 0 : propStatus.status) !== 'static' &&
                (propStatus === null || propStatus === void 0 ? void 0 : propStatus.status) !== 'keyframed') {
                return undefined;
            }
            return remotion_1.Internals.getEffectiveVisualModeValue({
                propStatus,
                dragOverrideValue: dragOverrides[key],
                defaultValue: undefined,
                frame: sourceFrame,
                shouldResortToDefaultValueIfUndefined: false,
            });
        });
        const effectValues = {};
        for (const [fieldKey, fieldSchema] of Object.entries(activeSchema)) {
            if (fieldSchema.type === 'hidden') {
                continue;
            }
            const propStatus = effectStatus.props[fieldKey];
            if ((propStatus === null || propStatus === void 0 ? void 0 : propStatus.status) !== 'static' &&
                (propStatus === null || propStatus === void 0 ? void 0 : propStatus.status) !== 'keyframed') {
                continue;
            }
            effectValues[fieldKey] = remotion_1.Internals.getEffectiveVisualModeValue({
                propStatus,
                dragOverrideValue: dragOverrides[fieldKey],
                defaultValue: fieldSchema.default,
                frame: sourceFrame,
                shouldResortToDefaultValueIfUndefined: true,
            });
        }
        for (const [fieldKey, fieldSchema] of Object.entries(activeSchema)) {
            if (fieldSchema.type !== 'uv-coordinate') {
                continue;
            }
            const propStatus = effectStatus.props[fieldKey];
            if ((propStatus === null || propStatus === void 0 ? void 0 : propStatus.status) !== 'static' &&
                (propStatus === null || propStatus === void 0 ? void 0 : propStatus.status) !== 'keyframed') {
                continue;
            }
            const effectiveValue = effectValues[fieldKey];
            const value = parseUvCoordinate(effectiveValue);
            if (value === null) {
                continue;
            }
            handles.push({
                clientId,
                propStatus,
                effectIndex,
                effectValues,
                ellipseControls: getUvEllipseControls({
                    activeSchema,
                    effectStatus,
                    effectValues,
                    fieldSchema,
                }),
                fieldDefault: fieldSchema.default,
                fieldKey,
                fieldSchema,
                isSelected: selectedFields.fieldKeys.has(fieldKey),
                nodePath,
                schema: effect.schema,
                sourceFrame,
                value,
            });
        }
    }
    return handles;
};
exports.getSelectedUvHandles = getSelectedUvHandles;
