// @remotion/studio/dist/components/SelectedOutlineUvControls.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.SelectedOutlineUvHandleCircleLayer = exports.SelectedOutlineUvHandleConnectionLayer = exports.getSelectedOutlineUvHandleTimelineSelection = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const react_1 = __importStar(require("react"));
const remotion_1 = require("remotion");
const colors_1 = require("../helpers/colors");
const ForceSpecificCursor_1 = require("./ForceSpecificCursor");
const selected_outline_drag_1 = require("./selected-outline-drag");
const selected_outline_measurement_1 = require("./selected-outline-measurement");
const selected_outline_uv_1 = require("./selected-outline-uv");
const call_add_keyframe_1 = require("./Timeline/call-add-keyframe");
const save_effect_prop_1 = require("./Timeline/save-effect-prop");
const getSvgPointFromPointerEvent = ({ event, rect, }) => {
    return {
        x: event.clientX - rect.left,
        y: event.clientY - rect.top,
    };
};
const uvHandleRadius = 4.25;
const selectedUvHandleRadius = 6.8;
const uvHandleStyle = {
    filter: 'drop-shadow(0 1px 2px rgba(0, 0, 0, 0.28))',
};
const getSelectedOutlineUvHandleTimelineSelection = ({ effectIndex, fieldKey, nodePathInfo, }) => {
    return {
        type: 'sequence-effect-prop',
        nodePathInfo: {
            ...nodePathInfo,
            auxiliaryKeys: ['effects', String(effectIndex), fieldKey],
        },
        i: effectIndex,
        key: fieldKey,
    };
};
exports.getSelectedOutlineUvHandleTimelineSelection = getSelectedOutlineUvHandleTimelineSelection;
const SelectedUvHandleConnectionLines = ({ handles, outline }) => {
    const lines = (0, react_1.useMemo)(() => (0, selected_outline_uv_1.getUvHandleConnectionLines)({ handles, points: outline.points }), [handles, outline.points]);
    return (jsx_runtime_1.jsx(jsx_runtime_1.Fragment, { children: lines.map((line) => (jsx_runtime_1.jsx("line", { x1: line.from.x, y1: line.from.y, x2: line.to.x, y2: line.to.y, stroke: colors_1.BLUE, strokeWidth: 2, vectorEffect: "non-scaling-stroke", pointerEvents: "none" }, line.key))) }));
};
const SelectedUvHandleCircle = ({ handle, nodePathInfo, onDraggingChange, onSelect, outline }) => {
    const { setEffectDragOverrides, clearEffectDragOverrides, setPropStatuses } = (0, react_1.useContext)(remotion_1.Internals.VisualModeSettersContext);
    const position = (0, react_1.useMemo)(() => (0, selected_outline_uv_1.getUvHandlePosition)(outline.points, handle.value), [handle.value, outline.points]);
    const onPointerDown = react_1.default.useCallback((event) => {
        if (event.button !== 0) {
            return;
        }
        event.preventDefault();
        event.stopPropagation();
        const interaction = (0, selected_outline_measurement_1.getOutlineSelectionInteraction)(event);
        onSelect((0, exports.getSelectedOutlineUvHandleTimelineSelection)({
            effectIndex: handle.effectIndex,
            fieldKey: handle.fieldKey,
            nodePathInfo,
        }), interaction);
        if (interaction.shiftKey || interaction.toggleKey) {
            return;
        }
        const svg = event.currentTarget.ownerSVGElement;
        if (svg === null) {
            return;
        }
        const svgRect = svg.getBoundingClientRect();
        const startPointerX = event.clientX;
        const startPointerY = event.clientY;
        let lastValue = null;
        let dragging = false;
        const defaultValue = handle.fieldDefault !== undefined
            ? JSON.stringify(handle.fieldDefault)
            : null;
        const updateFromPointerEvent = (pointerEvent) => {
            const point = getSvgPointFromPointerEvent({
                event: pointerEvent,
                rect: svgRect,
            });
            const rawUv = (0, selected_outline_uv_1.getUvCoordinateForPoint)(outline.points, point);
            const snappedUv = (0, selected_outline_drag_1.snapSelectedOutlineUv)({
                point,
                points: outline.points,
                uv: rawUv,
            });
            const nextValue = (0, selected_outline_uv_1.roundUvCoordinate)((0, selected_outline_uv_1.constrainUv)(snappedUv, handle.fieldSchema), handle.fieldSchema);
            lastValue = nextValue;
            setEffectDragOverrides(handle.nodePath, handle.effectIndex, handle.fieldKey, handle.propStatus.status === 'keyframed'
                ? remotion_1.Internals.makeKeyframedDragOverride({
                    status: handle.propStatus,
                    frame: handle.sourceFrame,
                    value: nextValue,
                })
                : remotion_1.Internals.makeStaticDragOverride(nextValue));
        };
        const updateDragFromPointerEvent = (pointerEvent) => {
            if (!dragging) {
                const deltaX = pointerEvent.clientX - startPointerX;
                const deltaY = pointerEvent.clientY - startPointerY;
                if (!(0, selected_outline_drag_1.isSelectedOutlineDragPastThreshold)({ deltaX, deltaY })) {
                    return;
                }
                dragging = true;
                (0, ForceSpecificCursor_1.forceSpecificCursor)('default');
                onDraggingChange(true);
            }
            updateFromPointerEvent(pointerEvent);
        };
        const onPointerMove = (moveEvent) => {
            moveEvent.preventDefault();
            updateDragFromPointerEvent(moveEvent);
        };
        const onPointerUp = () => {
            window.removeEventListener('pointermove', onPointerMove);
            window.removeEventListener('pointerup', onPointerUp);
            window.removeEventListener('pointercancel', onPointerUp);
            if (dragging) {
                (0, ForceSpecificCursor_1.stopForcingSpecificCursor)();
                onDraggingChange(false);
            }
            const stringifiedValue = lastValue === null ? null : JSON.stringify(lastValue);
            const shouldSave = (() => {
                if (lastValue === null) {
                    return false;
                }
                if (handle.propStatus.status === 'keyframed') {
                    return !(0, selected_outline_uv_1.tuplesEqual)(handle.value, lastValue);
                }
                return (!(0, selected_outline_uv_1.tuplesEqual)(handle.propStatus.codeValue, lastValue) &&
                    !(defaultValue === stringifiedValue &&
                        handle.propStatus.codeValue === undefined));
            })();
            if (!shouldSave) {
                clearEffectDragOverrides(handle.nodePath, handle.effectIndex);
                return;
            }
            (handle.propStatus.status === 'keyframed'
                ? (0, call_add_keyframe_1.callAddEffectKeyframe)({
                    fileName: handle.nodePath.absolutePath,
                    nodePath: handle.nodePath,
                    effectIndex: handle.effectIndex,
                    fieldKey: handle.fieldKey,
                    sourceFrame: handle.sourceFrame,
                    value: lastValue,
                    schema: handle.schema,
                    setPropStatuses,
                    clientId: handle.clientId,
                })
                : (0, save_effect_prop_1.saveEffectProp)({
                    type: 'value',
                    fileName: handle.nodePath.absolutePath,
                    nodePath: handle.nodePath,
                    effectIndex: handle.effectIndex,
                    fieldKey: handle.fieldKey,
                    value: lastValue,
                    defaultValue,
                    schema: handle.schema,
                    setPropStatuses,
                    clientId: handle.clientId,
                })).finally(() => {
                clearEffectDragOverrides(handle.nodePath, handle.effectIndex);
            });
        };
        window.addEventListener('pointermove', onPointerMove);
        window.addEventListener('pointerup', onPointerUp);
        window.addEventListener('pointercancel', onPointerUp);
    }, [
        clearEffectDragOverrides,
        handle,
        nodePathInfo,
        onDraggingChange,
        onSelect,
        outline.points,
        setPropStatuses,
        setEffectDragOverrides,
    ]);
    return (jsx_runtime_1.jsx("circle", { cx: position.x, cy: position.y, r: handle.isSelected ? selectedUvHandleRadius : uvHandleRadius, fill: "white", stroke: colors_1.BLUE, strokeWidth: 2, style: uvHandleStyle, vectorEffect: "non-scaling-stroke", pointerEvents: "all", cursor: "default", onPointerDown: onPointerDown }));
};
const SelectedOutlineUvHandleConnectionLayer = ({ outline, target }) => {
    if (!(target === null || target === void 0 ? void 0 : target.containsSelection) || target.uvHandles.length === 0) {
        return null;
    }
    return (jsx_runtime_1.jsx(SelectedUvHandleConnectionLines, { handles: target.uvHandles, outline: outline }));
};
exports.SelectedOutlineUvHandleConnectionLayer = SelectedOutlineUvHandleConnectionLayer;
const SelectedOutlineUvHandleCircleLayer = ({ onDraggingChange, onSelect, outline, target }) => {
    if (!(target === null || target === void 0 ? void 0 : target.containsSelection) || target.uvHandles.length === 0) {
        return null;
    }
    return (jsx_runtime_1.jsx(jsx_runtime_1.Fragment, { children: target.uvHandles.map((handle) => (jsx_runtime_1.jsx(SelectedUvHandleCircle, { handle: handle, nodePathInfo: target.nodePathInfo, onDraggingChange: onDraggingChange, onSelect: onSelect, outline: outline }, `${handle.effectIndex}-${handle.fieldKey}`))) }));
};
exports.SelectedOutlineUvHandleCircleLayer = SelectedOutlineUvHandleCircleLayer;
