// @remotion/studio/dist/components/StudioCanvasCapture.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.StudioCanvasCapture = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const canvas_capture_1 = require("@remotion/canvas-capture");
const react_1 = require("react");
const use_keybinding_1 = require("../helpers/use-keybinding");
const logCaptureError = (message, err) => {
    // eslint-disable-next-line no-console
    console.error(message, err instanceof Error ? err.message : String(err));
};
const StudioCanvasCapture = ({ children, density }) => {
    const captureRef = (0, react_1.useRef)(null);
    const isSupported = (0, react_1.useMemo)(() => (0, canvas_capture_1.isHtmlInCanvasAvailable)(), []);
    const keybindings = (0, use_keybinding_1.useKeybinding)();
    (0, react_1.useEffect)(() => {
        if (!isSupported) {
            return;
        }
        const binding = keybindings.registerKeybinding({
            event: 'keydown',
            key: 'p',
            commandCtrlKey: true,
            callback: () => {
                var _a;
                (_a = captureRef.current) === null || _a === void 0 ? void 0 : _a.toggleRecording().catch((err) => {
                    logCaptureError('Could not toggle Studio canvas recording:', err);
                });
            },
            preventDefault: true,
            triggerIfInputFieldFocused: true,
            keepRegisteredWhenNotHighestContext: true,
        });
        return () => {
            binding.unregister();
        };
    }, [isSupported, keybindings]);
    return (jsx_runtime_1.jsx(canvas_capture_1.HtmlInCanvasCapture, { ref: captureRef, density: density, filename: "remotion-studio-canvas-recording.webm", children: children }));
};
exports.StudioCanvasCapture = StudioCanvasCapture;
