// @remotion/studio/dist/components/TopPanel.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TopPanel = exports.useResponsiveSidebarStatus = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const react_1 = __importStar(require("react"));
const remotion_1 = require("remotion");
const mobile_layout_1 = require("../helpers/mobile-layout");
const use_breakpoint_1 = require("../helpers/use-breakpoint");
const editor_rulers_1 = require("../state/editor-rulers");
const sidebar_1 = require("../state/sidebar");
const CanvasIfSizeIsAvailable_1 = require("./CanvasIfSizeIsAvailable");
const CurrentCompositionSideEffects_1 = require("./CurrentCompositionSideEffects");
const use_is_ruler_visible_1 = require("./EditorRuler/use-is-ruler-visible");
const ExplorerPanel_1 = require("./ExplorerPanel");
const MobilePanel_1 = __importDefault(require("./MobilePanel"));
const ObserveDefaultPropsContext_1 = require("./ObserveDefaultPropsContext");
const OptionsPanel_1 = require("./OptionsPanel");
const PreviewToolbar_1 = require("./PreviewToolbar");
const SplitterContainer_1 = require("./Splitter/SplitterContainer");
const SplitterElement_1 = require("./Splitter/SplitterElement");
const SplitterHandle_1 = require("./Splitter/SplitterHandle");
const container = {
    height: '100%',
    width: '100%',
    display: 'flex',
    flexDirection: 'column',
    flex: 1,
};
const row = {
    display: 'flex',
    flexDirection: 'row',
    flex: 1,
    minHeight: 0,
};
const useResponsiveSidebarStatus = () => {
    const { sidebarCollapsedStateLeft } = (0, react_1.useContext)(sidebar_1.SidebarContext);
    const responsiveLeftStatus = (0, use_breakpoint_1.useBreakpoint)(1200) ? 'collapsed' : 'expanded';
    const actualStateLeft = (0, react_1.useMemo)(() => {
        if (sidebarCollapsedStateLeft === 'collapsed') {
            return 'collapsed';
        }
        if (sidebarCollapsedStateLeft === 'expanded') {
            return 'expanded';
        }
        return responsiveLeftStatus;
    }, [sidebarCollapsedStateLeft, responsiveLeftStatus]);
    return actualStateLeft;
};
exports.useResponsiveSidebarStatus = useResponsiveSidebarStatus;
const TopPanelInner = ({ readOnlyStudio, onMounted, drawRef, bufferStateDelayInMilliseconds }) => {
    const { setSidebarCollapsedState, sidebarCollapsedStateRight } = (0, react_1.useContext)(sidebar_1.SidebarContext);
    const rulersAreVisible = (0, use_is_ruler_visible_1.useIsRulerVisible)();
    const { canvasContent } = (0, react_1.useContext)(remotion_1.Internals.CompositionManager);
    const actualStateLeft = (0, exports.useResponsiveSidebarStatus)();
    const actualStateRight = (0, react_1.useMemo)(() => {
        if (sidebarCollapsedStateRight === 'collapsed') {
            return 'collapsed';
        }
        return 'expanded';
    }, [sidebarCollapsedStateRight]);
    (0, react_1.useEffect)(() => {
        onMounted();
    }, [onMounted]);
    const canvasContainerStyle = (0, react_1.useMemo)(() => ({
        flex: 1,
        display: 'flex',
        paddingTop: rulersAreVisible ? editor_rulers_1.RULER_WIDTH : 0,
        paddingLeft: rulersAreVisible ? editor_rulers_1.RULER_WIDTH : 0,
    }), [rulersAreVisible]);
    const onCollapseLeft = (0, react_1.useCallback)(() => {
        setSidebarCollapsedState({ left: 'collapsed', right: null });
    }, [setSidebarCollapsedState]);
    const onCollapseRight = (0, react_1.useCallback)(() => {
        setSidebarCollapsedState({ left: null, right: 'collapsed' });
    }, [setSidebarCollapsedState]);
    const isMobileLayout = (0, mobile_layout_1.useMobileLayout)();
    return (jsx_runtime_1.jsx(ObserveDefaultPropsContext_1.ObserveDefaultProps, { compositionId: (canvasContent === null || canvasContent === void 0 ? void 0 : canvasContent.type) === 'composition'
            ? canvasContent.compositionId
            : null, readOnlyStudio: readOnlyStudio, children: jsx_runtime_1.jsxs("div", { style: container, children: [
                jsx_runtime_1.jsx("div", { style: row, children: jsx_runtime_1.jsxs(SplitterContainer_1.SplitterContainer, { minFlex: 0.15, maxFlex: 0.4, defaultFlex: 0.2, id: "sidebar-to-preview", orientation: "vertical", children: [actualStateLeft === 'expanded' ? (isMobileLayout ? (jsx_runtime_1.jsx(MobilePanel_1.default, { onClose: onCollapseLeft, children: jsx_runtime_1.jsx(ExplorerPanel_1.ExplorerPanel, { readOnlyStudio: readOnlyStudio }) })) : (jsx_runtime_1.jsx(SplitterElement_1.SplitterElement, { sticky: null, type: "flexer", children: jsx_runtime_1.jsx(ExplorerPanel_1.ExplorerPanel, { readOnlyStudio: readOnlyStudio }) }))) : null, actualStateLeft === 'expanded' ? (jsx_runtime_1.jsx(SplitterHandle_1.SplitterHandle, { allowToCollapse: "left", onCollapse: onCollapseLeft })) : null, jsx_runtime_1.jsx(SplitterElement_1.SplitterElement, { sticky: null, type: "anti-flexer", children: jsx_runtime_1.jsxs(SplitterContainer_1.SplitterContainer, { minFlex: 0.5, maxFlex: 0.8, defaultFlex: 0.7, id: "canvas-to-right-sidebar", orientation: "vertical", children: [
                                        jsx_runtime_1.jsx(SplitterElement_1.SplitterElement, { sticky: null, type: "flexer", children: jsx_runtime_1.jsx("div", { ref: drawRef, style: canvasContainerStyle, children: jsx_runtime_1.jsx(CanvasIfSizeIsAvailable_1.CanvasIfSizeIsAvailable, {}) }) }), actualStateRight === 'expanded' ? (jsx_runtime_1.jsx(SplitterHandle_1.SplitterHandle, { allowToCollapse: "right", onCollapse: onCollapseRight })) : null, actualStateRight === 'expanded' ? (isMobileLayout ? (jsx_runtime_1.jsx(MobilePanel_1.default, { onClose: onCollapseRight, children: jsx_runtime_1.jsx(OptionsPanel_1.OptionsPanel, { readOnlyStudio: readOnlyStudio }) })) : (jsx_runtime_1.jsx(SplitterElement_1.SplitterElement, { sticky: null, type: "anti-flexer", children: jsx_runtime_1.jsx(OptionsPanel_1.OptionsPanel, { readOnlyStudio: readOnlyStudio }) }))) : null] }) })
                        ] }) }), jsx_runtime_1.jsx(PreviewToolbar_1.PreviewToolbar, { bufferStateDelayInMilliseconds: bufferStateDelayInMilliseconds, readOnlyStudio: readOnlyStudio }), jsx_runtime_1.jsx(CurrentCompositionSideEffects_1.TitleUpdater, {})
            ] }) }));
};
exports.TopPanel = react_1.default.memo(TopPanelInner);
