// @remotion/studio-server/dist/codemods/add-effect.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.addEffect = exports.makeConfigObjectExpression = exports.ensureEffectImport = exports.assertValidEffect = void 0;
const studio_shared_1 = require("@remotion/studio-shared");
const recast = __importStar(require("recast"));
const imports_1 = require("../helpers/imports");
const can_update_sequence_props_1 = require("../preview-server/routes/can-update-sequence-props");
const format_file_content_1 = require("./format-file-content");
const parse_ast_1 = require("./parse-ast");
const update_effect_props_1 = require("./update-effect-props/update-effect-props");
const b = recast.types.builders;
const identifierRegex = /^[A-Za-z_$][0-9A-Za-z_$]*$/;
const assertValidEffect = ({ effectName, effectImportPath, }) => {
    if (!identifierRegex.test(effectName)) {
        throw new Error(`Invalid effect name "${effectName}"`);
    }
    const allowedImport = effectImportPath.startsWith('@remotion/effects/') ||
        effectImportPath === '@remotion/light-leaks' ||
        effectImportPath === '@remotion/starburst';
    if (!allowedImport) {
        throw new Error(`Unsupported effect import "${effectImportPath}"`);
    }
};
exports.assertValidEffect = assertValidEffect;
const parseValueExpression = (value) => {
    const code = `a = ${(0, studio_shared_1.stringifyDefaultProps)({ props: value, enumPaths: [] })}`;
    const ast = (0, parse_ast_1.parseAst)(code);
    const stmt = ast.program.body[0];
    if (stmt.type !== 'ExpressionStatement' ||
        stmt.expression.type !== 'AssignmentExpression') {
        throw new Error('Failed to parse effect config value');
    }
    return stmt.expression.right;
};
const declarationBindsName = (declaration, name) => {
    var _a, _b;
    if (declaration.type === 'ClassDeclaration') {
        return ((_a = declaration.id) === null || _a === void 0 ? void 0 : _a.name) === name;
    }
    if (declaration.type === 'FunctionDeclaration') {
        return ((_b = declaration.id) === null || _b === void 0 ? void 0 : _b.name) === name;
    }
    return declaration.declarations.some((variableDeclaration) => {
        return (variableDeclaration.id.type === 'Identifier' &&
            variableDeclaration.id.name === name);
    });
};
const hasTopLevelBinding = ({ ast, name }) => {
    return ast.program.body.some((node) => {
        var _a;
        if (node.type === 'FunctionDeclaration' ||
            node.type === 'ClassDeclaration' ||
            node.type === 'VariableDeclaration') {
            return declarationBindsName(node, name);
        }
        if (node.type === 'ExportNamedDeclaration' &&
            node.declaration &&
            (node.declaration.type === 'FunctionDeclaration' ||
                node.declaration.type === 'ClassDeclaration' ||
                node.declaration.type === 'VariableDeclaration')) {
            return declarationBindsName(node.declaration, name);
        }
        if (node.type !== 'ImportDeclaration') {
            return false;
        }
        return (_a = node.specifiers) === null || _a === void 0 ? void 0 : _a.some((specifier) => {
            var _a;
            return ((_a = specifier.local) === null || _a === void 0 ? void 0 : _a.name) === name;
        });
    });
};
const getAvailableLocalName = ({ ast, effectName, }) => {
    if (!hasTopLevelBinding({ ast, name: effectName })) {
        return effectName;
    }
    const base = `${effectName}Effect`;
    if (!hasTopLevelBinding({ ast, name: base })) {
        return base;
    }
    for (let i = 2; i < 100; i++) {
        const candidate = `${base}${i}`;
        if (!hasTopLevelBinding({ ast, name: candidate })) {
            return candidate;
        }
    }
    throw new Error(`Cannot find a local name for ${effectName}`);
};
const ensureEffectImport = ({ ast, effectName, effectImportPath, }) => {
    const localName = getAvailableLocalName({ ast, effectName });
    return (0, imports_1.ensureNamedImport)({
        ast,
        importedName: effectName,
        sourcePath: effectImportPath,
        localName,
    });
};
exports.ensureEffectImport = ensureEffectImport;
const getEffectsArray = (attr) => {
    if (!attr.value || attr.value.type !== 'JSXExpressionContainer') {
        throw new Error('Cannot add effect: effects prop is not an array');
    }
    const expr = attr.value.expression;
    if (expr.type !== 'ArrayExpression') {
        throw new Error('Cannot add effect: effects prop is not an array');
    }
    return expr;
};
const makeEffectsAttr = (array) => {
    return b.jsxAttribute(b.jsxIdentifier('effects'), b.jsxExpressionContainer(array));
};
const makeConfigObjectExpression = (config) => {
    return b.objectExpression(Object.entries(config).map(([key, value]) => {
        const keyNode = identifierRegex.test(key)
            ? b.identifier(key)
            : b.stringLiteral(key);
        return b.objectProperty(keyNode, parseValueExpression(value));
    }));
};
exports.makeConfigObjectExpression = makeConfigObjectExpression;
const getJsxTagLabel = (name) => {
    if (name.type === 'JSXIdentifier') {
        return `<${name.name}>`;
    }
    return 'element';
};
const addEffect = async ({ input, sequenceNodePath, effectName, effectImportPath, effectConfig, prettierConfigOverride, }) => {
    var _a;
    var _b, _c;
    (0, exports.assertValidEffect)({ effectName, effectImportPath });
    const ast = (0, parse_ast_1.parseAst)(input);
    const jsx = (0, can_update_sequence_props_1.findJsxElementAtNodePath)(ast, sequenceNodePath);
    if (!jsx) {
        throw new Error('Could not find a JSX element at the specified location to add effect');
    }
    const localName = (0, exports.ensureEffectImport)({ ast, effectName, effectImportPath });
    const effectCall = b.callExpression(b.identifier(localName), [
        (0, exports.makeConfigObjectExpression)(effectConfig),
    ]);
    const attr = (0, update_effect_props_1.findEffectsAttr)((_b = jsx.attributes) !== null && _b !== void 0 ? _b : []);
    if (attr) {
        getEffectsArray(attr).elements.push(effectCall);
    }
    else {
        const effectsArray = b.arrayExpression([effectCall]);
        jsx.attributes.push(makeEffectsAttr(effectsArray));
    }
    const finalFile = (0, parse_ast_1.serializeAst)(ast);
    const { output, formatted } = await (0, format_file_content_1.formatFileContent)({
        input: finalFile,
        prettierConfigOverride,
    });
    return {
        output,
        formatted,
        effectLabel: `${effectName}()`,
        nodeLabel: getJsxTagLabel(jsx.name),
        logLine: (_c = (_a = jsx.loc) === null || _a === void 0 ? void 0 : _a.start.line) !== null && _c !== void 0 ? _c : 1,
    };
};
exports.addEffect = addEffect;
