// @remotion/studio-server/dist/codemods/delete-effect.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.deleteEffect = exports.deleteEffects = void 0;
const can_update_sequence_props_1 = require("../preview-server/routes/can-update-sequence-props");
const format_file_content_1 = require("./format-file-content");
const parse_ast_1 = require("./parse-ast");
const update_effect_props_1 = require("./update-effect-props/update-effect-props");
const getEffectsArray = (attr) => {
    if (!attr.value || attr.value.type !== 'JSXExpressionContainer') {
        throw new Error('Cannot delete effect: effects prop is not an array');
    }
    const expr = attr.value.expression;
    if (expr.type !== 'ArrayExpression') {
        throw new Error('Cannot delete effect: effects prop is not an array');
    }
    return expr;
};
const getAllEffectLabels = (effectsArray, attr) => {
    if (effectsArray.elements.length === 0) {
        throw new Error('Cannot delete effect: no effects found');
    }
    const elements = (0, update_effect_props_1.enumerateEffectArrayElements)(effectsArray);
    return {
        effectLabels: elements.map((effect) => effect.kind === 'call' ? `${effect.callee}()` : 'effect'),
        logLines: elements.map((effect) => {
            var _a, _b, _c;
            var _d, _e, _f;
            return effect.kind === 'call'
                ? ((_e = (_d = (_a = effect.node.loc) === null || _a === void 0 ? void 0 : _a.start.line) !== null && _d !== void 0 ? _d : (_b = attr.loc) === null || _b === void 0 ? void 0 : _b.start.line) !== null && _e !== void 0 ? _e : 1)
                : ((_f = (_c = attr.loc) === null || _c === void 0 ? void 0 : _c.start.line) !== null && _f !== void 0 ? _f : 1);
        }),
    };
};
const deleteEffects = async ({ input, effects, prettierConfigOverride, }) => {
    var _a, _b;
    var _c, _d, _e;
    if (effects.length === 0) {
        throw new Error('No effects were specified for deletion');
    }
    const ast = (0, parse_ast_1.parseAst)(input);
    const deletionsByAttr = new Map();
    for (const effect of effects) {
        const { sequenceNodePath } = effect;
        const jsx = (0, can_update_sequence_props_1.findJsxElementAtNodePath)(ast, sequenceNodePath);
        if (!jsx) {
            throw new Error('Could not find a JSX element at the specified location to delete effect');
        }
        const attr = (0, update_effect_props_1.findEffectsAttr)((_c = jsx.attributes) !== null && _c !== void 0 ? _c : []);
        if (!attr) {
            throw new Error('Could not find effects on the target JSX element');
        }
        const effectsArray = getEffectsArray(attr);
        const existingDeletion = deletionsByAttr.get(attr);
        const deletion = existingDeletion !== null && existingDeletion !== void 0 ? existingDeletion : {
            jsxAttributes: jsx.attributes,
            attr,
            effectsArray,
            allEffects: false,
            effectIndices: new Set(),
            effectLabels: [],
            logLines: [],
        };
        deletionsByAttr.set(attr, deletion);
        if (effect.type === 'all-effects') {
            const { effectLabels, logLines } = getAllEffectLabels(effectsArray, attr);
            deletion.allEffects = true;
            deletion.effectIndices.clear();
            deletion.effectLabels = effectLabels;
            deletion.logLines = logLines;
            continue;
        }
        const { effectIndex } = effect;
        if (deletion.allEffects || deletion.effectIndices.has(effectIndex)) {
            continue;
        }
        const found = (0, update_effect_props_1.findEffectCallExpression)({ attr, effectIndex });
        if (found.kind === 'error') {
            throw new Error(`Cannot delete effect: ${found.reason}`);
        }
        deletion.effectIndices.add(effectIndex);
        deletion.effectLabels.push(`${found.callee}()`);
        deletion.logLines.push((_e = (_d = (_a = found.call.loc) === null || _a === void 0 ? void 0 : _a.start.line) !== null && _d !== void 0 ? _d : (_b = jsx.loc) === null || _b === void 0 ? void 0 : _b.start.line) !== null && _e !== void 0 ? _e : 1);
    }
    for (const deletion of deletionsByAttr.values()) {
        if (deletion.allEffects) {
            const attrIndex = deletion.jsxAttributes.indexOf(deletion.attr);
            if (attrIndex !== -1) {
                deletion.jsxAttributes.splice(attrIndex, 1);
            }
            continue;
        }
        for (const effectIndex of [...deletion.effectIndices].sort((a, b) => b - a)) {
            deletion.effectsArray.elements.splice(effectIndex, 1);
        }
        if (deletion.effectsArray.elements.length === 0) {
            const attrIndex = deletion.jsxAttributes.indexOf(deletion.attr);
            if (attrIndex !== -1) {
                deletion.jsxAttributes.splice(attrIndex, 1);
            }
        }
    }
    const finalFile = (0, parse_ast_1.serializeAst)(ast);
    const { output, formatted } = await (0, format_file_content_1.formatFileContent)({
        input: finalFile,
        prettierConfigOverride,
    });
    return {
        output,
        formatted,
        effectLabels: [...deletionsByAttr.values()].flatMap((deletion) => deletion.effectLabels),
        logLines: [...deletionsByAttr.values()].flatMap((deletion) => deletion.logLines),
    };
};
exports.deleteEffects = deleteEffects;
const deleteEffect = async ({ input, sequenceNodePath, effectIndex, prettierConfigOverride, }) => {
    const { output, formatted, effectLabels, logLines } = await (0, exports.deleteEffects)({
        input,
        effects: [{ type: 'single-effect', sequenceNodePath, effectIndex }],
        prettierConfigOverride,
    });
    return {
        output,
        formatted,
        effectLabel: effectLabels[0],
        logLine: logLines[0],
    };
};
exports.deleteEffect = deleteEffect;
