// @remotion/studio-server/dist/codemods/delete-jsx-node.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.deleteJsxNode = exports.deleteJsxNodes = exports.deleteJsxElementAtPath = exports.findJsxElementPathForDeletion = exports.getJsxElementTagLabel = void 0;
const recast = __importStar(require("recast"));
const get_ast_node_path_1 = require("../helpers/get-ast-node-path");
const format_file_content_1 = require("./format-file-content");
const parse_ast_1 = require("./parse-ast");
const { builders: b, namedTypes } = recast.types;
/** Recast `builders.nullLiteral()` is typed against ast-types; normalize to Babel `Expression`. */
const nullLiteralExpr = () => b.nullLiteral();
const jsxOpeningNameToString = (name) => {
    if (name.type === 'JSXIdentifier') {
        return name.name;
    }
    if (name.type === 'JSXNamespacedName') {
        return `${name.namespace.name}:${name.name.name}`;
    }
    if (name.type === 'JSXMemberExpression') {
        const memberToString = (expr) => {
            if (expr.type === 'JSXIdentifier') {
                return expr.name;
            }
            return `${memberToString(expr.object)}.${expr.property.name}`;
        };
        return memberToString(name);
    }
    return 'Unknown';
};
/** e.g. `<Video>` or `<Remotion.Sequence>` for logs and undo copy. */
const getJsxElementTagLabel = (element) => {
    return `<${jsxOpeningNameToString(element.openingElement.name)}>`;
};
exports.getJsxElementTagLabel = getJsxElementTagLabel;
const findJsxElementPathForDeletion = (ast, nodePath) => {
    const current = (0, get_ast_node_path_1.getAstNodePath)(ast, nodePath);
    if (!current) {
        return null;
    }
    if (namedTypes.JSXOpeningElement.check(current.value)) {
        const parent = current.parentPath;
        if (parent && namedTypes.JSXElement.check(parent.value)) {
            return parent;
        }
        return null;
    }
    if (namedTypes.JSXElement.check(current.value)) {
        return current;
    }
    return null;
};
exports.findJsxElementPathForDeletion = findJsxElementPathForDeletion;
const replaceInLogicalExpression = (parent, node) => {
    const nl = nullLiteralExpr();
    if (parent.left === node) {
        parent.left = nl;
        return true;
    }
    if (parent.right === node) {
        parent.right = nl;
        return true;
    }
    return false;
};
const replaceInConditionalExpression = (parent, node) => {
    const nl = nullLiteralExpr();
    if (parent.consequent === node) {
        parent.consequent = nl;
        return true;
    }
    if (parent.alternate === node) {
        parent.alternate = nl;
        return true;
    }
    return false;
};
const replaceInArrowFunctionExpression = (parent, node) => {
    if (parent.body === node) {
        parent.body = nullLiteralExpr();
        return true;
    }
    return false;
};
const replaceInReturnStatement = (parent, node) => {
    if (parent.argument === node) {
        parent.argument = nullLiteralExpr();
        return true;
    }
    return false;
};
const replaceInCallExpression = (parent, node) => {
    const idx = parent.arguments.indexOf(node);
    if (idx !== -1) {
        parent.arguments[idx] = nullLiteralExpr();
        return true;
    }
    return false;
};
const replaceInOptionalCallExpression = (parent, node) => {
    const idx = parent.arguments.indexOf(node);
    if (idx !== -1) {
        parent.arguments[idx] = nullLiteralExpr();
        return true;
    }
    return false;
};
const replaceInArrayExpression = (parent, node) => {
    const idx = parent.elements.indexOf(node);
    if (idx !== -1) {
        parent.elements[idx] = nullLiteralExpr();
        return true;
    }
    return false;
};
const replaceInAssignmentExpression = (parent, node) => {
    if (parent.right === node) {
        parent.right = nullLiteralExpr();
        return true;
    }
    return false;
};
const replaceInVariableDeclarator = (parent, node) => {
    if (parent.init === node) {
        parent.init = nullLiteralExpr();
        return true;
    }
    return false;
};
const replaceInExportDefaultDeclaration = (parent, node) => {
    if (parent.declaration === node) {
        parent.declaration = nullLiteralExpr();
        return true;
    }
    return false;
};
const replaceInSequenceExpression = (parent, node) => {
    const idx = parent.expressions.indexOf(node);
    if (idx !== -1) {
        parent.expressions[idx] = nullLiteralExpr();
        return true;
    }
    return false;
};
const replaceInNewExpression = (parent, node) => {
    const idx = parent.arguments.indexOf(node);
    if (idx !== -1) {
        parent.arguments[idx] = nullLiteralExpr();
        return true;
    }
    return false;
};
const replaceInExpressionStatement = (parent, node) => {
    if (parent.expression === node) {
        parent.expression = nullLiteralExpr();
        return true;
    }
    return false;
};
const replaceInParenthesizedExpression = (parent, node) => {
    if (parent.expression === node) {
        parent.expression = nullLiteralExpr();
        return true;
    }
    return false;
};
const replaceInJSXExpressionContainer = (parent, node) => {
    if (parent.expression === node) {
        parent.expression = nullLiteralExpr();
        return true;
    }
    return false;
};
const replaceInTSAsExpression = (parent, node) => {
    if (parent.expression === node) {
        parent.expression = nullLiteralExpr();
        return true;
    }
    return false;
};
const replaceInJSXParentChildren = (parent, node) => {
    const idx = parent.children.indexOf(node);
    if (idx !== -1) {
        parent.children.splice(idx, 1);
        return true;
    }
    return false;
};
const replaceNodeWithNull = (parentNode, node) => {
    if (namedTypes.LogicalExpression.check(parentNode)) {
        return replaceInLogicalExpression(parentNode, node);
    }
    if (namedTypes.ConditionalExpression.check(parentNode)) {
        return replaceInConditionalExpression(parentNode, node);
    }
    if (namedTypes.ArrowFunctionExpression.check(parentNode)) {
        return replaceInArrowFunctionExpression(parentNode, node);
    }
    if (namedTypes.ReturnStatement.check(parentNode)) {
        return replaceInReturnStatement(parentNode, node);
    }
    if (namedTypes.CallExpression.check(parentNode)) {
        return replaceInCallExpression(parentNode, node);
    }
    if (parentNode.type === 'OptionalCallExpression') {
        return replaceInOptionalCallExpression(parentNode, node);
    }
    if (namedTypes.ArrayExpression.check(parentNode)) {
        return replaceInArrayExpression(parentNode, node);
    }
    if (namedTypes.AssignmentExpression.check(parentNode)) {
        return replaceInAssignmentExpression(parentNode, node);
    }
    if (namedTypes.VariableDeclarator.check(parentNode)) {
        return replaceInVariableDeclarator(parentNode, node);
    }
    if (namedTypes.ExportDefaultDeclaration.check(parentNode)) {
        return replaceInExportDefaultDeclaration(parentNode, node);
    }
    if (namedTypes.SequenceExpression.check(parentNode)) {
        return replaceInSequenceExpression(parentNode, node);
    }
    if (namedTypes.NewExpression.check(parentNode)) {
        return replaceInNewExpression(parentNode, node);
    }
    if (namedTypes.ExpressionStatement.check(parentNode)) {
        return replaceInExpressionStatement(parentNode, node);
    }
    if (namedTypes.ParenthesizedExpression.check(parentNode)) {
        return replaceInParenthesizedExpression(parentNode, node);
    }
    if (namedTypes.JSXExpressionContainer.check(parentNode)) {
        return replaceInJSXExpressionContainer(parentNode, node);
    }
    if (namedTypes.TSAsExpression.check(parentNode)) {
        return replaceInTSAsExpression(parentNode, node);
    }
    if (namedTypes.JSXElement.check(parentNode)) {
        return replaceInJSXParentChildren(parentNode, node);
    }
    if (namedTypes.JSXFragment.check(parentNode)) {
        return replaceInJSXParentChildren(parentNode, node);
    }
    return false;
};
const deleteJsxElementAtPath = (jsxPath) => {
    const { node, parentPath } = jsxPath;
    if (!parentPath) {
        throw new Error('Cannot delete JSX element with no parent');
    }
    const jsxNode = node;
    if (replaceNodeWithNull(parentPath.node, jsxNode)) {
        return;
    }
    // Recast can replace this node in arbitrary parent contexts.
    jsxPath.replace(b.nullLiteral());
};
exports.deleteJsxElementAtPath = deleteJsxElementAtPath;
const deleteJsxNodes = async ({ input, nodePaths, prettierConfigOverride, }) => {
    if (nodePaths.length === 0) {
        throw new Error('No JSX nodes were specified for deletion');
    }
    const ast = (0, parse_ast_1.parseAst)(input);
    const pathsToDelete = nodePaths.map((nodePath) => {
        var _a, _b;
        var _c, _d;
        const jsxPath = (0, exports.findJsxElementPathForDeletion)(ast, nodePath);
        if (!jsxPath) {
            throw new Error('Could not find a JSX element at the specified location to delete');
        }
        const jsxElement = jsxPath.node;
        return {
            jsxPath,
            nodeLabel: (0, exports.getJsxElementTagLabel)(jsxElement),
            logLine: (_d = (_c = (_a = jsxElement.openingElement.loc) === null || _a === void 0 ? void 0 : _a.start.line) !== null && _c !== void 0 ? _c : (_b = jsxElement.loc) === null || _b === void 0 ? void 0 : _b.start.line) !== null && _d !== void 0 ? _d : 1,
        };
    });
    for (const { jsxPath } of pathsToDelete) {
        (0, exports.deleteJsxElementAtPath)(jsxPath);
    }
    const finalFile = (0, parse_ast_1.serializeAst)(ast);
    const { output, formatted } = await (0, format_file_content_1.formatFileContent)({
        input: finalFile,
        prettierConfigOverride,
    });
    return {
        output,
        formatted,
        nodeLabels: pathsToDelete.map(({ nodeLabel }) => nodeLabel),
        logLines: pathsToDelete.map(({ logLine }) => logLine),
    };
};
exports.deleteJsxNodes = deleteJsxNodes;
const deleteJsxNode = async ({ input, nodePath, prettierConfigOverride, }) => {
    const { output, formatted, nodeLabels, logLines } = await (0, exports.deleteJsxNodes)({
        input,
        nodePaths: [nodePath],
        prettierConfigOverride,
    });
    return {
        output,
        formatted,
        nodeLabel: nodeLabels[0],
        logLine: logLines[0],
    };
};
exports.deleteJsxNode = deleteJsxNode;
