// @remotion/studio-server/dist/codemods/duplicate-jsx-node.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.duplicateJsxNode = exports.duplicateJsxElementAtPath = void 0;
const types_1 = require("@babel/types");
const recast = __importStar(require("recast"));
const delete_jsx_node_1 = require("./delete-jsx-node");
const format_file_content_1 = require("./format-file-content");
const parse_ast_1 = require("./parse-ast");
const { namedTypes } = recast.types;
const makeFragment = (first, second) => ({
    type: 'JSXFragment',
    openingFragment: { type: 'JSXOpeningFragment' },
    closingFragment: { type: 'JSXClosingFragment' },
    children: [first, second],
});
const uniquifyNamePropOnClone = (clone) => {
    for (const attr of clone.openingElement.attributes) {
        if (attr.type !== 'JSXAttribute' || attr.name.type !== 'JSXIdentifier') {
            continue;
        }
        if (attr.name.name !== 'name') {
            continue;
        }
        if (!attr.value) {
            continue;
        }
        if (attr.value.type === 'StringLiteral') {
            attr.value.value = `${attr.value.value}-copy`;
            return;
        }
        if (attr.value.type === 'JSXExpressionContainer' &&
            attr.value.expression.type === 'StringLiteral') {
            attr.value.expression.value = `${attr.value.expression.value}-copy`;
            return;
        }
    }
};
const duplicateInLogicalExpression = (parent, node, clone) => {
    const frag = makeFragment(node, clone);
    if (parent.left === node) {
        parent.left = frag;
        return true;
    }
    if (parent.right === node) {
        parent.right = frag;
        return true;
    }
    return false;
};
const duplicateInConditionalExpression = (parent, node, clone) => {
    const frag = makeFragment(node, clone);
    if (parent.consequent === node) {
        parent.consequent = frag;
        return true;
    }
    if (parent.alternate === node) {
        parent.alternate = frag;
        return true;
    }
    return false;
};
const duplicateInArrowFunctionExpression = (parent, node, clone) => {
    const frag = makeFragment(node, clone);
    if (parent.body === node) {
        parent.body = frag;
        return true;
    }
    return false;
};
const duplicateInReturnStatement = (parent, node, clone) => {
    const frag = makeFragment(node, clone);
    if (parent.argument === node) {
        parent.argument = frag;
        return true;
    }
    return false;
};
const duplicateInCallExpression = (parent, node, clone) => {
    const idx = parent.arguments.indexOf(node);
    if (idx !== -1) {
        parent.arguments.splice(idx + 1, 0, clone);
        return true;
    }
    return false;
};
const duplicateInOptionalCallExpression = (parent, node, clone) => {
    const idx = parent.arguments.indexOf(node);
    if (idx !== -1) {
        parent.arguments.splice(idx + 1, 0, clone);
        return true;
    }
    return false;
};
const duplicateInArrayExpression = (parent, node, clone) => {
    const idx = parent.elements.indexOf(node);
    if (idx !== -1) {
        parent.elements.splice(idx + 1, 0, clone);
        return true;
    }
    return false;
};
const duplicateInAssignmentExpression = (parent, node, clone) => {
    const frag = makeFragment(node, clone);
    if (parent.right === node) {
        parent.right = frag;
        return true;
    }
    return false;
};
const duplicateInVariableDeclarator = (parent, node, clone) => {
    const frag = makeFragment(node, clone);
    if (parent.init === node) {
        parent.init = frag;
        return true;
    }
    return false;
};
const duplicateInExportDefaultDeclaration = (parent, node, clone) => {
    const frag = makeFragment(node, clone);
    if (parent.declaration === node) {
        parent.declaration =
            frag;
        return true;
    }
    return false;
};
const duplicateInSequenceExpression = (parent, node, clone) => {
    const idx = parent.expressions.indexOf(node);
    if (idx !== -1) {
        parent.expressions.splice(idx + 1, 0, clone);
        return true;
    }
    return false;
};
const duplicateInNewExpression = (parent, node, clone) => {
    const idx = parent.arguments.indexOf(node);
    if (idx !== -1) {
        parent.arguments.splice(idx + 1, 0, clone);
        return true;
    }
    return false;
};
const duplicateInExpressionStatement = (parent, node, clone) => {
    const frag = makeFragment(node, clone);
    if (parent.expression === node) {
        parent.expression = frag;
        return true;
    }
    return false;
};
const duplicateInParenthesizedExpression = (parent, node, clone) => {
    const frag = makeFragment(node, clone);
    if (parent.expression === node) {
        parent.expression = frag;
        return true;
    }
    return false;
};
const duplicateInJSXExpressionContainer = (parent, node, clone) => {
    const frag = makeFragment(node, clone);
    if (parent.expression === node) {
        parent.expression = frag;
        return true;
    }
    return false;
};
const duplicateInTSAsExpression = (parent, node, clone) => {
    const frag = makeFragment(node, clone);
    if (parent.expression === node) {
        parent.expression = frag;
        return true;
    }
    return false;
};
const duplicateInJSXParentChildren = (parent, node, clone) => {
    const idx = parent.children.indexOf(node);
    if (idx !== -1) {
        parent.children.splice(idx + 1, 0, clone);
        return true;
    }
    return false;
};
const insertDuplicateForParent = (parentNode, node, clone) => {
    if (namedTypes.LogicalExpression.check(parentNode)) {
        return duplicateInLogicalExpression(parentNode, node, clone);
    }
    if (namedTypes.ConditionalExpression.check(parentNode)) {
        return duplicateInConditionalExpression(parentNode, node, clone);
    }
    if (namedTypes.ArrowFunctionExpression.check(parentNode)) {
        return duplicateInArrowFunctionExpression(parentNode, node, clone);
    }
    if (namedTypes.ReturnStatement.check(parentNode)) {
        return duplicateInReturnStatement(parentNode, node, clone);
    }
    if (namedTypes.CallExpression.check(parentNode)) {
        return duplicateInCallExpression(parentNode, node, clone);
    }
    if (parentNode.type === 'OptionalCallExpression') {
        return duplicateInOptionalCallExpression(parentNode, node, clone);
    }
    if (namedTypes.ArrayExpression.check(parentNode)) {
        return duplicateInArrayExpression(parentNode, node, clone);
    }
    if (namedTypes.AssignmentExpression.check(parentNode)) {
        return duplicateInAssignmentExpression(parentNode, node, clone);
    }
    if (namedTypes.VariableDeclarator.check(parentNode)) {
        return duplicateInVariableDeclarator(parentNode, node, clone);
    }
    if (namedTypes.ExportDefaultDeclaration.check(parentNode)) {
        return duplicateInExportDefaultDeclaration(parentNode, node, clone);
    }
    if (namedTypes.SequenceExpression.check(parentNode)) {
        return duplicateInSequenceExpression(parentNode, node, clone);
    }
    if (namedTypes.NewExpression.check(parentNode)) {
        return duplicateInNewExpression(parentNode, node, clone);
    }
    if (namedTypes.ExpressionStatement.check(parentNode)) {
        return duplicateInExpressionStatement(parentNode, node, clone);
    }
    if (namedTypes.ParenthesizedExpression.check(parentNode)) {
        return duplicateInParenthesizedExpression(parentNode, node, clone);
    }
    if (namedTypes.JSXExpressionContainer.check(parentNode)) {
        return duplicateInJSXExpressionContainer(parentNode, node, clone);
    }
    if (namedTypes.TSAsExpression.check(parentNode)) {
        return duplicateInTSAsExpression(parentNode, node, clone);
    }
    if (namedTypes.JSXElement.check(parentNode)) {
        return duplicateInJSXParentChildren(parentNode, node, clone);
    }
    if (namedTypes.JSXFragment.check(parentNode)) {
        return duplicateInJSXParentChildren(parentNode, node, clone);
    }
    return false;
};
const duplicateJsxElementAtPath = (jsxPath) => {
    const { node, parentPath } = jsxPath;
    if (!parentPath) {
        throw new Error('Cannot duplicate JSX element with no parent');
    }
    const jsxNode = node;
    const clone = (0, types_1.cloneNode)(jsxNode, true);
    uniquifyNamePropOnClone(clone);
    if (insertDuplicateForParent(parentPath.node, jsxNode, clone)) {
        return;
    }
    jsxPath.replace(makeFragment(jsxNode, clone));
};
exports.duplicateJsxElementAtPath = duplicateJsxElementAtPath;
const duplicateJsxNode = async ({ input, nodePath, prettierConfigOverride, }) => {
    var _a, _b;
    var _c, _d;
    const ast = (0, parse_ast_1.parseAst)(input);
    const jsxPath = (0, delete_jsx_node_1.findJsxElementPathForDeletion)(ast, nodePath);
    if (!jsxPath) {
        throw new Error('Could not find a JSX element at the specified location to duplicate');
    }
    const jsxElement = jsxPath.node;
    const nodeLabel = (0, delete_jsx_node_1.getJsxElementTagLabel)(jsxElement);
    const logLine = (_d = (_c = (_a = jsxElement.openingElement.loc) === null || _a === void 0 ? void 0 : _a.start.line) !== null && _c !== void 0 ? _c : (_b = jsxElement.loc) === null || _b === void 0 ? void 0 : _b.start.line) !== null && _d !== void 0 ? _d : 1;
    (0, exports.duplicateJsxElementAtPath)(jsxPath);
    const finalFile = (0, parse_ast_1.serializeAst)(ast);
    const { output, formatted } = await (0, format_file_content_1.formatFileContent)({
        input: finalFile,
        prettierConfigOverride,
    });
    return {
        output,
        formatted,
        nodeLabel,
        logLine,
    };
};
exports.duplicateJsxNode = duplicateJsxNode;
