// @remotion/studio-server/dist/codemods/duplicate-effect.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.duplicateEffect = exports.duplicateEffects = void 0;
const types_1 = require("@babel/types");
const can_update_sequence_props_1 = require("../preview-server/routes/can-update-sequence-props");
const format_file_content_1 = require("./format-file-content");
const parse_ast_1 = require("./parse-ast");
const update_effect_props_1 = require("./update-effect-props/update-effect-props");
const getEffectsArray = (attr) => {
    if (!attr.value || attr.value.type !== 'JSXExpressionContainer') {
        throw new Error('Cannot duplicate effect: effects prop is not an array');
    }
    const expr = attr.value.expression;
    if (expr.type !== 'ArrayExpression') {
        throw new Error('Cannot duplicate effect: effects prop is not an array');
    }
    return expr;
};
const duplicateEffects = async ({ input, effects, prettierConfigOverride, }) => {
    var _a, _b;
    var _c, _d, _e;
    if (effects.length === 0) {
        throw new Error('No effects were specified for duplication');
    }
    const ast = (0, parse_ast_1.parseAst)(input);
    const duplicationsByAttr = new Map();
    for (const effect of effects) {
        const { sequenceNodePath, effectIndex } = effect;
        const jsx = (0, can_update_sequence_props_1.findJsxElementAtNodePath)(ast, sequenceNodePath);
        if (!jsx) {
            throw new Error('Could not find a JSX element at the specified location to duplicate effect');
        }
        const attr = (0, update_effect_props_1.findEffectsAttr)((_c = jsx.attributes) !== null && _c !== void 0 ? _c : []);
        if (!attr) {
            throw new Error('Could not find effects on the target JSX element');
        }
        const effectsArray = getEffectsArray(attr);
        const existingDuplication = duplicationsByAttr.get(attr);
        const duplication = existingDuplication !== null && existingDuplication !== void 0 ? existingDuplication : {
            effectsArray,
            effectIndices: new Set(),
            effectLabels: [],
            logLines: [],
        };
        duplicationsByAttr.set(attr, duplication);
        if (duplication.effectIndices.has(effectIndex)) {
            continue;
        }
        const found = (0, update_effect_props_1.findEffectCallExpression)({ attr, effectIndex });
        if (found.kind === 'error') {
            throw new Error(`Cannot duplicate effect: ${found.reason}`);
        }
        duplication.effectIndices.add(effectIndex);
        duplication.effectLabels.push(`${found.callee}()`);
        duplication.logLines.push((_e = (_d = (_a = found.call.loc) === null || _a === void 0 ? void 0 : _a.start.line) !== null && _d !== void 0 ? _d : (_b = jsx.loc) === null || _b === void 0 ? void 0 : _b.start.line) !== null && _e !== void 0 ? _e : 1);
    }
    for (const duplication of duplicationsByAttr.values()) {
        for (const effectIndex of [...duplication.effectIndices].sort((a, b) => b - a)) {
            const effect = duplication.effectsArray.elements[effectIndex];
            if (!effect || effect.type !== 'CallExpression') {
                throw new Error('Cannot duplicate effect: not-call-expression');
            }
            duplication.effectsArray.elements.splice(effectIndex + 1, 0, (0, types_1.cloneNode)(effect, true));
        }
    }
    const finalFile = (0, parse_ast_1.serializeAst)(ast);
    const { output, formatted } = await (0, format_file_content_1.formatFileContent)({
        input: finalFile,
        prettierConfigOverride,
    });
    return {
        output,
        formatted,
        effectLabels: [...duplicationsByAttr.values()].flatMap((duplication) => duplication.effectLabels),
        logLines: [...duplicationsByAttr.values()].flatMap((duplication) => duplication.logLines),
    };
};
exports.duplicateEffects = duplicateEffects;
const duplicateEffect = async ({ input, sequenceNodePath, effectIndex, prettierConfigOverride, }) => {
    const { output, formatted, effectLabels, logLines } = await (0, exports.duplicateEffects)({
        input,
        effects: [{ sequenceNodePath, effectIndex }],
        prettierConfigOverride,
    });
    return {
        output,
        formatted,
        effectLabel: effectLabels[0],
        logLine: logLines[0],
    };
};
exports.duplicateEffect = duplicateEffect;
