// @remotion/studio-server/dist/codemods/effect-param-expression.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.makeParamExpression = exports.ensureRemotionImportLocalNames = exports.getRequiredRemotionImports = exports.getRequiredRemotionImportsForEffectParams = void 0;
const recast = __importStar(require("recast"));
const imports_1 = require("../helpers/imports");
const update_nested_prop_1 = require("./update-nested-prop");
const b = recast.types.builders;
const isNonLinearEasing = (easing) => easing.type !== 'linear';
const keyframedParamNeedsEasingImport = (param) => param.easing.some(isNonLinearEasing);
const getRequiredRemotionImportsForEffectParams = (params) => {
    const requiredImports = new Set();
    for (const param of params) {
        if (param.type !== 'keyframed') {
            continue;
        }
        requiredImports.add('useCurrentFrame');
        requiredImports.add(param.interpolationFunction);
        if (keyframedParamNeedsEasingImport(param)) {
            requiredImports.add('Easing');
        }
    }
    return requiredImports;
};
exports.getRequiredRemotionImportsForEffectParams = getRequiredRemotionImportsForEffectParams;
const getRequiredRemotionImports = (effects) => {
    return (0, exports.getRequiredRemotionImportsForEffectParams)(effects.flatMap((effect) => Object.values(effect.params)));
};
exports.getRequiredRemotionImports = getRequiredRemotionImports;
const ensureRemotionImportLocalNames = ({ ast, requiredImports, }) => {
    const localNames = {};
    for (const importedName of requiredImports) {
        localNames[importedName] = (0, imports_1.ensureNamedImport)({
            ast,
            importedName,
            sourcePath: 'remotion',
            localName: importedName,
        });
    }
    return localNames;
};
exports.ensureRemotionImportLocalNames = ensureRemotionImportLocalNames;
const makeEasingExpression = ({ easing, easingLocalName, }) => {
    switch (easing.type) {
        case 'linear':
            return b.memberExpression(b.identifier(easingLocalName), b.identifier('linear'));
        case 'spring':
            return b.callExpression(b.memberExpression(b.identifier(easingLocalName), b.identifier('spring')), [
                b.objectExpression([
                    b.objectProperty(b.identifier('damping'), (0, update_nested_prop_1.parseValueExpression)(easing.damping)),
                    b.objectProperty(b.identifier('mass'), (0, update_nested_prop_1.parseValueExpression)(easing.mass)),
                    b.objectProperty(b.identifier('stiffness'), (0, update_nested_prop_1.parseValueExpression)(easing.stiffness)),
                    ...(easing.allowTail === null
                        ? []
                        : [
                            b.objectProperty(b.identifier('allowTail'), b.booleanLiteral(easing.allowTail)),
                        ]),
                    ...(easing.durationRestThreshold === null
                        ? []
                        : [
                            b.objectProperty(b.identifier('durationRestThreshold'), (0, update_nested_prop_1.parseValueExpression)(easing.durationRestThreshold)),
                        ]),
                    b.objectProperty(b.identifier('overshootClamping'), b.booleanLiteral(easing.overshootClamping)),
                ]),
            ]);
        case 'bezier':
            return b.callExpression(b.memberExpression(b.identifier(easingLocalName), b.identifier('bezier')), [easing.x1, easing.y1, easing.x2, easing.y2].map((value) => (0, update_nested_prop_1.parseValueExpression)(value)));
        default:
            throw new Error(`Unsupported easing: ${JSON.stringify(easing)}`);
    }
};
const makeKeyframedOptions = ({ param, remotionLocalNames, }) => {
    var _a;
    const properties = [];
    if (param.interpolationFunction !== 'interpolateColors') {
        if (param.clamping.left !== 'extend') {
            properties.push(b.objectProperty(b.identifier('extrapolateLeft'), b.stringLiteral(param.clamping.left)));
        }
        if (param.clamping.right !== 'extend') {
            properties.push(b.objectProperty(b.identifier('extrapolateRight'), b.stringLiteral(param.clamping.right)));
        }
    }
    if (keyframedParamNeedsEasingImport(param)) {
        const easingLocalName = (_a = remotionLocalNames.Easing) !== null && _a !== void 0 ? _a : 'Easing';
        properties.push(b.objectProperty(b.identifier('easing'), b.arrayExpression(param.easing.map((easing) => makeEasingExpression({ easing, easingLocalName })))));
    }
    if (param.posterize !== undefined) {
        properties.push(b.objectProperty(b.identifier('posterize'), (0, update_nested_prop_1.parseValueExpression)(param.posterize)));
    }
    if (properties.length === 0) {
        return null;
    }
    return b.objectExpression(properties);
};
const makeKeyframedExpression = ({ param, remotionLocalNames, }) => {
    var _a;
    const expectedEasingCount = Math.max(0, param.keyframes.length - 1);
    if (param.easing.length !== expectedEasingCount) {
        throw new Error('Cannot paste keyframed effect: invalid easing metadata');
    }
    const calleeName = (_a = remotionLocalNames[param.interpolationFunction]) !== null && _a !== void 0 ? _a : param.interpolationFunction;
    const args = [
        b.identifier('frame'),
        b.arrayExpression(param.keyframes.map((keyframe) => (0, update_nested_prop_1.parseValueExpression)(keyframe.frame))),
        b.arrayExpression(param.keyframes.map((keyframe) => (0, update_nested_prop_1.parseValueExpression)(keyframe.value))),
    ];
    const options = makeKeyframedOptions({ param, remotionLocalNames });
    if (options) {
        args.push(options);
    }
    return b.callExpression(b.identifier(calleeName), args);
};
const makeParamExpression = ({ param, remotionLocalNames, }) => {
    if (param.type === 'static') {
        return (0, update_nested_prop_1.parseValueExpression)(param.value);
    }
    return makeKeyframedExpression({ param, remotionLocalNames });
};
exports.makeParamExpression = makeParamExpression;
