// @remotion/studio-server/dist/codemods/paste-effects.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.pasteEffects = void 0;
const recast = __importStar(require("recast"));
const get_ast_node_path_1 = require("../helpers/get-ast-node-path");
const can_update_sequence_props_1 = require("../preview-server/routes/can-update-sequence-props");
const add_effect_1 = require("./add-effect");
const effect_param_expression_1 = require("./effect-param-expression");
const format_file_content_1 = require("./format-file-content");
const parse_ast_1 = require("./parse-ast");
const update_effect_props_1 = require("./update-effect-props/update-effect-props");
const ensure_imports_and_frame_hook_1 = require("./update-keyframes/ensure-imports-and-frame-hook");
const b = recast.types.builders;
const identifierRegex = /^[A-Za-z_$][0-9A-Za-z_$]*$/;
const getEffectsArray = (attr) => {
    if (!attr.value || attr.value.type !== 'JSXExpressionContainer') {
        throw new Error('Cannot append effects: effects prop is not an array');
    }
    const expr = attr.value.expression;
    if (expr.type !== 'ArrayExpression') {
        throw new Error('Cannot append effects: effects prop is not an array');
    }
    return expr;
};
const makeEffectsAttr = (array) => {
    return b.jsxAttribute(b.jsxIdentifier('effects'), b.jsxExpressionContainer(array));
};
const makeEffectsArray = (calls) => {
    return b.arrayExpression(calls);
};
const makeParamsObjectExpression = ({ params, remotionLocalNames, }) => {
    return b.objectExpression(Object.entries(params).map(([key, param]) => {
        const keyNode = identifierRegex.test(key)
            ? b.identifier(key)
            : b.stringLiteral(key);
        return b.objectProperty(keyNode, (0, effect_param_expression_1.makeParamExpression)({ param, remotionLocalNames }));
    }));
};
const makeEffectCall = ({ ast, effect, remotionLocalNames, }) => {
    (0, add_effect_1.assertValidEffect)({
        effectName: effect.callee,
        effectImportPath: effect.importPath,
    });
    const localName = (0, add_effect_1.ensureEffectImport)({
        ast,
        effectName: effect.callee,
        effectImportPath: effect.importPath,
    });
    return b.callExpression(b.identifier(localName), [
        makeParamsObjectExpression({
            params: effect.params,
            remotionLocalNames,
        }),
    ]);
};
const removeEffectsAttr = (attributes, attr) => {
    const index = attributes.indexOf(attr);
    if (index !== -1) {
        attributes.splice(index, 1);
    }
};
const pasteEffects = async ({ input, targetSequenceNodePath, type, effects, prettierConfigOverride, }) => {
    var _a;
    var _b, _c, _d;
    const ast = (0, parse_ast_1.parseAst)(input);
    const targetJsxPath = (0, get_ast_node_path_1.getAstNodePath)(ast, targetSequenceNodePath);
    const targetJsx = (0, can_update_sequence_props_1.findJsxElementAtNodePath)(ast, targetSequenceNodePath);
    if (!targetJsx || !targetJsxPath) {
        throw new Error('Could not find a JSX element at the specified location to paste effects');
    }
    const requiredRemotionImports = (0, effect_param_expression_1.getRequiredRemotionImports)(effects);
    const remotionLocalNames = (0, effect_param_expression_1.ensureRemotionImportLocalNames)({
        ast,
        requiredImports: requiredRemotionImports,
    });
    if (requiredRemotionImports.has('useCurrentFrame')) {
        const fnPath = (0, ensure_imports_and_frame_hook_1.findEnclosingFunctionPath)(targetJsxPath);
        if (fnPath) {
            (0, ensure_imports_and_frame_hook_1.ensureUseCurrentFrameHook)(fnPath, (_b = remotionLocalNames.useCurrentFrame) !== null && _b !== void 0 ? _b : 'useCurrentFrame');
        }
    }
    const effectCalls = effects.map((effect) => makeEffectCall({ ast, effect, remotionLocalNames }));
    const effectLabels = effects.map((effect) => `${effect.callee}()`);
    const existingAttr = (0, update_effect_props_1.findEffectsAttr)((_c = targetJsx.attributes) !== null && _c !== void 0 ? _c : []);
    if (type === 'effects-replacing') {
        if (existingAttr) {
            removeEffectsAttr(targetJsx.attributes, existingAttr);
        }
        if (effectCalls.length > 0) {
            targetJsx.attributes.push(makeEffectsAttr(makeEffectsArray(effectCalls)));
        }
    }
    else if (existingAttr) {
        getEffectsArray(existingAttr).elements.push(...effectCalls);
    }
    else if (effectCalls.length > 0) {
        targetJsx.attributes.push(makeEffectsAttr(makeEffectsArray(effectCalls)));
    }
    else {
        throw new Error('Cannot paste effects: no effects were copied');
    }
    const finalFile = (0, parse_ast_1.serializeAst)(ast);
    const { output, formatted } = await (0, format_file_content_1.formatFileContent)({
        input: finalFile,
        prettierConfigOverride,
    });
    return {
        output,
        formatted,
        effectLabels,
        logLine: (_d = (_a = targetJsx.loc) === null || _a === void 0 ? void 0 : _a.start.line) !== null && _d !== void 0 ? _d : 1,
    };
};
exports.pasteEffects = pasteEffects;
