// @remotion/studio-server/dist/codemods/update-default-props.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.getCompositionDefaultPropsLine = exports.updateDefaultProps = void 0;
const studio_shared_1 = require("@remotion/studio-shared");
const recast = __importStar(require("recast"));
const format_inline_content_1 = require("./format-inline-content");
const parse_ast_1 = require("./parse-ast");
// Recast uses tabWidth=4 for column counting, so columns don't
// correspond to character indices when tabs are present.
// This converts a recast loc (line/column) to a character offset.
const RECAST_TAB_WIDTH = 4;
const recastLocToOffset = (input, loc) => {
    const lines = input.split('\n');
    let offset = 0;
    for (let i = 0; i < loc.line - 1; i++) {
        offset += lines[i].length + 1;
    }
    // Convert recast's tab-expanded column to character index
    const line = lines[loc.line - 1];
    let col = 0;
    for (let i = 0; i < line.length; i++) {
        if (col >= loc.column) {
            return offset + i;
        }
        col += line[i] === '\t' ? RECAST_TAB_WIDTH : 1;
    }
    return offset + line.length;
};
const updateDefaultProps = async ({ input, compositionId, newDefaultProps, enumPaths, }) => {
    const ast = (0, parse_ast_1.parseAst)(input);
    const stringified = (0, studio_shared_1.stringifyDefaultProps)({
        props: newDefaultProps,
        enumPaths,
    });
    let replaceStart;
    let replaceEnd;
    recast.types.visit(ast, {
        visitJSXElement(path) {
            var _a, _b;
            const { openingElement } = path.node;
            //	1: ensure its the element we're looking for
            const openingName = openingElement.name;
            if (openingName.type !== 'JSXIdentifier' &&
                openingName.type !== 'JSXNamespacedName') {
                this.traverse(path); // Continue traversing the AST
                return;
            }
            if (openingName.name !== 'Composition' && openingName.name !== 'Still') {
                this.traverse(path); // Continue traversing the AST
                return;
            }
            if (!((_a = openingElement.attributes) === null || _a === void 0 ? void 0 : _a.some((attr) => {
                if (attr.type === 'JSXSpreadAttribute') {
                    return;
                }
                if (!attr.value) {
                    return;
                }
                if (attr.value.type === 'JSXElement') {
                    return;
                }
                if (attr.value.type === 'JSXExpressionContainer') {
                    return;
                }
                if (attr.value.type === 'JSXFragment') {
                    return;
                }
                return attr.name.name === 'id' && attr.value.value === compositionId;
            }))) {
                this.traverse(path); // Continue traversing the AST
                return;
            }
            //	2: Find the defaultProps attribute and handle related errors
            const defaultPropsAttr = openingElement.attributes.find((attr) => {
                if (attr.type === 'JSXSpreadAttribute') {
                    this.traverse(path); // Continue traversing the AST
                    return;
                }
                return attr.name.name === 'defaultProps';
            });
            if (!defaultPropsAttr) {
                throw new Error(`No \`defaultProps\` prop found in the <Composition/> tag with the ID "${compositionId}".`);
            }
            if (defaultPropsAttr.type === 'JSXSpreadAttribute') {
                this.traverse(path); // Continue traversing the AST
                return;
            }
            //	3: ensure only hardcoded values are provided
            if (!defaultPropsAttr.value ||
                defaultPropsAttr.value.type === 'JSXElement' ||
                defaultPropsAttr.value.type === 'JSXText' ||
                defaultPropsAttr.value.type === 'StringLiteral' ||
                defaultPropsAttr.value.type === 'NumericLiteral' ||
                defaultPropsAttr.value.type === 'BigIntLiteral' ||
                defaultPropsAttr.value.type === 'DecimalLiteral' ||
                defaultPropsAttr.value.type === 'NullLiteral' ||
                defaultPropsAttr.value.type === 'BooleanLiteral' ||
                defaultPropsAttr.value.type === 'RegExpLiteral' ||
                defaultPropsAttr.value.type === 'JSXFragment' ||
                defaultPropsAttr.value.type === 'Literal') {
                throw new Error(`\`defaultProps\` prop must be a hardcoded value in the <Composition/> tag, but it is a ${(_b = defaultPropsAttr.value) === null || _b === void 0 ? void 0 : _b.type}".`);
            }
            const defaultPropsValue = defaultPropsAttr.value.expression;
            if (defaultPropsValue.type !== 'ObjectExpression' &&
                defaultPropsValue.type !== 'TSAsExpression') {
                throw new Error(`\`defaultProps\` prop must be a hardcoded value in the <Composition/> tag with the ID "${compositionId}".`);
            }
            // Capture source positions for direct string replacement
            // instead of modifying the AST and serializing (avoids recast artifacts)
            const valueLoc = defaultPropsAttr.value.loc;
            if (!valueLoc) {
                throw new Error('Could not determine source location of defaultProps');
            }
            replaceStart = recastLocToOffset(input, valueLoc.start);
            replaceEnd = recastLocToOffset(input, valueLoc.end);
            this.traverse(path); // Continue traversing the AST
        },
    });
    if (replaceStart === undefined || replaceEnd === undefined) {
        throw new Error(`Could not find defaultProps for composition "${compositionId}"`);
    }
    // linePrefix includes the JSX container opening brace
    const lineStart = input.lastIndexOf('\n', replaceStart) + 1;
    const linePrefix = input.substring(lineStart, replaceStart + 1);
    const { formatted, didFormat } = await (0, format_inline_content_1.formatInlineContent)({
        inlineContent: stringified,
        linePrefix,
        endOfLine: 'auto',
    });
    // Replace the JSX expression container in the original input
    const output = input.substring(0, replaceStart) +
        '{' +
        formatted +
        '}' +
        input.substring(replaceEnd);
    return { output, formatted: didFormat };
};
exports.updateDefaultProps = updateDefaultProps;
/** Line of the matching `<Composition>` / `<Still>` opening tag (for log links). */
const getCompositionDefaultPropsLine = ({ input, compositionId, }) => {
    const ast = (0, parse_ast_1.parseAst)(input);
    let line = 1;
    let found = false;
    recast.types.visit(ast, {
        visitJSXElement(path) {
            var _a, _b, _c;
            var _d, _e;
            if (found) {
                this.traverse(path);
                return;
            }
            const { openingElement } = path.node;
            const openingName = openingElement.name;
            if (openingName.type !== 'JSXIdentifier' &&
                openingName.type !== 'JSXNamespacedName') {
                this.traverse(path);
                return;
            }
            if (openingName.name !== 'Composition' && openingName.name !== 'Still') {
                this.traverse(path);
                return;
            }
            if (!((_a = openingElement.attributes) === null || _a === void 0 ? void 0 : _a.some((attr) => {
                if (attr.type === 'JSXSpreadAttribute') {
                    return;
                }
                if (!attr.value) {
                    return;
                }
                if (attr.value.type === 'JSXElement') {
                    return;
                }
                if (attr.value.type === 'JSXExpressionContainer') {
                    return;
                }
                if (attr.value.type === 'JSXFragment') {
                    return;
                }
                return attr.name.name === 'id' && attr.value.value === compositionId;
            }))) {
                this.traverse(path);
                return;
            }
            found = true;
            line = (_e = (_d = (_b = openingElement.loc) === null || _b === void 0 ? void 0 : _b.start.line) !== null && _d !== void 0 ? _d : (_c = path.node.loc) === null || _c === void 0 ? void 0 : _c.start.line) !== null && _e !== void 0 ? _e : 1;
            this.traverse(path);
        },
    });
    return line;
};
exports.getCompositionDefaultPropsLine = getCompositionDefaultPropsLine;
