// @remotion/studio-server/dist/preview-server/routes/add-keyframes.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.addKeyframesHandler = exports.addKeyframes = void 0;
const node_fs_1 = require("node:fs");
const renderer_1 = require("@remotion/renderer");
const update_keyframes_1 = require("../../codemods/update-keyframes/update-keyframes");
const file_watcher_1 = require("../../file-watcher");
const resolve_file_inside_project_1 = require("../../helpers/resolve-file-inside-project");
const undo_stack_1 = require("../undo-stack");
const watch_ignore_next_change_1 = require("../watch-ignore-next-change");
const log_effect_update_1 = require("./log-updates/log-effect-update");
const log_update_1 = require("./log-updates/log-update");
const save_props_mutex_1 = require("./save-props-mutex");
const groupBy = (items, getKey) => {
    var _a;
    const groups = new Map();
    for (const item of items) {
        const key = getKey(item);
        const group = (_a = groups.get(key)) !== null && _a !== void 0 ? _a : [];
        group.push(item);
        groups.set(key, group);
    }
    return [...groups.values()];
};
const getBatchDescription = ({ totalKeyframes, firstKeyframe, }) => {
    if (totalKeyframes === 1) {
        return {
            undoMessage: `↩️  ${firstKeyframe.key} keyframe removed at frame ${firstKeyframe.frame}`,
            redoMessage: `↪️  ${firstKeyframe.key} keyframe added at frame ${firstKeyframe.frame}`,
        };
    }
    return {
        undoMessage: `↩️  ${totalKeyframes} keyframes removed`,
        redoMessage: `↪️  ${totalKeyframes} keyframes added`,
    };
};
const addKeyframes = async ({ sequenceKeyframes, effectKeyframes, clientId, remotionRoot, logLevel, }) => {
    var _a, _b;
    const totalKeyframes = sequenceKeyframes.length + effectKeyframes.length;
    if (totalKeyframes === 0) {
        throw new Error('No keyframes were specified for adding');
    }
    const fileGroups = new Map();
    for (const [index, keyframe] of sequenceKeyframes.entries()) {
        const { absolutePath, fileRelativeToRoot } = (0, resolve_file_inside_project_1.resolveFileInsideProject)({
            remotionRoot,
            fileName: keyframe.fileName,
            action: 'modify',
        });
        const group = (_a = fileGroups.get(absolutePath)) !== null && _a !== void 0 ? _a : {
            fileRelativeToRoot,
            sequenceKeyframes: [],
            effectKeyframes: [],
        };
        group.sequenceKeyframes.push({
            ...keyframe,
            index,
            absolutePath,
            fileRelativeToRoot,
        });
        fileGroups.set(absolutePath, group);
    }
    for (const [index, keyframe] of effectKeyframes.entries()) {
        const { absolutePath, fileRelativeToRoot } = (0, resolve_file_inside_project_1.resolveFileInsideProject)({
            remotionRoot,
            fileName: keyframe.fileName,
            action: 'modify',
        });
        const group = (_b = fileGroups.get(absolutePath)) !== null && _b !== void 0 ? _b : {
            fileRelativeToRoot,
            sequenceKeyframes: [],
            effectKeyframes: [],
        };
        group.effectKeyframes.push({
            ...keyframe,
            index,
            absolutePath,
            fileRelativeToRoot,
        });
        fileGroups.set(absolutePath, group);
    }
    const snapshots = [];
    const sequenceLogs = [];
    const effectLogs = [];
    for (const [absolutePath, group] of fileGroups) {
        const fileContents = (0, node_fs_1.readFileSync)(absolutePath, 'utf-8');
        let output = fileContents;
        let firstLogLine = Number.POSITIVE_INFINITY;
        for (const keyframeGroup of groupBy(group.sequenceKeyframes, (keyframe) => JSON.stringify(keyframe.nodePath.nodePath))) {
            const [firstSequenceKeyframe] = keyframeGroup;
            if (!firstSequenceKeyframe) {
                continue;
            }
            const result = await (0, update_keyframes_1.updateSequenceKeyframes)({
                input: output,
                nodePath: firstSequenceKeyframe.nodePath.nodePath,
                schema: firstSequenceKeyframe.schema,
                updates: keyframeGroup.map((keyframe) => ({
                    key: keyframe.key,
                    operation: {
                        type: 'add',
                        frame: keyframe.frame,
                        value: JSON.parse(keyframe.value),
                    },
                })),
            });
            output = result.output;
            firstLogLine = Math.min(firstLogLine, result.logLine);
            for (const [keyframeIndex, keyframe] of keyframeGroup.entries()) {
                sequenceLogs.push({
                    fileRelativeToRoot: keyframe.fileRelativeToRoot,
                    line: result.logLine,
                    key: keyframe.key,
                    oldValueString: result.oldValueStrings[keyframeIndex],
                    newValueString: result.newValueStrings[keyframeIndex],
                    formatted: result.formatted,
                });
            }
        }
        for (const keyframeGroup of groupBy(group.effectKeyframes, (keyframe) => `${JSON.stringify(keyframe.sequenceNodePath.nodePath)}:${keyframe.effectIndex}`)) {
            const [firstEffectKeyframe] = keyframeGroup;
            if (!firstEffectKeyframe) {
                continue;
            }
            const result = await (0, update_keyframes_1.updateEffectKeyframes)({
                input: output,
                sequenceNodePath: firstEffectKeyframe.sequenceNodePath.nodePath,
                effectIndex: firstEffectKeyframe.effectIndex,
                schema: firstEffectKeyframe.schema,
                updates: keyframeGroup.map((keyframe) => ({
                    key: keyframe.key,
                    operation: {
                        type: 'add',
                        frame: keyframe.frame,
                        value: JSON.parse(keyframe.value),
                    },
                })),
            });
            output = result.output;
            firstLogLine = Math.min(firstLogLine, result.logLine);
            for (const [keyframeIndex, keyframe] of keyframeGroup.entries()) {
                effectLogs.push({
                    fileRelativeToRoot: keyframe.fileRelativeToRoot,
                    line: result.logLine,
                    effectName: result.effectCallee,
                    propKey: keyframe.key,
                    oldValueString: result.oldValueStrings[keyframeIndex],
                    newValueString: result.newValueStrings[keyframeIndex],
                    formatted: result.formatted,
                });
            }
        }
        snapshots.push({
            filePath: absolutePath,
            oldContents: fileContents,
            newContents: output,
            logLine: Number.isFinite(firstLogLine) ? firstLogLine : 1,
        });
    }
    const [firstKeyframe] = sequenceKeyframes.length > 0 ? sequenceKeyframes : effectKeyframes;
    if (!firstKeyframe) {
        throw new Error('No keyframes were specified for adding');
    }
    (0, undo_stack_1.pushTransactionToUndoStack)({
        snapshots,
        logLevel,
        remotionRoot,
        description: getBatchDescription({ totalKeyframes, firstKeyframe }),
        entryType: sequenceKeyframes.length > 0 && effectKeyframes.length > 0
            ? // Dead code for now: sequence props and effect props cannot be selected
                // together yet. Keep the mixed transaction type for the planned UI.
                'keyframe-add'
            : sequenceKeyframes.length > 0
                ? 'sequence-props'
                : 'effect-props',
        suppressHmrOnFileRestore: true,
    });
    for (const snapshot of snapshots) {
        (0, undo_stack_1.suppressUndoStackInvalidation)(snapshot.filePath);
        (0, watch_ignore_next_change_1.suppressBundlerUpdateForFile)(snapshot.filePath);
        (0, file_watcher_1.writeFileAndNotifyFileWatchers)(snapshot.filePath, snapshot.newContents, clientId);
    }
    for (const log of sequenceLogs) {
        (0, log_update_1.logUpdate)({
            fileRelativeToRoot: log.fileRelativeToRoot,
            line: log.line,
            key: log.key,
            oldValueString: log.oldValueString,
            newValueString: log.newValueString,
            defaultValueString: null,
            formatted: log.formatted,
            logLevel,
            removedProps: [],
            addedProps: [],
        });
    }
    for (const log of effectLogs) {
        (0, log_effect_update_1.logEffectUpdate)({
            fileRelativeToRoot: log.fileRelativeToRoot,
            line: log.line,
            effectName: log.effectName,
            propKey: log.propKey,
            oldValueString: log.oldValueString,
            newValueString: log.newValueString,
            defaultValueString: null,
            formatted: log.formatted,
            logLevel,
            removedProps: [],
            addedProps: [],
        });
    }
    (0, undo_stack_1.printUndoHint)(logLevel);
};
exports.addKeyframes = addKeyframes;
const addKeyframesHandler = ({ input: { sequenceKeyframes, effectKeyframes, clientId }, remotionRoot, logLevel, }) => (0, save_props_mutex_1.withSavePropsLock)(async () => {
    renderer_1.RenderInternals.Log.trace({ indent: false, logLevel }, `[add-keyframes] Received request to add ${sequenceKeyframes.length + effectKeyframes.length} keyframe(s)`);
    await (0, exports.addKeyframes)({
        sequenceKeyframes,
        effectKeyframes,
        clientId,
        remotionRoot,
        logLevel,
    });
    return { success: true };
});
exports.addKeyframesHandler = addKeyframesHandler;
