// @remotion/studio-server/dist/preview-server/routes/add-effect-keyframe.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.addEffectKeyframeHandler = void 0;
const node_fs_1 = require("node:fs");
const renderer_1 = require("@remotion/renderer");
const studio_shared_1 = require("@remotion/studio-shared");
const parse_ast_1 = require("../../codemods/parse-ast");
const update_keyframes_1 = require("../../codemods/update-keyframes/update-keyframes");
const file_watcher_1 = require("../../file-watcher");
const resolve_file_inside_project_1 = require("../../helpers/resolve-file-inside-project");
const undo_stack_1 = require("../undo-stack");
const watch_ignore_next_change_1 = require("../watch-ignore-next-change");
const can_update_effect_props_1 = require("./can-update-effect-props");
const can_update_sequence_props_1 = require("./can-update-sequence-props");
const log_effect_update_1 = require("./log-updates/log-effect-update");
const save_props_mutex_1 = require("./save-props-mutex");
const addEffectKeyframeHandler = ({ input: { fileName, sequenceNodePath, effectIndex, key, frame, value, schema, clientId, }, remotionRoot, logLevel, }) => (0, save_props_mutex_1.withSavePropsLock)(async () => {
    renderer_1.RenderInternals.Log.trace({ indent: false, logLevel }, `[add-effect-keyframe] Received request for fileName="${fileName}" effectIndex=${effectIndex} key="${key}" frame=${frame}`);
    const { absolutePath, fileRelativeToRoot } = (0, resolve_file_inside_project_1.resolveFileInsideProject)({
        remotionRoot,
        fileName,
        action: 'modify',
    });
    const fileContents = (0, node_fs_1.readFileSync)(absolutePath, 'utf-8');
    const parsedValue = JSON.parse(value);
    const { output, oldValueStrings, newValueStrings, formatted, logLine, effectCallee, updatedSequenceNodePath, } = await (0, update_keyframes_1.updateEffectKeyframes)({
        input: fileContents,
        sequenceNodePath: sequenceNodePath.nodePath,
        effectIndex,
        schema,
        updates: [
            {
                key,
                operation: {
                    type: 'add',
                    frame,
                    value: parsedValue,
                },
            },
        ],
    });
    const oldValueString = oldValueStrings[0];
    const newValueString = newValueStrings[0];
    const undoPropChange = `${key} keyframe removed at frame ${frame}`;
    const redoPropChange = `${key} keyframe added at frame ${frame}`;
    (0, undo_stack_1.pushToUndoStack)({
        filePath: absolutePath,
        oldContents: fileContents,
        newContents: null,
        logLevel,
        remotionRoot,
        logLine,
        description: {
            undoMessage: `↩️  ${undoPropChange}`,
            redoMessage: `↪️  ${redoPropChange}`,
        },
        entryType: 'effect-props',
        suppressHmrOnFileRestore: true,
    });
    (0, undo_stack_1.suppressUndoStackInvalidation)(absolutePath);
    (0, watch_ignore_next_change_1.suppressBundlerUpdateForFile)(absolutePath);
    (0, file_watcher_1.writeFileAndNotifyFileWatchers)(absolutePath, output, clientId);
    (0, log_effect_update_1.logEffectUpdate)({
        fileRelativeToRoot,
        line: logLine,
        effectName: effectCallee,
        propKey: key,
        oldValueString,
        newValueString,
        defaultValueString: null,
        formatted,
        logLevel,
        removedProps: [],
        addedProps: [],
    });
    (0, undo_stack_1.printUndoHint)(logLevel);
    const ast = (0, parse_ast_1.parseAst)((0, node_fs_1.readFileSync)(absolutePath, 'utf-8'));
    const jsx = (0, can_update_sequence_props_1.findJsxElementAtNodePath)(ast, updatedSequenceNodePath);
    if (!jsx) {
        return {
            canUpdate: false,
            effectIndex,
            reason: 'not-found',
        };
    }
    return (0, can_update_effect_props_1.computeEffectPropStatus)({
        ast,
        jsx,
        effectIndex,
        keys: (0, studio_shared_1.getAllSchemaKeys)(schema),
    });
});
exports.addEffectKeyframeHandler = addEffectKeyframeHandler;
