// @remotion/studio-server/dist/preview-server/routes/add-effect.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.addEffectHandler = void 0;
const node_fs_1 = require("node:fs");
const renderer_1 = require("@remotion/renderer");
const add_effect_1 = require("../../codemods/add-effect");
const file_watcher_1 = require("../../file-watcher");
const resolve_file_inside_project_1 = require("../../helpers/resolve-file-inside-project");
const format_log_file_location_1 = require("../format-log-file-location");
const undo_stack_1 = require("../undo-stack");
const formatting_1 = require("./log-updates/formatting");
const log_update_1 = require("./log-updates/log-update");
const addEffectHandler = async ({ input: { fileName, sequenceNodePath, effectName, effectImportPath, effectConfig, clientId, }, remotionRoot, logLevel, }) => {
    try {
        renderer_1.RenderInternals.Log.trace({ indent: false, logLevel }, `[add-effect] Received request for fileName="${fileName}" effect="${effectName}"`);
        const { absolutePath, fileRelativeToRoot } = (0, resolve_file_inside_project_1.resolveFileInsideProject)({
            remotionRoot,
            fileName,
            action: 'modify',
        });
        const fileContents = (0, node_fs_1.readFileSync)(absolutePath, 'utf-8');
        const { output, formatted, effectLabel, nodeLabel, logLine } = await (0, add_effect_1.addEffect)({
            input: fileContents,
            sequenceNodePath: sequenceNodePath.nodePath,
            effectName,
            effectImportPath,
            effectConfig,
        });
        (0, undo_stack_1.pushToUndoStack)({
            filePath: absolutePath,
            oldContents: fileContents,
            newContents: null,
            logLevel,
            remotionRoot,
            logLine,
            description: {
                undoMessage: `↩️  Addition of ${effectLabel} to ${nodeLabel}`,
                redoMessage: `↪️  Addition of ${effectLabel} to ${nodeLabel}`,
            },
            entryType: 'add-effect',
            suppressHmrOnFileRestore: false,
        });
        (0, undo_stack_1.suppressUndoStackInvalidation)(absolutePath);
        (0, file_watcher_1.writeFileAndNotifyFileWatchers)(absolutePath, output, clientId);
        const locationLabel = (0, format_log_file_location_1.formatLogFileLocation)({
            remotionRoot,
            absolutePath,
            line: logLine,
        });
        renderer_1.RenderInternals.Log.info({ indent: false, logLevel }, `${renderer_1.RenderInternals.chalk.blueBright(`${locationLabel}`)} Added ${(0, formatting_1.attrName)(effectLabel)} to ${nodeLabel}`);
        if (!formatted) {
            (0, log_update_1.warnAboutPrettierOnce)(logLevel);
        }
        renderer_1.RenderInternals.Log.verbose({ indent: false, logLevel }, `[add-effect] Wrote ${fileRelativeToRoot}${formatted ? ' (formatted)' : ''}`);
        (0, undo_stack_1.printUndoHint)(logLevel);
        return {
            success: true,
        };
    }
    catch (err) {
        return {
            success: false,
            reason: err.message,
            stack: err.stack,
        };
    }
};
exports.addEffectHandler = addEffectHandler;
