// @remotion/studio-server/dist/preview-server/routes/apply-visual-control-change.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.applyVisualControlHandler = void 0;
const node_fs_1 = require("node:fs");
const renderer_1 = require("@remotion/renderer");
const recast = __importStar(require("recast"));
const parse_ast_1 = require("../../codemods/parse-ast");
const recast_mods_1 = require("../../codemods/recast-mods");
const file_watcher_1 = require("../../file-watcher");
const resolve_file_inside_project_1 = require("../../helpers/resolve-file-inside-project");
const format_log_file_location_1 = require("../format-log-file-location");
const live_events_1 = require("../live-events");
const undo_stack_1 = require("../undo-stack");
const watch_ignore_next_change_1 = require("../watch-ignore-next-change");
const log_update_1 = require("./log-updates/log-update");
const getVisualControlChangeLine = (file, changeId) => {
    let line = 1;
    recast.types.visit(file.program, {
        visitCallExpression(callPath) {
            var _a;
            var _b;
            const { node } = callPath;
            if (node.callee.type === 'Identifier' &&
                node.callee.name === 'visualControl') {
                const firstArg = node.arguments[0];
                if ((firstArg === null || firstArg === void 0 ? void 0 : firstArg.type) === 'StringLiteral' && firstArg.value === changeId) {
                    line = (_b = (_a = node.loc) === null || _a === void 0 ? void 0 : _a.start.line) !== null && _b !== void 0 ? _b : 1;
                    return false;
                }
            }
            this.traverse(callPath);
        },
    });
    return line;
};
const applyVisualControlHandler = async ({ input: { fileName, changes }, remotionRoot, logLevel }) => {
    renderer_1.RenderInternals.Log.trace({ indent: false, logLevel }, `[apply-visual-control] Received request for ${fileName} with ${changes.length} changes`);
    const { absolutePath } = (0, resolve_file_inside_project_1.resolveFileInsideProject)({
        remotionRoot,
        fileName,
        action: 'apply visual control change to',
    });
    const fileContents = (0, node_fs_1.readFileSync)(absolutePath, 'utf-8');
    const ast = (0, parse_ast_1.parseAst)(fileContents);
    const logLine = changes.length > 0 ? getVisualControlChangeLine(ast, changes[0].id) : 1;
    const { newAst, changesMade } = (0, recast_mods_1.applyCodemod)({
        file: ast,
        codeMod: {
            type: 'apply-visual-control',
            changes,
        },
    });
    if (changesMade.length === 0) {
        throw new Error('No changes were made to the file');
    }
    let output = (0, parse_ast_1.serializeAst)(newAst);
    let formatted = false;
    try {
        const prettier = await Promise.resolve().then(() => __importStar(require('prettier')));
        const { format, resolveConfig, resolveConfigFile } = prettier;
        const configFilePath = await resolveConfigFile();
        if (configFilePath) {
            const prettierConfig = await resolveConfig(configFilePath);
            if (prettierConfig) {
                output = await format(output, {
                    ...prettierConfig,
                    filepath: 'test.tsx',
                    plugins: [],
                    endOfLine: 'auto',
                });
                formatted = true;
            }
        }
    }
    catch (_a) {
        // Prettier not available, use unformatted output
    }
    (0, undo_stack_1.pushToUndoStack)({
        filePath: absolutePath,
        oldContents: fileContents,
        newContents: null,
        logLevel,
        remotionRoot,
        logLine,
        description: {
            undoMessage: '↩️  Visual control change',
            redoMessage: '↪️  Visual control change',
        },
        entryType: 'visual-control',
        suppressHmrOnFileRestore: true,
    });
    (0, undo_stack_1.suppressUndoStackInvalidation)(absolutePath);
    (0, watch_ignore_next_change_1.suppressBundlerUpdateForFile)(absolutePath);
    (0, file_watcher_1.writeFileAndNotifyFileWatchers)(absolutePath, output, undefined);
    (0, live_events_1.waitForLiveEventsListener)().then((listener) => {
        listener.sendEventToClient({
            type: 'visual-control-values-changed',
            values: changes.map((change) => ({
                id: change.id,
                value: change.newValueIsUndefined
                    ? null
                    : JSON.parse(change.newValueSerialized),
                isUndefined: change.newValueIsUndefined,
            })),
        });
    });
    const locationLabel = (0, format_log_file_location_1.formatLogFileLocation)({
        remotionRoot,
        absolutePath,
        line: logLine,
    });
    renderer_1.RenderInternals.Log.info({ indent: false, logLevel }, `${renderer_1.RenderInternals.chalk.blueBright(`${locationLabel}`)} Applied visual control changes`);
    if (!formatted) {
        (0, log_update_1.warnAboutPrettierOnce)(logLevel);
    }
    (0, undo_stack_1.printUndoHint)(logLevel);
    return {
        success: true,
    };
};
exports.applyVisualControlHandler = applyVisualControlHandler;
