// @remotion/studio-server/dist/preview-server/routes/can-update-sequence-props.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.computeSequencePropsStatusFromFilenameByLine = exports.computeSequencePropsStatus = exports.computeSequencePropsStatusFromContent = exports.lineColumnToNodePath = exports.findNodePathForJsxElement = exports.findJsxElementAtNodePath = exports.getComputedStatus = exports.extractStaticValue = exports.isStaticValue = void 0;
const node_fs_1 = require("node:fs");
const renderer_1 = require("@remotion/renderer");
const studio_shared_1 = require("@remotion/studio-shared");
const recast = __importStar(require("recast"));
const no_react_1 = require("remotion/no-react");
const parse_ast_1 = require("../../codemods/parse-ast");
const get_ast_node_path_1 = require("../../helpers/get-ast-node-path");
const import_agnostic_node_path_1 = require("../../helpers/import-agnostic-node-path");
const resolve_file_inside_project_1 = require("../../helpers/resolve-file-inside-project");
const jsx_component_identity_1 = require("../jsx-component-identity");
const jsx_element_not_found_at_location_error_1 = require("../jsx-element-not-found-at-location-error");
const can_update_effect_props_1 = require("./can-update-effect-props");
const staticStatus = (codeValue) => ({
    status: 'static',
    codeValue,
});
const computedStatus = () => ({
    status: 'computed',
});
// Mirrors the encoding that staticFile() from "remotion" applies at runtime
const encodeStaticFilePath = (path) => {
    return path.split('/').map(encodeURIComponent).join('/');
};
// Detects calls that the Studio knows how to serialize back to source:
// staticFile("...") and new Date("...")
const getSpecialValueCall = (node) => {
    if (node.type === 'CallExpression') {
        const call = node;
        if (call.callee.type === 'Identifier' &&
            call.callee.name === 'staticFile' &&
            call.arguments.length === 1 &&
            call.arguments[0].type === 'StringLiteral') {
            return { type: 'static-file', value: call.arguments[0].value };
        }
        return null;
    }
    if (node.type === 'NewExpression') {
        const newExpr = node;
        if (newExpr.callee.type === 'Identifier' &&
            newExpr.callee.name === 'Date' &&
            newExpr.arguments.length === 1 &&
            newExpr.arguments[0].type === 'StringLiteral') {
            return { type: 'date', value: newExpr.arguments[0].value };
        }
        return null;
    }
    return null;
};
const defaultStaticValueOptions = {
    allowSpecialValues: false,
};
const isStaticValue = (node, options = defaultStaticValueOptions) => {
    if (options.allowSpecialValues && getSpecialValueCall(node) !== null) {
        return true;
    }
    switch (node.type) {
        case 'NumericLiteral':
        case 'StringLiteral':
        case 'BooleanLiteral':
        case 'NullLiteral':
            return true;
        case 'UnaryExpression':
            return ((node.operator === '-' ||
                node.operator === '+') &&
                node.argument.type === 'NumericLiteral');
        case 'TSAsExpression':
            return (0, exports.isStaticValue)(node.expression, options);
        case 'ArrayExpression':
            return node.elements.every((el) => el !== null &&
                el.type !== 'SpreadElement' &&
                (0, exports.isStaticValue)(el, options));
        case 'ObjectExpression':
            return node.properties.every((prop) => prop.type === 'ObjectProperty' &&
                (0, exports.isStaticValue)(prop.value, options));
        default:
            return false;
    }
};
exports.isStaticValue = isStaticValue;
const extractStaticValue = (node, options = defaultStaticValueOptions) => {
    if (options.allowSpecialValues) {
        const specialValue = getSpecialValueCall(node);
        if (specialValue !== null) {
            if (specialValue.type === 'static-file') {
                return `${no_react_1.NoReactInternals.FILE_TOKEN}${encodeStaticFilePath(specialValue.value)}`;
            }
            return `${no_react_1.NoReactInternals.DATE_TOKEN}${specialValue.value}`;
        }
    }
    switch (node.type) {
        case 'NumericLiteral':
        case 'StringLiteral':
        case 'BooleanLiteral':
            return node.value;
        case 'NullLiteral':
            return null;
        case 'UnaryExpression': {
            const un = node;
            if (un.argument.type === 'NumericLiteral') {
                const val = un.argument.value;
                return un.operator === '-' ? -val : val;
            }
            return undefined;
        }
        case 'TSAsExpression':
            return (0, exports.extractStaticValue)(node.expression, options);
        case 'ArrayExpression':
            return node.elements.map((el) => {
                if (el === null || el.type === 'SpreadElement') {
                    return undefined;
                }
                return (0, exports.extractStaticValue)(el, options);
            });
        case 'ObjectExpression': {
            const obj = node;
            const result = {};
            for (const prop of obj.properties) {
                if (prop.type === 'ObjectProperty') {
                    const p = prop;
                    const key = p.key.type === 'Identifier'
                        ? p.key.name
                        : p.key.type === 'StringLiteral' ||
                            p.key.type === 'NumericLiteral'
                            ? String(p.key.value)
                            : undefined;
                    if (key !== undefined) {
                        result[key] = (0, exports.extractStaticValue)(p.value, options);
                    }
                }
            }
            return result;
        }
        default:
            return undefined;
    }
};
exports.extractStaticValue = extractStaticValue;
const getNumericValue = (node) => {
    if (node.type === 'NumericLiteral') {
        return node.value;
    }
    if (node.type === 'UnaryExpression' &&
        (node.operator === '-' || node.operator === '+') &&
        node.argument.type === 'NumericLiteral') {
        return node.operator === '-' ? -node.argument.value : node.argument.value;
    }
    if (node.type === 'TSAsExpression') {
        return getNumericValue(node.expression);
    }
    return null;
};
const getExtrapolateType = (node) => {
    if (node.type === 'StringLiteral') {
        if (node.value === 'extend' ||
            node.value === 'identity' ||
            node.value === 'clamp' ||
            node.value === 'wrap') {
            return node.value;
        }
        return null;
    }
    if (node.type === 'TSAsExpression') {
        return getExtrapolateType(node.expression);
    }
    return null;
};
const getKeyframeEasing = (node) => {
    if (node.type === 'TSAsExpression') {
        return getKeyframeEasing(node.expression);
    }
    if (node.type === 'MemberExpression' &&
        node.object.type === 'Identifier' &&
        node.object.name === 'Easing' &&
        node.property.type === 'Identifier' &&
        node.property.name === 'linear' &&
        node.computed === false) {
        return { type: 'linear' };
    }
    if (node.type !== 'CallExpression' ||
        node.callee.type !== 'MemberExpression' ||
        node.callee.object.type !== 'Identifier' ||
        node.callee.object.name !== 'Easing' ||
        node.callee.property.type !== 'Identifier' ||
        node.callee.computed) {
        return null;
    }
    if (node.callee.property.name === 'spring') {
        if (node.arguments.length > 1) {
            return null;
        }
        const springConfig = node.arguments[0];
        if ((springConfig === null || springConfig === void 0 ? void 0 : springConfig.type) === 'ArgumentPlaceholder' ||
            (springConfig === null || springConfig === void 0 ? void 0 : springConfig.type) === 'JSXNamespacedName' ||
            (springConfig === null || springConfig === void 0 ? void 0 : springConfig.type) === 'SpreadElement') {
            return null;
        }
        return (0, studio_shared_1.parseSpringEasingConfig)(springConfig);
    }
    if (node.callee.property.name !== 'bezier' || node.arguments.length !== 4) {
        return null;
    }
    const values = node.arguments.map((arg) => {
        if (arg.type === 'ArgumentPlaceholder' ||
            arg.type === 'JSXNamespacedName' ||
            arg.type === 'SpreadElement') {
            return null;
        }
        return getNumericValue(arg);
    });
    if (values.some((v) => v === null)) {
        return null;
    }
    const [x1, y1, x2, y2] = values;
    return { type: 'bezier', x1, y1, x2, y2 };
};
const getKeyframeEasingArray = ({ easingNode, segments, }) => {
    if (segments === 0) {
        return [];
    }
    if (easingNode.type === 'TSAsExpression') {
        return getKeyframeEasingArray({
            easingNode: easingNode.expression,
            segments,
        });
    }
    if (easingNode.type === 'ArrayExpression') {
        if (easingNode.elements.length !== segments) {
            return null;
        }
        const parsed = easingNode.elements.map((element) => {
            if (!element || element.type === 'SpreadElement') {
                return null;
            }
            return getKeyframeEasing(element);
        });
        if (parsed.some((value) => value === null)) {
            return null;
        }
        return parsed;
    }
    const easing = getKeyframeEasing(easingNode);
    if (!easing) {
        return null;
    }
    return new Array(segments).fill(easing);
};
const getInterpolationMetadata = (interpolationFunction, callExpression, keyframeCount) => {
    const segments = Math.max(0, keyframeCount - 1);
    const defaultClamping = interpolationFunction === 'interpolateColors'
        ? {
            left: 'clamp',
            right: 'clamp',
        }
        : {
            left: 'extend',
            right: 'extend',
        };
    const defaults = {
        easing: Array.from({ length: segments }, () => studio_shared_1.LINEAR_KEYFRAME_EASING),
        clamping: defaultClamping,
        posterize: undefined,
    };
    const optionsArg = callExpression.arguments[3];
    if (!optionsArg) {
        return defaults;
    }
    if (optionsArg.type !== 'ObjectExpression') {
        return null;
    }
    let { easing } = defaults;
    let { clamping } = defaults;
    let posterize;
    for (const property of optionsArg.properties) {
        if (property.type !== 'ObjectProperty' || property.computed) {
            return null;
        }
        const key = property.key.type === 'Identifier'
            ? property.key.name
            : property.key.type === 'StringLiteral'
                ? property.key.value
                : null;
        if (!key) {
            return null;
        }
        const value = property.value;
        if (key === 'easing') {
            const parsedEasing = getKeyframeEasingArray({
                easingNode: value,
                segments,
            });
            if (!parsedEasing) {
                return null;
            }
            easing = parsedEasing;
            continue;
        }
        if (key === 'extrapolateLeft' || key === 'extrapolateRight') {
            if (interpolationFunction === 'interpolateColors') {
                return null;
            }
            const extrapolateType = getExtrapolateType(value);
            if (!extrapolateType) {
                return null;
            }
            clamping =
                key === 'extrapolateLeft'
                    ? { ...clamping, left: extrapolateType }
                    : { ...clamping, right: extrapolateType };
            continue;
        }
        if (key === 'posterize') {
            const parsedPosterize = getNumericValue(value);
            if (parsedPosterize === null ||
                !Number.isFinite(parsedPosterize) ||
                parsedPosterize <= 0) {
                return null;
            }
            posterize = parsedPosterize;
            continue;
        }
        return null;
    }
    return {
        easing,
        clamping,
        posterize,
    };
};
const getInterpolationKeyframes = (node, ast) => {
    if (node.type === 'TSAsExpression') {
        return getInterpolationKeyframes(node.expression, ast);
    }
    if (node.type === 'CallExpression' &&
        node.callee.type === 'Identifier' &&
        node.callee.name === 'String' &&
        node.arguments.length === 1 &&
        node.arguments[0].type !== 'ArgumentPlaceholder' &&
        node.arguments[0].type !== 'JSXNamespacedName' &&
        node.arguments[0].type !== 'SpreadElement') {
        const interpolation = getInterpolationKeyframes(node.arguments[0], ast);
        if (!interpolation) {
            return undefined;
        }
        return {
            ...interpolation,
            keyframes: interpolation.keyframes.map((keyframe) => ({
                ...keyframe,
                value: String(keyframe.value),
            })),
        };
    }
    if (node.type !== 'CallExpression') {
        return undefined;
    }
    const callExpression = node;
    if (callExpression.callee.type !== 'Identifier' ||
        !(0, studio_shared_1.isKeyframeInterpolationFunction)(callExpression.callee.name)) {
        return undefined;
    }
    const interpolationFunction = callExpression.callee.name;
    const frameArg = callExpression.arguments[0];
    const inputArg = callExpression.arguments[1];
    const outputArg = callExpression.arguments[2];
    if (!frameArg ||
        frameArg.type === 'SpreadElement' ||
        !isCurrentFrameIdentifier(frameArg, ast) ||
        !inputArg ||
        !outputArg ||
        inputArg.type !== 'ArrayExpression' ||
        outputArg.type !== 'ArrayExpression') {
        return undefined;
    }
    if (inputArg.elements.length !== outputArg.elements.length) {
        return undefined;
    }
    const keyframes = [];
    for (let i = 0; i < inputArg.elements.length; i++) {
        const inputElement = inputArg.elements[i];
        const outputElement = outputArg.elements[i];
        if (!inputElement ||
            !outputElement ||
            inputElement.type === 'SpreadElement' ||
            outputElement.type === 'SpreadElement') {
            return undefined;
        }
        const frame = getNumericValue(inputElement);
        if (frame === null || !(0, exports.isStaticValue)(outputElement)) {
            return undefined;
        }
        keyframes.push({
            frame,
            value: (0, exports.extractStaticValue)(outputElement),
        });
    }
    if (keyframes.length === 0) {
        return undefined;
    }
    const metadata = getInterpolationMetadata(interpolationFunction, callExpression, keyframes.length);
    if (!metadata) {
        return undefined;
    }
    return {
        interpolationFunction,
        keyframes,
        easing: metadata.easing,
        clamping: metadata.clamping,
        posterize: metadata.posterize,
    };
};
const isUseCurrentFrameCall = (node) => {
    return (node.type === 'CallExpression' &&
        node.callee.type === 'Identifier' &&
        node.callee.name === 'useCurrentFrame' &&
        node.arguments.length === 0);
};
const isCurrentFrameIdentifier = (node, ast) => {
    if (node.type === 'TSAsExpression') {
        return isCurrentFrameIdentifier(node.expression, ast);
    }
    if (node.type !== 'Identifier') {
        return false;
    }
    let hasUseCurrentFrameDeclaration = false;
    let hasOtherDeclaration = false;
    recast.types.visit(ast, {
        visitVariableDeclarator(p) {
            const { id, init } = p.node;
            if (id.type !== 'Identifier' || id.name !== node.name) {
                return this.traverse(p);
            }
            if (init && isUseCurrentFrameCall(init)) {
                hasUseCurrentFrameDeclaration = true;
            }
            else {
                hasOtherDeclaration = true;
            }
            return false;
        },
    });
    return hasUseCurrentFrameDeclaration && !hasOtherDeclaration;
};
const getComputedStatus = (node, ast) => {
    const interpolation = getInterpolationKeyframes(node, ast);
    if (!interpolation) {
        return computedStatus();
    }
    return {
        status: 'keyframed',
        interpolationFunction: interpolation.interpolationFunction,
        keyframes: interpolation.keyframes,
        easing: interpolation.easing,
        clamping: interpolation.clamping,
        posterize: interpolation.posterize,
    };
};
exports.getComputedStatus = getComputedStatus;
const getPropsStatus = (jsxElement, ast) => {
    const props = {};
    for (const attr of jsxElement.attributes) {
        if (attr.type === 'JSXSpreadAttribute') {
            continue;
        }
        if (attr.name.type === 'JSXNamespacedName') {
            continue;
        }
        const { name } = attr.name;
        if (typeof name !== 'string') {
            continue;
        }
        const { value } = attr;
        if (!value) {
            props[name] = staticStatus(true);
            continue;
        }
        if (value.type === 'StringLiteral') {
            props[name] = staticStatus(value.value);
            continue;
        }
        if (value.type === 'JSXExpressionContainer') {
            const { expression } = value;
            if (expression.type === 'JSXEmptyExpression') {
                props[name] = computedStatus();
                continue;
            }
            if (!(0, exports.isStaticValue)(expression)) {
                props[name] = (0, exports.getComputedStatus)(expression, ast);
                continue;
            }
            props[name] = staticStatus((0, exports.extractStaticValue)(expression));
            continue;
        }
        props[name] = computedStatus();
    }
    return props;
};
const getNodePathForRecastPath = (
// eslint-disable-next-line @typescript-eslint/no-explicit-any
recastPath, ast) => {
    const segments = [];
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    let current = recastPath;
    while (current && current.parentPath) {
        segments.unshift(current.name);
        current = current.parentPath;
    }
    // Recast paths start with "root" which doesn't correspond to a real AST property
    if (segments.length > 0 && segments[0] === 'root') {
        return (0, import_agnostic_node_path_1.toImportAgnosticNodePath)({ ast, nodePath: segments.slice(1) });
    }
    return (0, import_agnostic_node_path_1.toImportAgnosticNodePath)({ ast, nodePath: segments });
};
const findJsxElementAtNodePath = (ast, nodePath) => {
    const current = (0, get_ast_node_path_1.getAstNodePath)(ast, nodePath);
    if (!current) {
        return null;
    }
    if (recast.types.namedTypes.JSXOpeningElement.check(current.value)) {
        return current.value;
    }
    return null;
};
exports.findJsxElementAtNodePath = findJsxElementAtNodePath;
const findNodePathForJsxElement = (ast, target) => {
    let foundPath = null;
    recast.types.visit(ast, {
        visitJSXOpeningElement(p) {
            if (p.node === target) {
                foundPath = getNodePathForRecastPath(p, ast);
                return false;
            }
            return this.traverse(p);
        },
    });
    return foundPath;
};
exports.findNodePathForJsxElement = findNodePathForJsxElement;
const lineColumnToNodePath = (ast, targetLine) => {
    let foundPath = null;
    recast.types.visit(ast, {
        visitJSXOpeningElement(p) {
            const { node } = p;
            if (node.loc && node.loc.start.line === targetLine) {
                foundPath = getNodePathForRecastPath(p, ast);
                return false;
            }
            return this.traverse(p);
        },
    });
    return foundPath;
};
exports.lineColumnToNodePath = lineColumnToNodePath;
const PIXEL_VALUE_REGEX = /^-?\d+(\.\d+)?px$/;
const isSupportedTranslateValue = (value) => {
    const parts = value.split(/\s+/);
    if (parts.length === 1 || parts.length === 2) {
        return parts.every((part) => PIXEL_VALUE_REGEX.test(part));
    }
    return false;
};
const validateStyleValue = (childKey, value) => {
    if (childKey === 'translate' && typeof value === 'string') {
        return isSupportedTranslateValue(value);
    }
    return true;
};
const getNestedPropStatus = (jsxElement, ast, parentKey, childKey) => {
    const attr = jsxElement.attributes.find((a) => a.type !== 'JSXSpreadAttribute' &&
        a.name.type !== 'JSXNamespacedName' &&
        a.name.name === parentKey);
    if (!attr || !attr.value) {
        // Parent attribute doesn't exist, nested prop can be added
        return staticStatus(undefined);
    }
    if (attr.value.type !== 'JSXExpressionContainer') {
        return computedStatus();
    }
    const { expression } = attr.value;
    if (expression.type === 'JSXEmptyExpression' ||
        expression.type !== 'ObjectExpression') {
        // Parent is not an object literal (e.g. style={myStyles})
        return computedStatus();
    }
    const objExpr = expression;
    const prop = objExpr.properties.find((p) => p.type === 'ObjectProperty' &&
        ((p.key.type === 'Identifier' && p.key.name === childKey) ||
            (p.key.type === 'StringLiteral' && p.key.value === childKey)));
    if (!prop) {
        // Property not set in the object, can be added
        return staticStatus(undefined);
    }
    const propValue = prop.value;
    if (!(0, exports.isStaticValue)(propValue)) {
        return (0, exports.getComputedStatus)(propValue, ast);
    }
    const propStatus = (0, exports.extractStaticValue)(propValue);
    if (!validateStyleValue(childKey, propStatus)) {
        return computedStatus();
    }
    return staticStatus(propStatus);
};
const computeEffectsForJsx = ({ ast, jsxElement, effects, }) => {
    return effects.map((effect, effectIndex) => (0, can_update_effect_props_1.computeEffectPropStatus)({
        ast,
        jsx: jsxElement,
        effectIndex,
        keys: effect,
    }));
};
const computeSequenceOnlyPropsRecord = ({ jsxElement, ast, keys, }) => {
    const allProps = getPropsStatus(jsxElement, ast);
    const filteredProps = {};
    for (const key of keys) {
        const dotIndex = key.indexOf('.');
        if (dotIndex !== -1) {
            filteredProps[key] = getNestedPropStatus(jsxElement, ast, key.slice(0, dotIndex), key.slice(dotIndex + 1));
        }
        else if (key in allProps) {
            filteredProps[key] = allProps[key];
        }
        else {
            filteredProps[key] = staticStatus(undefined);
        }
    }
    return filteredProps;
};
const computeSequencePropsStatusFromContent = ({ fileContents, nodePath, componentIdentity, keys, effects, }) => {
    const ast = (0, parse_ast_1.parseAst)(fileContents);
    const jsxElement = (0, exports.findJsxElementAtNodePath)(ast, nodePath);
    if (!jsxElement) {
        throw new jsx_element_not_found_at_location_error_1.JsxElementNotFoundAtLocationError();
    }
    if (!(0, jsx_component_identity_1.jsxComponentIdentitiesMatch)({
        expected: componentIdentity,
        actual: (0, jsx_component_identity_1.getJsxComponentIdentity)({ ast, jsxElement }),
    })) {
        throw new jsx_component_identity_1.JsxElementIdentityMismatchError();
    }
    const filteredProps = computeSequenceOnlyPropsRecord({ jsxElement, ast, keys });
    const effectsStatuses = computeEffectsForJsx({ ast, jsxElement, effects });
    return {
        canUpdate: true,
        props: filteredProps,
        effects: effectsStatuses,
    };
};
exports.computeSequencePropsStatusFromContent = computeSequencePropsStatusFromContent;
const computeSequencePropsStatus = ({ fileName, nodePath, componentIdentity, keys, effects, remotionRoot, }) => {
    const { absolutePath } = (0, resolve_file_inside_project_1.resolveFileInsideProject)({
        remotionRoot,
        fileName,
        action: 'read',
    });
    const fileContents = (0, node_fs_1.readFileSync)(absolutePath, 'utf-8');
    return (0, exports.computeSequencePropsStatusFromContent)({
        fileContents,
        nodePath,
        componentIdentity,
        keys,
        effects,
    });
};
exports.computeSequencePropsStatus = computeSequencePropsStatus;
const computeSequencePropsStatusFromFilenameByLine = ({ fileName, line, componentIdentity, keys, effects, remotionRoot, logLevel, }) => {
    try {
        const { absolutePath } = (0, resolve_file_inside_project_1.resolveFileInsideProject)({
            remotionRoot,
            fileName,
            action: 'read',
        });
        const fileContents = (0, node_fs_1.readFileSync)(absolutePath, 'utf-8');
        const ast = (0, parse_ast_1.parseAst)(fileContents);
        const resolvedNodePath = (0, exports.lineColumnToNodePath)(ast, line);
        if (!resolvedNodePath) {
            return {
                status: {
                    canUpdate: false,
                    reason: 'not-found',
                },
                success: false,
            };
        }
        return {
            status: (0, exports.computeSequencePropsStatus)({
                fileName,
                nodePath: resolvedNodePath,
                componentIdentity,
                keys,
                effects,
                remotionRoot,
            }),
            nodePath: {
                absolutePath,
                nodePath: resolvedNodePath,
                sequenceKeys: keys,
                effectKeys: effects,
            },
            success: true,
        };
    }
    catch (err) {
        renderer_1.RenderInternals.Log.error({ indent: false, logLevel }, err);
        return {
            status: {
                canUpdate: false,
                reason: 'error',
            },
            success: false,
        };
    }
};
exports.computeSequencePropsStatusFromFilenameByLine = computeSequencePropsStatusFromFilenameByLine;
