// @remotion/studio-server/dist/preview-server/routes/can-update-default-props.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.computeCanUpdateDefaultProps = exports.computeCanUpdateDefaultPropsFromContent = exports.checkIfTypeScriptFile = void 0;
const node_fs_1 = require("node:fs");
const recast = __importStar(require("recast"));
const parse_ast_1 = require("../../codemods/parse-ast");
const project_info_1 = require("../project-info");
const can_update_sequence_props_1 = require("./can-update-sequence-props");
const checkIfTypeScriptFile = (file) => {
    if (!file.endsWith('.tsx') &&
        !file.endsWith('.ts') &&
        !file.endsWith('.mtsx') &&
        !file.endsWith('.mts')) {
        throw new Error('Cannot update Root file if not using TypeScript');
    }
};
exports.checkIfTypeScriptFile = checkIfTypeScriptFile;
const extractDefaultPropsFromSource = (input, compositionId) => {
    const ast = (0, parse_ast_1.parseAst)(input);
    let result = null;
    recast.types.visit(ast, {
        visitJSXElement(path) {
            var _a;
            const { openingElement } = path.node;
            const openingName = openingElement.name;
            if (openingName.type !== 'JSXIdentifier' &&
                openingName.type !== 'JSXNamespacedName') {
                this.traverse(path);
                return;
            }
            if (openingName.name !== 'Composition' && openingName.name !== 'Still') {
                this.traverse(path);
                return;
            }
            if (!((_a = openingElement.attributes) === null || _a === void 0 ? void 0 : _a.some((attr) => {
                if (attr.type === 'JSXSpreadAttribute') {
                    return;
                }
                if (!attr.value) {
                    return;
                }
                if (attr.value.type === 'JSXElement') {
                    return;
                }
                if (attr.value.type === 'JSXExpressionContainer') {
                    return;
                }
                if (attr.value.type === 'JSXFragment') {
                    return;
                }
                return attr.name.name === 'id' && attr.value.value === compositionId;
            }))) {
                this.traverse(path);
                return;
            }
            const defaultPropsAttr = openingElement.attributes.find((attr) => {
                if (attr.type === 'JSXSpreadAttribute') {
                    return;
                }
                return attr.name.name === 'defaultProps';
            });
            if (!defaultPropsAttr || defaultPropsAttr.type === 'JSXSpreadAttribute') {
                this.traverse(path);
                return;
            }
            if (!defaultPropsAttr.value ||
                defaultPropsAttr.value.type !== 'JSXExpressionContainer') {
                this.traverse(path);
                return;
            }
            const { expression } = defaultPropsAttr.value;
            if (expression.type === 'JSXEmptyExpression') {
                this.traverse(path);
                return;
            }
            if ((0, can_update_sequence_props_1.isStaticValue)(expression, {
                allowSpecialValues: true,
            })) {
                const value = (0, can_update_sequence_props_1.extractStaticValue)(expression, {
                    allowSpecialValues: true,
                });
                if (value !== null &&
                    typeof value === 'object' &&
                    !Array.isArray(value)) {
                    result = value;
                }
            }
            this.traverse(path);
        },
    });
    return result;
};
const computeCanUpdateDefaultPropsFromContent = (content, compositionId) => {
    try {
        const currentDefaultProps = extractDefaultPropsFromSource(content, compositionId);
        if (currentDefaultProps === null) {
            throw new Error(`Could not find or extract defaultProps for composition "${compositionId}"`);
        }
        return {
            canUpdate: true,
            currentDefaultProps,
        };
    }
    catch (err) {
        return {
            canUpdate: false,
            reason: err.message,
        };
    }
};
exports.computeCanUpdateDefaultPropsFromContent = computeCanUpdateDefaultPropsFromContent;
const computeCanUpdateDefaultProps = async ({ compositionId, remotionRoot, entryPoint, }) => {
    try {
        const projectInfo = await (0, project_info_1.getProjectInfo)(remotionRoot, entryPoint);
        if (!projectInfo.rootFile) {
            throw new Error('Cannot find root file in project');
        }
        (0, exports.checkIfTypeScriptFile)(projectInfo.rootFile);
        const input = (0, node_fs_1.readFileSync)(projectInfo.rootFile, 'utf-8');
        const result = (0, exports.computeCanUpdateDefaultPropsFromContent)(input, compositionId);
        return {
            result,
            rootFile: projectInfo.rootFile,
        };
    }
    catch (err) {
        return {
            result: {
                canUpdate: false,
                reason: err.message,
            },
            rootFile: null,
        };
    }
};
exports.computeCanUpdateDefaultProps = computeCanUpdateDefaultProps;
