// @remotion/studio-server/dist/preview-server/routes/can-update-effect-props.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.computeEffectPropsStatusesFromFile = exports.computeEffectPropsStatusesFromContent = exports.computeEffectPropStatus = void 0;
const node_fs_1 = require("node:fs");
const parse_ast_1 = require("../../codemods/parse-ast");
const update_effect_props_1 = require("../../codemods/update-effect-props/update-effect-props");
const resolve_file_inside_project_1 = require("../../helpers/resolve-file-inside-project");
const can_update_sequence_props_1 = require("./can-update-sequence-props");
const staticStatus = (codeValue) => ({
    status: 'static',
    codeValue,
});
const findEffectsAttr = (jsx) => {
    for (const attr of jsx.attributes) {
        if (attr.type !== 'JSXAttribute') {
            continue;
        }
        if (attr.name.type === 'JSXIdentifier' && attr.name.name === 'effects') {
            return attr;
        }
    }
    return null;
};
const getEffectsArrayElements = (attr) => {
    if (!attr || !attr.value || attr.value.type !== 'JSXExpressionContainer') {
        return null;
    }
    const expr = attr.value.expression;
    if (expr.type !== 'ArrayExpression') {
        return null;
    }
    return (0, update_effect_props_1.enumerateEffectArrayElements)(expr);
};
const getImportedName = (specifier) => {
    var _a, _b;
    if (!specifier.imported) {
        return null;
    }
    if (specifier.imported.type === 'Identifier') {
        return (_a = specifier.imported.name) !== null && _a !== void 0 ? _a : null;
    }
    return (_b = specifier.imported.value) !== null && _b !== void 0 ? _b : null;
};
const resolveEffectImport = ({ ast, call, fallbackCallee, }) => {
    var _a, _b;
    var _c;
    const { callee } = call;
    if (callee.type === 'Identifier') {
        const localName = callee.name;
        for (const node of ast.program.body) {
            if (node.type !== 'ImportDeclaration') {
                continue;
            }
            const matchingSpecifier = (_a = node.specifiers) === null || _a === void 0 ? void 0 : _a.find((specifier) => {
                var _a;
                return (specifier.type === 'ImportSpecifier' &&
                    ((_a = specifier.local) === null || _a === void 0 ? void 0 : _a.name) === localName);
            });
            if ((matchingSpecifier === null || matchingSpecifier === void 0 ? void 0 : matchingSpecifier.type) === 'ImportSpecifier') {
                return {
                    callee: (_c = getImportedName(matchingSpecifier)) !== null && _c !== void 0 ? _c : fallbackCallee,
                    importPath: String(node.source.value),
                };
            }
        }
    }
    if (callee.type === 'MemberExpression' &&
        callee.object.type === 'Identifier' &&
        callee.property.type === 'Identifier' &&
        !callee.computed) {
        const namespaceName = callee.object.name;
        for (const node of ast.program.body) {
            if (node.type !== 'ImportDeclaration') {
                continue;
            }
            const matchingSpecifier = (_b = node.specifiers) === null || _b === void 0 ? void 0 : _b.find((specifier) => {
                var _a;
                return (specifier.type === 'ImportNamespaceSpecifier' &&
                    ((_a = specifier.local) === null || _a === void 0 ? void 0 : _a.name) === namespaceName);
            });
            if (matchingSpecifier) {
                return {
                    callee: callee.property.name,
                    importPath: String(node.source.value),
                };
            }
        }
    }
    return { callee: fallbackCallee, importPath: null };
};
const getPropsFromObjectExpression = ({ ast, objExpr, keys, }) => {
    const out = {};
    for (const key of keys) {
        const prop = objExpr.properties.find((p) => p.type === 'ObjectProperty' &&
            ((p.key.type === 'Identifier' && p.key.name === key) ||
                (p.key.type === 'StringLiteral' &&
                    p.key.value === key)));
        if (!prop) {
            out[key] = staticStatus(undefined);
            continue;
        }
        const valueExpr = prop.value;
        if (!(0, can_update_sequence_props_1.isStaticValue)(valueExpr)) {
            out[key] = (0, can_update_sequence_props_1.getComputedStatus)(valueExpr, ast);
            continue;
        }
        out[key] = {
            status: 'static',
            codeValue: (0, can_update_sequence_props_1.extractStaticValue)(valueExpr),
        };
    }
    return out;
};
const computeEffectPropStatus = ({ ast, jsx, effectIndex, keys, }) => {
    const attr = findEffectsAttr(jsx);
    const elements = getEffectsArrayElements(attr);
    if (!elements) {
        return {
            canUpdate: false,
            effectIndex,
            reason: 'not-found',
        };
    }
    if (effectIndex < 0 || effectIndex >= elements.length) {
        return {
            canUpdate: false,
            effectIndex,
            reason: 'not-found',
        };
    }
    const target = elements[effectIndex];
    if (target.kind !== 'call') {
        return {
            canUpdate: false,
            effectIndex,
            reason: 'not-call-expression',
        };
    }
    const call = target.node;
    const effectImport = resolveEffectImport({
        ast,
        call,
        fallbackCallee: target.callee,
    });
    if (call.arguments.length === 0) {
        const emptyProps = {};
        for (const key of keys) {
            emptyProps[key] = staticStatus(undefined);
        }
        return {
            canUpdate: true,
            callee: effectImport.callee,
            importPath: effectImport.importPath,
            effectIndex,
            props: emptyProps,
        };
    }
    const firstArg = call.arguments[0];
    if (firstArg.type !== 'ObjectExpression') {
        return {
            canUpdate: false,
            effectIndex,
            reason: 'computed',
        };
    }
    const resolvedProps = getPropsFromObjectExpression({
        ast,
        objExpr: firstArg,
        keys,
    });
    return {
        canUpdate: true,
        effectIndex,
        callee: effectImport.callee,
        importPath: effectImport.importPath,
        props: resolvedProps,
    };
};
exports.computeEffectPropStatus = computeEffectPropStatus;
const computeEffectPropsStatusesFromContent = ({ fileContents, sequenceNodePath, effects, keysFor, }) => {
    const ast = (0, parse_ast_1.parseAst)(fileContents);
    const jsx = (0, can_update_sequence_props_1.findJsxElementAtNodePath)(ast, sequenceNodePath);
    if (!jsx) {
        return effects.map((_effect, effectIndex) => ({
            canUpdate: false,
            effectIndex,
            reason: 'not-found',
        }));
    }
    return effects.map((effect, effectIndex) => (0, exports.computeEffectPropStatus)({
        ast,
        jsx,
        effectIndex,
        keys: keysFor(effect),
    }));
};
exports.computeEffectPropsStatusesFromContent = computeEffectPropsStatusesFromContent;
const computeEffectPropsStatusesFromFile = ({ fileName, sequenceNodePath, effects, keysFor, remotionRoot, }) => {
    const { absolutePath } = (0, resolve_file_inside_project_1.resolveFileInsideProject)({
        remotionRoot,
        fileName,
        action: 'read',
    });
    const fileContents = (0, node_fs_1.readFileSync)(absolutePath, 'utf-8');
    return (0, exports.computeEffectPropsStatusesFromContent)({
        fileContents,
        sequenceNodePath,
        effects,
        keysFor,
    });
};
exports.computeEffectPropsStatusesFromFile = computeEffectPropsStatusesFromFile;
