// @remotion/studio-server/dist/preview-server/routes/composition-component-info.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.compositionComponentInfoHandler = void 0;
const resolve_composition_component_1 = require("../../helpers/resolve-composition-component");
const compositionComponentInfoHandler = async ({ input: { compositionFile, compositionId }, remotionRoot }) => {
    if (typeof compositionFile !== 'string') {
        throw new TypeError('Need to pass compositionFile');
    }
    if (typeof compositionId !== 'string') {
        throw new TypeError('Need to pass compositionId');
    }
    const location = await (0, resolve_composition_component_1.resolveCompositionComponent)({
        remotionRoot,
        compositionFile,
        compositionId,
    });
    return {
        location: {
            source: location.source,
            line: location.line,
            column: location.column,
        },
        canAddSequence: location.canAddSequence,
    };
};
exports.compositionComponentInfoHandler = compositionComponentInfoHandler;
