// @remotion/studio-server/dist/preview-server/routes/delete-effect.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.deleteEffectHandler = void 0;
const node_fs_1 = require("node:fs");
const renderer_1 = require("@remotion/renderer");
const delete_effect_1 = require("../../codemods/delete-effect");
const file_watcher_1 = require("../../file-watcher");
const resolve_file_inside_project_1 = require("../../helpers/resolve-file-inside-project");
const format_log_file_location_1 = require("../format-log-file-location");
const undo_stack_1 = require("../undo-stack");
const formatting_1 = require("./log-updates/formatting");
const log_update_1 = require("./log-updates/log-update");
const getDeletedEffectDescription = (effectLabels) => {
    if (effectLabels.length === 1) {
        return effectLabels[0];
    }
    return `${effectLabels.length} effects`;
};
const deleteEffectHandler = async ({ input: effects, remotionRoot, logLevel }) => {
    var _a;
    try {
        if (effects.length === 0) {
            throw new Error('No effects were specified for deletion');
        }
        renderer_1.RenderInternals.Log.trace({ indent: false, logLevel }, `[delete-effect] Received request to delete ${effects.length} effect target${effects.length === 1 ? '' : 's'}`);
        const itemsByFileName = new Map();
        for (const item of effects) {
            const fileItems = (_a = itemsByFileName.get(item.fileName)) !== null && _a !== void 0 ? _a : [];
            fileItems.push(item);
            itemsByFileName.set(item.fileName, fileItems);
        }
        const updates = await Promise.all([...itemsByFileName.entries()].map(async ([fileName, fileItems]) => {
            const { absolutePath, fileRelativeToRoot } = (0, resolve_file_inside_project_1.resolveFileInsideProject)({
                remotionRoot,
                fileName,
                action: 'modify',
            });
            const fileContents = (0, node_fs_1.readFileSync)(absolutePath, 'utf-8');
            const { output, formatted, effectLabels, logLines } = await (0, delete_effect_1.deleteEffects)({
                input: fileContents,
                effects: fileItems.map((item) => item.type === 'single-effect'
                    ? {
                        type: 'single-effect',
                        sequenceNodePath: item.sequenceNodePath.nodePath,
                        effectIndex: item.effectIndex,
                    }
                    : {
                        type: 'all-effects',
                        sequenceNodePath: item.sequenceNodePath.nodePath,
                    }),
            });
            return {
                absolutePath,
                fileRelativeToRoot,
                fileContents,
                output,
                formatted,
                effectLabels,
                logLine: Math.min(...logLines),
            };
        }));
        for (const update of updates) {
            const deletedEffectDescription = getDeletedEffectDescription(update.effectLabels);
            (0, undo_stack_1.pushToUndoStack)({
                filePath: update.absolutePath,
                oldContents: update.fileContents,
                newContents: null,
                logLevel,
                remotionRoot,
                logLine: update.logLine,
                description: {
                    undoMessage: `↩️  Deletion of ${deletedEffectDescription}`,
                    redoMessage: `↪️  Deletion of ${deletedEffectDescription}`,
                },
                entryType: 'delete-effect',
                suppressHmrOnFileRestore: false,
            });
            (0, undo_stack_1.suppressUndoStackInvalidation)(update.absolutePath);
            (0, file_watcher_1.writeFileAndNotifyFileWatchers)(update.absolutePath, update.output, undefined);
            const locationLabel = (0, format_log_file_location_1.formatLogFileLocation)({
                remotionRoot,
                absolutePath: update.absolutePath,
                line: update.logLine,
            });
            renderer_1.RenderInternals.Log.info({ indent: false, logLevel }, `${renderer_1.RenderInternals.chalk.blueBright(`${locationLabel}`)} ${(0, formatting_1.strikeThroughOrRemovedPrefix)(deletedEffectDescription)}`);
            if (!update.formatted) {
                (0, log_update_1.warnAboutPrettierOnce)(logLevel);
            }
            renderer_1.RenderInternals.Log.verbose({ indent: false, logLevel }, `[delete-effect] Wrote ${update.fileRelativeToRoot}${update.formatted ? ' (formatted)' : ''}`);
        }
        (0, undo_stack_1.printUndoHint)(logLevel);
        return {
            success: true,
        };
    }
    catch (err) {
        return {
            success: false,
            reason: err.message,
            stack: err.stack,
        };
    }
};
exports.deleteEffectHandler = deleteEffectHandler;
