// @remotion/studio-server/dist/preview-server/routes/delete-sequence-keyframe.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.deleteSequenceKeyframeHandler = void 0;
const node_fs_1 = require("node:fs");
const renderer_1 = require("@remotion/renderer");
const studio_shared_1 = require("@remotion/studio-shared");
const update_keyframes_1 = require("../../codemods/update-keyframes/update-keyframes");
const file_watcher_1 = require("../../file-watcher");
const resolve_file_inside_project_1 = require("../../helpers/resolve-file-inside-project");
const undo_stack_1 = require("../undo-stack");
const watch_ignore_next_change_1 = require("../watch-ignore-next-change");
const can_update_sequence_props_1 = require("./can-update-sequence-props");
const log_update_1 = require("./log-updates/log-update");
const save_props_mutex_1 = require("./save-props-mutex");
const deleteSequenceKeyframeHandler = ({ input: { fileName, nodePath, key, frame, schema, clientId }, remotionRoot, logLevel, }) => (0, save_props_mutex_1.withSavePropsLock)(async () => {
    renderer_1.RenderInternals.Log.trace({ indent: false, logLevel }, `[delete-sequence-keyframe] Received request for fileName="${fileName}" key="${key}" frame=${frame}`);
    const { absolutePath, fileRelativeToRoot } = (0, resolve_file_inside_project_1.resolveFileInsideProject)({
        remotionRoot,
        fileName,
        action: 'modify',
    });
    const fileContents = (0, node_fs_1.readFileSync)(absolutePath, 'utf-8');
    const { output, oldValueStrings, newValueStrings, formatted, logLine } = await (0, update_keyframes_1.updateSequenceKeyframes)({
        input: fileContents,
        nodePath: nodePath.nodePath,
        updates: [
            {
                key,
                operation: {
                    type: 'remove',
                    frame,
                },
            },
        ],
    });
    const oldValueString = oldValueStrings[0];
    const newValueString = newValueStrings[0];
    const undoPropChange = `${key} keyframe restored at frame ${frame}`;
    const redoPropChange = `${key} keyframe deleted at frame ${frame}`;
    (0, undo_stack_1.pushToUndoStack)({
        filePath: absolutePath,
        oldContents: fileContents,
        newContents: null,
        logLevel,
        remotionRoot,
        logLine,
        description: {
            undoMessage: `↩️  ${undoPropChange}`,
            redoMessage: `↪️  ${redoPropChange}`,
        },
        entryType: 'sequence-props',
        suppressHmrOnFileRestore: true,
    });
    (0, undo_stack_1.suppressUndoStackInvalidation)(absolutePath);
    (0, watch_ignore_next_change_1.suppressBundlerUpdateForFile)(absolutePath);
    (0, file_watcher_1.writeFileAndNotifyFileWatchers)(absolutePath, output, clientId);
    (0, log_update_1.logUpdate)({
        fileRelativeToRoot,
        line: logLine,
        key,
        oldValueString,
        newValueString,
        defaultValueString: null,
        formatted,
        logLevel,
        removedProps: [],
        addedProps: [],
    });
    (0, undo_stack_1.printUndoHint)(logLevel);
    const status = (0, can_update_sequence_props_1.computeSequencePropsStatusFromContent)({
        fileContents: output,
        keys: (0, studio_shared_1.getAllSchemaKeys)(schema),
        nodePath: nodePath.nodePath,
        effects: [],
    });
    return {
        canUpdate: true,
        props: status.props,
        results: [{ fileName, nodePath, props: status.props }],
    };
});
exports.deleteSequenceKeyframeHandler = deleteSequenceKeyframeHandler;
