// @remotion/studio-server/dist/preview-server/routes/delete-jsx-node.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.deleteJsxNodeHandler = void 0;
const node_fs_1 = require("node:fs");
const renderer_1 = require("@remotion/renderer");
const delete_jsx_node_1 = require("../../codemods/delete-jsx-node");
const file_watcher_1 = require("../../file-watcher");
const resolve_file_inside_project_1 = require("../../helpers/resolve-file-inside-project");
const format_log_file_location_1 = require("../format-log-file-location");
const undo_stack_1 = require("../undo-stack");
const log_update_1 = require("./log-updates/log-update");
const getDeletedNodeDescription = (nodeLabels) => {
    if (nodeLabels.length === 1) {
        return nodeLabels[0];
    }
    return `${nodeLabels.length} JSX nodes`;
};
const deleteJsxNodeHandler = async ({ input: { nodes }, remotionRoot, logLevel }) => {
    var _a;
    try {
        if (nodes.length === 0) {
            throw new Error('No JSX nodes were specified for deletion');
        }
        renderer_1.RenderInternals.Log.trace({ indent: false, logLevel }, `[delete-jsx-node] Received request to delete ${nodes.length} JSX node${nodes.length === 1 ? '' : 's'}`);
        const itemsByFileName = new Map();
        for (const item of nodes) {
            const fileItems = (_a = itemsByFileName.get(item.fileName)) !== null && _a !== void 0 ? _a : [];
            fileItems.push(item);
            itemsByFileName.set(item.fileName, fileItems);
        }
        const updates = await Promise.all([...itemsByFileName.entries()].map(async ([fileName, fileItems]) => {
            const { absolutePath, fileRelativeToRoot } = (0, resolve_file_inside_project_1.resolveFileInsideProject)({
                remotionRoot,
                fileName,
                action: 'modify',
            });
            const fileContents = (0, node_fs_1.readFileSync)(absolutePath, 'utf-8');
            const { output, formatted, nodeLabels, logLines } = await (0, delete_jsx_node_1.deleteJsxNodes)({
                input: fileContents,
                nodePaths: fileItems.map((item) => item.nodePath),
            });
            return {
                absolutePath,
                fileRelativeToRoot,
                fileContents,
                output,
                formatted,
                nodeLabels,
                logLine: Math.min(...logLines),
            };
        }));
        for (const update of updates) {
            const deletedNodeDescription = getDeletedNodeDescription(update.nodeLabels);
            (0, undo_stack_1.pushToUndoStack)({
                filePath: update.absolutePath,
                oldContents: update.fileContents,
                newContents: null,
                logLevel,
                remotionRoot,
                logLine: update.logLine,
                description: {
                    undoMessage: `↩️  Deletion of ${deletedNodeDescription}`,
                    redoMessage: `↪️  Deletion of ${deletedNodeDescription}`,
                },
                entryType: 'delete-jsx-node',
                suppressHmrOnFileRestore: false,
            });
            (0, undo_stack_1.suppressUndoStackInvalidation)(update.absolutePath);
            (0, file_watcher_1.writeFileAndNotifyFileWatchers)(update.absolutePath, update.output, undefined);
            const locationLabel = (0, format_log_file_location_1.formatLogFileLocation)({
                remotionRoot,
                absolutePath: update.absolutePath,
                line: update.logLine,
            });
            renderer_1.RenderInternals.Log.info({ indent: false, logLevel }, `${renderer_1.RenderInternals.chalk.blueBright(`${locationLabel}`)} Deleted ${deletedNodeDescription}`);
            if (!update.formatted) {
                (0, log_update_1.warnAboutPrettierOnce)(logLevel);
            }
            renderer_1.RenderInternals.Log.verbose({ indent: false, logLevel }, `[delete-jsx-node] Wrote ${update.fileRelativeToRoot}${update.formatted ? ' (formatted)' : ''}`);
        }
        (0, undo_stack_1.printUndoHint)(logLevel);
        return {
            success: true,
        };
    }
    catch (err) {
        return {
            success: false,
            reason: err.message,
            stack: err.stack,
        };
    }
};
exports.deleteJsxNodeHandler = deleteJsxNodeHandler;
