// @remotion/studio-server/dist/preview-server/routes/duplicate-jsx-node.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.duplicateJsxNodeHandler = void 0;
const node_fs_1 = require("node:fs");
const renderer_1 = require("@remotion/renderer");
const duplicate_jsx_node_1 = require("../../codemods/duplicate-jsx-node");
const file_watcher_1 = require("../../file-watcher");
const resolve_file_inside_project_1 = require("../../helpers/resolve-file-inside-project");
const format_log_file_location_1 = require("../format-log-file-location");
const undo_stack_1 = require("../undo-stack");
const log_update_1 = require("./log-updates/log-update");
const duplicateJsxNodeHandler = async ({ input: { fileName, nodePath }, remotionRoot, logLevel }) => {
    try {
        renderer_1.RenderInternals.Log.trace({ indent: false, logLevel }, `[duplicate-jsx-node] Received request for fileName="${fileName}"`);
        const { absolutePath, fileRelativeToRoot } = (0, resolve_file_inside_project_1.resolveFileInsideProject)({
            remotionRoot,
            fileName,
            action: 'modify',
        });
        const fileContents = (0, node_fs_1.readFileSync)(absolutePath, 'utf-8');
        const { output, formatted, nodeLabel, logLine } = await (0, duplicate_jsx_node_1.duplicateJsxNode)({
            input: fileContents,
            nodePath,
        });
        (0, undo_stack_1.pushToUndoStack)({
            filePath: absolutePath,
            oldContents: fileContents,
            newContents: null,
            logLevel,
            remotionRoot,
            logLine,
            description: {
                undoMessage: `↩️  Duplication of ${nodeLabel}`,
                redoMessage: `↪️  Duplication of ${nodeLabel}`,
            },
            entryType: 'duplicate-jsx-node',
            suppressHmrOnFileRestore: false,
        });
        (0, undo_stack_1.suppressUndoStackInvalidation)(absolutePath);
        (0, file_watcher_1.writeFileAndNotifyFileWatchers)(absolutePath, output, undefined);
        const locationLabel = (0, format_log_file_location_1.formatLogFileLocation)({
            remotionRoot,
            absolutePath,
            line: logLine,
        });
        renderer_1.RenderInternals.Log.info({ indent: false, logLevel }, `${renderer_1.RenderInternals.chalk.blueBright(`${locationLabel}`)} Duplicated ${nodeLabel}`);
        if (!formatted) {
            (0, log_update_1.warnAboutPrettierOnce)(logLevel);
        }
        renderer_1.RenderInternals.Log.verbose({ indent: false, logLevel }, `[duplicate-jsx-node] Wrote ${fileRelativeToRoot}${formatted ? ' (formatted)' : ''}`);
        (0, undo_stack_1.printUndoHint)(logLevel);
        return {
            success: true,
        };
    }
    catch (err) {
        return {
            success: false,
            reason: err.message,
            stack: err.stack,
        };
    }
};
exports.duplicateJsxNodeHandler = duplicateJsxNodeHandler;
