// @remotion/studio-server/dist/preview-server/routes/insert-jsx-element.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.insertJsxElementHandler = void 0;
const renderer_1 = require("@remotion/renderer");
const studio_shared_1 = require("@remotion/studio-shared");
const file_watcher_1 = require("../../file-watcher");
const resolve_composition_component_1 = require("../../helpers/resolve-composition-component");
const format_log_file_location_1 = require("../format-log-file-location");
const undo_stack_1 = require("../undo-stack");
const log_update_1 = require("./log-updates/log-update");
const save_props_mutex_1 = require("./save-props-mutex");
const validateDimension = (name, value) => {
    if (!Number.isFinite(value) || value < 1) {
        throw new Error(`${name} must be a positive number`);
    }
};
const validatePosition = (position) => {
    if (position === null) {
        return;
    }
    if (!Number.isFinite(position.x) || !Number.isFinite(position.y)) {
        throw new Error('Position must be finite');
    }
};
const validateElement = (element) => {
    validatePosition(element.position);
    if (element.type === 'solid') {
        validateDimension('width', element.width);
        validateDimension('height', element.height);
        return;
    }
    if (element.type === 'asset') {
        if (element.srcType === 'remote') {
            if (!(0, studio_shared_1.isUrl)(element.src)) {
                throw new Error('Remote asset source must be a URL');
            }
        }
        else if (!element.src || element.src.includes('\\')) {
            throw new Error('Asset path must be a static file path');
        }
        if (element.dimensions) {
            validateDimension('width', element.dimensions.width);
            validateDimension('height', element.dimensions.height);
        }
        return;
    }
    if (element.type === 'component') {
        if (!(0, studio_shared_1.isComponentIdentifier)(element.componentName)) {
            throw new Error('Unsupported component name');
        }
        if (!(0, studio_shared_1.isComponentIdentifier)(element.importName)) {
            throw new Error('Unsupported component import name');
        }
        if (!(0, studio_shared_1.isComponentImportPath)(element.importPath)) {
            throw new Error('Unsupported component import path');
        }
        if (!(0, studio_shared_1.areComponentProps)(element.props)) {
            throw new Error('Unsupported component props');
        }
        return;
    }
    throw new Error('Unsupported element type');
};
const getElementLabel = (element) => {
    if (element.type === 'solid') {
        return '<Solid>';
    }
    if (element.type === 'asset') {
        if (element.assetType === 'image') {
            return '<Img>';
        }
        if (element.assetType === 'video') {
            return '<Video>';
        }
        if (element.assetType === 'gif') {
            return '<Gif>';
        }
        if (element.assetType === 'animated-image') {
            return '<AnimatedImage>';
        }
        if (element.assetType === 'audio') {
            return '<Audio>';
        }
    }
    if (element.type === 'component') {
        return `<${element.componentName}>`;
    }
    throw new Error('Unsupported element type');
};
const insertJsxElementHandler = ({ input: { compositionFile, compositionId, element }, remotionRoot, logLevel, }) => (0, save_props_mutex_1.withSavePropsLock)(async () => {
    try {
        validateElement(element);
        const elementLabel = getElementLabel(element);
        renderer_1.RenderInternals.Log.trace({ indent: false, logLevel }, `[insert-jsx-element] Received request for compositionFile="${compositionFile}" compositionId="${compositionId}" element="${element.type}"`);
        const { fileName, source, oldContents, output, formatted, logLine } = await (0, resolve_composition_component_1.insertJsxElementIntoComposition)({
            remotionRoot,
            compositionFile,
            compositionId,
            element,
            prettierConfigOverride: null,
        });
        (0, undo_stack_1.pushToUndoStack)({
            filePath: fileName,
            oldContents,
            newContents: output,
            logLevel,
            remotionRoot,
            logLine,
            description: {
                undoMessage: `↩️  Added ${elementLabel}`,
                redoMessage: `↪️  Added ${elementLabel}`,
            },
            entryType: 'insert-jsx-element',
            suppressHmrOnFileRestore: false,
        });
        (0, undo_stack_1.suppressUndoStackInvalidation)(fileName);
        (0, file_watcher_1.writeFileAndNotifyFileWatchers)(fileName, output, undefined);
        const locationLabel = (0, format_log_file_location_1.formatLogFileLocation)({
            remotionRoot,
            absolutePath: fileName,
            line: logLine,
        });
        renderer_1.RenderInternals.Log.info({ indent: false, logLevel }, `${renderer_1.RenderInternals.chalk.blueBright(`${locationLabel}`)} Added ${elementLabel}`);
        if (!formatted) {
            (0, log_update_1.warnAboutPrettierOnce)(logLevel);
        }
        renderer_1.RenderInternals.Log.verbose({ indent: false, logLevel }, `[insert-jsx-element] Wrote ${source}${formatted ? ' (formatted)' : ''}`);
        (0, undo_stack_1.printUndoHint)(logLevel);
        return {
            success: true,
        };
    }
    catch (err) {
        return {
            success: false,
            reason: err.message,
            stack: err.stack,
        };
    }
});
exports.insertJsxElementHandler = insertJsxElementHandler;
