// @remotion/studio-server/dist/preview-server/routes/install-dependency.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.handleInstallPackage = void 0;
const node_child_process_1 = require("node:child_process");
const renderer_1 = require("@remotion/renderer");
const studio_shared_1 = require("@remotion/studio-shared");
const version_1 = require("remotion/version");
const install_command_1 = require("../../helpers/install-command");
const get_package_manager_1 = require("../get-package-manager");
const extraPackageNames = studio_shared_1.extraPackages.map((pkg) => pkg.name);
const isExtraPackage = (packageName) => {
    return extraPackageNames.includes(packageName);
};
const getExtraPackageVersion = (packageName) => {
    const pkg = studio_shared_1.extraPackages.find((p) => p.name === packageName);
    return pkg ? pkg.version : null;
};
const handleInstallPackage = async ({ logLevel, remotionRoot, input: { packageNames } }) => {
    for (const packageName of packageNames) {
        if (!packageName.startsWith('@remotion/') && !isExtraPackage(packageName)) {
            return Promise.reject(new Error(`Package ${packageName} is not allowed to be installed.`));
        }
    }
    const manager = (0, get_package_manager_1.getPackageManager)({
        remotionRoot,
        packageManager: undefined,
        dirUp: 0,
        logLevel,
    });
    if (manager === 'unknown') {
        throw new Error(`No lockfile was found in your project (one of ${get_package_manager_1.lockFilePaths
            .map((p) => p.path)
            .join(', ')}). Install dependencies using your favorite manager!`);
    }
    // Build packages with appropriate versions
    const packagesWithVersions = packageNames.map((pkg) => {
        const extraVersion = getExtraPackageVersion(pkg);
        if (extraVersion) {
            return `${pkg}@${extraVersion}`;
        }
        return `${pkg}@${version_1.VERSION}`;
    });
    const command = (0, install_command_1.getInstallCommand)({
        manager: manager.manager,
        packages: packagesWithVersions,
        version: '',
        additionalArgs: [],
    });
    renderer_1.RenderInternals.Log.info({ indent: false, logLevel }, renderer_1.RenderInternals.chalk.gray(`╭─  ${manager.manager} ${command.join(' ')}`));
    const time = Date.now();
    try {
        await new Promise((resolve, reject) => {
            const cmd = (0, node_child_process_1.spawn)(manager.manager, command, {});
            cmd.stdout.on('data', (d) => {
                const splitted = d.toString().trim().split('\n');
                splitted.forEach((line) => {
                    renderer_1.RenderInternals.Log.info({ indent: true, logLevel }, line);
                });
            });
            cmd.stdout.on('end', () => {
                resolve();
            });
            cmd.on('close', (code, signal) => {
                if (code === 0) {
                    resolve();
                }
                else {
                    reject(new Error(`Command exited with code ${code} and signal ${signal}`));
                }
            });
        });
        const timeEnd = Date.now();
        renderer_1.RenderInternals.Log.info({ indent: false, logLevel }, renderer_1.RenderInternals.chalk.gray('╰─ '), `Done in ${timeEnd - time}ms`);
        return Promise.resolve({});
    }
    catch (err) {
        const timeEnd = Date.now();
        renderer_1.RenderInternals.Log.info({ indent: false, logLevel }, renderer_1.RenderInternals.chalk.gray('╰─ '), renderer_1.RenderInternals.chalk.red(`Errored in ${timeEnd - time}ms`));
        return Promise.reject(err);
    }
};
exports.handleInstallPackage = handleInstallPackage;
