// @remotion/studio-server/dist/preview-server/routes/open-in-editor.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.openInEditorHandler = exports.getEditorName = void 0;
const node_path_1 = __importDefault(require("node:path"));
const open_in_editor_1 = require("../../helpers/open-in-editor");
const editorGuess = (0, open_in_editor_1.guessEditor)();
const getEditorName = async () => {
    const [edit] = await editorGuess;
    return (0, open_in_editor_1.getDisplayNameForEditor)(edit ? edit.command : null);
};
exports.getEditorName = getEditorName;
const openInEditorHandler = async ({ input, remotionRoot, logLevel }) => {
    try {
        if (!('stack' in input)) {
            throw new TypeError('Need to pass stack');
        }
        const { stack } = input;
        const guess = await editorGuess;
        const didOpen = await (0, open_in_editor_1.launchEditor)({
            colNumber: stack.originalColumnNumber,
            editor: guess[0],
            fileName: node_path_1.default.resolve(remotionRoot, stack.originalFileName),
            lineNumber: stack.originalLineNumber,
            vsCodeNewWindow: false,
            logLevel,
        });
        return {
            success: didOpen,
        };
    }
    catch (_a) {
        return {
            success: false,
        };
    }
};
exports.openInEditorHandler = openInEditorHandler;
