// @remotion/studio-server/dist/preview-server/routes/reorder-effect.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.reorderEffectHandler = void 0;
const node_fs_1 = require("node:fs");
const renderer_1 = require("@remotion/renderer");
const reorder_effect_1 = require("../../codemods/reorder-effect");
const file_watcher_1 = require("../../file-watcher");
const resolve_file_inside_project_1 = require("../../helpers/resolve-file-inside-project");
const format_log_file_location_1 = require("../format-log-file-location");
const undo_stack_1 = require("../undo-stack");
const formatting_1 = require("./log-updates/formatting");
const log_update_1 = require("./log-updates/log-update");
const reorderEffectHandler = async ({ input: { fileName, sequenceNodePath, fromIndex, toIndex, clientId }, remotionRoot, logLevel, }) => {
    try {
        renderer_1.RenderInternals.Log.trace({ indent: false, logLevel }, `[reorder-effect] Received request for fileName="${fileName}" fromIndex=${fromIndex} toIndex=${toIndex}`);
        const { absolutePath, fileRelativeToRoot } = (0, resolve_file_inside_project_1.resolveFileInsideProject)({
            remotionRoot,
            fileName,
            action: 'modify',
        });
        const fileContents = (0, node_fs_1.readFileSync)(absolutePath, 'utf-8');
        const { output, formatted, effectLabel, logLine } = await (0, reorder_effect_1.reorderEffect)({
            input: fileContents,
            sequenceNodePath: sequenceNodePath.nodePath,
            fromIndex,
            toIndex,
        });
        (0, undo_stack_1.pushToUndoStack)({
            filePath: absolutePath,
            oldContents: fileContents,
            newContents: null,
            logLevel,
            remotionRoot,
            logLine,
            description: {
                undoMessage: `↩️  Reordering of ${effectLabel}`,
                redoMessage: `↪️  Reordering of ${effectLabel}`,
            },
            entryType: 'reorder-effect',
            suppressHmrOnFileRestore: false,
        });
        (0, undo_stack_1.suppressUndoStackInvalidation)(absolutePath);
        (0, file_watcher_1.writeFileAndNotifyFileWatchers)(absolutePath, output, clientId);
        const locationLabel = (0, format_log_file_location_1.formatLogFileLocation)({
            remotionRoot,
            absolutePath,
            line: logLine,
        });
        renderer_1.RenderInternals.Log.info({ indent: false, logLevel }, `${renderer_1.RenderInternals.chalk.blueBright(`${locationLabel}`)} Reordered ${(0, formatting_1.attrName)(effectLabel)}`);
        if (!formatted) {
            (0, log_update_1.warnAboutPrettierOnce)(logLevel);
        }
        renderer_1.RenderInternals.Log.verbose({ indent: false, logLevel }, `[reorder-effect] Wrote ${fileRelativeToRoot}${formatted ? ' (formatted)' : ''}`);
        (0, undo_stack_1.printUndoHint)(logLevel);
        return {
            success: true,
        };
    }
    catch (err) {
        return {
            success: false,
            reason: err.message,
            stack: err.stack,
        };
    }
};
exports.reorderEffectHandler = reorderEffectHandler;
