// @remotion/studio-server/dist/preview-server/routes/save-sequence-props.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.saveSequencePropsHandler = exports.shouldSuppressHmrForSequencePropEdits = exports.convertSequencePropEditToCodemodChange = void 0;
const node_fs_1 = require("node:fs");
const renderer_1 = require("@remotion/renderer");
const studio_shared_1 = require("@remotion/studio-shared");
const update_sequence_props_1 = require("../../codemods/update-sequence-props/update-sequence-props");
const file_watcher_1 = require("../../file-watcher");
const resolve_file_inside_project_1 = require("../../helpers/resolve-file-inside-project");
const undo_stack_1 = require("../undo-stack");
const watch_ignore_next_change_1 = require("../watch-ignore-next-change");
const can_update_sequence_props_1 = require("./can-update-sequence-props");
const log_update_1 = require("./log-updates/log-update");
const save_props_mutex_1 = require("./save-props-mutex");
const convertSequencePropEditToCodemodChange = (edit) => {
    return {
        nodePath: edit.nodePath.nodePath,
        updates: [
            {
                key: edit.key,
                value: edit.value,
                defaultValue: edit.defaultValue,
            },
        ],
        schema: edit.schema,
    };
};
exports.convertSequencePropEditToCodemodChange = convertSequencePropEditToCodemodChange;
const shouldSuppressHmrForSequencePropEdits = (edits) => {
    return edits.every((edit) => edit.key !== 'showInTimeline');
};
exports.shouldSuppressHmrForSequencePropEdits = shouldSuppressHmrForSequencePropEdits;
const saveSequencePropsHandler = ({ input: { edits, clientId, undoLabel, redoLabel }, remotionRoot, logLevel, }) => (0, save_props_mutex_1.withSavePropsLock)(async () => {
    var _a;
    if (edits.length === 0) {
        throw new Error('No sequence prop edits to save');
    }
    renderer_1.RenderInternals.Log.trace({ indent: false, logLevel }, `[save-sequence-props] Received request with ${edits.length} edit(s)`);
    const editGroups = new Map();
    for (const [index, edit] of edits.entries()) {
        const parsedValue = JSON.parse(edit.value);
        const parsedDefaultValue = edit.defaultValue !== null ? JSON.parse(edit.defaultValue) : null;
        const { absolutePath, fileRelativeToRoot } = (0, resolve_file_inside_project_1.resolveFileInsideProject)({
            remotionRoot,
            fileName: edit.fileName,
            action: 'modify',
        });
        const group = (_a = editGroups.get(absolutePath)) !== null && _a !== void 0 ? _a : {
            fileRelativeToRoot,
            edits: [],
        };
        group.edits.push({
            index,
            fileName: edit.fileName,
            nodePath: edit.nodePath,
            key: edit.key,
            value: parsedValue,
            valueString: JSON.stringify(parsedValue),
            defaultValue: parsedDefaultValue,
            defaultValueString: parsedDefaultValue !== null
                ? JSON.stringify(parsedDefaultValue)
                : null,
            schema: edit.schema,
        });
        editGroups.set(absolutePath, group);
    }
    const snapshots = [];
    const outputByPath = new Map();
    const resultByIndex = new Map();
    for (const [absolutePath, group] of editGroups) {
        const fileContents = (0, node_fs_1.readFileSync)(absolutePath, 'utf-8');
        const { output, formatted, results: updateResults, } = await (0, update_sequence_props_1.updateMultipleSequenceProps)({
            input: fileContents,
            changes: group.edits.map(exports.convertSequencePropEditToCodemodChange),
            prettierConfigOverride: null,
        });
        const [{ logLine: firstLogLine }] = updateResults;
        outputByPath.set(absolutePath, output);
        snapshots.push({
            filePath: absolutePath,
            oldContents: fileContents,
            newContents: output,
            logLine: firstLogLine,
        });
        for (const [resultIndex, result] of updateResults.entries()) {
            const edit = group.edits[resultIndex];
            resultByIndex.set(edit.index, {
                oldValueString: result.oldValueStrings[0],
                logLine: result.logLine,
                removedProps: result.removedProps,
                formatted,
            });
        }
    }
    const undoMessage = `↩️  ${undoLabel}`;
    const redoMessage = `↪️  ${redoLabel}`;
    const suppressHmr = (0, exports.shouldSuppressHmrForSequencePropEdits)(edits);
    (0, undo_stack_1.pushTransactionToUndoStack)({
        snapshots,
        logLevel,
        remotionRoot,
        description: { undoMessage, redoMessage },
        entryType: 'sequence-props',
        suppressHmrOnFileRestore: suppressHmr,
    });
    for (const [absolutePath, output] of outputByPath) {
        (0, undo_stack_1.suppressUndoStackInvalidation)(absolutePath);
        if (suppressHmr) {
            (0, watch_ignore_next_change_1.suppressBundlerUpdateForFile)(absolutePath);
        }
        (0, file_watcher_1.writeFileAndNotifyFileWatchers)(absolutePath, output, clientId);
    }
    for (const { edits: groupEdits, fileRelativeToRoot } of editGroups.values()) {
        for (const edit of groupEdits) {
            const result = resultByIndex.get(edit.index);
            if (!result) {
                throw new Error('Could not compute sequence prop edit result');
            }
            (0, log_update_1.logUpdate)({
                fileRelativeToRoot,
                line: result.logLine,
                key: edit.key,
                oldValueString: result.oldValueString,
                newValueString: edit.valueString,
                defaultValueString: edit.defaultValueString,
                formatted: result.formatted,
                logLevel,
                removedProps: result.removedProps,
                addedProps: [],
            });
        }
    }
    (0, undo_stack_1.printUndoHint)(logLevel);
    const results = edits.map((edit) => {
        const { absolutePath } = (0, resolve_file_inside_project_1.resolveFileInsideProject)({
            remotionRoot,
            fileName: edit.fileName,
            action: 'modify',
        });
        const output = outputByPath.get(absolutePath);
        if (!output) {
            throw new Error('Could not compute sequence prop edit status');
        }
        const newStatus = (0, can_update_sequence_props_1.computeSequencePropsStatusFromContent)({
            fileContents: output,
            keys: (0, studio_shared_1.getAllSchemaKeys)(edit.schema),
            nodePath: edit.nodePath.nodePath,
            componentIdentity: null,
            effects: [],
        });
        return {
            fileName: edit.fileName,
            nodePath: edit.nodePath,
            props: newStatus.props,
        };
    });
    return {
        canUpdate: true,
        props: results[0].props,
        results,
    };
});
exports.saveSequencePropsHandler = saveSequencePropsHandler;
