// @remotion/studio-server/dist/preview-server/routes/save-effect-props.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.saveEffectPropsHandler = void 0;
const node_fs_1 = require("node:fs");
const renderer_1 = require("@remotion/renderer");
const studio_shared_1 = require("@remotion/studio-shared");
const parse_ast_1 = require("../../codemods/parse-ast");
const update_effect_props_1 = require("../../codemods/update-effect-props/update-effect-props");
const file_watcher_1 = require("../../file-watcher");
const resolve_file_inside_project_1 = require("../../helpers/resolve-file-inside-project");
const undo_stack_1 = require("../undo-stack");
const watch_ignore_next_change_1 = require("../watch-ignore-next-change");
const can_update_effect_props_1 = require("./can-update-effect-props");
const can_update_sequence_props_1 = require("./can-update-sequence-props");
const format_effect_prop_change_1 = require("./log-updates/format-effect-prop-change");
const log_effect_update_1 = require("./log-updates/log-effect-update");
const log_update_1 = require("./log-updates/log-update");
const save_props_mutex_1 = require("./save-props-mutex");
const saveEffectPropsHandler = ({ input, remotionRoot, logLevel }) => (0, save_props_mutex_1.withSavePropsLock)(async () => {
    const { fileName, sequenceNodePath, effectIndex, key, defaultValue, schema, clientId, } = input;
    renderer_1.RenderInternals.Log.trace({ indent: false, logLevel }, `[save-effect-props] Received request for fileName="${fileName}" effectIndex=${effectIndex} key="${key}"`);
    const { absolutePath, fileRelativeToRoot } = (0, resolve_file_inside_project_1.resolveFileInsideProject)({
        remotionRoot,
        fileName,
        action: 'modify',
    });
    const fileContents = (0, node_fs_1.readFileSync)(absolutePath, 'utf-8');
    const parsedDefault = defaultValue !== null ? JSON.parse(defaultValue) : null;
    const update = (() => {
        switch (input.type) {
            case 'value':
                return {
                    key,
                    value: JSON.parse(input.value),
                    defaultValue: parsedDefault,
                };
            case 'effect-param':
                return {
                    key,
                    effectParam: input.effectParam,
                    defaultValue: parsedDefault,
                };
            default:
                throw new Error(`Unsupported save effect props request type: ${input.type}`);
        }
    })();
    const { output, oldValueString, newValueString, formatted, logLine, effectCallee, removedProps, } = await (0, update_effect_props_1.updateEffectProps)({
        input: fileContents,
        sequenceNodePath: sequenceNodePath.nodePath,
        effectIndex,
        update,
        schema,
    });
    const defaultValueString = parsedDefault !== null ? JSON.stringify(parsedDefault) : null;
    const normalizedOld = (0, log_update_1.normalizeQuotes)(oldValueString);
    const normalizedNew = (0, log_update_1.normalizeQuotes)(newValueString);
    const normalizedDefault = defaultValueString !== null ? (0, log_update_1.normalizeQuotes)(defaultValueString) : null;
    const normalizedRemovedProps = removedProps.map((prop) => ({
        ...prop,
        valueString: (0, log_update_1.normalizeQuotes)(prop.valueString),
    }));
    const undoPropChange = (0, format_effect_prop_change_1.formatEffectPropChange)({
        effectName: effectCallee,
        key,
        oldValueString: normalizedNew,
        newValueString: normalizedOld,
        defaultValueString: normalizedDefault,
        removedProps: [],
        addedProps: normalizedRemovedProps,
    });
    const redoPropChange = (0, format_effect_prop_change_1.formatEffectPropChange)({
        effectName: effectCallee,
        key,
        oldValueString: normalizedOld,
        newValueString: normalizedNew,
        defaultValueString: normalizedDefault,
        removedProps: normalizedRemovedProps,
        addedProps: [],
    });
    (0, undo_stack_1.pushToUndoStack)({
        filePath: absolutePath,
        oldContents: fileContents,
        newContents: null,
        logLevel,
        remotionRoot,
        logLine,
        description: {
            undoMessage: `↩️  ${undoPropChange}`,
            redoMessage: `↪️  ${redoPropChange}`,
        },
        entryType: 'effect-props',
        suppressHmrOnFileRestore: true,
    });
    (0, undo_stack_1.suppressUndoStackInvalidation)(absolutePath);
    (0, watch_ignore_next_change_1.suppressBundlerUpdateForFile)(absolutePath);
    (0, file_watcher_1.writeFileAndNotifyFileWatchers)(absolutePath, output, clientId);
    (0, log_effect_update_1.logEffectUpdate)({
        fileRelativeToRoot,
        line: logLine,
        effectName: effectCallee,
        propKey: key,
        oldValueString,
        newValueString,
        defaultValueString,
        formatted,
        logLevel,
        removedProps,
        addedProps: [],
    });
    (0, undo_stack_1.printUndoHint)(logLevel);
    const ast = (0, parse_ast_1.parseAst)((0, node_fs_1.readFileSync)(absolutePath, 'utf-8'));
    const jsx = (0, can_update_sequence_props_1.findJsxElementAtNodePath)(ast, sequenceNodePath.nodePath);
    if (!jsx) {
        return {
            canUpdate: false,
            effectIndex,
            reason: 'not-found',
        };
    }
    return (0, can_update_effect_props_1.computeEffectPropStatus)({
        ast,
        jsx,
        effectIndex,
        keys: (0, studio_shared_1.getAllSchemaKeys)(schema),
    });
});
exports.saveEffectPropsHandler = saveEffectPropsHandler;
