// @remotion/studio-server/dist/preview-server/routes/update-default-props.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.updateDefaultPropsHandler = void 0;
const node_fs_1 = require("node:fs");
const renderer_1 = require("@remotion/renderer");
const update_default_props_1 = require("../../codemods/update-default-props");
const file_watcher_1 = require("../../file-watcher");
const format_log_file_location_1 = require("../format-log-file-location");
const project_info_1 = require("../project-info");
const undo_stack_1 = require("../undo-stack");
const watch_ignore_next_change_1 = require("../watch-ignore-next-change");
const can_update_default_props_1 = require("./can-update-default-props");
const log_update_1 = require("./log-updates/log-update");
const updateDefaultPropsHandler = async ({ input: { compositionId, defaultProps, enumPaths }, remotionRoot, entryPoint, logLevel, }) => {
    try {
        renderer_1.RenderInternals.Log.trace({ indent: false, logLevel }, `[update-default-props] Received request for compositionId="${compositionId}"`);
        const projectInfo = await (0, project_info_1.getProjectInfo)(remotionRoot, entryPoint);
        if (!projectInfo.rootFile) {
            throw new Error('Cannot find root file in project');
        }
        (0, can_update_default_props_1.checkIfTypeScriptFile)(projectInfo.rootFile);
        const fileContents = (0, node_fs_1.readFileSync)(projectInfo.rootFile, 'utf-8');
        const logLine = (0, update_default_props_1.getCompositionDefaultPropsLine)({
            input: fileContents,
            compositionId,
        });
        const { output, formatted } = await (0, update_default_props_1.updateDefaultProps)({
            compositionId,
            input: fileContents,
            newDefaultProps: JSON.parse(defaultProps),
            enumPaths,
        });
        (0, undo_stack_1.pushToUndoStack)({
            filePath: projectInfo.rootFile,
            oldContents: fileContents,
            newContents: null,
            logLevel,
            remotionRoot,
            logLine,
            description: {
                undoMessage: `↩️  default props update for "${compositionId}"`,
                redoMessage: `↪️  default props update for "${compositionId}"`,
            },
            entryType: 'default-props',
            suppressHmrOnFileRestore: true,
        });
        (0, undo_stack_1.suppressUndoStackInvalidation)(projectInfo.rootFile);
        (0, watch_ignore_next_change_1.suppressBundlerUpdateForFile)(projectInfo.rootFile);
        (0, file_watcher_1.writeFileAndNotifyFileWatchers)(projectInfo.rootFile, output, undefined);
        const locationLabel = (0, format_log_file_location_1.formatLogFileLocation)({
            remotionRoot,
            absolutePath: projectInfo.rootFile,
            line: logLine,
        });
        renderer_1.RenderInternals.Log.info({ indent: false, logLevel }, `${renderer_1.RenderInternals.chalk.blueBright(`${locationLabel}`)} Updated default props for "${compositionId}"`);
        if (!formatted) {
            (0, log_update_1.warnAboutPrettierOnce)(logLevel);
        }
        (0, undo_stack_1.printUndoHint)(logLevel);
        return {
            success: true,
        };
    }
    catch (err) {
        return {
            success: false,
            reason: err.message,
            stack: err.stack,
        };
    }
};
exports.updateDefaultPropsHandler = updateDefaultPropsHandler;
