// @remotion/studio-server/dist/preview-server/routes/update-sequence-keyframe-settings.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.updateSequenceKeyframeSettingsHandler = void 0;
const node_fs_1 = require("node:fs");
const renderer_1 = require("@remotion/renderer");
const studio_shared_1 = require("@remotion/studio-shared");
const update_keyframes_1 = require("../../codemods/update-keyframes/update-keyframes");
const file_watcher_1 = require("../../file-watcher");
const resolve_file_inside_project_1 = require("../../helpers/resolve-file-inside-project");
const undo_stack_1 = require("../undo-stack");
const watch_ignore_next_change_1 = require("../watch-ignore-next-change");
const can_update_sequence_props_1 = require("./can-update-sequence-props");
const log_update_1 = require("./log-updates/log-update");
const save_props_mutex_1 = require("./save-props-mutex");
const updateSequenceKeyframeSettingsHandler = ({ input: { fileName, nodePath, key, settings, schema, clientId }, remotionRoot, logLevel, }) => (0, save_props_mutex_1.withSavePropsLock)(async () => {
    renderer_1.RenderInternals.Log.trace({ indent: false, logLevel }, `[update-sequence-keyframe-settings] Received request for fileName="${fileName}" key="${key}"`);
    const { absolutePath, fileRelativeToRoot } = (0, resolve_file_inside_project_1.resolveFileInsideProject)({
        remotionRoot,
        fileName,
        action: 'modify',
    });
    const fileContents = (0, node_fs_1.readFileSync)(absolutePath, 'utf-8');
    const { output, oldValueStrings, newValueStrings, formatted, logLine, updatedNodePath, } = await (0, update_keyframes_1.updateSequenceKeyframes)({
        input: fileContents,
        nodePath: nodePath.nodePath,
        schema,
        updates: [
            {
                key,
                operation: settings.type === 'settings'
                    ? {
                        type: 'settings',
                        clamping: settings.clamping,
                        posterize: settings.posterize,
                    }
                    : {
                        type: 'easing',
                        segmentIndex: settings.segmentIndex,
                        easing: settings.easing,
                    },
            },
        ],
    });
    const undoPropChange = `${key} keyframe settings reverted`;
    const redoPropChange = `${key} keyframe settings updated`;
    (0, undo_stack_1.pushToUndoStack)({
        filePath: absolutePath,
        oldContents: fileContents,
        newContents: null,
        logLevel,
        remotionRoot,
        logLine,
        description: {
            undoMessage: `↩️  ${undoPropChange}`,
            redoMessage: `↪️  ${redoPropChange}`,
        },
        entryType: 'sequence-props',
        suppressHmrOnFileRestore: true,
    });
    (0, undo_stack_1.suppressUndoStackInvalidation)(absolutePath);
    (0, watch_ignore_next_change_1.suppressBundlerUpdateForFile)(absolutePath);
    (0, file_watcher_1.writeFileAndNotifyFileWatchers)(absolutePath, output, clientId);
    (0, log_update_1.logUpdate)({
        fileRelativeToRoot,
        line: logLine,
        key,
        oldValueString: oldValueStrings[0],
        newValueString: newValueStrings[0],
        defaultValueString: null,
        formatted,
        logLevel,
        removedProps: [],
        addedProps: [],
    });
    (0, undo_stack_1.printUndoHint)(logLevel);
    const status = (0, can_update_sequence_props_1.computeSequencePropsStatusFromContent)({
        fileContents: output,
        keys: (0, studio_shared_1.getAllSchemaKeys)(schema),
        nodePath: updatedNodePath,
        componentIdentity: null,
        effects: [],
    });
    const updatedSubscriptionKey = { ...nodePath, nodePath: updatedNodePath };
    return {
        canUpdate: true,
        props: status.props,
        results: [
            { fileName, nodePath: updatedSubscriptionKey, props: status.props },
        ],
    };
});
exports.updateSequenceKeyframeSettingsHandler = updateSequenceKeyframeSettingsHandler;
